<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 3 Using R &amp; Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 3 Using R &amp; Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Using R &amp; Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-02-25">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="learning-r-and-more.html">
<link rel="next" href="data-infrastructure.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and More</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting Started with R</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#nice-companions-for-r"><i class="fa fa-check"></i><b>2.2</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#shellbash"><i class="fa fa-check"></i><b>2.2.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="2.2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#git"><i class="fa fa-check"></i><b>2.2.2</b> Git</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#going-beyond-the-basics"><i class="fa fa-check"></i><b>2.3</b> Going Beyond the Basics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>2.3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#other-languages"><i class="fa fa-check"></i><b>2.3.2</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html"><i class="fa fa-check"></i><b>3</b> Using R &amp; Bioconductor</a><ul>
<li class="chapter" data-level="3.1" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#a-brief-overview"><i class="fa fa-check"></i><b>3.1</b> A Brief Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#prerequisites"><i class="fa fa-check"></i><b>3.2</b> Prerequisites</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#installing-r"><i class="fa fa-check"></i><b>3.2.1</b> Installing R</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>3.2.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>3.3</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="3.4" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#bioconductor-documentation"><i class="fa fa-check"></i><b>3.4</b> Bioconductor Documentation</a></li>
<li class="chapter" data-level="3.5" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#becoming-a-bioconductor-expert"><i class="fa fa-check"></i><b>3.5</b> Becoming a Bioconductor Expert</a></li>
<li class="chapter" data-level="3.6" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#conventions-used-in-this-book"><i class="fa fa-check"></i><b>3.6</b> Conventions used in this book</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-object"><i class="fa fa-check"></i><b>4.1</b> The <em>SingleCellExperiment</em> Object</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-sce"><i class="fa fa-check"></i><b>4.2</b> Working with <em>SCE</em></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-r-and-bioconductor" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Using R &amp; Bioconductor</h1>
<div id="a-brief-overview" class="section level2">
<h2><span class="header-section-number">3.1</span> A Brief Overview</h2>
<p><a href="http://www.r-project.org/">R</a> is a high-level programming language that was initially designed for statistical applications. While there is much to be said <a href="https://www.r-project.org/about.html">about R</a> as a programming language, one of the key advantages of using R is that it is highly extensible through <em>packages</em>. Packages are collections of functions, data, and documentation that extends the capabilities of base R. The ease of development and distribution of packages for R has made it a rich environment for many fields of study and application.</p>
<p>One of the primary ways in which packages are distributed is through centralized repositories. The first R repository a user typically runs into is the <a href="https://cran.r-project.org/mirrors.html">Comprehensive R Archive Network</a> (CRAN), which hosts over 13,000 packages to date, and is home to many of the most popular R packages.</p>
<p>Similar to CRAN, <a href="https://bioconductor.org">Bioconductor</a> is a repository of R packages as well. However, whereas CRAN is a general purpose repository, Bioconductor focuses on software tailored for genomic analysis. Furthermore, Bioconductor has stricter requirements for a package to be accepted into the repository. Of particular interest to us is the inclusion of <a href="#learning-more">high quality documentation</a> and the use of common <a href="data-infrastructure.html#data-infrastructure">data infrastructure</a> to promote package interoperability.</p>
</div>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">3.2</span> Prerequisites</h2>
<div id="installing-r" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Installing R</h3>
<p>Prior to getting started with this book, some prior programming experience with R is helpful. Check out the <a href="learning-r-and-more.html#learning-r-and-more"><em>Learning R and More</em></a> chapter for a list of resources to get started with R and other useful tools for bioinformatic analysis.</p>
<p>To follow along with the analysis workflows in this book on your personal computer, it is first necessary to install the <a href="http://www.r-project.org/">R</a> programming language. Additionally, we recommend a graphical user interface such as <a href="http://www.rstudio.com/download">RStudio</a> for programming in R and visualization. For more details, please see the online book <a href="https://r4ds.had.co.nz/introduction.html#prerequisites"><em>R for Data Science</em> prerequisites</a> section.</p>
<p>A special note for Mac OS users: we highly recommend using the <a href="https://brew.sh/">Homebrew</a> package manager for managing your R installation. Follow the website directions to install homebrew, and install R via the commandline with <code>brew install R</code>, and it will automatically configure your installation for you. Upgrading to new R versions can also be done by running <code>brew upgrade</code>.</p>
</div>
<div id="installing-r-bioconductor-packages" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Installing R &amp; Bioconductor Packages</h3>
<p>After installing R, the next step is to install R packages. In the R console, you can install packages from CRAN via the <code>install.packages()</code> function. In order to <a href="https://www.bioconductor.org/install/">install Bioconductor packages</a>, we will first need the <em>BiocManager</em> package which is hosted on CRAN. This can be done by running:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</code></pre>
<p>The <em>BiocManager</em> package makes it easy to install packages from the Bioconductor repository. For example, to install the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><em>SingleCellExperiment</em></a> package, we run:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## the command below is a one-line shortcut for:</span>
<span class="co">## library(BiocManager)</span>
<span class="co">## install(&quot;SingleCellExperiment&quot;)</span>
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SingleCellExperiment&quot;</span>)</code></pre>
<p>Throughout the book, we can load packages via the <code>library()</code> function, which by convention usually comes at the top of scripts to alert readers as to what packages are required. For example, to load the <em>SingleCellExperiment</em> package, we run:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)</code></pre>
<p>Many packages will be referenced throughout the book within the workflows, and similar to the above, can be installed using the <code>BiocManager::install()</code> function.</p>
</div>
</div>
<div id="getting-help-in-and-out-of-r" class="section level2">
<h2><span class="header-section-number">3.3</span> Getting Help In (and Out) of R</h2>
<p>One of the most helpful parts of R is being able to get help <em>inside</em> of R. For example, to get the manual associated with a function, class, dataset, or package, you can prepend the code of interest with a <code>?</code> to retrieve the relevant help page. For example, to get information about the <code>data.frame()</code> function, the <em>SingleCellExperiment</em> class, the in-built <em>iris</em> dataset, or for the <em>BiocManager</em> package, you can type:</p>
<pre class="sourceCode r"><code class="sourceCode r">?data.frame
?SingleCellExperiment
?iris
?BiocManager</code></pre>
<p>Beyond R, there are myriad online resources to get help. The R for Data Science book has a great section dedicated to looking for help <a href="https://r4ds.had.co.nz/introduction.html#getting-help-and-learning-more">outside of R</a>.</p>
</div>
<div id="bioconductor-documentation" class="section level2">
<h2><span class="header-section-number">3.4</span> Bioconductor Documentation</h2>
<p>One of the key tenets of Bioconductor software is the rich documentation of packages and workflows. Here, we cover some of the key resources</p>
<p>Bioconductor package vignettes and reference manuals
Bioconductor BiocViews and workflows
Bioconductor forum
Stackoverflow and Google for more generic R programming advice</p>
</div>
<div id="becoming-a-bioconductor-expert" class="section level2">
<h2><span class="header-section-number">3.5</span> Becoming a Bioconductor Expert</h2>
<p>While <a href="using-r-and-bioconductor.html#bioconductor-documentation">learning to use Bioconductor</a> tools is a very welcoming experience, there is unfortunately, no central resource for navigating the plethora of gotchas and paradigms associated with developing for Bioconductor. Based on conversations with folks involved in developing for Bioconductor, much of this knowledge is hard won and fairly spread out. This however is beginning to change with more recent efforts led by the Bioconductor team, and this book represents an earnest effort towards addressing this concern.</p>
<p>The <a href="https://github.com/Bioconductor/BiocWorkshops">BiocWorkshops</a> repository under the <a href="https://github.com/Bioconductor/">Bioconductor Github</a> provides a book composed of workshops that have been hosted by Bioconductor team members and contributors. These workshops center around learning, using, and developing for Bioconductor.</p>
<p>A host of topics are also available via the <a href="http://bioconductor.org/help/course-materials/">Learn</a> module on the Bioconductor website as well.</p>
</div>
<div id="conventions-used-in-this-book" class="section level2">
<h2><span class="header-section-number">3.6</span> Conventions used in this book</h2>
<p>Talk about code font, code blocks, function references, objects, classes.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="learning-r-and-more.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-infrastructure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
