<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 6 Workflow: Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 6 Workflow: Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Workflow: Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-02-28">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="about-the-data.html">
<link rel="next" href="workflow-clustering.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-essentials-of-sce"><i class="fa fa-check"></i><b>4.2</b> The Essentials of <code>sce</code></a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Workflow: Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#package-requirements"><i class="fa fa-check"></i><b>6.1</b> Package Requirements</a></li>
<li class="chapter" data-level="6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#loading-the-data"><i class="fa fa-check"></i><b>6.2</b> Loading the Data</a></li>
<li class="chapter" data-level="6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#preprocessing"><i class="fa fa-check"></i><b>6.3</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#working-with-common-genes"><i class="fa fa-check"></i><b>6.3.1</b> Working with Common Genes</a></li>
<li class="chapter" data-level="6.3.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>6.3.2</b> Cell and Gene Quality Control</a></li>
<li class="chapter" data-level="6.3.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#normalization"><i class="fa fa-check"></i><b>6.3.3</b> Normalization</a></li>
<li class="chapter" data-level="6.3.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#multibatch-normalization"><i class="fa fa-check"></i><b>6.3.4</b> Multibatch Normalization</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a></li>
<li class="chapter" data-level="6.5" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#combining-the-datasets"><i class="fa fa-check"></i><b>6.5</b> Combining the Datasets</a></li>
<li class="chapter" data-level="6.6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#integrating-datasets"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#naive-approach"><i class="fa fa-check"></i><b>6.6.1</b> Naive Approach</a></li>
<li class="chapter" data-level="6.6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#limma-batch-correction"><i class="fa fa-check"></i><b>6.6.2</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="6.6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#mnn-approach"><i class="fa fa-check"></i><b>6.6.3</b> MNN Approach</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#session-info"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="workflow-clustering.html"><a href="workflow-clustering.html"><i class="fa fa-check"></i><b>7</b> Workflow: Clustering</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="workflow-integrating-datasets" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Workflow: Integrating Datasets</h1>
<p>The purpose of this case study is to demonstrate how to integrate multiple scRNA-seq datasets using R/Bioconductor packages. In this workflow, we go from preprocessing the data to integrating the data and visualizing it in a reduced dimensionality space to showcase the success of the integration approach using mutual nearest neighbors relative to a naive approach. This approach is helpful for ameliorating batch effects that can be introduced when combining data from different sequencing runs and/or platforms.</p>
<p>Here, we will be combining two datasets from 10X Genomics PBMC Data. One dataset is comprised of 3000 PBMCs from a healthy donor, and the other dataset is comprised of 4000 PBMCs from a different healthy donor. Our goal is to produce an integrated representation of this data to facilitate downstream analysis, such as clustering.</p>
<div id="package-requirements" class="section level2">
<h2><span class="header-section-number">6.1</span> Package Requirements</h2>
<p>These packages will be required for working through the vignette, and can be installed by running the code below. The data that we will be using comes from the <code>TENxPBMCData</code> package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## required</span>
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&#39;scater&#39;</span>, <span class="st">&#39;scran&#39;</span>, <span class="st">&#39;limma&#39;</span>, <span class="st">&#39;TENxPBMCData&#39;</span>))

<span class="co">## suggested</span>
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&#39;BiocParallel&#39;</span>, <span class="st">&#39;BiocNeighbors&#39;</span>))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)
<span class="kw">library</span>(scran)
<span class="kw">library</span>(limma)
<span class="kw">library</span>(TENxPBMCData)
<span class="kw">library</span>(BiocParallel)
<span class="kw">library</span>(BiocNeighbors)</code></pre>
</div>
<div id="loading-the-data" class="section level2">
<h2><span class="header-section-number">6.2</span> Loading the Data</h2>
<p>Here we will be combining two different runs of scRNA-seq data - each from different healthy donors, and comprised of either 3000 cells (<code>pbmc3k</code>) or 4000 cells (<code>pbmc4k</code>). Note that these objects are already <code>SingleCellExperiment</code> objects.</p>
<pre class="sourceCode r"><code class="sourceCode r">pbmc3k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc3k&#39;</span>)
pbmc4k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)</code></pre>
</div>
<div id="preprocessing" class="section level2">
<h2><span class="header-section-number">6.3</span> Preprocessing</h2>
<p>Here we walk through the steps required to produce a clean expression matrix, taking the raw count data through to a normalized representation.</p>
<div id="working-with-common-genes" class="section level3">
<h3><span class="header-section-number">6.3.1</span> Working with Common Genes</h3>
<p>First, we find intersection of gene names and keep only the entries that are in common between the two datasets. We then reduce each of the individual datasets down to these matching entries (<code>keep_genes</code>) by subsetting.</p>
<pre class="sourceCode r"><code class="sourceCode r">keep_genes &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(pbmc3k), <span class="kw">rownames</span>(pbmc4k))
pbmc3k &lt;-<span class="st"> </span>pbmc3k[<span class="kw">match</span>(keep_genes, <span class="kw">rownames</span>(pbmc3k)), ]
pbmc4k &lt;-<span class="st"> </span>pbmc4k[<span class="kw">na.omit</span>(<span class="kw">match</span>(keep_genes, <span class="kw">rownames</span>(pbmc4k))), ]</code></pre>
</div>
<div id="cell-and-gene-quality-control" class="section level3">
<h3><span class="header-section-number">6.3.2</span> Cell and Gene Quality Control</h3>
<p>First, for the combined data <code>sce</code> and the individual datasets <code>pbmc3k</code> and <code>pbmc4k</code>, we calculate essential quality control characteristics using the <code>scater</code> function <code>calculateQCMetrics()</code>. We then determine cells low quality cells by finding outliers with uncharacteristically low total counts or total number of features (genes) detected. We automate this into a function.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## For pbmc3k</span>
pbmc3k &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(pbmc3k)
low_lib_pbmc3k &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(pbmc3k<span class="op">$</span>log10_total_counts, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmad=</span><span class="dv">3</span>)
low_genes_pbmc3k &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(pbmc3k<span class="op">$</span>log10_total_features_by_counts, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmad=</span><span class="dv">3</span>)

<span class="co">## For pbmc4k</span>
pbmc4k &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(pbmc4k)
low_lib_pbmc4k &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(pbmc4k<span class="op">$</span>log10_total_counts, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmad=</span><span class="dv">3</span>)
low_genes_pbmc4k &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(pbmc4k<span class="op">$</span>log10_total_features_by_counts, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmad=</span><span class="dv">3</span>)</code></pre>
<p>These results flag approximately 30 to 100 cells for removal from each of the datasets. We can then further subset our data to remove these cells by running the following:</p>
<pre class="sourceCode r"><code class="sourceCode r">pbmc3k &lt;-<span class="st"> </span>pbmc3k[,<span class="op">!</span>(low_lib_pbmc3k <span class="op">|</span><span class="st"> </span>low_genes_pbmc3k)]
pbmc4k &lt;-<span class="st"> </span>pbmc4k[,<span class="op">!</span>(low_lib_pbmc4k <span class="op">|</span><span class="st"> </span>low_genes_pbmc4k)]</code></pre>
</div>
<div id="normalization" class="section level3">
<h3><span class="header-section-number">6.3.3</span> Normalization</h3>
<p>From here, we now compute the size factors using the <code>scran</code> package’s <code>computeSumFactors()</code> function, and apply the size factors via the <code>scran</code> package’s <code>normalize()</code> function to produce a new assay, <code>logcounts</code>, within each <code>SingleCellExperiment</code> object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## compute the sizeFactors</span>
pbmc3k &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(pbmc3k)
pbmc4k &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(pbmc4k)

<span class="co">## Normalize (using already calculated size factors)</span>
pbmc3k &lt;-<span class="st"> </span><span class="kw">normalize</span>(pbmc3k)
pbmc4k &lt;-<span class="st"> </span><span class="kw">normalize</span>(pbmc4k)</code></pre>
</div>
<div id="multibatch-normalization" class="section level3">
<h3><span class="header-section-number">6.3.4</span> Multibatch Normalization</h3>
<p>We also rescale each batch to adjust for differences in sequencing depth between batches. The <code>multiBatchNorm()</code> function from the <code>scran</code> package recomputes log-normalized expression values after adjusting the size factors for systematic differences in coverage between <code>SingleCellExperiment</code> objects. The previously computed size factors only remove biases between cells within a single batch. This improves the quality of the correction step by removing one aspect of the technical differences between batches.</p>
<pre class="sourceCode r"><code class="sourceCode r">rescaled &lt;-<span class="st"> </span><span class="kw">multiBatchNorm</span>(pbmc3k, pbmc4k)
pbmc3k &lt;-<span class="st"> </span>rescaled[[<span class="dv">1</span>]]
pbmc4k &lt;-<span class="st"> </span>rescaled[[<span class="dv">2</span>]]</code></pre>
</div>
</div>
<div id="feature-selection" class="section level2">
<h2><span class="header-section-number">6.4</span> Feature Selection</h2>
<p>A key step across many data integration methods is the selection of informative features across the different experiments. This helps to speed up computation and possibly improve the resulting integration.</p>
<p>Once again, we rely on the <code>scran</code> package to identify the genes with the highest biological coefficient of variation, using the <code>trendVar()</code> and <code>decomposeVar()</code> functions to calculate the per gene variance and separate it into technical versus biological components. We perform this for each individual dataset:</p>
<pre class="sourceCode r"><code class="sourceCode r">fit_pbmc3k &lt;-<span class="st"> </span><span class="kw">trendVar</span>(pbmc3k, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>) 
dec_pbmc3k &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(pbmc3k, fit_pbmc3k)
dec_pbmc3k<span class="op">$</span>Symbol_TENx &lt;-<span class="st"> </span><span class="kw">rowData</span>(pbmc3k)<span class="op">$</span>Symbol_TENx
dec_pbmc3k &lt;-<span class="st"> </span>dec_pbmc3k[<span class="kw">order</span>(dec_pbmc3k<span class="op">$</span>bio, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>), ]

fit_pbmc4k &lt;-<span class="st"> </span><span class="kw">trendVar</span>(pbmc4k, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>) 
dec_pbmc4k &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(pbmc4k, fit_pbmc4k)
dec_pbmc4k<span class="op">$</span>Symbol_TENx &lt;-<span class="st"> </span><span class="kw">rowData</span>(pbmc4k)<span class="op">$</span>Symbol_TENx
dec_pbmc4k &lt;-<span class="st"> </span>dec_pbmc4k[<span class="kw">order</span>(dec_pbmc4k<span class="op">$</span>bio, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>), ]</code></pre>
<p>Then select the most informative genes that are shared across <em>both</em> datasets:</p>
<pre class="sourceCode r"><code class="sourceCode r">universe &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(dec_pbmc3k), <span class="kw">rownames</span>(dec_pbmc4k))
mean.bio &lt;-<span class="st"> </span>(dec_pbmc3k[universe,<span class="st">&quot;bio&quot;</span>] <span class="op">+</span><span class="st"> </span>dec_pbmc4k[universe,<span class="st">&quot;bio&quot;</span>])<span class="op">/</span><span class="dv">2</span>
hvg_genes &lt;-<span class="st"> </span>universe[mean.bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</code></pre>
</div>
<div id="combining-the-datasets" class="section level2">
<h2><span class="header-section-number">6.5</span> Combining the Datasets</h2>
<p>Finally, we combine the datasets into a unified <code>SingleCellExperiment</code> object for the downstream integration approaches, now that the data has been normalized (both within and between datasets) and the shared most informative features have been identified.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## total raw counts</span>
counts_pbmc &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">counts</span>(pbmc3k), <span class="kw">counts</span>(pbmc4k))

<span class="co">## total normalized counts (with multibatch normalization)</span>
logcounts_pbmc &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">logcounts</span>(pbmc3k), <span class="kw">logcounts</span>(pbmc4k))

sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>( 
    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> counts_pbmc, <span class="dt">logcounts =</span> logcounts_pbmc),  
    <span class="dt">rowData =</span> <span class="kw">rowData</span>(pbmc3k), <span class="co"># same as rowData(pbmc4k) </span>
    <span class="dt">colData =</span> <span class="kw">rbind</span>(<span class="kw">colData</span>(pbmc3k), <span class="kw">colData</span>(pbmc4k)) 
) </code></pre>
<p>For safekeeping, we will also store the <code>hvg_genes</code> from the prior section into the <code>sce</code> object via:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes &lt;-<span class="st"> </span>hvg_genes</code></pre>
</div>
<div id="integrating-datasets" class="section level2">
<h2><span class="header-section-number">6.6</span> Integrating Datasets</h2>
<p>Here we will now be comparing the results of different approaches to integration.</p>
<div id="naive-approach" class="section level3">
<h3><span class="header-section-number">6.6.1</span> Naive Approach</h3>
<p>The naive approach simply entails visualizing the combined <code>sce</code> object post-normalization with no attempt at batch correction. Here we manually calculate the PCA on the normalized data (retrieved via <code>logcounts(sce)</code> or, similarly, via <code>assay(sce, "logcounts")</code>, and then assign the result into the <code>reducedDim</code> slot of <code>sce</code>, naming it <code>"PCA_naive"</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">px &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce)[hvg_genes, ]))
<span class="kw">reducedDim</span>(sce, <span class="st">&quot;PCA_naive&quot;</span>) &lt;-<span class="st"> </span>px<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;PCA_naive&quot;</span>,
               <span class="dt">colour_by=</span><span class="st">&quot;Sample&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;PCA Without batch correction&quot;</span>)</code></pre>
<p><img src="OSCA_files/figure-html/unnamed-chunk-47-1.png" width="672" /></p>
</div>
<div id="limma-batch-correction" class="section level3">
<h3><span class="header-section-number">6.6.2</span> Limma Batch Correction</h3>
<p>The <code>limma</code> package, a popular framework for the statistical analysis of RNA-seq, has a function <code>removeBatchEffect()</code> which will be used here to correct the normalized expression matrix <code>logcounts</code> across the two batches. The result will be assigned into the <code>assays</code> slot of the <code>sce</code> object as <code>limma_corrected</code>, and then used for PCA, saving the result in the <code>reducedDim</code> slot as <code>"PCA_limma"</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">limma_corrected &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">removeBatchEffect</span>(<span class="kw">logcounts</span>(sce), <span class="dt">batch =</span> sce<span class="op">$</span>Sample)
<span class="kw">assay</span>(sce, <span class="st">&quot;logcounts_limma&quot;</span>) &lt;-<span class="st"> </span>limma_corrected <span class="co">## add new assay</span>

pl &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(<span class="kw">assay</span>(sce, <span class="st">&#39;logcounts_limma&#39;</span>)[hvg_genes, ]))
<span class="kw">reducedDim</span>(sce, <span class="st">&quot;PCA_limma&quot;</span>) &lt;-<span class="st"> </span>pl<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;PCA_limma&quot;</span>,
               <span class="dt">colour_by=</span><span class="st">&quot;Sample&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;PCA With limma removeBatchEffect() correction&quot;</span>)</code></pre>
<p><img src="OSCA_files/figure-html/unnamed-chunk-49-1.png" width="672" /></p>
</div>
<div id="mnn-approach" class="section level3">
<h3><span class="header-section-number">6.6.3</span> MNN Approach</h3>
<p>The mutual nearest neighbors (MNN) approach within the <code>scran</code> package utilizes a novel approach to adjust for batch effects. The <code>fastMNN()</code> function returns a representation of the data with reduced dimensionality, which can be used in a similar fashion to other lower-dimensional representations such as PCA. In particular, this representation can be used for downstream methods such as clustering.</p>
<p>Where <code>fastMNN()</code> differs from other integration methods such as the limma approach above is that it does <em>not</em> produce a batch-corrected expression matrix. Thus, the result from <code>fastMNN()</code> should solely be treated as a reduced dimensionality representation, suitable for direct plotting, TSNE/UMAP, clustering, and trajectory analysis that relies on such results. The already (batch) normalized (via <code>normalize()</code> and <code>multiBatchNorm()</code>) can be supplied to other statistical frameworks that are better suited to handle batch effects, such as in the case of differential expression.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Basic method to run - not run</span>
mnn_out &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(sce[hvg_genes, sce<span class="op">$</span>Sample <span class="op">==</span><span class="st"> &quot;pbmc3k&quot;</span>],
                  sce[hvg_genes, sce<span class="op">$</span>Sample <span class="op">==</span><span class="st"> &quot;pbmc4k&quot;</span>],
                  <span class="co">## subset.row = hvg_genes, ## same as subsetting above</span>
                  <span class="dt">k =</span> <span class="dv">20</span>, <span class="dt">d =</span> <span class="dv">50</span>, <span class="dt">approximate =</span> <span class="ot">TRUE</span>,
                  <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
                  <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">multiCoreParam</span>())</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Adding parallelization and Annoy method for approximate nearest neighbors</span>
<span class="co">## this makes fastMNN significantly faster on large data</span>
mnn_out &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(sce[hvg_genes, sce<span class="op">$</span>Sample <span class="op">==</span><span class="st"> &quot;pbmc3k&quot;</span>],
                  sce[hvg_genes, sce<span class="op">$</span>Sample <span class="op">==</span><span class="st"> &quot;pbmc4k&quot;</span>],
                  <span class="co">## subset.row = hvg_genes, ## same as subsetting above</span>
                  <span class="dt">k =</span> <span class="dv">20</span>, <span class="dt">d =</span> <span class="dv">50</span>, <span class="dt">approximate =</span> <span class="ot">TRUE</span>,
                  <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
                  <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>(<span class="dv">8</span>))

<span class="kw">reducedDim</span>(sce, <span class="st">&quot;MNN&quot;</span>) &lt;-<span class="st"> </span>mnn_out<span class="op">$</span>correct</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;MNN&quot;</span>,
                    <span class="dt">colour_by=</span><span class="st">&quot;Sample&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;MNN Ouput Reduced Dimensions&quot;</span>)</code></pre>
<p><img src="OSCA_files/figure-html/unnamed-chunk-52-1.png" width="672" /></p>
</div>
</div>
<div id="session-info" class="section level2">
<h2><span class="header-section-number">6.7</span> Session Info</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre>
<pre><code>## R Under development (unstable) (2019-01-13 r75986)
## Platform: x86_64-apple-darwin18.2.0 (64-bit)
## Running under: macOS Mojave 10.14.3
## 
## Matrix products: default
## BLAS/LAPACK: /usr/local/Cellar/openblas/0.3.5/lib/libopenblasp-r0.3.5.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BiocNeighbors_1.1.11        TENxPBMCData_1.1.0         
##  [3] HDF5Array_1.11.10           rhdf5_2.27.12              
##  [5] limma_3.39.12               scran_1.11.20              
##  [7] scater_1.11.10              ggplot2_3.1.0              
##  [9] SingleCellExperiment_1.5.2  SummarizedExperiment_1.13.0
## [11] DelayedArray_0.9.8          BiocParallel_1.17.17       
## [13] matrixStats_0.54.0          Biobase_2.43.1             
## [15] GenomicRanges_1.35.1        GenomeInfoDb_1.19.2        
## [17] IRanges_2.17.4              S4Vectors_0.21.10          
## [19] BiocGenerics_0.29.1         here_0.1                   
## [21] fs_1.2.6                    usethis_1.4.0              
## [23] devtools_2.0.1             
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  bit64_0.9-7                  
##  [3] httr_1.4.0                    rprojroot_1.3-2              
##  [5] dynamicTreeCut_1.63-1         tools_3.6.0                  
##  [7] backports_1.1.3               irlba_2.3.3                  
##  [9] R6_2.4.0                      vipor_0.4.5                  
## [11] DBI_1.0.0                     lazyeval_0.2.1               
## [13] colorspace_1.4-0              withr_2.1.2                  
## [15] tidyselect_0.2.5              gridExtra_2.3                
## [17] prettyunits_1.0.2             processx_3.2.1               
## [19] curl_3.3                      bit_1.1-14                   
## [21] compiler_3.6.0                cli_1.0.1                    
## [23] desc_1.2.0                    bookdown_0.9                 
## [25] labeling_0.3                  scales_1.0.0                 
## [27] callr_3.1.1                   stringr_1.4.0                
## [29] digest_0.6.18                 rmarkdown_1.11               
## [31] XVector_0.23.0                htmltools_0.3.6              
## [33] pkgconfig_2.0.2               sessioninfo_1.1.1            
## [35] rlang_0.3.1                   rstudioapi_0.9.0             
## [37] RSQLite_2.1.1                 shiny_1.2.0                  
## [39] DelayedMatrixStats_1.5.2      dplyr_0.8.0.1                
## [41] RCurl_1.95-4.11               magrittr_1.5                 
## [43] GenomeInfoDbData_1.2.0        Matrix_1.2-15                
## [45] Rcpp_1.0.0                    ggbeeswarm_0.6.0             
## [47] munsell_0.5.0                 Rhdf5lib_1.5.1               
## [49] viridis_0.5.1                 stringi_1.3.1                
## [51] yaml_2.2.0                    edgeR_3.25.3                 
## [53] zlibbioc_1.29.0               Rtsne_0.15                   
## [55] pkgbuild_1.0.2                plyr_1.8.4                   
## [57] AnnotationHub_2.15.7          grid_3.6.0                   
## [59] blob_1.1.1                    promises_1.0.1               
## [61] ExperimentHub_1.9.1           crayon_1.3.4                 
## [63] lattice_0.20-38               cowplot_0.9.4                
## [65] beachmat_1.5.4                locfit_1.5-9.1               
## [67] knitr_1.21                    ps_1.3.0                     
## [69] pillar_1.3.1                  igraph_1.2.4                 
## [71] pkgload_1.0.2                 glue_1.3.0                   
## [73] evaluate_0.13                 remotes_2.0.2                
## [75] BiocManager_1.30.4            httpuv_1.4.5.1               
## [77] testthat_2.0.1                gtable_0.2.0                 
## [79] purrr_0.3.0                   assertthat_0.2.0             
## [81] xfun_0.5                      mime_0.6                     
## [83] xtable_1.8-3                  later_0.8.0                  
## [85] viridisLite_0.3.0             tibble_2.0.1                 
## [87] AnnotationDbi_1.45.0          beeswarm_0.2.3               
## [89] memoise_1.1.0                 statmod_1.4.30               
## [91] interactiveDisplayBase_1.21.0</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(sce, <span class="st">&quot;_rfiles/_data/integration_sce.rds&quot;</span>, <span class="dt">compress =</span> <span class="st">&quot;xz&quot;</span>)</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="about-the-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="workflow-clustering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
