<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 3 Clustering and Differential Expression Workflow | Orchestrating Single-Cell Analysis</title>
  <meta name="description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 3 Clustering and Differential Expression Workflow | Orchestrating Single-Cell Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Clustering and Differential Expression Workflow | Orchestrating Single-Cell Analysis" />
  
  <meta name="twitter:description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-01-08">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="intro.html">
<link rel="next" href="trajectory-analysis-workflow.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Prerequisites</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#prerequisites-1"><i class="fa fa-check"></i><b>2.1</b> Prerequisites</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html"><i class="fa fa-check"></i><b>3</b> Clustering and Differential Expression Workflow</a><ul>
<li class="chapter" data-level="3.1" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#introduction"><i class="fa fa-check"></i><b>3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#learning-objectives"><i class="fa fa-check"></i><b>3.2</b> Learning objectives</a></li>
<li class="chapter" data-level="3.3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#package-requirements"><i class="fa fa-check"></i><b>3.3</b> Package requirements</a></li>
<li class="chapter" data-level="3.4" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#preamble"><i class="fa fa-check"></i><b>3.4</b> Preamble</a></li>
<li class="chapter" data-level="3.5" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#data"><i class="fa fa-check"></i><b>3.5</b> Data</a><ul>
<li class="chapter" data-level="3.5.1" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#description-of-data"><i class="fa fa-check"></i><b>3.5.1</b> Description of Data</a></li>
<li class="chapter" data-level="3.5.2" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#loading-the-data"><i class="fa fa-check"></i><b>3.5.2</b> Loading the Data</a></li>
<li class="chapter" data-level="3.5.3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#exploring-the-raw-data"><i class="fa fa-check"></i><b>3.5.3</b> Exploring the Raw Data</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#preprocessing"><i class="fa fa-check"></i><b>3.6</b> Preprocessing</a><ul>
<li class="chapter" data-level="3.6.1" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#rename-features"><i class="fa fa-check"></i><b>3.6.1</b> Rename features</a></li>
<li class="chapter" data-level="3.6.2" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>3.6.2</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="3.6.3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#normalization"><i class="fa fa-check"></i><b>3.6.3</b> Normalization</a></li>
<li class="chapter" data-level="3.6.4" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#feature-selection"><i class="fa fa-check"></i><b>3.6.4</b> Feature selection</a></li>
<li class="chapter" data-level="3.6.5" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#dimensionality-reduction"><i class="fa fa-check"></i><b>3.6.5</b> Dimensionality reduction</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#clustering"><i class="fa fa-check"></i><b>3.7</b> Clustering</a></li>
<li class="chapter" data-level="3.8" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#differential-expression"><i class="fa fa-check"></i><b>3.8</b> Differential Expression</a></li>
<li class="chapter" data-level="3.9" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html"><i class="fa fa-check"></i><b>4</b> Trajectory Analysis Workflow</a><ul>
<li class="chapter" data-level="4.1" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#learning-objectives-1"><i class="fa fa-check"></i><b>4.2</b> Learning objectives</a></li>
<li class="chapter" data-level="4.3" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#package-requirements-1"><i class="fa fa-check"></i><b>4.3</b> Package requirements</a></li>
<li class="chapter" data-level="4.4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#loading-the-data-1"><i class="fa fa-check"></i><b>4.4</b> Loading the data</a></li>
<li class="chapter" data-level="4.5" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#preprocessing-1"><i class="fa fa-check"></i><b>4.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="4.5.1" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#cell-and-gene-quality-control-1"><i class="fa fa-check"></i><b>4.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="4.5.2" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#normalization-1"><i class="fa fa-check"></i><b>4.5.2</b> Normalization</a></li>
<li class="chapter" data-level="4.5.3" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>4.5.3</b> Feature selection</a></li>
<li class="chapter" data-level="4.5.4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>4.5.4</b> Dimensionality reduction</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#trajectory-analysis"><i class="fa fa-check"></i><b>4.6</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="4.7" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#session-info-1"><i class="fa fa-check"></i><b>4.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html"><i class="fa fa-check"></i><b>5</b> Geneset Analysis and Gene Expression Networks</a><ul>
<li class="chapter" data-level="5.1" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#learning-objectives-2"><i class="fa fa-check"></i><b>5.2</b> Learning objectives</a></li>
<li class="chapter" data-level="5.3" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#package-requirements-2"><i class="fa fa-check"></i><b>5.3</b> Package requirements</a></li>
<li class="chapter" data-level="5.4" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#loading-the-data-2"><i class="fa fa-check"></i><b>5.4</b> Loading the data</a></li>
<li class="chapter" data-level="5.5" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#preprocessing-2"><i class="fa fa-check"></i><b>5.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="5.5.1" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#cell-and-gene-quality-control-2"><i class="fa fa-check"></i><b>5.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="5.5.2" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#normalization-2"><i class="fa fa-check"></i><b>5.5.2</b> Normalization</a></li>
<li class="chapter" data-level="5.5.3" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#feature-selection-2"><i class="fa fa-check"></i><b>5.5.3</b> Feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#geneset-analysis"><i class="fa fa-check"></i><b>5.6</b> Geneset Analysis</a></li>
<li class="chapter" data-level="5.7" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#gene-expression-networks"><i class="fa fa-check"></i><b>5.7</b> Gene Expression Networks</a></li>
<li class="chapter" data-level="5.8" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#session-info-2"><i class="fa fa-check"></i><b>5.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="integrating-datasets.html"><a href="integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#introduction-3"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#learning-objectives-3"><i class="fa fa-check"></i><b>6.2</b> Learning objectives</a></li>
<li class="chapter" data-level="6.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#package-requirements-3"><i class="fa fa-check"></i><b>6.3</b> Package requirements</a></li>
<li class="chapter" data-level="6.4" data-path="integrating-datasets.html"><a href="integrating-datasets.html#data-1"><i class="fa fa-check"></i><b>6.4</b> Data</a><ul>
<li class="chapter" data-level="6.4.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#description-of-data-1"><i class="fa fa-check"></i><b>6.4.1</b> Description of Data</a></li>
<li class="chapter" data-level="6.4.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#loading-the-data-3"><i class="fa fa-check"></i><b>6.4.2</b> Loading the Data</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="integrating-datasets.html"><a href="integrating-datasets.html#preprocessing-3"><i class="fa fa-check"></i><b>6.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.5.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#cell-and-gene-quality-control-3"><i class="fa fa-check"></i><b>6.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="6.5.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#normalization-3"><i class="fa fa-check"></i><b>6.5.2</b> Normalization</a></li>
<li class="chapter" data-level="6.5.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#feature-selection-3"><i class="fa fa-check"></i><b>6.5.3</b> Feature selection</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="integrating-datasets.html"><a href="integrating-datasets.html#integrating-datasets-1"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a></li>
<li class="chapter" data-level="6.7" data-path="integrating-datasets.html"><a href="integrating-datasets.html#session-info-3"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="clustering-and-differential-expression-workflow" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Clustering and Differential Expression Workflow</h1>
<p><em>Author</em>: Robert A. Amezquita, Fred Hutchinson Cancer Research Center</p>
<div id="introduction" class="section level2">
<h2><span class="header-section-number">3.1</span> Introduction</h2>
<p>Introduce the main purpose of this vignette in paragraph form, why these steps are important, what the end products allow for.</p>
</div>
<div id="learning-objectives" class="section level2">
<h2><span class="header-section-number">3.2</span> Learning objectives</h2>
<ul>
<li>In bullet form, illustrate what the main outputs are of this vignette</li>
<li>A couple of bullet points are fine</li>
</ul>
</div>
<div id="package-requirements" class="section level2">
<h2><span class="header-section-number">3.3</span> Package requirements</h2>
<p>Packages that are required to go through the vignette and how to install them, should largely be through BiocManager interface.</p>
<pre class="sourceCode r"><code class="sourceCode r">bioc_pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;TENxPBMCData&quot;</span>,
         <span class="st">&quot;scater&quot;</span>)

cran_pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tidyverse&quot;</span>)

BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs)
<span class="kw">install.packages</span>(cran_pkgs)</code></pre>
</div>
<div id="preamble" class="section level2">
<h2><span class="header-section-number">3.4</span> Preamble</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)</code></pre>
</div>
<div id="data" class="section level2">
<h2><span class="header-section-number">3.5</span> Data</h2>
<div id="description-of-data" class="section level3">
<h3><span class="header-section-number">3.5.1</span> Description of Data</h3>
</div>
<div id="loading-the-data" class="section level3">
<h3><span class="header-section-number">3.5.2</span> Loading the Data</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TENxPBMCData)
pbmc4k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)

<span class="co">## optional: increases memory for loading data into memory</span>
<span class="co">## options(DelayedArray.block.size=8e9) </span></code></pre>
</div>
<div id="exploring-the-raw-data" class="section level3">
<h3><span class="header-section-number">3.5.3</span> Exploring the Raw Data</h3>
<p>Initial state of the rowdata and coldata.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(pbmc4k)[<span class="dv">1</span>, ]</code></pre>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-5">Table 3.1: </span>Column data of SingleCellExperiment object.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:left;">
Sequence
</th>
<th style="text-align:right;">
Library
</th>
<th style="text-align:left;">
Cell_ranger_version
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
AAACCTGAGAAGGCCT-1
</td>
<td style="text-align:left;">
AAACCTGAGAAGGCCT
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v2.1
</td>
</tr>
<tr>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
AAACCTGAGACAGACC-1
</td>
<td style="text-align:left;">
AAACCTGAGACAGACC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v2.1
</td>
</tr>
</tbody>
</table>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Tissue_status
</th>
<th style="text-align:left;">
Barcode_type
</th>
<th style="text-align:left;">
Chemistry
</th>
<th style="text-align:left;">
Sequence_platform
</th>
<th style="text-align:left;">
Individual
</th>
<th style="text-align:left;">
Date_published
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Chromium
</td>
<td style="text-align:left;">
Chromium_v2
</td>
<td style="text-align:left;">
Hiseq4000
</td>
<td style="text-align:left;">
HealthyDonor1
</td>
<td style="text-align:left;">
2017-11-08
</td>
</tr>
<tr>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Chromium
</td>
<td style="text-align:left;">
Chromium_v2
</td>
<td style="text-align:left;">
Hiseq4000
</td>
<td style="text-align:left;">
HealthyDonor1
</td>
<td style="text-align:left;">
2017-11-08
</td>
</tr>
</tbody>
</table>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowData</span>(pbmc4k)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]</code></pre>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-7">Table 3.2: </span>Row data of SingleCellExperiment object.
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
ENSEMBL_ID
</th>
<th style="text-align:left;">
Symbol_TENx
</th>
<th style="text-align:left;">
Symbol
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSG00000243485
</td>
<td style="text-align:left;">
ENSG00000243485
</td>
<td style="text-align:left;">
RP11-34P13.3
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000237613
</td>
<td style="text-align:left;">
ENSG00000237613
</td>
<td style="text-align:left;">
FAM138A
</td>
<td style="text-align:left;">
FAM138A
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000186092
</td>
<td style="text-align:left;">
ENSG00000186092
</td>
<td style="text-align:left;">
OR4F5
</td>
<td style="text-align:left;">
OR4F5
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000238009
</td>
<td style="text-align:left;">
ENSG00000238009
</td>
<td style="text-align:left;">
RP11-34P13.7
</td>
<td style="text-align:left;">
LOC100996442
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000239945
</td>
<td style="text-align:left;">
ENSG00000239945
</td>
<td style="text-align:left;">
RP11-34P13.8
</td>
<td style="text-align:left;">
NA
</td>
</tr>
</tbody>
</table>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">counts</span>(pbmc4k)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-9">Table 3.3: </span>Count data of SingleCellExperiment object.
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
<th style="text-align:right;">
V4
</th>
<th style="text-align:right;">
V5
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSG00000243485
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000237613
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000186092
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="preprocessing" class="section level2">
<h2><span class="header-section-number">3.6</span> Preprocessing</h2>
<div id="rename-features" class="section level3">
<h3><span class="header-section-number">3.6.1</span> Rename features</h3>
<p>Remaps rownames to unique TENx based symbols.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rownames</span>(pbmc4k) &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">uniquifyFeatureNames</span>(<span class="dt">ID =</span> <span class="kw">rowData</span>(pbmc4k)<span class="op">$</span>ENSEMBL_ID,
                                               <span class="dt">names =</span> <span class="kw">rowData</span>(pbmc4k)<span class="op">$</span>Symbol_TENx)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">counts</span>(pbmc4k)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">.pretty_table</span>(<span class="st">&quot;Count data with transformed row identifiers.&quot;</span>)</code></pre>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-11">Table 3.4: </span>Count data with transformed row identifiers.
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
<th style="text-align:right;">
V4
</th>
<th style="text-align:right;">
V5
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
RP11-34P13.3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
FAM138A
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
OR4F5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
RP11-34P13.7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
RP11-34P13.8
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
<div id="cell-and-gene-quality-control" class="section level3">
<h3><span class="header-section-number">3.6.2</span> Cell and gene quality control</h3>
<p>Initial QC metrics appended to the SCE object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)
pbmc4k &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">calculateQCMetrics</span>(pbmc4k)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">colData</span>(pbmc4k)[<span class="dv">1</span>, ], <span class="dt">digits =</span> <span class="dv">2</span>, <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
            <span class="dt">caption =</span> <span class="st">&quot;Column data with new entries from `scater::calculateQCMetrics`.&quot;</span>)</code></pre>
<table>
<caption>
<span id="tab:unnamed-chunk-13">Table 3.5: </span>Column data with new entries from <code>scater::calculateQCMetrics</code>.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:left;">
Sequence
</th>
<th style="text-align:right;">
Library
</th>
<th style="text-align:left;">
Cell_ranger_version
</th>
<th style="text-align:left;">
Tissue_status
</th>
<th style="text-align:left;">
Barcode_type
</th>
<th style="text-align:left;">
Chemistry
</th>
<th style="text-align:left;">
Sequence_platform
</th>
<th style="text-align:left;">
Individual
</th>
<th style="text-align:left;">
Date_published
</th>
<th style="text-align:left;">
is_cell_control
</th>
<th style="text-align:right;">
total_features_by_counts
</th>
<th style="text-align:right;">
log10_total_features_by_counts
</th>
<th style="text-align:right;">
total_counts
</th>
<th style="text-align:right;">
log10_total_counts
</th>
<th style="text-align:right;">
pct_counts_in_top_50_features
</th>
<th style="text-align:right;">
pct_counts_in_top_100_features
</th>
<th style="text-align:right;">
pct_counts_in_top_200_features
</th>
<th style="text-align:right;">
pct_counts_in_top_500_features
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
AAACCTGAGAAGGCCT-1
</td>
<td style="text-align:left;">
AAACCTGAGAAGGCCT
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v2.1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Chromium
</td>
<td style="text-align:left;">
Chromium_v2
</td>
<td style="text-align:left;">
Hiseq4000
</td>
<td style="text-align:left;">
HealthyDonor1
</td>
<td style="text-align:left;">
2017-11-08
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
748
</td>
<td style="text-align:right;">
2.87
</td>
<td style="text-align:right;">
1738
</td>
<td style="text-align:right;">
3.24
</td>
<td style="text-align:right;">
38
</td>
<td style="text-align:right;">
52.1
</td>
<td style="text-align:right;">
67
</td>
<td style="text-align:right;">
85.7
</td>
</tr>
</tbody>
</table>
<pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">rowData</span>(pbmc4k)[<span class="dv">1</span>, ], <span class="dt">digits =</span> <span class="dv">2</span>, <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
            <span class="dt">caption =</span> <span class="st">&quot;Row data with new entries from `scater::calculateQCMetrics`.&quot;</span>)</code></pre>
<table>
<caption>
<span id="tab:unnamed-chunk-13">Table 3.5: </span>Row data with new entries from <code>scater::calculateQCMetrics</code>.
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
ENSEMBL_ID
</th>
<th style="text-align:left;">
Symbol_TENx
</th>
<th style="text-align:left;">
Symbol
</th>
<th style="text-align:left;">
is_feature_control
</th>
<th style="text-align:right;">
mean_counts
</th>
<th style="text-align:right;">
log10_mean_counts
</th>
<th style="text-align:right;">
n_cells_by_counts
</th>
<th style="text-align:right;">
pct_dropout_by_counts
</th>
<th style="text-align:right;">
total_counts
</th>
<th style="text-align:right;">
log10_total_counts
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
RP11-34P13.3
</td>
<td style="text-align:left;">
ENSG00000243485
</td>
<td style="text-align:left;">
RP11-34P13.3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
<pre class="sourceCode r"><code class="sourceCode r">scater<span class="op">::</span><span class="kw">plotExprsFreqVsMean</span>(pbmc4k)</code></pre>
<p><img src="orchestratingSingleCellAnalysis_files/figure-html/unnamed-chunk-14-1.png" width="1152" /></p>
<pre class="sourceCode r"><code class="sourceCode r">scater<span class="op">::</span><span class="kw">plotRowData</span>(pbmc4k, <span class="dt">x =</span> <span class="st">&#39;n_cells_by_counts&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;mean_counts&#39;</span>)</code></pre>
<p><img src="orchestratingSingleCellAnalysis_files/figure-html/unnamed-chunk-14-2.png" width="1152" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DropletUtils)
bcrank &lt;-<span class="st"> </span>DropletUtils<span class="op">::</span><span class="kw">barcodeRanks</span>(<span class="kw">as.matrix</span>(<span class="kw">counts</span>(pbmc4k)))

<span class="co">## this takes a while...might not ever finish</span>
<span class="co">## e.out &lt;- DropletUtils::emptyDrops(as.matrix(counts(pbmc4k)))</span>


<span class="kw">plot</span>(bcrank<span class="op">$</span>rank, bcrank<span class="op">$</span>total, <span class="dt">log =</span> <span class="st">&#39;xy&#39;</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">metadata</span>(bcrank)<span class="op">$</span>inflection, <span class="dt">col =</span> <span class="st">&#39;darkgreen&#39;</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">metadata</span>(bcrank)<span class="op">$</span>knee, <span class="dt">col =</span> <span class="st">&#39;dodgerblue&#39;</span>) <span class="co"># use as filtering threshold!</span></code></pre>
<p><img src="orchestratingSingleCellAnalysis_files/figure-html/unnamed-chunk-14-3.png" width="1152" /></p>
</div>
<div id="normalization" class="section level3">
<h3><span class="header-section-number">3.6.3</span> Normalization</h3>
<p>Application of normalization approach, even if its just a log normalization on the count data. Cell cycle normalization may be applicable here.</p>
</div>
<div id="feature-selection" class="section level3">
<h3><span class="header-section-number">3.6.4</span> Feature selection</h3>
<p>Identifying the subset of the clean matrix to work with.</p>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3><span class="header-section-number">3.6.5</span> Dimensionality reduction</h3>
<p>Calculating PCs and UMAP/tSNE representation.</p>
</div>
</div>
<div id="clustering" class="section level2">
<h2><span class="header-section-number">3.7</span> Clustering</h2>
<p>Strategy 1 of interest. Use BiocParallel. May need to save intermediate data within package to facilitate fast construction of Rmd.</p>
<ul>
<li>SC3</li>
<li>clusterExperiment</li>
<li>clustree</li>
<li>BEARscc # uses ERCC spike ins</li>
<li>BiocNeighbors # more for devs</li>
</ul>
</div>
<div id="differential-expression" class="section level2">
<h2><span class="header-section-number">3.8</span> Differential Expression</h2>
<p>Strategy 2 of interest. Differential expression approaches. Use one of the outputs from above to perform differential expression analyses comparison.</p>
<ul>
<li>zinbwave + edgeR/DESeq2 (may exclude as workflow is very different or highlight as an “advanced section”)</li>
<li>MAST</li>
<li>scDD</li>
<li>SC3</li>
</ul>
</div>
<div id="session-info" class="section level2">
<h2><span class="header-section-number">3.9</span> Session Info</h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="trajectory-analysis-workflow.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["orchestratingSingleCellAnalysis.pdf", "orchestratingSingleCellAnalysis.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
