# Clustering and Differential Expression Workflow

_Author_: Robert A. Amezquita, Fred Hutchinson Cancer Research Center

## Introduction

Introduce the main purpose of this vignette in paragraph form, why these steps are important, what the end products allow for.

## Learning objectives

* In bullet form, illustrate what the main outputs are of this vignette
* A couple of bullet points are fine

## Package requirements

Packages that are required to go through the vignette and how to install them, should largely be through BiocManager interface.


```r
BiocManager::install(c('SC3', 'clusterExperiment', 'BiocParallel', 'BEARscc', 'zinbwave', 'edgeR', 'DESeq2', 'MAST', 'scDD'))
```

## Loading the data

Here the data should be loaded if in package form and briefly explored to illustrate the contents of the data. 

If data comes from "raw" form, then a simple pipeline for creating the processed data should be illustrated as succinctly as possible.


```r
library(TENxPBMCData)
```

```
## Loading required package: SingleCellExperiment
```

```
## Loading required package: SummarizedExperiment
```

```
## Loading required package: GenomicRanges
```

```
## Loading required package: stats4
```

```
## Loading required package: BiocGenerics
```

```
## Loading required package: parallel
```

```
## 
## Attaching package: 'BiocGenerics'
```

```
## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
```

```
## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs
```

```
## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, Map, mapply, match, mget, order, paste,
##     pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,
##     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,
##     tapply, union, unique, unsplit, which, which.max, which.min
```

```
## Loading required package: S4Vectors
```

```
## 
## Attaching package: 'S4Vectors'
```

```
## The following object is masked from 'package:base':
## 
##     expand.grid
```

```
## Loading required package: IRanges
```

```
## Loading required package: GenomeInfoDb
```

```
## Loading required package: Biobase
```

```
## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.
```

```
## Loading required package: DelayedArray
```

```
## Loading required package: matrixStats
```

```
## 
## Attaching package: 'matrixStats'
```

```
## The following objects are masked from 'package:Biobase':
## 
##     anyMissing, rowMedians
```

```
## Loading required package: BiocParallel
```

```
## 
## Attaching package: 'DelayedArray'
```

```
## The following objects are masked from 'package:matrixStats':
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges
```

```
## The following objects are masked from 'package:base':
## 
##     aperm, apply
```

```
## Loading required package: HDF5Array
```

```
## Loading required package: rhdf5
```

```r
TENxPBMCData('pbmc4k')
```

```
## snapshotDate(): 2019-01-04
```

```
## see ?TENxPBMCData and browseVignettes('TENxPBMCData') for documentation
```

```
## downloading 0 resources
```

```
## loading from cache 
##     '/Users/ramezqui//.ExperimentHub/1613'
```

```
## class: SingleCellExperiment 
## dim: 33694 4340 
## metadata(0):
## assays(1): counts
## rownames(33694): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000277475 ENSG00000268674
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(11): Sample Barcode ... Individual Date_published
## reducedDimNames(0):
## spikeNames(0):
```

## Preprocessing

All steps required to preprocess the data into the clean expression matrix, split it up into subsections as necessary. Subsection splits should be informed by the paper (sub)sections. This is all steps prior to the main strategy of interest.

### Cell and gene quality control

Removal of "bad cells", low abundance genes. 

### Normalization

Application of normalization approach, even if its just a log normalization on the count data. Cell cycle normalization may be applicable here.

### Feature selection

Identifying the subset of the clean matrix to work with.

### Dimensionality reduction

Calculating PCs and UMAP/tSNE representation.

## Clustering

Strategy 1 of interest. Use BiocParallel. May need to save intermediate data within package to facilitate fast construction of Rmd.

* SC3
* clusterExperiment
* clustree
* BEARscc # uses ERCC spike ins
* BiocNeighbors # more for devs

## Differential Expression

Strategy 2 of interest. Differential expression approaches. Use one of the outputs from above to perform differential expression analyses comparison.

* zinbwave + edgeR/DESeq2 (may exclude as workflow is very different or highlight as an "advanced section")
* MAST
* scDD
* SC3

## Session Info


