<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 A Basic Analysis | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 A Basic Analysis | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 A Basic Analysis | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Robert A. Amezquita" />
<meta name="author" content="Stephanie C. Hicks" />


<meta name="date" content="2019-06-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="analytical-workflow-overview.html">
<link rel="next" href="adaptations-for-large-scale-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.8</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>6</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>6.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="6.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>6.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="6.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>6.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="6.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>6.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>6.3</b> Data Processing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>6.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="6.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>6.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="6.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>6.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="6.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>6.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="6.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>6.4.1</b> Clustering</a></li>
<li class="chapter" data-level="6.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>6.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="6.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>6.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>6.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="6.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>6.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>7</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="7.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>7.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="7.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>7.3</b> On-Disk Data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="annotation-2.html"><a href="annotation-2.html"><i class="fa fa-check"></i><b>8</b> Annotation</a></li>
<li class="chapter" data-level="9" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>9</b> Clustering</a><ul>
<li class="chapter" data-level="9.1" data-path="clustering-2.html"><a href="clustering-2.html#understanding-the-data"><i class="fa fa-check"></i><b>9.1</b> Understanding the Data</a></li>
<li class="chapter" data-level="9.2" data-path="clustering-2.html"><a href="clustering-2.html#manual-clustering"><i class="fa fa-check"></i><b>9.2</b> Manual Clustering</a><ul>
<li class="chapter" data-level="9.2.1" data-path="clustering-2.html"><a href="clustering-2.html#varying-k"><i class="fa fa-check"></i><b>9.2.1</b> Varying <code>k</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="clustering-2.html"><a href="clustering-2.html#varying-clustering-algorithms"><i class="fa fa-check"></i><b>9.2.2</b> Varying Clustering Algorithms</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="clustering-2.html"><a href="clustering-2.html#automated-clustering"><i class="fa fa-check"></i><b>9.3</b> Automated Clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="differential-expression-2.html"><a href="differential-expression-2.html"><i class="fa fa-check"></i><b>10</b> Differential Expression</a></li>
<li class="chapter" data-level="11" data-path="imputation-1.html"><a href="imputation-1.html"><i class="fa fa-check"></i><b>11</b> Imputation</a></li>
<li class="chapter" data-level="12" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>12</b> Integrating Datasets</a></li>
<li class="chapter" data-level="13" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>13</b> Quality Control</a></li>
<li class="chapter" data-level="14" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>14</b> Sharing and Interactive Interfaces</a></li>
<li class="chapter" data-level="15" data-path="trajectory-analysis-1.html"><a href="trajectory-analysis-1.html"><i class="fa fa-check"></i><b>15</b> Trajectory Analysis</a></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="16" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>16</b> About the Data</a><ul>
<li class="chapter" data-level="16.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>16.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="16.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>16.2</b> Cellbench_data</a></li>
<li class="chapter" data-level="16.3" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>16.3</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>17</b> About the Contributors</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="a-basic-analysis" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> A Basic Analysis</h1>
<p>In this chapter, we will walk through a minimal analysis of a simple scRNA-seq dataset in order to acquaint you with the overall framework of scRNA-seq in code terms.</p>
<p>Where relevant within each part of this basic workflow, we will refer the interested reader to the associated chapter to learn advanced or alternate ways of performing a given task.</p>
<p>Put another way, the workflow demonstrated in this chapter is written with the aim of <em>simplicity</em>, and thus will likely require nontrivial tweaking of parameters or alternate methods in real-world analyses.</p>
<p>One note: in this workflow, we will be loading libraries as they become necessary to clearly link libraries to their respective functions, which usually runs counter to the norm of loading libraries first, at the top of the analysis script.</p>
<div id="preprocessing-import-to-r" class="section level2">
<h2><span class="header-section-number">6.1</span> Preprocessing &amp; Import to R</h2>
<p>We will assume here that sequencing alignment and quantification of the data into a counts matrix, as well as the subsequent import to R has already been performed since this is highly platform- or technology-dependent.</p>
<p>Note that for 10X Genomics data (which is used in this example workflow), the counts matrix and associated metadata (cell barcodes, data path, etc.) can be imported via the <code>DropletUtils</code> package’s <code>read10xCounts()</code> function. For data processed through <code>Salmon</code>/<code>Alevin</code>/<code>Kallisto</code>, we recommend checking out the <code>tximport</code>/<code>tximeta</code> Bioconductor packages. These are either imported as <code>SingleCellExperiment</code> or as a counts matrix which can be then coerced into a <code>SingleCellExperiment</code> object as demonstrated below.</p>
</div>
<div id="constructing-the-singlecellexperiment" class="section level2">
<h2><span class="header-section-number">6.2</span> Constructing the SingleCellExperiment</h2>
<div id="from-scratch" class="section level3">
<h3><span class="header-section-number">6.2.1</span> From Scratch</h3>
<p>Below we show an example of creating a <code>SingleCellExperiment</code> class object from a counts matrix and associated experimental metadata.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)

## More realistic: read in your experimental design metadata
## If its per cell metadata, make sure it lines up with your
## counts matrix row IDs correctly
## my_metadata &lt;- read.csv(&quot;my_metadata.csv&quot;) 

## Example data
ncells &lt;-<span class="st"> </span><span class="dv">100</span>
my_counts_matrix &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">20000</span>, <span class="dv">5</span>), <span class="dt">ncol =</span> ncells)
my_metadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">genotype =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&#39;A&#39;</span>, <span class="st">&#39;B&#39;</span>), <span class="dt">each =</span> <span class="dv">50</span>),
                          <span class="dt">experiment_id =</span> <span class="st">&#39;Experiment1&#39;</span>)

## Construct the sce object manually
sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> my_counts_matrix),
                            <span class="dt">colData =</span> my_metadata)

## Manually adding a variable that is the same across all cells
<span class="kw">colData</span>(sce) &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">colData</span>(sce), <span class="dt">date =</span> <span class="st">&#39;2020-01-01&#39;</span>)

sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 200 100 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(0):
## colnames: NULL
## colData names(3): genotype experiment_id date
## reducedDimNames(0):
## spikeNames(0):</code></pre>
</div>
<div id="from-publicly-available-data" class="section level3">
<h3><span class="header-section-number">6.2.2</span> From Publicly Available Data</h3>
<p>From here on out, we will be working with a small example dataset from the <code>TENxPBMCData</code> Bioconductor package which has already been packaged into a <code>SingleCellExperiment</code> class object:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TENxPBMCData)
sce &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc3k&#39;</span>)

sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 32738 2700 
## metadata(0):
## assays(1): counts
## rownames(32738): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000215616 ENSG00000215611
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(11): Sample Barcode ... Individual Date_published
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>One decision that should be made early on in the analysis is what row identifier to identify genes. Depending on how the data is imported, the <code>rowData</code> component may already have additional annotation information, such as multiple row mappings. For our new <code>sce</code> object from the <code>pbmc3k</code> dataset, we can take a look at <code>rowData</code> to see our options:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowData</span>(sce)</code></pre>
<pre><code>## DataFrame with 32738 rows and 3 columns
##                      ENSEMBL_ID  Symbol_TENx       Symbol
##                     &lt;character&gt;  &lt;character&gt;  &lt;character&gt;
## ENSG00000243485 ENSG00000243485   MIR1302-10           NA
## ENSG00000237613 ENSG00000237613      FAM138A      FAM138A
## ENSG00000186092 ENSG00000186092        OR4F5        OR4F5
## ENSG00000238009 ENSG00000238009 RP11-34P13.7 LOC100996442
## ENSG00000239945 ENSG00000239945 RP11-34P13.8           NA
## ...                         ...          ...          ...
## ENSG00000215635 ENSG00000215635   AC145205.1           NA
## ENSG00000268590 ENSG00000268590        BAGE5           NA
## ENSG00000251180 ENSG00000251180   CU459201.1           NA
## ENSG00000215616 ENSG00000215616   AC002321.2           NA
## ENSG00000215611 ENSG00000215611   AC002321.1           NA</code></pre>
<p>We see that we could choose between <code>ENSEMBL_ID</code> (the default), <code>Symbol_TENx</code>, and <code>Symbol</code>. For ease of readability and subsetting, we will utilize the <code>Symbol_TENx</code> identifier as our object’s rownames, making it possible to subset the <code>sce</code> with gene symbols as in <code>sce[&quot;CD8A&quot;, ]</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">## reassign rownames
<span class="kw">rownames</span>(sce) &lt;-<span class="st"> </span><span class="kw">rowData</span>(sce)[, <span class="st">&quot;Symbol_TENx&quot;</span>]</code></pre>
<p>Now, while this seems to work just fine, eventually we may run into an issue because we actually have duplicated row names here. Depending on how a downstream function is coded, this may cause an esoteric error to pop-up. In fact, here we have about 100 duplicates.</p>
<p>We can avoid future errors (and many headaches) by removing duplicates before any analysis:</p>
<pre class="sourceCode r"><code class="sourceCode r">## counts dupes from top to bottom to make a logical vector
dupes &lt;-<span class="st"> </span><span class="kw">duplicated</span>(<span class="kw">rownames</span>(sce))

sce &lt;-<span class="st"> </span>sce[<span class="op">!</span>dupes, ]</code></pre>
<p>Keep in mind, the above is likely the most inelegant solution to the problem. Other methods could include, from the duplicated set of genes, choosing the one with the highest expression, aggregating the counts per cell, or keeping them all by adding an additional suffix to make the row names unique. Each has its own tradeoffs, so we leave this choice up to the diligent reader.</p>
<p>And one more bit of preprocessing to prevent a potential downstream error is to assign our columns proper names. We can grab the barcodes of each cell from <code>colData</code> and assign them as column names as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(sce) &lt;-<span class="st"> </span>sce<span class="op">$</span>Barcode</code></pre>
</div>
</div>
<div id="data-processing-1" class="section level2">
<h2><span class="header-section-number">6.3</span> Data Processing</h2>
<p>The aim of this section is to form the basis for more interesting downstream analyses. Thus, the objective here is to transform the data into a “clean” expression matrix that has been normalized and freed of technical artifacts, as well as a dimensionality reduction representation that can be used in subsequent analyses and visualization.</p>
<div id="quality-control-metrics-1" class="section level3">
<h3><span class="header-section-number">6.3.1</span> Quality Control Metrics</h3>
<p>The first step is to ensure that our dataset only contains viable cells, e.g. droplets that contain proper mRNA libraries.</p>
<p>One way to do that is to use the popular “knee plot”, which shows the relationship between the log rank vs the log total counts, and then calculate where the “knee” of the plot is. We use the <code>DropletUtils</code> package to demonstrate this in our example PBMC dataset.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DropletUtils)

## Calculate the rank vs total counts per cell
br &lt;-<span class="st"> </span><span class="kw">barcodeRanks</span>(<span class="kw">counts</span>(sce))

## Create the knee plot
<span class="kw">plot</span>(<span class="kw">log10</span>(br<span class="op">$</span>rank), <span class="kw">log10</span>(br<span class="op">$</span>total))
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">log10</span>(<span class="kw">metadata</span>(br)<span class="op">$</span>knee))</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-7"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-7-1.png" alt="Barcode rank (aka knee) plot showing log10-rank by log10-total counts relationship and calculated knee (horizontal line)." width="672" />
<p class="caption">
Figure 6.1: Barcode rank (aka knee) plot showing log10-rank by log10-total counts relationship and calculated knee (horizontal line).
</p>
</div>
<pre class="sourceCode r"><code class="sourceCode r">## Save the calculated knee from `barcodeRanks()`
knee &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">metadata</span>(br)<span class="op">$</span>knee)</code></pre>
<p>We see that the knee calculated via this method (horizontal line) is at 1740, or on the log scale, 3.2405.</p>
<p>This can be used as a filter to remove cells that are likely to be empty droplets. Before we do that, we will finish calculating other quality control (QC) metrics via the <code>scater</code> package and show the results from the first three cells.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)

sce &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce)</code></pre>
<p>We can display some of the calculated QC metrics appended to the <code>colData</code> component - there are a number of other columns present, but for brevity will only show two pertinent ones.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(sce)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&quot;log10_total_features_by_counts&quot;</span>, <span class="st">&quot;log10_total_counts&quot;</span>)]</code></pre>
<pre><code>## DataFrame with 3 rows and 2 columns
##                  log10_total_features_by_counts log10_total_counts
##                                       &lt;numeric&gt;          &lt;numeric&gt;
## AAACATACAACCAC-1                           2.89               3.38
## AAACATTGAGCTAC-1                           3.13               3.69
## AAACATTGATCAGC-1                           3.05                3.5</code></pre>
<p>We can further inspect these cells based on their total counts as well as vs the total features detected by counts (e.g. the number of genes that have nonzero counts).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(sce<span class="op">$</span>log10_total_counts, <span class="dt">breaks =</span> <span class="dv">100</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> knee)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-11"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-11-1.png" alt="Histogram of the log10 total counts with the calculated knee from above (vertical line)." width="672" />
<p class="caption">
Figure 6.2: Histogram of the log10 total counts with the calculated knee from above (vertical line).
</p>
</div>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">smoothScatter</span>(sce<span class="op">$</span>log10_total_counts, sce<span class="op">$</span>log10_total_features_by_counts, <span class="dt">nbin =</span> <span class="dv">250</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> knee)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-12"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-12-1.png" alt="Smoothed scatter plot of the log10-total counts vs the log10-total features detected by counts with the calculated knee from above (vertical line)." width="672" />
<p class="caption">
Figure 6.3: Smoothed scatter plot of the log10-total counts vs the log10-total features detected by counts with the calculated knee from above (vertical line).
</p>
</div>
<p>While there are various ways to filter cells, here we actually will not need to perform any filtering, as the data has already undergone a stringent quality control, and thus all the cells can be considered high quality.</p>
<p>For the sake of completeness, we will demonstrate here - without evaluating - how to subset based on the previously calculated barcode ranks knee:</p>
<pre class="sourceCode r"><code class="sourceCode r">## not run
sce &lt;-<span class="st"> </span>sce[, sce<span class="op">$</span>log10_total_counts <span class="op">&gt;</span><span class="st"> </span>knee]</code></pre>
</div>
<div id="normalizing-data-1" class="section level3">
<h3><span class="header-section-number">6.3.2</span> Normalizing Data</h3>
<p>Next up we will transform the primary data, the counts, into a (log) normalized version. In this section, we will use the <code>scran</code> package throughout.</p>
<p>First however, we will need to calculate scaling factors per cell. This function relies on an initial “quick and dirty” clustering to get roughly similar pools of cells. These are used to generate pool-based estimates, from which the subsequent cell-based size factors are generated. To learn more about the method, see the <code>?computeSumFactors</code> documentation.</p>
<p>For now, we will perform a simpler normalization, using the library sizes per cell to create a log-normalized expression matrix:</p>
<!-- 's `quickCluster()` function to do the initial clustering. Then the (cell) size factors are used to log-normalize the counts data via -->
<!-- ## quick_clusters <- quickCluster(sce, use.ranks = FALSE) -->
<!-- ## sce <- computeSumFactors(sce, clusters = quick_clusters) -->
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">normalize</span>(sce)</code></pre>
<p>We can see below that we now have two assays, <code>counts</code> and <code>logcounts</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assays</span>(sce)</code></pre>
<pre><code>## List of length 2
## names(2): counts logcounts</code></pre>
</div>
<div id="feature-selection-1" class="section level3">
<h3><span class="header-section-number">6.3.3</span> Feature Selection</h3>
<p>This section will use the <code>scran</code> package, as we select for informative genes by selecting for those with high coefficients of biological variation.</p>
<p>Since this experiment does not have spike-ins, we will fit the mean-variance trend across the endogenous genes.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce, <span class="dt">use.spikes =</span> <span class="ot">FALSE</span>)

<span class="kw">plot</span>(fit<span class="op">$</span>mean, fit<span class="op">$</span>var)
<span class="kw">curve</span>(fit<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-16"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-16-1.png" alt="Mean-variance trend line fit by scran package trendVar() function." width="672" />
<p class="caption">
Figure 6.4: Mean-variance trend line fit by scran package trendVar() function.
</p>
</div>
<p>We can see that the trend line goes through the central mass of genes, and thus continue on with looking at the decomposed variance. In this method, it is assumed that the total variance is the sum of the technical and biological variance, where the technical variance can be determined by interpolating the fitted trend at the mean log-count for that gene. Thus the biological variance is the total variance minus this interpolated (technical) variance.</p>
<p>We can then rank and choose genes which have a biological coefficient of variance greater than zero.</p>
<pre class="sourceCode r"><code class="sourceCode r">dec &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce, fit)
dec &lt;-<span class="st"> </span>dec[<span class="kw">order</span>(dec<span class="op">$</span>bio, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>), ] <span class="co"># order by bio var</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r">dec[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]</code></pre>
<pre><code>##          mean total   bio  tech p.value FDR
## LYZ     1.629 3.925 3.116 0.809       0   0
## S100A9  1.068 3.352 2.628 0.725       0   0
## HLA-DRA 1.543 3.071 2.278 0.793       0   0
## FTL     3.637 2.941 2.203 0.738       0   0
## CD74    2.209 2.696 1.900 0.796       0   0</code></pre>
<p>The total number of genes with biological variance greater than zero as 3770.</p>
<p>Alternatively, we could use the p-value/FDR as a way to rank our genes, but do note the following (from the <a href="https://bioconductor.org/packages/release/workflows/vignettes/simpleSingleCell/inst/doc/var.html#23_testing_for_significantly_positive_biological_components"><code>simpleSingleCell</code> vignette</a>:</p>
<blockquote>
<p>“Ranking based on p-value tends to prioritize HVGs that are more likely to be true positives but, at the same time, less likely to be interesting. This is because the ratio can be very large for HVGs that have very low total variance and do not contribute much to the cell-cell heterogeneity.”</p>
</blockquote>
<p>However we choose, we can save these highly variable genes and use them for subsequent analyses:</p>
<pre class="sourceCode r"><code class="sourceCode r">hvg_genes &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec)[dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</code></pre>
<p>For the purpose of sharing and saving this list of genes, we can stash the result into the <code>metadata</code> component of our <code>sce</code> object as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes &lt;-<span class="st"> </span>hvg_genes
<span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</code></pre>
<pre><code>##  [1] &quot;LYZ&quot;     &quot;S100A9&quot;  &quot;HLA-DRA&quot; &quot;FTL&quot;     &quot;CD74&quot;    &quot;CST3&quot;    &quot;S100A8&quot; 
##  [8] &quot;TYROBP&quot;  &quot;NKG7&quot;    &quot;FTH1&quot;</code></pre>
<p>The <code>metadata</code> component can hold any object, as it is a list container. Any results that you’d like to keep are safe to store here, and a great way to save or share intermediate results that would otherwise be kept in separate objects.</p>
</div>
<div id="dimensionality-reduction-1" class="section level3">
<h3><span class="header-section-number">6.3.4</span> Dimensionality Reduction</h3>
<p>We now can perform dimensionality reduction using our highly variable genes (<code>hvg_genes</code>) subset. To do this, we will first calculate the PCA representation via the <code>runPCA()</code> function from the <code>scater</code> package. We will calculate 50 components on our highly variable genes:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents =</span> <span class="dv">50</span>,
              <span class="dt">feature_set =</span> hvg_genes)</code></pre>
<p>The results of these calculations will be stored in the <code>reducedDims</code> component. This method saves the percent variance explained per component as an attribute, which can be accessed as follows, and subsequently plot the “elbow plot”:</p>
<pre class="sourceCode r"><code class="sourceCode r">## access the attribute where percentVar is saved in reducedDim
pct_var_explained &lt;-<span class="st"> </span><span class="kw">attr</span>(<span class="kw">reducedDim</span>(sce, <span class="st">&#39;PCA&#39;</span>), <span class="st">&#39;percentVar&#39;</span>)

<span class="kw">plot</span>(pct_var_explained) <span class="co"># elbow plot</span></code></pre>
<p><img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<p>To calculate a 2-dimensional representation of the data, we will use the top 20 components of our PCA result to compute the UMAP representation.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&#39;PCA&#39;</span>, <span class="dt">n_dimred =</span> <span class="dv">20</span>)

<span class="kw">plotUMAP</span>(sce)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-24"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-24-1.png" alt="UMAP plot." width="672" />
<p class="caption">
Figure 6.5: UMAP plot.
</p>
</div>
<p>With that, we have a canvas on which to paint our downstream analyses.</p>
</div>
</div>
<div id="downstream-statistical-analyses-1" class="section level2">
<h2><span class="header-section-number">6.4</span> Downstream Statistical Analyses</h2>
<p>There are a plethora of potential downstream analyses to run, the choice of which is highly dependent on the biological objective. For this example dataset, our aim will be to identify the key cell types via a combination of clustering and differential expression.</p>
<div id="clustering-1" class="section level3">
<h3><span class="header-section-number">6.4.1</span> Clustering</h3>
<p>Based on our earlier UMAP plot, it appears that we have a few distinct clusters. To do this computationally, we can utilize the <code>scran</code> package to:</p>
<ul>
<li>build a shared nearest neighbor (SNN) graph</li>
<li>calculate based on the SNN graph the most representative clusters</li>
</ul>
<p>In this first step, we will specify that we will consider <code>k</code> nearest neighbors, and <code>d</code> dimensions from the PCA calculation as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1234</span>) <span class="co"># to make results reproducible</span>
snng &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce, <span class="dt">k =</span> <span class="dv">50</span>, <span class="dt">d =</span> <span class="dv">20</span>)</code></pre>
<p>Following the graph construction, we can calculate the clusters using a variety of different graph-based methods from the <code>igraph</code> package. Here, we use the louvain method to determine our cell’s cluster memberships.</p>
<pre class="sourceCode r"><code class="sourceCode r">snng_clusters &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(snng)</code></pre>
<p>We see that we have the following numbers of cells per cluster:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(snng_clusters<span class="op">$</span>membership)</code></pre>
<pre><code>## 
##   1   2   3   4   5 
## 687 350 556 528 579</code></pre>
<p>To view this result graphically on the UMAP plot, we first assign the result to the <code>colData</code> component as a new column, and specify this as our color variable in the <code>plotUMAP()</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(sce)<span class="op">$</span>clusters &lt;-<span class="st"> </span><span class="kw">as.factor</span>(snng_clusters<span class="op">$</span>membership)
<span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;clusters&#39;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-28"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-28-1.png" alt="UMAP plot showing calculated clusters." width="672" />
<p class="caption">
Figure 6.6: UMAP plot showing calculated clusters.
</p>
</div>
<p>Naturally, this result will change as we tweak the number of <code>k</code> neighbors to consider and with the specific clustering algorithm, but for now we will go onwards to find markers of each of our clusters.</p>
</div>
<div id="differential-expression-1" class="section level3">
<h3><span class="header-section-number">6.4.2</span> Differential Expression</h3>
<p>In this section, we will look to identify genes that are unique to each of our clusters. To accomplish this, we will lean on the <code>scran</code> package to perform the analysis, and then the <code>scater</code> package to visualize the results.</p>
<p>For this analysis, we will limit ourselves to a top subset of highly variable genes in our <code>hvg_genes</code> set, purely for the sake of computation time. Furthermore, we will limit our consideration to genes with an <em>increased</em> log fold-change of at least 1.5 versus other clusters. We will also use the <code>BiocParallel</code> package to parallelize the computation and speed up our processing via the <code>BPPARAM</code> argument.</p>
<pre class="sourceCode r"><code class="sourceCode r">markers &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce, <span class="dt">clusters =</span> <span class="kw">colData</span>(sce)<span class="op">$</span>clusters,
                       <span class="dt">subset.row =</span> hvg_genes[<span class="dv">1</span><span class="op">:</span><span class="dv">250</span>],
                       <span class="dt">lfc =</span> <span class="fl">1.5</span>, <span class="dt">direction =</span> <span class="st">&#39;up&#39;</span>, <span class="dt">log.p =</span> <span class="ot">TRUE</span>, 
                       <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())</code></pre>
<p>We can view the top 5 markers that are differentially expressed (by our specified metrics):</p>
<pre class="sourceCode r"><code class="sourceCode r">markers[[<span class="dv">1</span>]][<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]</code></pre>
<pre><code>##        Top log.p.value log.FDR logFC.2 logFC.3 logFC.4 logFC.5
## CST3     1      -831.4  -825.9   3.446   3.470   3.413   3.480
## TYROBP   1      -778.7  -773.8   3.325   3.324   2.691   3.301
## LYZ      2      -526.9  -522.8   4.028   4.011   4.072   4.000
## FTL      3      -667.3  -662.9   3.110   3.634   3.463   3.534
## FTH1     4      -474.3  -470.6   2.771   3.118   3.100   2.755</code></pre>
<p>We can see that CD3D, a marker of T cells, is one of our top differentially expressed genes in cluster 1. We can plot the expression of this gene across all our clusters as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotExpression</span>(sce, <span class="st">&#39;CD3D&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;clusters&#39;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-32"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-32-1.png" alt="Violin plots of CD3D expression across clusters." width="672" />
<p class="caption">
Figure 6.7: Violin plots of CD3D expression across clusters.
</p>
</div>
<p>This plot highlights that CD3D is more highly expressed in cluster 1 relative to <em>some</em> of the other clusters, but not all. This can also be seen from our raw output above, where the log fold-change is calculated with respect to each cluster. There, we see that the log fold-change for CD3D is very high only relative to clusters 2 and 3 (meeting our cutoff of 1.5).</p>
</div>
<div id="annotation-1" class="section level3">
<h3><span class="header-section-number">6.4.3</span> Annotation</h3>
<div id="a-manual-approach" class="section level4">
<h4><span class="header-section-number">6.4.3.1</span> A Manual Approach</h4>
<p>To finish off our the downstream analysis section here, we will look to annotate our clusters with a cell type designation, based on publicly available knowledge.</p>
<p>Before we do that, let’s get a broader view of our top differentially expressed genes. To do this, we can iterate over the list-object returned by <code>findMarkers</code> to get the top 10 genes per cluster, and then plot these genes in a heatmap.</p>
<pre class="sourceCode r"><code class="sourceCode r">## grab the top 10 genes per cluster (e.g. within each list component)
genes &lt;-<span class="st"> </span><span class="kw">lapply</span>(markers, <span class="cf">function</span>(x) {
    <span class="kw">rownames</span>(x)[x<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">10</span>]
})

## uniqify the set of genes returned, after coercing to a vector
genes &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">unlist</span>(genes))

<span class="kw">plotHeatmap</span>(sce, genes,
            <span class="dt">colour_columns_by =</span> <span class="st">&quot;clusters&quot;</span>,
            <span class="dt">show_colnames =</span> <span class="ot">FALSE</span>,
            <span class="dt">clustering_method =</span> <span class="st">&#39;ward.D2&#39;</span>,
            <span class="dt">fontsize_row =</span> <span class="dv">6</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-33"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-33-1.png" alt="Heatmap showing top differentially expressed genes across the clusters." width="672" />
<p class="caption">
Figure 6.8: Heatmap showing top differentially expressed genes across the clusters.
</p>
</div>
<p>Based on the heatmap output (and <em>a priori</em> knowledge), we can make some observations:</p>
<ul>
<li>CD79A/CD79B, markers of B cells, are uniquely and highly expressed in cluster 2</li>
<li>HLA genes, present on antigen presenting cells (APCs), are highly expressed across clusters 2 and 3</li>
<li>LYZ, a marker of dendritic cells (an APC), is highly expressed in cluster 3</li>
<li>Granzymes A and B (GZMA/GZMB), and NKG7, markers of cytotoxic cells such as CD8s and NK cells, are highly expressed within (a subset of cluster 4)</li>
<li>CD3D/CD3E, markers of T cells, are expressed across clusters 5, 1, and 4</li>
</ul>
<p>Finally, we can view a selection of the genes mentioned above on our previous UMAP plot:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&quot;CD79A&quot;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-341"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-34-1.png" alt="Various UMAP plots showing the expression of select cell-type specific genes." width="672" />
<p class="caption">
Figure 6.9: Various UMAP plots showing the expression of select cell-type specific genes.
</p>
</div>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&quot;LYZ&quot;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-342"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-34-2.png" alt="Various UMAP plots showing the expression of select cell-type specific genes." width="672" />
<p class="caption">
Figure 6.10: Various UMAP plots showing the expression of select cell-type specific genes.
</p>
</div>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&quot;NKG7&quot;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-343"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-34-3.png" alt="Various UMAP plots showing the expression of select cell-type specific genes." width="672" />
<p class="caption">
Figure 6.11: Various UMAP plots showing the expression of select cell-type specific genes.
</p>
</div>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&quot;CD3D&quot;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-344"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-34-4.png" alt="Various UMAP plots showing the expression of select cell-type specific genes." width="672" />
<p class="caption">
Figure 6.12: Various UMAP plots showing the expression of select cell-type specific genes.
</p>
</div>
<p>Combining the information derived from our heatmap and viewing these genes on our UMAP, we can come to the following conclusion:</p>
<ul>
<li>Cluster 2 is likely to be B cells</li>
<li>Cluster 3 is likely to be dendritic cells</li>
<li>Clusters 1, 5, 4 appear to represent a spectrum of cells with cytotoxic capabilities, likely composed of a combination of T cells and NK cells,</li>
<li>Cluster 4 exhibits an strong NK cell signature on the basis of NKG7</li>
</ul>
<p>Now that we’ve manually sorted our dataset on the basis of prior knowledge, let’s try a more automated approach using publicly available markers.</p>
</div>
<div id="an-automated-approach" class="section level4">
<h4><span class="header-section-number">6.4.3.2</span> An Automated Approach</h4>
<p>Manually classifying cell types present in an scRNA-seq experiment can be prone to bias in terms of how a label is selected. Thus have emerged automated classification approaches which take a measured approach to the labeling of cell types.</p>
<p>One such approach - <code>cellassign</code> - applies labels in a single-cell manner based on a gene by cell type “marker matrix”. Here, we utilize an existing gene by cell type annotation from a publication by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1070-5">Becht et al. (2016)</a> which categorizes genes into cell types based on the specificity of their expression.</p>
<p>Let’s first construct a marker matrix loosely inspired by the <a href="https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html">Seurat PBMC 3k tutorial</a>:</p>
<pre class="sourceCode r"><code class="sourceCode r">anno &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
    <span class="dt">SYMBOL =</span> <span class="kw">c</span>(
        <span class="st">&#39;IL7R&#39;</span>, <span class="st">&#39;CCR7&#39;</span>, <span class="st">&#39;CD4&#39;</span>, <span class="st">&#39;CD3D&#39;</span>, <span class="st">&#39;CD3E&#39;</span>,
        <span class="st">&#39;CD14&#39;</span>, <span class="st">&#39;LYZ&#39;</span>,
        <span class="st">&#39;MS4A1&#39;</span>, <span class="st">&#39;CD79A&#39;</span>, <span class="st">&#39;CD79B&#39;</span>,
        <span class="st">&#39;CD8A&#39;</span>, <span class="st">&#39;CD8B&#39;</span>, <span class="st">&#39;CD3D&#39;</span>, <span class="st">&#39;CD3E&#39;</span>,
        <span class="st">&#39;GNLY&#39;</span>, <span class="st">&#39;NKG7&#39;</span>,
        <span class="st">&#39;FCER1A&#39;</span>, <span class="st">&#39;CST3&#39;</span>, <span class="st">&#39;ITGAX&#39;</span>
    ),
    <span class="dt">cell_type =</span> <span class="kw">c</span>(
        <span class="kw">rep</span>(<span class="st">&#39;CD4 T cell&#39;</span>, <span class="dv">5</span>),
        <span class="kw">rep</span>(<span class="st">&#39;Monocyte&#39;</span>, <span class="dv">2</span>),
        <span class="kw">rep</span>(<span class="st">&#39;B cell&#39;</span>, <span class="dv">3</span>),
        <span class="kw">rep</span>(<span class="st">&#39;CD8 T cell&#39;</span>, <span class="dv">4</span>),
        <span class="kw">rep</span>(<span class="st">&#39;NK cell&#39;</span>, <span class="dv">2</span>),
        <span class="kw">rep</span>(<span class="st">&#39;Dendritic cell&#39;</span>, <span class="dv">3</span>)
    )
)</code></pre>
<p>Lastly, we’ll need to reformat this matrix to fit the expectations of <code>cellassign</code>, chiefly to convert the annotation into a binary matrix of genes (rows) by cell types (columns):</p>
<pre class="sourceCode r"><code class="sourceCode r">## construct rho (binary marker matrix)
tmp &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">spread</span>(anno, cell_type, cell_type)
rho &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(tmp[, <span class="dv">-1</span>]), <span class="dv">0</span>, <span class="dv">1</span>)
<span class="kw">rownames</span>(rho) &lt;-<span class="st"> </span>tmp<span class="op">$</span>SYMBOL

## remove entries that are not present in our dataset
rho &lt;-<span class="st"> </span>rho[<span class="kw">rownames</span>(rho) <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(sce), ]

rho[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, ]</code></pre>
<pre><code>##      B cell CD4 T cell CD8 T cell Dendritic cell Monocyte NK cell
## CCR7      0          1          0              0        0       0
## CD14      0          0          0              0        1       0
## CD3D      0          1          1              0        0       0</code></pre>
<p>We can then run the <code>cellassign</code> method to produce cell type labels on a per cell basis:</p>
<pre class="sourceCode r"><code class="sourceCode r">## not run - results pulled from a prior run
## devtools::install_github(&#39;Irrationone/cellassign&#39;)
<span class="kw">library</span>(cellassign)
<span class="kw">library</span>(tensorflow)

<span class="kw">set.seed</span>(<span class="dv">1234</span>)
reticulate<span class="op">::</span><span class="kw">py_set_seed</span>(<span class="dv">1234</span>)
fit &lt;-<span class="st"> </span><span class="kw">cellassign</span>(sce[<span class="kw">rownames</span>(rho), ],
                  <span class="dt">marker_gene_info =</span> rho,
                  <span class="dt">s =</span> <span class="kw">sizeFactors</span>(sce))

## add cell type info into colData
<span class="kw">colData</span>(sce)<span class="op">$</span>cellassign_type &lt;-<span class="st"> </span>fit<span class="op">$</span>cell_type</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## plot the cellassign results on UMAP
<span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;cellassign_type&#39;</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-40"></span>
<img src="P2_W1.basic-analysis_files/figure-html/unnamed-chunk-40-1.png" alt="UMAP showing the results of automated label assignment as performed by cellassign." width="672" />
<p class="caption">
Figure 6.13: UMAP showing the results of automated label assignment as performed by cellassign.
</p>
</div>
<p>In practice, some combination of the above manual and automated classification schema will likely be necessary to properly annotate an scRNA-seq dataset.</p>
</div>
</div>
</div>
<div id="accessible-reproducible-analysis-1" class="section level2">
<h2><span class="header-section-number">6.5</span> Accessible &amp; Reproducible Analysis</h2>
<p>In collaborative settings, it is essential to share data and analyses. Thanks to the <code>SingleCellExperiment</code> class, most of if not all analysis steps performed can be recorded. These outputs are accessible through not only R, but also via graphical user interfaces as well that broaden the potential viewing audience.</p>
<div id="interactive-data-visualization-1" class="section level3">
<h3><span class="header-section-number">6.5.1</span> Interactive Data Visualization</h3>
<p>Interactive exploration and visualization is a great way for collaborators to learn more about scRNA-seq data and analyses. In particular the <code>iSEE</code> package has been especially designed for viewing and sharing scRNA-seq.</p>
<pre class="sourceCode r"><code class="sourceCode r">## not run
<span class="kw">library</span>(iSEE)
<span class="kw">iSEE</span>(sce)</code></pre>
<p>Based on the example analyses, we task the interested reader to assess the previous section’s automatic annotation relative to the clustering results using <code>iSEE</code>.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="analytical-workflow-overview.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="adaptations-for-large-scale-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W1.basic-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

</body>

</html>
