<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Adaptations for Large-scale Data | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Adaptations for Large-scale Data | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Adaptations for Large-scale Data | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Robert A. Amezquita" />
<meta name="author" content="Stephanie C. Hicks" />


<meta name="date" content="2019-06-21" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="a-basic-analysis.html">
<link rel="next" href="annotation-2.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.8</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>6</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>6.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="6.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>6.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="6.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>6.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="6.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>6.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>6.3</b> Data Processing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>6.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="6.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>6.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="6.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>6.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="6.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>6.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="6.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>6.4.1</b> Clustering</a></li>
<li class="chapter" data-level="6.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>6.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="6.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>6.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>6.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="6.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>6.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>7</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="7.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>7.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="7.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>7.3</b> On-Disk Data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="annotation-2.html"><a href="annotation-2.html"><i class="fa fa-check"></i><b>8</b> Annotation</a></li>
<li class="chapter" data-level="9" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>9</b> Clustering</a><ul>
<li class="chapter" data-level="9.1" data-path="clustering-2.html"><a href="clustering-2.html#understanding-the-data"><i class="fa fa-check"></i><b>9.1</b> Understanding the Data</a></li>
<li class="chapter" data-level="9.2" data-path="clustering-2.html"><a href="clustering-2.html#manual-clustering"><i class="fa fa-check"></i><b>9.2</b> Manual Clustering</a><ul>
<li class="chapter" data-level="9.2.1" data-path="clustering-2.html"><a href="clustering-2.html#varying-k"><i class="fa fa-check"></i><b>9.2.1</b> Varying <code>k</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="clustering-2.html"><a href="clustering-2.html#varying-clustering-algorithms"><i class="fa fa-check"></i><b>9.2.2</b> Varying Clustering Algorithms</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="clustering-2.html"><a href="clustering-2.html#automated-clustering"><i class="fa fa-check"></i><b>9.3</b> Automated Clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="differential-expression-2.html"><a href="differential-expression-2.html"><i class="fa fa-check"></i><b>10</b> Differential Expression</a></li>
<li class="chapter" data-level="11" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>11</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="11.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preprocessing-1"><i class="fa fa-check"></i><b>11.1</b> Preprocessing</a></li>
<li class="chapter" data-level="11.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#fast-mnn"><i class="fa fa-check"></i><b>11.2</b> Fast MNN</a></li>
<li class="chapter" data-level="11.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#harmony"><i class="fa fa-check"></i><b>11.3</b> Harmony</a></li>
<li class="chapter" data-level="11.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#advanced-mnn-workflow"><i class="fa fa-check"></i><b>11.4</b> Advanced MNN Workflow</a><ul>
<li class="chapter" data-level="11.4.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#feature-selection-2"><i class="fa fa-check"></i><b>11.4.1</b> Feature Selection</a></li>
<li class="chapter" data-level="11.4.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>11.4.2</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#naive-method-without-correction"><i class="fa fa-check"></i><b>11.5</b> Naive Method Without Correction</a></li>
<li class="chapter" data-level="11.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#limma-batch-correction"><i class="fa fa-check"></i><b>11.6</b> Limma Batch Correction</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>12</b> Quality Control</a></li>
<li class="chapter" data-level="13" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>13</b> Sharing and Interactive Interfaces</a></li>
<li class="chapter" data-level="14" data-path="trajectory-analysis-1.html"><a href="trajectory-analysis-1.html"><i class="fa fa-check"></i><b>14</b> Trajectory Analysis</a></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>15</b> About the Data</a><ul>
<li class="chapter" data-level="15.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>15.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="15.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>15.2</b> Cellbench_data</a></li>
<li class="chapter" data-level="15.3" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>15.3</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>16</b> About the Contributors</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="adaptations-for-large-scale-data" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Adaptations for Large-scale Data</h1>
<p>Large datasets such as the Human Cell Atlas (with over 1.3 million cells) have the potential to benefit from special adaptations that enable analysis in compute-constrained environments (such as personal laptops). Here, we briefly cover topics that aim to ease working with scRNA-seq data to make it faster and more tractable.</p>
<p>Do note however that these adaptations do not universally result in improved computational efficiency. For example, parallelization does incur costs, and disk-backed data representation will generally be slower than purely in-memory representations. In both cases, hardware specifics may dictate any potential gains in efficiency, as can be imagined in the case of a solid state drive (SSD) being faster for disk-backed data representations relative to a hard disk drive (HDD).</p>
<p>Thus, with the right compute environment, data, and task, these adaptations can yield significant computational improvements.</p>
<div id="approximate-methods" class="section level2">
<h2><span class="header-section-number">7.1</span> Approximate Methods</h2>
<p>A general approach that works across all scales of data involves changing the task itself. Some methods - such as PCA or nearest neighbor searches - have been extended to include versions that provide approximate results. Generally, these adaptations result in acceptable losses in accuracy for significant computational gains. In some cases, approximate methods may even be desirable for the results themselves, as has been shown by the FIt-SNE approach.</p>
<p>Some example packages that provide approximate versions of popular methods include:</p>
<ul>
<li><em>BiocSingular</em> via the <code>IrlbaParam()</code> and <code>RandomParam()</code> for approximate singular value decomposition (SVD)</li>
<li><em>BiocNeighbors</em> via the <code>AnnoyParam()</code> for approximate nearest neighbor searches</li>
</ul>
<p>These packages provide users (and developers) a common interface, enabling modular swapping of key algorithms within functions. For example, we can see the immediate benefit of using an approximate method for PCA as provided by the <code>BSPARAM</code> argument in the following code, which utilizes the <em>bench</em> package for profiling:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scater)

## Simulate a dataset with 1k genes and 1k cells
mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="fl">1e6</span>, <span class="dv">100</span>), <span class="dt">nrow =</span> <span class="dv">1000</span>)
tiny_sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> mat))
tiny_sce &lt;-<span class="st"> </span><span class="kw">normalize</span>(tiny_sce)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bench)
<span class="kw">library</span>(BiocSingular)

## simple function to show only cols of interest
.show_bench &lt;-<span class="st"> </span><span class="cf">function</span>(b) {
    b[, <span class="kw">c</span>(<span class="st">&#39;expression&#39;</span>, <span class="st">&#39;min&#39;</span>, <span class="st">&#39;median&#39;</span>, <span class="st">&#39;mem_alloc&#39;</span>)]
}

bm &lt;-<span class="st"> </span>bench<span class="op">::</span><span class="kw">mark</span>(
           <span class="kw">runPCA</span>(tiny_sce, <span class="dt">BSPARAM =</span> <span class="kw">IrlbaParam</span>()),
           <span class="kw">runPCA</span>(tiny_sce, <span class="dt">BSPARAM =</span> <span class="kw">ExactParam</span>()), 
           <span class="dt">check =</span> <span class="ot">FALSE</span>)

<span class="kw">.show_bench</span>(bm)</code></pre>
<pre><code>## # A tibble: 2 x 4
##   expression                                    min   median mem_alloc
##   &lt;bch:expr&gt;                               &lt;bch:tm&gt; &lt;bch:tm&gt; &lt;bch:byt&gt;
## 1 runPCA(tiny_sce, BSPARAM = IrlbaParam())   41.8ms   42.2ms        0B
## 2 runPCA(tiny_sce, BSPARAM = ExactParam())    613ms    613ms        0B</code></pre>
<p>We can see that the approximate, irlba based implementation is about 6 times faster than the exact version in this case.</p>
<p>We can also try this with nearest neighbor searches. Here we provide the <code>BNPARAM</code> argument to build a shared nearest neighbors graph using different algorithms under the hood:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
<span class="kw">library</span>(BiocNeighbors)

## Calculate PCA before testing NN back-ends
tiny_sce_pca &lt;-<span class="st"> </span><span class="kw">runPCA</span>(tiny_sce, <span class="dt">BSPARAM =</span> <span class="kw">IrlbaParam</span>())

bm &lt;-<span class="st"> </span>bench<span class="op">::</span><span class="kw">mark</span>(
           <span class="kw">buildSNNGraph</span>(tiny_sce_pca, <span class="dt">BNPARAM =</span> <span class="kw">AnnoyParam</span>()),
           <span class="kw">buildSNNGraph</span>(tiny_sce_pca, <span class="dt">BNPARAM =</span> <span class="kw">KmknnParam</span>()),
           <span class="dt">check =</span> <span class="ot">FALSE</span>)

<span class="kw">.show_bench</span>(bm)</code></pre>
<pre><code>## # A tibble: 2 x 4
##   expression                                              min  median
##   &lt;bch:expr&gt;                                          &lt;bch:t&gt; &lt;bch:t&gt;
## 1 buildSNNGraph(tiny_sce_pca, BNPARAM = AnnoyParam()) 679.1ms 679.1ms
## 2 buildSNNGraph(tiny_sce_pca, BNPARAM = KmknnParam())   1.41s   1.41s
## # … with 1 more variable: mem_alloc &lt;bch:byt&gt;</code></pre>
<p>We can see from the above benchmark that in our tiny dataset, we don’t see much, if any, benefit of using an approximate method (via the <code>AnnoyParam()</code>, which uses the Annoy library).</p>
<p>However, if we increase our dataset to something larger..</p>
<pre class="sourceCode r"><code class="sourceCode r">## Simulate a dataset with 1k genes and 10k cells
mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="fl">10e6</span>, <span class="dv">100</span>), <span class="dt">nrow =</span> <span class="dv">1000</span>)
big_sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> mat))
big_sce &lt;-<span class="st"> </span><span class="kw">normalize</span>(big_sce)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Calculate PCA before testing NN back-ends
big_sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(big_sce, <span class="dt">BSPARAM =</span> <span class="kw">IrlbaParam</span>())

## NN search
bm &lt;-<span class="st"> </span>bench<span class="op">::</span><span class="kw">mark</span>(
           <span class="kw">buildSNNGraph</span>(big_sce, <span class="dt">BNPARAM =</span> <span class="kw">AnnoyParam</span>()),
           <span class="kw">buildSNNGraph</span>(big_sce, <span class="dt">BNPARAM =</span> <span class="kw">KmknnParam</span>()),
           <span class="dt">check =</span> <span class="ot">FALSE</span>)

<span class="kw">.show_bench</span>(bm)</code></pre>
<pre><code>## # A tibble: 2 x 4
##   expression                                          min  median mem_alloc
##   &lt;bch:expr&gt;                                     &lt;bch:tm&gt; &lt;bch:t&gt; &lt;bch:byt&gt;
## 1 buildSNNGraph(big_sce, BNPARAM = AnnoyParam())    5.94s   5.94s        0B
## 2 buildSNNGraph(big_sce, BNPARAM = KmknnParam())   11.41s  11.41s        0B</code></pre>
<p>We see (more consistently) that we have realized some gains from using the approximate-based nearest neighbors search.</p>
</div>
<div id="parallelization" class="section level2">
<h2><span class="header-section-number">7.2</span> Parallelization</h2>
<p>Many tasks that arise in the analysis of scRNA-seq data are able to be parallelized. In other words, the problem can be broken down into smaller pieces that can be solved independently. Parallel computing can be performed in computing environments with access to multiple cores.</p>
<p>Bioconductor has reduced the complexity of implementing parallelized software via the <a href="https://bioconductor.org/packages/BiocParallel"><em>BiocParallel</em></a> package. This enables a common interface across Bioconductor software packages for parallel computing.</p>
<p>Across many Bioconductor packages, an argument that will often be present in functions that are parallelizable is the <code>BPPARAM</code> argument. Here, it is possible to specify the parallel back-end that can be used for evaluation.</p>
<p>Again, note that parallelization does incur an overhead cost in splitting up the data, sending it off, and combining the results, and thus your mileage may vary depending on the specifics.</p>
<p>Below is an example usage of the <em>BiocParallel</em> library supplying <code>BPPARAM</code> argument for constructing the UMAP representation from the PCA results:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BiocParallel)

bm &lt;-<span class="st"> </span>bench<span class="op">::</span><span class="kw">mark</span>(
           <span class="kw">runUMAP</span>(tiny_sce_pca, <span class="dt">BPPARAM =</span> <span class="kw">SerialParam</span>()),           
           <span class="kw">runUMAP</span>(tiny_sce_pca, <span class="dt">BPPARAM =</span> <span class="kw">MulticoreParam</span>()),
           <span class="dt">check =</span> <span class="ot">FALSE</span>)

<span class="kw">.show_bench</span>(bm)</code></pre>
<pre><code>## # A tibble: 2 x 4
##   expression                                           min median mem_alloc
##   &lt;bch:expr&gt;                                        &lt;bch:&gt; &lt;bch:&gt; &lt;bch:byt&gt;
## 1 runUMAP(tiny_sce_pca, BPPARAM = SerialParam())      3.7s   3.7s        0B
## 2 runUMAP(tiny_sce_pca, BPPARAM = MulticoreParam())  4.15s  4.15s        0B</code></pre>
<p>On this dataset and equipment, there’s not much, if any, benefit to parallelization, but your mileage may vary dependent on those aspects.</p>
</div>
<div id="on-disk-data" class="section level2">
<h2><span class="header-section-number">7.3</span> On-Disk Data</h2>
<p>The matrix of data from a single-cell experiment can be on the order of tens to hundreds of gigabytes, depending on the number of features and cells measured. A standard approach to import and represent these matrices has been to load the entire data set into memory using either the <em>matrix</em> object in base R or sparse and dense matrix classes from the <em>Matrix</em> R package. This approach however may prove intractable in computing environments with limited memory.</p>
<p>Disk-backed representations such as HDF5 free us from having to load an entire dataset into memory, and thus make it possible to work with large-scale scRNA-seq data.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rhdf5)
<span class="kw">library</span>(HDF5Array)</code></pre>
<!-- ## Notes on Data Representation -->
<!-- However, the explosion of data from single-cell assays has led to alternative data representations that are compatible with high-performance C++ code used for computationally intensive tasks in existing Bioconductor packages. This enables memory-efficient data manipulation and operations in R. For example, the *beachmat* Bioconductor package is a C++ interface for accessing single-cell data that is interoperable with sparse, dense and file-backed matrices, such as the HDF5 file format, which allows users to only load a subset of the data into memory at a time. In addition, Bioconductor has developed the infrastructure to read and write HDF5 files from R using the *rhdf5* package, to efficiently work with data in HDF5 files using array-like containers in the *HDF5Array*, *DelayedArray*, and *DelayedMatrixStats* packages. Much of this infrastructure was motivated by previous work done in the context of flow cytometry and whole genome bisulfite sequencing. -->
<!-- At a low-level, the main interface between HDF5 and Bioconductor is implemented in the packages `rhdf5`, which provides read/write functionalities, `Rhdf5lib`, which provides C and C++ HDF5 libraries, and `beachmat`, which provides a consistent C++ class interface for a variety of commonly used matrix types, including sparse and HDF5-backed matrices. These packages are useful for developers that want to develop methods able to interact with HDF5 data sets. -->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="a-basic-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="annotation-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_WX.adaptations-for-large-scale-data.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

</body>

</html>
