<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Clustering | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Clustering | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Clustering | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-07-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="normalization.html">
<link rel="next" href="marker-gene-detection.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="6.3" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i><b>6.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i><b>7.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#with-adaptive-thresholds"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i><b>8.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#why-log-transform"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="9.8" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i><b>9.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="clustering-2.html"><a href="clustering-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="clustering-2.html"><a href="clustering-2.html#comments-on-truth"><i class="fa fa-check"></i><b>10.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="10.3" data-path="clustering-2.html"><a href="clustering-2.html#graph-based-clustering"><i class="fa fa-check"></i><b>10.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="10.3.1" data-path="clustering-2.html"><a href="clustering-2.html#background-1"><i class="fa fa-check"></i><b>10.3.1</b> Background</a></li>
<li class="chapter" data-level="10.3.2" data-path="clustering-2.html"><a href="clustering-2.html#implementation"><i class="fa fa-check"></i><b>10.3.2</b> Implementation</a></li>
<li class="chapter" data-level="10.3.3" data-path="clustering-2.html"><a href="clustering-2.html#other-parameters"><i class="fa fa-check"></i><b>10.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="10.3.4" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>10.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering-2.html"><a href="clustering-2.html#k-means-clustering"><i class="fa fa-check"></i><b>10.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="10.4.1" data-path="clustering-2.html"><a href="clustering-2.html#background-2"><i class="fa fa-check"></i><b>10.4.1</b> Background</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering-2.html"><a href="clustering-2.html#base-implementation"><i class="fa fa-check"></i><b>10.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="10.4.3" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>10.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering-2.html"><a href="clustering-2.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="10.5.1" data-path="clustering-2.html"><a href="clustering-2.html#background-3"><i class="fa fa-check"></i><b>10.5.1</b> Background</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering-2.html"><a href="clustering-2.html#implementation-1"><i class="fa fa-check"></i><b>10.5.2</b> Implementation</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>10.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering-2.html"><a href="clustering-2.html#session-info-4"><i class="fa fa-check"></i><b>10.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>11</b> Marker gene detection</a><ul>
<li class="chapter" data-level="11.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-3"><i class="fa fa-check"></i><b>11.1</b> Motivation</a></li>
<li class="chapter" data-level="11.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>11.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="11.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>11.2.1</b> Standard application</a></li>
<li class="chapter" data-level="11.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>11.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="11.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>11.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>11.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="11.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>11.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="11.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>11.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="11.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>11.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>11.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="11.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>11.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="11.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>11.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="11.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>11.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="11.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>11.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="11.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>11.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-5"><i class="fa fa-check"></i><b>11.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>12</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="12.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preprocessing-1"><i class="fa fa-check"></i><b>12.1</b> Preprocessing</a></li>
<li class="chapter" data-level="12.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#fast-mnn"><i class="fa fa-check"></i><b>12.2</b> Fast MNN</a></li>
<li class="chapter" data-level="12.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#harmony"><i class="fa fa-check"></i><b>12.3</b> Harmony</a></li>
<li class="chapter" data-level="12.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#advanced-mnn-workflow"><i class="fa fa-check"></i><b>12.4</b> Advanced MNN Workflow</a><ul>
<li class="chapter" data-level="12.4.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>12.4.1</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#naive-method-without-correction"><i class="fa fa-check"></i><b>12.5</b> Naive Method Without Correction</a></li>
<li class="chapter" data-level="12.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#limma-batch-correction"><i class="fa fa-check"></i><b>12.6</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="12.7" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-6"><i class="fa fa-check"></i><b>12.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>13</b> Sharing and Interactive Interfaces</a><ul>
<li class="chapter" data-level="13.1" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html#session-info-7"><i class="fa fa-check"></i><b>13.1</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>14</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="14.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>14.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="14.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>14.2</b> Parallelization</a></li>
<li class="chapter" data-level="14.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>14.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="14.4" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-8"><i class="fa fa-check"></i><b>14.4</b> Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>15</b> About the Data</a><ul>
<li class="chapter" data-level="15.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>15.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="15.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>15.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>16</b> About the Contributors</a><ul>
<li class="chapter" data-level="16.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>16.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="16.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>16.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="16.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>16.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="16.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>16.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>17</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="clustering-2" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Clustering</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-2" class="section level2">
<h2><span class="header-section-number">10.1</span> Motivation</h2>
<p>Clustering is an unsupervised learning procedure that is used in scRNA-seq data analysis to empirically define groups of cells with similar expression profiles.
Its primary purpose is to summarize the data in a digestible format for human interpretation.
This allows us to describe population heterogeneity in terms of discrete labels that are easily understood, rather than attempting to comprehend the high-dimensional manifold on which the cells truly reside.
After annotation based on marker genes, the clusters can be treated as proxies for more abstract biological concepts such as cell types or states.
Clustering is thus a critical step for extracting biological insights from scRNA-seq data.
Here, we demonstrate the application of several commonly used methods with the 10X PBMC dataset.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb201-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb201-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb201-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="st">&quot;raw_data&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb201-4" data-line-number="4">raw.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</a>
<a class="sourceLine" id="cb201-5" data-line-number="5">    <span class="st">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</a>
<a class="sourceLine" id="cb201-6" data-line-number="6"><span class="kw">untar</span>(raw.path, <span class="dt">exdir=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k&quot;</span>))</a>
<a class="sourceLine" id="cb201-7" data-line-number="7"></a>
<a class="sourceLine" id="cb201-8" data-line-number="8"><span class="kw">library</span>(DropletUtils)</a>
<a class="sourceLine" id="cb201-9" data-line-number="9">fname &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</a>
<a class="sourceLine" id="cb201-10" data-line-number="10">sce.pbmc &lt;-<span class="st"> </span><span class="kw">read10xCounts</span>(fname, <span class="dt">col.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb201-11" data-line-number="11"></a>
<a class="sourceLine" id="cb201-12" data-line-number="12">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb201-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb201-14" data-line-number="14"><span class="kw">rownames</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb201-15" data-line-number="15"></a>
<a class="sourceLine" id="cb201-16" data-line-number="16"><span class="kw">library</span>(EnsDb.Hsapiens.v86)</a>
<a class="sourceLine" id="cb201-17" data-line-number="17">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(EnsDb.Hsapiens.v86, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb201-18" data-line-number="18">    <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb201-19" data-line-number="19"></a>
<a class="sourceLine" id="cb201-20" data-line-number="20">### cell-detection <span class="al">###</span></a>
<a class="sourceLine" id="cb201-21" data-line-number="21"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb201-22" data-line-number="22">e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb201-23" data-line-number="23">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="kw">which</span>(e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>)]</a>
<a class="sourceLine" id="cb201-24" data-line-number="24"></a>
<a class="sourceLine" id="cb201-25" data-line-number="25">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb201-26" data-line-number="26">sce.pbmc &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce.pbmc, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(location<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb201-27" data-line-number="27">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.pbmc<span class="op">$</span>pct_counts_Mito, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb201-28" data-line-number="28">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb201-29" data-line-number="29"></a>
<a class="sourceLine" id="cb201-30" data-line-number="30">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb201-31" data-line-number="31"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb201-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb201-33" data-line-number="33">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb201-34" data-line-number="34">sce.pbmc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.pbmc, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">cluster=</span>clusters)</a>
<a class="sourceLine" id="cb201-35" data-line-number="35">sce.pbmc &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb201-36" data-line-number="36"></a>
<a class="sourceLine" id="cb201-37" data-line-number="37">### feature-selection <span class="al">###</span></a>
<a class="sourceLine" id="cb201-38" data-line-number="38">fit.pbmc &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce.pbmc, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb201-39" data-line-number="39">dec.pbmc &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(<span class="dt">fit=</span>fit.pbmc)</a>
<a class="sourceLine" id="cb201-40" data-line-number="40">o &lt;-<span class="st"> </span><span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb201-41" data-line-number="41">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[<span class="kw">head</span>(o, <span class="dv">2000</span>)]</a>
<a class="sourceLine" id="cb201-42" data-line-number="42"></a>
<a class="sourceLine" id="cb201-43" data-line-number="43">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb201-44" data-line-number="44"><span class="kw">set.seed</span>(<span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb201-45" data-line-number="45">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce.pbmc, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">ncomponents=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb201-46" data-line-number="46">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>())</a>
<a class="sourceLine" id="cb201-47" data-line-number="47"></a>
<a class="sourceLine" id="cb201-48" data-line-number="48"><span class="kw">set.seed</span>(<span class="dv">100000</span>)</a>
<a class="sourceLine" id="cb201-49" data-line-number="49">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.pbmc, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb201-50" data-line-number="50"></a>
<a class="sourceLine" id="cb201-51" data-line-number="51"><span class="kw">set.seed</span>(<span class="dv">1000000</span>)</a>
<a class="sourceLine" id="cb201-52" data-line-number="52">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce.pbmc, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb202-1" data-line-number="1">sce.pbmc</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 3922 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
## rowData names(10): ID Symbol ... total_counts log10_total_counts
## colnames(3922): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCACAGGTCCAC-1 TTTGTCATCCCAAGAT-1
## colData names(38): Sample Barcode ...
##   pct_counts_in_top_200_features_Mito
##   pct_counts_in_top_500_features_Mito
## reducedDimNames(3): PCA TSNE UMAP
## spikeNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="comments-on-truth" class="section level2">
<h2><span class="header-section-number">10.2</span> Comments on Truth</h2>
<p>At this point<a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a>, it is worth stressing the distinction between clusters and cell types.
The former is an empirical construct while the latter is a biological truth (albeit a vaguely defined one).
For this reason, questions like “what is the true number of clusters?” are usually meaningless.
We can define as many clusters as we like, with whatever algorithm we like - each clustering will represent its own partitioning of the high-dimensional expression space, and is as “real” as any other clustering.</p>
<p>A more relevant question is “how well do the clusters approximate the cell types?”
Unfortunately, this is difficult to answer given the context-dependent interpretation of biological truth.
Some analysts will be satisfied with resolution of the major cell types; other analysts may want resolution of subtypes; and others still may require resolution of different states (e.g., metabolic activity, stress) within those subtypes.
Two clusterings can also be highly inconsistent yet both valid, simply partitioning the cells based on different aspects of biology.
Indeed, asking for an unqualified “best” clustering is akin to asking for the best magnification on a microscope without any context.</p>
<p>It is helpful to realize that clustering, like a microscope, is simply a tool to explore the data.
We can zoom in and out by changing the resolution of the clustering parameters, and we can experiment with different clustering algorithms to obtain alternative perspectives of the data.
This iterative approach is entirely permissible for data exploration, which constitutes the majority of all scRNA-seq data analysis.
Now, if we were operating in a formal hypothesis testing framework, this parameter fiddling would constitute data dredging and be penalized by the multiple testing correction - but we can afford to be more relaxed for exploratory data analysis, given that we did not have a firm quantitative hypothesis in the first place<a href="#fn11" class="footnote-ref" id="fnref11"><sup>11</sup></a>!</p>
</div>
<div id="graph-based-clustering" class="section level2">
<h2><span class="header-section-number">10.3</span> Graph-based clustering</h2>
<div id="background-1" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Background</h3>
<p>Popularized by its use in <em><a href="https://CRAN.R-project.org/package=Seurat">Seurat</a></em>, graph-based clustering is a flexible and scalable technique for clustering large scRNA-seq datasets.
We first build a graph where each node is a cell that is connected to its nearest neighbours in the high-dimensional space.
Edges are weighted based on the similarity between the cells involved, with higher weight given to cells that are more closely related.
We then apply algorithms to identify “communities” of cells that are more connected to cells in the same community than they are to cells of different communities.
Each community represents a cluster that we can use for downstream interpretation.</p>
<p>The major advantage of graph-based clustering lies in its scalability.
It only requires a <span class="math inline">\(k\)</span>-nearest neighbor search that can be done in log-linear time (on average), in contrast to hierachical clustering methods that are quadratic with respect to the number of cells.
Graph construction avoids making strong assumptions about the shape of the clusters or the distribution of cells within each cluster, compared to other methods like <span class="math inline">\(k\)</span>-means (that favor spherical clusters) or Gaussian mixture models (that require normality).
From a practical perspective, each cell is forcibly connected to a minimum number of neighboring cells, which reduces the risk of generating many uninformative clusters consisting of one or two outlier cells.</p>
<p>The main drawback of graph-based methods is that, after graph construction, no information is retained about relationships beyond the neighbouring cells<a href="#fn12" class="footnote-ref" id="fnref12"><sup>12</sup></a>.
This has some practical consequences in datasets that exhibit differences in cell density, as more steps through the graph are required to move the same distance through a region of higher cell density.
From the perspective of community detection algorithms, this effect “inflates” the high-density regions such that any internal substructure or noise is more likely to cause formation of subclusters.
The resolution of clustering thus becomes dependent on the density of cells, which can occasionally be misleading if overstates the heterogeneity in the data.</p>
</div>
<div id="implementation" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Implementation</h3>
<p>There are several considerations in the practical execution of a graph-based clustering method:</p>
<ul>
<li>How many neighbors are considered when constructing the graph.</li>
<li>What scheme is used to weight the edges.</li>
<li>Which community detection algorithm is used to define the clusters.</li>
</ul>
<p>For example, the following code uses the 10 nearest neighbors of each cell to construct a shared nearest neighbor graph.
Two cells are connected by an edge if any of their nearest neighbors are shared,
with the edge weight defined from the highest average rank of the shared neighbors <span class="citation">(Xu and Su <a href="#ref-xu2015identification">2015</a>)</span>.
The Walktrap method from the <em><a href="https://CRAN.R-project.org/package=igraph">igraph</a></em> package is then used to identify communities.
All calculations are performed using the top PCs to take advantage of data compression and denoising.</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb204-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb204-2" data-line-number="2">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb204-3" data-line-number="3">clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb204-4" data-line-number="4"><span class="kw">table</span>(clust)</a></code></pre></div>
<pre><code>## clust
##   1   2   3   4   5   6   7   8   9  10  11  12 
## 792 519 568 549 117 202  89 943  55  35  17  36</code></pre>
<p>We assign the cluster assignments back into our <code>SingleCellExperiment</code> object as a factor in the column metadata.
This allows us to conveniently visualize the distribution of clusters in a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="clustering-2.html#fig:tsne-clust-graph">10.1</a>).</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb206-1" data-line-number="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb206-2" data-line-number="2">sce.pbmc<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)</a>
<a class="sourceLine" id="cb206-3" data-line-number="3"><span class="kw">plotReducedDim</span>(sce.pbmc, <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-clust-graph"></span>
<img src="P2_W06.clustering_files/figure-html/tsne-clust-graph-1.png" alt="$t$-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from graph-based clustering." width="672" />
<p class="caption">
Figure 10.1: <span class="math inline">\(t\)</span>-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from graph-based clustering.
</p>
</div>
<p>The most important parameter is <code>k</code>, i.e., the number of nearest neighbors used to construct the graph.
This controls the resolution of the clustering, with higher <code>k</code> yielding a more inter-connected graph and larger clusters.
Users can exploit this by experimenting with different values of <code>k</code> to obtain a satisfactory resolution.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" data-line-number="1"><span class="co"># More resolved.</span></a>
<a class="sourceLine" id="cb207-2" data-line-number="2">g<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">k=</span><span class="dv">5</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb207-3" data-line-number="3">clust<span class="fl">.5</span> &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g<span class="fl">.5</span>)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb207-4" data-line-number="4"><span class="kw">table</span>(clust<span class="fl">.5</span>)</a></code></pre></div>
<pre><code>## clust.5
##   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16 
## 123 798  26 559 505 200  91 505 986  27  50   9  10   9   7  17</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb209-1" data-line-number="1"><span class="co"># Less resolved.</span></a>
<a class="sourceLine" id="cb209-2" data-line-number="2">g<span class="fl">.50</span> &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">k=</span><span class="dv">50</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb209-3" data-line-number="3">clust<span class="fl">.50</span> &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g<span class="fl">.50</span>)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb209-4" data-line-number="4"><span class="kw">table</span>(clust<span class="fl">.50</span>)</a></code></pre></div>
<pre><code>## clust.50
##    1    2    3    4    5    6    7    8 
##  181  839  572  116  173  518  488 1035</code></pre>
</div>
<div id="other-parameters" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Other parameters</h3>
<p>Further tweaking can be done by changing the weighting scheme.
Setting <code>type=&quot;number&quot;</code> will weight edges based on the number of nearest neighbors that are shared between two cells.
Similarly, <code>type=&quot;jaccard&quot;</code> will weight edges according to the Jaccard index of the two sets of neighbors.
We can also disable weighting altogether by using <code>buildKNNGraph()</code>, which is occasionally useful for downstream graph operations that do not support weights.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb211-1" data-line-number="1">g.num &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">type=</span><span class="st">&quot;number&quot;</span>)</a>
<a class="sourceLine" id="cb211-2" data-line-number="2">g.jaccard &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">type=</span><span class="st">&quot;jaccard&quot;</span>)</a>
<a class="sourceLine" id="cb211-3" data-line-number="3">g.none &lt;-<span class="st"> </span><span class="kw">buildKNNGraph</span>(sce.pbmc, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)</a></code></pre></div>
<p>All of these <code>g</code> variables are <code>graph</code> objects from the <em><a href="https://CRAN.R-project.org/package=igraph">igraph</a></em> package.
They can be used with any of the community detection algorithms provided by <em><a href="https://CRAN.R-project.org/package=igraph">igraph</a></em>.
We have already mentioned the Walktrap approach, but many others are available to choose from:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb212-1" data-line-number="1">clust.louvain &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb212-2" data-line-number="2">clust.infomap &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_infomap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb212-3" data-line-number="3">clust.fast &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_fast_greedy</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb212-4" data-line-number="4">clust.labprop &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_label_prop</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb212-5" data-line-number="5">clust.eigen &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_leading_eigen</span>(g)<span class="op">$</span>membership</a></code></pre></div>
<p>It is then straightforward to compare two clustering strategies to see how they differ.
For example, the results below suggest that Louvain is similar to Walktrap;
fast-greedy yields coarser clusters; and Infomap provides higher resolution.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb213-1" data-line-number="1"><span class="kw">table</span>(<span class="dt">Louvain=</span>clust.louvain, <span class="dt">Walktrap=</span>clust)</a></code></pre></div>
<pre><code>##        Walktrap
## Louvain   1   2   3   4   5   6   7   8   9  10  11  12
##      1    0   0   0   0   0   0   0   0   0   0   0  36
##      2    0   0   0   0   0 187   0   0   0   0   0   0
##      3    0   0 546   2   0  15   0   0   0   0   0   0
##      4    0   0   8  23   0   0   0 943   0   0   0   0
##      5   32   0   0   1   0   0  89   0   0  35   0   0
##      6    0 519   0   0   0   0   0   0   0   0   0   0
##      7  760   0   0   0   4   0   0   0   1   0   0   0
##      8    0   0   0   0 113   0   0   0   0   0   0   0
##      9    0   0   0   0   0   0   0   0   0   0  17   0
##      10   0   0  13   8   0   0   0   0  54   0   0   0
##      11   0   0   1 515   0   0   0   0   0   0   0   0</code></pre>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb215-1" data-line-number="1"><span class="kw">table</span>(<span class="dt">Infomap=</span>clust.infomap, <span class="dt">Walktrap=</span>clust)</a></code></pre></div>
<pre><code>##        Walktrap
## Infomap   1   2   3   4   5   6   7   8   9  10  11  12
##      1    0   0   2  24   0   0   0 943   0   0   0   0
##      2  483   0   0   0   0   0   0   0   0   0   0   0
##      3    0   0   1 375   0   0   0   0   0   0   0   0
##      4    0 295   0   0   0   0   0   0   0   0   0   0
##      5    0   0 254   0   0   2   0   0   0   0   0   0
##      6  246   0   0   0   0   0   0   0   0   0   0   0
##      7    0 211   0   0   0   0   0   0   0   0   0   0
##      8    0   0   0   0   0 187   0   0   0   0   0   0
##      9    0   0 168   4   0   0   0   0   0   0   0   0
##      10   0   0 120   0   0  13   0   0   0   0   0   0
##      11   0   0   0 137   0   0   0   0   0   0   0   0
##      12   0   0   0   0  94   0   0   0   0   0   0   0
##      13   0   0   0   0   0   0  89   0   0   0   0   0
##      14   0   0   0   0   0   0   0   0  55   0   0   0
##      15   0   0   0   0   0   0   0   0   0   0   0  36
##      16  20   0   0   0  23   0   0   0   0   0   0   0
##  [ reached getOption(&quot;max.print&quot;) -- omitted 6 rows ]</code></pre>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb217-1" data-line-number="1"><span class="kw">table</span>(<span class="dt">Fast=</span>clust.fast, <span class="dt">Walktrap=</span>clust)</a></code></pre></div>
<pre><code>##     Walktrap
## Fast   1   2   3   4   5   6   7   8   9  10  11  12
##    1   0   0   0   4   0   0   0 880   0   0   0   0
##    2   0   0   0   0  88   0   0   0   0   0   0   0
##    3   0   0   0   0   0   0   0   0   0   0  17   0
##    4   3   3 568 545   0 202   0  63  55   1   0   0
##    5 789   0   0   0  29   0  89   0   0  34   0   0
##    6   0   0   0   0   0   0   0   0   0   0   0  36
##    7   0 516   0   0   0   0   0   0   0   0   0   0</code></pre>
<p>Pipelines involving <em><a href="https://bioconductor.org/packages/3.10/scran">scran</a></em> default to rank-based weights followed by Walktrap clustering.
In contrast, <em><a href="https://CRAN.R-project.org/package=Seurat">Seurat</a></em> uses Jaccard-based weights followed by Louvain clustering.
Both of these strategies work well, and it is likely that the same could be said for many other combinations of weighting schemes and community detection algorithms<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a>.</p>
</div>
<div id="assessing-cluster-separation" class="section level3">
<h3><span class="header-section-number">10.3.4</span> Assessing cluster separation</h3>
<p>When dealing with graphs, the modularity is a natural metric for evaluating the separation between communities/clusters.
This is defined as the (scaled) difference between the observed total weight of edges between nodes in the same cluster and the expected total weight if edge weights were randomly distributed across all pairs of nodes.
Larger modularity values indicate that there most edges occur within clusters, suggesting that the clusters are sufficiently well separated to avoid edges forming between neighboring cells in different clusters.</p>
<p>The standard approach is to report a single modularity value for a clustering on a given graph.
This is useful for comparing different clusterings on the same graph - and indeed, some community detection algorithms are designed with the aim of maximizing the modularity - but it is less helpful for interpreting a given clustering.
Rather, we use the <code>clusterModularity()</code> function, which returns modularity scores for each cluster and cluster pair.
More precisely, when <code>get.values=TRUE</code>, it returns the observed and expected sum of weights that are used to compute the standard modularity score.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb219-1" data-line-number="1">mod &lt;-<span class="st"> </span><span class="kw">clusterModularity</span>(g, clust, <span class="dt">get.values=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb219-2" data-line-number="2"><span class="kw">names</span>(mod)</a></code></pre></div>
<pre><code>## [1] &quot;observed&quot; &quot;expected&quot;</code></pre>
<p>In each matrix, each row/column corresponds to a cluster, and each entry of the matrix contains the total weight of edges between cells in the respective clusters.
A dataset containing well-separated clusters should contain most of the observed total weight on the diagonal entries, i.e., most edges occur between cells in the same cluster.
We visualize this by computing the log-ratio of observed to expected weights (Figure <a href="clustering-2.html#fig:cluster-mod">10.2</a>).
(We use the log-ratio instead of the difference as the latter’s scale depends on the number of cells in each cluster, meaning that any visualization would be dominated by large differences for large clusters.)</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb221-1" data-line-number="1">ratio &lt;-<span class="st"> </span>mod<span class="op">$</span>observed<span class="op">/</span>mod<span class="op">$</span>expected</a>
<a class="sourceLine" id="cb221-2" data-line-number="2"></a>
<a class="sourceLine" id="cb221-3" data-line-number="3"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb221-4" data-line-number="4"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(ratio<span class="op">+</span><span class="dv">1</span>), <span class="dt">cluster_rows=</span><span class="ot">FALSE</span>, <span class="dt">cluster_cols=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb221-5" data-line-number="5">    <span class="dt">color=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>))(<span class="dv">100</span>))</a></code></pre></div>
<div class="figure"><span id="fig:cluster-mod"></span>
<img src="P2_W06.clustering_files/figure-html/cluster-mod-1.png" alt="Heatmap of the log~2~-ratio of the total weight between nodes in the same cluster or in different clusters, relative to the total weight expected under a null model of random links." width="672" />
<p class="caption">
Figure 10.2: Heatmap of the log<sub>2</sub>-ratio of the total weight between nodes in the same cluster or in different clusters, relative to the total weight expected under a null model of random links.
</p>
</div>
<p>Figure <a href="clustering-2.html#fig:cluster-mod">10.2</a> shows that the weight is mostly concentrated on the diagonal.
The few off-diagonal contributions represent closely related clusters with (relatively) many inter-connecting edges.
This is not a problem, as poorer separation is to be expected when resolution increases.
In fact, this can assist interpretation by knowing which clusters are more closely related to each other.</p>
<p>One useful approach is to use the <code>ratio</code> matrix to form another graph where the nodes are clusters rather than cells.
Edges between nodes are weighted according to the ratio of observed to expected edge weights between cells in those clusters.
We can then repeat our graph operations on this new cluster-level graph.
For example, we could obtain clusters of clusters, or we could simply create a new cluster-based layout for visualization (Figure <a href="clustering-2.html#fig:cluster-graph">10.3</a>).
This is analogous to the “graph abstraction” approach described by <span class="citation">Wolf et al. (<a href="#ref-wolf2017graph">2017</a>)</span>.</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb222-1" data-line-number="1">cluster.gr &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_adjacency_matrix</span>(ratio, </a>
<a class="sourceLine" id="cb222-2" data-line-number="2">    <span class="dt">mode=</span><span class="st">&quot;undirected&quot;</span>, <span class="dt">weighted=</span><span class="ot">TRUE</span>, <span class="dt">diag=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb222-3" data-line-number="3"><span class="kw">plot</span>(cluster.gr, <span class="dt">edge.width=</span>igraph<span class="op">::</span><span class="kw">E</span>(cluster.gr)<span class="op">$</span>weight<span class="op">*</span><span class="dv">10</span>)  </a></code></pre></div>
<div class="figure"><span id="fig:cluster-graph"></span>
<img src="P2_W06.clustering_files/figure-html/cluster-graph-1.png" alt="Force-directed layout showing the relationships between clusters based on the ratio of observed to expected total weights between nodes in different clusters. The thickness of the edge between a pair of clusters is proportional to the corresponding ratio." width="672" />
<p class="caption">
Figure 10.3: Force-directed layout showing the relationships between clusters based on the ratio of observed to expected total weights between nodes in different clusters. The thickness of the edge between a pair of clusters is proportional to the corresponding ratio.
</p>
</div>
<p>Incidentally, some readers may have noticed that all <em><a href="https://CRAN.R-project.org/package=igraph">igraph</a></em> commands were prefixed with <code>igraph::</code>.
We have done this deliberately to avoid bringing <code>igraph::normalize</code> into the global namespace.
Rather unfortunately, this <code>normalize</code> function accepts any argument and returns <code>NULL</code>, which causes difficult-to-diagnose bugs when it overwrites our intended <code>normalize</code> from <em><a href="https://bioconductor.org/packages/3.10/scater">scater</a></em>.</p>
</div>
</div>
<div id="k-means-clustering" class="section level2">
<h2><span class="header-section-number">10.4</span> <span class="math inline">\(k\)</span>-means clustering</h2>
<div id="background-2" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Background</h3>
<p><span class="math inline">\(k\)</span>-means clustering is a classic technique that aims to partition cells into <span class="math inline">\(k\)</span> clusters.
Each cell is assigned to the cluster with the closest centroid, which is done by minimizing the within-cluster sum of squares using a random starting configuration for the <span class="math inline">\(k\)</span> centroids.
The main advantage of this approach lies in its speed, given the simplicity and ease of implementation of the algorithm.
However, it has a number of serious shortcomings:</p>
<ul>
<li>It implicitly favours spherical clusters of equal radius.
This can result in inappropriate partitionings on real datasets that contain groupings with irregular sizes and shapes.</li>
<li>The number of clusters <span class="math inline">\(k\)</span> must be specified beforehand.
This strongly affects the algorithm such that an inappropriate choice will inevitably lead to a poor clustering.
(For example, setting <span class="math inline">\(k\)</span> to be below the number of cell types will always lead to co-clustering of two cell types, regardless of how well separated they are.
In contrast, other methods like graph-based clustering will respect strong separation even if the resolution is set to a low value.)</li>
<li>It is dependent on the randomly chosen initial coordinates.
This requires multiple runs to verify that the clustering is stable.</li>
</ul>
<p>These issues reduce the appeal of <span class="math inline">\(k\)</span>-means clustering for scRNA-seq data analysis.
Despite this, it is still used widely as it is fast and - frankly - often good enough.</p>
<p>Even if we were to forgo <span class="math inline">\(k\)</span>-means as our clustering method of choice, it is still one of the best approaches for sample-based data compression.
In this application, we set <span class="math inline">\(k\)</span> to a large value such as the square root of the number of cells to obtain fine-grained clusters.
These are not meant to be interpreted directly, but rather, the centroids are treated as “samples” for further analyses.
The idea here is to obtain a single representative of each region of the expression space, reducing the number of samples and computational work in later steps like, e.g., trajectory reconstruction.
This approach will also eliminate differences in cell density across the expression space, ensuring that the most abundant cell type does not dominate downstream results.</p>
</div>
<div id="base-implementation" class="section level3">
<h3><span class="header-section-number">10.4.2</span> Base implementation</h3>
<p>Base R provides the <code>kmeans()</code> function that does as its name suggests.
We call this on our top PCs to obtain a clustering for a specified number of clusters in the <code>centers=</code> argument.
This requires a random seed to ensure that the results are reproducible.
In general, the <span class="math inline">\(k\)</span>-means clusters correspond to the visual clusters on the <span class="math inline">\(t\)</span>-SNE plot in Figure <a href="clustering-2.html#fig:tsne-clust-kmeans">10.4</a>, though there are some divergences that are not apparent in, e.g., Figure <a href="clustering-2.html#fig:tsne-clust-graph">10.1</a>.
(This is at least partially due to the fact that <span class="math inline">\(t\)</span>-SNE is itself graph-based and so will naturally agree more with a graph-based clustering strategy.)</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb223-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb223-2" data-line-number="2">clust.kmeans &lt;-<span class="st"> </span><span class="kw">kmeans</span>(<span class="kw">reducedDim</span>(sce.pbmc, <span class="st">&quot;PCA&quot;</span>), <span class="dt">centers=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb223-3" data-line-number="3"><span class="kw">table</span>(clust.kmeans<span class="op">$</span>cluster)</a></code></pre></div>
<pre><code>## 
##    1    2    3    4    5    6    7    8    9   10 
##  458  207   23  517    1  568   36 1033 1057   22</code></pre>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb225-1" data-line-number="1">sce.pbmc<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust.kmeans<span class="op">$</span>cluster)</a>
<a class="sourceLine" id="cb225-2" data-line-number="2"><span class="kw">plotReducedDim</span>(sce.pbmc, <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-clust-kmeans"></span>
<img src="P2_W06.clustering_files/figure-html/tsne-clust-kmeans-1.png" alt="$t$-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from $k$-means clustering." width="672" />
<p class="caption">
Figure 10.4: <span class="math inline">\(t\)</span>-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from <span class="math inline">\(k\)</span>-means clustering.
</p>
</div>
<p>We obtain a “reasonable” choice of <span class="math inline">\(k\)</span> by computing the gap statistic using methods from the <em><a href="https://CRAN.R-project.org/package=cluster">cluster</a></em> package.
This is the log-ratio of the expected to observed within-cluster sum of squares, where the expected value is computed by randomly distributingn cells throughout high-dimensional space using the bounding box of the original data.
A larger gap statistic represents a lower observed sum of squares - and thus better clustering - compared to a population with no structure.
Ideally, we would choose the <span class="math inline">\(k\)</span> that maximizes the gap statistic, but this is often unhelpful as the tendency of <span class="math inline">\(k\)</span>-means to favour spherical clusters results in a large choice <span class="math inline">\(k\)</span> to capture different cluster shapes.
Instead, we choose the most parsimonious <span class="math inline">\(k\)</span> beyond which the increases in the gap statistic are considered insignificant (Figure <a href="#fig:kmeans-gap"><strong>??</strong></a>).</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb226-1" data-line-number="1"><span class="kw">library</span>(cluster)</a>
<a class="sourceLine" id="cb226-2" data-line-number="2">gaps &lt;-<span class="st"> </span><span class="kw">clusGap</span>(<span class="kw">reducedDim</span>(sce.pbmc, <span class="st">&quot;PCA&quot;</span>), kmeans, <span class="dt">K.max=</span><span class="dv">20</span>)</a>
<a class="sourceLine" id="cb226-3" data-line-number="3">best.k &lt;-<span class="st"> </span><span class="kw">maxSE</span>(gaps<span class="op">$</span>Tab[,<span class="st">&quot;gap&quot;</span>], gaps<span class="op">$</span>Tab[,<span class="st">&quot;SE.sim&quot;</span>])</a>
<a class="sourceLine" id="cb226-4" data-line-number="4">best.k</a></code></pre></div>
<pre><code>## [1] 9</code></pre>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb228-1" data-line-number="1"><span class="kw">plot</span>(gaps<span class="op">$</span>Tab[,<span class="st">&quot;gap&quot;</span>], <span class="dt">xlab=</span><span class="st">&quot;Number of clusters&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Gap statistic&quot;</span>)</a>
<a class="sourceLine" id="cb228-2" data-line-number="2"><span class="kw">abline</span>(<span class="dt">v=</span>best.k, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:kmeans-gp"></span>
<img src="P2_W06.clustering_files/figure-html/kmeans-gp-1.png" alt="Gap statistic with respect to increasing number of $k$-means clusters in the 10X PBMC dataset. The red line represents the chosen $k$." width="672" />
<p class="caption">
Figure 10.5: Gap statistic with respect to increasing number of <span class="math inline">\(k\)</span>-means clusters in the 10X PBMC dataset. The red line represents the chosen <span class="math inline">\(k\)</span>.
</p>
</div>
<p>It is then straightforwad to repeat the clustering with the <code>best.k</code> (Figure <a href="clustering-2.html#fig:tsne-clust-kmeans-best">10.6</a>.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb229-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb229-2" data-line-number="2">clust.kmeans2 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(<span class="kw">reducedDim</span>(sce.pbmc, <span class="st">&quot;PCA&quot;</span>), <span class="dt">centers=</span>best.k)</a>
<a class="sourceLine" id="cb229-3" data-line-number="3"><span class="kw">table</span>(clust.kmeans2<span class="op">$</span>cluster)</a></code></pre></div>
<pre><code>## 
##    1    2    3    4    5    6    7    8    9 
##  458  207   24  517   22  568   36 1033 1057</code></pre>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb231-1" data-line-number="1">sce.pbmc<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust.kmeans2<span class="op">$</span>cluster)</a>
<a class="sourceLine" id="cb231-2" data-line-number="2"><span class="kw">plotReducedDim</span>(sce.pbmc, <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-clust-kmeans-best"></span>
<img src="P2_W06.clustering_files/figure-html/tsne-clust-kmeans-best-1.png" alt="$t$-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from $k$-means clustering. Here, $k$ is chosen based on the gap statistic." width="672" />
<p class="caption">
Figure 10.6: <span class="math inline">\(t\)</span>-SNE plot of the 10X PBMC dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from <span class="math inline">\(k\)</span>-means clustering. Here, <span class="math inline">\(k\)</span> is chosen based on the gap statistic.
</p>
</div>
</div>
<div id="assessing-cluster-separation-1" class="section level3">
<h3><span class="header-section-number">10.4.3</span> Assessing cluster separation</h3>
<p>It is informative to examine the within-cluster sum of squares for each cluster.
Clusters are more likely to have low sum of squares if they have no internal structure and are separated from other clusters (such that there are not many cells on the boundaries between clusters, which results in a higher sum of squares from the centroid).
As an aside, clusters with the largest sum of squares do not appear to be the most dispersed in Figure <a href="#fig:pbmc-kmeans-best"><strong>??</strong></a>, which highlights the risks of attempting to quantitatively interpret <span class="math inline">\(t\)</span>-SNE plots.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb232-1" data-line-number="1">tab &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">WithinSS=</span>clust.kmeans2<span class="op">$</span>withinss, </a>
<a class="sourceLine" id="cb232-2" data-line-number="2">    <span class="dt">NCells=</span><span class="kw">tabulate</span>(clust.kmeans2<span class="op">$</span>cluster))</a>
<a class="sourceLine" id="cb232-3" data-line-number="3">tab</a></code></pre></div>
<pre><code>##   WithinSS NCells
## 1    11113    458
## 2     5715    207
## 3   291364     24
## 4    15365    517
## 5   276270     22
## 6    18199    568
## 7     1533     36
## 8    15074   1033
## 9    73975   1057</code></pre>
<p>To explore the relationships between <span class="math inline">\(k\)</span>-means clusters, a natural approach is to compute distances between their centroids.
This directly lends itself to visualization as a tree after hierarchical clustering (Figure <a href="clustering-2.html#fig:kmeans-tree">10.7</a>).</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb234-1" data-line-number="1">cent.tree &lt;-<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(clust.kmeans2<span class="op">$</span>centers), <span class="st">&quot;ward.D2&quot;</span>)</a>
<a class="sourceLine" id="cb234-2" data-line-number="2"><span class="kw">plot</span>(cent.tree)</a></code></pre></div>
<div class="figure"><span id="fig:kmeans-tree"></span>
<img src="P2_W06.clustering_files/figure-html/kmeans-tree-1.png" alt="Hierarchy of $k$-means cluster centroids, using Ward's minimum variance method." width="672" />
<p class="caption">
Figure 10.7: Hierarchy of <span class="math inline">\(k\)</span>-means cluster centroids, using Ward’s minimum variance method.
</p>
</div>
</div>
</div>
<div id="hierarchical-clustering" class="section level2">
<h2><span class="header-section-number">10.5</span> Hierarchical clustering</h2>
<div id="background-3" class="section level3">
<h3><span class="header-section-number">10.5.1</span> Background</h3>
<p>Hierarchical clustering is an ancient technique that aims to generate a dendrogram containing a hierarchy of samples.
This is most commonly done by greedily agglomerating samples into clusters, then agglomerating those clusters into larger clusters, and so on until all samples belong to a single cluster.
Variants of hierarchical clustering methods primarily differ in how they choose to perform the agglomerations.
For example, complete linkage aims to merge clusters with the smallest maximum distance between their elements, while Ward’s method aims to minimize the increase in within-cluster variance.</p>
<p>In the context of scRNA-seq, the main advantage of hierarchical clustering lies in the production of the dendrogram.
This is a rich summary that describes not only the relationships between cells but also the relationships between clusters at varying resolution.
Users can easily “cut” the tree at different heights to define clusters with different granularity, where clusters defined at high resolution are guaranteed to be nested within those defined at a lower resolution.
The dendrogram is also a natural representation of the data in situations where cells have descended from a relatively recent common ancestor.</p>
<p>In practice, hierachical clustering is too slow to be used for anything but the smallest scRNA-seq datasets.
Most variants require a cell-cell distance matrix that is prohibitively expensive to compute for many cells.
Greedy agglomeration is also likely to result in a quantitatively suboptimal partitioning (as defined by the agglomeration measure) at higher levels of the dendrogram when the number of cells and merge steps is high.
Nonetheless, we will still demonstrate the application of hierarchical clustering here, as it is useful for squeezing more information out of datasets with very few cells<a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a>.</p>
</div>
<div id="implementation-1" class="section level3">
<h3><span class="header-section-number">10.5.2</span> Implementation</h3>
<p>As the PBMC dataset is too large, we will demonstrate on the 416B dataset instead.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb235-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb235-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb235-3" data-line-number="3">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">LunSpikeInData</span>(<span class="dt">which=</span><span class="st">&quot;416b&quot;</span>) </a>
<a class="sourceLine" id="cb235-4" data-line-number="4"></a>
<a class="sourceLine" id="cb235-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb235-6" data-line-number="6"><span class="kw">library</span>(org.Mm.eg.db)</a>
<a class="sourceLine" id="cb235-7" data-line-number="7">symb &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Mm.eg.db, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b), </a>
<a class="sourceLine" id="cb235-8" data-line-number="8">    <span class="dt">keytype=</span><span class="st">&quot;ENSEMBL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SYMBOL&quot;</span>)</a>
<a class="sourceLine" id="cb235-9" data-line-number="9"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL &lt;-<span class="st"> </span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb235-10" data-line-number="10"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL &lt;-<span class="st"> </span>symb</a>
<a class="sourceLine" id="cb235-11" data-line-number="11"></a>
<a class="sourceLine" id="cb235-12" data-line-number="12"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb235-13" data-line-number="13"><span class="kw">rownames</span>(sce<span class="fl">.416</span>b) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL, </a>
<a class="sourceLine" id="cb235-14" data-line-number="14">    <span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL)</a>
<a class="sourceLine" id="cb235-15" data-line-number="15"></a>
<a class="sourceLine" id="cb235-16" data-line-number="16"><span class="kw">library</span>(TxDb.Mmusculus.UCSC.mm10.ensGene)</a>
<a class="sourceLine" id="cb235-17" data-line-number="17">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(TxDb.Mmusculus.UCSC.mm10.ensGene, </a>
<a class="sourceLine" id="cb235-18" data-line-number="18">    <span class="dt">keys=</span><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL, </a>
<a class="sourceLine" id="cb235-19" data-line-number="19">    <span class="dt">column=</span><span class="st">&quot;CDSCHROM&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb235-20" data-line-number="20"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>CHR &lt;-<span class="st"> </span>location</a>
<a class="sourceLine" id="cb235-21" data-line-number="21"></a>
<a class="sourceLine" id="cb235-22" data-line-number="22">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb235-23" data-line-number="23">mito &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>CHR<span class="op">==</span><span class="st">&quot;chrM&quot;</span>)</a>
<a class="sourceLine" id="cb235-24" data-line-number="24">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce<span class="fl">.416</span>b, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mt=</span>mito))</a>
<a class="sourceLine" id="cb235-25" data-line-number="25">libsize.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_counts, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, </a>
<a class="sourceLine" id="cb235-26" data-line-number="26">    <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>sce<span class="fl">.416</span>b<span class="op">$</span>PlateOnco)</a>
<a class="sourceLine" id="cb235-27" data-line-number="27">feature.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_features_by_counts, <span class="dt">nmads=</span><span class="dv">3</span>, </a>
<a class="sourceLine" id="cb235-28" data-line-number="28">    <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>sce<span class="fl">.416</span>b<span class="op">$</span>PlateOnco)</a>
<a class="sourceLine" id="cb235-29" data-line-number="29">spike.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>pct_counts_ERCC, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>,</a>
<a class="sourceLine" id="cb235-30" data-line-number="30">    <span class="dt">batch=</span>sce<span class="fl">.416</span>b<span class="op">$</span>PlateOnco)</a>
<a class="sourceLine" id="cb235-31" data-line-number="31">keep &lt;-<span class="st"> </span><span class="op">!</span>(libsize.drop <span class="op">|</span><span class="st"> </span>feature.drop <span class="op">|</span><span class="st"> </span>spike.drop)</a>
<a class="sourceLine" id="cb235-32" data-line-number="32">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span>sce<span class="fl">.416</span>b[,keep]</a>
<a class="sourceLine" id="cb235-33" data-line-number="33"></a>
<a class="sourceLine" id="cb235-34" data-line-number="34">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb235-35" data-line-number="35"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb235-36" data-line-number="36">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb235-37" data-line-number="37">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSpikeFactors</span>(sce<span class="fl">.416</span>b, <span class="dt">general.use=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb235-38" data-line-number="38">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb235-39" data-line-number="39"></a>
<a class="sourceLine" id="cb235-40" data-line-number="40">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb235-41" data-line-number="41">fit<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce<span class="fl">.416</span>b, <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>Plate)</a>
<a class="sourceLine" id="cb235-42" data-line-number="42">dec<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce<span class="fl">.416</span>b, fit<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb235-43" data-line-number="43"></a>
<a class="sourceLine" id="cb235-44" data-line-number="44"><span class="co"># Taking all of the genes with positive biological components.</span></a>
<a class="sourceLine" id="cb235-45" data-line-number="45">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec<span class="fl">.416</span>b)[dec<span class="fl">.416</span>b<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb235-46" data-line-number="46"></a>
<a class="sourceLine" id="cb235-47" data-line-number="47">### batch-correction <span class="al">###</span></a>
<a class="sourceLine" id="cb235-48" data-line-number="48"><span class="co"># Composition of the plates is expected to be the same,</span></a>
<a class="sourceLine" id="cb235-49" data-line-number="49"><span class="co"># hence the use of a simple &#39;removeBatchEffect&#39;.</span></a>
<a class="sourceLine" id="cb235-50" data-line-number="50"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb235-51" data-line-number="51"><span class="kw">assay</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;corrected&quot;</span>) &lt;-<span class="st"> </span><span class="kw">removeBatchEffect</span>(<span class="kw">logcounts</span>(sce<span class="fl">.416</span>b), </a>
<a class="sourceLine" id="cb235-52" data-line-number="52">    <span class="dt">design=</span><span class="kw">model.matrix</span>(<span class="op">~</span>sce<span class="fl">.416</span>b<span class="op">$</span>phenotype), <span class="dt">batch=</span>sce<span class="fl">.416</span>b<span class="op">$</span>Plate)</a>
<a class="sourceLine" id="cb235-53" data-line-number="53"></a>
<a class="sourceLine" id="cb235-54" data-line-number="54">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb235-55" data-line-number="55">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sce<span class="fl">.416</span>b, <span class="dt">technical=</span>dec<span class="fl">.416</span>b, </a>
<a class="sourceLine" id="cb235-56" data-line-number="56">    <span class="dt">subset.row=</span>chosen.hvgs, <span class="dt">assay.type=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb235-57" data-line-number="57"></a>
<a class="sourceLine" id="cb235-58" data-line-number="58"><span class="kw">set.seed</span>(<span class="dv">1010</span>)</a>
<a class="sourceLine" id="cb235-59" data-line-number="59">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce<span class="fl">.416</span>b, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb236-1" data-line-number="1">sce<span class="fl">.416</span>b</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 46703 188 
## metadata(1): log.exprs.offset
## assays(3): counts logcounts corrected
## rownames(46703): ENSMUSG00000102693 ENSMUSG00000064842 ... SIRV7
##   CBFB-MYH11-mcherry
## rowData names(14): Length ENSEMBL ... total_counts
##   log10_total_counts
## colnames(188): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1
##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...
##   SLX-11312.N712_S507.H5H5YBBXX.s_8.r_1
##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1
## colData names(63): Source Name cell line ...
##   pct_counts_in_top_200_features_SIRV
##   pct_counts_in_top_500_features_SIRV
## reducedDimNames(2): PCA TSNE
## spikeNames(2): ERCC SIRV
## altExpNames(0):</code></pre>
<p>We compute a cell-cell distance matrix using the top PCs and we apply hierarchical clustering with Ward’s method.
This aims to minimize the increase in the intra-cluster variance at each merge step.
The resulting tree in Figure <a href="clustering-2.html#fig:dend-416b">10.8</a> shows a clear split in the population caused by oncogene induction.</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb238-1" data-line-number="1">dist<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">reducedDim</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;PCA&quot;</span>))</a>
<a class="sourceLine" id="cb238-2" data-line-number="2">tree<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">hclust</span>(dist<span class="fl">.416</span>b, <span class="st">&quot;ward.D2&quot;</span>)</a>
<a class="sourceLine" id="cb238-3" data-line-number="3"></a>
<a class="sourceLine" id="cb238-4" data-line-number="4"><span class="co"># Making a prettier plot.</span></a>
<a class="sourceLine" id="cb238-5" data-line-number="5"><span class="kw">library</span>(dendextend)</a>
<a class="sourceLine" id="cb238-6" data-line-number="6">tree<span class="fl">.416</span>b<span class="op">$</span>labels &lt;-<span class="st"> </span><span class="kw">seq_along</span>(tree<span class="fl">.416</span>b<span class="op">$</span>labels)</a>
<a class="sourceLine" id="cb238-7" data-line-number="7">dend &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(tree<span class="fl">.416</span>b, <span class="dt">hang=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb238-8" data-line-number="8"></a>
<a class="sourceLine" id="cb238-9" data-line-number="9">combined.fac &lt;-<span class="st"> </span><span class="kw">paste0</span>(sce<span class="fl">.416</span>b<span class="op">$</span>Plate, <span class="st">&quot;.&quot;</span>, sce<span class="fl">.416</span>b<span class="op">$</span>Oncogene)</a>
<a class="sourceLine" id="cb238-10" data-line-number="10"><span class="kw">labels_colors</span>(dend) &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb238-11" data-line-number="11">    <span class="st">`</span><span class="dt">20160113.control</span><span class="st">`</span>=<span class="st">&quot;blue&quot;</span>,</a>
<a class="sourceLine" id="cb238-12" data-line-number="12">    <span class="st">`</span><span class="dt">20160113.induced</span><span class="st">`</span>=<span class="st">&quot;red&quot;</span>,</a>
<a class="sourceLine" id="cb238-13" data-line-number="13">    <span class="st">`</span><span class="dt">20160325.control</span><span class="st">`</span>=<span class="st">&quot;dodgerblue&quot;</span>,</a>
<a class="sourceLine" id="cb238-14" data-line-number="14">    <span class="st">`</span><span class="dt">20160325.induced</span><span class="st">`</span>=<span class="st">&quot;salmon&quot;</span></a>
<a class="sourceLine" id="cb238-15" data-line-number="15">)[combined.fac][<span class="kw">order.dendrogram</span>(dend)]</a>
<a class="sourceLine" id="cb238-16" data-line-number="16"></a>
<a class="sourceLine" id="cb238-17" data-line-number="17"><span class="kw">plot</span>(dend)</a></code></pre></div>
<div class="figure"><span id="fig:dend-416b"></span>
<img src="P2_W06.clustering_files/figure-html/dend-416b-1.png" alt="Hierarchy of cells in the 416B data set after hierarchical clustering, where each leaf node is a cell that is coloured according to its oncogene induction status (red is induced, blue is control) and plate of origin (light or dark)." width="672" />
<p class="caption">
Figure 10.8: Hierarchy of cells in the 416B data set after hierarchical clustering, where each leaf node is a cell that is coloured according to its oncogene induction status (red is induced, blue is control) and plate of origin (light or dark).
</p>
</div>
<p>To obtain explicit clusters, we “cut” the tree by removing internal branches such that every subtree represents a distinct cluster.
This is most simply done by removing internal branches above a certain height of the tree, as performed by the <code>cutree()</code> function.
We generally prefer to use the <em><a href="https://CRAN.R-project.org/package=dynamicTreeCut">dynamicTreeCut</a></em> package, which uses the shape of the tree to choose a suitable partitioning (Figure <a href="#dend-cluster"><strong>??</strong></a>).</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb239-1" data-line-number="1"><span class="kw">library</span>(dynamicTreeCut)</a>
<a class="sourceLine" id="cb239-2" data-line-number="2"></a>
<a class="sourceLine" id="cb239-3" data-line-number="3"><span class="co"># minClusterSize needs to be turned down for small datasets.</span></a>
<a class="sourceLine" id="cb239-4" data-line-number="4"><span class="co"># deepSplit controls the resolution of the partitioning.</span></a>
<a class="sourceLine" id="cb239-5" data-line-number="5">clust<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">cutreeDynamic</span>(tree<span class="fl">.416</span>b, <span class="dt">distM=</span><span class="kw">as.matrix</span>(dist<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb239-6" data-line-number="6">    <span class="dt">minClusterSize=</span><span class="dv">10</span>, <span class="dt">deepSplit=</span><span class="dv">1</span>)</a></code></pre></div>
<pre><code>##  ..cutHeight not given, setting it to 1300  ===&gt;  99% of the (truncated) height range in dendro.
##  ..done.</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb241-1" data-line-number="1"><span class="kw">table</span>(clust<span class="fl">.416</span>b)</a></code></pre></div>
<pre><code>## clust.416b
##  1  2  3  4  5  6 
## 45 44 37 36 13 13</code></pre>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb243-1" data-line-number="1"><span class="kw">labels_colors</span>(dend) &lt;-<span class="st"> </span>clust<span class="fl">.416</span>b[<span class="kw">order.dendrogram</span>(dend)]</a>
<a class="sourceLine" id="cb243-2" data-line-number="2"><span class="kw">plot</span>(dend)</a></code></pre></div>
<div class="figure"><span id="fig:dend-cluster"></span>
<img src="P2_W06.clustering_files/figure-html/dend-cluster-1.png" alt="Hierarchy of cells in the 416B data set after hierarchical clustering, where each leaf node is a cell that is coloured according to its assigned cluster identity from a dynamic tree cut." width="672" />
<p class="caption">
Figure 10.9: Hierarchy of cells in the 416B data set after hierarchical clustering, where each leaf node is a cell that is coloured according to its assigned cluster identity from a dynamic tree cut.
</p>
</div>
<p>This generally corresponds well to the grouping of cells on a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="clustering-2.html#fig:tsne-416b">10.10</a>).</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb244-1" data-line-number="1">sce<span class="fl">.416</span>b<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb244-2" data-line-number="2"><span class="kw">plotReducedDim</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-416b"></span>
<img src="P2_W06.clustering_files/figure-html/tsne-416b-1.png" alt="$t$-SNE plot of the 416B dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from hierarchical clustering." width="672" />
<p class="caption">
Figure 10.10: <span class="math inline">\(t\)</span>-SNE plot of the 416B dataset, where each point represents a cell and is coloured according to the identity of the assigned cluster from hierarchical clustering.
</p>
</div>
</div>
<div id="assessing-cluster-separation-2" class="section level3">
<h3><span class="header-section-number">10.5.3</span> Assessing cluster separation</h3>
<p>We check the separation of the clusters using the silhouette width (Figure <a href="clustering-2.html#fig:silhouette416b">10.11</a>).
For each cell, we compute the average distance to cells in each other cluster.
We then compute the minimum of these average distances across all clusters, as well as the average distance to cells in the same cluster.
The silhouette width for each cell is defined as the difference between these two values divided by their maximum.
Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters.</p>
<p>Each cluster would ideally contain large positive silhouette widths, indicating that it is well-separated from other clusters.
In Figure <a href="clustering-2.html#fig:silhouette416b">10.11</a>, some clusters are well-separated while others have a substantial proportion of negative widths.
These can arise from the presence of internal subclusters, which inflates the within-cluster distance; or overclustering, where cells at the boundary of a partition are closer to the neighboring cluster than their own cluster.</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb245-1" data-line-number="1">sil &lt;-<span class="st"> </span><span class="kw">silhouette</span>(clust<span class="fl">.416</span>b, <span class="dt">dist =</span> dist<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb245-2" data-line-number="2"><span class="kw">plot</span>(sil)</a></code></pre></div>
<div class="figure"><span id="fig:silhouette416b"></span>
<img src="P2_W06.clustering_files/figure-html/silhouette416b-1.png" alt="Silhouette widths for cells in each cluster in the 416B dataset. Each bar represents a cell, grouped by the cluster to which it is assigned." width="672" />
<p class="caption">
Figure 10.11: Silhouette widths for cells in each cluster in the 416B dataset. Each bar represents a cell, grouped by the cluster to which it is assigned.
</p>
</div>
<p>For a more detailed examination, we identify the closest neighboring cluster for cells with negative widths.
This provides a perspective on the relationships between clusters that is closer to the raw data than the dendrogram in Figure <a href="clustering-2.html#fig:dend-cluster">10.9</a>.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb246-1" data-line-number="1">neg.widths &lt;-<span class="st"> </span>sil[,<span class="dv">3</span>] <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb246-2" data-line-number="2"><span class="kw">table</span>(<span class="dt">Cluster=</span>sil[neg.widths,<span class="dv">1</span>], <span class="dt">Neighbor=</span>sil[neg.widths,<span class="dv">2</span>])</a></code></pre></div>
<pre><code>##        Neighbor
## Cluster  1  2  4  5  6
##       1  0  0  1  0  0
##       2  0  0  0  4  0
##       3  0 32  0  0  5
##       6  1  0  2  0  0</code></pre>
<p>The average silhouette width across all cells can also be used to choose clustering parameters.
The aim is to maximize the average silhouette width in order to obtain well-separated clusters.
This can be helpful to automatically obtain a “reasonable” clustering, though in practice, the clustering that yields the strongest separation is usually trivial and not biologically informative.</p>
</div>
</div>
<div id="session-info-4" class="section level2">
<h2><span class="header-section-number">10.6</span> Session Info</h2>
<pre><code>## &lt;button class=&quot;aaron-collapse&quot;&gt;View session info&lt;/button&gt;
## &lt;div class=&quot;aaron-content&quot;&gt;
## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.6 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] dynamicTreeCut_1.63-1       dendextend_1.12.0          
##  [3] cluster_2.1.0               pheatmap_1.0.12            
##  [5] scater_1.13.9               ggplot2_3.2.0              
##  [7] scran_1.13.9                SingleCellExperiment_1.5.3 
##  [9] SummarizedExperiment_1.15.5 DelayedArray_0.11.4        
## [11] BiocParallel_1.19.0         matrixStats_0.54.0         
## [13] Biobase_2.45.0              GenomicRanges_1.37.14      
## [15] GenomeInfoDb_1.21.1         IRanges_2.19.10            
## [17] S4Vectors_0.23.17           BiocGenerics_0.31.5        
## [19] BiocStyle_2.13.2            Cairo_1.5-10               
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.5.1            edgeR_3.27.6            
##  [3] BiocSingular_1.1.5       viridisLite_0.3.0       
##  [5] DelayedMatrixStats_1.7.1 assertthat_0.2.1        
##  [7] statmod_1.4.32           BiocManager_1.30.4      
##  [9] dqrng_0.2.1              GenomeInfoDbData_1.2.1  
## [11] vipor_0.4.5              yaml_2.2.0              
## [13] pillar_1.4.2             lattice_0.20-38         
## [15] glue_1.3.1               limma_3.41.6            
## [17] digest_0.6.20            RColorBrewer_1.1-2      
## [19] XVector_0.25.0           colorspace_1.4-1        
## [21] htmltools_0.3.6          Matrix_1.2-17           
## [23] pkgconfig_2.0.2          bookdown_0.12           
## [25] zlibbioc_1.31.0          purrr_0.3.2             
## [27] scales_1.0.0             tibble_2.1.3            
## [29] withr_2.1.2              lazyeval_0.2.2          
## [31] magrittr_1.5             crayon_1.3.4            
## [33] evaluate_0.14            beeswarm_0.2.3          
## [35] tools_3.6.0              stringr_1.4.0           
## [37] munsell_0.5.0            locfit_1.5-9.1          
## [39] irlba_2.3.3              compiler_3.6.0          
## [41] rsvd_1.0.1               rlang_0.4.0             
## [43] grid_3.6.0               RCurl_1.95-4.12         
## [45] BiocNeighbors_1.3.2      igraph_1.2.4.1          
## [47] bitops_1.0-6             rmarkdown_1.14          
## [49] gtable_0.3.0             R6_2.4.0                
## [51] gridExtra_2.3            knitr_1.23              
## [53] dplyr_0.8.3              stringi_1.4.3           
## [55] ggbeeswarm_0.6.0         Rcpp_1.0.1              
## [57] tidyselect_0.2.5         xfun_0.8                
## &lt;/div&gt;</code></pre>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-wolf2017graph">
<p>Wolf, F. Alexander, Fiona Hamey, Mireya Plass, Jordi Solana, Joakim S. Dahlin, Berthold Gottgens, Nikolaus Rajewsky, Lukas Simon, and Fabian J. Theis. 2017. “Graph Abstraction Reconciles Clustering with Trajectory Inference Through a Topology Preserving Map of Single Cells.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/208819" class="uri">https://doi.org/10.1101/208819</a>.</p>
</div>
<div id="ref-xu2015identification">
<p>Xu, C., and Z. Su. 2015. “Identification of cell types from single-cell transcriptomes using a novel clustering method.” <em>Bioinformatics</em> 31 (12):1974–80.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="10">
<li id="fn10"><p>Pretty deep title here. I’m not a Doctor of Philosophy for nothing, y’know.<a href="clustering-2.html#fnref10" class="footnote-back">↩</a></p></li>
<li id="fn11"><p>“Oh!”, you say, “But my null hypothesis is that T cell abundance is the same before and after drug treatment! Sounds pretty clear to me.” Sure, if you can define what a T cell is. Quantitatively, mind you - what is the minimum CD3 read/UMI count required for a cell to be a T cell? If you can do that, you don’t need clustering - head straight to Chapter ??? for cell annotation. (Do not pass GO. Do not collect $200.)<a href="clustering-2.html#fnref11" class="footnote-back">↩</a></p></li>
<li id="fn12"><p>Sten Linarrsson talked about this in SCG2018, but I don’t know where that work ended up. So this is what passes as a reference for the time being.<a href="clustering-2.html#fnref12" class="footnote-back">↩</a></p></li>
<li id="fn13"><p>In any case, the biggest effect lies in varying <code>k</code>, so if you’re fiddling with these other choices, you should probably do something else that’s more productive. Like folding your socks.<a href="clustering-2.html#fnref13" class="footnote-back">↩</a></p></li>
<li id="fn14"><p>And, hey, this book isn’t going to add padding to itself. Gotta write something to look productive.<a href="clustering-2.html#fnref14" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="normalization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="marker-gene-detection.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W06.clustering.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
