<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 4 Data Infrastructure | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 4 Data Infrastructure | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Data Infrastructure | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-03-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="beyond-r-basics.html">
<link rel="next" href="about-the-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-essentials-of-sce"><i class="fa fa-check"></i><b>4.2</b> The Essentials of <code>sce</code></a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Workflow: Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#package-requirements"><i class="fa fa-check"></i><b>6.1</b> Package Requirements</a></li>
<li class="chapter" data-level="6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#loading-the-data"><i class="fa fa-check"></i><b>6.2</b> Loading the Data</a></li>
<li class="chapter" data-level="6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#preprocessing"><i class="fa fa-check"></i><b>6.3</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#working-with-common-genes"><i class="fa fa-check"></i><b>6.3.1</b> Working with Common Genes</a></li>
<li class="chapter" data-level="6.3.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>6.3.2</b> Cell and Gene Quality Control</a></li>
<li class="chapter" data-level="6.3.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#normalization"><i class="fa fa-check"></i><b>6.3.3</b> Normalization</a></li>
<li class="chapter" data-level="6.3.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#multibatch-normalization"><i class="fa fa-check"></i><b>6.3.4</b> Multibatch Normalization</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a></li>
<li class="chapter" data-level="6.5" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#combining-the-datasets"><i class="fa fa-check"></i><b>6.5</b> Combining the Datasets</a></li>
<li class="chapter" data-level="6.6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#integrating-datasets"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#naive-approach"><i class="fa fa-check"></i><b>6.6.1</b> Naive Approach</a></li>
<li class="chapter" data-level="6.6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#limma-batch-correction"><i class="fa fa-check"></i><b>6.6.2</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="6.6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#mnn-approach"><i class="fa fa-check"></i><b>6.6.3</b> MNN Approach</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#session-info"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="workflow-clustering.html"><a href="workflow-clustering.html"><i class="fa fa-check"></i><b>7</b> Workflow: Clustering</a><ul>
<li class="chapter" data-level="7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#package-requirements-1"><i class="fa fa-check"></i><b>7.1</b> Package Requirements</a></li>
<li class="chapter" data-level="7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#loading-the-data-1"><i class="fa fa-check"></i><b>7.2</b> Loading the Data</a></li>
<li class="chapter" data-level="7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#understanding-the-data"><i class="fa fa-check"></i><b>7.3</b> Understanding the Data</a></li>
<li class="chapter" data-level="7.4" data-path="workflow-clustering.html"><a href="workflow-clustering.html#preprocessing-1"><i class="fa fa-check"></i><b>7.4</b> Preprocessing</a></li>
<li class="chapter" data-level="7.5" data-path="workflow-clustering.html"><a href="workflow-clustering.html#feature-selection-1"><i class="fa fa-check"></i><b>7.5</b> Feature Selection</a></li>
<li class="chapter" data-level="7.6" data-path="workflow-clustering.html"><a href="workflow-clustering.html#dimension-reduction"><i class="fa fa-check"></i><b>7.6</b> Dimension Reduction</a></li>
<li class="chapter" data-level="7.7" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clustering"><i class="fa fa-check"></i><b>7.7</b> Clustering</a><ul>
<li class="chapter" data-level="7.7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#sc3"><i class="fa fa-check"></i><b>7.7.1</b> SC3</a></li>
<li class="chapter" data-level="7.7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clusterexperiment"><i class="fa fa-check"></i><b>7.7.2</b> clusterExperiment</a></li>
<li class="chapter" data-level="7.7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#biocneighbors"><i class="fa fa-check"></i><b>7.7.3</b> BiocNeighbors</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-infrastructure" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Data Infrastructure</h1>
<p>One of the advantages of using Bioconductor packages is that they utilize common data infrastructures which makes analyses interoperable across various packages. Furthermore, much engineering effort is put into making this infrastructure robust and scalable. Here, we describe the <em>SingleCellExperiment</em> object (or <code>sce</code> in shorthand) in detail to describe how it is constructed, utilized in downstream analysis, and how it stores various types of primary data and metadata.</p>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">4.1</span> Prerequisites</h2>
<p>The Bioconductor package <code>SingleCellExperiment</code> provides the <code>SingleCellExperiment</code> class for usage. While the package is implicitly installed and loaded when using any package that depends on the <code>sce</code> class, it can be explicitly installed (and loaded) as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&#39;SingleCellExperiment&#39;</span>)</code></pre>
<p>Additionally, we use some functions from the <code>scater</code> and <code>scran</code> packages, as well as the CRAN package <code>uwot</code> (which conveniently can also be installed through <code>BiocManager</code>). These functions will be accessed through the <code>&lt;package&gt;::&lt;function&gt;</code> convention as needed.</p>
<pre class="sourceCode r"><code class="sourceCode r">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&#39;scater&#39;</span>, <span class="st">&#39;scran&#39;</span>, <span class="st">&#39;uwot&#39;</span>))</code></pre>
<p>For this session, all we will need loaded is the <code>SingleCellExperiment</code> package:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)</code></pre>
</div>
<div id="the-essentials-of-sce" class="section level2">
<h2><span class="header-section-number">4.2</span> The Essentials of <code>sce</code></h2>
<div class="figure">
<img src="_img/singlecellexperiment.png" alt="Figure 1. Overview of the SingleCellExperiment object" />
<p class="caption">Figure 1. Overview of the SingleCellExperiment object</p>
</div>
<div id="primary-data-the-assays-slot" class="section level3">
<h3><span class="header-section-number">4.2.1</span> Primary Data: The <code>assays</code> Slot</h3>
<p>The <code>SingleCellExperiment</code> (<code>sce</code>) object is the basis of single-cell analytical applications based in Bioconductor. The <code>sce</code> object is an <a href="https://adv-r.hadley.nz/s4.html">S4 object</a>, which in essence provides a more formalized approach towards construction and accession of data compared to other methods available in R. The utility of S4 comes from validity checks that ensure that safe data manipulation, and most important for our discussion, from its extensibility through <em>slots</em>.</p>
<p>If we imagine the <code>sce</code> object to be a ship, the <em>slots</em> of <code>sce</code> can be thought of as individual cargo boxes - each exists as a separate entity <em>within</em> the <code>sce</code> object. Furthermore, each slot contains data that arrives in its own format. To extend the metaphor, we can imagine that different variations of cargo boxes are required for fruits versus bricks. In the case of <code>sce</code>, certain slots expect numeric matrices, whereas others may expect data frames.</p>
<p>To construct a rudimentary <code>sce</code> object, all we need is a single slot:</p>
<ul>
<li><code>assays</code> slot: contains primary data such as counts in list, where each entry of the list is in a matrix format, where rows correspond to features (genes) and columns correspond to samples (cells) (<em>Figure 1A, blue box</em>)</li>
</ul>
<p>Let’s start simple by generating three cells worth of count data across ten genes.</p>
<pre class="sourceCode r"><code class="sourceCode r">counts_matrix &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">cell_1 =</span> <span class="kw">rpois</span>(<span class="dv">10</span>, <span class="dv">10</span>), 
                    <span class="dt">cell_2 =</span> <span class="kw">rpois</span>(<span class="dv">10</span>, <span class="dv">10</span>), 
                    <span class="dt">cell_3 =</span> <span class="kw">rpois</span>(<span class="dv">10</span>, <span class="dv">30</span>))
<span class="kw">rownames</span>(counts_matrix) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;gene_&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)
counts_matrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(counts_matrix) <span class="co"># must be a matrix object!</span></code></pre>
<p>From this, we can now construct our first <code>SingleCellExperiment</code> object, using the defined <em>constructor</em>, <code>SingleCellExperiment()</code>. Note that we provide our data as a <em>named</em> <em>list</em>, and each entry of the list is a <em>matrix</em>. Here, we name the <code>counts_matrix</code> entry as simply <code>counts</code> within the list.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> counts_matrix))</code></pre>
<p>To inspect the object, we can simply type <code>sce</code> into the console to see some pertinent information, which will display an overview of the various slots available to us (which may or may not have any data).</p>
<pre class="sourceCode r"><code class="sourceCode r">sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>To access the count data we just supplied, we can do any one of the following:</p>
<ul>
<li><code>assay(sce, "counts")</code> - this is the most general method, where we can supply the name of the assay as the second argument.</li>
<li><code>counts(sce)</code> - this is the same as the above, but <em>only</em> works for assays with the special name <code>"counts"</code>.</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">counts</span>(sce)</code></pre>
<pre><code>##         cell_1 cell_2 cell_3
## gene_1      10      9     30
## gene_2      13     21     35
## gene_3       6     11     20
## gene_4       9     10     38
## gene_5       8     14     24
## gene_6       9     14     24
## gene_7      13      4     22
## gene_8      14      9     25
## gene_9      12     12     36
## gene_10     10     16     38</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## assay(sce, &quot;counts&quot;) ## same as above in this special case</span></code></pre>
</div>
<div id="extending-the-assays-slot" class="section level3">
<h3><span class="header-section-number">4.2.2</span> Extending the <code>assays</code> Slot</h3>
<p>What makes the <code>assay</code> slot especially powerful is that it can hold <em>multiple</em> representations of the primary data. This is especially useful for storing a <em>normalized</em> version of the data. We can do just that as shown below, using the <code>scran</code> and <code>scater</code> packages to compute a log-count normalized representation of the initial primary data.</p>
<p>Note that here, we overwrite our previous <code>sce</code> upon reassigning the results to <code>sce</code> - this is because these functions <em>return</em> a <code>SingleCellExperiment</code> object. Some functions - especially those outside of single-cell oriented Bioconductor packages - do not, in which case you will need to append your results to the <code>sce</code> object (see below).</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scran<span class="op">::</span><span class="kw">computeSumFactors</span>(sce)
sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">normalize</span>(sce)</code></pre>
<p>Viewing the object again, we see that these functions added some new entries:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 10 3 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>Specifically, we see that the <code>assays</code> slot has grown to be comprised of two entries: <code>counts</code> (our initial data) and <code>logcounts</code> (the normalized data). Similar to <code>counts</code>, the <code>logcounts</code> name is a special name which lets us access it simply by typing <code>logcounts(sce)</code>, although the longhand version works just as well.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">logcounts</span>(sce)</code></pre>
<pre><code>##         cell_1 cell_2 cell_3
## gene_1    4.13   3.80   4.22
## gene_2    4.49   4.96   4.43
## gene_3    3.45   4.07   3.68
## gene_4    3.99   3.94   4.55
## gene_5    3.83   4.40   3.92
## gene_6    3.99   4.40   3.92
## gene_7    4.49   2.75   3.80
## gene_8    4.59   3.80   3.98
## gene_9    4.38   4.19   4.47
## gene_10   4.13   4.58   4.55</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## assay(sce, &quot;logcounts&quot;) ## same as above</span></code></pre>
<p>Notice that the data before had a severe discrepancy in counts between cells 1/2 versus 3, and that normalization has ameliorated this difference.</p>
<p>To look at all the available assays within <code>sce</code>, we can type:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assays</span>(sce)</code></pre>
<pre><code>## List of length 2
## names(2): counts logcounts</code></pre>
<p>While the functions above demonstrate automatic addition of assays to our <code>sce</code> object, there may be cases where we want to perform our own calculations and save the result into the <code>assays</code> slot. In particular, this is important for using functions that do <em>not</em> return your <code>SingleCellExperiment</code> object.</p>
<p>Let’s append a new version of the data that has been offset by <code>+100</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">counts_<span class="dv">100</span> &lt;-<span class="st"> </span><span class="kw">assay</span>(sce, <span class="st">&quot;counts&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="dv">100</span>
<span class="kw">assay</span>(sce, <span class="st">&quot;counts_100&quot;</span>) &lt;-<span class="st"> </span>counts_<span class="dv">100</span> <span class="co"># assign a new entry to assays slot</span></code></pre>
<p>Then we can use the accessor <code>assays()</code> (notice this is plural!) to see all our entries into the <code>assay</code> slot that we have made so far. Note that to see the <em>names</em> of all the assays, we use the plural <code>assays()</code> accessor, and to retrieve a single assay entry (as a matrix) we use the singular <code>assay()</code> accessor, providing the name of the assay we wish to retrieve as above.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assays</span>(sce)</code></pre>
<pre><code>## List of length 3
## names(3): counts logcounts counts_100</code></pre>
<p>These entries are also seen on the default view of <code>sce</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 10 3 
## metadata(1): log.exprs.offset
## assays(3): counts logcounts counts_100
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>This sort of extension of the <code>assays</code> slot is represented graphically in <em>Figure 1B (dark blue box)</em>, showing the addition of the <code>logcounts</code> matrix into the <code>assays</code> slot.</p>
<p>In a similar manner, many of the slots of <code>sce</code> are extendable through assignment as shown above, thus allowing for myriad custom functionality as needed for interoperability with functions outside of single-cell oriented Bioconductor packages.</p>
</div>
<div id="column-metadata-coldata-slot" class="section level3">
<h3><span class="header-section-number">4.2.3</span> Column (Meta)Data: <code>colData</code> Slot</h3>
<p>To further annotate our <code>sce</code> object, one of the first and most useful pieces of information is adding on metadata that describes the columns of our primary data, e.g. describing the samples or cells of our experiment. This data is entered into the <code>colData</code> slot:</p>
<ul>
<li><code>colData</code> slot: metadata that describes that samples (cells) provided as a <code>data.frame</code> or (<code>DataFrame</code> if appending), where rows correspond to cells, and columns correspond to the sample (cells) metadata features (e.g. id, batch, author, etc.) (<em>Figure 1A, orange box</em>).</li>
</ul>
<p>So, let’s come up with some metadata for the cells, starting with a batch variable, where cells 1 and 2 are in batch 1, and cell 3 is from batch 2.</p>
<pre class="sourceCode r"><code class="sourceCode r">cell_metadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">batch =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">rownames</span>(cell_metadata) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;cell_&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)</code></pre>
<p>Now, we can take two approaches - either append the <code>cell_metadata</code> to our existing <code>sce</code>, or start from scratch via the <code>SingleCellExperiment()</code> constructor and provide it from the get go. We’ll start from scratch for now, but will also show how to append the data as well:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## From scratch:</span>
sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> counts_matrix),
                           <span class="dt">colData =</span> cell_metadata)

<span class="co">## Appending to existing object (requires DataFrame() coercion)</span>
<span class="co">## colData(sce) &lt;- DataFrame(cell_metadata)</span></code></pre>
<p>Similar to <code>assays</code>, we can see our colData is now populated from the default view of <code>sce</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(1): batch
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>And furthermore access our column (meta)data with the accessor, <code>colData()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(sce)</code></pre>
<pre><code>## DataFrame with 3 rows and 1 column
##            batch
##        &lt;numeric&gt;
## cell_1         1
## cell_2         1
## cell_3         2</code></pre>
<p>Finally, some packages automatically add to the <code>colData</code> slot, for example, the <code>scater</code> package features a function, <code>calculateQCMetrics()</code>, which appends a lot of quality control data. Here we show the first five columns of <code>colData(sce)</code> with the quality control metrics appended to it.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">calculateQCMetrics</span>(sce)
<span class="kw">colData</span>(sce)[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre>
<pre><code>## DataFrame with 3 rows and 5 columns
##            batch is_cell_control total_features_by_counts
##        &lt;numeric&gt;       &lt;logical&gt;                &lt;integer&gt;
## cell_1         1           FALSE                       10
## cell_2         1           FALSE                       10
## cell_3         2           FALSE                       10
##        log10_total_features_by_counts total_counts
##                             &lt;numeric&gt;    &lt;integer&gt;
## cell_1               1.04139268515823          104
## cell_2               1.04139268515823          120
## cell_3               1.04139268515823          292</code></pre>
<div id="using-coldata-for-subsetting" class="section level4">
<h4><span class="header-section-number">4.2.3.1</span> Using <code>colData</code> for Subsetting</h4>
<p>A common operation with <code>colData</code> is its use in subsetting. One simple way to access <code>colData</code> is through the use of the <code>$</code> operator, which is a shortcut for accessing a variable within the <code>colData</code> slot:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce<span class="op">$</span>batch</code></pre>
<pre><code>## [1] 1 1 2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## colData(sce)$batch # same as above</span></code></pre>
<p>If we only wanted cells within batch 1, we could subset our sce object as follows (remember, we subset on the <em>columns</em> in this case because we are filtering by cells/samples here).</p>
<pre class="sourceCode r"><code class="sourceCode r">sce[, sce<span class="op">$</span>batch <span class="op">==</span><span class="st"> </span><span class="dv">1</span>]</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 10 2 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(7): is_feature_control mean_counts ... total_counts
##   log10_total_counts
## colnames(2): cell_1 cell_2
## colData names(10): batch is_cell_control ...
##   pct_counts_in_top_200_features pct_counts_in_top_500_features
## reducedDimNames(0):
## spikeNames(0):</code></pre>
</div>
</div>
<div id="feature-metadata-rowdatarowranges" class="section level3">
<h3><span class="header-section-number">4.2.4</span> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></h3>
<p>Lastly, the rows also have their own metadata slot to store information that pertains to the features of the <code>sce</code> object:</p>
<ul>
<li><code>rowData</code> slot: contains data in a <code>data.frame</code> (<code>DataFrame</code>) format that describes aspects of the data corresponding to the rows of the primary data (<em>Figure 1A, green box</em>).</li>
</ul>
<p>Furthermore, there is a special slot which pertains to features with genomic coordinates:</p>
<ul>
<li><code>rowRanges</code> slot: contains data in a <code>GRangesList</code> (where each entry is a <code>GenomicRanges</code> format) that describes the chromosome, start, and end coordinates of the features (genes, genomic regions).</li>
</ul>
<p>Both of these can be accessed via their respective accessors, <code>rowRanges()</code> and <code>rowData()</code>. In our case, <code>rowRanges(sce)</code> produces an empty list:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowRanges</span>(sce) <span class="co"># empty</span></code></pre>
<pre><code>## GRangesList object of length 10:
## $gene_1 
## GRanges object with 0 ranges and 0 metadata columns:
##    seqnames    ranges strand
##       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
## 
## $gene_2 
## GRanges object with 0 ranges and 0 metadata columns:
##      seqnames ranges strand
## 
## $gene_3 
## GRanges object with 0 ranges and 0 metadata columns:
##      seqnames ranges strand
## 
## ...
## &lt;7 more elements&gt;
## -------
## seqinfo: no sequences</code></pre>
<p>However, our call to <code>calculateQCMetrics(sce)</code> in the prior section filled in the <code>rowData</code> slot of our <code>sce</code> object, as we can see below (only the first three columns are shown for brevity):</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowData</span>(sce)[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre>
<pre><code>## DataFrame with 10 rows and 3 columns
##         is_feature_control      mean_counts log10_mean_counts
##                  &lt;logical&gt;        &lt;numeric&gt;         &lt;numeric&gt;
## gene_1               FALSE 16.3333333333333  1.23888208891514
## gene_2               FALSE               23  1.38021124171161
## gene_3               FALSE 12.3333333333333   1.1249387366083
## gene_4               FALSE               19  1.30102999566398
## gene_5               FALSE 15.3333333333333  1.21307482530885
## gene_6               FALSE 15.6666666666667  1.22184874961636
## gene_7               FALSE               13  1.14612803567824
## gene_8               FALSE               16  1.23044892137827
## gene_9               FALSE               20  1.32221929473392
## gene_10              FALSE 21.3333333333333  1.34895354798116</code></pre>
<p>In a similar fashion to the <code>colData</code> slot, such feature metadata could be provided at the onset when creating the <code>SingleCellExperiment</code> object, which we leave up to the reader as an exercise.</p>
<div id="subsetting-with-on-rows" class="section level4">
<h4><span class="header-section-number">4.2.4.1</span> Subsetting with on Rows</h4>
<p>To subset an <code>sce</code> object down at the feature/gene level, we can do a row subsetting operation similar to other R objects, by supplying either numeric indices or a vector of names:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce[<span class="kw">c</span>(<span class="st">&quot;gene_1&quot;</span>, <span class="st">&quot;gene_4&quot;</span>), ]</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 2 3 
## metadata(0):
## assays(1): counts
## rownames(2): gene_1 gene_4
## rowData names(7): is_feature_control mean_counts ... total_counts
##   log10_total_counts
## colnames(3): cell_1 cell_2 cell_3
## colData names(10): batch is_cell_control ...
##   pct_counts_in_top_200_features pct_counts_in_top_500_features
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## sce[c(1, 4), ] # same as above in this case</span></code></pre>
</div>
</div>
<div id="size-factors-slot-sizefactors" class="section level3">
<h3><span class="header-section-number">4.2.5</span> Size Factors Slot: <code>sizeFactors</code></h3>
<p>Briefly, we already encountered this via the <code>scran::computeSumFactors(sce)</code> call, which adds a <code>sizeFactors</code> slot:</p>
<ul>
<li><code>sizeFactors</code> slot: contains information in a numeric vector regarding the sample/cell normalization factors used to produce a normalize data representation (<em>Figure 1B, brown box</em>)</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scran<span class="op">::</span><span class="kw">computeSumFactors</span>(sce)
sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">normalize</span>(sce)
<span class="kw">sizeFactors</span>(sce)</code></pre>
<pre><code>## [1] 0.605 0.698 1.698</code></pre>
</div>
</div>
<div id="a-brief-recap-from-se-to-sce" class="section level2">
<h2><span class="header-section-number">4.3</span> A Brief Recap: From <code>se</code> to <code>sce</code></h2>
<p>So far, we have covered the <code>assays</code> (primary data), <code>colData</code> (sample metadata), <code>rowData</code>/<code>rowRanges</code> (feature metadata), and <code>sizeFactors</code> slots of <code>SingleCellExperiment</code>.</p>
<p>What is important to note is that the <code>SingleCellExperiment</code> class <em>derives</em> from the <code>SummarizedExperiment</code> (<code>se</code>) class, its predecessor, and in particular inherits the aforementioned slots. As such, much of the <code>SummarizedExperiment</code> functionality is retained in <code>SingleCellExperiment</code>. This allows existing methods that work with <code>SummarizedExperiment</code> to work similarly on <code>SingleCellExperiment</code> objects.</p>
<p>So what’s new about the <code>SingleCellExperiment</code> class then? For our discussion, the most important change is the addition of a new slot called <code>reducedDims</code>.</p>
</div>
<div id="the-reduceddims-slot" class="section level2">
<h2><span class="header-section-number">4.4</span> The <code>reducedDims</code> Slot</h2>
<p>The <code>reducedDims</code> slot is a new addition which is specially designed to store the reduced dimensionality representations of primary data, such as PCA, tSNE, UMAP, and others.</p>
<ul>
<li><code>reducedDims</code> slot: contains a list of numeric <code>matrix</code> entries which describe dimensionality reduced representations of the primary data, such that rows represent the columns of the primary data (aka the samples/cells), and columns represent the dimensions</li>
</ul>
<p>Most importantly, just like the <code>assays</code> slot, the <code>reducedDims</code> slot can hold a list of many entries. So, it can hold a PCA, TSNE, and UMAP representation of a given dataset all within the <code>reducedDims</code> slot.</p>
<p>In our example, we can calculate a PCA representation of our data as follows using the <code>scater</code> package function <code>runPCA()</code>. We see that the <code>sce</code> now shows a new <code>reducedDim</code> and that the accessor <code>reducedDim()</code> produces the results of running PCA on the normalized data from <code>logcounts(sce)</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runPCA</span>(sce)
<span class="kw">reducedDim</span>(sce, <span class="st">&quot;PCA&quot;</span>)</code></pre>
<pre><code>##          PC1    PC2
## cell_1 -1.91 -1.412
## cell_2  3.18 -0.202
## cell_3 -1.27  1.614
## attr(,&quot;percentVar&quot;)
## [1] 0.768 0.232</code></pre>
<p>From this, we can also calculate a tSNE representation using the <code>scater</code> package function <code>runTSNE()</code>, and see that it can be seen both in the default view of <code>sce</code> and via accession:</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runTSNE</span>(sce, <span class="dt">perplexity =</span> <span class="fl">0.1</span>)</code></pre>
<pre><code>## Perplexity should be lower than K!</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">reducedDim</span>(sce, <span class="st">&quot;TSNE&quot;</span>)</code></pre>
<pre><code>##         [,1]  [,2]
## cell_1 -5513 -1430
## cell_2  3994 -4059
## cell_3  1519  5489</code></pre>
<p>We can view the names of all our entries in the <code>reducedDims</code> slot via the accessor, <code>reducedDims()</code> (notice that this is plural, and thus not the same as <code>reducedDim()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">reducedDims</span>(sce)</code></pre>
<pre><code>## List of length 2
## names(2): PCA TSNE</code></pre>
<p>Now, say we have a different dimensionality reduction approach which has not yet been implemented with <code>SingleCellExperiment</code> objects in mind. For example, let’s say we want to try the <code>umap()</code> function as implemented in the <code>uwot</code> package (which is a much faster version of the default <code>umap</code> implementation currently in <code>scater</code>).</p>
<p>Similar to how we extended the <code>assays</code> slot with our own custom entry of <code>counts_100</code>, we can do similarly for the <code>reducedDims</code> slot:</p>
<pre class="sourceCode r"><code class="sourceCode r">u &lt;-<span class="st"> </span>uwot<span class="op">::</span><span class="kw">umap</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce)), <span class="dt">n_neighbors =</span> <span class="dv">2</span>)
<span class="kw">reducedDim</span>(sce, <span class="st">&quot;UMAP_uwot&quot;</span>) &lt;-<span class="st"> </span>u

<span class="kw">reducedDim</span>(sce, <span class="st">&quot;UMAP_uwot&quot;</span>)</code></pre>
<pre><code>##          [,1]   [,2]
## cell_1 -0.302 -0.565
## cell_2  0.595  0.299
## cell_3 -0.293  0.266
## attr(,&quot;scaled:center&quot;)
## [1] -8.61 -2.30</code></pre>
<p>And we can also see its entry when we look at the <code>reducedDims()</code> accessor output:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">reducedDims</span>(sce)</code></pre>
<pre><code>## List of length 3
## names(3): PCA TSNE UMAP_uwot</code></pre>
</div>
<div id="one-more-thing-metadata-slot" class="section level2">
<h2><span class="header-section-number">4.5</span> One More Thing: <code>metadata</code> Slot</h2>
<p>Some analyses produce results that do not fit into the aforementioned slots. Thankfully, there is a slot just for this type of messy data, and in fact, can accommodate any type of data, so long as it is in a named list:</p>
<ul>
<li><code>metadata</code> slot: a named list of entries, where each entry in the list can be anything you want it to be</li>
</ul>
<p>For example, say we have some favorite genes, such as highly variable genes, we want to save inside of <code>sce</code> for use in our analysis at a later point. We can do this simply by appending to the metadata slot as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">my_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene_1&quot;</span>, <span class="st">&quot;gene_5&quot;</span>)
<span class="kw">metadata</span>(sce) &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">favorite_genes =</span> my_genes)
<span class="kw">metadata</span>(sce)</code></pre>
<pre><code>## $favorite_genes
## [1] &quot;gene_1&quot; &quot;gene_5&quot;</code></pre>
<p>Similarly, we can append more information via the <code>$</code> operator:</p>
<pre class="sourceCode r"><code class="sourceCode r">your_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene_4&quot;</span>, <span class="st">&quot;gene_8&quot;</span>)
<span class="kw">metadata</span>(sce)<span class="op">$</span>your_genes &lt;-<span class="st"> </span>your_genes
<span class="kw">metadata</span>(sce)</code></pre>
<pre><code>## $favorite_genes
## [1] &quot;gene_1&quot; &quot;gene_5&quot;
## 
## $your_genes
## [1] &quot;gene_4&quot; &quot;gene_8&quot;</code></pre>
</div>
<div id="about-spike-ins" class="section level2">
<h2><span class="header-section-number">4.6</span> About Spike-Ins</h2>
<p>You might have noticed that the <code>sce</code> default view produces an entry with <code>spikeNames</code>. The <code>SingleCellExperiment</code> object contains some special considerations for experiments with spike-in (ERCC) controls. We leave this to the interested reader to learn more about in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html"><code>SingleCellExperiment</code> introductory vignette</a>.</p>
</div>
<div id="working-with-singlecellexperiment" class="section level2">
<h2><span class="header-section-number">4.7</span> Working with <em>SingleCellExperiment</em></h2>
<p><em>Figure 1C</em> shows an example workflow that uses the <code>SingleCellExperiment</code> object as its base, and similar to our walkthrough of the <code>sce</code> class above, continually appends new entries to save the results of the analysis. In the workflows that follow in this book, we will be similarly appending to an initial <code>sce</code> object many of our analytical results.</p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="beyond-r-basics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="about-the-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
