<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Dimensionality reduction | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Dimensionality reduction | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Dimensionality reduction | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-07-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="feature-selection-2.html">
<link rel="next" href="clustering-1.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="6.3" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i><b>6.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i><b>7.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i><b>8.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="9.8" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i><b>9.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="feature-selection-2.html"><a href="feature-selection-2.html"><i class="fa fa-check"></i><b>10</b> Feature selection</a><ul>
<li class="chapter" data-level="10.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>10.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="10.2.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>10.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="10.2.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#coefficient-of-variation"><i class="fa fa-check"></i><b>10.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="10.2.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#sec:spikeins"><i class="fa fa-check"></i><b>10.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="10.2.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>10.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#selecting-highly-variable-genes"><i class="fa fa-check"></i><b>10.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="10.3.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>10.3.1</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="10.3.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-a-fixed-threshold"><i class="fa fa-check"></i><b>10.3.2</b> Based on a fixed threshold</a></li>
<li class="chapter" data-level="10.3.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#keeping-all-genes-above-the-trend"><i class="fa fa-check"></i><b>10.3.3</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>10.4</b> Putting it all together</a></li>
<li class="chapter" data-level="10.5" data-path="feature-selection-2.html"><a href="feature-selection-2.html#session-info-4"><i class="fa fa-check"></i><b>10.5</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html"><i class="fa fa-check"></i><b>11</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="11.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#overview-1"><i class="fa fa-check"></i><b>11.1</b> Overview</a></li>
<li class="chapter" data-level="11.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#principal-components-analysis"><i class="fa fa-check"></i><b>11.2</b> Principal components analysis</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#background-1"><i class="fa fa-check"></i><b>11.2.1</b> Background</a></li>
<li class="chapter" data-level="11.2.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#performing-the-pca"><i class="fa fa-check"></i><b>11.2.2</b> Performing the PCA</a></li>
<li class="chapter" data-level="11.2.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>11.2.3</b> Choosing the number of PCs</a></li>
<li class="chapter" data-level="11.2.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#putting-it-together"><i class="fa fa-check"></i><b>11.2.4</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>11.3</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="11.3.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#motivation-4"><i class="fa fa-check"></i><b>11.3.1</b> Motivation</a></li>
<li class="chapter" data-level="11.3.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#visualizating-with-pca"><i class="fa fa-check"></i><b>11.3.2</b> Visualizating with PCA</a></li>
<li class="chapter" data-level="11.3.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>11.3.3</b> t-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="11.3.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>11.3.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="11.3.5" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#interpreting-the-plots"><i class="fa fa-check"></i><b>11.3.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#session-info-5"><i class="fa fa-check"></i><b>11.4</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="clustering-1.html"><a href="clustering-1.html"><i class="fa fa-check"></i><b>12</b> Clustering</a><ul>
<li class="chapter" data-level="12.1" data-path="clustering-1.html"><a href="clustering-1.html#motivation-5"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="clustering-1.html"><a href="clustering-1.html#comments-on-truth"><i class="fa fa-check"></i><b>12.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="12.3" data-path="clustering-1.html"><a href="clustering-1.html#graph-based-clustering"><i class="fa fa-check"></i><b>12.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="12.3.1" data-path="clustering-1.html"><a href="clustering-1.html#background-2"><i class="fa fa-check"></i><b>12.3.1</b> Background</a></li>
<li class="chapter" data-level="12.3.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation"><i class="fa fa-check"></i><b>12.3.2</b> Implementation</a></li>
<li class="chapter" data-level="12.3.3" data-path="clustering-1.html"><a href="clustering-1.html#other-parameters"><i class="fa fa-check"></i><b>12.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="12.3.4" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>12.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="clustering-1.html"><a href="clustering-1.html#k-means-clustering"><i class="fa fa-check"></i><b>12.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="12.4.1" data-path="clustering-1.html"><a href="clustering-1.html#background-3"><i class="fa fa-check"></i><b>12.4.1</b> Background</a></li>
<li class="chapter" data-level="12.4.2" data-path="clustering-1.html"><a href="clustering-1.html#base-implementation"><i class="fa fa-check"></i><b>12.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="12.4.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>12.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="clustering-1.html"><a href="clustering-1.html#hierarchical-clustering"><i class="fa fa-check"></i><b>12.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="12.5.1" data-path="clustering-1.html"><a href="clustering-1.html#background-4"><i class="fa fa-check"></i><b>12.5.1</b> Background</a></li>
<li class="chapter" data-level="12.5.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation-1"><i class="fa fa-check"></i><b>12.5.2</b> Implementation</a></li>
<li class="chapter" data-level="12.5.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>12.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="clustering-1.html"><a href="clustering-1.html#session-info-6"><i class="fa fa-check"></i><b>12.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>13</b> Marker gene detection</a><ul>
<li class="chapter" data-level="13.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-6"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>13.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="13.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>13.2.1</b> Standard application</a></li>
<li class="chapter" data-level="13.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>13.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="13.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>13.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>13.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="13.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>13.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="13.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>13.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="13.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>13.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>13.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="13.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>13.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="13.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>13.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="13.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>13.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="13.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>13.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="13.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>13.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-7"><i class="fa fa-check"></i><b>13.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>14</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="14.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#motivation-7"><i class="fa fa-check"></i><b>14.1</b> Motivation</a></li>
<li class="chapter" data-level="14.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#simple-two-batch-example"><i class="fa fa-check"></i><b>14.2</b> Simple two batch example</a><ul>
<li class="chapter" data-level="14.2.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>14.2.1</b> Setting up the data</a></li>
<li class="chapter" data-level="14.2.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#performing-mnn-correction"><i class="fa fa-check"></i><b>14.2.2</b> Performing MNN correction</a></li>
<li class="chapter" data-level="14.2.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#correction-diagnostics"><i class="fa fa-check"></i><b>14.2.3</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-8"><i class="fa fa-check"></i><b>14.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>15</b> Sharing and Interactive Interfaces</a><ul>
<li class="chapter" data-level="15.1" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html#session-info-9"><i class="fa fa-check"></i><b>15.1</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>16</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="16.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>16.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="16.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>16.2</b> Parallelization</a></li>
<li class="chapter" data-level="16.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>16.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="16.4" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-10"><i class="fa fa-check"></i><b>16.4</b> Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>17</b> About the Data</a><ul>
<li class="chapter" data-level="17.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>17.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="17.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>17.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>18</b> About the Contributors</a><ul>
<li class="chapter" data-level="18.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>18.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="18.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>18.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="18.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>18.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="18.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>18.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>19</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dimensionality-reduction-2" class="section level1">
<h1><span class="header-section-number">Chapter 11</span> Dimensionality reduction</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="overview-1" class="section level2">
<h2><span class="header-section-number">11.1</span> Overview</h2>
<p>Many scRNA-seq analysis procedures involve comparing cells based on their expression values across multiple genes.
For example, clustering aims to identify cells with similar transcriptomic profiles by computing Euclidean distances across genes.
In these applications, each individual gene represents a dimension of the data.
More intuitively, if we had a scRNA-seq data set with two genes, we could make a two-dimensional plot where each axis represents the expression of one gene and each point in the plot represents a cell.
This concept can be extended to data sets with thousands of genes where each cell’s expression profile defines its location in the high-dimensional expression space.</p>
<p>Dimensionality reduction aims to reduce the number of separate dimensions<a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a> in the data.
This is possible because different genes are correlated if they are affected by the same biological process.
Thus, we do not need to store separate information for individual genes, but can instead compress multiple features into a single dimension, e.g., an “eigengene” <span class="citation">(<span class="citeproc-not-found" data-reference-id="langfelder2007eigengene"><strong>???</strong></span>)</span>.
This reduces computational work in downstream analyses, as calculations only need to be performed for a few dimensions rather than thousands of genes;
reduces noise by averaging across multiple genes to obtain a more precise representation of the patterns in the data;
and enables effective plotting of the data, for those of us who are not capable of visualizing more than 3 dimensions.</p>
<p>We will use the <span class="citation">Zeisel et al. (<a href="#ref-zeisel2015brain">2015</a>)</span> dataset to demonstrate the applications of various dimensionality reduction methods in this chapter.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb239-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb239-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb239-3" data-line-number="3">sce.zeisel &lt;-<span class="st"> </span><span class="kw">ZeiselBrainData</span>()</a>
<a class="sourceLine" id="cb239-4" data-line-number="4"></a>
<a class="sourceLine" id="cb239-5" data-line-number="5"><span class="co"># Removing repeats.</span></a>
<a class="sourceLine" id="cb239-6" data-line-number="6">sce.zeisel &lt;-<span class="st"> </span>sce.zeisel[<span class="kw">rowData</span>(sce.zeisel)<span class="op">$</span>featureType<span class="op">!=</span><span class="st">&quot;repeat&quot;</span>,]</a>
<a class="sourceLine" id="cb239-7" data-line-number="7"></a>
<a class="sourceLine" id="cb239-8" data-line-number="8"><span class="co"># Merging together redundant rows for the same gene.</span></a>
<a class="sourceLine" id="cb239-9" data-line-number="9">raw.names &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;_loc[0-9]+$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(sce.zeisel))</a>
<a class="sourceLine" id="cb239-10" data-line-number="10">new.counts &lt;-<span class="st"> </span><span class="kw">rowsum</span>(<span class="kw">counts</span>(sce.zeisel), <span class="dt">group=</span>raw.names, <span class="dt">reorder=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb239-11" data-line-number="11">sce.zeisel &lt;-<span class="st"> </span>sce.zeisel[<span class="kw">match</span>(<span class="kw">rownames</span>(new.counts), raw.names),]</a>
<a class="sourceLine" id="cb239-12" data-line-number="12"><span class="kw">counts</span>(sce.zeisel) &lt;-<span class="st"> </span>new.counts</a>
<a class="sourceLine" id="cb239-13" data-line-number="13"><span class="kw">rownames</span>(sce.zeisel) &lt;-<span class="st"> </span><span class="kw">rownames</span>(new.counts)</a>
<a class="sourceLine" id="cb239-14" data-line-number="14"></a>
<a class="sourceLine" id="cb239-15" data-line-number="15">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb239-16" data-line-number="16"><span class="co"># Adding Ensembl IDs.</span></a>
<a class="sourceLine" id="cb239-17" data-line-number="17"><span class="kw">library</span>(org.Mm.eg.db)</a>
<a class="sourceLine" id="cb239-18" data-line-number="18">ensembl &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Mm.eg.db, <span class="dt">keys=</span><span class="kw">rownames</span>(sce.zeisel), </a>
<a class="sourceLine" id="cb239-19" data-line-number="19">    <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;ENSEMBL&quot;</span>)</a>
<a class="sourceLine" id="cb239-20" data-line-number="20"><span class="kw">rowData</span>(sce.zeisel)<span class="op">$</span>ENSEMBL &lt;-<span class="st"> </span>ensembl</a>
<a class="sourceLine" id="cb239-21" data-line-number="21"></a>
<a class="sourceLine" id="cb239-22" data-line-number="22">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb239-23" data-line-number="23"><span class="co"># Skipping: assuming that QC was performed by the original authors.</span></a>
<a class="sourceLine" id="cb239-24" data-line-number="24"></a>
<a class="sourceLine" id="cb239-25" data-line-number="25">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb239-26" data-line-number="26"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb239-27" data-line-number="27"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb239-28" data-line-number="28">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.zeisel)</a>
<a class="sourceLine" id="cb239-29" data-line-number="29">sce.zeisel &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.zeisel, <span class="dt">cluster=</span>clusters, <span class="dt">min.mean=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb239-30" data-line-number="30">sce.zeisel &lt;-<span class="st"> </span><span class="kw">computeSpikeFactors</span>(sce.zeisel, <span class="dt">general.use=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb239-31" data-line-number="31">sce.zeisel &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.zeisel)</a>
<a class="sourceLine" id="cb239-32" data-line-number="32"></a>
<a class="sourceLine" id="cb239-33" data-line-number="33">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb239-34" data-line-number="34">fit.zeisel &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce.zeisel)</a>
<a class="sourceLine" id="cb239-35" data-line-number="35">dec.zeisel &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce.zeisel, fit.zeisel)</a>
<a class="sourceLine" id="cb239-36" data-line-number="36"></a>
<a class="sourceLine" id="cb239-37" data-line-number="37"><span class="co"># Taking 10% of the genes with the top biological components.</span></a>
<a class="sourceLine" id="cb239-38" data-line-number="38">ordered.genes &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.zeisel)[<span class="kw">order</span>(dec.zeisel<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)]</a>
<a class="sourceLine" id="cb239-39" data-line-number="39">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">head</span>(ordered.genes, <span class="fl">0.1</span> <span class="op">*</span><span class="st"> </span><span class="kw">nrow</span>(dec.zeisel))</a></code></pre></div>
</div>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb240-1" data-line-number="1">sce.zeisel</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 19896 3005 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(19896): Tspan12 Tshz1 ... ERCC-00170 ERCC-00171
## rowData names(2): featureType ENSEMBL
## colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12
##   1772058148_F03
## colData names(10): tissue group # ... level1class level2class
## reducedDimNames(0):
## spikeNames(1): ERCC</code></pre>
</div>
<div id="principal-components-analysis" class="section level2">
<h2><span class="header-section-number">11.2</span> Principal components analysis</h2>
<div id="background-1" class="section level3">
<h3><span class="header-section-number">11.2.1</span> Background</h3>
<p>Principal components analysis (PCA) discovers axes in high-dimensional space that capture the largest amount of variation.
This is best understood by imagining each axis as a line.
Say we draw a line anywhere, and we move all cells in our data set onto this line by the shortest path.
The variance captured by this axis is defined as the variance across cells along that line.
In PCA, the first axis (or “principal component”, PC) is chosen such that it captures the greatest variance across cells.
The next PC is chosen such that it is orthogonal to the first and captures the greatest remaining amount of variation, and so on.</p>
<p>By definition, the top PCs capture the dominant factors of heterogeneity in the data set.
Thus, we can perform dimensionality reduction by restricting downstream analyses to the top PCs.
This strategy is simple, highly effective and widely used throughout the data sciences.
It takes advantage of the well-studied theoretical properties of the PCA - namely, that a low-rank approximation formed from the top PCs is the optimal approximation of the original data for a given matrix rank.
It also allows us to use a wide range of fast PCA implementations for scalable and efficient data analysis.</p>
<p>When applying PCA to scRNA-seq data, our assumption is that biological processes affect multiple genes in a coordinated manner.
This means the the earlier PCs are likely to represent biological structure as more variation can be captured by considering the correlated behaviour of many genes.
By comparison, random (technical) noise is expected to affect each gene independently.
There is unlikely to be an axis that can capture random variation across many genes, suggesting that noise is mostly concentrated in the later PCs.
This motivates the use of the earlier PCs in our downstream analyses, which concentrates the biological signal to simultaneously reduce computational work and remove noise.</p>
</div>
<div id="performing-the-pca" class="section level3">
<h3><span class="header-section-number">11.2.2</span> Performing the PCA</h3>
<p>We perform the PCA on the log-normalized expression values using the HVGs we selected earlier.
By restricting to the HVGs, we reduce computational work and noise to improve the performance of the PCA.
(While PCA is robust to noise, an excess of it may cause the earlier PCs to capture noise instead of biological structure.)
The <code>runPCA()</code> function from <em><a href="https://bioconductor.org/packages/3.10/scater">scater</a></em> will compute the PCA and store the coordinates in the output object.</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb242-1" data-line-number="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb242-2" data-line-number="2">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce.zeisel, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">scale=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb242-3" data-line-number="3"><span class="kw">reducedDimNames</span>(sce.zeisel)</a></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;</code></pre>
<p>The PCA is based on a mathematical technique called the singular decomposition decomposition (SVD).
By default, <code>runPCA()</code> will use an exact SVD based on base R’s <code>svd()</code> function.
It then only keeps the first 50 PCs for downstream analysis.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb244-1" data-line-number="1"><span class="kw">dim</span>(<span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA&quot;</span>))</a></code></pre></div>
<pre><code>## [1] 3005   50</code></pre>
<p>For large data sets, greater efficiency is obtained by using approximate SVD algorithms that only compute the top PCs.
For example, we can use methods from the <em><a href="https://CRAN.R-project.org/package=irlba">irlba</a></em> package:</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb246-1" data-line-number="1"><span class="kw">library</span>(BiocSingular)</a>
<a class="sourceLine" id="cb246-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb246-3" data-line-number="3">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce.zeisel, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">scale=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb246-4" data-line-number="4">    <span class="dt">BSPARAM=</span><span class="kw">IrlbaParam</span>(), <span class="dt">name=</span><span class="st">&quot;IRLBA&quot;</span>)</a>
<a class="sourceLine" id="cb246-5" data-line-number="5"><span class="kw">reducedDimNames</span>(sce.zeisel)</a></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;   &quot;IRLBA&quot;</code></pre>
<p>The <code>BSPARAM=</code> argument from <em><a href="https://bioconductor.org/packages/3.10/BiocSingular">BiocSingular</a></em> provides a flexible interface for specifying different SVD algorithms.
Any function that accepts a <code>BSPARAM=</code> argument can often be sped up by using an alternative SVD algorithm.
Another strategy is to perform the SVD using methods from the <em><a href="https://CRAN.R-project.org/package=rsvd">rsvd</a></em> package:</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb248-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1001</span>) <span class="co"># Slightly different seed to add some variety!</span></a>
<a class="sourceLine" id="cb248-2" data-line-number="2">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce.zeisel, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">scale=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb248-3" data-line-number="3">    <span class="dt">BSPARAM=</span><span class="kw">RandomParam</span>(), <span class="dt">name=</span><span class="st">&quot;RSVD&quot;</span>)</a>
<a class="sourceLine" id="cb248-4" data-line-number="4"><span class="kw">reducedDimNames</span>(sce.zeisel)</a></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;   &quot;IRLBA&quot; &quot;RSVD&quot;</code></pre>
<p>The majority of these approximate methods are based on randomization and thus require <code>set.seed()</code> to obtain reproducible results<a href="#fn11" class="footnote-ref" id="fnref11"><sup>11</sup></a>.</p>
</div>
<div id="choosing-the-number-of-pcs" class="section level3">
<h3><span class="header-section-number">11.2.3</span> Choosing the number of PCs</h3>
<div id="motivation-3" class="section level4">
<h4><span class="header-section-number">11.2.3.1</span> Motivation</h4>
<p>How many of the top PCs should we retain for downstream analyses?
The choice of the number of PCs <span class="math inline">\(d\)</span> is a decision that is analogous to the choice of the number of HVGs to use.
Using more PCs will avoid discarding biological signal in later PCs, at the cost of retaining more noise.
Most practitioners will simply set <span class="math inline">\(d\)</span> to a “reasonable” but arbitrary value, typically ranging from 10 to 50.
This is often satisfactory<a href="#fn12" class="footnote-ref" id="fnref12"><sup>12</sup></a> provided it is coupled with sufficient testing of alternative values to explore other perspectives of the data at a different bias-variance trade-off.
Nonetheless, we will describe some more data-driven strategies to guide a suitable choice of <span class="math inline">\(d\)</span>.</p>
</div>
<div id="using-the-elbow-point" class="section level4">
<h4><span class="header-section-number">11.2.3.2</span> Using the elbow point</h4>
<p>A simple heuristic for choosing <span class="math inline">\(d\)</span> involves identifying the elbow point in the percentage of variance explained by successive PCs.
This refers to the “elbow” in the curve of a scree plot as shown in Figure <a href="dimensionality-reduction-2.html#fig:elbow">11.1</a>.</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb250-1" data-line-number="1"><span class="co"># Hoping to find this guy a better home.</span></a>
<a class="sourceLine" id="cb250-2" data-line-number="2">elbowFinder &lt;-<span class="st"> </span><span class="cf">function</span>(var.exp) {</a>
<a class="sourceLine" id="cb250-3" data-line-number="3">    var.exp &lt;-<span class="st"> </span><span class="kw">sort</span>(var.exp, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb250-4" data-line-number="4"></a>
<a class="sourceLine" id="cb250-5" data-line-number="5">    dy &lt;-<span class="st"> </span><span class="op">-</span><span class="kw">diff</span>(<span class="kw">range</span>(var.exp))</a>
<a class="sourceLine" id="cb250-6" data-line-number="6">    dx &lt;-<span class="st"> </span><span class="kw">length</span>(var.exp) <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb250-7" data-line-number="7">    l2 &lt;-<span class="st"> </span><span class="kw">sqrt</span>(dx<span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>dy<span class="op">^</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb250-8" data-line-number="8">    dx &lt;-<span class="st"> </span>dx<span class="op">/</span>l2</a>
<a class="sourceLine" id="cb250-9" data-line-number="9">    dy &lt;-<span class="st"> </span>dy<span class="op">/</span>l2</a>
<a class="sourceLine" id="cb250-10" data-line-number="10"></a>
<a class="sourceLine" id="cb250-11" data-line-number="11">    dy0 &lt;-<span class="st"> </span>var.exp <span class="op">-</span><span class="st"> </span>var.exp[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb250-12" data-line-number="12">    dx0 &lt;-<span class="st"> </span><span class="kw">seq_along</span>(var.exp) <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb250-13" data-line-number="13"></a>
<a class="sourceLine" id="cb250-14" data-line-number="14">    parallel.l2 &lt;-<span class="st"> </span><span class="kw">sqrt</span>((dx0 <span class="op">*</span><span class="st"> </span>dx)<span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>(dy0 <span class="op">*</span><span class="st"> </span>dy)<span class="op">^</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb250-15" data-line-number="15">    normal.x &lt;-<span class="st"> </span>dx0 <span class="op">-</span><span class="st"> </span>dx <span class="op">*</span><span class="st"> </span>parallel.l2</a>
<a class="sourceLine" id="cb250-16" data-line-number="16">    normal.y &lt;-<span class="st"> </span>dy0 <span class="op">-</span><span class="st"> </span>dy <span class="op">*</span><span class="st"> </span>parallel.l2</a>
<a class="sourceLine" id="cb250-17" data-line-number="17">    normal.l2 &lt;-<span class="st"> </span><span class="kw">sqrt</span>(normal.x<span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>normal.y<span class="op">^</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb250-18" data-line-number="18"></a>
<a class="sourceLine" id="cb250-19" data-line-number="19">    <span class="kw">which.max</span>(normal.l2)</a>
<a class="sourceLine" id="cb250-20" data-line-number="20">}</a>
<a class="sourceLine" id="cb250-21" data-line-number="21"></a>
<a class="sourceLine" id="cb250-22" data-line-number="22"><span class="co"># Percentage of variance explained is tucked away in the attributes.</span></a>
<a class="sourceLine" id="cb250-23" data-line-number="23">percent.var &lt;-<span class="st"> </span><span class="kw">attr</span>(<span class="kw">reducedDim</span>(sce.zeisel), <span class="st">&quot;percentVar&quot;</span>)</a>
<a class="sourceLine" id="cb250-24" data-line-number="24">chosen.elbow &lt;-<span class="st"> </span><span class="kw">elbowFinder</span>(percent.var)</a>
<a class="sourceLine" id="cb250-25" data-line-number="25">chosen.elbow</a></code></pre></div>
<pre><code>## [1] 6</code></pre>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb252-1" data-line-number="1"><span class="kw">plot</span>(percent.var, <span class="dt">xlab=</span><span class="st">&quot;PC&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance explained (%)&quot;</span>)</a>
<a class="sourceLine" id="cb252-2" data-line-number="2"><span class="kw">abline</span>(<span class="dt">v=</span>chosen.elbow, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:elbow"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/elbow-1.png" alt="Percentage of variance explained by successive PCs in the Zeisel brain data. The identified elbow point is marked with a red line." width="672" />
<p class="caption">
Figure 11.1: Percentage of variance explained by successive PCs in the Zeisel brain data. The identified elbow point is marked with a red line.
</p>
</div>
<p>Our assumption is that each of the top PCs capturing biological signal should explain much more variance than the remaining PCs.
Thus, there should be a sharp drop in the percentage of variance explained when we move past the last “biological” PC.
This manifests as an elbow in the scree plot, the location of which serves as a natural choice for <span class="math inline">\(d\)</span>.</p>
<p>From a practical perspective, the use of the elbow point tends to retain fewer PCs compared to other methods.
The definition of “much more variance” is relative so, in order to be retained, later PCs must explain a amount of variance that is comparable to that explained by the first few PCs.
Strong biological variation in the early PCs will shift the elbow to the left, potentially excluding weaker (but still interesting) variation in the next PCs immediately following the elbow.</p>
</div>
<div id="using-the-technical-noise" class="section level4">
<h4><span class="header-section-number">11.2.3.3</span> Using the technical noise</h4>
<p>Another strategy is to retain all PCs until the percentage of total variation explained reaches some threshold <span class="math inline">\(T\)</span>.
For example, one might retain the top set of PCs that explains 80% of the total variation in the data.
Of course, it would be pointless to swap one arbitrary parameter <span class="math inline">\(d\)</span> for another <span class="math inline">\(T\)</span>.
Instead, we define <span class="math inline">\(T\)</span> as the proportion of variance in the data that is attributed to the biological component.
This is done using the <code>denoisePCA()</code> function, which uses our estimates from <code>decomposeVar()</code> to determine <span class="math inline">\(d\)</span>.</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb253-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb253-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">001001001</span>)</a>
<a class="sourceLine" id="cb253-3" data-line-number="3"></a>
<a class="sourceLine" id="cb253-4" data-line-number="4"><span class="co"># Setting value=&quot;n&quot; to just report the number of PCs.</span></a>
<a class="sourceLine" id="cb253-5" data-line-number="5"><span class="co"># Setting max.rank=200 to circumvent the default max of 100.</span></a>
<a class="sourceLine" id="cb253-6" data-line-number="6"><span class="co"># Also setting BSPARAM to perform a fast SVD with IRLBA.</span></a>
<a class="sourceLine" id="cb253-7" data-line-number="7">denoised.n &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sce.zeisel, <span class="dt">technical=</span>dec.zeisel, <span class="dt">value=</span><span class="st">&#39;n&#39;</span>,</a>
<a class="sourceLine" id="cb253-8" data-line-number="8">    <span class="dt">subset.row=</span>chosen.hvgs, <span class="dt">BSPARAM=</span><span class="kw">IrlbaParam</span>(), <span class="dt">max.rank=</span><span class="dv">200</span>)</a>
<a class="sourceLine" id="cb253-9" data-line-number="9"><span class="kw">as.integer</span>(denoised.n)</a></code></pre></div>
<pre><code>## [1] 133</code></pre>
<p>This choice of <span class="math inline">\(d\)</span> represents the lower bound on the number of PCs required to retain all biological variation.
Any fewer PCs will definitely discard variance in the biological component.
In reality, the choice returned by <code>denoisePCA()</code> will only be optimal if the early PCs capture all the biological signal with minimal noise.
This is unlikely to be true as the PCA cannot distinguish between technical noise and weak biological signal in the later PCs.</p>
<p>From a practical perspective, the <code>denoisePCA()</code> approach retains more PCs than the elbow point method.
This is because the former does not compare PCs to each other and thus does not discard PCs corresponding to secondary factors of variation.
Whether this is a “better” outcome depends on the analyst’s willingness to increase noise in order to preserve weaker biological signals.</p>
</div>
<div id="based-on-population-structure" class="section level4">
<h4><span class="header-section-number">11.2.3.4</span> Based on population structure</h4>
<p>Yet another choice for <span class="math inline">\(d\)</span> uses information about the number of subpopulations in the data.
We assume that each subpopulation differs from the others along a different axis in the high-dimensional space.
This suggests that we should set <span class="math inline">\(d\)</span> to the number of unique subpopulations minus 1.
By doing so, we guarantee separation of all subpopulations while retaining as few dimensions (and noise) as possible.</p>
<p>The number of subpopulations is usually not known in advance so instead we use the number of clusters as a proxy.
However, this is itself dependent on the choice of <span class="math inline">\(d\)</span> as clustering is often performed on the PCs (see Chapter ???).
To handle this circularity, we test multiple choices for <span class="math inline">\(d\)</span> and we pick a value that is closest to the number of clusters minus 1.
This aims to achieve self-consistency in the number of PCs and clusters based on the reasoning above.</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb255-1" data-line-number="1">pcs &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(sce.zeisel)</a>
<a class="sourceLine" id="cb255-2" data-line-number="2">n &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dt">by=</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb255-3" data-line-number="3">collected &lt;-<span class="st"> </span><span class="kw">integer</span>(<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb255-4" data-line-number="4"></a>
<a class="sourceLine" id="cb255-5" data-line-number="5"><span class="co"># Using graph-based clustering here, but this can be</span></a>
<a class="sourceLine" id="cb255-6" data-line-number="6"><span class="co"># replaced with any clustering method of choice, provided</span></a>
<a class="sourceLine" id="cb255-7" data-line-number="7"><span class="co"># that the method determines the number of clusters by itself.</span></a>
<a class="sourceLine" id="cb255-8" data-line-number="8"><span class="cf">for</span> (i <span class="cf">in</span> n) {</a>
<a class="sourceLine" id="cb255-9" data-line-number="9">    g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(pcs[,<span class="kw">seq_len</span>(i)], <span class="dt">transposed=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb255-10" data-line-number="10">    clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb255-11" data-line-number="11">    collected &lt;-<span class="st"> </span><span class="kw">c</span>(collected, <span class="kw">length</span>(<span class="kw">unique</span>(clust)))</a>
<a class="sourceLine" id="cb255-12" data-line-number="12">}</a>
<a class="sourceLine" id="cb255-13" data-line-number="13"></a>
<a class="sourceLine" id="cb255-14" data-line-number="14"><span class="kw">data.frame</span>(<span class="dt">n.pcs=</span>n, <span class="dt">n.clusters=</span>collected)</a></code></pre></div>
<pre><code>##    n.pcs n.clusters
## 1     10         18
## 2     11         15
## 3     12         17
## 4     13         14
## 5     14         14
## 6     15         14
## 7     16         15
## 8     17         15
## 9     18         14
## 10    19         15
## 11    20         14</code></pre>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb257-1" data-line-number="1">n[collected<span class="dv">-1</span><span class="op">==</span>n]</a></code></pre></div>
<pre><code>## [1] 13</code></pre>
<p>This strategy is the most pragmatic as it directly addresses the role of the bias-variance trade-off in downstream analyses.
There is no need to preserve biological signal beyond what is distinguishable in later steps.
However, it involves strong assumptions about the nature of the biological differences between subpopulations - and indeed, discrete subpopulations may not even exist in studies of continuous processes like differentiation.</p>
</div>
</div>
<div id="putting-it-together" class="section level3">
<h3><span class="header-section-number">11.2.4</span> Putting it together</h3>
<p>Once we have chosen <span class="math inline">\(d\)</span>, enforcing it is as simple as subsetting the PC matrix by column.
We then use the <code>reducedDim()&lt;-</code> command to reassign the subsetted matrix back into the <code>SingleCellExperiment</code> object.
For example, if we were to take the top 20 PCs, we would do:</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb259-1" data-line-number="1"><span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA&quot;</span>) &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA&quot;</span>)[,<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]</a>
<a class="sourceLine" id="cb259-2" data-line-number="2"><span class="kw">ncol</span>(<span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA&quot;</span>))</a></code></pre></div>
<pre><code>## [1] 20</code></pre>
<p>Downstream applications that use the <code>&quot;PCA&quot;</code> dimensionality reduction results in <code>sce.zeisel</code> will subsequently operate on the first 20 PCs only.
Alternatively, some users may prefer to keep the full set of PCs, in which case the top set can be assigned to another name:</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb261-1" data-line-number="1"><span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA_20&quot;</span>) &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(sce.zeisel, <span class="st">&quot;PCA&quot;</span>)[,<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]</a>
<a class="sourceLine" id="cb261-2" data-line-number="2"><span class="kw">reducedDimNames</span>(sce.zeisel)</a></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;    &quot;IRLBA&quot;  &quot;RSVD&quot;   &quot;PCA_20&quot;</code></pre>
</div>
</div>
<div id="dimensionality-reduction-for-visualization" class="section level2">
<h2><span class="header-section-number">11.3</span> Dimensionality reduction for visualization</h2>
<div id="motivation-4" class="section level3">
<h3><span class="header-section-number">11.3.1</span> Motivation</h3>
<p>Another application of dimensionality reduction is to compress the data into 2 (sometimes 3) dimensions for plotting.
This serves a separate purpose to the PCA-based dimensionality reduction described above.
Algorithms are more than happy to operate on 10-50 PCs, but these are still too many dimensions for human comprehension.
Further dimensionality reduction strategies are required to pack the most salient features of the data into 2 or 3 dimensions, which we will discuss below.</p>
</div>
<div id="visualizating-with-pca" class="section level3">
<h3><span class="header-section-number">11.3.2</span> Visualizating with PCA</h3>
<p>The simplest visualization approach is to plot the top 2 PCs (Figure <a href="dimensionality-reduction-2.html#fig:zeisel-pca">11.2</a>):</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb263-1" data-line-number="1"><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:zeisel-pca"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/zeisel-pca-1.png" alt="PCA plot of the first two PCs in the Zeisel brain data. Each point is a cell, coloured according to the annotation provided by the original authors." width="672" />
<p class="caption">
Figure 11.2: PCA plot of the first two PCs in the Zeisel brain data. Each point is a cell, coloured according to the annotation provided by the original authors.
</p>
</div>
<p>The problem is that PCA is a linear technique, i.e., only variation along a line in high-dimensional space is captured by each PC.
As such, it cannot efficiently pack differences in <span class="math inline">\(d\)</span> dimensions into the first 2 PCs.
This is demonstrated in Figure <a href="dimensionality-reduction-2.html#fig:zeisel-pca">11.2</a> where the top two PCs fail to resolve some subpopulations identified by <span class="citation">Zeisel et al. (<a href="#ref-zeisel2015brain">2015</a>)</span>.
If the first PC is devoted to resolving the biggest difference between subpopulations, and the second PC is devoted to resolving the next biggest difference, then the remaining differences will not be visible in the plot.</p>
<p>One workaround is to plot several of the top PCs against each other in pairwise plots (Figure <a href="dimensionality-reduction-2.html#fig:zeisel-pca-multi">11.3</a>).
However, it is difficult to interpret multiple plots simultaneously, and even this approach is not sufficient to separate some of the annotated subpopulations.</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb264-1" data-line-number="1"><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">ncomponents=</span><span class="dv">4</span>,</a>
<a class="sourceLine" id="cb264-2" data-line-number="2">    <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:zeisel-pca-multi"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/zeisel-pca-multi-1.png" alt="PCA plot of the first two PCs in the Zeisel brain data. Each point is a cell, coloured according to the annotation provided by the original authors." width="672" />
<p class="caption">
Figure 11.3: PCA plot of the first two PCs in the Zeisel brain data. Each point is a cell, coloured according to the annotation provided by the original authors.
</p>
</div>
<p>There are some advantages to the PCA for visualization.
It is predictable and will not introduce artificial structure in the visualization.
It is also deterministic and robust to small changes in the input values.
However, as shown above, PCA is usually not satisfactory for visualization of complex populations.</p>
</div>
<div id="t-stochastic-neighbor-embedding" class="section level3">
<h3><span class="header-section-number">11.3.3</span> t-stochastic neighbor embedding</h3>
<p>The <em>de facto</em> standard for visualization of scRNA-seq data is the <span class="math inline">\(t\)</span>-stochastic neighbour embedding (<span class="math inline">\(t\)</span>-SNE) method <span class="citation">(Van der Maaten and Hinton <a href="#ref-van2008visualizing">2008</a>)</span>.
This attempts to find a low-dimensional representation of the data that preserves the distances between each point and its neighbours in the high-dimensional space.
Unlike PCA, it is not restricted to linear transformations, nor is it obliged to accurately represent distances between distance populations.
This means that it has much more freedom in how it arranges cells in low-dimensional space, enabling it to separate many distinct clusters in a complex population (Figure <a href="dimensionality-reduction-2.html#fig:tsne-brain">11.4</a>).</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb265-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">00101001101</span>)</a>
<a class="sourceLine" id="cb265-2" data-line-number="2">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb265-3" data-line-number="3"><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-brain"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/tsne-brain-1.png" alt="$t$-SNE plots constructed from the top PCs in the Zeisel brain dataset. Each point represents a cell, coloured according to the published annotation." width="672" />
<p class="caption">
Figure 11.4: <span class="math inline">\(t\)</span>-SNE plots constructed from the top PCs in the Zeisel brain dataset. Each point represents a cell, coloured according to the published annotation.
</p>
</div>
<p>One of the main disadvantages of <span class="math inline">\(t\)</span>-SNE is that it is much more computationally intensive than other visualization methods.
We mitigate this effect by performing the calculations on the top PCs with <code>use_dimred=&quot;PCA&quot;</code> in <code>runtTSNE()</code>.
This takes advantage of the PCA to compact the data and remove noise, followed by <span class="math inline">\(t\)</span>-SNE to create the visualization.
It is possible to run <span class="math inline">\(t\)</span>-SNE on the original expression matrix but this is less efficient.</p>
<p>Another issue with <span class="math inline">\(t\)</span>-SNE is that it requires the user to be aware of additional parameters (discussed <a href="http://distill.pub/2016/misread-tsne/">here</a> in some depth).
It involves a random initialization so we need to set the seed to ensure that the chosen results are reproducible.
The “perplexity” is another important parameter that determines the granularity of the visualization (Figure <a href="dimensionality-reduction-2.html#fig:tsne-perplexity">11.5</a>).
Low perplexities will favour resolution of finer structure, possibly to the point that the visualization is compromised by random noise.
Thus, it is advisable to test different perplexity values to ensure that the choice of perplexity does not drive the interpretation of the plot.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb266-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb266-2" data-line-number="2">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">perplexity=</span><span class="dv">5</span>)</a>
<a class="sourceLine" id="cb266-3" data-line-number="3">out5 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;TSNE&quot;</span>,</a>
<a class="sourceLine" id="cb266-4" data-line-number="4">    <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;perplexity = 5&quot;</span>)</a>
<a class="sourceLine" id="cb266-5" data-line-number="5"></a>
<a class="sourceLine" id="cb266-6" data-line-number="6"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb266-7" data-line-number="7">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">perplexity=</span><span class="dv">20</span>)</a>
<a class="sourceLine" id="cb266-8" data-line-number="8">out20 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;TSNE&quot;</span>,</a>
<a class="sourceLine" id="cb266-9" data-line-number="9">    <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;perplexity = 20&quot;</span>)</a>
<a class="sourceLine" id="cb266-10" data-line-number="10"></a>
<a class="sourceLine" id="cb266-11" data-line-number="11"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb266-12" data-line-number="12">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">perplexity=</span><span class="dv">80</span>)</a>
<a class="sourceLine" id="cb266-13" data-line-number="13">out80 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;TSNE&quot;</span>, </a>
<a class="sourceLine" id="cb266-14" data-line-number="14">    <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;perplexity = 80&quot;</span>)</a>
<a class="sourceLine" id="cb266-15" data-line-number="15"></a>
<a class="sourceLine" id="cb266-16" data-line-number="16"><span class="kw">multiplot</span>(out5, out20, out80, <span class="dt">cols=</span><span class="dv">3</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-perplexity"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/tsne-perplexity-1.png" alt="$t$-SNE plots constructed from the top PCs in the Zeisel brain dataset, using a range of perplexity values. Each point represents a cell, coloured according to its annotation." width="1440" />
<p class="caption">
Figure 11.5: <span class="math inline">\(t\)</span>-SNE plots constructed from the top PCs in the Zeisel brain dataset, using a range of perplexity values. Each point represents a cell, coloured according to its annotation.
</p>
</div>
<p>Despite its shortcomings, <span class="math inline">\(t\)</span>-SNE is proven tool for general-purpose visualization of scRNA-seq data.
It sees routine use in many analysis pipelines and will likely continue to do so for some time.</p>
</div>
<div id="uniform-manifold-approximation-and-projection" class="section level3">
<h3><span class="header-section-number">11.3.4</span> Uniform manifold approximation and projection</h3>
<p>The uniform manifold approximation and projection (UMAP) method <span class="citation">(<span class="citeproc-not-found" data-reference-id="mcinnes2018umap"><strong>???</strong></span>)</span> is an alternative to <span class="math inline">\(t\)</span>-SNE for non-linear dimensionality reduction.
It is roughly similar to <span class="math inline">\(t\)</span>-SNE in that it also tries to find a low-dimensional representation that preserves relationships between neighbors in high-dimensional space.
However, the two methods are based on different theory, represented by differences in the various graph weighting equations<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a>.
This manifests as a different visualization as shown in Figure <a href="dimensionality-reduction-2.html#fig:umap-brain">11.6</a>.</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb267-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1100101001</span>)</a>
<a class="sourceLine" id="cb267-2" data-line-number="2">sce.zeisel &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb267-3" data-line-number="3"><span class="kw">plotReducedDim</span>(sce.zeisel, <span class="dt">use_dimred=</span><span class="st">&quot;UMAP&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;level1class&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:umap-brain"></span>
<img src="P2_W07.reduced-dimensions_files/figure-html/umap-brain-1.png" alt="UMAP plots constructed from the top PCs in the Zeisel brain dataset. Each point represents a cell, coloured according to the published annotation." width="672" />
<p class="caption">
Figure 11.6: UMAP plots constructed from the top PCs in the Zeisel brain dataset. Each point represents a cell, coloured according to the published annotation.
</p>
</div>
<p>Compared to <span class="math inline">\(t\)</span>-SNE, the UMAP visualization tends to have more compact visual clusters with more empty space between them.
It also attempts to preserve more of the global structure than <span class="math inline">\(t\)</span>-SNE.
From a practical perspective, UMAP is much faster than <span class="math inline">\(t\)</span>-SNE, which may be an important consideration for large datasets.
(Nonetheless, we have still run UMAP on the top 20 PCs here for consistency.)
UMAP also involves a series of randomization steps so setting the seed is critical.</p>
<p>Like <span class="math inline">\(t\)</span>-SNE, UMAP has its own suite of hyperparameters that affect the visualization.
Of these, the number of neighbours (<code>n_neighbors</code>) and the minimum distance between embedded points (<code>min_dist</code>) have the greatest effect on the granularity of the output.
If these values are too low, random noise will be incorrectly treated as high-resolution structure, while values that are too high will discard fine structure altogether in favour of obtaining an accurate overview of the entire dataset.
Again, it is a good idea to test a range of values for these parameters to ensure that they do not compromise any conclusions drawn from a UMAP plot.</p>
<p>It is arguable whether the UMAP or <span class="math inline">\(t\)</span>-SNE visualizations are more useful or aesthetically pleasing.
However, it is clear that UMAP is much faster.
For that reason alone, it is increasingly displacing <span class="math inline">\(t\)</span>-SNE as the method of choice for visualizing large scRNA-seq data sets.</p>
</div>
<div id="interpreting-the-plots" class="section level3">
<h3><span class="header-section-number">11.3.5</span> Interpreting the plots</h3>
<p>Dimensionality reduction for visualization necessarily involves discarding information and distorting the distances between cells.
One might then ask whether the results of such extreme data compression can be trusted.
Indeed, it would probably be unwise to directly analyze the low-dimensional coordinates used for plotting.
Rather, the plots should be used to interpret/communicate the results of quantitative analyses based on a more accurate higher-rank representation of the data.</p>
<p>To illustrate, consider the interaction between clustering and <span class="math inline">\(t\)</span>-SNE.
As a general rule, we would not perform clustering on the <span class="math inline">\(t\)</span>-SNE coordinates.
Rather, we would cluster on the first 10-50 PCs (see Chapter ???) and then visualize the cluster identities on the <span class="math inline">\(t\)</span>-SNE plot.
This ensures that clustering makes use of the information that was lost during compression into two dimensions.
Given a discrepancy between the visible clusters on a <span class="math inline">\(t\)</span>-SNE plot and those identified by our clustering, we would be inclined to favour the latter.</p>
</div>
</div>
<div id="session-info-5" class="section level2">
<h2><span class="header-section-number">11.4</span> Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.6 LTS

Matrix products: default
BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scran_1.13.9                BiocSingular_1.1.5         
 [3] scater_1.13.9               ggplot2_3.2.0              
 [5] SingleCellExperiment_1.7.0  SummarizedExperiment_1.15.5
 [7] DelayedArray_0.11.4         BiocParallel_1.19.0        
 [9] matrixStats_0.54.0          Biobase_2.45.0             
[11] GenomicRanges_1.37.14       GenomeInfoDb_1.21.1        
[13] IRanges_2.19.10             S4Vectors_0.23.17          
[15] BiocGenerics_0.31.5         BiocStyle_2.13.2           
[17] Cairo_1.5-10               

loaded via a namespace (and not attached):
 [1] locfit_1.5-9.1           Rcpp_1.0.2              
 [3] rsvd_1.0.2               lattice_0.20-38         
 [5] assertthat_0.2.1         digest_0.6.20           
 [7] R6_2.4.0                 dynamicTreeCut_1.63-1   
 [9] evaluate_0.14            pillar_1.4.2            
[11] zlibbioc_1.31.0          rlang_0.4.0             
[13] lazyeval_0.2.2           irlba_2.3.3             
[15] Matrix_1.2-17            rmarkdown_1.14          
[17] BiocNeighbors_1.3.3      statmod_1.4.32          
[19] stringr_1.4.0            igraph_1.2.4.1          
[21] RCurl_1.95-4.12          munsell_0.5.0           
[23] compiler_3.6.0           vipor_0.4.5             
[25] xfun_0.8                 pkgconfig_2.0.2         
[27] ggbeeswarm_0.6.0         htmltools_0.3.6         
[29] tidyselect_0.2.5         tibble_2.1.3            
[31] gridExtra_2.3            GenomeInfoDbData_1.2.1  
[33] edgeR_3.27.9             bookdown_0.12           
[35] viridisLite_0.3.0        crayon_1.3.4            
[37] dplyr_0.8.3              withr_2.1.2             
[39] bitops_1.0-6             grid_3.6.0              
[41] gtable_0.3.0             magrittr_1.5            
[43] dqrng_0.2.1              scales_1.0.0            
[45] stringi_1.4.3            XVector_0.25.0          
[47] viridis_0.5.1            limma_3.41.15           
[49] DelayedMatrixStats_1.7.1 tools_3.6.0             
[51] glue_1.3.1               beeswarm_0.2.3          
[53] purrr_0.3.2              yaml_2.2.0              
[55] colorspace_1.4-1         BiocManager_1.30.4      
[57] knitr_1.23              </code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-van2008visualizing">
<p>Van der Maaten, L., and G. Hinton. 2008. “Visualizing Data Using T-SNE.” <em>J. Mach. Learn. Res.</em> 9 (2579-2605):85.</p>
</div>
<div id="ref-zeisel2015brain">
<p>Zeisel, A., A. B. Munoz-Manchado, S. Codeluppi, P. Lonnerberg, G. La Manno, A. Jureus, S. Marques, et al. 2015. “Brain structure. Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq.” <em>Science</em> 347 (6226):1138–42.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="10">
<li id="fn10"><p>Duh.<a href="dimensionality-reduction-2.html#fnref10" class="footnote-back">↩</a></p></li>
<li id="fn11"><p>Some of us leave coded messages in binary via the seeds. Just so you know.<a href="dimensionality-reduction-2.html#fnref11" class="footnote-back">↩</a></p></li>
<li id="fn12"><p>Or in other words: if the choice of <span class="math inline">\(d\)</span> is your biggest concern, you must lead a blessed life.<a href="dimensionality-reduction-2.html#fnref12" class="footnote-back">↩</a></p></li>
<li id="fn13"><p>That we will not discuss here, thank god.<a href="dimensionality-reduction-2.html#fnref13" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="feature-selection-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="clustering-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W07.reduced-dimensions.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
