<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-08-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="normalization.html"/>
<link rel="next" href="dimensionality-reduction-2.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="feature-selection-2.html"><a href="feature-selection-2.html"><i class="fa fa-check"></i><b>10</b> Feature selection</a><ul>
<li class="chapter" data-level="10.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>10.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="10.2.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>10.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="10.2.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#coefficient-of-variation"><i class="fa fa-check"></i><b>10.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="10.2.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#sec:spikeins"><i class="fa fa-check"></i><b>10.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="10.2.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>10.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#selecting-highly-variable-genes"><i class="fa fa-check"></i><b>10.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="10.3.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>10.3.1</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="10.3.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-a-fixed-threshold"><i class="fa fa-check"></i><b>10.3.2</b> Based on a fixed threshold</a></li>
<li class="chapter" data-level="10.3.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#keeping-all-genes-above-the-trend"><i class="fa fa-check"></i><b>10.3.3</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>10.4</b> Putting it all together</a></li>
<li class="chapter" data-level="" data-path="feature-selection-2.html"><a href="feature-selection-2.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="feature-selection-2.html"><a href="feature-selection-2.html#references-1"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html"><i class="fa fa-check"></i><b>11</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="11.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#overview-1"><i class="fa fa-check"></i><b>11.1</b> Overview</a></li>
<li class="chapter" data-level="11.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#principal-components-analysis"><i class="fa fa-check"></i><b>11.2</b> Principal components analysis</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#background-1"><i class="fa fa-check"></i><b>11.2.1</b> Background</a></li>
<li class="chapter" data-level="11.2.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#performing-the-pca"><i class="fa fa-check"></i><b>11.2.2</b> Performing the PCA</a></li>
<li class="chapter" data-level="11.2.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>11.2.3</b> Choosing the number of PCs</a></li>
<li class="chapter" data-level="11.2.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#putting-it-together"><i class="fa fa-check"></i><b>11.2.4</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>11.3</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="11.3.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#motivation-4"><i class="fa fa-check"></i><b>11.3.1</b> Motivation</a></li>
<li class="chapter" data-level="11.3.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#visualizating-with-pca"><i class="fa fa-check"></i><b>11.3.2</b> Visualizating with PCA</a></li>
<li class="chapter" data-level="11.3.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>11.3.3</b> t-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="11.3.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>11.3.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="11.3.5" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#interpreting-the-plots"><i class="fa fa-check"></i><b>11.3.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="clustering-1.html"><a href="clustering-1.html"><i class="fa fa-check"></i><b>12</b> Clustering</a><ul>
<li class="chapter" data-level="12.1" data-path="clustering-1.html"><a href="clustering-1.html#motivation-5"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="clustering-1.html"><a href="clustering-1.html#comments-on-truth"><i class="fa fa-check"></i><b>12.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="12.3" data-path="clustering-1.html"><a href="clustering-1.html#graph-based-clustering"><i class="fa fa-check"></i><b>12.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="12.3.1" data-path="clustering-1.html"><a href="clustering-1.html#background-2"><i class="fa fa-check"></i><b>12.3.1</b> Background</a></li>
<li class="chapter" data-level="12.3.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation"><i class="fa fa-check"></i><b>12.3.2</b> Implementation</a></li>
<li class="chapter" data-level="12.3.3" data-path="clustering-1.html"><a href="clustering-1.html#other-parameters"><i class="fa fa-check"></i><b>12.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="12.3.4" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>12.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="clustering-1.html"><a href="clustering-1.html#k-means-clustering"><i class="fa fa-check"></i><b>12.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="12.4.1" data-path="clustering-1.html"><a href="clustering-1.html#background-3"><i class="fa fa-check"></i><b>12.4.1</b> Background</a></li>
<li class="chapter" data-level="12.4.2" data-path="clustering-1.html"><a href="clustering-1.html#base-implementation"><i class="fa fa-check"></i><b>12.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="12.4.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>12.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="clustering-1.html"><a href="clustering-1.html#hierarchical-clustering"><i class="fa fa-check"></i><b>12.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="12.5.1" data-path="clustering-1.html"><a href="clustering-1.html#background-4"><i class="fa fa-check"></i><b>12.5.1</b> Background</a></li>
<li class="chapter" data-level="12.5.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation-1"><i class="fa fa-check"></i><b>12.5.2</b> Implementation</a></li>
<li class="chapter" data-level="12.5.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>12.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="clustering-1.html"><a href="clustering-1.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>13</b> Marker gene detection</a><ul>
<li class="chapter" data-level="13.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-6"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>13.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="13.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>13.2.1</b> Standard application</a></li>
<li class="chapter" data-level="13.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>13.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="13.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>13.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>13.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="13.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>13.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="13.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>13.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="13.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>13.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>13.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="13.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>13.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="13.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>13.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="13.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>13.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="13.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>13.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="13.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>13.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#references-2"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="doublet-detection.html"><a href="doublet-detection.html"><i class="fa fa-check"></i><b>14</b> Doublet detection</a><ul>
<li class="chapter" data-level="14.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-2"><i class="fa fa-check"></i><b>14.1</b> Overview</a></li>
<li class="chapter" data-level="14.2" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-with-clusters"><i class="fa fa-check"></i><b>14.2</b> Doublet detection with clusters</a><ul>
<li class="chapter" data-level="14.2.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-3"><i class="fa fa-check"></i><b>14.2.1</b> Overview</a></li>
<li class="chapter" data-level="14.2.2" data-path="doublet-detection.html"><a href="doublet-detection.html#application-to-the-mammary-data"><i class="fa fa-check"></i><b>14.2.2</b> Application to the mammary data</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-by-simulation"><i class="fa fa-check"></i><b>14.3</b> Doublet detection by simulation</a><ul>
<li class="chapter" data-level="14.3.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-4"><i class="fa fa-check"></i><b>14.3.1</b> Overview</a></li>
<li class="chapter" data-level="14.3.2" data-path="doublet-detection.html"><a href="doublet-detection.html#application-to-the-mammary-data-1"><i class="fa fa-check"></i><b>14.3.2</b> Application to the mammary data</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="doublet-detection.html"><a href="doublet-detection.html#strengths-and-weaknesses"><i class="fa fa-check"></i><b>14.4</b> Strengths and weaknesses</a></li>
<li class="chapter" data-level="14.5" data-path="doublet-detection.html"><a href="doublet-detection.html#further-comments"><i class="fa fa-check"></i><b>14.5</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="doublet-detection.html"><a href="doublet-detection.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="14.6" data-path="doublet-detection.html"><a href="doublet-detection.html#references-3"><i class="fa fa-check"></i><b>14.6</b> References</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>15</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="15.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#motivation-7"><i class="fa fa-check"></i><b>15.1</b> Motivation</a></li>
<li class="chapter" data-level="15.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>15.2</b> Setting up the data</a></li>
<li class="chapter" data-level="15.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#diagnosing-batch-effects"><i class="fa fa-check"></i><b>15.3</b> Diagnosing batch effects</a></li>
<li class="chapter" data-level="15.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#linear-regression"><i class="fa fa-check"></i><b>15.4</b> Linear regression</a></li>
<li class="chapter" data-level="15.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#performing-mnn-correction"><i class="fa fa-check"></i><b>15.5</b> Performing MNN correction</a><ul>
<li class="chapter" data-level="15.5.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#application-to-the-pbmc-data"><i class="fa fa-check"></i><b>15.5.1</b> Application to the PBMC data</a></li>
<li class="chapter" data-level="15.5.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#correction-diagnostics"><i class="fa fa-check"></i><b>15.5.2</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="15.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preserving-biological-heterogeneity"><i class="fa fa-check"></i><b>15.6</b> Preserving biological heterogeneity</a></li>
<li class="chapter" data-level="15.7" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#application-to-a-pancreas-dataset"><i class="fa fa-check"></i><b>15.7</b> Application to a pancreas dataset</a></li>
<li class="chapter" data-level="15.8" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#using-the-corrected-values"><i class="fa fa-check"></i><b>15.8</b> Using the corrected values</a></li>
<li class="chapter" data-level="" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#references-4"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html"><i class="fa fa-check"></i><b>16</b> Multi-sample comparisons</a><ul>
<li class="chapter" data-level="16.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-8"><i class="fa fa-check"></i><b>16.1</b> Motivation</a></li>
<li class="chapter" data-level="16.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#setting-up-the-data-2"><i class="fa fa-check"></i><b>16.2</b> Setting up the data</a></li>
<li class="chapter" data-level="16.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-expression-between-conditions"><i class="fa fa-check"></i><b>16.3</b> Differential expression between conditions</a><ul>
<li class="chapter" data-level="16.3.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#creating-pseudo-bulk-samples"><i class="fa fa-check"></i><b>16.3.1</b> Creating pseudo-bulk samples</a></li>
<li class="chapter" data-level="16.3.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-de-analysis"><i class="fa fa-check"></i><b>16.3.2</b> Performing the DE analysis</a></li>
<li class="chapter" data-level="16.3.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#putting-it-all-together-2"><i class="fa fa-check"></i><b>16.3.3</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-abundance-between-conditions"><i class="fa fa-check"></i><b>16.4</b> Differential abundance between conditions</a><ul>
<li class="chapter" data-level="16.4.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#overview-5"><i class="fa fa-check"></i><b>16.4.1</b> Overview</a></li>
<li class="chapter" data-level="16.4.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-da-analysis"><i class="fa fa-check"></i><b>16.4.2</b> Performing the DA analysis</a></li>
<li class="chapter" data-level="16.4.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#composition-effects"><i class="fa fa-check"></i><b>16.4.3</b> Handling composition effects</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#de-or-da-two-sides-of-the-same-coin"><i class="fa fa-check"></i><b>16.5</b> DE or DA? Two sides of the same coin</a></li>
<li class="chapter" data-level="16.6" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#avoiding-problems-with-ambient-rna"><i class="fa fa-check"></i><b>16.6</b> Avoiding problems with ambient RNA</a><ul>
<li class="chapter" data-level="16.6.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-9"><i class="fa fa-check"></i><b>16.6.1</b> Motivation</a></li>
<li class="chapter" data-level="16.6.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#discarding-ambient-degs"><i class="fa fa-check"></i><b>16.6.2</b> Discarding ambient DEGs</a></li>
<li class="chapter" data-level="16.6.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#subtracting-ambient-counts"><i class="fa fa-check"></i><b>16.6.3</b> Subtracting ambient counts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#references-5"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="interactive-sharing.html"><a href="interactive-sharing.html"><i class="fa fa-check"></i><b>17</b> Interactive Interfaces and Sharing</a><ul>
<li class="chapter" data-level="17.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#motivation-10"><i class="fa fa-check"></i><b>17.1</b> Motivation</a></li>
<li class="chapter" data-level="17.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#interactive-quickstart"><i class="fa fa-check"></i><b>17.2</b> Quick start</a></li>
<li class="chapter" data-level="17.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#isee-examples"><i class="fa fa-check"></i><b>17.3</b> Examples of usages for iSEE apps</a><ul>
<li class="chapter" data-level="17.3.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#quality-control-1"><i class="fa fa-check"></i><b>17.3.1</b> Quality control</a></li>
<li class="chapter" data-level="17.3.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#annotation-and-identification-of-cell-populations"><i class="fa fa-check"></i><b>17.3.2</b> Annotation and identification of cell populations</a></li>
<li class="chapter" data-level="17.3.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#collaborative-analysis"><i class="fa fa-check"></i><b>17.3.3</b> Collaborative analysis</a></li>
<li class="chapter" data-level="17.3.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#dissemination"><i class="fa fa-check"></i><b>17.3.4</b> Dissemination of analysis results</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#reproducibility"><i class="fa fa-check"></i><b>17.4</b> Reproducibility</a></li>
<li class="chapter" data-level="17.5" data-path="interactive-sharing.html"><a href="interactive-sharing.html#additional-resources"><i class="fa fa-check"></i><b>17.5</b> Additional resources</a></li>
<li class="chapter" data-level="" data-path="interactive-sharing.html"><a href="interactive-sharing.html#session-info-11"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="interactive-sharing.html"><a href="interactive-sharing.html#references-6"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>18</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="18.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>18.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="18.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>18.2</b> Parallelization</a></li>
<li class="chapter" data-level="18.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>18.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-12"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html"><i class="fa fa-check"></i><b>19</b> 416B Smart-seq2 dataset</a><ul>
<li class="chapter" data-level="19.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#introduction-2"><i class="fa fa-check"></i><b>19.1</b> Introduction</a></li>
<li class="chapter" data-level="19.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#analysis-code"><i class="fa fa-check"></i><b>19.2</b> Analysis code</a><ul>
<li class="chapter" data-level="19.2.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#data-loading"><i class="fa fa-check"></i><b>19.2.1</b> Data loading</a></li>
<li class="chapter" data-level="19.2.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#gene-annotation"><i class="fa fa-check"></i><b>19.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="19.2.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#quality-control-2"><i class="fa fa-check"></i><b>19.2.3</b> Quality control</a></li>
<li class="chapter" data-level="19.2.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#normalization-1"><i class="fa fa-check"></i><b>19.2.4</b> Normalization</a></li>
<li class="chapter" data-level="19.2.5" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#variance-modelling"><i class="fa fa-check"></i><b>19.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="19.2.6" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#batch-correction"><i class="fa fa-check"></i><b>19.2.6</b> Batch correction</a></li>
<li class="chapter" data-level="19.2.7" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>19.2.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="19.2.8" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#clustering-2"><i class="fa fa-check"></i><b>19.2.8</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#results"><i class="fa fa-check"></i><b>19.3</b> Results</a><ul>
<li class="chapter" data-level="19.3.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#quality-control-statistics"><i class="fa fa-check"></i><b>19.3.1</b> Quality control statistics</a></li>
<li class="chapter" data-level="19.3.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#normalization-2"><i class="fa fa-check"></i><b>19.3.2</b> Normalization</a></li>
<li class="chapter" data-level="19.3.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#variance-modelling-1"><i class="fa fa-check"></i><b>19.3.3</b> Variance modelling</a></li>
<li class="chapter" data-level="19.3.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#clustering-3"><i class="fa fa-check"></i><b>19.3.4</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#references-7"><i class="fa fa-check"></i><b>19.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html"><i class="fa fa-check"></i><b>20</b> Human pancreas Smart-seq2 dataset</a><ul>
<li class="chapter" data-level="20.1" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#introduction-3"><i class="fa fa-check"></i><b>20.1</b> Introduction</a></li>
<li class="chapter" data-level="20.2" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#analysis-code-1"><i class="fa fa-check"></i><b>20.2</b> Analysis code</a><ul>
<li class="chapter" data-level="20.2.1" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#data-loading-1"><i class="fa fa-check"></i><b>20.2.1</b> Data loading</a></li>
<li class="chapter" data-level="20.2.2" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#gene-annotation-1"><i class="fa fa-check"></i><b>20.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="20.2.3" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#sample-annotation"><i class="fa fa-check"></i><b>20.2.3</b> Sample annotation</a></li>
<li class="chapter" data-level="20.2.4" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#quality-control-3"><i class="fa fa-check"></i><b>20.2.4</b> Quality control</a></li>
<li class="chapter" data-level="20.2.5" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#normalization-3"><i class="fa fa-check"></i><b>20.2.5</b> Normalization</a></li>
<li class="chapter" data-level="20.2.6" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#variance-modelling-2"><i class="fa fa-check"></i><b>20.2.6</b> Variance modelling</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html"><i class="fa fa-check"></i><b>21</b> Human pancreas CEL-seq2 dataset</a><ul>
<li class="chapter" data-level="21.1" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#introduction-4"><i class="fa fa-check"></i><b>21.1</b> Introduction</a></li>
<li class="chapter" data-level="21.2" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#analysis-code-2"><i class="fa fa-check"></i><b>21.2</b> Analysis code</a><ul>
<li class="chapter" data-level="21.2.1" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#data-loading-2"><i class="fa fa-check"></i><b>21.2.1</b> Data loading</a></li>
<li class="chapter" data-level="21.2.2" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#gene-annotation-2"><i class="fa fa-check"></i><b>21.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="21.2.3" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#quality-control-4"><i class="fa fa-check"></i><b>21.2.3</b> Quality control</a></li>
<li class="chapter" data-level="21.2.4" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#normalization-4"><i class="fa fa-check"></i><b>21.2.4</b> Normalization</a></li>
<li class="chapter" data-level="21.2.5" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#variance-modelling-3"><i class="fa fa-check"></i><b>21.2.5</b> Variance modelling</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html"><i class="fa fa-check"></i><b>22</b> Chimeric embryo 10X dataset</a><ul>
<li class="chapter" data-level="22.1" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#introduction-5"><i class="fa fa-check"></i><b>22.1</b> Introduction</a></li>
<li class="chapter" data-level="22.2" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#analysis-code-3"><i class="fa fa-check"></i><b>22.2</b> Analysis code</a><ul>
<li class="chapter" data-level="22.2.1" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#data-loading-3"><i class="fa fa-check"></i><b>22.2.1</b> Data loading</a></li>
<li class="chapter" data-level="22.2.2" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#feature-annotation"><i class="fa fa-check"></i><b>22.2.2</b> Feature annotation</a></li>
<li class="chapter" data-level="22.2.3" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#quality-control-5"><i class="fa fa-check"></i><b>22.2.3</b> Quality control</a></li>
<li class="chapter" data-level="22.2.4" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#normalization-5"><i class="fa fa-check"></i><b>22.2.4</b> Normalization</a></li>
<li class="chapter" data-level="22.2.5" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#variance-modelling-4"><i class="fa fa-check"></i><b>22.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="22.2.6" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#merging"><i class="fa fa-check"></i><b>22.2.6</b> Merging</a></li>
<li class="chapter" data-level="22.2.7" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#clustering-4"><i class="fa fa-check"></i><b>22.2.7</b> Clustering</a></li>
<li class="chapter" data-level="22.2.8" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#dimensionality-reduction-4"><i class="fa fa-check"></i><b>22.2.8</b> Dimensionality reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html"><i class="fa fa-check"></i><b>23</b> Mammary gland dataset</a><ul>
<li class="chapter" data-level="23.1" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#introduction-6"><i class="fa fa-check"></i><b>23.1</b> Introduction</a></li>
<li class="chapter" data-level="23.2" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#analysis-code-4"><i class="fa fa-check"></i><b>23.2</b> Analysis code</a><ul>
<li class="chapter" data-level="23.2.1" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#data-loading-4"><i class="fa fa-check"></i><b>23.2.1</b> Data loading</a></li>
<li class="chapter" data-level="23.2.2" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#gene-annotation-3"><i class="fa fa-check"></i><b>23.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="23.2.3" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#quality-control-6"><i class="fa fa-check"></i><b>23.2.3</b> Quality control</a></li>
<li class="chapter" data-level="23.2.4" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#normalization-6"><i class="fa fa-check"></i><b>23.2.4</b> Normalization</a></li>
<li class="chapter" data-level="23.2.5" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#variance-modelling-5"><i class="fa fa-check"></i><b>23.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="23.2.6" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#dimensionality-reduction-5"><i class="fa fa-check"></i><b>23.2.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="23.2.7" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#clustering-5"><i class="fa fa-check"></i><b>23.2.7</b> Clustering</a></li>
<li class="chapter" data-level="23.2.8" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#marker-detection"><i class="fa fa-check"></i><b>23.2.8</b> Marker detection</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="24" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>24</b> About the Data</a><ul>
<li class="chapter" data-level="24.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>24.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="24.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>24.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>25</b> About the Contributors</a><ul>
<li class="chapter" data-level="25.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>25.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="25.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>25.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="25.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>25.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="25.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>25.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>26</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-selection-2" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Feature selection</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-2" class="section level2">
<h2><span class="header-section-number">10.1</span> Motivation</h2>
<p>We often use scRNA-seq data in exploratory analyses to characterize heterogeneity across cells.
Procedures like clustering and dimensionality reduction compare cells based on their gene expression profiles, which involves aggregating per-gene differences into a single (dis)similarity metric between a pair of cells.
The choice of genes to use in this calculation has a major impact on the behavior of the metric and the performance of downstream methods.
We want to select genes that contain useful information about the biology of the system while removing genes that contain random noise.
This aims to preserve interesting biological structure without the variance that obscures that structure.
It also reduces the size of the dataset to improve computational efficiency of later steps.</p>
<p>The simplest approach to feature selection is to select the most variable genes based on their expression across the population.
This assumes that genuine biological differences will manifest as increased variation in the affected genes, compared to other genes that are only affected by technical noise or a baseline level of “uninteresting” biological variation (e.g., from transcriptional bursting).
Several methods are available to quantify the variation per gene and to select the highly variable genes (HVGs), which we will discuss below.
For demonstration, we will use the 10X PBMC dataset:</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb203-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb203-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="st">&quot;raw_data&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb203-4" data-line-number="4">raw.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</a>
<a class="sourceLine" id="cb203-5" data-line-number="5">    <span class="st">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</a>
<a class="sourceLine" id="cb203-6" data-line-number="6"><span class="kw">untar</span>(raw.path, <span class="dt">exdir=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k&quot;</span>))</a>
<a class="sourceLine" id="cb203-7" data-line-number="7"></a>
<a class="sourceLine" id="cb203-8" data-line-number="8"><span class="kw">library</span>(DropletUtils)</a>
<a class="sourceLine" id="cb203-9" data-line-number="9">fname &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</a>
<a class="sourceLine" id="cb203-10" data-line-number="10">sce.pbmc &lt;-<span class="st"> </span><span class="kw">read10xCounts</span>(fname, <span class="dt">col.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb203-11" data-line-number="11"></a>
<a class="sourceLine" id="cb203-12" data-line-number="12">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb203-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb203-14" data-line-number="14"><span class="kw">rownames</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb203-15" data-line-number="15"></a>
<a class="sourceLine" id="cb203-16" data-line-number="16"><span class="kw">library</span>(EnsDb.Hsapiens.v86)</a>
<a class="sourceLine" id="cb203-17" data-line-number="17">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(EnsDb.Hsapiens.v86, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb203-18" data-line-number="18">    <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb203-19" data-line-number="19"></a>
<a class="sourceLine" id="cb203-20" data-line-number="20">### cell-detection <span class="al">###</span></a>
<a class="sourceLine" id="cb203-21" data-line-number="21"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb203-22" data-line-number="22">e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb203-23" data-line-number="23">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="kw">which</span>(e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>)]</a>
<a class="sourceLine" id="cb203-24" data-line-number="24"></a>
<a class="sourceLine" id="cb203-25" data-line-number="25">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb203-26" data-line-number="26">sce.pbmc &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce.pbmc, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(location<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb203-27" data-line-number="27">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.pbmc<span class="op">$</span>pct_counts_Mito, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb203-28" data-line-number="28">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb203-29" data-line-number="29"></a>
<a class="sourceLine" id="cb203-30" data-line-number="30">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb203-31" data-line-number="31"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb203-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb203-33" data-line-number="33">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb203-34" data-line-number="34">sce.pbmc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.pbmc, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">cluster=</span>clusters)</a>
<a class="sourceLine" id="cb203-35" data-line-number="35">sce.pbmc &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.pbmc)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb204-1" data-line-number="1">sce.pbmc</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 3922 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
## rowData names(10): ID Symbol ... total_counts log10_total_counts
## colnames(3922): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCACAGGTCCAC-1 TTTGTCATCCCAAGAT-1
## colData names(38): Sample Barcode ...
##   pct_counts_in_top_200_features_Mito
##   pct_counts_in_top_500_features_Mito
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):</code></pre>
<p>As well as the 416B dataset:</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb206-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb206-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb206-3" data-line-number="3">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">LunSpikeInData</span>(<span class="dt">which=</span><span class="st">&quot;416b&quot;</span>) </a>
<a class="sourceLine" id="cb206-4" data-line-number="4"></a>
<a class="sourceLine" id="cb206-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb206-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb206-7" data-line-number="7">ens.mm.v97 &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73905&quot;</span>]]</a>
<a class="sourceLine" id="cb206-8" data-line-number="8"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL &lt;-<span class="st"> </span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb206-9" data-line-number="9"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb206-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SYMBOL&quot;</span>)</a>
<a class="sourceLine" id="cb206-11" data-line-number="11"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb206-12" data-line-number="12">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>)</a>
<a class="sourceLine" id="cb206-13" data-line-number="13"></a>
<a class="sourceLine" id="cb206-14" data-line-number="14"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb206-15" data-line-number="15"><span class="kw">rownames</span>(sce<span class="fl">.416</span>b) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL, </a>
<a class="sourceLine" id="cb206-16" data-line-number="16">    <span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL)</a>
<a class="sourceLine" id="cb206-17" data-line-number="17"></a>
<a class="sourceLine" id="cb206-18" data-line-number="18">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb206-19" data-line-number="19">mito &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME<span class="op">==</span><span class="st">&quot;MT&quot;</span>)</a>
<a class="sourceLine" id="cb206-20" data-line-number="20">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce<span class="fl">.416</span>b, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mt=</span>mito))</a>
<a class="sourceLine" id="cb206-21" data-line-number="21">qc &lt;-<span class="st"> </span><span class="kw">quickCellQC</span>(stats, <span class="dt">percent_subsets=</span><span class="st">&quot;subsets_Mt_percent&quot;</span>)</a>
<a class="sourceLine" id="cb206-22" data-line-number="22">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span>sce<span class="fl">.416</span>b[,<span class="op">!</span>qc<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb206-23" data-line-number="23"></a>
<a class="sourceLine" id="cb206-24" data-line-number="24">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb206-25" data-line-number="25"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb206-26" data-line-number="26">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb206-27" data-line-number="27">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce<span class="fl">.416</span>b)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" data-line-number="1">sce<span class="fl">.416</span>b</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 46604 187 
## metadata(0):
## assays(2): counts logcounts
## rownames(46604): 4933401J01Rik Gm26206 ... CAAA01147332.1
##   CBFB-MYH11-mcherry
## rowData names(4): Length ENSEMBL SYMBOL SEQNAME
## colnames(187): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1
##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...
##   SLX-11312.N712_S507.H5H5YBBXX.s_8.r_1
##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1
## colData names(9): Source Name cell line ... spike-in addition
##   block
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(2): ERCC SIRV</code></pre>
</div>
<div id="quantifying-per-gene-variation" class="section level2">
<h2><span class="header-section-number">10.2</span> Quantifying per-gene variation</h2>
<div id="variance-of-the-log-counts" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Variance of the log-counts</h3>
<p>The simplest approach to quantifying per-gene variation is to simply compute the variance of the log-normalized expression values (referred to as “log-counts” for simplicity) for each gene across all cells in the population <span class="citation">(<span class="citeproc-not-found" data-reference-id="lun2016step"><strong>???</strong></span>)</span>.
This has an advantage in that the feature selection is based on the same log-values that are used for later downstream steps.
Genes with the largest variances in log-values will contribute the most to the Euclidean distances between cells.
By using log-values here, we ensure that our quantitative definition of heterogeneity is consistent throughout the entire analysis.</p>
<p>Calculation of the per-gene variance is simple but feature selection requires modelling of the mean-variance relationship.
As discussed briefly in the last chapter, the log-transformation does not achieve perfect variance stabilization.
This means that the variance of a gene is more affected by its abundance than the underlying biological heterogeneity.
To account for this effect, we use the <code>modelGeneVar()</code> function to fit a trend to the variance with respect to abundance across all genes (Figure <a href="feature-selection-2.html#fig:trend-plot-pbmc">10.1</a>).</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb209-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb209-2" data-line-number="2">dec.pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb209-3" data-line-number="3"></a>
<a class="sourceLine" id="cb209-4" data-line-number="4"><span class="co"># Visualizing the fit:</span></a>
<a class="sourceLine" id="cb209-5" data-line-number="5">fit.pbmc &lt;-<span class="st"> </span><span class="kw">metadata</span>(dec.pbmc)</a>
<a class="sourceLine" id="cb209-6" data-line-number="6"><span class="kw">plot</span>(fit.pbmc<span class="op">$</span>mean, fit.pbmc<span class="op">$</span>var, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb209-7" data-line-number="7">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb209-8" data-line-number="8"><span class="kw">curve</span>(fit.pbmc<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:trend-plot-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/trend-plot-pbmc-1.png" alt="Variance in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the trend fitted to all genes." width="672" />
<p class="caption">
Figure 10.1: Variance in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the trend fitted to all genes.
</p>
</div>
<p>At any given abundance, we assume that the expression profiles of most genes are dominated by random technical noise (see <a href="feature-selection-2.html#sec:spikeins">10.2.3</a> for details).
Under this assumption, our trend represents an estimate of the technical noise as a function of abundance.
We then break down the total variance of each gene into the technical component, i.e., the fitted value of the trend at that gene’s abundance;
and the biological component, defined as the difference between the total variance and the technical component.
This biological component represents the “interesting” variation for each gene and can be used as the metric for HVG selection.</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb210-1" data-line-number="1"><span class="co"># Ordering by most interesting genes for inspection.</span></a>
<a class="sourceLine" id="cb210-2" data-line-number="2">dec.pbmc[<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),] </a></code></pre></div>
<pre><code>## DataFrame with 33694 rows and 6 columns
##                     mean             total              tech
##                &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;
## LYZ     1.97720774557909   5.1093189779025 0.829103126835094
## S100A9  1.94951728004377  4.58067979752236 0.829355453195275
## S100A8  1.71813080049068    4.450555343632  0.82113988580669
## HLA-DRA 2.09809035649377  3.73632552542342 0.825379686156474
## CD74    2.89964299605066  3.32217806233448 0.794847543782114
## ...                  ...               ...               ...
## PTMA     3.8310477802539 0.473973817102772 0.742787431378562
## HLA-B   4.50257580785005 0.477758219250801 0.758429136842052
## EIF1    3.24339798459787 0.477492332280039 0.773251499236556
## TMSB4X  6.08574497419255 0.405491648922873 0.748371857455256
## B2M     5.95574224289044 0.303964766423101 0.720380629718446
##                        bio               p.value                   FDR
##                  &lt;numeric&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## LYZ        4.2802158510674 1.21228447173797e-263 1.19106949348255e-259
## S100A9    3.75132434432709 5.94502641052638e-203 3.89399229889478e-199
## S100A8    3.62941545782531 6.64639337130205e-194 3.26504074365213e-190
## HLA-DRA   2.91094583926695 2.66248562787844e-124 4.02444942983164e-121
## CD74      2.52733051855236 1.85527889318145e-101  1.73601096433407e-98
## ...                    ...                   ...                   ...
## PTMA     -0.26881361427579     0.992456099633289     0.999999999334437
## HLA-B   -0.280670917591251     0.993524809206667     0.999999999334437
## EIF1    -0.295759166956516     0.994893426785729     0.999999999334437
## TMSB4X  -0.342880208532382     0.998953636877237     0.999999999334437
## B2M     -0.416415863295345     0.999948154914924     0.999999999334437</code></pre>
<pre><code>## &lt;numeric&gt;

## 1.97720774557909

## 1.94951728004377

## 1.71813080049068

## 2.09809035649377

## 2.89964299605066

## ...

## 3.8310477802539

## 4.50257580785005

## 3.24339798459787

## 6.08574497419255

## 5.95574224289044

## &lt;numeric&gt;

## 5.1093189779025

## 4.58067979752236

## 4.450555343632

## 3.73632552542342

## 3.32217806233448

## ...

## 0.473973817102772

## 0.477758219250801

## 0.477492332280039

## 0.405491648922873

## 0.303964766423101

## &lt;numeric&gt;

## 0.829103126835094

## 0.829355453195275

## 0.82113988580669

## 0.825379686156474

## 0.794847543782114

## ...

## 0.742787431378562

## 0.758429136842052

## 0.773251499236556

## 0.748371857455256

## 0.720380629718446

## &lt;numeric&gt;

## 4.2802158510674

## 3.75132434432709

## 3.62941545782531

## 2.91094583926695

## 2.52733051855236

## ...

## -0.26881361427579

## -0.280670917591251

## -0.295759166956516

## -0.342880208532382

## -0.416415863295345

## &lt;numeric&gt;

## 1.21228447173797e-263

## 5.94502641052638e-203

## 6.64639337130205e-194

## 2.66248562787844e-124

## 1.85527889318145e-101

## ...

## 0.992456099633289

## 0.993524809206667

## 0.994893426785729

## 0.998953636877237

## 0.999948154914924

## &lt;numeric&gt;

## 1.19106949348255e-259

## 3.89399229889478e-199

## 3.26504074365213e-190

## 4.02444942983164e-121

## 1.73601096433407e-98

## ...

## 0.999999999334437

## 0.999999999334437

## 0.999999999334437

## 0.999999999334437

## 0.999999999334437</code></pre>
<p>(Careful readers will notice that some genes have negative biological components.
Negative components of variation have no obvious interpretation and can be ignored for most applications.
They are inevitable when fitting a trend to the per-gene variances, as approximately half of the genes will lie below the trend.)</p>
</div>
<div id="coefficient-of-variation" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Coefficient of variation</h3>
<p>An alternative approach to quantification uses the squared coefficient of variation (CV<sup>2</sup>) of the normalized expression values prior to log-transformation.
The CV<sup>2</sup> is a widely used metric for describing variation in non-negative data and is closely related to the dispersion parameter of the negative binomial distribution in packages like <em><a href="https://bioconductor.org/packages/3.10/edgeR">edgeR</a></em> and <em><a href="https://bioconductor.org/packages/3.10/DESeq2">DESeq2</a></em>.
We compute the CV<sup>2</sup> for each gene in the PBMC dataset using the <code>modelGeneCV2()</code> function, which provides a robust implementation of the approach described by <span class="citation">Brennecke et al. (<a href="#ref-brennecke2013accounting">2013</a>)</span>.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb213-1" data-line-number="1">dec.cv2.pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneCV2</span>(sce.pbmc)</a></code></pre></div>
<p>This allows us to model the mean-variance relationship when considering the relevance of each gene (Figure <a href="feature-selection-2.html#fig:cv2-pbmc">10.2</a>).
Again, our assumption is that most genes contain random noise and that the trend captures mostly technical variation.
Large CV<span class="math inline">\(^2\)</span> values that deviate strongly from the trend are likely to represent genes affected by biological structure.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb214-1" data-line-number="1">fit.cv2.pbmc &lt;-<span class="st"> </span><span class="kw">metadata</span>(dec.cv2.pbmc)</a>
<a class="sourceLine" id="cb214-2" data-line-number="2"><span class="kw">plot</span>(fit.cv2.pbmc<span class="op">$</span>mean, fit.cv2.pbmc<span class="op">$</span>cv2, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>)</a>
<a class="sourceLine" id="cb214-3" data-line-number="3"><span class="kw">curve</span>(fit.cv2.pbmc<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:cv2-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/cv2-pbmc-1.png" alt="CV^2^ in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the fitted trend." width="672" />
<p class="caption">
Figure 10.2: CV<sup>2</sup> in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the fitted trend.
</p>
</div>
<p>For each gene, we quantify the deviation from the trend in terms of the ratio of its CV<sup>2</sup> to the fitted value of trend at its abundance.
This is more appropriate than the directly subtracting the trend from the CV<sup>2</sup>, as the magnitude of the ratio is not affected by the mean.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb215-1" data-line-number="1">dec.cv2.pbmc[<span class="kw">order</span>(dec.cv2.pbmc<span class="op">$</span>ratio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 33694 rows and 5 columns
##                          mean            total            trend
##                     &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;
## HIST1H2AC   0.903832087554474  267.52946465042 1.56249347165887
## GNG11       0.689915788093492 219.246474356754 1.98426985636371
## PRTFDC1    0.0411712190890851 3039.89950832654 30.0648545637562
## TNNC2       0.102005431273823 1211.47780026804 12.2552987937789
## PF4          1.10708448892009  128.81923174922 1.31275273661753
## ...                       ...              ...              ...
## AC023491.2                  0              NaN              Inf
## AC233755.2                  0              NaN              Inf
## AC233755.1                  0              NaN              Inf
## AC213203.1                  0              NaN              Inf
## FAM231B                     0              NaN              Inf
##                       ratio   p.value
##                   &lt;numeric&gt; &lt;numeric&gt;
## HIST1H2AC  171.219572755328         0
## GNG11      110.492266792046         0
## PRTFDC1    101.111399088263         0
## TNNC2      98.8533874737529         0
## PF4         98.129090236093         0
## ...                     ...       ...
## AC023491.2              NaN       NaN
## AC233755.2              NaN       NaN
## AC233755.1              NaN       NaN
## AC213203.1              NaN       NaN
## FAM231B                 NaN       NaN</code></pre>
<pre><code>## &lt;numeric&gt;

## 0.903832087554474

## 0.689915788093492

## 0.0411712190890851

## 0.102005431273823

## 1.10708448892009

## ...

## 0

## 0

## 0

## 0

## 0

## &lt;numeric&gt;

## 267.52946465042

## 219.246474356754

## 3039.89950832654

## 1211.47780026804

## 128.81923174922

## ...

## NaN

## NaN

## NaN

## NaN

## NaN

## &lt;numeric&gt;

## 1.56249347165887

## 1.98426985636371

## 30.0648545637562

## 12.2552987937789

## 1.31275273661753

## ...

## Inf

## Inf

## Inf

## Inf

## Inf

## &lt;numeric&gt;

## 171.219572755328

## 110.492266792046

## 101.111399088263

## 98.8533874737529

## 98.129090236093

## ...

## NaN

## NaN

## NaN

## NaN

## NaN

## &lt;numeric&gt;

## 0

## 0

## 0

## 0

## 0

## ...

## NaN

## NaN

## NaN

## NaN

## NaN</code></pre>
<p>Both the CV<sup>2</sup> and the variance of log-counts are effective metrics for quantifying variation in gene expression.
The CV<sup>2</sup> tends to give higher rank to low-abundance HVGs driven by upregulation in rare subpopulations, for which the increase in variance on the raw scale is stronger than that on the log-scale.
However, the variation described by the CV<sup>2</sup> is less directly relevant to downstream procedures operating on the log-counts, and the reliance on the ratio can assign high rank to uninteresting genes with low absolute variance.
We generally prefer the use of the variance of log-counts and will use it in the following sections, though the many of the same principles apply to procedures based on the CV<sup>2</sup>.</p>
</div>
<div id="sec:spikeins" class="section level3">
<h3><span class="header-section-number">10.2.3</span> Quantifying technical noise</h3>
<p>The use of a trend fitted to endogenous genes assumes that the expression profiles of most genes are dominated by random technical noise.
In practice, all expressed genes will exhibit some non-zero level of biological variability due to events like transcriptional bursting.
This suggests that our estimates of the technical component are likely to be inflated.
It would be more appropriate to consider these estimates as technical noise plus “uninteresting” biological variation, under the assumption that most genes are unaffected by the relevant heterogeneity in the population.</p>
<p>This assumption is generally reasonable but may be problematic in some scenarios where many genes at a particular abundance are affected by a biological process.
For example, strong upregulation of cell type-specific genes may result in an enrichment of HVGs at high abundances.
This would inflate the fitted trend at high abundances and compromise the detection of the affected genes.
We can avoid this problem by fitting a mean-dependent trend to the variance of the spike-in transcripts, if they are available (Figure <a href="feature-selection-2.html#fig:spike-416b">10.3</a>).
The premise here is that spike-ins should not be affected by biological variation, so the fitted value of the spike-in trend should represent a better estimate of the technical component for each gene.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb218-1" data-line-number="1">dec.spike<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;ERCC&quot;</span>)</a>
<a class="sourceLine" id="cb218-2" data-line-number="2">dec.spike<span class="fl">.416</span>b[<span class="kw">order</span>(dec.spike<span class="fl">.416</span>b<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 46604 rows and 6 columns
##                      mean             total             tech
##                 &lt;numeric&gt;         &lt;numeric&gt;        &lt;numeric&gt;
## Lyz2     6.58452322626962   13.981115391828 1.57439813967028
## Ccl9     6.71231727838975  13.2123141538704 1.44517363924406
## Top2a    5.78674433920395   14.061964981076 2.55485592247822
## Cd200r3  4.87041205832691  15.5629113090432 4.12756395747085
## Ccnb2    5.90621334470428  13.3668217298191 2.37519789726572
## ...                   ...               ...              ...
## Gm8623   2.99687811926373 0.906393458047858 6.58595054641976
## Rpl5-ps2 3.60016942615025 0.611220853516254 6.32496226060295
## Gm26825  3.22224019884412 0.865862049948253 6.58121590339427
## Gm12816  2.91225548992418   0.8379080485496 6.56932341240521
## Rps12l1  3.14861797000057 0.741526126629146 6.59206832695328
##                        bio               p.value                   FDR
##                  &lt;numeric&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## Lyz2      12.4067172521577 2.67876851201925e-198 2.65689190250442e-195
## Ccl9      11.7671405146263 1.51573651575589e-211 1.71812342957396e-208
## Top2a     11.5071090585978  2.73161223403655e-66  7.22481084655623e-64
## Cd200r3   11.4353473515723  2.42835264947996e-26  1.64685203613165e-24
## Ccnb2     10.9916238325534  7.35199139996513e-70  2.16057781833049e-67
## ...                    ...                   ...                   ...
## Gm8623    -5.6795570883719     0.999490513869684     0.999795028709123
## Rpl5-ps2  -5.7137414070867     0.999710626616979     0.999795028709123
## Gm26825  -5.71535385344601     0.999530711943699     0.999795028709123
## Gm12816  -5.73141536385561     0.999555718435578     0.999795028709123
## Rps12l1  -5.85054220032413     0.999638884480882     0.999795028709123</code></pre>
<pre><code>## &lt;numeric&gt;

## 6.58452322626962

## 6.71231727838975

## 5.78674433920395

## 4.87041205832691

## 5.90621334470428

## ...

## 2.99687811926373

## 3.60016942615025

## 3.22224019884412

## 2.91225548992418

## 3.14861797000057

## &lt;numeric&gt;

## 13.981115391828

## 13.2123141538704

## 14.061964981076

## 15.5629113090432

## 13.3668217298191

## ...

## 0.906393458047858

## 0.611220853516254

## 0.865862049948253

## 0.8379080485496

## 0.741526126629146

## &lt;numeric&gt;

## 1.57439813967028

## 1.44517363924406

## 2.55485592247822

## 4.12756395747085

## 2.37519789726572

## ...

## 6.58595054641976

## 6.32496226060295

## 6.58121590339427

## 6.56932341240521

## 6.59206832695328

## &lt;numeric&gt;

## 12.4067172521577

## 11.7671405146263

## 11.5071090585978

## 11.4353473515723

## 10.9916238325534

## ...

## -5.6795570883719

## -5.7137414070867

## -5.71535385344601

## -5.73141536385561

## -5.85054220032413

## &lt;numeric&gt;

## 2.67876851201925e-198

## 1.51573651575589e-211

## 2.73161223403655e-66

## 2.42835264947996e-26

## 7.35199139996513e-70

## ...

## 0.999490513869684

## 0.999710626616979

## 0.999530711943699

## 0.999555718435578

## 0.999638884480882

## &lt;numeric&gt;

## 2.65689190250442e-195

## 1.71812342957396e-208

## 7.22481084655623e-64

## 1.64685203613165e-24

## 2.16057781833049e-67

## ...

## 0.999795028709123

## 0.999795028709123

## 0.999795028709123

## 0.999795028709123

## 0.999795028709123</code></pre>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb221-1" data-line-number="1"><span class="kw">plot</span>(dec.spike<span class="fl">.416</span>b<span class="op">$</span>mean, dec.spike<span class="fl">.416</span>b<span class="op">$</span>total, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb221-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb221-3" data-line-number="3">fit.spike<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">metadata</span>(dec.spike<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb221-4" data-line-number="4"><span class="kw">points</span>(fit.spike<span class="fl">.416</span>b<span class="op">$</span>mean, fit.spike<span class="fl">.416</span>b<span class="op">$</span>var, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</a>
<a class="sourceLine" id="cb221-5" data-line-number="5"><span class="kw">curve</span>(fit.spike<span class="fl">.416</span>b<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:spike-416b"></span>
<img src="P2_W06.feature-selection_files/figure-html/spike-416b-1.png" alt="Variance in the 416B data set as a function of the mean. Each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins." width="672" />
<p class="caption">
Figure 10.3: Variance in the 416B data set as a function of the mean. Each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins.
</p>
</div>
<p>In the absence of spike-in data, one can attempt to create a trend by making some distributional assumptions about the noise.
For example, UMI counts typically exhibit near-Poisson noise when only technical effects are considered.
This can be used to construct a mean-variance trend in the log-counts (Figure <a href="feature-selection-2.html#fig:tech-pbmc">10.4</a>) with the <code>modelGeneVarByPoisson()</code> function.
Note the increased residuals of the high-abundance genes, which can be interpreted as the amount of biological variation that was assumed to be “uninteresting” when fitting the gene-based trend in Figure <a href="feature-selection-2.html#fig:trend-plot-pbmc">10.1</a>.</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb222-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">0010101</span>)</a>
<a class="sourceLine" id="cb222-2" data-line-number="2">dec.pois.pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneVarByPoisson</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb222-3" data-line-number="3">dec.pois.pbmc &lt;-<span class="st"> </span>dec.pois.pbmc[<span class="kw">order</span>(dec.pois.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a>
<a class="sourceLine" id="cb222-4" data-line-number="4"><span class="kw">head</span>(dec.pois.pbmc)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                     mean            total              tech
##                &lt;numeric&gt;        &lt;numeric&gt;         &lt;numeric&gt;
## LYZ     1.97720774557909  5.1093189779025 0.625920315505478
## S100A9  1.94951728004377 4.58067979752236 0.631472547597445
## S100A8  1.71813080049068   4.450555343632 0.672535358554497
## HLA-DRA 2.09809035649377 3.73632552542342 0.600680256885254
## CD74    2.89964299605066 3.32217806233448 0.425592152350886
## CST3     1.4921009978511 2.96883317240039 0.697923385529405
##                      bio   p.value       FDR
##                &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## LYZ     4.48339866239702         0         0
## S100A9  3.94920724992492         0         0
## S100A8  3.77801998507751         0         0
## HLA-DRA 3.13564526853816         0         0
## CD74    2.89658590998359         0         0
## CST3    2.27090978687099         0         0</code></pre>
<pre><code>## &lt;numeric&gt;

## 1.97720774557909

## 1.94951728004377

## 1.71813080049068

## 2.09809035649377

## 2.89964299605066

## 1.4921009978511

## &lt;numeric&gt;

## 5.1093189779025

## 4.58067979752236

## 4.450555343632

## 3.73632552542342

## 3.32217806233448

## 2.96883317240039

## &lt;numeric&gt;

## 0.625920315505478

## 0.631472547597445

## 0.672535358554497

## 0.600680256885254

## 0.425592152350886

## 0.697923385529405

## &lt;numeric&gt;

## 4.48339866239702

## 3.94920724992492

## 3.77801998507751

## 3.13564526853816

## 2.89658590998359

## 2.27090978687099

## &lt;numeric&gt;

## 0

## 0

## 0

## 0

## 0

## 0

## &lt;numeric&gt;

## 0

## 0

## 0

## 0

## 0

## 0</code></pre>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb225-1" data-line-number="1"><span class="kw">plot</span>(dec.pois.pbmc<span class="op">$</span>mean, dec.pois.pbmc<span class="op">$</span>total, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb225-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb225-3" data-line-number="3"><span class="kw">curve</span>(<span class="kw">metadata</span>(dec.pois.pbmc)<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tech-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/tech-pbmc-1.png" alt="Variance of normalized log-expression values for each gene in the PBMC dataset, plotted against the mean log-expression. The blue line represents represents the mean-variance relationship corresponding to Poisson noise." width="672" />
<p class="caption">
Figure 10.4: Variance of normalized log-expression values for each gene in the PBMC dataset, plotted against the mean log-expression. The blue line represents represents the mean-variance relationship corresponding to Poisson noise.
</p>
</div>
</div>
<div id="accounting-for-blocking-factors" class="section level3">
<h3><span class="header-section-number">10.2.4</span> Accounting for blocking factors</h3>
<div id="fitting-block-specific-trends" class="section level4">
<h4><span class="header-section-number">10.2.4.1</span> Fitting block-specific trends</h4>
<p>Data containing multiple batches will often exhibit batch effects (see Chapter ??? for more details).
We are (usually) not interested in HVGs that are driven by batch effects.
Rather, we want to focus on genes that are highly variable within each batch.
This is naturally achieved by performing trend fitting and variance decomposition separately for each batch.
We demonstrate this approach by treating each plate (<code>block</code>) in the 416B dataset as a different batch, using the <code>modelGeneVarWithSpikes()</code> function.
(The same argument is available in all other variance-modelling functions.)</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb226-1" data-line-number="1">dec.block<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;ERCC&quot;</span>, <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb226-2" data-line-number="2"><span class="kw">head</span>(dec.block<span class="fl">.416</span>b[<span class="kw">order</span>(dec.block<span class="fl">.416</span>b<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>])</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                     mean            total             tech
##                &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;
## Lyz2     6.5888839665632 13.9879587012246 1.59005301830697
## Ccl9    6.71152694096627 13.2838732503136 1.38839690248626
## Cd200r3 4.87339059420103 15.6086389540576 4.18071722807439
## Top2a   5.79434982405204 13.8554180869451  2.7386542626093
## Ccnb2   5.91377600206442 13.1732321163212  2.5544833544482
## Hbb-bt  4.90872275033472 14.7067184567165 4.09107979580209
##                      bio               p.value                   FDR
##                &lt;numeric&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## Lyz2    12.3979056829176                     0                     0
## Ccl9    11.8954763478274                     0                     0
## Cd200r3 11.4279217259832  1.26914883229919e-55  8.09941522896778e-54
## Top2a   11.1167638243358  1.7661124143504e-127 3.65569912271278e-125
## Ccnb2    10.618748761873 4.39888780633588e-140 1.05768813476787e-137
## Hbb-bt  10.6156386609144  1.00779070037796e-47  5.35478790888324e-46</code></pre>
<pre><code>## &lt;numeric&gt;

## 6.5888839665632

## 6.71152694096627

## 4.87339059420103

## 5.79434982405204

## 5.91377600206442

## 4.90872275033472

## &lt;numeric&gt;

## 13.9879587012246

## 13.2838732503136

## 15.6086389540576

## 13.8554180869451

## 13.1732321163212

## 14.7067184567165

## &lt;numeric&gt;

## 1.59005301830697

## 1.38839690248626

## 4.18071722807439

## 2.7386542626093

## 2.5544833544482

## 4.09107979580209

## &lt;numeric&gt;

## 12.3979056829176

## 11.8954763478274

## 11.4279217259832

## 11.1167638243358

## 10.618748761873

## 10.6156386609144

## &lt;numeric&gt;

## 0

## 0

## 1.26914883229919e-55

## 1.7661124143504e-127

## 4.39888780633588e-140

## 1.00779070037796e-47

## &lt;numeric&gt;

## 0

## 0

## 8.09941522896778e-54

## 3.65569912271278e-125

## 1.05768813476787e-137

## 5.35478790888324e-46</code></pre>
<p>The use of a batch-specific trend fit is useful as it accommodates differences in the mean-variance trends between batches.
This is especially important if batches exhibit systematic technical differences, e.g., differences in coverage or in the amount of spike-in RNA added.
In this case, there are only minor differences between the trends in Figure <a href="feature-selection-2.html#fig:blocked-fit">10.5</a>, which indicates that the experiment was tightly replicated across plates.
The analysis of each plate yields estimates of the biological and technical components for each gene, which are averaged across plates to take advantage of information from multiple batches.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb229-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb229-2" data-line-number="2">blocked.stats &lt;-<span class="st"> </span>dec.block<span class="fl">.416</span>b<span class="op">$</span>per.block</a>
<a class="sourceLine" id="cb229-3" data-line-number="3"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">colnames</span>(blocked.stats)) {</a>
<a class="sourceLine" id="cb229-4" data-line-number="4">    current &lt;-<span class="st"> </span>blocked.stats[[i]]</a>
<a class="sourceLine" id="cb229-5" data-line-number="5">    <span class="kw">plot</span>(current<span class="op">$</span>mean, current<span class="op">$</span>total, <span class="dt">main=</span>i, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">cex=</span><span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb229-6" data-line-number="6">        <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb229-7" data-line-number="7">    curfit &lt;-<span class="st"> </span><span class="kw">metadata</span>(current)</a>
<a class="sourceLine" id="cb229-8" data-line-number="8">    <span class="kw">points</span>(curfit<span class="op">$</span>mean, curfit<span class="op">$</span>var, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</a>
<a class="sourceLine" id="cb229-9" data-line-number="9">    <span class="kw">curve</span>(curfit<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&#39;dodgerblue&#39;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>) </a>
<a class="sourceLine" id="cb229-10" data-line-number="10">}</a></code></pre></div>
<div class="figure"><span id="fig:blocked-fit"></span>
<img src="P2_W06.feature-selection_files/figure-html/blocked-fit-1.png" alt="Variance in the 416B data set as a function of the mean after blocking on the plate of origin. Each plot represents the results for a single plate, each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins." width="672" />
<p class="caption">
Figure 10.5: Variance in the 416B data set as a function of the mean after blocking on the plate of origin. Each plot represents the results for a single plate, each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins.
</p>
</div>
</div>
<div id="using-a-design-matrix" class="section level4">
<h4><span class="header-section-number">10.2.4.2</span> Using a design matrix</h4>
<p>The use of block-specific trends is the recommended approach for experiments with a single blocking factor.
However, this is not practical for studies involving a large number of blocking factors and/or covariates.
In such cases, we can use the <code>design=</code> argument to specify a design matrix with uninteresting factors of variation.
We illustrate again with the 416B data set, blocking on the plate of origin and oncogene induction.
(The same argument is available in <code>modelGeneVar()</code> when spike-ins are not available.)</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb230-1" data-line-number="1">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="kw">factor</span>(block) <span class="op">+</span><span class="st"> </span>phenotype, <span class="kw">colData</span>(sce<span class="fl">.416</span>b))</a>
<a class="sourceLine" id="cb230-2" data-line-number="2">dec.design<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;ERCC&quot;</span>, <span class="dt">design=</span>design)</a>
<a class="sourceLine" id="cb230-3" data-line-number="3">dec.design<span class="fl">.416</span>b[<span class="kw">order</span>(dec.design<span class="fl">.416</span>b<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 46604 rows and 6 columns
##                      mean             total             tech
##                 &lt;numeric&gt;         &lt;numeric&gt;        &lt;numeric&gt;
## Ccnb2    5.90621334470428  9.79715520072974 2.33818123384043
## Lyz2     6.58452322626962  8.90376336427706 1.52626961476178
## Gem      5.84717806481195  9.65752815506973 2.42726139963328
## Cenpa    5.76426929502564  8.75737959544375 2.55809652313982
## Idh1     5.97112724393234  8.35043772112997 2.24402124405226
## ...                   ...               ...              ...
## Gm5054   2.87548171848215 0.462387955705153 6.77447770216436
## Gm12191  3.59440418513899  0.16968946726796 6.50247698108384
## Gm7429   3.48699542530014 0.247924635935214 6.61767039856465
## Gm16378  2.86103400521603 0.207146235830467 6.76875864745726
## Rps2-ps2  3.0757346747959 0.199895150695701 6.80180881355727
##                        bio               p.value                  FDR
##                  &lt;numeric&gt;             &lt;numeric&gt;            &lt;numeric&gt;
## Ccnb2      7.4589739668893  1.22571991004624e-73 2.05472089709442e-71
## Lyz2      7.37749374951528 3.72834732625484e-166 2.4652661042825e-163
## Gem       7.23026675543645  1.75586377579736e-64 2.28396619229947e-62
## Cenpa     6.19928307230393  2.18851936966485e-43 1.46335716504219e-41
## Idh1      6.10641647707771  3.65933032533613e-54 3.52658700665187e-52
## ...                    ...                   ...                  ...
## Gm5054   -6.31208974645921     0.999999939108963    0.999999984005896
## Gm12191  -6.33278751381588     0.999999984005896    0.999999984005896
## Gm7429   -6.36974576262944     0.999999976938005    0.999999984005896
## Gm16378   -6.5616124116268       0.9999999814994    0.999999984005896
## Rps2-ps2 -6.60191366286157     0.999999982210137    0.999999984005896</code></pre>
<pre><code>## &lt;numeric&gt;

## 5.90621334470428

## 6.58452322626962

## 5.84717806481195

## 5.76426929502564

## 5.97112724393234

## ...

## 2.87548171848215

## 3.59440418513899

## 3.48699542530014

## 2.86103400521603

## 3.0757346747959

## &lt;numeric&gt;

## 9.79715520072974

## 8.90376336427706

## 9.65752815506973

## 8.75737959544375

## 8.35043772112997

## ...

## 0.462387955705153

## 0.16968946726796

## 0.247924635935214

## 0.207146235830467

## 0.199895150695701

## &lt;numeric&gt;

## 2.33818123384043

## 1.52626961476178

## 2.42726139963328

## 2.55809652313982

## 2.24402124405226

## ...

## 6.77447770216436

## 6.50247698108384

## 6.61767039856465

## 6.76875864745726

## 6.80180881355727

## &lt;numeric&gt;

## 7.4589739668893

## 7.37749374951528

## 7.23026675543645

## 6.19928307230393

## 6.10641647707771

## ...

## -6.31208974645921

## -6.33278751381588

## -6.36974576262944

## -6.5616124116268

## -6.60191366286157

## &lt;numeric&gt;

## 1.22571991004624e-73

## 3.72834732625484e-166

## 1.75586377579736e-64

## 2.18851936966485e-43

## 3.65933032533613e-54

## ...

## 0.999999939108963

## 0.999999984005896

## 0.999999976938005

## 0.9999999814994

## 0.999999982210137

## &lt;numeric&gt;

## 2.05472089709442e-71

## 2.4652661042825e-163

## 2.28396619229947e-62

## 1.46335716504219e-41

## 3.52658700665187e-52

## ...

## 0.999999984005896

## 0.999999984005896

## 0.999999984005896

## 0.999999984005896

## 0.999999984005896</code></pre>
<p>This strategy is simple but somewhat inaccurate as it does not consider the mean expression in each blocking level.
Briefly, the technical component is estimated as the fitted value of the trend at the average abundance for each gene.
However, the true technical component is the average of the fitted values at the per-block means, which may be quite different for strong batch effects and non-linear mean-variance relationships.
The <code>block=</code> approach is safer and should be preferred in all situations where it is applicable.</p>
</div>
</div>
</div>
<div id="selecting-highly-variable-genes" class="section level2">
<h2><span class="header-section-number">10.3</span> Selecting highly variable genes</h2>
<div id="based-on-the-largest-metrics" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Based on the largest metrics</h3>
<p>Once we have quantified the per-gene variation, the next step is to select the subset of HVGs to use in downstream analyses.
The simplest approach is to simply take the top <span class="math inline">\(X\)</span> genes with the largest values for the relevant measure of variation.
For <code>modelGeneVar()</code> and <code>modelGeneVarWithSpikes()</code>, this would be the genes with the largest biological components:</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb233-1" data-line-number="1">hvg.pbmc.var &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>), <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb233-2" data-line-number="2">hvg.pbmc.var &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[hvg.pbmc.var]</a></code></pre></div>
<p>For <code>modelGeneCV2()</code> (and its relative, <code>modelGeneCV2WithSpikes()</code>), this would instead be the genes with the largest ratios:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb234-1" data-line-number="1">hvg.pbmc.cv2 &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">order</span>(dec.cv2.pbmc<span class="op">$</span>ratio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>), <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb234-2" data-line-number="2">hvg.pbmc.cv2 &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[hvg.pbmc.cv2]</a></code></pre></div>
<p>The main advantage of this approach is that the user can easily control the number of genes retained.
This ensures that the computational complexity of downstream calculations is easily predicted.
It is also fairly easy to translate the choice of <span class="math inline">\(X\)</span> into a biological statement.
Recall our trend-fitting assumption that most genes are not differentially expressed between cell types or states in our population.
If we quantify this assumption into a statement that, e.g., no more than 5% of genes are differentially expressed, we can simply take the top 5% of genes with the largest variance.</p>
<p>The main disadvantage of this approach that it turns HVG selection into a competition between genes, whereby a subset of very highly variable genes can push other informative genes out of the top set.
This can be problematic if a single subpopulation is very different from the others.
In such cases, the top set will be dominated by differentially expressed genes involving the outlier subpopulation, compromising resolution of heterogeneity between the other populations.
Similar problems are encountered when the magnitude of the chosen variance measure varies with abundance.</p>
<p>The choice of <span class="math inline">\(X\)</span> is also fairly arbitrary, with any value from 500 to 5000 being considered “reasonable”.
We have chosen <span class="math inline">\(X=1000\)</span> in the code above though there is no particular <em>a priori</em> reason for doing so.
A larger <span class="math inline">\(X\)</span> will reduce the risk of discarding signal at the cost of increasing noise that obscures signal.
Our recommendation is to simply pick an arbitrary <span class="math inline">\(X\)</span> and proceed with the rest of the analysis, with the intention of testing other choices later, rather than spending much time worrying about obtaining the “optimal” value<a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a>.</p>
</div>
<div id="based-on-a-fixed-threshold" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Based on a fixed threshold</h3>
<p>Another approach to feature selection is to set a fixed threshold of one of the metrics.
This is most commonly done with the (adjusted) <span class="math inline">\(p\)</span>-value reported by each of the above methods.
The <span class="math inline">\(p\)</span>-value for each gene is generated by testing against the null hypothesis that the variance is equal to the trend.
For example, we might define our HVGs as all genes that have adjusted <span class="math inline">\(p\)</span>-values below 0.05.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb235-1" data-line-number="1"><span class="co"># &#39;which()&#39; automatically removes &#39;NA&#39; FDR values.</span></a>
<a class="sourceLine" id="cb235-2" data-line-number="2">hvg.pbmc.var<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[<span class="kw">which</span>(dec.pbmc<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)]</a>
<a class="sourceLine" id="cb235-3" data-line-number="3"><span class="kw">length</span>(hvg.pbmc.var<span class="fl">.2</span>)</a></code></pre></div>
<pre><code>## [1] 642</code></pre>
<p>This approach is simple to implement and - if the test holds its size - it controls the false discovery rate (FDR).
That is, it returns a subset of genes where the proportion of false positives is expected to be below the specified threshold.
This can occasionally be useful in applications where the HVGs themselves are of interest.
For example, if a collaborator were to take a list of HVGs back to the bench to verify the existence of heterogeneous expression for some of the genes, we would want to control the FDR in that list.</p>
<p>The downside of this approach is that it is less predictable than the top <span class="math inline">\(X\)</span> strategy.
The number of genes returned depends on the type II error rate of the test and the severity of the multiple testing correction.
One might obtain no genes or every gene at a given FDR threshold, depending on the circumstances.
Moreover, control of the FDR is usually not helpful at this stage of the analysis.
We are not interpreting the individual HVGs themselves but are only using them for feature selection prior to downstream steps.
There is no reason to think that a 5% threshold on the FDR yields a more suitable compromise between bias and noise.</p>
</div>
<div id="keeping-all-genes-above-the-trend" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Keeping all genes above the trend</h3>
<p>As the title suggests, this involves keeping all genes above the trend.
The rationale is to only remove the obviously uninteresting genes with variances below the trend.
By doing so, we avoid the need to make any judgement calls regarding what level of variation is interesting enough to retain.
This approach represents one extreme of the bias-variance trade-off where bias is minimized at the cost of maximizing noise.
For <code>decomposeVar()</code>, it equates to keeping all positive biological components:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb237-1" data-line-number="1">hvg.pbmc.var<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[dec.pbmc<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb237-2" data-line-number="2"><span class="kw">length</span>(hvg.pbmc.var<span class="fl">.3</span>)</a></code></pre></div>
<pre><code>## [1] 12776</code></pre>
<p>For <code>improvedCV2()</code>, this involves keeping all ratios above 1:</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb239-1" data-line-number="1"><span class="co"># &#39;which()&#39; automatically removes &#39;NA&#39; ratios.</span></a>
<a class="sourceLine" id="cb239-2" data-line-number="2">hvg.pbmc.cv2<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.cv2.pbmc)[<span class="kw">which</span>(dec.cv2.pbmc<span class="op">$</span>ratio <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)]</a>
<a class="sourceLine" id="cb239-3" data-line-number="3"><span class="kw">length</span>(hvg.pbmc.cv2<span class="fl">.3</span>)</a></code></pre></div>
<pre><code>## [1] 9390</code></pre>
<p>This strategy is the most conservative as it does not discard any potential biological signal.
Weak or secondary population structure is given the chance to manifest as the affected genes are retained.
This makes it useful for reliable automated processing of diverse data sets where the primary factor of variation in one data set is a secondary factor in another data set (and thus overlooked by the top <span class="math inline">\(X\)</span> approach).
The obvious cost is that more noise is also captured, which can reduce the resolution of otherwise well-separated populations.
From a practical perspective, the use of more genes involves more computational work in each downstream step.</p>
</div>
</div>
<div id="putting-it-all-together-1" class="section level2">
<h2><span class="header-section-number">10.4</span> Putting it all together</h2>
<p>The few lines of code below will select the top 10% of genes with the highest biological components.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb241-1" data-line-number="1">dec.pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb241-2" data-line-number="2">chosen &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)]</a>
<a class="sourceLine" id="cb241-3" data-line-number="3">chosen &lt;-<span class="st"> </span><span class="kw">head</span>(chosen, <span class="kw">nrow</span>(dec.pbmc) <span class="op">*</span><span class="st"> </span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb241-4" data-line-number="4"><span class="kw">length</span>(chosen)</a></code></pre></div>
<pre><code>## [1] 3369</code></pre>
<p>We can then subset the <code>SingleCellExperiment</code> to only retain our selection of HVGs.
This ensures that downstream methods will only use these genes for their calculations.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb243-1" data-line-number="1">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[chosen,]</a>
<a class="sourceLine" id="cb243-2" data-line-number="2"><span class="kw">dim</span>(sce.pbmc)</a></code></pre></div>
<pre><code>## [1] 3369 3922</code></pre>
<p>Alternatively, some methods may allow users to pass in the full <code>SingleCellExperiment</code> object and specify the genes to use via an extra argument like <code>subset.row=</code>.
This may be more convenient in the context of the overall analysis, where genes outside of this subset may still be of interest during DE analyses or for visualization.</p>
</div>
<div id="session-info-4" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scran_1.13.13               SingleCellExperiment_1.7.7 
 [3] SummarizedExperiment_1.15.8 DelayedArray_0.11.4        
 [5] BiocParallel_1.19.2         matrixStats_0.54.0         
 [7] Biobase_2.45.0              GenomicRanges_1.37.14      
 [9] GenomeInfoDb_1.21.1         IRanges_2.19.14            
[11] S4Vectors_0.23.20           BiocGenerics_0.31.5        
[13] BiocStyle_2.13.2            Cairo_1.5-10               

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2               rsvd_1.0.2              
 [3] locfit_1.5-9.1           lattice_0.20-38         
 [5] assertthat_0.2.1         digest_0.6.20           
 [7] R6_2.4.0                 evaluate_0.14           
 [9] highr_0.8                ggplot2_3.2.1           
[11] pillar_1.4.2             zlibbioc_1.31.0         
[13] rlang_0.4.0              lazyeval_0.2.2          
[15] irlba_2.3.3              Matrix_1.2-17           
[17] rmarkdown_1.13           BiocNeighbors_1.3.3     
[19] statmod_1.4.32           stringr_1.4.0           
[21] igraph_1.2.4.1           RCurl_1.95-4.12         
[23] munsell_0.5.0            vipor_0.4.5             
[25] compiler_3.6.1           BiocSingular_1.1.5      
[27] xfun_0.9                 pkgconfig_2.0.2         
[29] ggbeeswarm_0.6.0         htmltools_0.3.6         
[31] tidyselect_0.2.5         gridExtra_2.3           
[33] tibble_2.1.3             GenomeInfoDbData_1.2.1  
[35] bookdown_0.13            edgeR_3.27.12           
[37] viridisLite_0.3.0        crayon_1.3.4            
[39] dplyr_0.8.3              bitops_1.0-6            
[41] grid_3.6.1               gtable_0.3.0            
[43] magrittr_1.5             scales_1.0.0            
[45] dqrng_0.2.1              stringi_1.4.3           
[47] XVector_0.25.0           viridis_0.5.1           
[49] limma_3.41.15            scater_1.13.17          
[51] DelayedMatrixStats_1.7.1 tools_3.6.1             
[53] beeswarm_0.2.3           glue_1.3.1              
[55] purrr_0.3.2              yaml_2.2.0              
[57] colorspace_1.4-1         BiocManager_1.30.4      
[59] knitr_1.24              </code></pre>
</div>
</div>
<div id="references-1" class="section level2 unnumbered">
<h2>References</h2>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-brennecke2013accounting">
<p>Brennecke, P., S. Anders, J. K. Kim, A. A. Kołodziejczyk, X. Zhang, V. Proserpio, B. Baying, et al. 2013. “Accounting for technical noise in single-cell RNA-seq experiments.” <em>Nat. Methods</em> 10 (11):1093–5.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="9">
<li id="fn9"><p>When everything’s said and done, the “optimal” value is one that gets you a figure for your manuscript.<a href="feature-selection-2.html#fnref9" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="normalization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dimensionality-reduction-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W06.feature-selection.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
