<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Feature selection | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-07-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="normalization.html">
<link rel="next" href="dimensionality-reduction-2.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="6.3" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i><b>6.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i><b>7.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i><b>8.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="9.8" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i><b>9.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="feature-selection-2.html"><a href="feature-selection-2.html"><i class="fa fa-check"></i><b>10</b> Feature selection</a><ul>
<li class="chapter" data-level="10.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>10.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="10.2.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>10.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="10.2.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#coefficient-of-variation"><i class="fa fa-check"></i><b>10.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="10.2.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#sec:spikeins"><i class="fa fa-check"></i><b>10.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="10.2.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>10.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#selecting-highly-variable-genes"><i class="fa fa-check"></i><b>10.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="10.3.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>10.3.1</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="10.3.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-a-fixed-threshold"><i class="fa fa-check"></i><b>10.3.2</b> Based on a fixed threshold</a></li>
<li class="chapter" data-level="10.3.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#keeping-all-genes-above-the-trend"><i class="fa fa-check"></i><b>10.3.3</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>10.4</b> Putting it all together</a></li>
<li class="chapter" data-level="10.5" data-path="feature-selection-2.html"><a href="feature-selection-2.html#session-info-4"><i class="fa fa-check"></i><b>10.5</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html"><i class="fa fa-check"></i><b>11</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="11.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#overview-1"><i class="fa fa-check"></i><b>11.1</b> Overview</a></li>
<li class="chapter" data-level="11.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#principal-components-analysis"><i class="fa fa-check"></i><b>11.2</b> Principal components analysis</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#background-1"><i class="fa fa-check"></i><b>11.2.1</b> Background</a></li>
<li class="chapter" data-level="11.2.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#performing-the-pca"><i class="fa fa-check"></i><b>11.2.2</b> Performing the PCA</a></li>
<li class="chapter" data-level="11.2.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>11.2.3</b> Choosing the number of PCs</a></li>
<li class="chapter" data-level="11.2.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#putting-it-together"><i class="fa fa-check"></i><b>11.2.4</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>11.3</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="11.3.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#motivation-4"><i class="fa fa-check"></i><b>11.3.1</b> Motivation</a></li>
<li class="chapter" data-level="11.3.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#visualizating-with-pca"><i class="fa fa-check"></i><b>11.3.2</b> Visualizating with PCA</a></li>
<li class="chapter" data-level="11.3.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>11.3.3</b> t-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="11.3.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>11.3.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="11.3.5" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#interpreting-the-plots"><i class="fa fa-check"></i><b>11.3.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#session-info-5"><i class="fa fa-check"></i><b>11.4</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="clustering-1.html"><a href="clustering-1.html"><i class="fa fa-check"></i><b>12</b> Clustering</a><ul>
<li class="chapter" data-level="12.1" data-path="clustering-1.html"><a href="clustering-1.html#motivation-5"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="clustering-1.html"><a href="clustering-1.html#comments-on-truth"><i class="fa fa-check"></i><b>12.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="12.3" data-path="clustering-1.html"><a href="clustering-1.html#graph-based-clustering"><i class="fa fa-check"></i><b>12.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="12.3.1" data-path="clustering-1.html"><a href="clustering-1.html#background-2"><i class="fa fa-check"></i><b>12.3.1</b> Background</a></li>
<li class="chapter" data-level="12.3.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation"><i class="fa fa-check"></i><b>12.3.2</b> Implementation</a></li>
<li class="chapter" data-level="12.3.3" data-path="clustering-1.html"><a href="clustering-1.html#other-parameters"><i class="fa fa-check"></i><b>12.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="12.3.4" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>12.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="clustering-1.html"><a href="clustering-1.html#k-means-clustering"><i class="fa fa-check"></i><b>12.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="12.4.1" data-path="clustering-1.html"><a href="clustering-1.html#background-3"><i class="fa fa-check"></i><b>12.4.1</b> Background</a></li>
<li class="chapter" data-level="12.4.2" data-path="clustering-1.html"><a href="clustering-1.html#base-implementation"><i class="fa fa-check"></i><b>12.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="12.4.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>12.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="clustering-1.html"><a href="clustering-1.html#hierarchical-clustering"><i class="fa fa-check"></i><b>12.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="12.5.1" data-path="clustering-1.html"><a href="clustering-1.html#background-4"><i class="fa fa-check"></i><b>12.5.1</b> Background</a></li>
<li class="chapter" data-level="12.5.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation-1"><i class="fa fa-check"></i><b>12.5.2</b> Implementation</a></li>
<li class="chapter" data-level="12.5.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>12.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="clustering-1.html"><a href="clustering-1.html#session-info-6"><i class="fa fa-check"></i><b>12.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>13</b> Marker gene detection</a><ul>
<li class="chapter" data-level="13.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-6"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>13.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="13.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>13.2.1</b> Standard application</a></li>
<li class="chapter" data-level="13.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>13.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="13.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>13.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>13.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="13.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>13.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="13.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>13.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="13.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>13.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>13.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="13.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>13.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="13.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>13.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="13.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>13.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="13.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>13.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="13.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>13.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-7"><i class="fa fa-check"></i><b>13.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>14</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="14.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#motivation-7"><i class="fa fa-check"></i><b>14.1</b> Motivation</a></li>
<li class="chapter" data-level="14.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#simple-two-batch-example"><i class="fa fa-check"></i><b>14.2</b> Simple two batch example</a><ul>
<li class="chapter" data-level="14.2.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>14.2.1</b> Setting up the data</a></li>
<li class="chapter" data-level="14.2.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#performing-mnn-correction"><i class="fa fa-check"></i><b>14.2.2</b> Performing MNN correction</a></li>
<li class="chapter" data-level="14.2.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#correction-diagnostics"><i class="fa fa-check"></i><b>14.2.3</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-8"><i class="fa fa-check"></i><b>14.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>15</b> Sharing and Interactive Interfaces</a><ul>
<li class="chapter" data-level="15.1" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html#session-info-9"><i class="fa fa-check"></i><b>15.1</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>16</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="16.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>16.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="16.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>16.2</b> Parallelization</a></li>
<li class="chapter" data-level="16.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>16.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="16.4" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-10"><i class="fa fa-check"></i><b>16.4</b> Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>17</b> About the Data</a><ul>
<li class="chapter" data-level="17.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>17.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="17.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>17.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>18</b> About the Contributors</a><ul>
<li class="chapter" data-level="18.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>18.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="18.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>18.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="18.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>18.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="18.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>18.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>19</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-selection-2" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Feature selection</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-2" class="section level2">
<h2><span class="header-section-number">10.1</span> Motivation</h2>
<p>We often use scRNA-seq data in exploratory analyses to characterize heterogeneity across cells.
Procedures like clustering and dimensionality reduction compare cells based on their gene expression profiles, which involves aggregating per-gene differences into a single (dis)similarity metric between a pair of cells.
The choice of genes to use in this calculation has a major impact on the behavior of the metric and the performance of downstream methods.
We want to select genes that contain useful information about the biology of the system while removing genes that contain random noise.
This aims to preserve interesting biological structure without the variance that obscures that structure.
It also reduces the size of the dataset to improve computational efficiency of later steps.</p>
<p>The simplest approach to feature selection is to select the most variable genes based on their expression across the population.
This assumes that genuine biological differences will manifest as increased variation in the affected genes, compared to other genes that are only affected by technical noise or a baseline level of “uninteresting” biological variation (e.g., from transcriptional bursting).
Several methods are available to quantify the variation per gene and to select the highly variable genes (HVGs), which we will discuss below.
For demonstration, we will use the 10X PBMC dataset:</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb202-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb202-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb202-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="st">&quot;raw_data&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb202-4" data-line-number="4">raw.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</a>
<a class="sourceLine" id="cb202-5" data-line-number="5">    <span class="st">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</a>
<a class="sourceLine" id="cb202-6" data-line-number="6"><span class="kw">untar</span>(raw.path, <span class="dt">exdir=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k&quot;</span>))</a>
<a class="sourceLine" id="cb202-7" data-line-number="7"></a>
<a class="sourceLine" id="cb202-8" data-line-number="8"><span class="kw">library</span>(DropletUtils)</a>
<a class="sourceLine" id="cb202-9" data-line-number="9">fname &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</a>
<a class="sourceLine" id="cb202-10" data-line-number="10">sce.pbmc &lt;-<span class="st"> </span><span class="kw">read10xCounts</span>(fname, <span class="dt">col.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb202-11" data-line-number="11"></a>
<a class="sourceLine" id="cb202-12" data-line-number="12">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb202-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb202-14" data-line-number="14"><span class="kw">rownames</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb202-15" data-line-number="15"></a>
<a class="sourceLine" id="cb202-16" data-line-number="16"><span class="kw">library</span>(EnsDb.Hsapiens.v86)</a>
<a class="sourceLine" id="cb202-17" data-line-number="17">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(EnsDb.Hsapiens.v86, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb202-18" data-line-number="18">    <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb202-19" data-line-number="19"></a>
<a class="sourceLine" id="cb202-20" data-line-number="20">### cell-detection <span class="al">###</span></a>
<a class="sourceLine" id="cb202-21" data-line-number="21"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb202-22" data-line-number="22">e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb202-23" data-line-number="23">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="kw">which</span>(e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>)]</a>
<a class="sourceLine" id="cb202-24" data-line-number="24"></a>
<a class="sourceLine" id="cb202-25" data-line-number="25">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb202-26" data-line-number="26">sce.pbmc &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce.pbmc, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(location<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb202-27" data-line-number="27">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.pbmc<span class="op">$</span>pct_counts_Mito, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb202-28" data-line-number="28">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb202-29" data-line-number="29"></a>
<a class="sourceLine" id="cb202-30" data-line-number="30">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb202-31" data-line-number="31"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb202-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb202-33" data-line-number="33">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb202-34" data-line-number="34">sce.pbmc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.pbmc, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">cluster=</span>clusters)</a>
<a class="sourceLine" id="cb202-35" data-line-number="35">sce.pbmc &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.pbmc)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" data-line-number="1">sce.pbmc</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 3922 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
## rowData names(10): ID Symbol ... total_counts log10_total_counts
## colnames(3922): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCACAGGTCCAC-1 TTTGTCATCCCAAGAT-1
## colData names(38): Sample Barcode ...
##   pct_counts_in_top_200_features_Mito
##   pct_counts_in_top_500_features_Mito
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>As well as the 416B dataset:</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb205-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb205-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb205-3" data-line-number="3">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">LunSpikeInData</span>(<span class="dt">which=</span><span class="st">&quot;416b&quot;</span>) </a>
<a class="sourceLine" id="cb205-4" data-line-number="4"></a>
<a class="sourceLine" id="cb205-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb205-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb205-7" data-line-number="7">ens.mm.v97 &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73905&quot;</span>]]</a>
<a class="sourceLine" id="cb205-8" data-line-number="8"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL &lt;-<span class="st"> </span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb205-9" data-line-number="9"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb205-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SYMBOL&quot;</span>)</a>
<a class="sourceLine" id="cb205-11" data-line-number="11"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb205-12" data-line-number="12">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>)</a>
<a class="sourceLine" id="cb205-13" data-line-number="13"></a>
<a class="sourceLine" id="cb205-14" data-line-number="14"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb205-15" data-line-number="15"><span class="kw">rownames</span>(sce<span class="fl">.416</span>b) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL, </a>
<a class="sourceLine" id="cb205-16" data-line-number="16">    <span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL)</a>
<a class="sourceLine" id="cb205-17" data-line-number="17"></a>
<a class="sourceLine" id="cb205-18" data-line-number="18">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb205-19" data-line-number="19">mito &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME<span class="op">==</span><span class="st">&quot;MT&quot;</span>)</a>
<a class="sourceLine" id="cb205-20" data-line-number="20">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce<span class="fl">.416</span>b, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mt=</span>mito))</a>
<a class="sourceLine" id="cb205-21" data-line-number="21"></a>
<a class="sourceLine" id="cb205-22" data-line-number="22">combined &lt;-<span class="st"> </span><span class="kw">paste0</span>(sce<span class="fl">.416</span>b<span class="op">$</span>block, <span class="st">&quot;:&quot;</span>, sce<span class="fl">.416</span>b<span class="op">$</span>phenotype)</a>
<a class="sourceLine" id="cb205-23" data-line-number="23">libsize.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_counts, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, </a>
<a class="sourceLine" id="cb205-24" data-line-number="24">    <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb205-25" data-line-number="25">feature.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_features_by_counts, <span class="dt">nmads=</span><span class="dv">3</span>, </a>
<a class="sourceLine" id="cb205-26" data-line-number="26">    <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb205-27" data-line-number="27">spike.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>pct_counts_ERCC, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>,</a>
<a class="sourceLine" id="cb205-28" data-line-number="28">    <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb205-29" data-line-number="29"></a>
<a class="sourceLine" id="cb205-30" data-line-number="30">keep &lt;-<span class="st"> </span><span class="op">!</span>(libsize.drop <span class="op">|</span><span class="st"> </span>feature.drop <span class="op">|</span><span class="st"> </span>spike.drop)</a>
<a class="sourceLine" id="cb205-31" data-line-number="31">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span>sce<span class="fl">.416</span>b[,keep]</a>
<a class="sourceLine" id="cb205-32" data-line-number="32"></a>
<a class="sourceLine" id="cb205-33" data-line-number="33">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb205-34" data-line-number="34"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb205-35" data-line-number="35">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb205-36" data-line-number="36">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSpikeFactors</span>(sce<span class="fl">.416</span>b, <span class="dt">general.use=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb205-37" data-line-number="37">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce<span class="fl">.416</span>b)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb206-1" data-line-number="1">sce<span class="fl">.416</span>b</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 46703 188 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(46703): ENSMUSG00000102693 ENSMUSG00000064842 ... SIRV7
##   CBFB-MYH11-mcherry
## rowData names(14): Length ENSEMBL ... total_counts
##   log10_total_counts
## colnames(188): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1
##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...
##   SLX-11312.N712_S507.H5H5YBBXX.s_8.r_1
##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1
## colData names(63): Source Name cell line ...
##   pct_counts_in_top_200_features_SIRV
##   pct_counts_in_top_500_features_SIRV
## reducedDimNames(0):
## spikeNames(2): ERCC SIRV</code></pre>
</div>
<div id="quantifying-per-gene-variation" class="section level2">
<h2><span class="header-section-number">10.2</span> Quantifying per-gene variation</h2>
<div id="variance-of-the-log-counts" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Variance of the log-counts</h3>
<p>The simplest approach to quantifying per-gene variation is to simply compute the variance of the log-normalized expression values (referred to as “log-counts” for simplicity) for each gene across all cells in the population.
This has an advantage in that the feature selection is based on the same log-values that are used for later downstream steps.
Genes with the largest variances in log-values will contribute the most to the Euclidean distances between cells.
By using log-values here, we ensure that our quantitative definition of heterogeneity is consistent throughout the entire analysis.</p>
<p>Calculation of the per-gene variance is simple but feature selection requires modelling of the mean-variance relationship.
As discussed briefly in the last chapter, the log-transformation does not achieve perfect variance stabilization.
This means that the variance of a gene is more affected by its abundance than the underlying biological heterogeneity.
To account for this effect, we use the <code>trendVar()</code> function to fit a trend to the variance with respect to abundance across all genes (Figure <a href="feature-selection-2.html#fig:trend-plot-pbmc">10.1</a>).</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb208-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb208-2" data-line-number="2"></a>
<a class="sourceLine" id="cb208-3" data-line-number="3"><span class="co"># No spike-ins, so setting &#39;use.spikes=FALSE&#39;.</span></a>
<a class="sourceLine" id="cb208-4" data-line-number="4">fit.pbmc &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce.pbmc, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb208-5" data-line-number="5"></a>
<a class="sourceLine" id="cb208-6" data-line-number="6"><span class="kw">plot</span>(fit.pbmc<span class="op">$</span>mean, fit.pbmc<span class="op">$</span>var, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb208-7" data-line-number="7">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb208-8" data-line-number="8"><span class="kw">curve</span>(fit.pbmc<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:trend-plot-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/trend-plot-pbmc-1.png" alt="Variance in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the trend fitted to all genes." width="672" />
<p class="caption">
Figure 10.1: Variance in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the trend fitted to all genes.
</p>
</div>
<p>At any given abundance, we assume that the expression profiles of most genes are dominated by random technical noise (see <a href="feature-selection-2.html#sec:spikeins">10.2.3</a> for details).
Under this assumption, our trend represents an estimate of the technical noise as a function of abundance.
We then use the <code>decomposeVar()</code> function to break down the total variance of each gene into the technical component, i.e., the fitted value of the trend at that gene’s abundance;
and the biological component, defined as the difference between the total variance and the technical component.
This biological component represents the “interesting” variation for each gene and can be used as the metric for HVG selection.</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb209-1" data-line-number="1">dec.pbmc &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(<span class="dt">fit=</span>fit.pbmc)</a>
<a class="sourceLine" id="cb209-2" data-line-number="2"></a>
<a class="sourceLine" id="cb209-3" data-line-number="3"><span class="co"># Ordering by most interesting genes for inspection.</span></a>
<a class="sourceLine" id="cb209-4" data-line-number="4">dec.pbmc[<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),] </a></code></pre></div>
<pre><code>## DataFrame with 33694 rows and 6 columns
##                     mean             total                bio
##                &lt;numeric&gt;         &lt;numeric&gt;          &lt;numeric&gt;
## LYZ      1.9806721473164  5.12675017749729   4.23343856586781
## S100A9  1.95576910151395  4.62353584269683   3.73059071698752
## S100A8  1.72447657712873  4.49362852270639   3.60800589885247
## HLA-DRA 2.10051754708271  3.73885530331428   2.84304344288055
## CD74    2.90746230772648   3.3292370802074   2.30788747256288
## ...                  ...               ...                ...
## MT-CO2  4.19188502683674 0.544465230103077 -0.635044758465113
## PTMA    3.84384732827719 0.498416795149973 -0.650614922570656
## HLA-B   4.51931577517642 0.499881887649168 -0.701096896119891
## TMSB4X   6.1055295386377 0.470802972021666 -0.776306567179052
## B2M     5.97628515673904  0.34502768951129 -0.899609719380582
##                      tech   p.value       FDR
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## LYZ     0.893311611629482         0         0
## S100A9  0.892945125709309         0         0
## S100A8  0.885622623853927         0         0
## HLA-DRA 0.895811860433723         0         0
## CD74     1.02134960764452         0         0
## ...                   ...       ...       ...
## MT-CO2   1.17950998856819         1         1
## PTMA     1.14903171772063         1         1
## HLA-B    1.20097878376906         1         1
## TMSB4X   1.24710953920072         1         1
## B2M      1.24463740889187         1         1</code></pre>
<p>(Careful readers will notice that some genes have negative biological components.
Negative components of variation have no obvious interpretation and can be ignored for most applications.
They are inevitable when fitting a trend to the per-gene variances, as approximately half of the genes will lie below the trend.)</p>
</div>
<div id="coefficient-of-variation" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Coefficient of variation</h3>
<p>An alternative approach to quantification uses the squared coefficient of variation (CV<sup>2</sup>) of the normalized expression values prior to log-transformation.
The CV<sup>2</sup> is a widely used metric for describing variation in non-negative data and is closely related to the dispersion parameter of the negative binomial distribution in packages like <em><a href="https://bioconductor.org/packages/3.10/edgeR">edgeR</a></em> and <em><a href="https://bioconductor.org/packages/3.10/DESeq2">DESeq2</a></em>.
We compute the CV<sup>2</sup> for each gene in the PBMC dataset using the <code>improvedCV2()</code> function, which provides a more robust implementation of the approach described by <span class="citation">Brennecke et al. (<a href="#ref-brennecke2013accounting">2013</a>)</span>.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb211-1" data-line-number="1"><span class="co"># </span><span class="al">TODO</span><span class="co">: upgrade scran to use the latest version of this.</span></a>
<a class="sourceLine" id="cb211-2" data-line-number="2"><span class="co"># No spike-ins, so setting &#39;spike.type=NA&#39;.</span></a>
<a class="sourceLine" id="cb211-3" data-line-number="3">cv2.pbmc &lt;-<span class="st"> </span><span class="kw">improvedCV2</span>(sce.pbmc, <span class="dt">spike.type=</span><span class="ot">NA</span>) </a></code></pre></div>
<p>This allows us to model the mean-variance relationship when considering the relevance of each gene (Figure <a href="feature-selection-2.html#fig:cv2-pbmc">10.2</a>).
Again, our assumption is that most genes contain random noise and that the trend captures mostly technical variation.
Large CV<span class="math inline">\(^2\)</span> values that deviate strongly from the trend are likely to represent genes affected by biological structure.</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb212-1" data-line-number="1"><span class="kw">plot</span>(cv2.pbmc<span class="op">$</span>mean, cv2.pbmc<span class="op">$</span>cv2, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>)</a>
<a class="sourceLine" id="cb212-2" data-line-number="2">o &lt;-<span class="st"> </span><span class="kw">order</span>(cv2.pbmc<span class="op">$</span>mean)</a>
<a class="sourceLine" id="cb212-3" data-line-number="3"><span class="kw">lines</span>(cv2.pbmc<span class="op">$</span>mean[o], cv2.pbmc<span class="op">$</span>trend[o], <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:cv2-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/cv2-pbmc-1.png" alt="CV^2^ in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the fitted trend." width="672" />
<p class="caption">
Figure 10.2: CV<sup>2</sup> in the PBMC data set as a function of the mean. Each point represents a gene while the blue line represents the fitted trend.
</p>
</div>
<p>For each gene, we quantify the deviation from the trend in terms of the ratio of its CV<sup>2</sup> to the fitted value of trend at its abundance.
This is more appropriate than the directly subtracting the trend from the CV<sup>2</sup>, as the magnitude of the ratio is not affected by the mean.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb213-1" data-line-number="1">diff &lt;-<span class="st"> </span>cv2.pbmc<span class="op">$</span>cv2<span class="op">/</span>cv2.pbmc<span class="op">$</span>trend</a>
<a class="sourceLine" id="cb213-2" data-line-number="2">cv2.pbmc[<span class="kw">order</span>(diff, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 33694 rows and 6 columns
##                         mean              cv2             trend
##                    &lt;numeric&gt;        &lt;numeric&gt;         &lt;numeric&gt;
## PRTFDC1    0.349193768088249 3781.11597077882   3.7476790639203
## GNG11        2.0189430911897 415.713443913274 0.870369442833297
## RCAN1      0.194958683724188 3033.04562507858  6.50000581102497
## SDPR        1.60637825330209 467.847955624938  1.02491132642899
## BEND2      0.210970404091042 2644.94583077305  6.02707265545054
## ...                      ...              ...               ...
## AC023491.2                 0              NaN               Inf
## AC233755.2                 0              NaN               Inf
## AC233755.1                 0              NaN               Inf
## AC213203.1                 0              NaN               Inf
## FAM231B                    0              NaN               Inf
##                       ratio   p.value       FDR
##                   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## PRTFDC1    1008.92203048559         0         0
## GNG11      477.628721155479         0         0
## RCAN1      466.621986696395         0         0
## SDPR       456.476520027366         0         0
## BEND2      438.844192193554         0         0
## ...                     ...       ...       ...
## AC023491.2              NaN         1         1
## AC233755.2              NaN         1         1
## AC233755.1              NaN         1         1
## AC213203.1              NaN         1         1
## FAM231B                 NaN         1         1</code></pre>
<p>Both the CV<sup>2</sup> and the variance of log-counts are effective metrics for quantifying variation in gene expression.
The CV<sup>2</sup> tends to give higher rank to low-abundance HVGs driven by upregulation in rare subpopulations, for which the increase in variance on the raw scale is stronger than that on the log-scale.
However, the variation described by the CV<sup>2</sup> is less directly relevant to downstream procedures operating on the log-counts, and the reliance on the ratio can assign high rank to uninteresting genes with low absolute variance.
We generally prefer the use of the variance of log-counts and will use it in the following sections, though the many of the same principles apply to procedures based on the CV<sup>2</sup>.</p>
</div>
<div id="sec:spikeins" class="section level3">
<h3><span class="header-section-number">10.2.3</span> Quantifying technical noise</h3>
<p>The use of a trend fitted to endogenous genes assumes that the expression profiles of most genes are dominated by random technical noise.
In practice, all expressed genes will exhibit some non-zero level of biological variability due to events like transcriptional bursting.
This suggests that our estimates of the technical component are likely to be inflated.
It would be more appropriate to consider these estimates as technical noise plus “uninteresting” biological variation, under the assumption that most genes are unaffected by the relevant heterogeneity in the population.</p>
<p>This assumption is generally reasonable but may be problematic in some scenarios where many genes at a particular abundance are affected by a biological process.
For example, strong upregulation of cell type-specific genes may result in an enrichment of HVGs at high abundances.
This would inflate the fitted trend at high abundances and compromise the detection of the affected genes.
We can avoid this problem by fitting a mean-dependent trend to the variance of the spike-in transcripts, if they are available (Figure <a href="feature-selection-2.html#fig:spike-416b">10.3</a>).
The premise here is that spike-ins should not be affected by biological variation, so the fitted value of the spike-in trend should represent a better estimate of the technical component for each gene.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb215-1" data-line-number="1">fit<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb215-2" data-line-number="2">dec<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce<span class="fl">.416</span>b, fit<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb215-3" data-line-number="3">dec<span class="fl">.416</span>b[<span class="kw">order</span>(dec<span class="fl">.416</span>b<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 46703 rows and 6 columns
##                                mean             total               bio
##                           &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;
## Lyz2               6.57237822734111  13.9096268809543  12.0532209419185
## Ccl9               6.72849586510843  13.2061002387558   11.530457079581
## Top2a              5.77874153915785  13.9853581173099  10.9827877439828
## Cd200r3            4.84945724546366  15.5251666159147  10.9183993789504
## ENSMUSG00000096842 4.21058861964987  15.9330829391269  10.5640743777755
## ...                             ...               ...               ...
## ENSMUSG00000097554 3.21315842136385 0.869510710256039 -5.25440177296372
## ENSMUSG00000083011 2.91045997741136 0.828657162029132 -5.29256475478502
## ENSMUSG00000082064 3.60090056018739 0.607899262223165 -5.32548110628434
## ENSMUSG00000083007 2.71208464889622 0.762293733047904  -5.3432512239974
## ENSMUSG00000078087 3.14961465503528 0.738996070206387 -5.39306841637665
##                                tech               p.value
##                           &lt;numeric&gt;             &lt;numeric&gt;
## Lyz2               1.85640593903582 8.54995227701758e-185
## Ccl9               1.67564315917488 1.53437515087334e-198
## Top2a               3.0025703733271  9.88028287012378e-89
## Cd200r3            4.60676723696428  2.14422927380176e-49
## ENSMUSG00000096842 5.36900856135144  3.89126290980577e-38
## ...                             ...                   ...
## ENSMUSG00000097554 6.12391248321976                     1
## ENSMUSG00000083011 6.12122191681415                     1
## ENSMUSG00000082064  5.9333803685075                     1
## ENSMUSG00000083007 6.10554495704531                     1
## ENSMUSG00000078087 6.13206448658304                     1
##                                      FDR
##                                &lt;numeric&gt;
## Lyz2               2.65641317278752e-181
## Ccl9               7.94533550347791e-195
## Top2a               5.90334234460575e-86
## Cd200r3             3.85828807244237e-47
## ENSMUSG00000096842  3.98567948678215e-36
## ...                                  ...
## ENSMUSG00000097554                     1
## ENSMUSG00000083011                     1
## ENSMUSG00000082064                     1
## ENSMUSG00000083007                     1
## ENSMUSG00000078087                     1</code></pre>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb217-1" data-line-number="1"><span class="kw">plot</span>(dec<span class="fl">.416</span>b<span class="op">$</span>mean, dec<span class="fl">.416</span>b<span class="op">$</span>total, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb217-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb217-3" data-line-number="3">is.spike &lt;-<span class="st"> </span><span class="kw">isSpike</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb217-4" data-line-number="4"><span class="kw">points</span>(dec<span class="fl">.416</span>b<span class="op">$</span>mean[is.spike], dec<span class="fl">.416</span>b<span class="op">$</span>total[is.spike], <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</a>
<a class="sourceLine" id="cb217-5" data-line-number="5"><span class="kw">curve</span>(fit<span class="fl">.416</span>b<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure"><span id="fig:spike-416b"></span>
<img src="P2_W06.feature-selection_files/figure-html/spike-416b-1.png" alt="Variance in the 416B data set as a function of the mean. Each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins." width="672" />
<p class="caption">
Figure 10.3: Variance in the 416B data set as a function of the mean. Each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins.
</p>
</div>
<p>In the absence of spike-in data, one can attempt to create a trend by making some distributional assumptions about the noise.
For example, UMI counts typically exhibit near-Poisson noise when only technical effects are considered.
This can be used to construct a mean-variance trend in the log-counts (Figure <a href="feature-selection-2.html#fig:tech-pbmc">10.4</a>) with the <code>makeTechTrend()</code> function.
Note the increased residuals of the high-abundance genes, which can be interpreted as the amount of biological variation that was assumed to be “uninteresting” when fitting the gene-based trend in Figure <a href="feature-selection-2.html#fig:trend-plot-pbmc">10.1</a>.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb218-1" data-line-number="1">tech.trend &lt;-<span class="st"> </span><span class="kw">makeTechTrend</span>(<span class="dt">x=</span>sce.pbmc)</a>
<a class="sourceLine" id="cb218-2" data-line-number="2"></a>
<a class="sourceLine" id="cb218-3" data-line-number="3">fit.pbmc2 &lt;-<span class="st"> </span>fit.pbmc</a>
<a class="sourceLine" id="cb218-4" data-line-number="4">fit.pbmc2<span class="op">$</span>trend &lt;-<span class="st"> </span>tech.trend <span class="co"># overwrite trend.</span></a>
<a class="sourceLine" id="cb218-5" data-line-number="5">dec.pbmc2 &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(<span class="dt">fit=</span>fit.pbmc2) </a>
<a class="sourceLine" id="cb218-6" data-line-number="6">dec.pbmc2 &lt;-<span class="st"> </span>dec.pbmc2[<span class="kw">order</span>(dec.pbmc2<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a>
<a class="sourceLine" id="cb218-7" data-line-number="7"><span class="kw">head</span>(dec.pbmc2)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                     mean            total              bio
##                &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;
## LYZ      1.9806721473164 5.12675017749729  4.4957050381929
## S100A9  1.95576910151395 4.62353584269683 3.98858788456279
## S100A8  1.72447657712873 4.49362852270639 3.82909592851813
## HLA-DRA 2.10051754708271 3.73885530331428 3.12807677300474
## CD74    2.90746230772648  3.3292370802074 2.87761651579676
## CST3    1.49489535215485 2.97218277928419 2.29406052725448
##                      tech   p.value       FDR
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## LYZ     0.631045139304399         0         0
## S100A9  0.634947958134047         0         0
## S100A8  0.664532594188262         0         0
## HLA-DRA 0.610778530309538         0         0
## CD74    0.451620564410635         0         0
## CST3    0.678122252029714         0         0</code></pre>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb220-1" data-line-number="1"><span class="kw">plot</span>(dec.pbmc2<span class="op">$</span>mean, dec.pbmc2<span class="op">$</span>total, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb220-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb220-3" data-line-number="3"><span class="kw">curve</span>(fit.pbmc2<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tech-pbmc"></span>
<img src="P2_W06.feature-selection_files/figure-html/tech-pbmc-1.png" alt="Variance of normalized log-expression values for each gene in the PBMC dataset, plotted against the mean log-expression. The blue line represents the mean-dependent trend fitted to the variances, while the red line represents the Poisson noise." width="672" />
<p class="caption">
Figure 10.4: Variance of normalized log-expression values for each gene in the PBMC dataset, plotted against the mean log-expression. The blue line represents the mean-dependent trend fitted to the variances, while the red line represents the Poisson noise.
</p>
</div>
</div>
<div id="accounting-for-blocking-factors" class="section level3">
<h3><span class="header-section-number">10.2.4</span> Accounting for blocking factors</h3>
<div id="fitting-block-specific-trends" class="section level4">
<h4><span class="header-section-number">10.2.4.1</span> Fitting block-specific trends</h4>
<p>Data containing multiple batches will often exhibit batch effects (see Chapter ??? for more details).
We are (usually) not interested in HVGs that are driven by batch effects.
Rather, we want to focus on genes that are highly variable within each batch.
This is naturally achieved by performing trend fitting and variance decomposition separately for each batch.
We demonstrate this approach by treating each plate (<code>block</code>) in the 416B dataset as a different batch, using the <code>multiBlockVar()</code> function.</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb221-1" data-line-number="1">sce<span class="fl">.416</span>b<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">multiBlockNorm</span>(sce<span class="fl">.416</span>b, <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb221-2" data-line-number="2">dec<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">multiBlockVar</span>(sce<span class="fl">.416</span>b<span class="fl">.2</span>, <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb221-3" data-line-number="3"><span class="kw">head</span>(dec<span class="fl">.416</span>b[<span class="kw">order</span>(dec<span class="fl">.416</span>b<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>])</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                     mean            total              bio
##                &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;
## Lyz2    6.57237822734111 13.9153175317599 12.1371589280417
## Ccl9    6.72849586510843 13.2762218610643  11.752319895066
## Cd200r3 4.84945724546366 15.5808151041527 11.1878190309862
## Top2a   5.77874153915785 13.8393332839276 10.8570743303387
## Hbb-bt  4.93360124730523 14.7939723392898 10.5209124986672
## Ccnb2   5.89821323258258  13.144631142326 10.3507085507561
##                     tech               p.value                   FDR
##                &lt;numeric&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## Lyz2    1.77815860371824 3.50296135945991e-232 1.02032506997669e-228
## Ccl9     1.5239019659983 2.26265579092136e-233 7.02992069867327e-230
## Cd200r3 4.39299607316647  6.71476042638906e-58  1.14628093374152e-55
## Top2a   2.98225895358894  4.79084975637688e-99  2.59619490751381e-96
## Hbb-bt  4.27305984062261  5.27656858109607e-53  7.56643698933543e-51
## Ccnb2   2.79392259156986  1.2567728529748e-106 8.48849884638228e-104</code></pre>
<p>The use of a batch-specific trend fit is useful as it accommodates differences in the mean-variance trends between batches.
This is especially important if batches exhibit systematic technical differences, e.g., differences in coverage or in the amount of spike-in RNA added.
In this case, there are only minor differences between the trends in Figure <a href="feature-selection-2.html#fig:blocked-fit">10.5</a>, which indicates that the experiment was tightly replicated across plates.
The analysis of each plate yields estimates of the biological and technical components for each gene, which are averaged across plates to take advantage of information from multiple batches.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb223-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb223-2" data-line-number="2">blocked.stats &lt;-<span class="st"> </span>dec<span class="fl">.416</span>b<span class="op">$</span>per.block</a>
<a class="sourceLine" id="cb223-3" data-line-number="3"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">colnames</span>(blocked.stats)) {</a>
<a class="sourceLine" id="cb223-4" data-line-number="4">    current &lt;-<span class="st"> </span>blocked.stats[[i]]</a>
<a class="sourceLine" id="cb223-5" data-line-number="5">    <span class="kw">plot</span>(current<span class="op">$</span>mean, current<span class="op">$</span>total, <span class="dt">main=</span>i, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">cex=</span><span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb223-6" data-line-number="6">        <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb223-7" data-line-number="7">    is.spike &lt;-<span class="st"> </span><span class="kw">isSpike</span>(sce<span class="fl">.416</span>b<span class="fl">.2</span>)</a>
<a class="sourceLine" id="cb223-8" data-line-number="8">    <span class="kw">points</span>(current<span class="op">$</span>mean[is.spike], current<span class="op">$</span>total[is.spike], <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</a>
<a class="sourceLine" id="cb223-9" data-line-number="9">    <span class="kw">curve</span>(<span class="kw">metadata</span>(current)<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&#39;dodgerblue&#39;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>) </a>
<a class="sourceLine" id="cb223-10" data-line-number="10">}</a></code></pre></div>
<div class="figure"><span id="fig:blocked-fit"></span>
<img src="P2_W06.feature-selection_files/figure-html/blocked-fit-1.png" alt="Variance in the 416B data set as a function of the mean after blocking on the plate of origin. Each plot represents the results for a single plate, each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins." width="672" />
<p class="caption">
Figure 10.5: Variance in the 416B data set as a function of the mean after blocking on the plate of origin. Each plot represents the results for a single plate, each point represents a gene (black) or spike-in transcript (red) and the blue line represents the trend fitted to all spike-ins.
</p>
</div>
<p>Incidentally, we run <code>multiBlockNorm()</code> to adjust the size factors within each level of the blocking factor.
Specifically, the spike-in size factors across cells in a given batch is scaled so that the mean is equal to that of the gene-based size factors for the same set of cells.
Log-normalized expression values are then recalculated using these centred size factors.
This procedure ensures that the average abundances of the spike-in transcripts are comparable to the endogenous genes,
avoiding problems due to differences in the quantity of spike-in RNA between batches.
Of course, this is not necessary if no spike-ins were available, in which case we should set <code>trend.var=list(use.spikes=FALSE)</code> in the <code>multiBlockVar()</code> call.</p>
</div>
<div id="using-a-design-matrix" class="section level4">
<h4><span class="header-section-number">10.2.4.2</span> Using a design matrix</h4>
<p>The use of block-specific trends is the recommended approach for experiments with a single blocking factor.
However, this is not practical for studies involving a large number of blocking factors and/or covariates.
In such cases, we can use the <code>design=</code> argument to specify a design matrix with uninteresting factors of variation.
<code>decomposeVar()</code> will then focus on genes with large residual variances, i.e., beyond that explained by the design matrix.
We illustrate again with the 416B data set, blocking on the plate of origin and oncogene induction.</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb224-1" data-line-number="1">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="kw">factor</span>(block) <span class="op">+</span><span class="st"> </span>phenotype, <span class="kw">colData</span>(sce<span class="fl">.416</span>b))</a>
<a class="sourceLine" id="cb224-2" data-line-number="2">fit<span class="fl">.416</span>b<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce<span class="fl">.416</span>b, <span class="dt">design=</span>design)</a>
<a class="sourceLine" id="cb224-3" data-line-number="3">dec<span class="fl">.416</span>b<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce<span class="fl">.416</span>b, fit<span class="fl">.416</span>b<span class="fl">.2</span>)</a>
<a class="sourceLine" id="cb224-4" data-line-number="4">dec<span class="fl">.416</span>b<span class="fl">.2</span>[<span class="kw">order</span>(dec<span class="fl">.416</span>b<span class="fl">.2</span><span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a></code></pre></div>
<pre><code>## DataFrame with 46703 rows and 6 columns
##                                mean             total               bio
##                           &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;
## Ccnb2              5.89821323258258  9.79093386488465   7.0893697170353
## Lyz2               6.57237822734111  8.84765039552134  7.07843003454142
## Gem                5.86432958703213   9.6816178335956  6.92675481460231
## ENSMUSG00000076617 6.37255828888088  7.87882496529147  5.86185528899622
## Idh1               5.98634090489272  8.40011571572379  5.83482283821305
## ...                             ...               ...               ...
## Gm7429              3.4731788564726 0.253468796782048 -5.74461281024374
## ENSMUSG00000083061 3.57276296115205 0.169207180855588 -5.75979426776692
## ENSMUSG00000081740 2.54482110453333 0.271685903926373 -5.81003429108432
## ENSMUSG00000103903  3.0916185774635 0.198365087332871 -5.92335585818509
## ENSMUSG00000095762 2.84338170706645 0.205600789787391 -5.93183550024739
##                                tech              p.value
##                           &lt;numeric&gt;            &lt;numeric&gt;
## Ccnb2              2.70156414784935 3.19116390139292e-56
## Lyz2               1.76922036097993  8.9427450119608e-99
## Gem                2.75486301899329 4.98139083384899e-53
## ENSMUSG00000076617 2.01696967629525 1.37607077983069e-64
## Idh1               2.56529287751074 3.43618243897962e-46
## ...                             ...                  ...
## Gm7429             5.99808160702579                    1
## ENSMUSG00000083061 5.92900144862251                    1
## ENSMUSG00000081740 6.08172019501069                    1
## ENSMUSG00000103903 6.12172094551797                    1
## ENSMUSG00000095762 6.13743629003478                    1
##                                     FDR
##                               &lt;numeric&gt;
## Ccnb2              1.32786609339746e-53
## Lyz2               1.81203342842357e-95
## Gem                 1.7196499142274e-50
## ENSMUSG00000076617 8.43821087147755e-62
## Idh1               8.08787102960638e-44
## ...                                 ...
## Gm7429                                1
## ENSMUSG00000083061                    1
## ENSMUSG00000081740                    1
## ENSMUSG00000103903                    1
## ENSMUSG00000095762                    1</code></pre>
<p>This strategy is simple but somewhat inaccurate as it does not consider the mean expression in each blocking level.
Briefly, the technical component is estimated as the fitted value of the trend at the average abundance for each gene.
However, the true technical component is the average of the fitted values at the per-block means, which may be quite different for strong batch effects and non-linear mean-variance relationships.
The <code>multiBlockVar()</code> approach is safer and should be preferred in all situations where it is applicable.</p>
</div>
</div>
</div>
<div id="selecting-highly-variable-genes" class="section level2">
<h2><span class="header-section-number">10.3</span> Selecting highly variable genes</h2>
<div id="based-on-the-largest-metrics" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Based on the largest metrics</h3>
<p>Once we have quantified the per-gene variation, the next step is to select the subset of HVGs to use in downstream analyses.
The simplest approach is to simply take the top <span class="math inline">\(X\)</span> genes with the largest values for the relevant measure of variation.
For <code>decomposeVar()</code>, this would be the genes with the largest biological components:</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb226-1" data-line-number="1">hvg.pbmc.var &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>), <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb226-2" data-line-number="2">hvg.pbmc.var &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[hvg.pbmc.var]</a></code></pre></div>
<p>For <code>improvedCV2()</code>, this would instead be the genes with the largest ratios:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb227-1" data-line-number="1">hvg.pbmc.cv2 &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">order</span>(diff, <span class="dt">decreasing=</span><span class="ot">TRUE</span>), <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb227-2" data-line-number="2">hvg.pbmc.cv2 &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[hvg.pbmc.cv2]</a></code></pre></div>
<p>The main advantage of this approach is that the user can easily control the number of genes retained.
This ensures that the computational complexity of downstream calculations is easily predicted.
It is also fairly easy to translate the choice of <span class="math inline">\(X\)</span> into a biological statement.
Recall our trend-fitting assumption that most genes are not differentially expressed between cell types or states in our population.
If we quantify this assumption into a statement that, e.g., no more than 5% of genes are differentially expressed, we can simply take the top 5% of genes with the largest variance.</p>
<p>The main disadvantage of this approach that it turns HVG selection into a competition between genes, whereby a subset of very highly variable genes can push other informative genes out of the top set.
This can be problematic if a single subpopulation is very different from the others.
In such cases, the top set will be dominated by differentially expressed genes involving the outlier subpopulation, compromising resolution of heterogeneity between the other populations.
Similar problems are encountered when the magnitude of the chosen variance measure varies with abundance.</p>
<p>The choice of <span class="math inline">\(X\)</span> is also fairly arbitrary, with any value from 500 to 5000 being considered “reasonable”.
We have chosen <span class="math inline">\(X=1000\)</span> in the code above though there is no particular <em>a priori</em> reason for doing so.
A larger <span class="math inline">\(X\)</span> will reduce the risk of discarding signal at the cost of increasing noise that obscures signal.
Our recommendation is to simply pick an arbitrary <span class="math inline">\(X\)</span> and proceed with the rest of the analysis, with the intention of testing other choices later, rather than spending much time worrying about obtaining the “optimal” value<a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a>.</p>
</div>
<div id="based-on-a-fixed-threshold" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Based on a fixed threshold</h3>
<p>Another approach to feature selection is to set a fixed threshold of one of the metrics.
This is most commonly done with the (adjusted) <span class="math inline">\(p\)</span>-value reported by each of the above methods.
The <span class="math inline">\(p\)</span>-value for each gene is generated by testing against the null hypothesis that the variance is equal to the trend.
For example, we might define our HVGs as all genes that have adjusted <span class="math inline">\(p\)</span>-values below 0.05.</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb228-1" data-line-number="1">hvg.pbmc.var<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[dec.pbmc<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>]</a>
<a class="sourceLine" id="cb228-2" data-line-number="2"><span class="kw">length</span>(hvg.pbmc.var<span class="fl">.2</span>)</a></code></pre></div>
<pre><code>## [1] 4282</code></pre>
<p>This approach is simple to implement and - if the test holds its size - it controls the false discovery rate (FDR).
That is, it returns a subset of genes where the proportion of false positives is expected to be below the specified threshold.
This can occasionally be useful in applications where the HVGs themselves are of interest.
For example, if a collaborator were to take a list of HVGs back to the bench to verify the existence of heterogeneous expression for some of the genes, we would want to control the FDR in that list.</p>
<p>The downside of this approach is that it is less predictable than the top <span class="math inline">\(X\)</span> strategy.
The number of genes returned depends on the type II error rate of the test and the severity of the multiple testing correction.
One might obtain no genes or every gene at a given FDR threshold, depending on the circumstances.
Moreover, control of the FDR is usually not helpful at this stage of the analysis.
We are not interpreting the individual HVGs themselves but are only using them for feature selection prior to downstream steps.
There is no reason to think that a 5% threshold on the FDR yields a more suitable compromise between bias and noise.</p>
</div>
<div id="keeping-all-genes-above-the-trend" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Keeping all genes above the trend</h3>
<p>As the title suggests, this involves keeping all genes above the trend.
The rationale is to only remove the obviously uninteresting genes with variances below the trend.
By doing so, we avoid the need to make any judgement calls regarding what level of variation is interesting enough to retain.
This approach represents one extreme of the bias-variance trade-off where bias is minimized at the cost of maximizing noise.
For <code>decomposeVar()</code>, it equates to keeping all positive biological components:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb230-1" data-line-number="1">hvg.pbmc.var<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[dec.pbmc<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb230-2" data-line-number="2"><span class="kw">length</span>(hvg.pbmc.var<span class="fl">.3</span>)</a></code></pre></div>
<pre><code>## [1] 8110</code></pre>
<p>For <code>improvedCV2()</code>, this involves keeping all ratios above 1:</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb232-1" data-line-number="1">hvg.pbmc.cv2<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">rownames</span>(cv2.pbmc)[diff <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb232-2" data-line-number="2"><span class="kw">length</span>(hvg.pbmc.cv2<span class="fl">.3</span>)</a></code></pre></div>
<pre><code>## [1] 33694</code></pre>
<p>This strategy is the most conservative as it does not discard any potential biological signal.
Weak or secondary population structure is given the chance to manifest as the affected genes are retained.
This makes it useful for reliable automated processing of diverse data sets where the primary factor of variation in one data set is a secondary factor in another data set (and thus overlooked by the top <span class="math inline">\(X\)</span> approach).
The obvious cost is that more noise is also captured, which can reduce the resolution of otherwise well-separated populations.
From a practical perspective, the use of more genes involves more computational work in each downstream step.</p>
</div>
</div>
<div id="putting-it-all-together-1" class="section level2">
<h2><span class="header-section-number">10.4</span> Putting it all together</h2>
<p>The few lines of code below will select the top 10% of genes with the highest biological components.</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb234-1" data-line-number="1">fit.pbmc &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce.pbmc, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb234-2" data-line-number="2">dec.pbmc &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(<span class="dt">fit=</span>fit.pbmc)</a>
<a class="sourceLine" id="cb234-3" data-line-number="3">chosen &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[<span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)]</a>
<a class="sourceLine" id="cb234-4" data-line-number="4">chosen &lt;-<span class="st"> </span><span class="kw">head</span>(chosen, <span class="kw">nrow</span>(dec.pbmc) <span class="op">*</span><span class="st"> </span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb234-5" data-line-number="5"><span class="kw">length</span>(chosen)</a></code></pre></div>
<pre><code>## [1] 3369</code></pre>
<p>We can then subset the <code>SingleCellExperiment</code> to only retain our selection of HVGs.
This ensures that downstream methods will only use these genes for their calculations.</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb236-1" data-line-number="1">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[chosen,]</a>
<a class="sourceLine" id="cb236-2" data-line-number="2"><span class="kw">dim</span>(sce.pbmc)</a></code></pre></div>
<pre><code>## [1] 3369 3922</code></pre>
<p>Alternatively, some methods may allow users to pass in the full <code>SingleCellExperiment</code> object and specify the genes to use via an extra argument like <code>subset.row=</code>.
This may be more convenient in the context of the overall analysis, where genes outside of this subset may still be of interest during DE analyses or for visualization.</p>
</div>
<div id="session-info-4" class="section level2">
<h2><span class="header-section-number">10.5</span> Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.6 LTS

Matrix products: default
BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scran_1.13.9                SingleCellExperiment_1.7.0 
 [3] SummarizedExperiment_1.15.5 DelayedArray_0.11.4        
 [5] BiocParallel_1.19.0         matrixStats_0.54.0         
 [7] Biobase_2.45.0              GenomicRanges_1.37.14      
 [9] GenomeInfoDb_1.21.1         IRanges_2.19.10            
[11] S4Vectors_0.23.17           BiocGenerics_0.31.5        
[13] BiocStyle_2.13.2            Cairo_1.5-10               

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2               rsvd_1.0.2              
 [3] locfit_1.5-9.1           lattice_0.20-38         
 [5] assertthat_0.2.1         digest_0.6.20           
 [7] R6_2.4.0                 dynamicTreeCut_1.63-1   
 [9] evaluate_0.14            ggplot2_3.2.0           
[11] pillar_1.4.2             zlibbioc_1.31.0         
[13] rlang_0.4.0              lazyeval_0.2.2          
[15] irlba_2.3.3              Matrix_1.2-17           
[17] rmarkdown_1.14           BiocNeighbors_1.3.3     
[19] statmod_1.4.32           stringr_1.4.0           
[21] igraph_1.2.4.1           RCurl_1.95-4.12         
[23] munsell_0.5.0            vipor_0.4.5             
[25] compiler_3.6.0           BiocSingular_1.1.5      
[27] xfun_0.8                 pkgconfig_2.0.2         
[29] ggbeeswarm_0.6.0         htmltools_0.3.6         
[31] tidyselect_0.2.5         gridExtra_2.3           
[33] tibble_2.1.3             GenomeInfoDbData_1.2.1  
[35] bookdown_0.12            edgeR_3.27.9            
[37] viridisLite_0.3.0        crayon_1.3.4            
[39] dplyr_0.8.3              bitops_1.0-6            
[41] grid_3.6.0               gtable_0.3.0            
[43] magrittr_1.5             scales_1.0.0            
[45] dqrng_0.2.1              stringi_1.4.3           
[47] XVector_0.25.0           viridis_0.5.1           
[49] limma_3.41.15            scater_1.13.9           
[51] DelayedMatrixStats_1.7.1 tools_3.6.0             
[53] beeswarm_0.2.3           glue_1.3.1              
[55] purrr_0.3.2              yaml_2.2.0              
[57] colorspace_1.4-1         BiocManager_1.30.4      
[59] knitr_1.23              </code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-brennecke2013accounting">
<p>Brennecke, P., S. Anders, J. K. Kim, A. A. Kołodziejczyk, X. Zhang, V. Proserpio, B. Baying, et al. 2013. “Accounting for technical noise in single-cell RNA-seq experiments.” <em>Nat. Methods</em> 10 (11):1093–5.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="9">
<li id="fn9"><p>When everything’s said and done, the “optimal” value is one that gets you a figure for your manuscript.<a href="feature-selection-2.html#fnref9" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="normalization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dimensionality-reduction-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W06.feature-selection.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
