<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Robert A. Amezquita" />
<meta name="author" content="Stephanie C. Hicks" />


<meta name="date" content="2019-06-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="clustering-2.html">
<link rel="next" href="differential-expression-2.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#introducing-the-dataset"><i class="fa fa-check"></i><b>8.2</b> Introducing the dataset</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.3</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.4</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.4.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.4.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.4.2" data-path="quality-control.html"><a href="quality-control.html#with-adaptive-thresholds"><i class="fa fa-check"></i><b>8.4.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.4.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.4.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.5</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.6</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.6.1</b> Background</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.6.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.6.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.6.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.7</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.7.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.7.1</b> Overview</a></li>
<li class="chapter" data-level="8.7.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.7.2</b> Checking for discarded cell types</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>9</b> Clustering</a><ul>
<li class="chapter" data-level="9.1" data-path="clustering-2.html"><a href="clustering-2.html#loading-the-data"><i class="fa fa-check"></i><b>9.1</b> Loading the Data</a></li>
<li class="chapter" data-level="9.2" data-path="clustering-2.html"><a href="clustering-2.html#manual-clustering"><i class="fa fa-check"></i><b>9.2</b> Manual Clustering</a><ul>
<li class="chapter" data-level="9.2.1" data-path="clustering-2.html"><a href="clustering-2.html#varying-k"><i class="fa fa-check"></i><b>9.2.1</b> Varying <code>k</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="clustering-2.html"><a href="clustering-2.html#varying-clustering-algorithms"><i class="fa fa-check"></i><b>9.2.2</b> Varying Clustering Algorithms</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="clustering-2.html"><a href="clustering-2.html#automated-clustering"><i class="fa fa-check"></i><b>9.3</b> Automated Clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>10</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="10.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preprocessing-1"><i class="fa fa-check"></i><b>10.1</b> Preprocessing</a></li>
<li class="chapter" data-level="10.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#fast-mnn"><i class="fa fa-check"></i><b>10.2</b> Fast MNN</a></li>
<li class="chapter" data-level="10.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#harmony"><i class="fa fa-check"></i><b>10.3</b> Harmony</a></li>
<li class="chapter" data-level="10.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#advanced-mnn-workflow"><i class="fa fa-check"></i><b>10.4</b> Advanced MNN Workflow</a><ul>
<li class="chapter" data-level="10.4.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>10.4.1</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#naive-method-without-correction"><i class="fa fa-check"></i><b>10.5</b> Naive Method Without Correction</a></li>
<li class="chapter" data-level="10.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#limma-batch-correction"><i class="fa fa-check"></i><b>10.6</b> Limma Batch Correction</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-expression-2.html"><a href="differential-expression-2.html"><i class="fa fa-check"></i><b>11</b> Differential Expression</a></li>
<li class="chapter" data-level="12" data-path="trajectory-analysis-1.html"><a href="trajectory-analysis-1.html"><i class="fa fa-check"></i><b>12</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="13" data-path="annotation-2.html"><a href="annotation-2.html"><i class="fa fa-check"></i><b>13</b> Annotation</a></li>
<li class="chapter" data-level="14" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>14</b> Sharing and Interactive Interfaces</a></li>
<li class="chapter" data-level="15" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>15</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="15.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>15.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="15.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>15.2</b> Parallelization</a></li>
<li class="chapter" data-level="15.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>15.3</b> On-Disk Data</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="data-import-and-export.html"><a href="data-import-and-export.html"><i class="fa fa-check"></i><b>16</b> Data Import and Export</a></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>17</b> About the Data</a><ul>
<li class="chapter" data-level="17.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>17.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="17.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>17.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>18</b> About the Contributors</a><ul>
<li class="chapter" data-level="18.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>18.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="18.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>18.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="18.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>18.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="18.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>18.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="integrating-datasets-1" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Integrating Datasets</h1>
<p>As scRNA-seq approaches continue to gain in popularity, projects that combine independently generated datasets will become increasingly common. While generalized linear modeling frameworks can be used to integrate disparate datasets, the performance of these frameworks in the scRNA-seq context may be sub-optimal, often due to an underlying assumption that the composition of cell populations is either known or identical across batches of cells. This has inspired a new class of methods that leverage the unique properties of scRNA-seq to improve the amelioration of batch effects, which we demonstrate here.</p>
<p>Here we will work with two PBMC datasets from the <em>TENxPBMCData</em> package - the <code>pbmc3k</code> and <code>pbmc4k</code> datasets - that come from different donors, and integrate them to produce an integrated representation of the data.</p>
<div id="preprocessing-1" class="section level2">
<h2><span class="header-section-number">10.1</span> Preprocessing</h2>
<p>Before proceeding to the integration section, we will first load the data, and subsequently run it through a preprocessing pipeline that ensures we have a common set of genes across both experiments, filters out poor quality cells, and normalizes the expression matrix. The code for this is shown below for completeness.
Note that here we make use of the <em>purrr</em> package both for code succinctness and to apply the same function to each PBMC dataset contained within the <code>pbmc_l</code> list. (see base <code>lapply()</code> and <code>purrrr::map()</code> for more details).</p>
<p>Furthermore, to speed up computation, we will subsample each of these datasets down to 1000 cells each.</p>
<pre class="sourceCode r"><code class="sourceCode r">## Load the data
<span class="kw">library</span>(TENxPBMCData)
pbmc3k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc3k&#39;</span>)
pbmc4k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)
pbmc_l &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pbmc3k =</span> pbmc3k, <span class="dt">pbmc4k =</span> pbmc4k)

## Subsample data to 1000 cells each
<span class="kw">library</span>(purrr)
<span class="kw">set.seed</span>(<span class="dv">1234</span>)
pbmc_l &lt;-<span class="st"> </span><span class="kw">map</span>(pbmc_l, <span class="cf">function</span>(x) { x[, <span class="kw">sample</span>(<span class="kw">ncol</span>(x), <span class="dv">1000</span>)] })

## Retain common genes (rows); reorder rows by common genes
common_genes &lt;-<span class="st"> </span><span class="kw">reduce</span>(<span class="kw">map</span>(pbmc_l, rownames), intersect)
pbmc_l &lt;-<span class="st"> </span><span class="kw">map</span>(pbmc_l, <span class="cf">function</span>(x) { x[common_genes, ] })

## Save rowData from first object
## remove from each individual object else `cbind` will throw an error
rd &lt;-<span class="st"> </span><span class="kw">rowData</span>(pbmc_l[[<span class="dv">1</span>]])
pbmc_l &lt;-<span class="st"> </span><span class="kw">map</span>(pbmc_l, <span class="cf">function</span>(x) {
    <span class="kw">rowData</span>(x) &lt;-<span class="st"> </span><span class="ot">NULL</span>
    <span class="kw">return</span>(x)
})

## Combine the objects via `cbind`
sce &lt;-<span class="st"> </span><span class="kw">reduce</span>(pbmc_l, cbind)

## Replace rowData within combined object
## Uniquify row ids to use mostly human readable rownames
<span class="kw">rowData</span>(sce) &lt;-<span class="st"> </span>rd
<span class="kw">rownames</span>(sce) &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">uniquifyFeatureNames</span>(rd<span class="op">$</span>ENSEMBL_ID, rd<span class="op">$</span>Symbol_TENx)</code></pre>
<!-- ## HOTFIX NEEDED - RECAST ASSAY INTO MATRIX  -->
<pre class="sourceCode r"><code class="sourceCode r">## Calculate and apply QC metrics; filter out cells with low lib or low genes detected
<span class="kw">library</span>(scater)
sce &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce)
low_lib &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="op">$</span>log10_total_counts, <span class="dt">type =</span> <span class="st">&quot;lower&quot;</span>, <span class="dt">nmad =</span> <span class="dv">3</span>)
low_genes &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="op">$</span>log10_total_features_by_counts, <span class="dt">type =</span> <span class="st">&quot;lower&quot;</span>, <span class="dt">nmad =</span> <span class="dv">3</span>)
sce &lt;-<span class="st"> </span>sce[, <span class="op">!</span>(low_lib <span class="op">|</span><span class="st"> </span>low_genes)]

## normalize data
sce &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce)

## feature selection
<span class="kw">library</span>(scran)
fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce, <span class="dt">use.spikes =</span> <span class="ot">FALSE</span>)
dec &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce, fit)
hvg &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec)[dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>] <span class="co"># save gene names</span></code></pre>
<p>For faster calculations, here we will subset the <code>hvg</code> object to the top 2000 genes:</p>
<pre class="sourceCode r"><code class="sourceCode r">hvg &lt;-<span class="st"> </span>hvg[<span class="dv">1</span><span class="op">:</span><span class="dv">2000</span>]</code></pre>
<p>From the combined dataset encapsulated within <code>sce</code>, we note that the origin of each cell is annotated within <code>colData</code> under the column <code>Sample</code>. This will be used as the batch variable throughout.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(sce<span class="op">$</span>Sample) <span class="co"># alternately: colData(sce)$Sample</span></code></pre>
<pre><code>## 
## pbmc3k pbmc4k 
##    990   1000</code></pre>
</div>
<div id="fast-mnn" class="section level2">
<h2><span class="header-section-number">10.2</span> Fast MNN</h2>
<p>The <em>batchelor</em> package provides the <code>fastMNN()</code> function, which calculates a dimension reduced representation based on the original principal components derived directly from the normalized expression data. These MNN coordinates can then be used for further dimensionality reduction via UMAP as shown below.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(batchelor)

## create new object from sce for mnn simple workflow
sce_mnn_simple &lt;-<span class="st"> </span>sce

## run MNN with no advanced preprocessing
## BSPARAM/BNPARAM/BPPARAM explained in &quot;adaptations to large scale data&quot;
mnn_simple_out &lt;-<span class="st"> </span>batchelor<span class="op">::</span><span class="kw">fastMNN</span>(sce_mnn_simple,
                                     <span class="dt">subset.row =</span> hvg,
                                     <span class="dt">batch =</span> sce_mnn_simple<span class="op">$</span>Sample)
                                     ## BSPARAM = BiocSingular::IrlbaParam(),
                                     ## BNPARAM = BiocNeighbors::AnnoyParam())

## add MNN_simple coordinates to the new sce object
<span class="kw">reducedDim</span>(sce_mnn_simple, <span class="st">&#39;MNN&#39;</span>) &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(mnn_simple_out, <span class="st">&#39;corrected&#39;</span>)

## run UMAP on the &quot;simple MNN&quot; version
sce_mnn_simple &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce_mnn_simple, <span class="dt">use_dimred =</span> <span class="st">&#39;MNN&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(patchwork)
p1 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_mnn_simple, <span class="st">&#39;MNN&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
p2 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_mnn_simple, <span class="st">&#39;UMAP&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
<span class="kw">wrap_plots</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-7"></span>
<img src="P2_W3.integrating-datasets_files/figure-html/unnamed-chunk-7-1.png" alt="Simple MNN method. First two MNN coordinates (left) and UMAP derived from the full MNN coordinate set (right)." width="672" />
<p class="caption">
Figure 7.1: Simple MNN method. First two MNN coordinates (left) and UMAP derived from the full MNN coordinate set (right).
</p>
</div>
</div>
<div id="harmony" class="section level2">
<h2><span class="header-section-number">10.3</span> Harmony</h2>
<p>Harmony provides a unified framework for data visualization, analysis, and interpretation of scRNA-seq data. Included is a method for integrating different batches of scRNA-seq. This produces a new dimension reduction representation, saved as <code>HARMONY</code> within the <em>SingleCellExperiment</em> object, as shown below. Furthermore, this Harmony embedding can be used to produce a UMAP representation as well:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(harmony) <span class="co"># devtools::install_github(&#39;immunogenomics/harmony&#39;)</span>
<span class="kw">library</span>(BiocSingular)

## create a new sce for harmony
sce_harmony &lt;-<span class="st"> </span>sce

## Add PCA to sce, then run Harmony as its dependent on PCA
sce_harmony &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce_harmony,
                      <span class="dt">feature_set =</span> hvg,
                      <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(),
                      <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())
sce_harmony &lt;-<span class="st"> </span><span class="kw">RunHarmony</span>(sce_harmony, <span class="dt">group.by.vars =</span> <span class="st">&quot;Sample&quot;</span>)

## Calculate UMAP on Harmony embeddings
sce_harmony &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce_harmony, <span class="dt">use_dimred =</span> <span class="st">&#39;HARMONY&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_harmony, <span class="st">&#39;HARMONY&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
p2 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_harmony, <span class="st">&#39;UMAP&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
<span class="kw">wrap_plots</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-9"></span>
<img src="P2_W3.integrating-datasets_files/figure-html/unnamed-chunk-9-1.png" alt="Plot of the first two components of the Harmony embedding (left). UMAP calculated on the Harmony embeddings. (right)" width="672" />
<p class="caption">
Figure 10.1: Plot of the first two components of the Harmony embedding (left). UMAP calculated on the Harmony embeddings. (right)
</p>
</div>
</div>
<div id="advanced-mnn-workflow" class="section level2">
<h2><span class="header-section-number">10.4</span> Advanced MNN Workflow</h2>
<p>Here we work through a more advanced workflow using MNN as above, but this time walking through the various steps that are otherwise implicit within the <code>fastMNN()</code> function. While not always necessary, it may be informative for those interested in the finer details of a batch correction workflow. Note that this follows the</p>
<p>For more details, we refer to the <a href="https://github.com/MarioniLab/compareSingleCell/"><code>compareSingleCell</code> vignette</a> vignette by Aaron Lun.</p>
<!-- ### Feature Selection -->
<!-- We briefly highlight feature selection within the preprocessing, as the choice of genes used in the integration can have an outsize effect on end result.  -->
<!-- For brevity, here we use the `scran` package to identify the genes with biological coefficients of variation greater than zero, but leave the choice of method up to the reader. Note that, compared to the basic analysis workflow, we use the `multiBlockVar()` function and block on the origin (`pbmc3k` vs `pbmc4k`). This function models the variance of expression in each block separately. -->
<!-- Additionally, we turn off weighting such that each sample contributes equally to the results, regardless of the number of cells per sample. This prevents one condition from biasing the combined statistics.  -->
<!-- For more details, we refer to the [`compareSingleCell` merge](https://github.com/MarioniLab/compareSingleCell/blob/master/vignettes/embryo_merge.Rmd) vignette. -->
<pre class="sourceCode r"><code class="sourceCode r">## create a new sce for advanced mnn
sce_mnn_advanced &lt;-<span class="st"> </span>sce</code></pre>
<!-- ```{r} -->
<!-- library(scran) -->
<!-- dec <- multiBlockVar(sce_mnn_advanced, block = sce_mnn_advanced$Sample, -->
<!--                      make.tech.trend = TRUE, -->
<!--                      weighted = FALSE) -->
<!-- hvg <- rownames(dec)[dec$bio > 0] -->
<!-- ``` -->
<!-- In this example, we get back 6001 genes. -->
<!-- ```{r} -->
<!-- hvg_subset <- hvg[1:1000] -->
<!-- ``` -->
<div id="dimensionality-reduction-2" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Dimensionality Reduction</h3>
<p>Principal components analyses can be used to both reduce computational work and remove high-dimensional noise, as discussed in the <a href="https://bioconductor.org/packages/3.10/simpleSingleCell/vignettes/reads.html#denoising-expression-values-using-pca">simpleSingleCell vignette</a>.</p>
<p>To accomplish this, the <code>multiBatchPCA()</code> function from <em>scran</em> ensures that each sample contributes equally to the new coordinate space. For further discussion on this point, see the <a href="https://bioconductor.org/packages/3.10/simpleSingleCell/vignettes/batch.html#hierarchical-merging">simpleSingleCell vignette</a> section on hierarchical merging.</p>
<p>We then calculate the number of PCs to retain by comparing the variance explained per principal component versus the technical noise and total variance in the data. This will leave us with a subset of PCA components.</p>
<p>Note that we will use the features selected from the preproprocessing step earlier in this chapter.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(batchelor)
<span class="kw">library</span>(BiocSingular) <span class="co"># provide approximate pca via IrlbaParam()</span>

## Calculate PCA by batch
<span class="kw">set.seed</span>(<span class="dv">1234</span>) <span class="co"># for Random/IrlbaParam backends</span>

## run batchelor version of multiBatchPCA (use :: notation to override namespace)
pcs &lt;-<span class="st"> </span>batchelor<span class="op">::</span><span class="kw">multiBatchPCA</span>(sce_mnn_advanced,
                                <span class="dt">batch =</span> sce_mnn_advanced<span class="op">$</span>Sample,
                                <span class="dt">subset.row =</span> hvg,
                                <span class="dt">get.variance =</span> <span class="ot">TRUE</span>,
                                <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(),
                                <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

## Retain only the top PCs based on variance explained &gt; technical noise
retain &lt;-<span class="st"> </span><span class="kw">denoisePCANumber</span>(
    <span class="kw">metadata</span>(pcs)<span class="op">$</span>var.explained, <span class="co"># variance explained per PC.</span>
    <span class="kw">sum</span>(dec[hvg, ]<span class="op">$</span>tech),       <span class="co"># technical noise in subset of genes.</span>
    <span class="kw">metadata</span>(pcs)<span class="op">$</span>var.total      <span class="co"># total variance in the data</span>
)

retain</code></pre>
<pre><code>## [1] 3</code></pre>
<p>From the number of PCs to retain, we can filter the PCs as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">## Subset PCA number of components up to the number in retain
top_pcs &lt;-<span class="st"> </span><span class="kw">map</span>(<span class="kw">as.list</span>(pcs), <span class="op">~</span><span class="st"> </span>.[, <span class="dv">1</span><span class="op">:</span>retain])</code></pre>
<p>Then, using the retained components as input, we can input them directly into the <code>fastMNN()</code> method as follows, and furthermore produce a UMAP representation based on these MNN coordinates:</p>
<pre class="sourceCode r"><code class="sourceCode r">## Calculate MNN coordinates, add into reducedDims
mnn_advanced_out &lt;-<span class="st"> </span>batchelor<span class="op">::</span><span class="kw">fastMNN</span>(top_pcs<span class="op">$</span>pbmc3k,
                                       top_pcs<span class="op">$</span>pbmc4k,
                                       <span class="dt">pc.input =</span> <span class="ot">TRUE</span>)
<span class="kw">reducedDim</span>(sce_mnn_advanced, <span class="st">&#39;MNN&#39;</span>) &lt;-<span class="st"> </span>mnn_advanced_out<span class="op">$</span>corrected

## Run UMAP on MNN coordinates
sce_mnn_advanced &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce_mnn_advanced, <span class="dt">use_dimred =</span> <span class="st">&#39;MNN&#39;</span>)</code></pre>
<p>The MNN coordinates and the UMAP coordinates derived from them are visualized below:</p>
<pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_mnn_advanced, <span class="st">&#39;MNN&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
p2 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_mnn_advanced, <span class="st">&#39;UMAP&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
<span class="kw">wrap_plots</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-14"></span>
<img src="P2_W3.integrating-datasets_files/figure-html/unnamed-chunk-14-1.png" alt="Plot of first two MNN coordinates (left). UMAP calculated on the MNN coordinates (right). Colour denotes origin of sample." width="672" />
<p class="caption">
Figure 10.2: Plot of first two MNN coordinates (left). UMAP calculated on the MNN coordinates (right). Colour denotes origin of sample.
</p>
</div>
<p>Comparing to the simpler MNN workflow shown above (which only used the <code>fastMNN()</code> function on the barely processed <em>SingleCellExperiment</em> class object <code>sce</code>), we see that the two results are fairly comparable, differing only by the number of principal components and the selected number of features (here we chose 1000 for computational efficacy; we leave it to the interested reader to explore the results with a larger feature space).</p>
</div>
</div>
<div id="naive-method-without-correction" class="section level2">
<h2><span class="header-section-number">10.5</span> Naive Method Without Correction</h2>
<p>For completeness, we show here the results of performing the correction solely on the gene expression matrix.</p>
<pre class="sourceCode r"><code class="sourceCode r">## create a new sce for naive method
sce_naive &lt;-<span class="st"> </span>sce</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## run PCA on only the normalized gene expression data
sce_naive &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce_naive,
                    <span class="dt">feature_set =</span> hvg,
                    <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(),
                    <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

## Run UMAP on the norm gene exp. derived PCA coordinates - use only top PCs
sce_naive &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce_naive, <span class="dt">use_dimred =</span> <span class="st">&#39;PCA&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(sce_naive, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
p2 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_naive, <span class="st">&#39;UMAP&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
<span class="kw">wrap_plots</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-17"></span>
<img src="P2_W3.integrating-datasets_files/figure-html/unnamed-chunk-17-1.png" alt="First two components of PCA calculated directly from the gene expression matrix (left). UMAP calculated on the PCA derived from the normalized gene expression matrix (right)." width="672" />
<p class="caption">
Figure 10.3: First two components of PCA calculated directly from the gene expression matrix (left). UMAP calculated on the PCA derived from the normalized gene expression matrix (right).
</p>
</div>
<p>While the PCA looks well integrated, the UMAP representation - which encapsulates a fuller PCA space - distinctly shows the residual batch effect present.</p>
</div>
<div id="limma-batch-correction" class="section level2">
<h2><span class="header-section-number">10.6</span> Limma Batch Correction</h2>
<p>The <code>limma</code> package, a popular framework for the statistical analysis of RNA-seq, has a function <code>removeBatchEffect()</code> which will be used here to correct the normalized expression matrix <code>logcounts</code> across the two batches. The result will be assigned into the <code>assays</code> slot of the <code>sce</code> object as <code>limma_corrected</code>, and then used for PCA, saving the result in the <code>reducedDim</code> slot as <code>&quot;PCA_limma&quot;</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">## create a new sce for limma method
sce_limma &lt;-<span class="st"> </span>sce</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)
limma_corrected &lt;-<span class="st"> </span><span class="kw">removeBatchEffect</span>(<span class="kw">logcounts</span>(sce_limma), <span class="dt">batch =</span> sce_limma<span class="op">$</span>Sample)
<span class="kw">assay</span>(sce_limma, <span class="st">&quot;logcounts_limma&quot;</span>) &lt;-<span class="st"> </span>limma_corrected ## add new assay

## calc a PCA on limma values; save separate to prevent overwriting
sce_limma &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce_limma,
                    <span class="dt">feature_set =</span> hvg,
                    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_limma&quot;</span>,
                    <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(),
                    <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

## run UMAP; save separate to prevent overwriting
sce_limma &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce_limma, <span class="dt">use_dimred =</span> <span class="st">&#39;PCA&#39;</span>)</code></pre>
<p>The resulting PCA and UMAP from the limma batch correction method are shown below:</p>
<pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_limma, <span class="st">&#39;PCA&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
p2 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce_limma, <span class="st">&#39;UMAP&#39;</span>, <span class="dt">colour_by =</span> <span class="st">&#39;Sample&#39;</span>)
<span class="kw">wrap_plots</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-20"></span>
<img src="P2_W3.integrating-datasets_files/figure-html/unnamed-chunk-20-1.png" alt="PCA calculated on the limma batch-corrected gene expression matrix (left). UMAP calculated via the PCA derived from limma batch-corrected gene expression matrix (right)." width="672" />
<p class="caption">
Figure 10.4: PCA calculated on the limma batch-corrected gene expression matrix (left). UMAP calculated via the PCA derived from limma batch-corrected gene expression matrix (right).
</p>
</div>
<!-- ```{r} -->
<!-- ## chunk for combining 1k cells from pbmc3k and pbmc4k -->
<!-- ## ... -->
<!-- ## separate chunk -->
<!-- ## Create two SCEs, reset reducedDims -->
<!-- sce_mnn_25 <- sce -->
<!-- sce_mnn_7 <- sce -->
<!-- reducedDims(sce_mnn_25) <- reducedDims(sce_mnn_7) <- NULL -->
<!-- ## Run MNN with differing component numbers -->
<!-- retain <- 25 -->
<!-- top_pcs <- map(as.list(pcs), ~ .[, 1:retain]) -->
<!-- mnn <- fastMNN(top_pcs$pbmc3k, top_pcs$pbmc4k, pc.input = TRUE) -->
<!-- reducedDim(sce_mnn_25, 'MNN') <- mnn$corrected -->
<!-- sce_mnn_25 <- runUMAP(sce_mnn_25, use_dimred = 'MNN') -->
<!-- retain <- 7 -->
<!-- top_pcs <- map(as.list(pcs), ~ .[, 1:retain]) -->
<!-- mnn <- fastMNN(top_pcs$pbmc3k, top_pcs$pbmc4k, pc.input = TRUE) -->
<!-- reducedDim(sce_mnn_7, 'MNN') <- mnn$corrected -->
<!-- sce_mnn_7 <- runUMAP(sce_mnn_7, use_dimred = 'MNN') -->
<!-- ## Plot on plot -->
<!-- plotReducedDim(sce_mnn_25, use_dimred = 'UMAP', colour_by = 'Sample') -->
<!-- dev.new() -->
<!-- plotReducedDim(sce_mnn_7, use_dimred = 'UMAP', colour_by = 'Sample') -->
<!-- ``` -->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="clustering-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="differential-expression-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W3.integrating-datasets.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
