<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2020-05-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="cell-type-annotation.html"/>
<link rel="next" href="multi-sample-comparisons.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background"><i class="fa fa-check"></i><b>4.1</b> Background</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#storing-primary-experimental-data"><i class="fa fa-check"></i><b>4.2</b> Storing primary experimental data</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#filling-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Filling the <code>assays</code> slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#adding-more-assays"><i class="fa fa-check"></i><b>4.2.2</b> Adding more <code>assays</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#handling-metadata"><i class="fa fa-check"></i><b>4.3</b> Handling metadata</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#on-the-columns"><i class="fa fa-check"></i><b>4.3.1</b> On the columns</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#on-the-rows"><i class="fa fa-check"></i><b>4.3.2</b> On the rows</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#other-metadata"><i class="fa fa-check"></i><b>4.3.3</b> Other metadata</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#single-cell-specific-fields"><i class="fa fa-check"></i><b>4.4</b> Single-cell-specific fields</a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background-1"><i class="fa fa-check"></i><b>4.4.1</b> Background</a></li>
<li class="chapter" data-level="4.4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#dimensionality-reduction-results"><i class="fa fa-check"></i><b>4.4.2</b> Dimensionality reduction results</a></li>
<li class="chapter" data-level="4.4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#alternative-experiments"><i class="fa fa-check"></i><b>4.4.3</b> Alternative Experiments</a></li>
<li class="chapter" data-level="4.4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors"><i class="fa fa-check"></i><b>4.4.4</b> Size factors</a></li>
<li class="chapter" data-level="4.4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-labels"><i class="fa fa-check"></i><b>4.4.5</b> Column labels</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#conclusion"><i class="fa fa-check"></i><b>4.5</b> Conclusion</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="5" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>5</b> Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="overview.html"><a href="overview.html#introduction-1"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="overview.html"><a href="overview.html#experimental-design"><i class="fa fa-check"></i><b>5.2</b> Experimental Design</a></li>
<li class="chapter" data-level="5.3" data-path="overview.html"><a href="overview.html#obtaining-a-count-matrix"><i class="fa fa-check"></i><b>5.3</b> Obtaining a count matrix</a></li>
<li class="chapter" data-level="5.4" data-path="overview.html"><a href="overview.html#data-processing-and-downstream-analysis"><i class="fa fa-check"></i><b>5.4</b> Data processing and downstream analysis</a></li>
<li class="chapter" data-level="5.5" data-path="overview.html"><a href="overview.html#quick-start"><i class="fa fa-check"></i><b>5.5</b> Quick start</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>6</b> Quality Control</a><ul>
<li class="chapter" data-level="6.1" data-path="quality-control.html"><a href="quality-control.html#quality-control-motivation"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>6.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="6.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>6.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="6.3.1" data-path="quality-control.html"><a href="quality-control.html#fixed-qc"><i class="fa fa-check"></i><b>6.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="6.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>6.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="6.3.3" data-path="quality-control.html"><a href="quality-control.html#other-approaches"><i class="fa fa-check"></i><b>6.3.3</b> Other approaches</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="quality-control.html"><a href="quality-control.html#quality-control-plots"><i class="fa fa-check"></i><b>6.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="6.5" data-path="quality-control.html"><a href="quality-control.html#qc-droplets"><i class="fa fa-check"></i><b>6.5</b> Cell calling for droplet data</a><ul>
<li class="chapter" data-level="6.5.1" data-path="quality-control.html"><a href="quality-control.html#background-2"><i class="fa fa-check"></i><b>6.5.1</b> Background</a></li>
<li class="chapter" data-level="6.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>6.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="6.5.3" data-path="quality-control.html"><a href="quality-control.html#relationship-with-other-qc-metrics"><i class="fa fa-check"></i><b>6.5.3</b> Relationship with other QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="quality-control.html"><a href="quality-control.html#quality-control-discarded"><i class="fa fa-check"></i><b>6.6</b> Removing low-quality cells</a></li>
<li class="chapter" data-level="6.7" data-path="quality-control.html"><a href="quality-control.html#marking-qc"><i class="fa fa-check"></i><b>6.7</b> Marking low-quality cells</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>7</b> Normalization</a><ul>
<li class="chapter" data-level="7.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>7.1</b> Motivation</a></li>
<li class="chapter" data-level="7.2" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>7.2</b> Library size normalization</a></li>
<li class="chapter" data-level="7.3" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>7.3</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="7.4" data-path="normalization.html"><a href="normalization.html#spike-norm"><i class="fa fa-check"></i><b>7.4</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="7.5" data-path="normalization.html"><a href="normalization.html#applying-the-size-factors"><i class="fa fa-check"></i><b>7.5</b> Applying the size factors</a><ul>
<li class="chapter" data-level="7.5.1" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>7.5.1</b> Scaling and log-transforming</a></li>
<li class="chapter" data-level="7.5.2" data-path="normalization.html"><a href="normalization.html#downsampling-and-log-transforming"><i class="fa fa-check"></i><b>7.5.2</b> Downsampling and log-transforming</a></li>
<li class="chapter" data-level="7.5.3" data-path="normalization.html"><a href="normalization.html#other-options"><i class="fa fa-check"></i><b>7.5.3</b> Other options</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="normalization.html"><a href="normalization.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>8</b> Feature selection</a><ul>
<li class="chapter" data-level="8.1" data-path="feature-selection.html"><a href="feature-selection.html#motivation-1"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="feature-selection.html"><a href="feature-selection.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>8.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="8.2.1" data-path="feature-selection.html"><a href="feature-selection.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>8.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="8.2.2" data-path="feature-selection.html"><a href="feature-selection.html#coefficient-of-variation"><i class="fa fa-check"></i><b>8.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="8.2.3" data-path="feature-selection.html"><a href="feature-selection.html#sec:spikeins"><i class="fa fa-check"></i><b>8.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="8.2.4" data-path="feature-selection.html"><a href="feature-selection.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>8.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="feature-selection.html"><a href="feature-selection.html#hvg-selection"><i class="fa fa-check"></i><b>8.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="8.3.1" data-path="feature-selection.html"><a href="feature-selection.html#overview-1"><i class="fa fa-check"></i><b>8.3.1</b> Overview</a></li>
<li class="chapter" data-level="8.3.2" data-path="feature-selection.html"><a href="feature-selection.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>8.3.2</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="8.3.3" data-path="feature-selection.html"><a href="feature-selection.html#based-on-significance"><i class="fa fa-check"></i><b>8.3.3</b> Based on significance</a></li>
<li class="chapter" data-level="8.3.4" data-path="feature-selection.html"><a href="feature-selection.html#feature-selection-positive"><i class="fa fa-check"></i><b>8.3.4</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="feature-selection.html"><a href="feature-selection.html#apriori-hvgs"><i class="fa fa-check"></i><b>8.4</b> Selecting <em>a priori</em> genes of interest</a></li>
<li class="chapter" data-level="8.5" data-path="feature-selection.html"><a href="feature-selection.html#feature-selection-subsetting"><i class="fa fa-check"></i><b>8.5</b> Putting it all together</a></li>
<li class="chapter" data-level="" data-path="feature-selection.html"><a href="feature-selection.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>9</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="9.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#overview-2"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-components-analysis"><i class="fa fa-check"></i><b>9.2</b> Principal components analysis</a></li>
<li class="chapter" data-level="9.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>9.3</b> Choosing the number of PCs</a><ul>
<li class="chapter" data-level="9.3.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#motivation-2"><i class="fa fa-check"></i><b>9.3.1</b> Motivation</a></li>
<li class="chapter" data-level="9.3.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#using-the-elbow-point"><i class="fa fa-check"></i><b>9.3.2</b> Using the elbow point</a></li>
<li class="chapter" data-level="9.3.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#using-the-technical-noise"><i class="fa fa-check"></i><b>9.3.3</b> Using the technical noise</a></li>
<li class="chapter" data-level="9.3.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#based-on-population-structure"><i class="fa fa-check"></i><b>9.3.4</b> Based on population structure</a></li>
<li class="chapter" data-level="9.3.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#putting-it-together"><i class="fa fa-check"></i><b>9.3.5</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#non-negative-matrix-factorization"><i class="fa fa-check"></i><b>9.4</b> Non-negative matrix factorization</a></li>
<li class="chapter" data-level="9.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>9.5</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="9.5.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#motivation-3"><i class="fa fa-check"></i><b>9.5.1</b> Motivation</a></li>
<li class="chapter" data-level="9.5.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizing-with-pca"><i class="fa fa-check"></i><b>9.5.2</b> Visualizing with PCA</a></li>
<li class="chapter" data-level="9.5.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>9.5.3</b> <span class="math inline">\(t\)</span>-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="9.5.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>9.5.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="9.5.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualization-interpretation"><i class="fa fa-check"></i><b>9.5.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#motivation-4"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#what-is-the-true-clustering"><i class="fa fa-check"></i><b>10.2</b> What is the “true clustering”?</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#clustering-graph"><i class="fa fa-check"></i><b>10.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#background-3"><i class="fa fa-check"></i><b>10.3.1</b> Background</a></li>
<li class="chapter" data-level="10.3.2" data-path="clustering.html"><a href="clustering.html#implementation"><i class="fa fa-check"></i><b>10.3.2</b> Implementation</a></li>
<li class="chapter" data-level="10.3.3" data-path="clustering.html"><a href="clustering.html#other-parameters"><i class="fa fa-check"></i><b>10.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="10.3.4" data-path="clustering.html"><a href="clustering.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>10.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#background-4"><i class="fa fa-check"></i><b>10.4.1</b> Background</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#base-implementation"><i class="fa fa-check"></i><b>10.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="10.4.3" data-path="clustering.html"><a href="clustering.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>10.4.3</b> Assessing cluster separation</a></li>
<li class="chapter" data-level="10.4.4" data-path="clustering.html"><a href="clustering.html#in-two-step-procedures"><i class="fa fa-check"></i><b>10.4.4</b> In two-step procedures</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#background-5"><i class="fa fa-check"></i><b>10.5.1</b> Background</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#implementation-1"><i class="fa fa-check"></i><b>10.5.2</b> Implementation</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>10.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#cluster-bootstrapping"><i class="fa fa-check"></i><b>10.6</b> Evaluating cluster stability</a></li>
<li class="chapter" data-level="10.7" data-path="clustering.html"><a href="clustering.html#subclustering"><i class="fa fa-check"></i><b>10.7</b> Subclustering</a></li>
<li class="chapter" data-level="" data-path="clustering.html"><a href="clustering.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="marker-detection.html"><a href="marker-detection.html"><i class="fa fa-check"></i><b>11</b> Marker gene detection</a><ul>
<li class="chapter" data-level="11.1" data-path="marker-detection.html"><a href="marker-detection.html#motivation-5"><i class="fa fa-check"></i><b>11.1</b> Motivation</a></li>
<li class="chapter" data-level="11.2" data-path="marker-detection.html"><a href="marker-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>11.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="11.2.1" data-path="marker-detection.html"><a href="marker-detection.html#standard-application"><i class="fa fa-check"></i><b>11.2.1</b> Standard application</a></li>
<li class="chapter" data-level="11.2.2" data-path="marker-detection.html"><a href="marker-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>11.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="11.2.3" data-path="marker-detection.html"><a href="marker-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>11.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="marker-detection.html"><a href="marker-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>11.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="11.3.1" data-path="marker-detection.html"><a href="marker-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>11.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="11.3.2" data-path="marker-detection.html"><a href="marker-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>11.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="11.3.3" data-path="marker-detection.html"><a href="marker-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>11.3.3</b> Using custom DE methods</a></li>
<li class="chapter" data-level="11.3.4" data-path="marker-detection.html"><a href="marker-detection.html#combining-multiple-marker-statistics"><i class="fa fa-check"></i><b>11.3.4</b> Combining multiple marker statistics</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="marker-detection.html"><a href="marker-detection.html#marker-batch"><i class="fa fa-check"></i><b>11.4</b> Handling blocking factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="marker-detection.html"><a href="marker-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>11.4.1</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="11.4.2" data-path="marker-detection.html"><a href="marker-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>11.4.2</b> Using the <code>design=</code> argument</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="marker-detection.html"><a href="marker-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>11.5</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="11.5.1" data-path="marker-detection.html"><a href="marker-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>11.5.1</b> From data snooping</a></li>
<li class="chapter" data-level="11.5.2" data-path="marker-detection.html"><a href="marker-detection.html#false-replicates"><i class="fa fa-check"></i><b>11.5.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="marker-detection.html"><a href="marker-detection.html#further-comments"><i class="fa fa-check"></i><b>11.6</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="marker-detection.html"><a href="marker-detection.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html"><i class="fa fa-check"></i><b>12</b> Cell type annotation</a><ul>
<li class="chapter" data-level="12.1" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#motivation-6"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cell-labels-from-reference-data"><i class="fa fa-check"></i><b>12.2</b> Assigning cell labels from reference data</a><ul>
<li class="chapter" data-level="12.2.1" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#overview-3"><i class="fa fa-check"></i><b>12.2.1</b> Overview</a></li>
<li class="chapter" data-level="12.2.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#using-the-in-built-references"><i class="fa fa-check"></i><b>12.2.2</b> Using the in-built references</a></li>
<li class="chapter" data-level="12.2.3" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#using-custom-references"><i class="fa fa-check"></i><b>12.2.3</b> Using custom references</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cell-labels-from-gene-sets"><i class="fa fa-check"></i><b>12.3</b> Assigning cell labels from gene sets</a></li>
<li class="chapter" data-level="12.4" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cluster-labels-from-markers"><i class="fa fa-check"></i><b>12.4</b> Assigning cluster labels from markers</a></li>
<li class="chapter" data-level="12.5" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#computing-gene-set-activities"><i class="fa fa-check"></i><b>12.5</b> Computing gene set activities</a></li>
<li class="chapter" data-level="" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="integrating-datasets.html"><a href="integrating-datasets.html"><i class="fa fa-check"></i><b>13</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="13.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#motivation-7"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#setting-up-the-data"><i class="fa fa-check"></i><b>13.2</b> Setting up the data</a></li>
<li class="chapter" data-level="13.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#batch-diagnosis"><i class="fa fa-check"></i><b>13.3</b> Diagnosing batch effects</a></li>
<li class="chapter" data-level="13.4" data-path="integrating-datasets.html"><a href="integrating-datasets.html#linear-regression"><i class="fa fa-check"></i><b>13.4</b> Linear regression</a></li>
<li class="chapter" data-level="13.5" data-path="integrating-datasets.html"><a href="integrating-datasets.html#performing-mnn-correction"><i class="fa fa-check"></i><b>13.5</b> Performing MNN correction</a><ul>
<li class="chapter" data-level="13.5.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#algorithm-overview"><i class="fa fa-check"></i><b>13.5.1</b> Algorithm overview</a></li>
<li class="chapter" data-level="13.5.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#application-to-the-pbmc-data"><i class="fa fa-check"></i><b>13.5.2</b> Application to the PBMC data</a></li>
<li class="chapter" data-level="13.5.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#correction-diagnostics"><i class="fa fa-check"></i><b>13.5.3</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="13.6" data-path="integrating-datasets.html"><a href="integrating-datasets.html#preserving-biological-heterogeneity"><i class="fa fa-check"></i><b>13.6</b> Preserving biological heterogeneity</a><ul>
<li class="chapter" data-level="13.6.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#comparison-to-within-batch-clusters"><i class="fa fa-check"></i><b>13.6.1</b> Comparison to within-batch clusters</a></li>
<li class="chapter" data-level="13.6.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#integration-with-markers"><i class="fa fa-check"></i><b>13.6.2</b> Encouraging consistency with marker genes</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="integrating-datasets.html"><a href="integrating-datasets.html#using-corrected-values"><i class="fa fa-check"></i><b>13.7</b> Using the corrected values</a></li>
<li class="chapter" data-level="" data-path="integrating-datasets.html"><a href="integrating-datasets.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html"><i class="fa fa-check"></i><b>14</b> Multi-sample comparisons</a><ul>
<li class="chapter" data-level="14.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-8"><i class="fa fa-check"></i><b>14.1</b> Motivation</a></li>
<li class="chapter" data-level="14.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>14.2</b> Setting up the data</a></li>
<li class="chapter" data-level="14.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-expression-between-conditions"><i class="fa fa-check"></i><b>14.3</b> Differential expression between conditions</a><ul>
<li class="chapter" data-level="14.3.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#creating-pseudo-bulk-samples"><i class="fa fa-check"></i><b>14.3.1</b> Creating pseudo-bulk samples</a></li>
<li class="chapter" data-level="14.3.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-de-analysis"><i class="fa fa-check"></i><b>14.3.2</b> Performing the DE analysis</a></li>
<li class="chapter" data-level="14.3.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#putting-it-all-together"><i class="fa fa-check"></i><b>14.3.3</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-abundance"><i class="fa fa-check"></i><b>14.4</b> Differential abundance between conditions</a><ul>
<li class="chapter" data-level="14.4.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#overview-4"><i class="fa fa-check"></i><b>14.4.1</b> Overview</a></li>
<li class="chapter" data-level="14.4.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-da-analysis"><i class="fa fa-check"></i><b>14.4.2</b> Performing the DA analysis</a></li>
<li class="chapter" data-level="14.4.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#composition-effects"><i class="fa fa-check"></i><b>14.4.3</b> Handling composition effects</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#comments-on-interpretation"><i class="fa fa-check"></i><b>14.5</b> Comments on interpretation</a><ul>
<li class="chapter" data-level="14.5.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#de-da-duality"><i class="fa fa-check"></i><b>14.5.1</b> DE or DA? Two sides of the same coin</a></li>
<li class="chapter" data-level="14.5.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#sacrificing-differences"><i class="fa fa-check"></i><b>14.5.2</b> Sacrificing biology by integration</a></li>
</ul></li>
<li class="chapter" data-level="14.6" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#ambient-problems"><i class="fa fa-check"></i><b>14.6</b> Avoiding problems with ambient RNA</a><ul>
<li class="chapter" data-level="14.6.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-9"><i class="fa fa-check"></i><b>14.6.1</b> Motivation</a></li>
<li class="chapter" data-level="14.6.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#discarding-ambient-degs"><i class="fa fa-check"></i><b>14.6.2</b> Discarding ambient DEGs</a></li>
<li class="chapter" data-level="14.6.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#subtracting-ambient-counts"><i class="fa fa-check"></i><b>14.6.3</b> Subtracting ambient counts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="doublet-detection.html"><a href="doublet-detection.html"><i class="fa fa-check"></i><b>15</b> Doublet detection</a><ul>
<li class="chapter" data-level="15.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-5"><i class="fa fa-check"></i><b>15.1</b> Overview</a></li>
<li class="chapter" data-level="15.2" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-with-clusters"><i class="fa fa-check"></i><b>15.2</b> Doublet detection with clusters</a></li>
<li class="chapter" data-level="15.3" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-simulation"><i class="fa fa-check"></i><b>15.3</b> Doublet detection by simulation</a></li>
<li class="chapter" data-level="15.4" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-in-multiplexed-experiments"><i class="fa fa-check"></i><b>15.4</b> Doublet detection in multiplexed experiments</a><ul>
<li class="chapter" data-level="15.4.1" data-path="doublet-detection.html"><a href="doublet-detection.html#background-7"><i class="fa fa-check"></i><b>15.4.1</b> Background</a></li>
<li class="chapter" data-level="15.4.2" data-path="doublet-detection.html"><a href="doublet-detection.html#identifying-inter-sample-doublets"><i class="fa fa-check"></i><b>15.4.2</b> Identifying inter-sample doublets</a></li>
<li class="chapter" data-level="15.4.3" data-path="doublet-detection.html"><a href="doublet-detection.html#guilt-by-association-for-unmarked-doublets"><i class="fa fa-check"></i><b>15.4.3</b> Guilt by association for unmarked doublets</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="doublet-detection.html"><a href="doublet-detection.html#further-comments-1"><i class="fa fa-check"></i><b>15.5</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="doublet-detection.html"><a href="doublet-detection.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html"><i class="fa fa-check"></i><b>16</b> Cell cycle assignment</a><ul>
<li class="chapter" data-level="16.1" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#motivation-10"><i class="fa fa-check"></i><b>16.1</b> Motivation</a></li>
<li class="chapter" data-level="16.2" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-the-cyclins"><i class="fa fa-check"></i><b>16.2</b> Using the cyclins</a></li>
<li class="chapter" data-level="16.3" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-reference-profiles"><i class="fa fa-check"></i><b>16.3</b> Using reference profiles</a></li>
<li class="chapter" data-level="16.4" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-the-cyclone-classifier"><i class="fa fa-check"></i><b>16.4</b> Using the <code>cyclone()</code> classifier</a></li>
<li class="chapter" data-level="16.5" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#regressing-out-cell-cycle-phase"><i class="fa fa-check"></i><b>16.5</b> Regressing out cell cycle phase</a></li>
<li class="chapter" data-level="" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#session-info-11"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html"><i class="fa fa-check"></i><b>17</b> Trajectory Analysis</a><ul>
<li class="chapter" data-level="17.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#overview-6"><i class="fa fa-check"></i><b>17.1</b> Overview</a></li>
<li class="chapter" data-level="17.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#obtaining-pseudo-times"><i class="fa fa-check"></i><b>17.2</b> Obtaining pseudo-times</a><ul>
<li class="chapter" data-level="17.2.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#cluster-based-minimum-spanning-tree"><i class="fa fa-check"></i><b>17.2.1</b> Cluster-based minimum spanning tree</a></li>
<li class="chapter" data-level="17.2.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#principal-curves"><i class="fa fa-check"></i><b>17.2.2</b> Principal curves</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#characterizing-trajectories"><i class="fa fa-check"></i><b>17.3</b> Characterizing trajectories</a><ul>
<li class="chapter" data-level="17.3.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#changes-along-a-trajectory"><i class="fa fa-check"></i><b>17.3.1</b> Changes along a trajectory</a></li>
<li class="chapter" data-level="17.3.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#changes-between-lineages"><i class="fa fa-check"></i><b>17.3.2</b> Changes between lineages</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#finding-the-root"><i class="fa fa-check"></i><b>17.4</b> Finding the root</a><ul>
<li class="chapter" data-level="17.4.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#overview-7"><i class="fa fa-check"></i><b>17.4.1</b> Overview</a></li>
<li class="chapter" data-level="17.4.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#entropy-based-methods"><i class="fa fa-check"></i><b>17.4.2</b> Entropy-based methods</a></li>
<li class="chapter" data-level="17.4.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#rna-velocity"><i class="fa fa-check"></i><b>17.4.3</b> RNA velocity</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html"><i class="fa fa-check"></i><b>18</b> Integrating with protein abundance</a><ul>
<li class="chapter" data-level="18.1" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#motivation-11"><i class="fa fa-check"></i><b>18.1</b> Motivation</a></li>
<li class="chapter" data-level="18.2" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#preprocessing"><i class="fa fa-check"></i><b>18.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="18.2.1" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#setting-up-the-data-2"><i class="fa fa-check"></i><b>18.2.1</b> Setting up the data</a></li>
<li class="chapter" data-level="18.2.2" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#quality-control-1"><i class="fa fa-check"></i><b>18.2.2</b> Quality control</a></li>
<li class="chapter" data-level="18.2.3" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#normalization-1"><i class="fa fa-check"></i><b>18.2.3</b> Normalization</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#clustering-and-interpretation"><i class="fa fa-check"></i><b>18.3</b> Clustering and interpretation</a></li>
<li class="chapter" data-level="18.4" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#integration-with-gene-expression-data"><i class="fa fa-check"></i><b>18.4</b> Integration with gene expression data</a><ul>
<li class="chapter" data-level="18.4.1" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-subclustering"><i class="fa fa-check"></i><b>18.4.1</b> By subclustering</a></li>
<li class="chapter" data-level="18.4.2" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-combined-clustering"><i class="fa fa-check"></i><b>18.4.2</b> By combined clustering</a></li>
<li class="chapter" data-level="18.4.3" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-differential-testing"><i class="fa fa-check"></i><b>18.4.3</b> By differential testing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#session-info-12"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="repertoire-seq.html"><a href="repertoire-seq.html"><i class="fa fa-check"></i><b>19</b> Analyzing repertoire sequencing data</a><ul>
<li class="chapter" data-level="19.1" data-path="repertoire-seq.html"><a href="repertoire-seq.html#motivation-12"><i class="fa fa-check"></i><b>19.1</b> Motivation</a></li>
<li class="chapter" data-level="19.2" data-path="repertoire-seq.html"><a href="repertoire-seq.html#analyzing-the-t-cell-receptor-repertoire"><i class="fa fa-check"></i><b>19.2</b> Analyzing the T cell receptor repertoire</a><ul>
<li class="chapter" data-level="19.2.1" data-path="repertoire-seq.html"><a href="repertoire-seq.html#data-processing"><i class="fa fa-check"></i><b>19.2.1</b> Data processing</a></li>
<li class="chapter" data-level="19.2.2" data-path="repertoire-seq.html"><a href="repertoire-seq.html#basic-diagnostics"><i class="fa fa-check"></i><b>19.2.2</b> Basic diagnostics</a></li>
<li class="chapter" data-level="19.2.3" data-path="repertoire-seq.html"><a href="repertoire-seq.html#complex-diagnostics"><i class="fa fa-check"></i><b>19.2.3</b> Complex diagnostics</a></li>
<li class="chapter" data-level="19.2.4" data-path="repertoire-seq.html"><a href="repertoire-seq.html#quantifying-clonal-expansion"><i class="fa fa-check"></i><b>19.2.4</b> Quantifying clonal expansion</a></li>
<li class="chapter" data-level="19.2.5" data-path="repertoire-seq.html"><a href="repertoire-seq.html#quantifying-gene-expression-and-properties"><i class="fa fa-check"></i><b>19.2.5</b> Quantifying gene expression and properties</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="repertoire-seq.html"><a href="repertoire-seq.html#analyzing-the-immunoglobulin-repertoire"><i class="fa fa-check"></i><b>19.3</b> Analyzing the immunoglobulin repertoire</a></li>
<li class="chapter" data-level="19.4" data-path="repertoire-seq.html"><a href="repertoire-seq.html#multi-sample-analyses"><i class="fa fa-check"></i><b>19.4</b> Multi-sample analyses</a></li>
<li class="chapter" data-level="" data-path="repertoire-seq.html"><a href="repertoire-seq.html#session-info-13"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="interactive-sharing.html"><a href="interactive-sharing.html"><i class="fa fa-check"></i><b>20</b> Interactive data exploration</a><ul>
<li class="chapter" data-level="20.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#motivation-13"><i class="fa fa-check"></i><b>20.1</b> Motivation</a></li>
<li class="chapter" data-level="20.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#interactive-quickstart"><i class="fa fa-check"></i><b>20.2</b> Quick start</a></li>
<li class="chapter" data-level="20.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#isee-examples"><i class="fa fa-check"></i><b>20.3</b> Usage examples</a><ul>
<li class="chapter" data-level="20.3.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#quality-control-2"><i class="fa fa-check"></i><b>20.3.1</b> Quality control</a></li>
<li class="chapter" data-level="20.3.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#annotation-of-cell-populations"><i class="fa fa-check"></i><b>20.3.2</b> Annotation of cell populations</a></li>
<li class="chapter" data-level="20.3.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#querying-features-of-interest"><i class="fa fa-check"></i><b>20.3.3</b> Querying features of interest</a></li>
</ul></li>
<li class="chapter" data-level="20.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#reproducible-visualizations"><i class="fa fa-check"></i><b>20.4</b> Reproducible visualizations</a></li>
<li class="chapter" data-level="20.5" data-path="interactive-sharing.html"><a href="interactive-sharing.html#dissemination"><i class="fa fa-check"></i><b>20.5</b> Dissemination of analysis results</a></li>
<li class="chapter" data-level="20.6" data-path="interactive-sharing.html"><a href="interactive-sharing.html#additional-resources"><i class="fa fa-check"></i><b>20.6</b> Additional resources</a></li>
<li class="chapter" data-level="" data-path="interactive-sharing.html"><a href="interactive-sharing.html#session-info-14"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html"><i class="fa fa-check"></i><b>21</b> Dealing with big data</a><ul>
<li class="chapter" data-level="21.1" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#motivation-14"><i class="fa fa-check"></i><b>21.1</b> Motivation</a></li>
<li class="chapter" data-level="21.2" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#fast-approximations"><i class="fa fa-check"></i><b>21.2</b> Fast approximations</a><ul>
<li class="chapter" data-level="21.2.1" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#nearest-neighbor-searching"><i class="fa fa-check"></i><b>21.2.1</b> Nearest neighbor searching</a></li>
<li class="chapter" data-level="21.2.2" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#big-data-svd"><i class="fa fa-check"></i><b>21.2.2</b> Singular value decomposition</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#parallelization"><i class="fa fa-check"></i><b>21.3</b> Parallelization</a></li>
<li class="chapter" data-level="21.4" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#out-of-memory-representations"><i class="fa fa-check"></i><b>21.4</b> Out of memory representations</a></li>
<li class="chapter" data-level="" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#session-info-15"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="interoperability.html"><a href="interoperability.html"><i class="fa fa-check"></i><b>22</b> Interoperability</a><ul>
<li class="chapter" data-level="22.1" data-path="interoperability.html"><a href="interoperability.html#motivation-15"><i class="fa fa-check"></i><b>22.1</b> Motivation</a></li>
<li class="chapter" data-level="22.2" data-path="interoperability.html"><a href="interoperability.html#interchanging-with-seurat"><i class="fa fa-check"></i><b>22.2</b> Interchanging with <em>Seurat</em></a></li>
<li class="chapter" data-level="22.3" data-path="interoperability.html"><a href="interoperability.html#interchanging-with-scanpy"><i class="fa fa-check"></i><b>22.3</b> Interchanging with <em>scanpy</em></a></li>
<li class="chapter" data-level="" data-path="interoperability.html"><a href="interoperability.html#session-info-16"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="23" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html"><i class="fa fa-check"></i><b>23</b> Lun 416B cell line (Smart-seq2)</a><ul>
<li class="chapter" data-level="23.1" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#introduction-3"><i class="fa fa-check"></i><b>23.1</b> Introduction</a></li>
<li class="chapter" data-level="23.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>23.2</b> Data loading</a></li>
<li class="chapter" data-level="23.3" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#quality-control-3"><i class="fa fa-check"></i><b>23.3</b> Quality control</a></li>
<li class="chapter" data-level="23.4" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#normalization-2"><i class="fa fa-check"></i><b>23.4</b> Normalization</a></li>
<li class="chapter" data-level="23.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>23.5</b> Variance modelling</a></li>
<li class="chapter" data-level="23.6" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#batch-correction"><i class="fa fa-check"></i><b>23.6</b> Batch correction</a></li>
<li class="chapter" data-level="23.7" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>23.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="23.8" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#clustering-1"><i class="fa fa-check"></i><b>23.8</b> Clustering</a></li>
<li class="chapter" data-level="23.9" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#interpretation"><i class="fa fa-check"></i><b>23.9</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#session-info-17"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html"><i class="fa fa-check"></i><b>24</b> Zeisel mouse brain (STRT-Seq)</a><ul>
<li class="chapter" data-level="24.1" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#introduction-4"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#data-loading-1"><i class="fa fa-check"></i><b>24.2</b> Data loading</a></li>
<li class="chapter" data-level="24.3" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#quality-control-4"><i class="fa fa-check"></i><b>24.3</b> Quality control</a></li>
<li class="chapter" data-level="24.4" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#normalization-3"><i class="fa fa-check"></i><b>24.4</b> Normalization</a></li>
<li class="chapter" data-level="24.5" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#variance-modelling-1"><i class="fa fa-check"></i><b>24.5</b> Variance modelling</a></li>
<li class="chapter" data-level="24.6" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>24.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="24.7" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#clustering-2"><i class="fa fa-check"></i><b>24.7</b> Clustering</a></li>
<li class="chapter" data-level="24.8" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#interpretation-1"><i class="fa fa-check"></i><b>24.8</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html#session-info-18"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html"><i class="fa fa-check"></i><b>25</b> Unfiltered human PBMCs (10X Genomics)</a><ul>
<li class="chapter" data-level="25.1" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#introduction-5"><i class="fa fa-check"></i><b>25.1</b> Introduction</a></li>
<li class="chapter" data-level="25.2" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#data-loading-2"><i class="fa fa-check"></i><b>25.2</b> Data loading</a></li>
<li class="chapter" data-level="25.3" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#quality-control-5"><i class="fa fa-check"></i><b>25.3</b> Quality control</a></li>
<li class="chapter" data-level="25.4" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#normalization-4"><i class="fa fa-check"></i><b>25.4</b> Normalization</a></li>
<li class="chapter" data-level="25.5" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#variance-modelling-2"><i class="fa fa-check"></i><b>25.5</b> Variance modelling</a></li>
<li class="chapter" data-level="25.6" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>25.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="25.7" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#clustering-3"><i class="fa fa-check"></i><b>25.7</b> Clustering</a></li>
<li class="chapter" data-level="25.8" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#interpretation-2"><i class="fa fa-check"></i><b>25.8</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html#session-info-19"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html"><i class="fa fa-check"></i><b>26</b> Filtered human PBMCs (10X Genomics)</a><ul>
<li class="chapter" data-level="26.1" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#introduction-6"><i class="fa fa-check"></i><b>26.1</b> Introduction</a></li>
<li class="chapter" data-level="26.2" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-loading-3"><i class="fa fa-check"></i><b>26.2</b> Data loading</a></li>
<li class="chapter" data-level="26.3" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#quality-control-6"><i class="fa fa-check"></i><b>26.3</b> Quality control</a></li>
<li class="chapter" data-level="26.4" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#normalization-5"><i class="fa fa-check"></i><b>26.4</b> Normalization</a></li>
<li class="chapter" data-level="26.5" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#variance-modelling-3"><i class="fa fa-check"></i><b>26.5</b> Variance modelling</a></li>
<li class="chapter" data-level="26.6" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#dimensionality-reduction-4"><i class="fa fa-check"></i><b>26.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="26.7" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#clustering-4"><i class="fa fa-check"></i><b>26.7</b> Clustering</a></li>
<li class="chapter" data-level="26.8" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>26.8</b> Data integration</a></li>
<li class="chapter" data-level="" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#session-info-20"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html"><i class="fa fa-check"></i><b>27</b> Human PBMC 10X dataset (surface proteins)</a><ul>
<li class="chapter" data-level="27.1" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#introduction-7"><i class="fa fa-check"></i><b>27.1</b> Introduction</a></li>
<li class="chapter" data-level="27.2" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#analysis-code"><i class="fa fa-check"></i><b>27.2</b> Analysis code</a><ul>
<li class="chapter" data-level="27.2.1" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#data-loading-4"><i class="fa fa-check"></i><b>27.2.1</b> Data loading</a></li>
<li class="chapter" data-level="27.2.2" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#quality-control-7"><i class="fa fa-check"></i><b>27.2.2</b> Quality control</a></li>
<li class="chapter" data-level="27.2.3" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#normalization-6"><i class="fa fa-check"></i><b>27.2.3</b> Normalization</a></li>
<li class="chapter" data-level="27.2.4" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#dimensionality-reduction-5"><i class="fa fa-check"></i><b>27.2.4</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="27.2.5" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#clustering-5"><i class="fa fa-check"></i><b>27.2.5</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="human-pbmc-10x-dataset-surface-proteins.html"><a href="human-pbmc-10x-dataset-surface-proteins.html#session-info-21"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html"><i class="fa fa-check"></i><b>28</b> Grun human pancreas (CEL-seq2)</a><ul>
<li class="chapter" data-level="28.1" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#introduction-8"><i class="fa fa-check"></i><b>28.1</b> Introduction</a></li>
<li class="chapter" data-level="28.2" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#data-loading-5"><i class="fa fa-check"></i><b>28.2</b> Data loading</a></li>
<li class="chapter" data-level="28.3" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#quality-control-8"><i class="fa fa-check"></i><b>28.3</b> Quality control</a></li>
<li class="chapter" data-level="28.4" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#normalization-7"><i class="fa fa-check"></i><b>28.4</b> Normalization</a></li>
<li class="chapter" data-level="28.5" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#variance-modelling-4"><i class="fa fa-check"></i><b>28.5</b> Variance modelling</a></li>
<li class="chapter" data-level="28.6" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#data-integration-1"><i class="fa fa-check"></i><b>28.6</b> Data integration</a></li>
<li class="chapter" data-level="28.7" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#dimensionality-reduction-6"><i class="fa fa-check"></i><b>28.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="28.8" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#clustering-6"><i class="fa fa-check"></i><b>28.8</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html#session-info-22"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html"><i class="fa fa-check"></i><b>29</b> Muraro human pancreas (CEL-seq)</a><ul>
<li class="chapter" data-level="29.1" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#introduction-9"><i class="fa fa-check"></i><b>29.1</b> Introduction</a></li>
<li class="chapter" data-level="29.2" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#data-loading-6"><i class="fa fa-check"></i><b>29.2</b> Data loading</a></li>
<li class="chapter" data-level="29.3" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#quality-control-9"><i class="fa fa-check"></i><b>29.3</b> Quality control</a></li>
<li class="chapter" data-level="29.4" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#normalization-8"><i class="fa fa-check"></i><b>29.4</b> Normalization</a><ul>
<li class="chapter" data-level="29.4.1" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#variance-modelling-5"><i class="fa fa-check"></i><b>29.4.1</b> Variance modelling</a></li>
</ul></li>
<li class="chapter" data-level="29.5" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#data-integration-2"><i class="fa fa-check"></i><b>29.5</b> Data integration</a></li>
<li class="chapter" data-level="29.6" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#dimensionality-reduction-7"><i class="fa fa-check"></i><b>29.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="29.7" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#clustering-7"><i class="fa fa-check"></i><b>29.7</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html#session-info-23"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html"><i class="fa fa-check"></i><b>30</b> Lawlor human pancreas (SMARTer)</a><ul>
<li class="chapter" data-level="30.1" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#introduction-10"><i class="fa fa-check"></i><b>30.1</b> Introduction</a></li>
<li class="chapter" data-level="30.2" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#data-loading-7"><i class="fa fa-check"></i><b>30.2</b> Data loading</a></li>
<li class="chapter" data-level="30.3" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#quality-control-10"><i class="fa fa-check"></i><b>30.3</b> Quality control</a></li>
<li class="chapter" data-level="30.4" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#normalization-9"><i class="fa fa-check"></i><b>30.4</b> Normalization</a></li>
<li class="chapter" data-level="30.5" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#variance-modelling-6"><i class="fa fa-check"></i><b>30.5</b> Variance modelling</a></li>
<li class="chapter" data-level="30.6" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#dimensionality-reduction-8"><i class="fa fa-check"></i><b>30.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="30.7" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#clustering-8"><i class="fa fa-check"></i><b>30.7</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html#session-info-24"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html"><i class="fa fa-check"></i><b>31</b> Segerstolpe human pancreas (Smart-seq2)</a><ul>
<li class="chapter" data-level="31.1" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#introduction-11"><i class="fa fa-check"></i><b>31.1</b> Introduction</a></li>
<li class="chapter" data-level="31.2" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#data-loading-8"><i class="fa fa-check"></i><b>31.2</b> Data loading</a></li>
<li class="chapter" data-level="31.3" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#quality-control-11"><i class="fa fa-check"></i><b>31.3</b> Quality control</a><ul>
<li class="chapter" data-level="31.3.1" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#normalization-10"><i class="fa fa-check"></i><b>31.3.1</b> Normalization</a></li>
<li class="chapter" data-level="31.3.2" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#variance-modelling-7"><i class="fa fa-check"></i><b>31.3.2</b> Variance modelling</a></li>
<li class="chapter" data-level="31.3.3" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#dimensionality-reduction-9"><i class="fa fa-check"></i><b>31.3.3</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="31.3.4" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#clustering-9"><i class="fa fa-check"></i><b>31.3.4</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#session-info-25"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="merged-pancreas.html"><a href="merged-pancreas.html"><i class="fa fa-check"></i><b>32</b> Merged human pancreas datasets</a><ul>
<li class="chapter" data-level="32.1" data-path="merged-pancreas.html"><a href="merged-pancreas.html#introduction-12"><i class="fa fa-check"></i><b>32.1</b> Introduction</a></li>
<li class="chapter" data-level="32.2" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-good"><i class="fa fa-check"></i><b>32.2</b> The good</a></li>
<li class="chapter" data-level="32.3" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-bad"><i class="fa fa-check"></i><b>32.3</b> The bad</a></li>
<li class="chapter" data-level="32.4" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-ugly"><i class="fa fa-check"></i><b>32.4</b> The ugly</a></li>
<li class="chapter" data-level="" data-path="merged-pancreas.html"><a href="merged-pancreas.html#session-info-26"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="merged-hcsc.html"><a href="merged-hcsc.html"><i class="fa fa-check"></i><b>33</b> Grun mouse HSC (CEL-seq)</a><ul>
<li class="chapter" data-level="33.1" data-path="merged-hcsc.html"><a href="merged-hcsc.html#introduction-13"><i class="fa fa-check"></i><b>33.1</b> Introduction</a></li>
<li class="chapter" data-level="33.2" data-path="merged-hcsc.html"><a href="merged-hcsc.html#data-loading-9"><i class="fa fa-check"></i><b>33.2</b> Data loading</a></li>
<li class="chapter" data-level="33.3" data-path="merged-hcsc.html"><a href="merged-hcsc.html#quality-control-12"><i class="fa fa-check"></i><b>33.3</b> Quality control</a></li>
<li class="chapter" data-level="33.4" data-path="merged-hcsc.html"><a href="merged-hcsc.html#normalization-11"><i class="fa fa-check"></i><b>33.4</b> Normalization</a></li>
<li class="chapter" data-level="33.5" data-path="merged-hcsc.html"><a href="merged-hcsc.html#variance-modelling-8"><i class="fa fa-check"></i><b>33.5</b> Variance modelling</a></li>
<li class="chapter" data-level="33.6" data-path="merged-hcsc.html"><a href="merged-hcsc.html#dimensionality-reduction-10"><i class="fa fa-check"></i><b>33.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="33.7" data-path="merged-hcsc.html"><a href="merged-hcsc.html#clustering-10"><i class="fa fa-check"></i><b>33.7</b> Clustering</a></li>
<li class="chapter" data-level="33.8" data-path="merged-hcsc.html"><a href="merged-hcsc.html#marker-gene-detection"><i class="fa fa-check"></i><b>33.8</b> Marker gene detection</a></li>
<li class="chapter" data-level="" data-path="merged-hcsc.html"><a href="merged-hcsc.html#session-info-27"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html"><i class="fa fa-check"></i><b>34</b> Nestorowa mouse HSC (Smart-seq2)</a><ul>
<li class="chapter" data-level="34.1" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#introduction-14"><i class="fa fa-check"></i><b>34.1</b> Introduction</a></li>
<li class="chapter" data-level="34.2" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#data-loading-10"><i class="fa fa-check"></i><b>34.2</b> Data loading</a></li>
<li class="chapter" data-level="34.3" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#quality-control-13"><i class="fa fa-check"></i><b>34.3</b> Quality control</a></li>
<li class="chapter" data-level="34.4" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#normalization-12"><i class="fa fa-check"></i><b>34.4</b> Normalization</a></li>
<li class="chapter" data-level="34.5" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#variance-modelling-9"><i class="fa fa-check"></i><b>34.5</b> Variance modelling</a></li>
<li class="chapter" data-level="34.6" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#dimensionality-reduction-11"><i class="fa fa-check"></i><b>34.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="34.7" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#clustering-11"><i class="fa fa-check"></i><b>34.7</b> Clustering</a></li>
<li class="chapter" data-level="34.8" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#marker-gene-detection-1"><i class="fa fa-check"></i><b>34.8</b> Marker gene detection</a></li>
<li class="chapter" data-level="34.9" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#cell-type-annotation-1"><i class="fa fa-check"></i><b>34.9</b> Cell type annotation</a></li>
<li class="chapter" data-level="34.10" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#miscellaneous-analyses"><i class="fa fa-check"></i><b>34.10</b> Miscellaneous analyses</a></li>
<li class="chapter" data-level="" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#session-info-28"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><i class="fa fa-check"></i><b>35</b> Pijuan-Sala chimeric mouse embryo (10X Genomics)</a><ul>
<li class="chapter" data-level="35.1" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#introduction-15"><i class="fa fa-check"></i><b>35.1</b> Introduction</a></li>
<li class="chapter" data-level="35.2" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#data-loading-11"><i class="fa fa-check"></i><b>35.2</b> Data loading</a></li>
<li class="chapter" data-level="35.3" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#quality-control-14"><i class="fa fa-check"></i><b>35.3</b> Quality control</a></li>
<li class="chapter" data-level="35.4" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#normalization-13"><i class="fa fa-check"></i><b>35.4</b> Normalization</a></li>
<li class="chapter" data-level="35.5" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#variance-modelling-10"><i class="fa fa-check"></i><b>35.5</b> Variance modelling</a></li>
<li class="chapter" data-level="35.6" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#merging"><i class="fa fa-check"></i><b>35.6</b> Merging</a></li>
<li class="chapter" data-level="35.7" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#clustering-12"><i class="fa fa-check"></i><b>35.7</b> Clustering</a></li>
<li class="chapter" data-level="35.8" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#dimensionality-reduction-12"><i class="fa fa-check"></i><b>35.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#session-info-29"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html"><i class="fa fa-check"></i><b>36</b> Bach mouse mammary gland (10X Genomics)</a><ul>
<li class="chapter" data-level="36.1" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#introduction-16"><i class="fa fa-check"></i><b>36.1</b> Introduction</a></li>
<li class="chapter" data-level="36.2" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#data-loading-12"><i class="fa fa-check"></i><b>36.2</b> Data loading</a></li>
<li class="chapter" data-level="36.3" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#quality-control-15"><i class="fa fa-check"></i><b>36.3</b> Quality control</a></li>
<li class="chapter" data-level="36.4" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#normalization-14"><i class="fa fa-check"></i><b>36.4</b> Normalization</a></li>
<li class="chapter" data-level="36.5" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#variance-modelling-11"><i class="fa fa-check"></i><b>36.5</b> Variance modelling</a></li>
<li class="chapter" data-level="36.6" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#dimensionality-reduction-13"><i class="fa fa-check"></i><b>36.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="36.7" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#clustering-13"><i class="fa fa-check"></i><b>36.7</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html#session-info-30"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html"><i class="fa fa-check"></i><b>37</b> HCA human bone marrow (10X Genomics)</a><ul>
<li class="chapter" data-level="37.1" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#introduction-17"><i class="fa fa-check"></i><b>37.1</b> Introduction</a></li>
<li class="chapter" data-level="37.2" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#data-loading-13"><i class="fa fa-check"></i><b>37.2</b> Data loading</a></li>
<li class="chapter" data-level="37.3" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#quality-control-16"><i class="fa fa-check"></i><b>37.3</b> Quality control</a></li>
<li class="chapter" data-level="37.4" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#normalization-15"><i class="fa fa-check"></i><b>37.4</b> Normalization</a></li>
<li class="chapter" data-level="37.5" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#variance-modeling"><i class="fa fa-check"></i><b>37.5</b> Variance modeling</a></li>
<li class="chapter" data-level="37.6" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#data-integration-3"><i class="fa fa-check"></i><b>37.6</b> Data integration</a></li>
<li class="chapter" data-level="37.7" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#dimensionality-reduction-14"><i class="fa fa-check"></i><b>37.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="37.8" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#clustering-14"><i class="fa fa-check"></i><b>37.8</b> Clustering</a></li>
<li class="chapter" data-level="37.9" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#differential-expression"><i class="fa fa-check"></i><b>37.9</b> Differential expression</a></li>
<li class="chapter" data-level="37.10" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#cell-type-classification"><i class="fa fa-check"></i><b>37.10</b> Cell type classification</a></li>
<li class="chapter" data-level="" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#session-info-31"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="38" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>38</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun-phd"><em>Aaron Lun, PhD</em></a></li>
<li><a href="contributors.html#robert-amezquita-phd"><em>Robert Amezquita, PhD</em></a></li>
<li><a href="contributors.html#stephanie-hicks-phd"><em>Stephanie Hicks, PhD</em></a></li>
<li><a href="contributors.html#raphael-gottardo-phd"><em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="" data-path="contributors.html"><a href="contributors.html#other-notable-entities"><i class="fa fa-check"></i>Other notable entities</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>39</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="integrating-datasets" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> Integrating Datasets</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-7" class="section level2">
<h2><span class="header-section-number">13.1</span> Motivation</h2>
<p>Large single-cell RNA sequencing (scRNA-seq) projects usually need to generate data across multiple batches due to logistical constraints. However, the processing of different batches is often subject to uncontrollable differences, e.g., changes in operator, differences in reagent quality. This results in systematic differences in the observed expression in cells from different batches, which we refer to as “batch effects”. Batch effects are problematic as they can be major drivers of heterogeneity in the data, masking the relevant biological differences and complicating interpretation of the results.</p>
<p>Computational correction of these effects is critical for eliminating batch-to-batch variation, allowing data across multiple batches to be combined for common downstream analysis. However, existing methods based on linear models <span class="citation">(Ritchie et al. <a href="#ref-ritchie2015limma">2015</a>; Leek et al. <a href="#ref-leek2012sva">2012</a>)</span> assume that the composition of cell populations are either known or the same across batches. To overcome these limitations, bespoke methods have been developed for batch correction of single-cell data <span class="citation">(Haghverdi et al. <a href="#ref-haghverdi2018batch">2018</a>; Butler et al. <a href="#ref-butler2018integrating">2018</a>; Lin et al. <a href="#ref-lin2019scmerge">2019</a>)</span> that do not require <em>a priori</em> knowledge about the composition of the population. This allows them to be used in workflows for exploratory analyses of scRNA-seq data where such knowledge is usually unavailable.</p>
</div>
<div id="setting-up-the-data" class="section level2">
<h2><span class="header-section-number">13.2</span> Setting up the data</h2>
<p>To demonstrate, we will use two separate 10X Genomics PBMC datasets generated in two different batches. Each dataset was obtained from the <em><a href="https://bioconductor.org/packages/3.11/TENxPBMCData">TENxPBMCData</a></em> package and separately subjected to basic processing steps. Separate processing prior to the batch correction step is more convenient, scalable and (on occasion) more reliable. For example, outlier-based QC on the cells is more effective when performed within a batch (Section <a href="quality-control.html#qc-batch">6.3.2.3</a>). The same can also be said for trend fitting when modelling the mean-variance relationship (Section <a href="feature-selection.html#variance-batch">8.2.4.1</a>).</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#--- loading ---#</span>
<span class="kw">library</span>(TENxPBMCData)
all.sce &lt;-<span class="st"> </span><span class="kw">list</span>(
    <span class="dt">pbmc3k=</span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc3k&#39;</span>),
    <span class="dt">pbmc4k=</span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>),
    <span class="dt">pbmc8k=</span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc8k&#39;</span>)
)

<span class="co">#--- quality-control ---#</span>
<span class="kw">library</span>(scater)
stats &lt;-<span class="st"> </span>high.mito &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span> (n <span class="cf">in</span> <span class="kw">names</span>(all.sce)) {
    current &lt;-<span class="st"> </span>all.sce[[n]]
    is.mito &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;MT&quot;</span>, <span class="kw">rowData</span>(current)<span class="op">$</span>Symbol_TENx)
    stats[[n]] &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(current, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))
    high.mito[[n]] &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(stats[[n]]<span class="op">$</span>subsets_Mito_percent, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)
    all.sce[[n]] &lt;-<span class="st"> </span>current[,<span class="op">!</span>high.mito[[n]]]
}

<span class="co">#--- normalization ---#</span>
all.sce &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.sce, logNormCounts)

<span class="co">#--- variance-modelling ---#</span>
<span class="kw">library</span>(scran)
all.dec &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.sce, modelGeneVar)
all.hvgs &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.dec, getTopHVGs, <span class="dt">prop=</span><span class="fl">0.1</span>)

<span class="co">#--- dimensionality-reduction ---#</span>
<span class="kw">library</span>(BiocSingular)
<span class="kw">set.seed</span>(<span class="dv">10000</span>)
all.sce &lt;-<span class="st"> </span><span class="kw">mapply</span>(<span class="dt">FUN=</span>runPCA, <span class="dt">x=</span>all.sce, <span class="dt">subset_row=</span>all.hvgs, 
    <span class="dt">MoreArgs=</span><span class="kw">list</span>(<span class="dt">ncomponents=</span><span class="dv">25</span>, <span class="dt">BSPARAM=</span><span class="kw">RandomParam</span>()), 
    <span class="dt">SIMPLIFY=</span><span class="ot">FALSE</span>)

<span class="kw">set.seed</span>(<span class="dv">100000</span>)
all.sce &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.sce, runTSNE, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)

<span class="kw">set.seed</span>(<span class="dv">1000000</span>)
all.sce &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.sce, runUMAP, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)

<span class="co">#--- clustering ---#</span>
<span class="cf">for</span> (n <span class="cf">in</span> <span class="kw">names</span>(all.sce)) {
    g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(all.sce[[n]], <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred=</span><span class="st">&#39;PCA&#39;</span>)
    clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership
    <span class="kw">colLabels</span>(all.sce[[n]])  &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)
}</code></pre></div>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc3k &lt;-<span class="st"> </span>all.sce<span class="op">$</span>pbmc3k
dec3k &lt;-<span class="st"> </span>all.dec<span class="op">$</span>pbmc3k
pbmc3k</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 32738 2609 
## metadata(0):
## assays(2): counts logcounts
## rownames(32738): ENSG00000243485 ENSG00000237613 ... ENSG00000215616
##   ENSG00000215611
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(13): Sample Barcode ... sizeFactor label
## reducedDimNames(3): PCA TSNE UMAP
## altExpNames(0):</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc4k &lt;-<span class="st"> </span>all.sce<span class="op">$</span>pbmc4k
dec4k &lt;-<span class="st"> </span>all.dec<span class="op">$</span>pbmc4k
pbmc4k</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 4182 
## metadata(0):
## assays(2): counts logcounts
## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
##   ENSG00000268674
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(13): Sample Barcode ... sizeFactor label
## reducedDimNames(3): PCA TSNE UMAP
## altExpNames(0):</code></pre>
<p>To prepare for the batch correction:</p>
<ol style="list-style-type: decimal">
<li><p>We subset all batches to the common “universe” of features. In this case, it is straightforward as both batches use Ensembl gene annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">universe &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(pbmc3k), <span class="kw">rownames</span>(pbmc4k))
<span class="kw">length</span>(universe)</code></pre></div>
<pre><code>## [1] 31232</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Subsetting the SingleCellExperiment object.</span>
pbmc3k &lt;-<span class="st"> </span>pbmc3k[universe,]
pbmc4k &lt;-<span class="st"> </span>pbmc4k[universe,]

<span class="co"># Also subsetting the variance modelling results, for convenience.</span>
dec3k &lt;-<span class="st"> </span>dec3k[universe,]
dec4k &lt;-<span class="st"> </span>dec4k[universe,]</code></pre></div></li>
<li><p>We rescale each batch to adjust for differences in sequencing depth between batches. The <code>multiBatchNorm()</code> function recomputes log-normalized expression values after adjusting the size factors for systematic differences in coverage between <code>SingleCellExperiment</code> objects. (Size factors only remove biases between cells <em>within</em> a single batch.) This improves the quality of the correction by removing one aspect of the technical differences between batches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(batchelor)
rescaled &lt;-<span class="st"> </span><span class="kw">multiBatchNorm</span>(pbmc3k, pbmc4k)
pbmc3k &lt;-<span class="st"> </span>rescaled[[<span class="dv">1</span>]]
pbmc4k &lt;-<span class="st"> </span>rescaled[[<span class="dv">2</span>]]</code></pre></div></li>
<li><p>We perform feature selection by averaging the variance components across all batches with the <code>combineVar()</code> function. We compute the average as it is responsive to batch-specific HVGs while still preserving the within-batch ranking of genes. This allows us to use the same strategies described in Section <a href="feature-selection.html#hvg-selection">8.3</a> to select genes of interest. In contrast, approaches based on taking the intersection or union of HVGs across batches become increasingly conservative or liberal, respectively, with an increasing number of batches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
combined.dec &lt;-<span class="st"> </span><span class="kw">combineVar</span>(dec3k, dec4k)
chosen.hvgs &lt;-<span class="st"> </span>combined.dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>
<span class="kw">sum</span>(chosen.hvgs)</code></pre></div>
<pre><code>## [1] 13431</code></pre>
<p>When integrating datasets of variable composition, it is generally safer to err on the side of including more genes than are used in a single dataset analysis, to ensure that markers are retained for any dataset-specific subpopulations that might be present. For a top <span class="math inline">\(X\)</span> selection, this means using a larger <span class="math inline">\(X\)</span> (say, ~5000), or in this case, we simply take all genes above the trend. That said, many of the signal-to-noise considerations described in Section <a href="feature-selection.html#hvg-selection">8.3</a> still apply here, so some experimentation may be necessary for best results.</p>
<p>Alternatively, a more forceful approach to feature selection can be used based on marker genes from within-batch comparisons; this is discussed in more detail in Section <a href="integrating-datasets.html#integration-with-markers">13.6.2</a>.</p></li>
</ol>
</div>
<div id="batch-diagnosis" class="section level2">
<h2><span class="header-section-number">13.3</span> Diagnosing batch effects</h2>
<p>Before we actually perform any correction, it is worth examining whether there is any batch effect in this dataset. We combine the two <code>SingleCellExperiment</code>s and perform a PCA on the log-expression values for all genes with positive (average) biological components.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Synchronizing the metadata for cbind()ing.</span>
<span class="kw">rowData</span>(pbmc3k) &lt;-<span class="st"> </span><span class="kw">rowData</span>(pbmc4k)
pbmc3k<span class="op">$</span>batch &lt;-<span class="st"> &quot;3k&quot;</span>
pbmc4k<span class="op">$</span>batch &lt;-<span class="st"> &quot;4k&quot;</span>
uncorrected &lt;-<span class="st"> </span><span class="kw">cbind</span>(pbmc3k, pbmc4k)

<span class="co"># Using RandomParam() as it is more efficient for file-backed matrices.</span>
<span class="kw">library</span>(scater)
<span class="kw">set.seed</span>(<span class="dv">0010101010</span>)
uncorrected &lt;-<span class="st"> </span><span class="kw">runPCA</span>(uncorrected, <span class="dt">subset_row=</span>chosen.hvgs,
    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>())</code></pre></div>
<p>We use graph-based clustering on the components to obtain a summary of the population structure. As our two PBMC populations should be replicates, each cluster should ideally consist of cells from both batches. However, we instead see clusters that are comprised of cells from a single batch. This indicates that cells of the same type are artificially separated due to technical differences between batches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(uncorrected, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)
clusters &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership
tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters, <span class="dt">Batch=</span>uncorrected<span class="op">$</span>batch)
tab</code></pre></div>
<pre><code>##        Batch
## Cluster   3k   4k
##      1     1  781
##      2     0 1309
##      3     0  535
##      4    14   51
##      5     0  605
##      6   489    0
##      7     0  184
##      8  1272    0
##      9     0  414
##      10  151    0
##      11    0   50
##      12  155    0
##      13    0   65
##      14    0   61
##      15    0   88
##      16   30    0
##      17  339    0
##      18  145    0
##      19   11    3
##      20    2   36</code></pre>
<p>We can also visualize the corrected coordinates using a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="integrating-datasets.html#fig:tsne-pbmc-uncorrected">13.1</a>). The strong separation between cells from different batches is consistent with the clustering results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1111001</span>)
uncorrected &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(uncorrected, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)
<span class="kw">plotTSNE</span>(uncorrected, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-uncorrected"></span>
<img src="P2_W09.data-integration_files/figure-html/tsne-pbmc-uncorrected-1.png" alt="$t$-SNE plot of the PBMC datasets without any batch correction. Each point is a cell that is colored according to its batch of origin." width="672" />
<p class="caption">
Figure 13.1: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets without any batch correction. Each point is a cell that is colored according to its batch of origin.
</p>
</div>
<p>Of course, the other explanation for batch-specific clusters is that there are cell types that are unique to each batch. The degree of intermingling of cells from different batches is not an effective diagnostic when the batches involved might actually contain unique cell subpopulations (which is not a consideration in the PBMC dataset, but the same cannot be said in general). If a cluster only contains cells from a single batch, one can always debate whether that is caused by a failure of the correction method or if there is truly a batch-specific subpopulation. For example, do batch-specific metabolic or differentiation states represent distinct subpopulations? Or should they be merged together? We will not attempt to answer this here, only noting that each batch correction algorithm will make different (and possibly inappropriate) decisions on what constitutes “shared” and “unique” populations.</p>
</div>
<div id="linear-regression" class="section level2">
<h2><span class="header-section-number">13.4</span> Linear regression</h2>
<p>Batch effects in bulk RNA sequencing studies are commonly removed with linear regression. This involves fitting a linear model to each gene’s expression profile, setting the undesirable batch term to zero and recomputing the observations <em>sans</em> the batch effect, yielding a set of corrected expression values for downstream analyses. Linear modelling is the basis of the <code>removeBatchEffect()</code> function from the <em><a href="https://bioconductor.org/packages/3.11/limma">limma</a></em> package <span class="citation">(Ritchie et al. <a href="#ref-ritchie2015limma">2015</a>)</span> as well the <code>comBat()</code> function from the <em><a href="https://bioconductor.org/packages/3.11/sva">sva</a></em> package <span class="citation">(Leek et al. <a href="#ref-leek2012sva">2012</a>)</span>.</p>
<p>To use this approach in a scRNA-seq context, we assume that the composition of cell subpopulations is the same across batches. We also assume that the batch effect is additive, i.e., any batch-induced fold-change in expression is the same across different cell subpopulations for any given gene. These are strong assumptions as batches derived from different individuals will naturally exhibit variation in cell type abundances and expression. Nonetheless, they may be acceptable when dealing with batches that are technical replicates generated from the same population of cells. (In fact, when its assumptions hold, linear regression is the most statistically efficient as it uses information from all cells to compute the common batch vector.) Linear modelling can also accommodate situations where the composition is known <em>a priori</em> by including the cell type as a factor in the linear model, but this situation is even less common.</p>
<p>We use the <code>rescaleBatches()</code> function from the <em><a href="https://bioconductor.org/packages/3.11/batchelor">batchelor</a></em> package to remove the batch effect. This is roughly equivalent to applying a linear regression to the log-expression values per gene, with some adjustments to improve performance and efficiency. For each gene, the mean expression in each batch is scaled down until it is equal to the lowest mean across all batches. We deliberately choose to scale all expression values down as this mitigates differences in variance when batches lie at different positions on the mean-variance trend. (Specifically, the shrinkage effect of the pseudo-count is greater for smaller counts, suppressing any differences in variance across batches.) An additional feature of <code>rescaleBatches()</code> is that it will preserve sparsity in the input matrix for greater efficiency, whereas other methods like <code>removeBatchEffect()</code> will always return a dense matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(batchelor)
rescaled &lt;-<span class="st"> </span><span class="kw">rescaleBatches</span>(pbmc3k, pbmc4k)
rescaled</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 31232 6791 
## metadata(0):
## assays(1): corrected
## rownames(31232): ENSG00000243485 ENSG00000237613 ... ENSG00000198695
##   ENSG00000198727
## rowData names(0):
## colnames: NULL
## colData names(1): batch
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>After clustering, we observe that most clusters consist of mixtures of cells from the two replicate batches, consistent with the removal of the batch effect. This conclusion is supported by the apparent mixing of cells from different batches in Figure <a href="integrating-datasets.html#fig:tsne-pbmc-rescaled">13.2</a>. However, at least one batch-specific cluster is still present, indicating that the correction is not entirely complete. This is attributable to violation of one of the aforementioned assumptions, even in this simple case involving replicated batches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1010101010</span>) <span class="co"># To ensure reproducibility of IRLBA.</span>
rescaled &lt;-<span class="st"> </span><span class="kw">runPCA</span>(rescaled, <span class="dt">subset_row=</span>chosen.hvgs, <span class="dt">exprs_values=</span><span class="st">&quot;corrected&quot;</span>)

snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(rescaled, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)
clusters.resc &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership
tab.resc &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters.resc, <span class="dt">Batch=</span>rescaled<span class="op">$</span>batch)
tab.resc</code></pre></div>
<pre><code>##        Batch
## Cluster    1    2
##      1   278  525
##      2    16   23
##      3   337  606
##      4    43  748
##      5   604  529
##      6    22   71
##      7   188   48
##      8    25   49
##      9   263    0
##      10  123  135
##      11   16   85
##      12   11   57
##      13  116    6
##      14  455 1035
##      15    6   31
##      16   89  187
##      17    3   36
##      18    3    8
##      19   11    3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rescaled &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(rescaled, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)
rescaled<span class="op">$</span>batch &lt;-<span class="st"> </span><span class="kw">factor</span>(rescaled<span class="op">$</span>batch)
<span class="kw">plotTSNE</span>(rescaled, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-rescaled"></span>
<img src="P2_W09.data-integration_files/figure-html/tsne-pbmc-rescaled-1.png" alt="$t$-SNE plot of the PBMC datasets after correction with `rescaleBatches()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">
Figure 13.2: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets after correction with <code>rescaleBatches()</code>. Each point represents a cell and is colored according to the batch of origin.
</p>
</div>
<p>Alternatively, we could use the <code>regressBatches()</code> function to perform a more conventional linear regression for batch correction. This is subject to the same assumptions as described above for <code>rescaleBatches()</code>, though it has the additional disadvantage of discarding sparsity in the matrix of residuals. (We do put in some effort to mitigate the loss of efficiency by avoiding explicit calculation of the residuals, see <code>?ResidualMatrix</code> for details.)</p>
</div>
<div id="performing-mnn-correction" class="section level2">
<h2><span class="header-section-number">13.5</span> Performing MNN correction</h2>
<div id="algorithm-overview" class="section level3">
<h3><span class="header-section-number">13.5.1</span> Algorithm overview</h3>
<p>Consider a cell <span class="math inline">\(a\)</span> in batch <span class="math inline">\(A\)</span>, and identify the cells in batch <span class="math inline">\(B\)</span> that are nearest neighbors to <span class="math inline">\(a\)</span> in the expression space defined by the selected features. Repeat this for a cell <span class="math inline">\(b\)</span> in batch <span class="math inline">\(B\)</span>, identifying its nearest neighbors in <span class="math inline">\(A\)</span>. Mutual nearest neighbors are pairs of cells from different batches that belong in each other’s set of nearest neighbors. The reasoning is that MNN pairs represent cells from the same biological state prior to the application of a batch effect - see <span class="citation">Haghverdi et al. (<a href="#ref-haghverdi2018batch">2018</a>)</span> for full theoretical details. Thus, the difference between cells in MNN pairs can be used as an estimate of the batch effect, the subtraction of which yields batch-corrected values.</p>
<p>Compared to linear regression, MNN correction does not assume that the population composition is the same or known beforehand. This is because it learns the shared population structure via identification of MNN pairs and uses this information to obtain an appropriate estimate of the batch effect. Instead, the key assumption of MNN-based approaches is that the batch effect is orthogonal to the biology in high-dimensional expression space. Violations reduce the effectiveness and accuracy of the correction, with the most common case arising from variations in the direction of the batch effect between clusters. Nonetheless, the assumption is usually reasonable as a random vector is very likely to be orthogonal in high-dimensional space.</p>
</div>
<div id="application-to-the-pbmc-data" class="section level3">
<h3><span class="header-section-number">13.5.2</span> Application to the PBMC data</h3>
<p>The <em><a href="https://bioconductor.org/packages/3.11/batchelor">batchelor</a></em> package provides an implementation of the MNN approach via the <code>fastMNN()</code> function. (Unlike the MNN method originally described by <span class="citation">Haghverdi et al. (<a href="#ref-haghverdi2018batch">2018</a>)</span>, the <code>fastMNN()</code> function performs PCA to reduce the dimensions beforehand and speed up the downstream neighbor detection steps.) We apply it to our two PBMC batches to remove the batch effect across the highly variable genes in <code>chosen.hvgs</code>. To reduce computational work and technical noise, all cells in all batches are projected into the low-dimensional space defined by the top <code>d</code> principal components. Identification of MNNs and calculation of correction vectors are then performed in this low-dimensional space.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Using randomized SVD here, as this is faster than </span>
<span class="co"># irlba for file-backed matrices.</span>
<span class="kw">set.seed</span>(<span class="dv">1000101001</span>)
mnn.out &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(pbmc3k, pbmc4k, <span class="dt">d=</span><span class="dv">50</span>, <span class="dt">k=</span><span class="dv">20</span>, <span class="dt">subset.row=</span>chosen.hvgs,
    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>(<span class="dt">deferred=</span><span class="ot">TRUE</span>))
mnn.out</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 13431 6791 
## metadata(2): merge.info pca.info
## assays(1): reconstructed
## rownames(13431): ENSG00000239945 ENSG00000228463 ... ENSG00000198695
##   ENSG00000198727
## rowData names(1): rotation
## colnames: NULL
## colData names(1): batch
## reducedDimNames(1): corrected
## altExpNames(0):</code></pre>
<p>The function returns a <code>SingleCellExperiment</code> object containing corrected values for downstream analyses like clustering or visualization. Each column of <code>mnn.out</code> corresponds to a cell in one of the batches, while each row corresponds to an input gene in <code>chosen.hvgs</code>. The <code>batch</code> field in the column metadata contains a vector specifying the batch of origin of each cell.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(mnn.out<span class="op">$</span>batch) </code></pre></div>
<pre><code>## [1] 1 1 1 1 1 1</code></pre>
<p>The <code>corrected</code> matrix in the <code>reducedDims()</code> contains the low-dimensional corrected coordinates for all cells, which we will use in place of the PCs in our downstream analyses.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(<span class="kw">reducedDim</span>(mnn.out, <span class="st">&quot;corrected&quot;</span>))</code></pre></div>
<pre><code>## [1] 6791   50</code></pre>
<p>A <code>reconstructed</code> matrix in the <code>assays()</code> contains the corrected expression values for each gene in each cell, obtained by projecting the low-dimensional coordinates in <code>corrected</code> back into gene expression space. We do not recommend using this for anything other than visualization (Section <a href="integrating-datasets.html#using-corrected-values">13.7</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assay</span>(mnn.out, <span class="st">&quot;reconstructed&quot;</span>)</code></pre></div>
<pre><code>## &lt;13431 x 6791&gt; matrix of class LowRankMatrix and type &quot;double&quot;:
##                          [,1]          [,2]          [,3] ...       [,6790]
## ENSG00000239945 -2.522191e-06 -1.851424e-06 -1.198984e-05   .  1.831777e-06
## ENSG00000228463 -6.626821e-04 -6.724341e-04 -4.820230e-04   . -8.531370e-04
## ENSG00000237094 -8.077231e-05 -8.038006e-05 -9.630608e-05   .  7.261218e-06
## ENSG00000229905  3.838135e-06  6.179994e-06  5.432122e-06   .  8.533821e-06
## ENSG00000237491 -4.527173e-04 -3.178168e-04 -1.510488e-04   . -3.490597e-04
##             ...             .             .             .   .             .
## ENSG00000198840 -0.0296507947 -0.0340100587 -0.0502384904   . -0.0362883569
## ENSG00000212907 -0.0041681457 -0.0056569549 -0.0106420441   . -0.0083837431
## ENSG00000198886  0.0145358352  0.0200516598 -0.0307130838   . -0.0109253965
## ENSG00000198695  0.0014427168  0.0013489620  0.0001493234   . -0.0009825885
## ENSG00000198727  0.0152570264  0.0106167007 -0.0256449590   . -0.0227962145
##                       [,6791]
## ENSG00000239945 -3.640783e-06
## ENSG00000228463 -3.998685e-04
## ENSG00000237094 -4.093746e-05
## ENSG00000229905  3.484915e-06
## ENSG00000237491 -2.082230e-04
##             ...             .
## ENSG00000198840 -0.0183084463
## ENSG00000212907  0.0005995725
## ENSG00000198886 -0.0070064062
## ENSG00000198695 -0.0022712251
## ENSG00000198727 -0.0022897927</code></pre>
<p>The most relevant parameter for tuning <code>fastMNN()</code> is <code>k</code>, which specifies the number of nearest neighbors to consider when defining MNN pairs. This can be interpreted as the minimum anticipated frequency of any shared cell type or state in each batch. Increasing <code>k</code> will generally result in more aggressive merging as the algorithm is more generous in matching subpopulations across batches. It can occasionally be desirable to increase <code>k</code> if one clearly sees that the same cell types are not being adequately merged across batches.</p>
<p>See Chapter <a href="merged-pancreas.html#merged-pancreas">32</a> for an example of a more complex <code>fastMNN()</code> merge involving several human pancreas datasets generated by different authors on different patients with different technologies.</p>
</div>
<div id="correction-diagnostics" class="section level3">
<h3><span class="header-section-number">13.5.3</span> Correction diagnostics</h3>
<p>We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure. If the batch effect is successfully corrected, clusters corresponding to shared cell types or states should contain cells from multiple batches. We see that all clusters contain contributions from each batch after correction, consistent with our expectation that the two batches are replicates of each other.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(mnn.out, <span class="dt">use.dimred=</span><span class="st">&quot;corrected&quot;</span>)
clusters.mnn &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership
tab.mnn &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters.mnn, <span class="dt">Batch=</span>mnn.out<span class="op">$</span>batch)
tab.mnn</code></pre></div>
<pre><code>##        Batch
## Cluster    1    2
##      1   337  606
##      2   289  542
##      3   152  181
##      4    12    4
##      5   517  467
##      6    17   19
##      7   313  661
##      8   162  118
##      9    11   56
##      10  547 1083
##      11   17   59
##      12   16   58
##      13  144   93
##      14   67  191
##      15    4   36
##      16    4    8</code></pre>
<p>We can also visualize the corrected coordinates using a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="integrating-datasets.html#fig:tsne-pbmc-corrected">13.3</a>). The presence of visual clusters containing cells from both batches provides a comforting illusion that the correction was successful.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)
<span class="kw">set.seed</span>(<span class="dv">0010101010</span>)
mnn.out &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(mnn.out, <span class="dt">dimred=</span><span class="st">&quot;corrected&quot;</span>)

mnn.out<span class="op">$</span>batch &lt;-<span class="st"> </span><span class="kw">factor</span>(mnn.out<span class="op">$</span>batch)
<span class="kw">plotTSNE</span>(mnn.out, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-corrected"></span>
<img src="P2_W09.data-integration_files/figure-html/tsne-pbmc-corrected-1.png" alt="$t$-SNE plot of the PBMC datasets after MNN correction. Each point is a cell that is colored according to its batch of origin." width="672" />
<p class="caption">
Figure 13.3: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets after MNN correction. Each point is a cell that is colored according to its batch of origin.
</p>
</div>
<p>For <code>fastMNN()</code>, one useful diagnostic is the proportion of variance within each batch that is lost during MNN correction. Specifically, this refers to the within-batch variance that is removed during orthogonalization with respect to the average correction vector at each merge step. This is returned via the <code>lost.var</code> field in the metadata of <code>mnn.out</code>, which contains a matrix of the variance lost in each batch (column) at each merge step (row).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">metadata</span>(mnn.out)<span class="op">$</span>merge.info<span class="op">$</span>lost.var</code></pre></div>
<pre><code>##             [,1]        [,2]
## [1,] 0.006617087 0.003315395</code></pre>
<p>Large proportions of lost variance (&gt;10%) suggest that correction is removing genuine biological heterogeneity. This would occur due to violations of the assumption of orthogonality between the batch effect and the biological subspace <span class="citation">(Haghverdi et al. <a href="#ref-haghverdi2018batch">2018</a>)</span>. In this case, the proportion of lost variance is small, indicating that non-orthogonality is not a major concern.</p>
</div>
</div>
<div id="preserving-biological-heterogeneity" class="section level2">
<h2><span class="header-section-number">13.6</span> Preserving biological heterogeneity</h2>
<div id="comparison-to-within-batch-clusters" class="section level3">
<h3><span class="header-section-number">13.6.1</span> Comparison to within-batch clusters</h3>
<p>Another useful diagnostic check is to compare the clustering within each batch to the clustering of the merged data. Accurate data integration should preserve variance within each batch as there should be nothing to remove between cells in the same batch. This check complements the previously mentioned diagnostics that only focus on the removal of differences between batches. Specifically, it protects us against cases where the correction method simply aggregates all cells together, which would achieve perfect mixing but also discard the biological heterogeneity of interest.</p>
<p>Ideally, we should see a many-to-1 mapping where the across-batch clustering is nested inside the within-batch clusterings. This indicates that any within-batch structure was preserved after correction while acknowledging that greater resolution is possible with more cells. In practice, more discrepancies can be expected even when the correction is perfect, due to the existence of closely related clusters that were arbitrarily separated in the within-batch clustering. As a general rule, we can be satisfied with the correction if the vast majority of entries in Figure <a href="integrating-datasets.html#fig:heat-after-mnn">13.4</a> are zero, though this may depend on whether specific clusters of interest are gained or lost.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(pheatmap)

<span class="co"># For the first batch (adding +10 for a smoother color transition</span>
<span class="co"># from zero to non-zero counts for any given matrix entry).</span>
tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">paste</span>(<span class="st">&quot;after&quot;</span>, clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>]),
    <span class="kw">paste</span>(<span class="st">&quot;before&quot;</span>, <span class="kw">colLabels</span>(pbmc3k)))
heat3k &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(<span class="kw">log10</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">cluster_row=</span><span class="ot">FALSE</span>, <span class="dt">cluster_col=</span><span class="ot">FALSE</span>,
    <span class="dt">main=</span><span class="st">&quot;PBMC 3K comparison&quot;</span>, <span class="dt">silent=</span><span class="ot">TRUE</span>)

<span class="co"># For the second batch.</span>
tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">paste</span>(<span class="st">&quot;after&quot;</span>, clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">2</span>]),
    <span class="kw">paste</span>(<span class="st">&quot;before&quot;</span>, <span class="kw">colLabels</span>(pbmc4k)))
heat4k &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(<span class="kw">log10</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">cluster_row=</span><span class="ot">FALSE</span>, <span class="dt">cluster_col=</span><span class="ot">FALSE</span>,
    <span class="dt">main=</span><span class="st">&quot;PBMC 4K comparison&quot;</span>, <span class="dt">silent=</span><span class="ot">TRUE</span>)

gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(heat3k[[<span class="dv">4</span>]], heat4k[[<span class="dv">4</span>]])</code></pre></div>
<div class="figure"><span id="fig:heat-after-mnn"></span>
<img src="P2_W09.data-integration_files/figure-html/heat-after-mnn-1.png" alt="Comparison between the within-batch clusters and the across-batch clusters obtained after MNN correction. One heatmap is generated for each of the PBMC 3K and 4K datasets, where each entry is colored according to the number of cells with each pair of labels (before and after correction)." width="672" />
<p class="caption">
Figure 13.4: Comparison between the within-batch clusters and the across-batch clusters obtained after MNN correction. One heatmap is generated for each of the PBMC 3K and 4K datasets, where each entry is colored according to the number of cells with each pair of labels (before and after correction).
</p>
</div>
<p>Another evaluation approach is to compute the coassignment probabilities (Section <a href="clustering.html#cluster-bootstrapping">10.6</a>), i.e., the probability that cells from two within-batch clusters are clustered together in the across-batch clustering. High probabilities off the diagonal in Figure <a href="integrating-datasets.html#fig:coassign-after-mnn">13.5</a> indicate that within-batch clusters are merged in the across-batch analysis. We would generally expect low off-diagonal probabilities for most pairs of clusters, though this may not be reasonably possible if the within-batch clusters were poorly separated in the first place.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># For the first batch.</span>
tab &lt;-<span class="st"> </span><span class="kw">coassignProb</span>(<span class="kw">colLabels</span>(pbmc3k), clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>])
heat3k &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(tab, <span class="dt">cluster_row=</span><span class="ot">FALSE</span>, <span class="dt">cluster_col=</span><span class="ot">FALSE</span>,
    <span class="dt">col=</span><span class="kw">rev</span>(viridis<span class="op">::</span><span class="kw">magma</span>(<span class="dv">100</span>)), <span class="dt">main=</span><span class="st">&quot;PBMC 3K probabilities&quot;</span>, <span class="dt">silent=</span><span class="ot">TRUE</span>)

<span class="co"># For the second batch.</span>
tab &lt;-<span class="st"> </span><span class="kw">coassignProb</span>(<span class="kw">colLabels</span>(pbmc4k), clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">2</span>])
heat4k &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(tab, <span class="dt">cluster_row=</span><span class="ot">FALSE</span>, <span class="dt">cluster_col=</span><span class="ot">FALSE</span>,
    <span class="dt">col=</span><span class="kw">rev</span>(viridis<span class="op">::</span><span class="kw">magma</span>(<span class="dv">100</span>)), <span class="dt">main=</span><span class="st">&quot;PBMC 4K probabilities&quot;</span>, <span class="dt">silent=</span><span class="ot">TRUE</span>)

gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(heat3k[[<span class="dv">4</span>]], heat4k[[<span class="dv">4</span>]])</code></pre></div>
<div class="figure"><span id="fig:coassign-after-mnn"></span>
<img src="P2_W09.data-integration_files/figure-html/coassign-after-mnn-1.png" alt="Coassignment probabilities for the within-batch clusters, based on coassignment of cells in the across-batch clusters obtained after MNN correction. One heatmap is generated for each of the PBMC 3K and 4K datasets, where each entry is colored according to the coassignment probability between each pair of within-batch clusters." width="672" />
<p class="caption">
Figure 13.5: Coassignment probabilities for the within-batch clusters, based on coassignment of cells in the across-batch clusters obtained after MNN correction. One heatmap is generated for each of the PBMC 3K and 4K datasets, where each entry is colored according to the coassignment probability between each pair of within-batch clusters.
</p>
</div>
<p>Finally, we can summarize the agreement between clusterings by computing the Rand index. This provides a simple metric that we can use to assess the preservation of variation by different correction methods. Larger rand indices (i.e., closer to 1) are more desirable, though this must be balanced against the ability of each method to actually remove the batch effect.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(fossil)
ri3k &lt;-<span class="st"> </span><span class="kw">rand.index</span>(<span class="kw">as.integer</span>(clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>]),
    <span class="kw">as.integer</span>(<span class="kw">colLabels</span>(pbmc3k)))
ri3k</code></pre></div>
<pre><code>## [1] 0.9335226</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ri4k &lt;-<span class="st"> </span><span class="kw">rand.index</span>(<span class="kw">as.integer</span>(clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">2</span>]),
    <span class="kw">as.integer</span>(<span class="kw">colLabels</span>(pbmc4k)))
ri4k</code></pre></div>
<pre><code>## [1] 0.9575746</code></pre>
</div>
<div id="integration-with-markers" class="section level3">
<h3><span class="header-section-number">13.6.2</span> Encouraging consistency with marker genes</h3>
<p>In some situations, we will already have performed within-batch analyses to characterize salient aspects of population heterogeneity. This is not uncommon when merging datasets from different sources where each dataset has already been analyzed, annotated and interpreted separately. It is subsequently desirable for the integration procedure to retain these “known interesting” aspects of each dataset in the merged dataset. We can encourage this outcome by using the marker genes within each dataset as our selected feature set for <code>fastMNN()</code> and related methods. This focuses on the relevant heterogeneity and represents a semi-supervised approach that is a natural extension of the strategy described in Section <a href="feature-selection.html#apriori-hvgs">8.4</a>.</p>
<p>To illustrate, we apply this strategy to our PBMC datasets. We identify the top marker genes from pairwise Wilcoxon ranked sum tests between every pair of clusters <em>within</em> each batch, analogous to the method used by <em><a href="https://bioconductor.org/packages/3.11/SingleR">SingleR</a></em> (Chapter <a href="cell-type-annotation.html#cell-type-annotation">12</a>). In this case, we use the top 10 marker genes but any value can be used depending on the acceptable trade-off between signal and noise (and speed). We then take the union across all comparisons in all batches and use that in place of our HVG set in <code>fastMNN()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Recall that groups for marker detection</span>
<span class="co"># are automatically defined from &#39;colLabels()&#39;. </span>
stats3 &lt;-<span class="st"> </span><span class="kw">pairwiseWilcox</span>(pbmc3k, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)
markers3 &lt;-<span class="st"> </span><span class="kw">getTopMarkers</span>(stats3[[<span class="dv">1</span>]], stats3[[<span class="dv">2</span>]], <span class="dt">n=</span><span class="dv">10</span>)

stats4 &lt;-<span class="st"> </span><span class="kw">pairwiseWilcox</span>(pbmc4k, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)
markers4 &lt;-<span class="st"> </span><span class="kw">getTopMarkers</span>(stats4[[<span class="dv">1</span>]], stats4[[<span class="dv">2</span>]], <span class="dt">n=</span><span class="dv">10</span>)

marker.set &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">unlist</span>(<span class="kw">c</span>(<span class="kw">unlist</span>(markers3), <span class="kw">unlist</span>(markers4))))
<span class="kw">length</span>(marker.set) <span class="co"># getting the total number of genes selected in this manner.</span></code></pre></div>
<pre><code>## [1] 314</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1000110</span>)
mnn.out2 &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(pbmc3k, pbmc4k, <span class="dt">subset.row=</span>marker.set,
    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>(<span class="dt">deferred=</span><span class="ot">TRUE</span>))</code></pre></div>
<p>A quick inspection of Figure <a href="integrating-datasets.html#fig:tsne-pbmc-corrected-markers">13.6</a> indicates that the original within-batch structure is indeed preserved in the corrected data. This highlights the utility of a marker-based feature set for integrating datasets that have already been characterized separately in a manner that preserves existing interpretations of each dataset. We note that some within-batch clusters have merged, most likely due to the lack of robust separation in the first place, though this may also be treated as a diagnostic on the appropriateness of the integration depending on the context.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mnn.out2 &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(mnn.out2, <span class="dt">dimred=</span><span class="st">&quot;corrected&quot;</span>)
gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(
    <span class="kw">plotTSNE</span>(mnn.out2[,mnn.out2<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>], <span class="dt">colour_by=</span><span class="kw">I</span>(<span class="kw">colLabels</span>(pbmc3k))),
    <span class="kw">plotTSNE</span>(mnn.out2[,mnn.out2<span class="op">$</span>batch<span class="op">==</span><span class="dv">2</span>], <span class="dt">colour_by=</span><span class="kw">I</span>(<span class="kw">colLabels</span>(pbmc4k))),
    <span class="dt">ncol=</span><span class="dv">2</span>
)</code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-corrected-markers"></span>
<img src="P2_W09.data-integration_files/figure-html/tsne-pbmc-corrected-markers-1.png" alt="$t$-SNE plots of the merged PBMC datasets, where the merge was performed using only marker genes identified within each batch. Each point represents a cell that is colored by the assigned cluster from the within-batch analysis for the 3K (left) and 4K dataset (right)." width="672" />
<p class="caption">
Figure 13.6: <span class="math inline">\(t\)</span>-SNE plots of the merged PBMC datasets, where the merge was performed using only marker genes identified within each batch. Each point represents a cell that is colored by the assigned cluster from the within-batch analysis for the 3K (left) and 4K dataset (right).
</p>
</div>
</div>
</div>
<div id="using-corrected-values" class="section level2">
<h2><span class="header-section-number">13.7</span> Using the corrected values</h2>
<p>The greatest value of batch correction lies in facilitating cell-based analysis of population heterogeneity in a consistent manner across batches. Cluster 1 in batch A is the same as cluster 1 in batch B when the clustering is performed on the merged data. There is no need to identify mappings between separate clusterings, which might not even be possible when the clusters are not well-separated. The burden of interpretation is consolidated by generating a single set of clusters for all batches, rather than requiring separate examination of each batch’s clusters. Another benefit is that the available number of cells is increased when all batches are combined, which allows for greater resolution of population structure in downstream analyses. We previously demonstrated the application of clustering methods to the batch-corrected data, but the same principles apply for other analyses like trajectory reconstruction.</p>
<p>At this point, it is also tempting to use the corrected expression values for gene-based analyses like DE-based marker gene detection. This is not generally recommended as an arbitrary correction algorithm is not obliged to preserve the magnitude (or even direction) of differences in per-gene expression when attempting to align multiple batches. For example, cosine normalization in <code>fastMNN()</code> shrinks the magnitude of the expression values so that the computed log-fold changes have no obvious interpretation. Of greater concern is the possibility that the correction introduces artificial agreement across batches. To illustrate:</p>
<ol style="list-style-type: decimal">
<li>Consider a dataset (first batch) with two cell types, <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. Consider a second batch with the same cell types, denoted as <span class="math inline">\(A&#39;\)</span> and <span class="math inline">\(B&#39;\)</span>. Assume that, for some reason, gene <span class="math inline">\(X\)</span> is expressed in <span class="math inline">\(A\)</span> but not in <span class="math inline">\(A&#39;\)</span>, <span class="math inline">\(B\)</span> or <span class="math inline">\(B&#39;\)</span> - possibly due to some difference in how the cells were treated, or maybe due to a donor effect.</li>
<li>We then merge the batches together based on the shared cell types. This yields a result where <span class="math inline">\(A\)</span> and <span class="math inline">\(A&#39;\)</span> cells are intermingled and the difference due to <span class="math inline">\(X\)</span> is eliminated. One can debate whether this <em>should</em> be the case, but in general, it is necessary for batch correction methods to smooth over small biological differences (as discussed in Section <a href="integrating-datasets.html#batch-diagnosis">13.3</a>).</li>
<li>Now, if we corrected the second batch to the first, we must have coerced the expression values of <span class="math inline">\(X\)</span> in <span class="math inline">\(A&#39;\)</span> to non-zero values to align with those of <span class="math inline">\(A\)</span>, while leaving the expression of <span class="math inline">\(X\)</span> in <span class="math inline">\(B&#39;\)</span> and <span class="math inline">\(B\)</span> at zero. Thus, we have artificially introduced DE between <span class="math inline">\(A&#39;\)</span> and <span class="math inline">\(B&#39;\)</span> for <span class="math inline">\(X\)</span> in the second batch to align with the DE between <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> in the first batch. (The converse is also possible where DE in the first batch is artificially removed to align with the second batch, depending on the order of merges.)</li>
<li>The artificial DE has implications for the identification of the cell types and interpretation of the results. We would be misled into believing that both <span class="math inline">\(A\)</span> and <span class="math inline">\(A&#39;\)</span> are <span class="math inline">\(X\)</span>-positive, when in fact this is only true for <span class="math inline">\(A\)</span>. At best, this is only a minor error - after all, we do actually have <span class="math inline">\(X\)</span>-positive cells of that overall type, we simply do not see that <span class="math inline">\(A&#39;\)</span> is <span class="math inline">\(X\)</span>-negative. At worst, this can compromise the conclusions, e.g., if the first batch was drug treated and the second batch was a control, we might mistakenly think that a <span class="math inline">\(X\)</span>-positive population exists in the latter and conclude that our drug has no effect.</li>
</ol>
<!--
There are also some statistical considerations:

- It is usually inappropriate to perform DE analyses on batch-corrected values, due to the failure to model the uncertainty of the correction.
This usually results in loss of type I error control, i.e., more false positives than expected.
- The correction does not preserve the mean-variance relationship.
Applications of common DE methods like *[edgeR](https://bioconductor.org/packages/3.11/edgeR)* or *[limma](https://bioconductor.org/packages/3.11/limma)* are unlikely to be valid.

However, these are probably minor given that the marker gene procedure is not particularly rigorous in the first place!
-->
<p>Rather, it is preferable to perform DE analyses using the uncorrected expression values with blocking on the batch, as discussed in Section <a href="marker-detection.html#marker-batch">11.4</a>. This strategy is based on the expectation that any genuine DE between clusters should still be present in a within-batch comparison where batch effects are absent. It penalizes genes that exhibit inconsistent DE across batches, thus protecting against misleading conclusions when a population in one batch is aligned to a similar-but-not-identical population in another batch. We demonstrate this approach below using a blocked <span class="math inline">\(t\)</span>-test to detect markers in the PBMC dataset, where the presence of the same pattern across clusters within each batch (Figure <a href="integrating-datasets.html#fig:pbmc-marker-blocked">13.7</a>) is reassuring. If integration is performed across multiple conditions, it is even more important to use the uncorrected expression values for downstream analyses - see Section <a href="multi-sample-comparisons.html#sacrificing-differences">14.5.2</a> for a discussion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m.out &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(uncorrected, clusters.mnn, <span class="dt">block=</span>uncorrected<span class="op">$</span>batch,
    <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>, <span class="dt">lfc=</span><span class="dv">1</span>, <span class="dt">row.data=</span><span class="kw">rowData</span>(uncorrected)[,<span class="dv">3</span>,<span class="dt">drop=</span><span class="ot">FALSE</span>])

<span class="co"># A (probably activated?) T cell subtype of some sort:</span>
demo &lt;-<span class="st"> </span>m.out[[<span class="st">&quot;10&quot;</span>]]
<span class="kw">as.data.frame</span>(demo[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>,<span class="kw">c</span>(<span class="st">&quot;Symbol&quot;</span>, <span class="st">&quot;Top&quot;</span>, <span class="st">&quot;p.value&quot;</span>, <span class="st">&quot;FDR&quot;</span>)]) </code></pre></div>
<pre><code>##                 Symbol Top       p.value           FDR
## ENSG00000177954  RPS27   1 3.398543e-168 1.061433e-163
## ENSG00000227507    LTB   1 1.238347e-157 1.933803e-153
## ENSG00000167286   CD3D   1  9.136016e-89  4.076229e-85
## ENSG00000111716   LDHB   1  8.699141e-44  1.811277e-40
## ENSG00000008517   IL32   1  4.880406e-31  6.928402e-28
## ENSG00000172809  RPL38   1 8.726980e-143 6.814026e-139
## ENSG00000171223   JUNB   1  8.762155e-72  2.736596e-68
## ENSG00000071082  RPL31   2  8.612369e-78  2.988684e-74
## ENSG00000121966  CXCR4   2  2.369668e-07  1.321598e-04
## ENSG00000251562 MALAT1   2  3.618005e-33  5.649876e-30
## ENSG00000133639   BTG1   2  6.847076e-12  4.549955e-09
## ENSG00000170345    FOS   2  2.738128e-46  6.108372e-43
## ENSG00000129824 RPS4Y1   2 1.074722e-108 6.713145e-105
## ENSG00000177606    JUN   3  1.039395e-37  1.909553e-34
## ENSG00000112306  RPS12   3  1.655795e-33  2.721779e-30
## ENSG00000110700  RPS13   3  7.600217e-18  7.657096e-15
## ENSG00000198851   CD3E   3  1.057929e-36  1.835625e-33
## ENSG00000213741  RPS29   3 1.493885e-148 1.555234e-144
## ENSG00000116251  RPL22   4  3.992343e-25  4.795726e-22
## ENSG00000144713  RPL32   4  1.223940e-32  1.820290e-29</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotExpression</span>(uncorrected, <span class="dt">x=</span><span class="kw">I</span>(<span class="kw">factor</span>(clusters.mnn)), 
    <span class="dt">features=</span><span class="st">&quot;ENSG00000177954&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">facet_wrap</span>(<span class="op">~</span>colour_by)</code></pre></div>
<div class="figure"><span id="fig:pbmc-marker-blocked"></span>
<img src="P2_W09.data-integration_files/figure-html/pbmc-marker-blocked-1.png" alt="Distributions of RPSA uncorrected log-expression values within each cluster in each batch of the merged PBMC dataset." width="672" />
<p class="caption">
Figure 13.7: Distributions of RPSA uncorrected log-expression values within each cluster in each batch of the merged PBMC dataset.
</p>
</div>
<p>We suggest limiting the use of per-gene corrected values to visualization, e.g., when coloring points on a <span class="math inline">\(t\)</span>-SNE plot by per-cell expression. This can be more aesthetically pleasing than uncorrected expression values that may contain large shifts on the colour scale between cells in different batches. Use of the corrected values in any quantitative procedure should be treated with caution, and should be backed up by similar results from an analysis on the uncorrected values.</p>
</div>
<div id="session-info-8" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] fossil_0.4.0                shapefiles_0.7             
 [3] foreign_0.8-79              maps_3.3.0                 
 [5] sp_1.4-1                    pheatmap_1.0.12            
 [7] scater_1.16.0               ggplot2_3.3.0              
 [9] scran_1.16.0                batchelor_1.4.0            
[11] SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.1
[13] Biobase_2.48.0              GenomicRanges_1.40.0       
[15] GenomeInfoDb_1.24.0         HDF5Array_1.16.0           
[17] rhdf5_2.32.0                DelayedArray_0.14.0        
[19] IRanges_2.22.1              S4Vectors_0.26.0           
[21] BiocGenerics_0.34.0         matrixStats_0.56.0         
[23] OSCAUtils_0.0.2             BiocStyle_2.16.0           

loaded via a namespace (and not attached):
 [1] bitops_1.0-6              RColorBrewer_1.1-2       
 [3] tools_4.0.0               R6_2.4.1                 
 [5] irlba_2.3.3               vipor_0.4.5              
 [7] colorspace_1.4-1          withr_2.2.0              
 [9] tidyselect_1.1.0          gridExtra_2.3            
[11] processx_3.4.2            compiler_4.0.0           
[13] BiocNeighbors_1.6.0       labeling_0.3             
[15] bookdown_0.19             scales_1.1.1             
[17] callr_3.4.3               stringr_1.4.0            
[19] digest_0.6.25             rmarkdown_2.1            
[21] XVector_0.28.0            pkgconfig_2.0.3          
[23] htmltools_0.4.0           limma_3.44.1             
[25] highr_0.8                 rlang_0.4.6              
[27] DelayedMatrixStats_1.10.0 farver_2.0.3             
[29] BiocParallel_1.22.0       dplyr_0.8.5              
[31] RCurl_1.98-1.2            magrittr_1.5             
[33] BiocSingular_1.4.0        GenomeInfoDbData_1.2.3   
[35] Matrix_1.2-18             Rcpp_1.0.4.6             
[37] ggbeeswarm_0.6.0          munsell_0.5.0            
[39] Rhdf5lib_1.10.0           viridis_0.5.1            
[41] lifecycle_0.2.0           stringi_1.4.6            
[43] yaml_2.2.1                edgeR_3.30.0             
[45] zlibbioc_1.34.0           Rtsne_0.15               
[47] grid_4.0.0                dqrng_0.2.1              
[49] crayon_1.3.4              lattice_0.20-41          
[51] beachmat_2.4.0            cowplot_1.0.0            
[53] locfit_1.5-9.4            knitr_1.28               
[55] ps_1.3.3                  pillar_1.4.4             
[57] igraph_1.2.5              codetools_0.2-16         
[59] glue_1.4.1                evaluate_0.14            
[61] BiocManager_1.30.10       vctrs_0.3.0              
[63] gtable_0.3.0              purrr_0.3.4              
[65] assertthat_0.2.1          xfun_0.13                
[67] rsvd_1.0.3                viridisLite_0.3.0        
[69] tibble_3.0.1              beeswarm_0.2.3           
[71] statmod_1.4.34            ellipsis_0.3.1           </code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-butler2018integrating">
<p>Butler, A., P. Hoffman, P. Smibert, E. Papalexi, and R. Satija. 2018. “Integrating single-cell transcriptomic data across different conditions, technologies, and species.” <em>Nat. Biotechnol.</em> 36 (5): 411–20.</p>
</div>
<div id="ref-haghverdi2018batch">
<p>Haghverdi, L., A. T. L. Lun, M. D. Morgan, and J. C. Marioni. 2018. “Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors.” <em>Nat. Biotechnol.</em> 36 (5): 421–27.</p>
</div>
<div id="ref-leek2012sva">
<p>Leek, J. T., W. E. Johnson, H. S. Parker, A. E. Jaffe, and J. D. Storey. 2012. “The sva package for removing batch effects and other unwanted variation in high-throughput experiments.” <em>Bioinformatics</em> 28 (6): 882–83.</p>
</div>
<div id="ref-lin2019scmerge">
<p>Lin, Y., S. Ghazanfar, K. Y. X. Wang, J. A. Gagnon-Bartsch, K. K. Lo, X. Su, Z. G. Han, et al. 2019. “scMerge leverages factor analysis, stable expression, and pseudoreplication to merge multiple single-cell RNA-seq datasets.” <em>Proc. Natl. Acad. Sci. U.s.a.</em> 116 (20): 9775–84.</p>
</div>
<div id="ref-ritchie2015limma">
<p>Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K. Smyth. 2015. “limma powers differential expression analyses for RNA-sequencing and microarray studies.” <em>Nucleic Acids Res.</em> 43 (7): e47.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="cell-type-annotation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="multi-sample-comparisons.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
