<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 15 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 15 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 15 Integrating Datasets | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-08-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="doublet-detection.html"/>
<link rel="next" href="multi-sample-comparisons.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="feature-selection-2.html"><a href="feature-selection-2.html"><i class="fa fa-check"></i><b>10</b> Feature selection</a><ul>
<li class="chapter" data-level="10.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>10.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="10.2.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>10.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="10.2.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#coefficient-of-variation"><i class="fa fa-check"></i><b>10.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="10.2.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#sec:spikeins"><i class="fa fa-check"></i><b>10.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="10.2.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>10.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#selecting-highly-variable-genes"><i class="fa fa-check"></i><b>10.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="10.3.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>10.3.1</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="10.3.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-a-fixed-threshold"><i class="fa fa-check"></i><b>10.3.2</b> Based on a fixed threshold</a></li>
<li class="chapter" data-level="10.3.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#keeping-all-genes-above-the-trend"><i class="fa fa-check"></i><b>10.3.3</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>10.4</b> Putting it all together</a></li>
<li class="chapter" data-level="" data-path="feature-selection-2.html"><a href="feature-selection-2.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="feature-selection-2.html"><a href="feature-selection-2.html#references-1"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html"><i class="fa fa-check"></i><b>11</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="11.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#overview-1"><i class="fa fa-check"></i><b>11.1</b> Overview</a></li>
<li class="chapter" data-level="11.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#principal-components-analysis"><i class="fa fa-check"></i><b>11.2</b> Principal components analysis</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#background-1"><i class="fa fa-check"></i><b>11.2.1</b> Background</a></li>
<li class="chapter" data-level="11.2.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#performing-the-pca"><i class="fa fa-check"></i><b>11.2.2</b> Performing the PCA</a></li>
<li class="chapter" data-level="11.2.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>11.2.3</b> Choosing the number of PCs</a></li>
<li class="chapter" data-level="11.2.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#putting-it-together"><i class="fa fa-check"></i><b>11.2.4</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>11.3</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="11.3.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#motivation-4"><i class="fa fa-check"></i><b>11.3.1</b> Motivation</a></li>
<li class="chapter" data-level="11.3.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#visualizating-with-pca"><i class="fa fa-check"></i><b>11.3.2</b> Visualizating with PCA</a></li>
<li class="chapter" data-level="11.3.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>11.3.3</b> t-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="11.3.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>11.3.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="11.3.5" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#interpreting-the-plots"><i class="fa fa-check"></i><b>11.3.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="clustering-1.html"><a href="clustering-1.html"><i class="fa fa-check"></i><b>12</b> Clustering</a><ul>
<li class="chapter" data-level="12.1" data-path="clustering-1.html"><a href="clustering-1.html#motivation-5"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="clustering-1.html"><a href="clustering-1.html#comments-on-truth"><i class="fa fa-check"></i><b>12.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="12.3" data-path="clustering-1.html"><a href="clustering-1.html#graph-based-clustering"><i class="fa fa-check"></i><b>12.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="12.3.1" data-path="clustering-1.html"><a href="clustering-1.html#background-2"><i class="fa fa-check"></i><b>12.3.1</b> Background</a></li>
<li class="chapter" data-level="12.3.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation"><i class="fa fa-check"></i><b>12.3.2</b> Implementation</a></li>
<li class="chapter" data-level="12.3.3" data-path="clustering-1.html"><a href="clustering-1.html#other-parameters"><i class="fa fa-check"></i><b>12.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="12.3.4" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>12.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="clustering-1.html"><a href="clustering-1.html#k-means-clustering"><i class="fa fa-check"></i><b>12.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="12.4.1" data-path="clustering-1.html"><a href="clustering-1.html#background-3"><i class="fa fa-check"></i><b>12.4.1</b> Background</a></li>
<li class="chapter" data-level="12.4.2" data-path="clustering-1.html"><a href="clustering-1.html#base-implementation"><i class="fa fa-check"></i><b>12.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="12.4.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>12.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="clustering-1.html"><a href="clustering-1.html#hierarchical-clustering"><i class="fa fa-check"></i><b>12.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="12.5.1" data-path="clustering-1.html"><a href="clustering-1.html#background-4"><i class="fa fa-check"></i><b>12.5.1</b> Background</a></li>
<li class="chapter" data-level="12.5.2" data-path="clustering-1.html"><a href="clustering-1.html#implementation-1"><i class="fa fa-check"></i><b>12.5.2</b> Implementation</a></li>
<li class="chapter" data-level="12.5.3" data-path="clustering-1.html"><a href="clustering-1.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>12.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="clustering-1.html"><a href="clustering-1.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>13</b> Marker gene detection</a><ul>
<li class="chapter" data-level="13.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-6"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>13.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="13.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>13.2.1</b> Standard application</a></li>
<li class="chapter" data-level="13.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>13.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="13.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>13.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>13.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="13.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>13.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="13.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>13.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="13.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>13.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>13.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="13.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>13.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="13.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>13.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="13.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>13.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="13.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>13.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="13.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>13.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#references-2"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="doublet-detection.html"><a href="doublet-detection.html"><i class="fa fa-check"></i><b>14</b> Doublet detection</a><ul>
<li class="chapter" data-level="14.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-2"><i class="fa fa-check"></i><b>14.1</b> Overview</a></li>
<li class="chapter" data-level="14.2" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-with-clusters"><i class="fa fa-check"></i><b>14.2</b> Doublet detection with clusters</a><ul>
<li class="chapter" data-level="14.2.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-3"><i class="fa fa-check"></i><b>14.2.1</b> Overview</a></li>
<li class="chapter" data-level="14.2.2" data-path="doublet-detection.html"><a href="doublet-detection.html#application-to-the-mammary-data"><i class="fa fa-check"></i><b>14.2.2</b> Application to the mammary data</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-by-simulation"><i class="fa fa-check"></i><b>14.3</b> Doublet detection by simulation</a><ul>
<li class="chapter" data-level="14.3.1" data-path="doublet-detection.html"><a href="doublet-detection.html#overview-4"><i class="fa fa-check"></i><b>14.3.1</b> Overview</a></li>
<li class="chapter" data-level="14.3.2" data-path="doublet-detection.html"><a href="doublet-detection.html#application-to-the-mammary-data-1"><i class="fa fa-check"></i><b>14.3.2</b> Application to the mammary data</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="doublet-detection.html"><a href="doublet-detection.html#strengths-and-weaknesses"><i class="fa fa-check"></i><b>14.4</b> Strengths and weaknesses</a></li>
<li class="chapter" data-level="14.5" data-path="doublet-detection.html"><a href="doublet-detection.html#further-comments"><i class="fa fa-check"></i><b>14.5</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="doublet-detection.html"><a href="doublet-detection.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="14.6" data-path="doublet-detection.html"><a href="doublet-detection.html#references-3"><i class="fa fa-check"></i><b>14.6</b> References</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>15</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="15.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#motivation-7"><i class="fa fa-check"></i><b>15.1</b> Motivation</a></li>
<li class="chapter" data-level="15.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>15.2</b> Setting up the data</a></li>
<li class="chapter" data-level="15.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#diagnosing-batch-effects"><i class="fa fa-check"></i><b>15.3</b> Diagnosing batch effects</a></li>
<li class="chapter" data-level="15.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#linear-regression"><i class="fa fa-check"></i><b>15.4</b> Linear regression</a></li>
<li class="chapter" data-level="15.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#performing-mnn-correction"><i class="fa fa-check"></i><b>15.5</b> Performing MNN correction</a><ul>
<li class="chapter" data-level="15.5.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#application-to-the-pbmc-data"><i class="fa fa-check"></i><b>15.5.1</b> Application to the PBMC data</a></li>
<li class="chapter" data-level="15.5.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#correction-diagnostics"><i class="fa fa-check"></i><b>15.5.2</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="15.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preserving-biological-heterogeneity"><i class="fa fa-check"></i><b>15.6</b> Preserving biological heterogeneity</a></li>
<li class="chapter" data-level="15.7" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#application-to-a-pancreas-dataset"><i class="fa fa-check"></i><b>15.7</b> Application to a pancreas dataset</a></li>
<li class="chapter" data-level="15.8" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#using-the-corrected-values"><i class="fa fa-check"></i><b>15.8</b> Using the corrected values</a></li>
<li class="chapter" data-level="" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#references-4"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html"><i class="fa fa-check"></i><b>16</b> Multi-sample comparisons</a><ul>
<li class="chapter" data-level="16.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-8"><i class="fa fa-check"></i><b>16.1</b> Motivation</a></li>
<li class="chapter" data-level="16.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#setting-up-the-data-2"><i class="fa fa-check"></i><b>16.2</b> Setting up the data</a></li>
<li class="chapter" data-level="16.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-expression-between-conditions"><i class="fa fa-check"></i><b>16.3</b> Differential expression between conditions</a><ul>
<li class="chapter" data-level="16.3.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#creating-pseudo-bulk-samples"><i class="fa fa-check"></i><b>16.3.1</b> Creating pseudo-bulk samples</a></li>
<li class="chapter" data-level="16.3.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-de-analysis"><i class="fa fa-check"></i><b>16.3.2</b> Performing the DE analysis</a></li>
<li class="chapter" data-level="16.3.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#putting-it-all-together-2"><i class="fa fa-check"></i><b>16.3.3</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-abundance-between-conditions"><i class="fa fa-check"></i><b>16.4</b> Differential abundance between conditions</a><ul>
<li class="chapter" data-level="16.4.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#overview-5"><i class="fa fa-check"></i><b>16.4.1</b> Overview</a></li>
<li class="chapter" data-level="16.4.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-da-analysis"><i class="fa fa-check"></i><b>16.4.2</b> Performing the DA analysis</a></li>
<li class="chapter" data-level="16.4.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#composition-effects"><i class="fa fa-check"></i><b>16.4.3</b> Handling composition effects</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#de-or-da-two-sides-of-the-same-coin"><i class="fa fa-check"></i><b>16.5</b> DE or DA? Two sides of the same coin</a></li>
<li class="chapter" data-level="16.6" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#avoiding-problems-with-ambient-rna"><i class="fa fa-check"></i><b>16.6</b> Avoiding problems with ambient RNA</a><ul>
<li class="chapter" data-level="16.6.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#motivation-9"><i class="fa fa-check"></i><b>16.6.1</b> Motivation</a></li>
<li class="chapter" data-level="16.6.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#discarding-ambient-degs"><i class="fa fa-check"></i><b>16.6.2</b> Discarding ambient DEGs</a></li>
<li class="chapter" data-level="16.6.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#subtracting-ambient-counts"><i class="fa fa-check"></i><b>16.6.3</b> Subtracting ambient counts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#references-5"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="interactive-sharing.html"><a href="interactive-sharing.html"><i class="fa fa-check"></i><b>17</b> Interactive Interfaces and Sharing</a><ul>
<li class="chapter" data-level="17.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#motivation-10"><i class="fa fa-check"></i><b>17.1</b> Motivation</a></li>
<li class="chapter" data-level="17.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#interactive-quickstart"><i class="fa fa-check"></i><b>17.2</b> Quick start</a></li>
<li class="chapter" data-level="17.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#isee-examples"><i class="fa fa-check"></i><b>17.3</b> Examples of usages for iSEE apps</a><ul>
<li class="chapter" data-level="17.3.1" data-path="interactive-sharing.html"><a href="interactive-sharing.html#quality-control-1"><i class="fa fa-check"></i><b>17.3.1</b> Quality control</a></li>
<li class="chapter" data-level="17.3.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#annotation-and-identification-of-cell-populations"><i class="fa fa-check"></i><b>17.3.2</b> Annotation and identification of cell populations</a></li>
<li class="chapter" data-level="17.3.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#collaborative-analysis"><i class="fa fa-check"></i><b>17.3.3</b> Collaborative analysis</a></li>
<li class="chapter" data-level="17.3.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#dissemination"><i class="fa fa-check"></i><b>17.3.4</b> Dissemination of analysis results</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#reproducibility"><i class="fa fa-check"></i><b>17.4</b> Reproducibility</a></li>
<li class="chapter" data-level="17.5" data-path="interactive-sharing.html"><a href="interactive-sharing.html#additional-resources"><i class="fa fa-check"></i><b>17.5</b> Additional resources</a></li>
<li class="chapter" data-level="" data-path="interactive-sharing.html"><a href="interactive-sharing.html#session-info-11"><i class="fa fa-check"></i>Session Info</a></li>
<li class="chapter" data-level="" data-path="interactive-sharing.html"><a href="interactive-sharing.html#references-6"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>18</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="18.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>18.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="18.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>18.2</b> Parallelization</a></li>
<li class="chapter" data-level="18.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>18.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-12"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html"><i class="fa fa-check"></i><b>19</b> 416B Smart-seq2 dataset</a><ul>
<li class="chapter" data-level="19.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#introduction-2"><i class="fa fa-check"></i><b>19.1</b> Introduction</a></li>
<li class="chapter" data-level="19.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#analysis-code"><i class="fa fa-check"></i><b>19.2</b> Analysis code</a><ul>
<li class="chapter" data-level="19.2.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#data-loading"><i class="fa fa-check"></i><b>19.2.1</b> Data loading</a></li>
<li class="chapter" data-level="19.2.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#gene-annotation"><i class="fa fa-check"></i><b>19.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="19.2.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#quality-control-2"><i class="fa fa-check"></i><b>19.2.3</b> Quality control</a></li>
<li class="chapter" data-level="19.2.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#normalization-1"><i class="fa fa-check"></i><b>19.2.4</b> Normalization</a></li>
<li class="chapter" data-level="19.2.5" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#variance-modelling"><i class="fa fa-check"></i><b>19.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="19.2.6" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#batch-correction"><i class="fa fa-check"></i><b>19.2.6</b> Batch correction</a></li>
<li class="chapter" data-level="19.2.7" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>19.2.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="19.2.8" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#clustering-2"><i class="fa fa-check"></i><b>19.2.8</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#results"><i class="fa fa-check"></i><b>19.3</b> Results</a><ul>
<li class="chapter" data-level="19.3.1" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#quality-control-statistics"><i class="fa fa-check"></i><b>19.3.1</b> Quality control statistics</a></li>
<li class="chapter" data-level="19.3.2" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#normalization-2"><i class="fa fa-check"></i><b>19.3.2</b> Normalization</a></li>
<li class="chapter" data-level="19.3.3" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#variance-modelling-1"><i class="fa fa-check"></i><b>19.3.3</b> Variance modelling</a></li>
<li class="chapter" data-level="19.3.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#clustering-3"><i class="fa fa-check"></i><b>19.3.4</b> Clustering</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="b-smart-seq2-dataset.html"><a href="b-smart-seq2-dataset.html#references-7"><i class="fa fa-check"></i><b>19.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html"><i class="fa fa-check"></i><b>20</b> Human pancreas Smart-seq2 dataset</a><ul>
<li class="chapter" data-level="20.1" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#introduction-3"><i class="fa fa-check"></i><b>20.1</b> Introduction</a></li>
<li class="chapter" data-level="20.2" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#analysis-code-1"><i class="fa fa-check"></i><b>20.2</b> Analysis code</a><ul>
<li class="chapter" data-level="20.2.1" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#data-loading-1"><i class="fa fa-check"></i><b>20.2.1</b> Data loading</a></li>
<li class="chapter" data-level="20.2.2" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#gene-annotation-1"><i class="fa fa-check"></i><b>20.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="20.2.3" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#sample-annotation"><i class="fa fa-check"></i><b>20.2.3</b> Sample annotation</a></li>
<li class="chapter" data-level="20.2.4" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#quality-control-3"><i class="fa fa-check"></i><b>20.2.4</b> Quality control</a></li>
<li class="chapter" data-level="20.2.5" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#normalization-3"><i class="fa fa-check"></i><b>20.2.5</b> Normalization</a></li>
<li class="chapter" data-level="20.2.6" data-path="human-pancreas-smart-seq2-dataset.html"><a href="human-pancreas-smart-seq2-dataset.html#variance-modelling-2"><i class="fa fa-check"></i><b>20.2.6</b> Variance modelling</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html"><i class="fa fa-check"></i><b>21</b> Human pancreas CEL-seq2 dataset</a><ul>
<li class="chapter" data-level="21.1" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#introduction-4"><i class="fa fa-check"></i><b>21.1</b> Introduction</a></li>
<li class="chapter" data-level="21.2" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#analysis-code-2"><i class="fa fa-check"></i><b>21.2</b> Analysis code</a><ul>
<li class="chapter" data-level="21.2.1" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#data-loading-2"><i class="fa fa-check"></i><b>21.2.1</b> Data loading</a></li>
<li class="chapter" data-level="21.2.2" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#gene-annotation-2"><i class="fa fa-check"></i><b>21.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="21.2.3" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#quality-control-4"><i class="fa fa-check"></i><b>21.2.3</b> Quality control</a></li>
<li class="chapter" data-level="21.2.4" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#normalization-4"><i class="fa fa-check"></i><b>21.2.4</b> Normalization</a></li>
<li class="chapter" data-level="21.2.5" data-path="human-pancreas-cel-seq2-dataset.html"><a href="human-pancreas-cel-seq2-dataset.html#variance-modelling-3"><i class="fa fa-check"></i><b>21.2.5</b> Variance modelling</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html"><i class="fa fa-check"></i><b>22</b> Chimeric embryo 10X dataset</a><ul>
<li class="chapter" data-level="22.1" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#introduction-5"><i class="fa fa-check"></i><b>22.1</b> Introduction</a></li>
<li class="chapter" data-level="22.2" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#analysis-code-3"><i class="fa fa-check"></i><b>22.2</b> Analysis code</a><ul>
<li class="chapter" data-level="22.2.1" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#data-loading-3"><i class="fa fa-check"></i><b>22.2.1</b> Data loading</a></li>
<li class="chapter" data-level="22.2.2" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#feature-annotation"><i class="fa fa-check"></i><b>22.2.2</b> Feature annotation</a></li>
<li class="chapter" data-level="22.2.3" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#quality-control-5"><i class="fa fa-check"></i><b>22.2.3</b> Quality control</a></li>
<li class="chapter" data-level="22.2.4" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#normalization-5"><i class="fa fa-check"></i><b>22.2.4</b> Normalization</a></li>
<li class="chapter" data-level="22.2.5" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#variance-modelling-4"><i class="fa fa-check"></i><b>22.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="22.2.6" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#merging"><i class="fa fa-check"></i><b>22.2.6</b> Merging</a></li>
<li class="chapter" data-level="22.2.7" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#clustering-4"><i class="fa fa-check"></i><b>22.2.7</b> Clustering</a></li>
<li class="chapter" data-level="22.2.8" data-path="chimeric-embryo-10x-dataset.html"><a href="chimeric-embryo-10x-dataset.html#dimensionality-reduction-4"><i class="fa fa-check"></i><b>22.2.8</b> Dimensionality reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html"><i class="fa fa-check"></i><b>23</b> Mammary gland dataset</a><ul>
<li class="chapter" data-level="23.1" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#introduction-6"><i class="fa fa-check"></i><b>23.1</b> Introduction</a></li>
<li class="chapter" data-level="23.2" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#analysis-code-4"><i class="fa fa-check"></i><b>23.2</b> Analysis code</a><ul>
<li class="chapter" data-level="23.2.1" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#data-loading-4"><i class="fa fa-check"></i><b>23.2.1</b> Data loading</a></li>
<li class="chapter" data-level="23.2.2" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#gene-annotation-3"><i class="fa fa-check"></i><b>23.2.2</b> Gene annotation</a></li>
<li class="chapter" data-level="23.2.3" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#quality-control-6"><i class="fa fa-check"></i><b>23.2.3</b> Quality control</a></li>
<li class="chapter" data-level="23.2.4" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#normalization-6"><i class="fa fa-check"></i><b>23.2.4</b> Normalization</a></li>
<li class="chapter" data-level="23.2.5" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#variance-modelling-5"><i class="fa fa-check"></i><b>23.2.5</b> Variance modelling</a></li>
<li class="chapter" data-level="23.2.6" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#dimensionality-reduction-5"><i class="fa fa-check"></i><b>23.2.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="23.2.7" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#clustering-5"><i class="fa fa-check"></i><b>23.2.7</b> Clustering</a></li>
<li class="chapter" data-level="23.2.8" data-path="mammary-gland-dataset.html"><a href="mammary-gland-dataset.html#marker-detection"><i class="fa fa-check"></i><b>23.2.8</b> Marker detection</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="24" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>24</b> About the Data</a><ul>
<li class="chapter" data-level="24.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>24.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="24.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>24.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>25</b> About the Contributors</a><ul>
<li class="chapter" data-level="25.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>25.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="25.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>25.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="25.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>25.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="25.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>25.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>26</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="integrating-datasets-1" class="section level1">
<h1><span class="header-section-number">Chapter 15</span> Integrating Datasets</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-7" class="section level2">
<h2><span class="header-section-number">15.1</span> Motivation</h2>
<p>Large single-cell RNA sequencing (scRNA-seq) projects usually need to generate data across multiple batches due to logistical constraints.
However, the processing of different batches is often subject to uncontrollable differences, e.g., changes in operator, differences in reagent quality.
This results in systematic differences in the observed expression in cells from different batches, which we refer to as “batch effects”.
Batch effects are problematic as they can be major drivers of heterogeneity in the data, masking the relevant biological differences and complicating interpretation of the results.</p>
<p>Computational correction of these effects is critical for eliminating batch-to-batch variation, allowing data across multiple batches to be combined for common downstream analysis.
However, existing methods based on linear models <span class="citation">(Ritchie et al. <a href="#ref-ritchie2015limma">2015</a>; Leek et al. <a href="#ref-leek2012sva">2012</a>)</span> assume that the composition of cell populations are either known or the same across batches.
To overcome these limitations, bespoke methods have been developed for batch correction of single-cell data <span class="citation">(Haghverdi et al. <a href="#ref-haghverdi2018batch">2018</a>; Butler et al. <a href="#ref-butler2018integrating">2018</a>; Lin et al. <a href="#ref-lin2019scmerge">2019</a>)</span> that do not require <em>a priori</em> knowledge about the composition of the population.
This allows them to be used in workflows for exploratory analyses of scRNA-seq data where such knowledge is usually unavailable.</p>
</div>
<div id="setting-up-the-data-1" class="section level2">
<h2><span class="header-section-number">15.2</span> Setting up the data</h2>
<p>To demonstrate, we will use two separate 10X Genomics PBMC datasets generated in two different batches.
Each dataset was obtained from the <em><a href="https://bioconductor.org/packages/3.10/TENxPBMCData">TENxPBMCData</a></em> package and separately subjected to basic processing steps.
Separate processing prior to the batch correction step is more convenient, scalable and (on occasion) more reliable.
For example, outlier-based QC on the cells is more effective when performed within a batch (Chapter ???).
The same can also be said for trend fitting when modelling the mean-variance relationship (Section ???).</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb396-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb396-2" data-line-number="2"><span class="kw">library</span>(TENxPBMCData)</a>
<a class="sourceLine" id="cb396-3" data-line-number="3">pbmc3k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc3k&#39;</span>)</a>
<a class="sourceLine" id="cb396-4" data-line-number="4"></a>
<a class="sourceLine" id="cb396-5" data-line-number="5">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb396-6" data-line-number="6">is.mito &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;MT&quot;</span>, <span class="kw">rowData</span>(pbmc3k)<span class="op">$</span>Symbol_TENx)</a>
<a class="sourceLine" id="cb396-7" data-line-number="7"></a>
<a class="sourceLine" id="cb396-8" data-line-number="8"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb396-9" data-line-number="9">qc.stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(pbmc3k, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</a>
<a class="sourceLine" id="cb396-10" data-line-number="10">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(qc.stats<span class="op">$</span>subsets_Mito_percent, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb396-11" data-line-number="11">pbmc3k &lt;-<span class="st"> </span>pbmc3k[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb396-12" data-line-number="12"></a>
<a class="sourceLine" id="cb396-13" data-line-number="13">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb396-14" data-line-number="14">pbmc3k &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(pbmc3k)</a>
<a class="sourceLine" id="cb396-15" data-line-number="15"></a>
<a class="sourceLine" id="cb396-16" data-line-number="16">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb396-17" data-line-number="17"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb396-18" data-line-number="18">dec3k &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(pbmc3k)</a>
<a class="sourceLine" id="cb396-19" data-line-number="19"></a>
<a class="sourceLine" id="cb396-20" data-line-number="20">### feature-selection <span class="al">###</span></a>
<a class="sourceLine" id="cb396-21" data-line-number="21">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">which</span>(dec3k<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb396-22" data-line-number="22"></a>
<a class="sourceLine" id="cb396-23" data-line-number="23">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb396-24" data-line-number="24"><span class="co"># Using randomized SVD, which is more efficient for file-backed matrices.</span></a>
<a class="sourceLine" id="cb396-25" data-line-number="25"><span class="kw">set.seed</span>(<span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb396-26" data-line-number="26">pbmc3k &lt;-<span class="st"> </span><span class="kw">runPCA</span>(pbmc3k, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">ncomponents=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb396-27" data-line-number="27">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>())</a>
<a class="sourceLine" id="cb396-28" data-line-number="28"></a>
<a class="sourceLine" id="cb396-29" data-line-number="29"><span class="kw">set.seed</span>(<span class="dv">100000</span>)</a>
<a class="sourceLine" id="cb396-30" data-line-number="30">pbmc3k &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(pbmc3k, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb396-31" data-line-number="31"></a>
<a class="sourceLine" id="cb396-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000000</span>)</a>
<a class="sourceLine" id="cb396-33" data-line-number="33">pbmc3k &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(pbmc3k, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb396-34" data-line-number="34"></a>
<a class="sourceLine" id="cb396-35" data-line-number="35">### clustering <span class="al">###</span></a>
<a class="sourceLine" id="cb396-36" data-line-number="36">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(pbmc3k, <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb396-37" data-line-number="37">clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb396-38" data-line-number="38">pbmc3k<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb397-1" data-line-number="1">pbmc3k</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 32738 2609 
## metadata(0):
## assays(2): counts logcounts
## rownames(32738): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000215616 ENSG00000215611
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(12): Sample Barcode ... Date_published cluster
## reducedDimNames(3): PCA TSNE UMAP
## spikeNames(0):
## altExpNames(0):</code></pre>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb399-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb399-2" data-line-number="2"><span class="kw">library</span>(TENxPBMCData)</a>
<a class="sourceLine" id="cb399-3" data-line-number="3">pbmc4k &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)</a>
<a class="sourceLine" id="cb399-4" data-line-number="4"></a>
<a class="sourceLine" id="cb399-5" data-line-number="5">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb399-6" data-line-number="6">is.mito &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;MT&quot;</span>, <span class="kw">rowData</span>(pbmc4k)<span class="op">$</span>Symbol_TENx)</a>
<a class="sourceLine" id="cb399-7" data-line-number="7"></a>
<a class="sourceLine" id="cb399-8" data-line-number="8"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb399-9" data-line-number="9">qc.stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(pbmc4k, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</a>
<a class="sourceLine" id="cb399-10" data-line-number="10">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(qc.stats<span class="op">$</span>subsets_Mito_percent, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb399-11" data-line-number="11">pbmc4k &lt;-<span class="st"> </span>pbmc4k[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb399-12" data-line-number="12"></a>
<a class="sourceLine" id="cb399-13" data-line-number="13">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb399-14" data-line-number="14">pbmc4k &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(pbmc4k)</a>
<a class="sourceLine" id="cb399-15" data-line-number="15"></a>
<a class="sourceLine" id="cb399-16" data-line-number="16">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb399-17" data-line-number="17"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb399-18" data-line-number="18">dec4k &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(pbmc4k)</a>
<a class="sourceLine" id="cb399-19" data-line-number="19"></a>
<a class="sourceLine" id="cb399-20" data-line-number="20">### feature-selection <span class="al">###</span></a>
<a class="sourceLine" id="cb399-21" data-line-number="21">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">which</span>(dec4k<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb399-22" data-line-number="22"></a>
<a class="sourceLine" id="cb399-23" data-line-number="23">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb399-24" data-line-number="24"><span class="co"># Using randomized SVD, which is more efficient for file-backed matrices.</span></a>
<a class="sourceLine" id="cb399-25" data-line-number="25"><span class="kw">set.seed</span>(<span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb399-26" data-line-number="26">pbmc4k &lt;-<span class="st"> </span><span class="kw">runPCA</span>(pbmc4k, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">ncomponents=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb399-27" data-line-number="27">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>())</a>
<a class="sourceLine" id="cb399-28" data-line-number="28"></a>
<a class="sourceLine" id="cb399-29" data-line-number="29"><span class="kw">set.seed</span>(<span class="dv">100000</span>)</a>
<a class="sourceLine" id="cb399-30" data-line-number="30">pbmc4k &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(pbmc4k, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb399-31" data-line-number="31"></a>
<a class="sourceLine" id="cb399-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000000</span>)</a>
<a class="sourceLine" id="cb399-33" data-line-number="33">pbmc4k &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(pbmc4k, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb399-34" data-line-number="34"></a>
<a class="sourceLine" id="cb399-35" data-line-number="35">### clustering <span class="al">###</span></a>
<a class="sourceLine" id="cb399-36" data-line-number="36">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(pbmc4k, <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb399-37" data-line-number="37">clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb399-38" data-line-number="38">pbmc4k<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb400-1" data-line-number="1">pbmc4k</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 4182 
## metadata(0):
## assays(2): counts logcounts
## rownames(33694): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000277475 ENSG00000268674
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(12): Sample Barcode ... Date_published cluster
## reducedDimNames(3): PCA TSNE UMAP
## spikeNames(0):
## altExpNames(0):</code></pre>
<p>To prepare for the batch correction:</p>
<ol style="list-style-type: decimal">
<li><p>We subset all batches to the common “universe” of features.
In this case, it is straightforward as both batches use Ensembl gene annotation<a href="#fn22" class="footnote-ref" id="fnref22"><sup>22</sup></a>.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb402-1" data-line-number="1">universe &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(pbmc3k), <span class="kw">rownames</span>(pbmc4k))</a>
<a class="sourceLine" id="cb402-2" data-line-number="2"><span class="kw">length</span>(universe)</a></code></pre></div>
<pre><code>## [1] 31232</code></pre></li>
<li><p>We rescale each batch to adjust for differences in sequencing depth between batches.
The <code>multiBatchNorm()</code> function recomputes log-normalized expression values after adjusting the size factors for systematic differences in coverage between <code>SingleCellExperiment</code> objects.
(Size factors only remove biases between cells <em>within</em> a single batch.)
This improves the quality of the correction by removing one aspect of the technical differences between batches.</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb404-1" data-line-number="1"><span class="kw">library</span>(batchelor)</a>
<a class="sourceLine" id="cb404-2" data-line-number="2">rescaled &lt;-<span class="st"> </span><span class="kw">multiBatchNorm</span>(pbmc3k[universe,], pbmc4k[universe,])</a>
<a class="sourceLine" id="cb404-3" data-line-number="3">pbmc3k &lt;-<span class="st"> </span>rescaled[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb404-4" data-line-number="4">pbmc4k &lt;-<span class="st"> </span>rescaled[[<span class="dv">2</span>]]</a></code></pre></div></li>
<li><p>We perform feature selection by averaging the variance components across all batches with the <code>combineVar()</code> function.
We use the average as it is responsive to batch-specific HVGs while still preserving the within-batch ranking of genes.
This allows us to use the same strategies described in Chapter ??? to select genes of interest.
For example, if we were to retain all genes with positive biological components, convergence of the average towards zero for non-HVGs ensures that the expected number of retained genes is stable with more batches.
In contrast, approaches based on taking the intersection or union of HVGs across batches become increasingly conservative or liberal, respectively, with an increasing number of batches.</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb405-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb405-2" data-line-number="2">combined.dec &lt;-<span class="st"> </span><span class="kw">combineVar</span>(dec3k[universe,], dec4k[universe,])</a>
<a class="sourceLine" id="cb405-3" data-line-number="3">chosen.hvgs &lt;-<span class="st"> </span>combined.dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb405-4" data-line-number="4">combined.dec[,<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</a></code></pre></div>
<pre><code>## DataFrame with 31232 rows and 6 columns
##                                 mean                total
##                            &lt;numeric&gt;            &lt;numeric&gt;
## ENSG00000243485                    0                    0
## ENSG00000237613                    0                    0
## ENSG00000186092                    0                    0
## ENSG00000238009  0.00104386992482179  0.00106138029098745
## ENSG00000239945 0.000258798374840956 0.000281948894897105
## ...                              ...                  ...
## ENSG00000212907    0.535233858583492    0.437804768744041
## ENSG00000198886     3.01692549039158    0.617031726421764
## ENSG00000198786     1.54198723959989    0.617834240701208
## ENSG00000198695    0.127831201202011    0.129776187258376
## ENSG00000198727     2.56240219438806    0.717304616025944
##                                 tech                  bio
##                            &lt;numeric&gt;            &lt;numeric&gt;
## ENSG00000243485                    0                    0
## ENSG00000237613                    0                    0
## ENSG00000186092                    0                    0
## ENSG00000238009  0.00106590149862648 -4.5212076390281e-06
## ENSG00000239945 0.000264260487890133  1.7688407006972e-05
## ...                              ...                  ...
## ENSG00000212907    0.408040682015157   0.0297640867288842
## ENSG00000198886    0.596769334163596   0.0202623922581677
## ENSG00000198786    0.628483569105677  -0.0106493284044691
## ENSG00000198695    0.129278188915683 0.000497998342692874
## ENSG00000198727    0.691490878290747    0.025813737735197
##                           p.value               FDR
##                         &lt;numeric&gt;         &lt;numeric&gt;
## ENSG00000243485                NA                NA
## ENSG00000237613                NA                NA
## ENSG00000186092                NA                NA
## ENSG00000238009 0.512246219727119 0.834032704052722
## ENSG00000239945 0.314021257793438 0.834032704052722
## ...                           ...               ...
## ENSG00000212907  0.40460723763713 0.834032704052722
## ENSG00000198886 0.215888744541045 0.834032704052722
## ENSG00000198786 0.642327056158103 0.834072154688035
## ENSG00000198695 0.585074623235346 0.834032704052722
## ENSG00000198727 0.419767773986605 0.834032704052722</code></pre>
<pre><code>## &lt;numeric&gt;

## 0

## 0

## 0

## 0.00104386992482179

## 0.000258798374840956

## ...

## 0.535233858583492

## 3.01692549039158

## 1.54198723959989

## 0.127831201202011

## 2.56240219438806

## &lt;numeric&gt;

## 0

## 0

## 0

## 0.00106138029098745

## 0.000281948894897105

## ...

## 0.437804768744041

## 0.617031726421764

## 0.617834240701208

## 0.129776187258376

## 0.717304616025944

## &lt;numeric&gt;

## 0

## 0

## 0

## 0.00106590149862648

## 0.000264260487890133

## ...

## 0.408040682015157

## 0.596769334163596

## 0.628483569105677

## 0.129278188915683

## 0.691490878290747

## &lt;numeric&gt;

## 0

## 0

## 0

## -4.5212076390281e-06

## 1.7688407006972e-05

## ...

## 0.0297640867288842

## 0.0202623922581677

## -0.0106493284044691

## 0.000497998342692874

## 0.025813737735197

## &lt;numeric&gt;

## NA

## NA

## NA

## 0.512246219727119

## 0.314021257793438

## ...

## 0.40460723763713

## 0.215888744541045

## 0.642327056158103

## 0.585074623235346

## 0.419767773986605

## &lt;numeric&gt;

## NA

## NA

## NA

## 0.834032704052722

## 0.834032704052722

## ...

## 0.834032704052722

## 0.834032704052722

## 0.834072154688035

## 0.834032704052722

## 0.834032704052722</code></pre></li>
</ol>
</div>
<div id="diagnosing-batch-effects" class="section level2">
<h2><span class="header-section-number">15.3</span> Diagnosing batch effects</h2>
<p>Before we actually perform any correction, it is worth examining whether there is any batch effect in this dataset.
We combine the two <code>SingleCellExperiment</code>s and perform a PCA on the log-expression values for all genes with positive (average) biological components.</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb408-1" data-line-number="1"><span class="co"># Synchronizing the metadata for cbind()ing.</span></a>
<a class="sourceLine" id="cb408-2" data-line-number="2"><span class="kw">rowData</span>(pbmc3k) &lt;-<span class="st"> </span><span class="kw">rowData</span>(pbmc4k)</a>
<a class="sourceLine" id="cb408-3" data-line-number="3">pbmc3k<span class="op">$</span>batch &lt;-<span class="st"> &quot;3k&quot;</span></a>
<a class="sourceLine" id="cb408-4" data-line-number="4">pbmc4k<span class="op">$</span>batch &lt;-<span class="st"> &quot;4k&quot;</span></a>
<a class="sourceLine" id="cb408-5" data-line-number="5">combined &lt;-<span class="st"> </span><span class="kw">cbind</span>(pbmc3k, pbmc4k)</a>
<a class="sourceLine" id="cb408-6" data-line-number="6"></a>
<a class="sourceLine" id="cb408-7" data-line-number="7"><span class="co"># Using RandomParam() as it is more efficient for file-backed matrices.</span></a>
<a class="sourceLine" id="cb408-8" data-line-number="8"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb408-9" data-line-number="9"><span class="kw">set.seed</span>(<span class="dv">0010101010</span>)</a>
<a class="sourceLine" id="cb408-10" data-line-number="10">combined &lt;-<span class="st"> </span><span class="kw">runPCA</span>(combined, <span class="dt">subset_row=</span>chosen.hvgs,</a>
<a class="sourceLine" id="cb408-11" data-line-number="11">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>())</a></code></pre></div>
<p>We use graph-based clustering on the components to obtain a summary of the population structure.
As our two PBMC populations should be replicates, each cluster should ideally consist of cells from both batches.
However, we instead see clusters that are comprised of cells from a single batch.
This indicates that cells of the same type are artificially separated due to technical differences between batches.</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb409-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb409-2" data-line-number="2">snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(combined, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb409-3" data-line-number="3">clusters &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb409-4" data-line-number="4">tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters, <span class="dt">Batch=</span>combined<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb409-5" data-line-number="5">tab</a></code></pre></div>
<pre><code>##        Batch
## Cluster   3k   4k
##      1     0  126
##      2    12  459
##      3     1  776
##      4     0 1310
##      5   500    0
##      6     0  536
##      7     0  606
##      8  1296    0
##      9     0  176
##      10    0   54
##      11  149    0
##      12   30    1
##      13    0   89
##      14  131    0
##      15  342    0
##      16    1   10
##      17  134    0
##      18   11    3
##      19    2   36</code></pre>
<p>We can also visualize the corrected coordinates using a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="integrating-datasets-1.html#fig:tsne-pbmc-uncorrected">15.1</a>).
The strong separation between cells from different batches is consistent with the clustering results.</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb411-1" data-line-number="1">combined &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(combined, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb411-2" data-line-number="2"><span class="kw">plotTSNE</span>(combined, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-uncorrected"></span>
<img src="P2_W11.data-integration_files/figure-html/tsne-pbmc-uncorrected-1.png" alt="$t$-SNE plot of the PBMC datasets without any batch correction. Each point is a cell that is colored according to its batch of origin." width="672" />
<p class="caption">
Figure 15.1: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets without any batch correction. Each point is a cell that is colored according to its batch of origin.
</p>
</div>
<p>Of course, the other explanation for batch-specific clusters is that there are cell types that are unique to each batch.
The degree of intermingling of cells from different batches is not an effective diagnostic when the batches involved might actually contain unique cell subpopulations (which is not a consideration in the PBMC dataset, but the same cannot be said in general).
If a cluster only contains cells from a single batch, one can always debate whether that is caused by a failure of the correction method or if there is truly a batch-specific subpopulation.
For example, do batch-specific metabolic or differentiation states represent distinct subpopulations?
Or should they be merged together?
We will not attempt to answer this here, only noting that each batch correction algorithm will make different (and possibly inappropriate) decisions on what constitutes “shared” and “unique” populations.</p>
</div>
<div id="linear-regression" class="section level2">
<h2><span class="header-section-number">15.4</span> Linear regression</h2>
<p>Batch effects in bulk RNA sequencing studies are commonly removed with linear regression.
This involves fitting a linear model to each gene’s expression profile, setting the undesirable batch term to zero and recomputing the observations <em>sans</em> the batch effect, yielding a set of corrected expression values for downstream analyses.
Linear modelling is the basis of the <code>removeBatchEffect()</code> function from the <em><a href="https://bioconductor.org/packages/3.10/limma">limma</a></em> package <span class="citation">(Ritchie et al. <a href="#ref-ritchie2015limma">2015</a>)</span> as well the <code>comBat()</code> function from the <em><a href="https://bioconductor.org/packages/3.10/sva">sva</a></em> package <span class="citation">(Leek et al. <a href="#ref-leek2012sva">2012</a>)</span>.</p>
<p>To use this approach in a scRNA-seq context, we assume that the composition of cell subpopulations is the same across batches.
We also assume that the batch effect is additive, i.e., any batch-induced fold-change in expression is the same across different cell subpopulations for any given gene.
These are strong assumptions as batches derived from different individuals will naturally exhibit variation in cell type abundances and expression.
Nonetheless, they may be acceptable when dealing with batches that are technical replicates generated from the same population of cells.
Linear modelling can also accommodate situations where the composition is known <em>a priori</em> by including the cell type as a factor in the linear model, but this situation is even less common<a href="#fn23" class="footnote-ref" id="fnref23"><sup>23</sup></a>.</p>
<p>We use the <code>rescaleBatches()</code> function from the <em><a href="https://bioconductor.org/packages/3.10/batchelor">batchelor</a></em> package to remove the batch effect.
This is roughly equivalent to applying a linear regression to the log-expression values per gene, with some adjustments to improve performance and efficiency.
For each gene, the mean expression in each batch is scaled down until it is equal to the lowest mean across all batches.
We deliberately choose to scale all expression values down as this mitigates differences in variance when batches lie at different positions on the mean-variance trend.
(Specifically, the shrinkage effect of the pseudo-count is greater for smaller counts, suppressing any differences in variance across batches.)
An additional feature of <code>rescaleBatches()</code> is that it will preserve sparsity in the input matrix for greater efficiency, whereas other methods like <code>removeBatchEffect()</code> will always return a dense matrix.</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb412-1" data-line-number="1"><span class="kw">library</span>(batchelor)</a>
<a class="sourceLine" id="cb412-2" data-line-number="2">rescaled &lt;-<span class="st"> </span><span class="kw">rescaleBatches</span>(pbmc3k, pbmc4k)</a>
<a class="sourceLine" id="cb412-3" data-line-number="3">rescaled</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 31232 6791 
## metadata(0):
## assays(1): corrected
## rownames(31232): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000198695 ENSG00000198727
## rowData names(0):
## colnames: NULL
## colData names(1): batch
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):</code></pre>
<p>After clustering, we observe that most clusters consist of mixtures of cells from the two replicate batches, consistent with the removal of the batch effect.
This conclusion is supported by the apparent mixing of cells from different batches in Figure <a href="integrating-datasets-1.html#fig:tsne-pbmc-rescaled">15.2</a>.
However, at least one batch-specific cluster is still present, indicating that the correction is not entirely complete.
This is attributable to violation of one of the aforementioned assumptions, even in this simple case involving replicated batches.</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb414-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1010101010</span>)</a>
<a class="sourceLine" id="cb414-2" data-line-number="2">rescaled &lt;-<span class="st"> </span><span class="kw">runPCA</span>(rescaled, <span class="dt">subset_row=</span>chosen.hvgs,</a>
<a class="sourceLine" id="cb414-3" data-line-number="3">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(), <span class="dt">exprs_values=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb414-4" data-line-number="4"></a>
<a class="sourceLine" id="cb414-5" data-line-number="5">snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(rescaled, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb414-6" data-line-number="6">clusters.resc &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb414-7" data-line-number="7">tab.resc &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters.resc, <span class="dt">Batch=</span>rescaled<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb414-8" data-line-number="8">tab.resc</a></code></pre></div>
<pre><code>##        Batch
## Cluster   1   2
##      1  272 523
##      2  336 606
##      3  126 266
##      4  643 560
##      5   19  47
##      6   12   3
##      7  313   0
##      8    8  50
##      9   19  58
##      10  15  70
##      11 131 154
##      12  37 511
##      13  10  83
##      14 100 207
##      15 137   8
##      16  16  25
##      17 397 964
##      18   3  36
##      19   4   8
##      20  11   3</code></pre>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb416-1" data-line-number="1">rescaled &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(rescaled, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb416-2" data-line-number="2">rescaled<span class="op">$</span>batch &lt;-<span class="st"> </span><span class="kw">factor</span>(rescaled<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb416-3" data-line-number="3"><span class="kw">plotTSNE</span>(rescaled, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-rescaled"></span>
<img src="P2_W11.data-integration_files/figure-html/tsne-pbmc-rescaled-1.png" alt="$t$-SNE plot of the PBMC datasets after correction with `rescaleBatches()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">
Figure 15.2: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets after correction with <code>rescaleBatches()</code>. Each point represents a cell and is colored according to the batch of origin.
</p>
</div>
</div>
<div id="performing-mnn-correction" class="section level2">
<h2><span class="header-section-number">15.5</span> Performing MNN correction</h2>
<div id="application-to-the-pbmc-data" class="section level3">
<h3><span class="header-section-number">15.5.1</span> Application to the PBMC data</h3>
<p>Consider a cell <span class="math inline">\(a\)</span> in batch <span class="math inline">\(A\)</span>, and identify the cells in batch <span class="math inline">\(B\)</span> that are nearest neighbours to <span class="math inline">\(a\)</span> in the expression space defined by the selected features.
Repeat this for a cell <span class="math inline">\(b\)</span> in batch <span class="math inline">\(B\)</span>, identifying its nearest neighbours in <span class="math inline">\(A\)</span>.
Mutual nearest neighbours are pairs of cells from different batches that belong in each other’s set of nearest neighbours.
The reasoning is that MNN pairs represent cells from the same biological state prior to the application of a batch effect - see <span class="citation">Haghverdi et al. (<a href="#ref-haghverdi2018batch">2018</a>)</span> for full theoretical details.
Thus, the difference between cells in MNN pairs can be used as an estimate of the batch effect, the subtraction of which yields batch-corrected values.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.10/batchelor">batchelor</a></em> package provides an implementation of the MNN approach via the <code>fastMNN()</code> function.
(Unlike the MNN method described by <span class="citation">Haghverdi et al. (<a href="#ref-haghverdi2018batch">2018</a>)</span>, the <code>fastMNN()</code> function performs PCA to reduce the dimensions beforehand and speed up the downstream neighbor detection steps.)
We apply it to our two PBMC batches to remove the batch effect across the highly variable genes in <code>chosen</code>.
To reduce computational work and technical noise, all cells in all batches are projected into the low-dimensional space defined by the top <code>d</code> principal components.
Identification of MNNs and calculation of correction vectors are then performed in this low-dimensional space.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb417-1" data-line-number="1"><span class="co"># Using randomized SVD here, as this is faster than </span></a>
<a class="sourceLine" id="cb417-2" data-line-number="2"><span class="co"># irlba for file-backed matrices.</span></a>
<a class="sourceLine" id="cb417-3" data-line-number="3"><span class="kw">set.seed</span>(<span class="dv">1000101001</span>)</a>
<a class="sourceLine" id="cb417-4" data-line-number="4">mnn.out &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(pbmc3k, pbmc4k, <span class="dt">d=</span><span class="dv">50</span>, <span class="dt">k=</span><span class="dv">20</span>,</a>
<a class="sourceLine" id="cb417-5" data-line-number="5">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">RandomParam</span>(<span class="dt">deferred=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb417-6" data-line-number="6">mnn.out</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 31232 6791 
## metadata(1): merge.info
## assays(1): reconstructed
## rownames(31232): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000198695 ENSG00000198727
## rowData names(1): rotation
## colnames: NULL
## colData names(1): batch
## reducedDimNames(1): corrected
## spikeNames(0):
## altExpNames(0):</code></pre>
<p>The function returns a <code>SingleCellExperiment</code> object containing corrected values for downstream analyses like clustering or visualization.
Each column of <code>mnn.out</code> corresponds to a cell in one of the batches, while each row corresponds to an input gene in <code>chosen</code>.
The <code>batch</code> field in the column metadata contains a vector specifying the batch of origin of each cell.</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb419-1" data-line-number="1"><span class="kw">head</span>(mnn.out<span class="op">$</span>batch) </a></code></pre></div>
<pre><code>## [1] 1 1 1 1 1 1</code></pre>
<p>The <code>corrected</code> matrix in the <code>reducedDims()</code> contains the low-dimensional corrected coordinates for all cells, which we will use in place of the PCs in our downstream analyses.</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb421-1" data-line-number="1"><span class="kw">dim</span>(<span class="kw">reducedDim</span>(mnn.out, <span class="st">&quot;corrected&quot;</span>))</a></code></pre></div>
<pre><code>## [1] 6791   50</code></pre>
<p>The most relevant parameter for tuning <code>fastMNN()</code> is <code>k</code>, which specifies the number of nearest neighbours to consider when defining MNN pairs.
This can be interpreted as the minimum anticipated frequency of any shared cell type or state in each batch.
Increasing <code>k</code> will generally result in more aggressive merging as the algorithm is more generous in matching subpopulations across batches.
It can occasionally be desirable to increase <code>k</code> if one clearly sees that the same cell types are not being adequately merged across batches.</p>
</div>
<div id="correction-diagnostics" class="section level3">
<h3><span class="header-section-number">15.5.2</span> Correction diagnostics</h3>
<p>We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure.
If the batch effect is successfully corrected, clusters corresponding to shared cell types or states should contain cells from multiple batches.
We see that all clusters contain contributions from each batch after correction, consistent with our expectation that the two batches are replicates of each other.</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb423-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb423-2" data-line-number="2">snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(mnn.out, <span class="dt">use.dimred=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb423-3" data-line-number="3">clusters.mnn &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb423-4" data-line-number="4">tab.mnn &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters.mnn, <span class="dt">Batch=</span>mnn.out<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb423-5" data-line-number="5">tab.mnn</a></code></pre></div>
<pre><code>##        Batch
## Cluster    1    2
##      1   152  181
##      2   331  589
##      3   531 1205
##      4   277  502
##      5   283  597
##      6   617  579
##      7   203  181
##      8    13   51
##      9    17   69
##      10    7   18
##      11   11   16
##      12   18   61
##      13  131   86
##      14    3   36
##      15    4    8
##      16   11    3</code></pre>
<p>We can also visualize the corrected coordinates using a <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="integrating-datasets-1.html#fig:tsne-pbmc-corrected">15.3</a>).
The presence of visual clusters containing cells from both batches provides a comforting illusion that the correction was successful.</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb425-1" data-line-number="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb425-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">0010101010</span>)</a>
<a class="sourceLine" id="cb425-3" data-line-number="3">mnn.out &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(mnn.out, <span class="dt">use_dimred=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb425-4" data-line-number="4"></a>
<a class="sourceLine" id="cb425-5" data-line-number="5">mnn.out<span class="op">$</span>batch &lt;-<span class="st"> </span><span class="kw">factor</span>(mnn.out<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb425-6" data-line-number="6"><span class="kw">plotTSNE</span>(mnn.out, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-pbmc-corrected"></span>
<img src="P2_W11.data-integration_files/figure-html/tsne-pbmc-corrected-1.png" alt="$t$-SNE plot of the PBMC datasets after MNN correction. Each point is a cell that is colored according to its batch of origin." width="672" />
<p class="caption">
Figure 15.3: <span class="math inline">\(t\)</span>-SNE plot of the PBMC datasets after MNN correction. Each point is a cell that is colored according to its batch of origin.
</p>
</div>
<p>For <code>fastMNN()</code>, one useful diagnostic is the proportion of variance within each batch that is lost during MNN correction.
Specifically, this refers to the within-batch variance that is removed during orthogonalization with respect to the average correction vector at each merge step.
This is returned via the <code>lost.var</code> field in the metadata of <code>mnn.out</code>, which contains a matrix of the variance lost in each batch (column) at each merge step (row).</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb426-1" data-line-number="1"><span class="kw">metadata</span>(mnn.out)<span class="op">$</span>merge.info<span class="op">$</span>lost.var</a></code></pre></div>
<pre><code>##          [,1]     [,2]
## [1,] 0.004368 0.003173</code></pre>
<p>Large proportions of lost variance suggest that correction is removing genuine biological heterogeneity.
This would occur due to violations of the assumption of orthogonality between the batch effect and the biological subspace <span class="citation">(Haghverdi et al. <a href="#ref-haghverdi2018batch">2018</a>)</span>.
In this case, the proportion of lost variance is small, indicating that non-orthogonality is not a major concern.</p>
</div>
</div>
<div id="preserving-biological-heterogeneity" class="section level2">
<h2><span class="header-section-number">15.6</span> Preserving biological heterogeneity</h2>
<p>Another useful diagnostic check is to compare the clustering within each batch to the clustering of the merged data.
Accurate data integration should preserve variance within each batch as there should be nothing to remove between cells in the same batch.
This check complements the previously mentioned diagnostics that only focus on the removal of differences between batches.
Specifically, it protects us against cases where the correction method simply aggregates all cells together, which would achieve perfect mixing but also discard the biological heterogeneity of interest.</p>
<p>Ideally, we should see a many-to-1 mapping where the across-batch clustering is nested inside the within-batch clusterings.
This indicates that any within-batch structure was preserved after correction while acknowledging that greater resolution is possible with more cells.
In practice, more discrepancies can be expected even when the correction is perfect, due to the existence of closely related clusters that were arbitrarily separated in the within-batch clustering.
As a general rule, we can be satisfied with the correction if the vast majority of entries of the <code>table()</code>s below are zero, though this may depend on whether specific clusters of interest are gained or lost.</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb428-1" data-line-number="1"><span class="co"># For the first batch.</span></a>
<a class="sourceLine" id="cb428-2" data-line-number="2"><span class="kw">table</span>(<span class="dt">New=</span>clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>], <span class="dt">Old=</span>pbmc3k<span class="op">$</span>cluster)</a></code></pre></div>
<pre><code>##     Old
## New    1   2   3   4   5   6   7   8   9
##   1    0   1   0   0   0 151   0   0   0
##   2    0   0   3   0   0   0   0 328   0
##   3    0   8  17 506   0   0   0   0   0
##   4    0 273   0   3   0   1   0   0   0
##   5  281   0   0   0   0   0   2   0   0
##   6    0  46 475  96   0   0   0   0   0
##   7  179   0   0   0   0   0  24   0   0
##   8    9   0   0   0   0   0   4   0   0
##   9    1   0   0   0  16   0   0   0   0
##   10   0   1   1   1   0   0   0   4   0
##   11   0   0  11   0   0   0   0   0   0
##   12   0   0   0   0  18   0   0   0   0
##   13   1   0   0   0   0   0 130   0   0
##   14   0   0   3   0   0   0   0   0   0
##   15   0   0   0   4   0   0   0   0   0
##   16   0   0   0   0   0   0   0   0  11</code></pre>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb430-1" data-line-number="1"><span class="co"># For the second batch.</span></a>
<a class="sourceLine" id="cb430-2" data-line-number="2"><span class="kw">table</span>(<span class="dt">New=</span>clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">2</span>], <span class="dt">Old=</span>pbmc4k<span class="op">$</span>cluster)</a></code></pre></div>
<pre><code>##     Old
## New    1   2   3   4   5   6   7   8   9  10  11  12
##   1    0   0   0   0   0 181   0   0   0   0   0   0
##   2    0   0   0   0 221   0 368   0   0   0   0   0
##   3   20   0   0   3   0   0   0   0 997 185   0   0
##   4    0   1   0 496   0   5   0   0   0   0   0   0
##   5    0 594   0   0   0   0   0   3   0   0   0   0
##   6  458   0   0  66   0   0   0   0  39  16   0   0
##   7    0 179   1   0   0   0   0   1   0   0   0   0
##   8    0   6   0   0   0   0   0  45   0   0   0   0
##   9    0   4  65   0   0   0   0   0   0   0   0   0
##   10   0   0   0   1   4   1  12   0   0   0   0   0
##   11  14   0   0   1   1   0   0   0   0   0   0   0
##   12   0   0  61   0   0   0   0   0   0   0   0   0
##   13   0   9   0   0   0   0   0   0   0   0  77   0
##   14   0   0   0   0   0   0   0   0   0   0   0  36
##   15   0   0   0   0   0   0   8   0   0   0   0   0
##   16   0   3   0   0   0   0   0   0   0   0   0   0</code></pre>
<p>We can summarize the agreement between clusterings by computing the Rand index.
This provides a simple metric that we can use to assess the preservation of variation by different correction methods.
Larger Rand indices are more desirable, though this must be balanced against the ability of each method to actually remove the batch effect.</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb432-1" data-line-number="1"><span class="kw">library</span>(fossil)</a>
<a class="sourceLine" id="cb432-2" data-line-number="2"><span class="kw">rand.index</span>(<span class="kw">as.integer</span>(clusters.mnn[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb432-3" data-line-number="3">    <span class="kw">as.integer</span>(pbmc3k<span class="op">$</span>cluster))</a></code></pre></div>
<pre><code>## [1] 0.9297</code></pre>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb434-1" data-line-number="1"><span class="kw">rand.index</span>(<span class="kw">as.integer</span>(clusters.resc[rescaled<span class="op">$</span>batch<span class="op">==</span><span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb434-2" data-line-number="2">    <span class="kw">as.integer</span>(pbmc3k<span class="op">$</span>cluster))</a></code></pre></div>
<pre><code>## [1] 0.9102</code></pre>
</div>
<div id="application-to-a-pancreas-dataset" class="section level2">
<h2><span class="header-section-number">15.7</span> Application to a pancreas dataset</h2>
<p>We perform another demonstration with two human pancreas CEL-seq(2) datasets <span class="citation">(Muraro et al. <a href="#ref-muraro2016singlecell">2016</a>; Grun et al. <a href="#ref-grun2016denovo">2016</a>)</span>.
This is a more challenging application than the PBMC dataset as it involves different patients and protocols.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb436-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb436-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb436-3" data-line-number="3">sce.grun &lt;-<span class="st"> </span><span class="kw">GrunPancreasData</span>()</a>
<a class="sourceLine" id="cb436-4" data-line-number="4"></a>
<a class="sourceLine" id="cb436-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb436-6" data-line-number="6"><span class="co"># Converting back to Ensembl identifiers.</span></a>
<a class="sourceLine" id="cb436-7" data-line-number="7"><span class="kw">library</span>(org.Hs.eg.db)</a>
<a class="sourceLine" id="cb436-8" data-line-number="8">gene.symb &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;__chr.*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(sce.grun))</a>
<a class="sourceLine" id="cb436-9" data-line-number="9">gene.ids &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Hs.eg.db, <span class="dt">keys=</span>gene.symb, </a>
<a class="sourceLine" id="cb436-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;ENSEMBL&quot;</span>)</a>
<a class="sourceLine" id="cb436-11" data-line-number="11"></a>
<a class="sourceLine" id="cb436-12" data-line-number="12"><span class="co"># Removing duplicated genes or genes without Ensembl IDs.</span></a>
<a class="sourceLine" id="cb436-13" data-line-number="13">keep &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(gene.ids) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(gene.ids)</a>
<a class="sourceLine" id="cb436-14" data-line-number="14">sce.grun &lt;-<span class="st"> </span>sce.grun[keep,]</a>
<a class="sourceLine" id="cb436-15" data-line-number="15"><span class="kw">rownames</span>(sce.grun) &lt;-<span class="st"> </span>gene.ids[keep]</a>
<a class="sourceLine" id="cb436-16" data-line-number="16"></a>
<a class="sourceLine" id="cb436-17" data-line-number="17">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb436-18" data-line-number="18"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb436-19" data-line-number="19">QC &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.grun)</a>
<a class="sourceLine" id="cb436-20" data-line-number="20">filter &lt;-<span class="st"> </span><span class="kw">quickCellQC</span>(QC, <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>, <span class="dt">nmads=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb436-21" data-line-number="21">sce.grun &lt;-<span class="st"> </span>sce.grun[,<span class="op">!</span>filter<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb436-22" data-line-number="22"></a>
<a class="sourceLine" id="cb436-23" data-line-number="23">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb436-24" data-line-number="24"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb436-25" data-line-number="25"><span class="kw">set.seed</span>(<span class="dv">1000</span>) <span class="co"># for irlba. </span></a>
<a class="sourceLine" id="cb436-26" data-line-number="26">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.grun)</a>
<a class="sourceLine" id="cb436-27" data-line-number="27">sce.grun &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.grun, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb436-28" data-line-number="28">sce.grun &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.grun)</a>
<a class="sourceLine" id="cb436-29" data-line-number="29"></a>
<a class="sourceLine" id="cb436-30" data-line-number="30">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb436-31" data-line-number="31"><span class="co"># Blocking on a combined plate and donor factor.</span></a>
<a class="sourceLine" id="cb436-32" data-line-number="32">block &lt;-<span class="st"> </span><span class="kw">paste0</span>(sce.grun<span class="op">$</span>sample, <span class="st">&quot;_&quot;</span>, sce.grun<span class="op">$</span>donor)</a>
<a class="sourceLine" id="cb436-33" data-line-number="33">dec.grun &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sce.grun, <span class="dt">spikes=</span><span class="st">&quot;ERCC&quot;</span>, <span class="dt">block=</span>block)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb437-1" data-line-number="1">sce.grun</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 17692 1290 
## metadata(0):
## assays(2): counts logcounts
## rownames(17692): ENSG00000268895 ENSG00000121410 ...
##   ENSG00000074755 ENSG00000036549
## rowData names(2): symbol chr
## colnames(1290): D2ex_1 D2ex_2 ... D17TGFB_94 D17TGFB_95
## colData names(2): donor sample
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(1): ERCC</code></pre>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb439-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb439-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb439-3" data-line-number="3">sce.muraro &lt;-<span class="st"> </span><span class="kw">MuraroPancreasData</span>()</a>
<a class="sourceLine" id="cb439-4" data-line-number="4"></a>
<a class="sourceLine" id="cb439-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb439-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb439-7" data-line-number="7">edb &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73881&quot;</span>]]</a>
<a class="sourceLine" id="cb439-8" data-line-number="8">gene.symb &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;__chr.*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(sce.muraro))</a>
<a class="sourceLine" id="cb439-9" data-line-number="9">gene.ids &lt;-<span class="st"> </span><span class="kw">mapIds</span>(edb, <span class="dt">keys=</span>gene.symb, </a>
<a class="sourceLine" id="cb439-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb439-11" data-line-number="11"></a>
<a class="sourceLine" id="cb439-12" data-line-number="12"><span class="co"># Removing duplicated genes or genes without Ensembl IDs.</span></a>
<a class="sourceLine" id="cb439-13" data-line-number="13">keep &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(gene.ids) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(gene.ids)</a>
<a class="sourceLine" id="cb439-14" data-line-number="14">sce.muraro &lt;-<span class="st"> </span>sce.muraro[keep,]</a>
<a class="sourceLine" id="cb439-15" data-line-number="15"><span class="kw">rownames</span>(sce.muraro) &lt;-<span class="st"> </span>gene.ids[keep]</a>
<a class="sourceLine" id="cb439-16" data-line-number="16"></a>
<a class="sourceLine" id="cb439-17" data-line-number="17">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb439-18" data-line-number="18"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb439-19" data-line-number="19">qcstats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.muraro)</a>
<a class="sourceLine" id="cb439-20" data-line-number="20">filtered &lt;-<span class="st"> </span><span class="kw">quickCellQC</span>(qcstats, <span class="dt">nmads=</span><span class="dv">3</span>,</a>
<a class="sourceLine" id="cb439-21" data-line-number="21">    <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>)</a>
<a class="sourceLine" id="cb439-22" data-line-number="22">sce.muraro &lt;-<span class="st"> </span>sce.muraro[,<span class="op">!</span>filtered<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb439-23" data-line-number="23"></a>
<a class="sourceLine" id="cb439-24" data-line-number="24">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb439-25" data-line-number="25"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb439-26" data-line-number="26"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb439-27" data-line-number="27">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.muraro)</a>
<a class="sourceLine" id="cb439-28" data-line-number="28">sce.muraro &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.muraro, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb439-29" data-line-number="29">sce.muraro &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.muraro)</a>
<a class="sourceLine" id="cb439-30" data-line-number="30"></a>
<a class="sourceLine" id="cb439-31" data-line-number="31">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb439-32" data-line-number="32"><span class="co"># Blocking on a combined plate and donor factor.</span></a>
<a class="sourceLine" id="cb439-33" data-line-number="33">block &lt;-<span class="st"> </span><span class="kw">paste0</span>(sce.muraro<span class="op">$</span>plate, <span class="st">&quot;_&quot;</span>, sce.muraro<span class="op">$</span>donor)</a>
<a class="sourceLine" id="cb439-34" data-line-number="34">dec.muraro &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sce.muraro, <span class="st">&quot;ERCC&quot;</span>, <span class="dt">block=</span>block)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb440-1" data-line-number="1">sce.muraro</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 16940 2346 
## metadata(0):
## assays(2): counts logcounts
## rownames(16940): ENSG00000268895 ENSG00000121410 ...
##   ENSG00000159840 ENSG00000074755
## rowData names(2): symbol chr
## colnames(2346): D28-1_1 D28-1_2 ... D30-8_93 D30-8_94
## colData names(3): label donor plate
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(1): ERCC</code></pre>
<p>We subset both batches to their common universe of genes;
adjust their scaling to equalize sequencing coverage (not really necessary in this case, as the coverage is already similar, but we will do so anyway for consistency);
and select those genes with positive average biological components for further use.</p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb442-1" data-line-number="1">universe &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(sce.grun), <span class="kw">rownames</span>(sce.muraro))</a>
<a class="sourceLine" id="cb442-2" data-line-number="2">universe &lt;-<span class="st"> </span>universe[<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&quot;^ERCC&quot;</span>, universe)]</a>
<a class="sourceLine" id="cb442-3" data-line-number="3"></a>
<a class="sourceLine" id="cb442-4" data-line-number="4">normed.pancreas &lt;-<span class="st"> </span><span class="kw">multiBatchNorm</span>(sce.grun[universe,], sce.muraro[universe,])</a>
<a class="sourceLine" id="cb442-5" data-line-number="5">sce.grun &lt;-<span class="st"> </span>normed.pancreas[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb442-6" data-line-number="6">sce.muraro &lt;-<span class="st"> </span>normed.pancreas[[<span class="dv">2</span>]]</a>
<a class="sourceLine" id="cb442-7" data-line-number="7"></a>
<a class="sourceLine" id="cb442-8" data-line-number="8">combined.pan &lt;-<span class="st"> </span><span class="kw">combineVar</span>(dec.grun[universe,], dec.muraro[universe,])</a>
<a class="sourceLine" id="cb442-9" data-line-number="9">chosen.genes &lt;-<span class="st"> </span>universe[combined.pan<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a></code></pre></div>
<p>We observe that <code>rescaleBatches()</code> is unable to align cells from different batches in Figure <a href="integrating-datasets-1.html#fig:tsne-pancreas-rescaled">15.4</a>.
This is attributable to differences in population composition between batches, with additional complications from non-linearities in the batch effect, e.g., when the magnitude or direction of the batch effect differs between cell types.</p>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb443-1" data-line-number="1">rescaled.pancreas &lt;-<span class="st"> </span><span class="kw">rescaleBatches</span>(sce.grun, sce.muraro)</a>
<a class="sourceLine" id="cb443-2" data-line-number="2">rescaled.pancreas &lt;-<span class="st"> </span><span class="kw">runPCA</span>(rescaled.pancreas, <span class="dt">subset_row=</span>chosen.genes,</a>
<a class="sourceLine" id="cb443-3" data-line-number="3">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(), <span class="dt">exprs_values=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb443-4" data-line-number="4"></a>
<a class="sourceLine" id="cb443-5" data-line-number="5">rescaled.pancreas &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(rescaled.pancreas, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb443-6" data-line-number="6"><span class="kw">plotTSNE</span>(rescaled.pancreas, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-pancreas-rescaled"></span>
<img src="P2_W11.data-integration_files/figure-html/tsne-pancreas-rescaled-1.png" alt="$t$-SNE plot of the pancreas datasets after correction with `rescaleBatches()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">
Figure 15.4: <span class="math inline">\(t\)</span>-SNE plot of the pancreas datasets after correction with <code>rescaleBatches()</code>. Each point represents a cell and is colored according to the batch of origin.
</p>
</div>
<p>Here, we use <code>fastMNN()</code> to merge together the two human pancreas datasets described earlier.
Clustering on the merged datasets yields fewer batch-specific clusters, which is recapitulated as greater intermingling between batches in Figure <a href="integrating-datasets-1.html#fig:tsne-pancreas-mnn">15.5</a>.
This improvement over Figure <a href="integrating-datasets-1.html#fig:tsne-pancreas-rescaled">15.4</a> represents the ability of <code>fastMNN()</code> to adapt to more complex situations involving differences in population composition between batches.</p>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb444-1" data-line-number="1">mnn.pancreas &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(sce.grun, sce.muraro, <span class="dt">subset.row=</span>chosen.genes)</a>
<a class="sourceLine" id="cb444-2" data-line-number="2"></a>
<a class="sourceLine" id="cb444-3" data-line-number="3">snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(mnn.pancreas, <span class="dt">use.dimred=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb444-4" data-line-number="4">clusters &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb444-5" data-line-number="5">tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Cluster=</span>clusters, <span class="dt">Batch=</span>mnn.pancreas<span class="op">$</span>batch)</a>
<a class="sourceLine" id="cb444-6" data-line-number="6">tab</a></code></pre></div>
<pre><code>##        Batch
## Cluster   1   2
##      1  320 279
##      2  343 265
##      3  209 847
##      4   61 197
##      5  161 399
##      6   42   1
##      7   25 108
##      8   22 127
##      9   59  80
##      10  33   0
##      11   0  18
##      12   8   4
##      13   7  21</code></pre>
<div class="sourceCode" id="cb446"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb446-1" data-line-number="1">mnn.pancreas &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(mnn.pancreas, <span class="dt">dimred=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb446-2" data-line-number="2"><span class="kw">plotTSNE</span>(mnn.pancreas, <span class="dt">colour_by=</span><span class="st">&quot;batch&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:tsne-pancreas-mnn"></span>
<img src="P2_W11.data-integration_files/figure-html/tsne-pancreas-mnn-1.png" alt="$t$-SNE plot of the pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">
Figure 15.5: <span class="math inline">\(t\)</span>-SNE plot of the pancreas datasets after correction with <code>fastMNN()</code>. Each point represents a cell and is colored according to the batch of origin.
</p>
</div>
</div>
<div id="using-the-corrected-values" class="section level2">
<h2><span class="header-section-number">15.8</span> Using the corrected values</h2>
<p>The greatest value of batch correction lies in facilitating cell-based analysis of population heterogeneity in a consistent manner across batches.
Cluster 1 in batch A is the same as cluster 1 in batch B when the clustering is performed on the merged data.
There is no need to identify mappings between separate clusterings, which might not even be possible when the clusters are not well-separated.
The burden of interpretation is consolidated by generating a single set of clusters for all batches, rather than requiring separate examination of each batch’s clusters.
Another benefit is that the available number of cells is increased when all batches are combined, which allows for greater resolution of population structure in downstream analyses<a href="#fn24" class="footnote-ref" id="fnref24"><sup>24</sup></a>.
We previously demonstrated the application of clustering methods to the batch-corrected data, but the same principles apply for other analyses like trajectory reconstruction.</p>
<p>At this point, it is also tempting to use the batch-corrected values for gene-based analyses like DE-based marker gene detection.
This is not generally recommended as an arbitrary correction algorithm is not obliged to preserve the magnitude (or even direction) of differences in per-gene expression when attempting to align multiple batches.
For example, cosine normalization in <code>fastMNN()</code> shrinks the magnitude of the expression values so that any subsequent log-fold change calculations are meaningless.
Rather, it is preferable to perform DE analyses using the uncorrected expression values with blocking on the batch, as discussed i Chapter ???.
This strategy is based on the expectation that any genuine DE between clusters should still be present in a within-batch comparison where batch effects are absent.
It penalizes genes that exhibit inconsistent DE across batches, thus protecting against misleading conclusions when a population in one batch is aligned to a similar-but-not-identical population in another batch.</p>
</div>
<div id="session-info-9" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] fossil_0.3.7                shapefiles_0.7             
 [3] foreign_0.8-71              maps_3.3.0                 
 [5] sp_1.3-1                    scater_1.13.17             
 [7] ggplot2_3.2.1               scran_1.13.13              
 [9] batchelor_1.1.10            SingleCellExperiment_1.7.7 
[11] SummarizedExperiment_1.15.8 Biobase_2.45.0             
[13] GenomicRanges_1.37.14       GenomeInfoDb_1.21.1        
[15] HDF5Array_1.13.5            rhdf5_2.29.0               
[17] DelayedArray_0.11.4         BiocParallel_1.19.2        
[19] IRanges_2.19.14             S4Vectors_0.23.20          
[21] BiocGenerics_0.31.5         matrixStats_0.54.0         
[23] BiocStyle_2.13.2            Cairo_1.5-10               

loaded via a namespace (and not attached):
 [1] viridis_0.5.1            edgeR_3.27.12           
 [3] BiocSingular_1.1.5       viridisLite_0.3.0       
 [5] DelayedMatrixStats_1.7.1 assertthat_0.2.1        
 [7] statmod_1.4.32           highr_0.8               
 [9] BiocManager_1.30.4       dqrng_0.2.1             
[11] GenomeInfoDbData_1.2.1   vipor_0.4.5             
[13] yaml_2.2.0               pillar_1.4.2            
[15] lattice_0.20-38          beachmat_2.1.2          
[17] glue_1.3.1               limma_3.41.15           
[19] digest_0.6.20            XVector_0.25.0          
[21] colorspace_1.4-1         cowplot_0.9.4           
[23] htmltools_0.3.6          Matrix_1.2-17           
[25] pkgconfig_2.0.2          bookdown_0.13           
[27] zlibbioc_1.31.0          purrr_0.3.2             
[29] scales_1.0.0             Rtsne_0.15              
[31] tibble_2.1.3             withr_2.1.2             
[33] lazyeval_0.2.2           magrittr_1.5            
[35] crayon_1.3.4             evaluate_0.14           
[37] beeswarm_0.2.3           tools_3.6.1             
[39] stringr_1.4.0            Rhdf5lib_1.7.4          
[41] munsell_0.5.0            locfit_1.5-9.1          
[43] irlba_2.3.3              compiler_3.6.1          
[45] rsvd_1.0.2               rlang_0.4.0             
[47] grid_3.6.1               RCurl_1.95-4.12         
[49] BiocNeighbors_1.3.3      igraph_1.2.4.1          
[51] labeling_0.3             bitops_1.0-6            
[53] rmarkdown_1.13           gtable_0.3.0            
[55] R6_2.4.0                 gridExtra_2.3           
[57] knitr_1.24               dplyr_0.8.3             
[59] stringi_1.4.3            ggbeeswarm_0.6.0        
[61] Rcpp_1.0.2               tidyselect_0.2.5        
[63] xfun_0.9                </code></pre>
</div>
</div>
<div id="references-4" class="section level2 unnumbered">
<h2>References</h2>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-butler2018integrating">
<p>Butler, A., P. Hoffman, P. Smibert, E. Papalexi, and R. Satija. 2018. “Integrating single-cell transcriptomic data across different conditions, technologies, and species.” <em>Nat. Biotechnol.</em> 36 (5):411–20.</p>
</div>
<div id="ref-grun2016denovo">
<p>Grun, D., M. J. Muraro, J. C. Boisset, K. Wiebrands, A. Lyubimova, G. Dharmadhikari, M. van den Born, et al. 2016. “De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data.” <em>Cell Stem Cell</em> 19 (2):266–77.</p>
</div>
<div id="ref-haghverdi2018batch">
<p>Haghverdi, L., A. T. L. Lun, M. D. Morgan, and J. C. Marioni. 2018. “Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors.” <em>Nat. Biotechnol.</em> 36 (5):421–27.</p>
</div>
<div id="ref-leek2012sva">
<p>Leek, J. T., W. E. Johnson, H. S. Parker, A. E. Jaffe, and J. D. Storey. 2012. “The sva package for removing batch effects and other unwanted variation in high-throughput experiments.” <em>Bioinformatics</em> 28 (6):882–83.</p>
</div>
<div id="ref-lin2019scmerge">
<p>Lin, Y., S. Ghazanfar, K. Y. X. Wang, J. A. Gagnon-Bartsch, K. K. Lo, X. Su, Z. G. Han, et al. 2019. “scMerge leverages factor analysis, stable expression, and pseudoreplication to merge multiple single-cell RNA-seq datasets.” <em>Proc. Natl. Acad. Sci. U.S.A.</em> 116 (20):9775–84.</p>
</div>
<div id="ref-muraro2016singlecell">
<p>Muraro, M. J., G. Dharmadhikari, D. Grun, N. Groen, T. Dielen, E. Jansen, L. van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Syst</em> 3 (4):385–94.</p>
</div>
<div id="ref-ritchie2015limma">
<p>Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K. Smyth. 2015. “limma powers differential expression analyses for RNA-sequencing and microarray studies.” <em>Nucleic Acids Res.</em> 43 (7):e47.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="22">
<li id="fn22"><p>This step can be much, much, much more painful. As is often said, biologists would rather share a toothbrush than nomenclature.<a href="integrating-datasets-1.html#fnref22" class="footnote-back">↩</a></p></li>
<li id="fn23"><p>If I already knew the type and state of each cell, why would I waste money sequencing them?<a href="integrating-datasets-1.html#fnref23" class="footnote-back">↩</a></p></li>
<li id="fn24"><p>And a nice <span class="math inline">\(t\)</span>-SNE plot for Figure 1. Hey, those atlas papers almost write themselves!<a href="integrating-datasets-1.html#fnref24" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="doublet-detection.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="multi-sample-comparisons.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W11.data-integration.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
