<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 1 Introduction | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 1 Introduction | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Introduction | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Robert A. Amezquita">
<meta name="author" content="Bioconductor">


<meta name="date" content="2019-03-05">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="learning-r-and-more.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Orchestrating Single-Cell Analysis with Bioconductor</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-essentials-of-sce"><i class="fa fa-check"></i><b>4.2</b> The Essentials of <code>sce</code></a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Workflow: Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#package-requirements"><i class="fa fa-check"></i><b>6.1</b> Package Requirements</a></li>
<li class="chapter" data-level="6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#loading-the-data"><i class="fa fa-check"></i><b>6.2</b> Loading the Data</a></li>
<li class="chapter" data-level="6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#preprocessing"><i class="fa fa-check"></i><b>6.3</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#working-with-common-genes"><i class="fa fa-check"></i><b>6.3.1</b> Working with Common Genes</a></li>
<li class="chapter" data-level="6.3.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>6.3.2</b> Cell and Gene Quality Control</a></li>
<li class="chapter" data-level="6.3.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#normalization"><i class="fa fa-check"></i><b>6.3.3</b> Normalization</a></li>
<li class="chapter" data-level="6.3.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#multibatch-normalization"><i class="fa fa-check"></i><b>6.3.4</b> Multibatch Normalization</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a></li>
<li class="chapter" data-level="6.5" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#combining-the-datasets"><i class="fa fa-check"></i><b>6.5</b> Combining the Datasets</a></li>
<li class="chapter" data-level="6.6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#integrating-datasets"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#naive-approach"><i class="fa fa-check"></i><b>6.6.1</b> Naive Approach</a></li>
<li class="chapter" data-level="6.6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#limma-batch-correction"><i class="fa fa-check"></i><b>6.6.2</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="6.6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#mnn-approach"><i class="fa fa-check"></i><b>6.6.3</b> MNN Approach</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#session-info"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="workflow-clustering.html"><a href="workflow-clustering.html"><i class="fa fa-check"></i><b>7</b> Workflow: Clustering</a><ul>
<li class="chapter" data-level="7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#package-requirements-1"><i class="fa fa-check"></i><b>7.1</b> Package Requirements</a></li>
<li class="chapter" data-level="7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#loading-the-data-1"><i class="fa fa-check"></i><b>7.2</b> Loading the Data</a></li>
<li class="chapter" data-level="7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#understanding-the-data"><i class="fa fa-check"></i><b>7.3</b> Understanding the Data</a></li>
<li class="chapter" data-level="7.4" data-path="workflow-clustering.html"><a href="workflow-clustering.html#preprocessing-1"><i class="fa fa-check"></i><b>7.4</b> Preprocessing</a></li>
<li class="chapter" data-level="7.5" data-path="workflow-clustering.html"><a href="workflow-clustering.html#feature-selection-1"><i class="fa fa-check"></i><b>7.5</b> Feature Selection</a></li>
<li class="chapter" data-level="7.6" data-path="workflow-clustering.html"><a href="workflow-clustering.html#dimension-reduction"><i class="fa fa-check"></i><b>7.6</b> Dimension Reduction</a></li>
<li class="chapter" data-level="7.7" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clustering"><i class="fa fa-check"></i><b>7.7</b> Clustering</a><ul>
<li class="chapter" data-level="7.7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#sc3"><i class="fa fa-check"></i><b>7.7.1</b> SC3</a></li>
<li class="chapter" data-level="7.7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clusterexperiment"><i class="fa fa-check"></i><b>7.7.2</b> clusterExperiment</a></li>
<li class="chapter" data-level="7.7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#biocneighbors"><i class="fa fa-check"></i><b>7.7.3</b> BiocNeighbors</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p><a href="https://bioconductor.org">Bioconductor</a> is an open source, open development software project to provide tools for the analysis and comprehension of high-throughput genomic data. It is based primarily on the <a href="http://www.r-project.org/">R</a> programming language.</p>
<div id="what-you-will-learn" class="section level2">
<h2><span class="header-section-number">1.1</span> What you will learn</h2>
<p>The goal of this book is to provide a solid foundation in the usage of Bioconductor tools for single-cell RNA-seq analysis by walking through various steps of typical workflows using example datasets. We strive to tackle key concepts covered in the manuscript, <strong>“Orchestrating Single-Cell Analysis with Bioconductor”</strong>, with each workflow covering these in varying detail, as well as essential preliminaries that are important for following along with the workflows on your own.</p>
<div id="preliminaries" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Preliminaries</h3>
<p>For those unfamiliar with R (and those looking to learn more), we recommend reading the <a href="learning-r-and-more.html#learning-r-and-more"><em>Learning R and More</em></a> chapter, which first and foremost covers how to get started with R. We point to many great online resources for learning R, as well as related tools that are nice to know for bioinformatic analysis. For advanced users, we also point to some extra resources that go beyond the basics. While we provide an extensive list of learning resources for the interested audience in this chapter, we only ask for <a href="learning-r-and-more.html#getting-started-with-r">some familiarity with R</a> before going to the next section.</p>
<p>We then briefly cover getting started with <a href="#using-r-and-bioconductor"><em>Using R and Bioconductor</em></a>. Bioconductor, being its own repository, has a unique set of tools, documentation, resources, and practices that benefit from some extra explanation.</p>
<p><a href="data-infrastructure.html#data-infrastructure"><em>Data Infrastructure</em></a> merits a separate chapter. The reason for this is that common data containers are an essential part of Bioconductor workflows because they enable interoperability across packages, allowing for “plug and play” usage of cutting-edge tools. Specifically, here we cover the <em>SingleCellExperiment</em> class in depth, as it has become the working standard for Bioconductor based single-cell analysis packages.</p>
<p>Finally, before diving into the various workflows, armed with knowledge about the <em>SingleCellExperiment</em> class, we briefly discuss the datasets that will be used throughout the book in <a href="about-the-data.html#about-the-data"><em>About the Data</em></a>.</p>
</div>
<div id="workflows" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Workflows</h3>
<p>All workflows begin with data import and subsequent <em>quality control and normalization</em>, going from a raw (count) expression matrix to a clean one. This includes adjusting for experimental factors and possibly even latent factors. Using the clean expression matrix, <em>feature selection</em> strategies can be applied to select the features (genes) driving heterogeneity. Furthermore, these features can then be used to perform <em>dimensionality reduction</em>, which enables downstream analysis that would not otherwise be possible and visualization on 2- or 3-dimensions.</p>
<p>From there, the workflows largely focus on differing downstream analyses. <em>Clustering</em> details how to segment a scRNA-seq dataset, and <em>differential expression</em> provides a means to determine what drives the differences between different groups of cells. <em>Integrating datasets</em> walks through merging scRNA-seq datasets, an area of need as the number of scRNA-seq datasets continues to grow and comparisons between datasets must be done. Finally, we touch upon how to work with <em>large scale data</em>, specifically where it becomes impractical or impossible to work with data solely in-memory.</p>
<p>As an added bonus, we dedicate a chapter to <em>interactive visualization</em>, which focuses on using the <em>iSEE</em> package to enable active exploration of a single cell experiment’s data.</p>
</div>
</div>
<div id="what-you-wont-learn" class="section level2">
<h2><span class="header-section-number">1.2</span> What you won’t learn</h2>
<p>The field of bioinformatic analysis is large and filled with many potential trajectories depending on the biological system being studied and technology being deployed. Here, we only briefly survey some of the many tools available for the analysis of scRNA-seq, focusing on Bioconductor packages. It is impossible to thoroughly review the plethora of tools available through R and Bioconductor for biological analysis in one book, but we hope to provide the means for further exploration on your own.</p>
<p>Thus, it goes without saying that you may not learn the optimal workflow for your own data from our examples - while we strive to provide high quality templates, they should be treated as just that - a template from which to extend upon for your own analyses.</p>
</div>
<div id="who-we-wrote-this-for" class="section level2">
<h2><span class="header-section-number">1.3</span> Who we wrote this for</h2>
<p>We’ve written this book with the interested experimental biologist in mind, and do our best to make few assumptions on previous programming or statistical experience. Likewise, we also welcome more seasoned bioinformaticians who are looking for a starting point from which to dive into single-cell RNA-seq analysis. As such, we welcome any and all feedback for improving this book to help increase accessibility and refine technical details.</p>
</div>
<div id="why-we-wrote-this" class="section level2">
<h2><span class="header-section-number">1.4</span> Why we wrote this</h2>
<p>This book was conceived in the fall of 2018, as single-cell RNA-seq analysis continued its rise in prominence in the field of biology. With its rapid growth, and the ongoing developments within Bioconductor tailored specifically for scRNA-seq, it became apparent that an update to the <a href="https://www.nature.com/articles/nmeth.3252">Orchestrating high-throughput genomic analysis with Bioconductor</a> paper was necessary for the age of single-cell studies.</p>
<p>We strive to highlight the fantastic software by people who call Bioconductor home for their tools, and in the process hope to showcase the Bioconductor community at large in continually pushing forward the field of biological analysis.</p>
</div>
<div id="acknowledgements" class="section level2">
<h2><span class="header-section-number">1.5</span> Acknowledgements</h2>
<p>We would like to thank all Bioconductor contributors for their efforts in creating the definitive leading-edge repository of software for biological analysis. It is truly extraordinary to chart the growth of Bioconductor over the years. We are thankful for the wonderful community of scientists and developers alike that together make the Bioconductor community special.</p>
<p>We would first and foremost like to thank the Bioconductor core team and the emerging targets subcommittee for commissioning this work, Raphael Gottardo for his continuous mentorship of the writing herein, and all our contributors to the companion manuscript of this book.</p>
<p>We’d also like to thank Garret Grolemund and Hadley Wickham for their book, <a href="https://r4ds.had.co.nz/index.html">R for Data Science</a>, from which we drew stylistic and teaching inspiration.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="learning-r-and-more.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
