<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 2 Learning R and More | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 2 Learning R and More | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Learning R and More | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-02-07">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="using-r-and-bioconductor.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and More</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting Started with R</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#nice-companions-for-r"><i class="fa fa-check"></i><b>2.2</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#shellbash"><i class="fa fa-check"></i><b>2.2.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="2.2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#git"><i class="fa fa-check"></i><b>2.2.2</b> Git</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#going-beyond-the-basics"><i class="fa fa-check"></i><b>2.3</b> Going Beyond the Basics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>2.3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#other-languages"><i class="fa fa-check"></i><b>2.3.2</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html"><i class="fa fa-check"></i><b>3</b> Using R &amp; Bioconductor</a><ul>
<li class="chapter" data-level="3.1" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#a-brief-overview"><i class="fa fa-check"></i><b>3.1</b> A Brief Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#prerequisites"><i class="fa fa-check"></i><b>3.2</b> Prerequisites</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#installing-r"><i class="fa fa-check"></i><b>3.2.1</b> Installing R</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>3.2.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>3.3</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="3.4" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#bioconductor-documentation"><i class="fa fa-check"></i><b>3.4</b> Bioconductor Documentation</a></li>
<li class="chapter" data-level="3.5" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#becoming-a-bioconductor-expert"><i class="fa fa-check"></i><b>3.5</b> Becoming a Bioconductor Expert</a></li>
<li class="chapter" data-level="3.6" data-path="using-r-and-bioconductor.html"><a href="using-r-and-bioconductor.html#conventions-used-in-this-book"><i class="fa fa-check"></i><b>3.6</b> Conventions used in this book</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-object"><i class="fa fa-check"></i><b>4.1</b> The <em>SingleCellExperiment</em> Object</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-sce"><i class="fa fa-check"></i><b>4.2</b> Working with <em>SCE</em></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="learning-r-and-more" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Learning R and More</h1>
<p>There exist many resources for learning R and other tools that may be needed to run particular bioinformatic analyses. Here we have laid out a general order for getting started with bioinformatics with an R centric approach, from top to bottom.</p>
<p>For those who are just getting started - start with the first section, <a href="learning-r-and-more.html#getting-started-with-r"><em>Getting Started with R</em></a>. With just this section, you’ll have all you need to run the latest and greatest from Bioconductor and develop your own analytical workflows.</p>
<p>Once you are comfortable with R, check out learning some <a href="learning-r-and-more.html#nice-companions-for-r"><em>Nice Companions for R</em></a>, which includes Shell/Bash and Git, as this will greatly expand the horizon of bioinformatic tools available to you, enhancing your ability to work with raw data using shell-based tools and helping you keep track of your code over time (with the added bonus of keeping it safe!).</p>
<p>For those seeking to advance their knowledge of R further, we recommend checking out the <a href="learning-r-and-more.html#going-beyond-the-basics"><em>Going Beyond the Basics</em></a> for resources to become an R expert and the benefits of checking out other languages. However, note that while learning more advanced concepts in R is useful, it is by no means necessary to get started. Checking out the resources in <a href="learning-r-and-more.html#getting-started-with-r"><em>Getting Started with R</em></a> is more than enough to get going.</p>
<p>Before programming on your own hardware, read ahead for some brief notes on installing R and Bioconductor packages in the <a href="using-r-and-bioconductor.html#prerequisites"><em>Prerequisites</em></a> section.</p>
<div id="getting-started-with-r" class="section level2">
<h2><span class="header-section-number">2.1</span> Getting Started with R</h2>
<p>To learn more about programming with R, we highly recommend checking out <a href="https://www.datacamp.com/">Datacamp</a>, which includes both introductory and advanced courses within the <a href="https://www.datacamp.com/courses/tech:r">R track</a>. Datacamp is all online with many free courses, with videos and a code editor/console that promotes an interactive learning experience. What we like about Datacamp is that it is more focused on topics and programming paradigms that center around data science, which is especially helpful for getting started with R.</p>
<p>Beyond just Datacamp, a mainstay resource for learning R is the <a href="https://r4ds.had.co.nz/">R for Data Science</a> book. This book illustrates R programming through the exploration of various data science concepts - transformation, visualization, exploration, and more.</p>
</div>
<div id="nice-companions-for-r" class="section level2">
<h2><span class="header-section-number">2.2</span> Nice Companions for R</h2>
<p>While not essential for our purposes, many bioinformatic tools require knowledge beyond just R to install, run, and import their results into R for further analysis. The most important of which are basic knowledge of the Shell/Bash utilities, for working with bioinformatic pipelines and troubleshooting (R package) installation issues, and Git for code version control.</p>
<div id="shellbash" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Shell/Bash</h3>
<p>Datacamp and other interactive online resources such as <a href="https://www.codecademy.com/catalog/subject/all">Codecademy</a> are great places to learn some of these extra skills. We highly recommend learning <a href="https://www.datacamp.com/courses/tech:shell">Shell/Bash</a>, as it is the starting point for most bioinformatic processing pipelines.</p>
</div>
<div id="git" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Git</h3>
<p>We would recommend learning <a href="https://www.datacamp.com/courses/tech:git">Git</a> next, a system for code versioning control which underlies the popular <a href="https://Github.com">Github</a> repository, where many of the most popular open source tools are hosted. Learning Git is essential for not only keeping track of your own code, but also for using, managing, and contributing to open source software projects.</p>
<p>For a more R centric look at using Git (and Github), we highly recommend checking out <a href="https://happygitwithr.com/">Happy Git and Github for the useR</a>.</p>
</div>
</div>
<div id="going-beyond-the-basics" class="section level2">
<h2><span class="header-section-number">2.3</span> Going Beyond the Basics</h2>
<p>Naturally, there are myriad other things to advance your skills in R, data science, and bioinformatics. Here we’ll briefly touch on this before moving on.</p>
<div id="becoming-an-r-expert" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Becoming an R Expert</h3>
<p>For a deeper dive into the finer details of the R programming language, the <a href="https://adv-r.hadley.nz/">Advanced R</a>. While targeted at more experienced R users and programmers, this book represents a comprehensive compendium of more advanced concepts, and touches on some of the paradigms used extensively by developers throughout Bioconductor, specifically <a href="https://adv-r.hadley.nz/s4.html">programming with S4</a>.</p>
<p>Eventually, you’ll reach the point where you have your own collection of functions, datasets, and reach the point where you will be writing your own packages. Luckily, there’s a guide for just that, with the book <a href="http://r-pkgs.had.co.nz/">R Packages</a>. Packages are great even if just for personal use, and of course, with some polishing, can eventually become a package available on CRAN or Bioconductor. Furthermore, they are also a great way of putting together code associated with a manuscript, promoting reproducible, accessible computing practices, something we all strive for in our work.</p>
<p>Finally, we save the most intriguing resource for last - another book for those on the road to becoming an R expert is <a href="https://www.burns-stat.com/pages/Tutor/R_inferno.pdf">R Inferno</a>, which dives into many of the unique quirks of R. <em>Warning: this book goes <em>very</em> deep into the painstaking details of R.</em></p>
</div>
<div id="other-languages" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Other Languages</h3>
<p>Another frequent question that comes up is “What else should I learn besides R?” Firstly, we believe that <a href="learning-r-and-more.html#getting-started-with-r">honing your R skills</a> is first and foremost, and beyond just R, learning Shell/Bash and Git covered in the <a href="#nice-friends-for-r"><em>Nice Friends for R</em></a> section are already a great start. For those just getting started, these skills should become comfortable in practice before moving on.</p>
<p>However, there are indeed benefits to going beyond just R. At a basic level, learning other programming languages helps broaden one’s perspective - similar to learning multiple spoken or written languages, learning about other programming languages (even if only in a cursory manner) helps one identify broader patterns that may be applicable across languages.</p>
<p>At an applied level, work within and outside of R has made it ever more friendly now than ever before with multi-lingual setups and teams, enabling the use of the best tool for the job at hand. For example, <a href="https://www.python.org/">Python</a> is another popular language used in both data science and a broader array of applications as well. R now supports a native Python interface via the <a href="https://github.com/rstudio/reticulate"><em>reticulate</em></a> package, enabling access to tools developed originally in Python such as the popular <a href="https://tensorflow.rstudio.com/">TensorFlow</a> framework for machine learning applications. <a href="http://www.cplusplus.com/">C++</a> is frequently used natively in R as well via <a href="http://www.rcpp.org/">Rcpp</a> in packages to massively accelerate computations. Finally, multiple langauges are supported in code documents and reports through <a href="https://rmarkdown.rstudio.com/lesson-5.html">R Markdown</a>.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-r-and-bioconductor.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
