<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 2 Learning R and Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 2 Learning R and Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Learning R and Bioconductor | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-03-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="beyond-r-basics.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-essentials-of-sce"><i class="fa fa-check"></i><b>4.2</b> The Essentials of <code>sce</code></a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Workflow: Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#package-requirements"><i class="fa fa-check"></i><b>6.1</b> Package Requirements</a></li>
<li class="chapter" data-level="6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#loading-the-data"><i class="fa fa-check"></i><b>6.2</b> Loading the Data</a></li>
<li class="chapter" data-level="6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#preprocessing"><i class="fa fa-check"></i><b>6.3</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#working-with-common-genes"><i class="fa fa-check"></i><b>6.3.1</b> Working with Common Genes</a></li>
<li class="chapter" data-level="6.3.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>6.3.2</b> Cell and Gene Quality Control</a></li>
<li class="chapter" data-level="6.3.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#normalization"><i class="fa fa-check"></i><b>6.3.3</b> Normalization</a></li>
<li class="chapter" data-level="6.3.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#multibatch-normalization"><i class="fa fa-check"></i><b>6.3.4</b> Multibatch Normalization</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a></li>
<li class="chapter" data-level="6.5" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#combining-the-datasets"><i class="fa fa-check"></i><b>6.5</b> Combining the Datasets</a></li>
<li class="chapter" data-level="6.6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#integrating-datasets"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#naive-approach"><i class="fa fa-check"></i><b>6.6.1</b> Naive Approach</a></li>
<li class="chapter" data-level="6.6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#limma-batch-correction"><i class="fa fa-check"></i><b>6.6.2</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="6.6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#mnn-approach"><i class="fa fa-check"></i><b>6.6.3</b> MNN Approach</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#session-info"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="workflow-clustering.html"><a href="workflow-clustering.html"><i class="fa fa-check"></i><b>7</b> Workflow: Clustering</a><ul>
<li class="chapter" data-level="7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#package-requirements-1"><i class="fa fa-check"></i><b>7.1</b> Package Requirements</a></li>
<li class="chapter" data-level="7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#loading-the-data-1"><i class="fa fa-check"></i><b>7.2</b> Loading the Data</a></li>
<li class="chapter" data-level="7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#understanding-the-data"><i class="fa fa-check"></i><b>7.3</b> Understanding the Data</a></li>
<li class="chapter" data-level="7.4" data-path="workflow-clustering.html"><a href="workflow-clustering.html#preprocessing-1"><i class="fa fa-check"></i><b>7.4</b> Preprocessing</a></li>
<li class="chapter" data-level="7.5" data-path="workflow-clustering.html"><a href="workflow-clustering.html#feature-selection-1"><i class="fa fa-check"></i><b>7.5</b> Feature Selection</a></li>
<li class="chapter" data-level="7.6" data-path="workflow-clustering.html"><a href="workflow-clustering.html#dimension-reduction"><i class="fa fa-check"></i><b>7.6</b> Dimension Reduction</a></li>
<li class="chapter" data-level="7.7" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clustering"><i class="fa fa-check"></i><b>7.7</b> Clustering</a><ul>
<li class="chapter" data-level="7.7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#sc3"><i class="fa fa-check"></i><b>7.7.1</b> SC3</a></li>
<li class="chapter" data-level="7.7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clusterexperiment"><i class="fa fa-check"></i><b>7.7.2</b> clusterExperiment</a></li>
<li class="chapter" data-level="7.7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#biocneighbors"><i class="fa fa-check"></i><b>7.7.3</b> BiocNeighbors</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="learning-r-and-more" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Learning R and Bioconductor</h1>
<p>In this section, we outline various resources for learning R and Bioconductor. We provide a brief set of instructions for installing R on your own machine, and then cover how to get help for functions, packages, and Bioconductor-specific resources for learning more.</p>
<div id="the-benefits-of-r-and-bioconductor" class="section level2">
<h2><span class="header-section-number">2.1</span> The Benefits of R and Bioconductor</h2>
<p><a href="http://www.r-project.org/">R</a> is a high-level programming language that was initially designed for statistical applications. While there is much to be said <a href="https://www.r-project.org/about.html">about R</a> as a programming language, one of the key advantages of using R is that it is highly extensible through <em>packages</em>. Packages are collections of functions, data, and documentation that extends the capabilities of base R. The ease of development and distribution of packages for R has made it a rich environment for many fields of study and application.</p>
<p>One of the primary ways in which packages are distributed is through centralized repositories. The first R repository a user typically runs into is the <a href="https://cran.r-project.org/mirrors.html">Comprehensive R Archive Network</a> (CRAN), which hosts over 13,000 packages to date, and is home to many of the most popular R packages.</p>
<p>Similar to CRAN, <a href="https://bioconductor.org">Bioconductor</a> is a repository of R packages as well. However, whereas CRAN is a general purpose repository, Bioconductor focuses on software tailored for genomic analysis. Furthermore, Bioconductor has stricter requirements for a package to be accepted into the repository. Of particular interest to us is the inclusion of <a href="#learning-more">high quality documentation</a> and the use of common <a href="data-infrastructure.html#data-infrastructure">data infrastructure</a> to promote package interoperability.</p>
<p>In order to use these packages from CRAN and Bioconductor, and start programming with R to follow along in these workflows, some knowledge of R is helpful. Here we outline resources to guide you through learning the basics.</p>
</div>
<div id="getting-started-with-r" class="section level2">
<h2><span class="header-section-number">2.2</span> Learning R Online</h2>
<p>To learn more about programming with R, we highly recommend checking out the online courses offered by <a href="https://www.datacamp.com/">Datacamp</a>, which includes both introductory and advanced courses within the <a href="https://www.datacamp.com/courses/tech:r">R track</a>. Datacamp is all online with many free courses, with videos and a code editor/console that promotes an interactive learning experience. What we like about Datacamp is that it is more focused on topics and programming paradigms that center around data science, which is especially helpful for getting started with R.</p>
<p>Beyond just Datacamp, a mainstay resource for learning R is the <a href="https://r4ds.had.co.nz/">R for Data Science</a> book. This book illustrates R programming through the exploration of various data science concepts - transformation, visualization, exploration, and more.</p>
</div>
<div id="running-r-locally" class="section level2">
<h2><span class="header-section-number">2.3</span> Running R Locally</h2>
<p>While learning R through online resources is a great way to start with R, as it requires minimal knowledge to start up, at some point, it will be desirable to have a local installation - on your own hardware - of R. This will allow you to install and maintain your own software and code, and furthermore allow you to create a personalized workspace.</p>
<div id="installing-r" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Installing R</h3>
<p>Prior to getting started with this book, some prior programming experience with R is helpful. Check out the <a href="learning-r-and-more.html#learning-r-and-more"><em>Learning R and More</em></a> chapter for a list of resources to get started with R and other useful tools for bioinformatic analysis.</p>
<p>To follow along with the analysis workflows in this book on your personal computer, it is first necessary to install the <a href="http://www.r-project.org/">R</a> programming language. Additionally, we recommend a graphical user interface such as <a href="http://www.rstudio.com/download">RStudio</a> for programming in R and visualization. RStudio features many helpful tools, such as code completion and an interactive data viewer to name but two. For more details, please see the online book <a href="https://r4ds.had.co.nz/introduction.html#prerequisites"><em>R for Data Science</em> prerequisites</a> section for more information about installing R and using RStudio.</p>
<div id="for-macoslinux-users" class="section level4">
<h4><span class="header-section-number">2.3.1.1</span> For MacOS/Linux Users</h4>
<p>A special note for MacOS/Linux users: we highly recommend using a package manager to manage your R installation. This differs across different Linux distributions, but for MacOS we highly recommend the <a href="https://brew.sh/">Homebrew</a> package manager. Follow the website directions to install homebrew, and install R via the commandline with <code>brew install R</code>, and it will automatically configure your installation for you. Upgrading to new R versions can be done by running <code>brew upgrade</code>.</p>
</div>
</div>
<div id="installing-r-bioconductor-packages" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Installing R &amp; Bioconductor Packages</h3>
<p>After installing R, the next step is to install R packages. In the R console, you can install packages from CRAN via the <code>install.packages()</code> function. In order to <a href="https://www.bioconductor.org/install/">install Bioconductor packages</a>, we will first need the <em>BiocManager</em> package which is hosted on CRAN. This can be done by running:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</code></pre>
<p>The <em>BiocManager</em> package makes it easy to install packages from the Bioconductor repository. For example, to install the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><em>SingleCellExperiment</em></a> package, we run:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## the command below is a one-line shortcut for:</span>
<span class="co">## library(BiocManager)</span>
<span class="co">## install(&quot;SingleCellExperiment&quot;)</span>
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SingleCellExperiment&quot;</span>)</code></pre>
<p>Throughout the book, we can load packages via the <code>library()</code> function, which by convention usually comes at the top of scripts to alert readers as to what packages are required. For example, to load the <em>SingleCellExperiment</em> package, we run:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)</code></pre>
<p>Many packages will be referenced throughout the book within the workflows, and similar to the above, can be installed using the <code>BiocManager::install()</code> function.</p>
</div>
</div>
<div id="getting-help-in-and-out-of-r" class="section level2">
<h2><span class="header-section-number">2.4</span> Getting Help In (and Out) of R</h2>
<p>One of the most helpful parts of R is being able to get help <em>inside</em> of R. For example, to get the manual associated with a function, class, dataset, or package, you can prepend the code of interest with a <code>?</code> to retrieve the relevant help page. For example, to get information about the <code>data.frame()</code> function, the <em>SingleCellExperiment</em> class, the in-built <em>iris</em> dataset, or for the <em>BiocManager</em> package, you can type:</p>
<pre class="sourceCode r"><code class="sourceCode r">?data.frame
?SingleCellExperiment
?iris
?BiocManager</code></pre>
<p>Beyond the R console, there are myriad online resources to get help. The R for Data Science book has a great section dedicated to looking for help <a href="https://r4ds.had.co.nz/introduction.html#getting-help-and-learning-more">outside of R</a>. In particular, <a href="https://stackoverflow.com/questions/tagged/r">Stackoverflow’s R tag</a> is a helpful resource for asking and exploring general R programming questions.</p>
</div>
<div id="bioconductor-documentation" class="section level2">
<h2><span class="header-section-number">2.5</span> Bioconductor Help</h2>
<p>One of the key tenets of Bioconductor software that makes it stand out from CRAN is the required documentation of packages and workflows. In addition, Bioconductor hosts a Bioconductor-specific support site that has grown into a valuable resource of its own, thanks to the work of dedicated volunteers.</p>
<div id="bioconductor-packages" class="section level3">
<h3><span class="header-section-number">2.5.1</span> Bioconductor Packages</h3>
<p>Each package hosted on Bioconductor has a dedicated page with various resources. For an example, looking at the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html"><code>scater</code></a> package page on Bioconductor, we see that it contains:</p>
<ul>
<li>a brief description of the package at the top, in addition to the authors, maintainer, and an associated citation</li>
<li>installation instructions that can be cut and paste into your R console</li>
<li>documentation - vignettes, reference manual, news</li>
</ul>
<p>Here, the most important information comes from the documentation section. Every package in Bioconductor is <em>required</em> to be submitted with a <em>vignette</em> - a document showcasing basic functionality of the package. Typically, these vignettes have a descriptive title that summarizes the main objective of the vignette. These vignettes are a great resource for learning how to operate the essential functionality of the package.</p>
<p>The <em>reference manual</em> contains a comprehensive listing of all the functions available in the package. This is a compilation of each function’s <em>manual</em>, aka help pages, which can be accessed programmatically in the R console via <code>?&lt;function&gt;</code>.</p>
<p>Finally, the <em>NEWS</em> file contains notes from the authors which highlight changes across different versions of the package. This is a great way of tracking changes, especially functions that are added, removed, or deprecated, in order to keep your scripts current with new versions of dependent packages.</p>
<p>Below this, the <em>Details</em> section covers finer nuances of the package, mostly relating to its relationship to other packages:</p>
<ul>
<li>upstream dependencies (<em>Depends</em>, <em>Imports</em>, <em>Suggests</em> fields): packages that are imported upon loading the given package</li>
<li>downstream dependencies (<em>Depends On Me</em>, <em>Imports Me</em>, <em>Suggests Me</em>): packages that import the given package when loaded</li>
</ul>
<p>For example, we can see that an entry called <a href="https://bioconductor.org/packages/release/workflows/html/simpleSingleCell.html"><em>simpleSingle</em></a> in the <em>Depends On Me</em> field on the <code>scater</code> page takes us to a step-by-step workflow for low-level analysis of single-cell RNA-seq data.</p>
<p>One additional <em>Details</em> entry, the <em>biocViews</em>, is helpful for looking at how the authors annotate their package. For example, for the <code>scater</code> package, we see that it is associated with <code>DataImport</code>, <code>DimensionReduction</code>, <code>GeneExpression</code>, <code>RNASeq</code>, and <code>SingleCell</code>, to name but some of its many annotations. We cover <em>biocViews</em> in more detail.</p>
</div>
<div id="biocviews" class="section level3">
<h3><span class="header-section-number">2.5.2</span> biocViews</h3>
<p>To find packages via the Bioconductor website, one useful resource is the <a href="https://bioconductor.org/packages/release/BiocViews.html">BiocViews</a> page, which provides a hierarchically organized view of annotations associated with Bioconductor packages.</p>
<p>Under the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">“Software”</a> label for example (which is comprised of most of the Bioconductor packages), there exist many different views to explore packages. For example, we can inspect based on the associated <a href="https://bioconductor.org/packages/release/BiocViews.html#___Technology">“Technology”</a>, and explore <a href="https://bioconductor.org/packages/release/BiocViews.html#___Sequencing">“Sequencing”</a> associated packages, and furthermore subset based on <a href="https://bioconductor.org/packages/release/BiocViews.html#___RNASeq">“RNASeq”</a>.</p>
<p>Another area of particular interest is the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Workflow">“Workflow”</a> view, which provides Bioconductor packages that illustrate an analytical workflow. For example, the <a href="https://bioconductor.org/packages/release/BiocViews.html#___SingleCellWorkflow">“SingleCellWorkflow”</a> contains the aforementioned tutorial, encapsulated in the <em>simpleSingleCell</em> package.</p>
</div>
<div id="bioconductor-forums" class="section level3">
<h3><span class="header-section-number">2.5.3</span> Bioconductor Forums</h3>
<p>The <a href="https://support.bioconductor.org/">Bioconductor support site</a> contains a Stackoverflow-style question and answer support site that is actively contributed to from both users and package developers. Thanks to the work of dedicated volunteers, there are ample questions to explore to learn more about Bioconductor specific workflows.</p>
<p>Another way to connect with the Bioconductor community is through <a href="https://bioc-community.herokuapp.com">Slack</a>, which hosts various channels dedicated to packages and workflows. The Bioc-community Slack is a great way to stay in the loop on the latest developments happening across Bioconductor, and we recommend exploring the “Channels” section to find topics of interest.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="beyond-r-basics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
