<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Marker gene detection | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Marker gene detection | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Marker gene detection | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-07-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="clustering-2.html">
<link rel="next" href="integrating-datasets-1.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
<li class="chapter" data-level="6.3" data-path="quick-start.html"><a href="quick-start.html#session-info"><i class="fa fa-check"></i><b>6.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#session-info-1"><i class="fa fa-check"></i><b>7.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.3.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>8.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.3.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.3.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.5</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.5.1</b> Background</a></li>
<li class="chapter" data-level="8.5.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.5.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.5.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.5.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.6</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.6.1</b> Overview</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.6.2</b> Checking for discarded cell types</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i><b>8.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#putting-it-all-together"><i class="fa fa-check"></i><b>9.6</b> Putting it all together</a></li>
<li class="chapter" data-level="9.7" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>9.7</b> Why (log-)transform?</a></li>
<li class="chapter" data-level="9.8" data-path="normalization.html"><a href="normalization.html#session-info-3"><i class="fa fa-check"></i><b>9.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="feature-selection-2.html"><a href="feature-selection-2.html"><i class="fa fa-check"></i><b>10</b> Feature selection</a><ul>
<li class="chapter" data-level="10.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#motivation-2"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>10.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="10.2.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>10.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="10.2.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#coefficient-of-variation"><i class="fa fa-check"></i><b>10.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="10.2.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#sec:spikeins"><i class="fa fa-check"></i><b>10.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="10.2.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>10.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#selecting-highly-variable-genes"><i class="fa fa-check"></i><b>10.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="10.3.1" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>10.3.1</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="10.3.2" data-path="feature-selection-2.html"><a href="feature-selection-2.html#based-on-a-fixed-threshold"><i class="fa fa-check"></i><b>10.3.2</b> Based on a fixed threshold</a></li>
<li class="chapter" data-level="10.3.3" data-path="feature-selection-2.html"><a href="feature-selection-2.html#keeping-all-genes-above-the-trend"><i class="fa fa-check"></i><b>10.3.3</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="feature-selection-2.html"><a href="feature-selection-2.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>10.4</b> Putting it all together</a></li>
<li class="chapter" data-level="10.5" data-path="feature-selection-2.html"><a href="feature-selection-2.html#session-info-4"><i class="fa fa-check"></i><b>10.5</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html"><i class="fa fa-check"></i><b>11</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="11.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#overview-1"><i class="fa fa-check"></i><b>11.1</b> Overview</a></li>
<li class="chapter" data-level="11.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#principal-components-analysis"><i class="fa fa-check"></i><b>11.2</b> Principal components analysis</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#background-1"><i class="fa fa-check"></i><b>11.2.1</b> Background</a></li>
<li class="chapter" data-level="11.2.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#performing-the-pca"><i class="fa fa-check"></i><b>11.2.2</b> Performing the PCA</a></li>
<li class="chapter" data-level="11.2.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>11.2.3</b> Choosing the number of PCs</a></li>
<li class="chapter" data-level="11.2.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#putting-it-together"><i class="fa fa-check"></i><b>11.2.4</b> Putting it together</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>11.3</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="11.3.1" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#motivation-4"><i class="fa fa-check"></i><b>11.3.1</b> Motivation</a></li>
<li class="chapter" data-level="11.3.2" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#visualizating-with-pca"><i class="fa fa-check"></i><b>11.3.2</b> Visualizating with PCA</a></li>
<li class="chapter" data-level="11.3.3" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>11.3.3</b> t-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="11.3.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>11.3.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="11.3.5" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#interpreting-the-plots"><i class="fa fa-check"></i><b>11.3.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="dimensionality-reduction-2.html"><a href="dimensionality-reduction-2.html#session-info-5"><i class="fa fa-check"></i><b>11.4</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>12</b> Clustering</a><ul>
<li class="chapter" data-level="12.1" data-path="clustering-2.html"><a href="clustering-2.html#motivation-5"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="clustering-2.html"><a href="clustering-2.html#comments-on-truth"><i class="fa fa-check"></i><b>12.2</b> Comments on Truth</a></li>
<li class="chapter" data-level="12.3" data-path="clustering-2.html"><a href="clustering-2.html#graph-based-clustering"><i class="fa fa-check"></i><b>12.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="12.3.1" data-path="clustering-2.html"><a href="clustering-2.html#background-2"><i class="fa fa-check"></i><b>12.3.1</b> Background</a></li>
<li class="chapter" data-level="12.3.2" data-path="clustering-2.html"><a href="clustering-2.html#implementation"><i class="fa fa-check"></i><b>12.3.2</b> Implementation</a></li>
<li class="chapter" data-level="12.3.3" data-path="clustering-2.html"><a href="clustering-2.html#other-parameters"><i class="fa fa-check"></i><b>12.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="12.3.4" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>12.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="clustering-2.html"><a href="clustering-2.html#k-means-clustering"><i class="fa fa-check"></i><b>12.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="12.4.1" data-path="clustering-2.html"><a href="clustering-2.html#background-3"><i class="fa fa-check"></i><b>12.4.1</b> Background</a></li>
<li class="chapter" data-level="12.4.2" data-path="clustering-2.html"><a href="clustering-2.html#base-implementation"><i class="fa fa-check"></i><b>12.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="12.4.3" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>12.4.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="clustering-2.html"><a href="clustering-2.html#hierarchical-clustering"><i class="fa fa-check"></i><b>12.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="12.5.1" data-path="clustering-2.html"><a href="clustering-2.html#background-4"><i class="fa fa-check"></i><b>12.5.1</b> Background</a></li>
<li class="chapter" data-level="12.5.2" data-path="clustering-2.html"><a href="clustering-2.html#implementation-1"><i class="fa fa-check"></i><b>12.5.2</b> Implementation</a></li>
<li class="chapter" data-level="12.5.3" data-path="clustering-2.html"><a href="clustering-2.html#assessing-cluster-separation-2"><i class="fa fa-check"></i><b>12.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="clustering-2.html"><a href="clustering-2.html#session-info-6"><i class="fa fa-check"></i><b>12.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html"><i class="fa fa-check"></i><b>13</b> Marker gene detection</a><ul>
<li class="chapter" data-level="13.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#motivation-6"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-pairwise-t-tests"><i class="fa fa-check"></i><b>13.2</b> Using pairwise <span class="math inline">\(t\)</span>-tests</a><ul>
<li class="chapter" data-level="13.2.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#standard-application"><i class="fa fa-check"></i><b>13.2.1</b> Standard application</a></li>
<li class="chapter" data-level="13.2.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>13.2.2</b> Using the log-fold change</a></li>
<li class="chapter" data-level="13.2.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#finding-cluster-specific-markers"><i class="fa fa-check"></i><b>13.2.3</b> Finding cluster-specific markers</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#alternative-testing-regimes"><i class="fa fa-check"></i><b>13.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="13.3.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>13.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="13.3.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>13.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="13.3.3" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>13.3.3</b> Using custom DE methods</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#handling-blocking-factors"><i class="fa fa-check"></i><b>13.4</b> Handling blocking factors</a></li>
<li class="chapter" data-level="13.5" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>13.5</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="13.6" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>13.6</b> Using the <code>design=</code> argument</a></li>
<li class="chapter" data-level="13.7" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#invalidity-of-p-values"><i class="fa fa-check"></i><b>13.7</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="13.7.1" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>13.7.1</b> From data snooping</a></li>
<li class="chapter" data-level="13.7.2" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#nature-of-replication"><i class="fa fa-check"></i><b>13.7.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="marker-gene-detection.html"><a href="marker-gene-detection.html#session-info-7"><i class="fa fa-check"></i><b>13.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>14</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="14.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#motivation-7"><i class="fa fa-check"></i><b>14.1</b> Motivation</a></li>
<li class="chapter" data-level="14.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#simple-two-batch-example"><i class="fa fa-check"></i><b>14.2</b> Simple two batch example</a><ul>
<li class="chapter" data-level="14.2.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#setting-up-the-data-1"><i class="fa fa-check"></i><b>14.2.1</b> Setting up the data</a></li>
<li class="chapter" data-level="14.2.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#performing-mnn-correction"><i class="fa fa-check"></i><b>14.2.2</b> Performing MNN correction</a></li>
<li class="chapter" data-level="14.2.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#correction-diagnostics"><i class="fa fa-check"></i><b>14.2.3</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#session-info-8"><i class="fa fa-check"></i><b>14.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>15</b> Sharing and Interactive Interfaces</a><ul>
<li class="chapter" data-level="15.1" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html#session-info-9"><i class="fa fa-check"></i><b>15.1</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>16</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="16.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>16.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="16.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>16.2</b> Parallelization</a></li>
<li class="chapter" data-level="16.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>16.3</b> On-Disk Data</a></li>
<li class="chapter" data-level="16.4" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#session-info-10"><i class="fa fa-check"></i><b>16.4</b> Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>17</b> About the Data</a><ul>
<li class="chapter" data-level="17.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>17.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="17.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>17.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>18</b> About the Contributors</a><ul>
<li class="chapter" data-level="18.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>18.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="18.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>18.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="18.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>18.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="18.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>18.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>19</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="marker-gene-detection" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> Marker gene detection</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-6" class="section level2">
<h2><span class="header-section-number">13.1</span> Motivation</h2>
<p>To interpret our clustering results from Chapter ???, we identify the genes that drive separation between clusters.
These marker genes allow us to assign biological meaning to each cluster based on their functional annotation.
In the most obvious case, the marker genes for each cluster are <em>a priori</em> associated with particular cell types, allowing us to treat the clustering as a proxy for cell type identity.
The same principle can be applied to more subtle differences in activation status or differentiation state.</p>
<p>Identification of marker genes is usually based around the retrospective detection of differential expression between clusters.
Genes that are more strongly DE are more likely to have driven cluster separation in the first place.
Several different statistical tests are available to quantify the differences in expression profiles, and different approaches can be used to consolidate test results into a single ranking of genes for each cluster.
These choices parametrize the theroetical differences between the various marker detection strategies presented in this chapter.
We will demonstrate using the 10X PBMC dataset:</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb319-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb319-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb319-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="st">&quot;raw_data&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb319-4" data-line-number="4">raw.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</a>
<a class="sourceLine" id="cb319-5" data-line-number="5">    <span class="st">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</a>
<a class="sourceLine" id="cb319-6" data-line-number="6"><span class="kw">untar</span>(raw.path, <span class="dt">exdir=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k&quot;</span>))</a>
<a class="sourceLine" id="cb319-7" data-line-number="7"></a>
<a class="sourceLine" id="cb319-8" data-line-number="8"><span class="kw">library</span>(DropletUtils)</a>
<a class="sourceLine" id="cb319-9" data-line-number="9">fname &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</a>
<a class="sourceLine" id="cb319-10" data-line-number="10">sce.pbmc &lt;-<span class="st"> </span><span class="kw">read10xCounts</span>(fname, <span class="dt">col.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb319-11" data-line-number="11"></a>
<a class="sourceLine" id="cb319-12" data-line-number="12">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb319-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb319-14" data-line-number="14"><span class="kw">rownames</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb319-15" data-line-number="15"></a>
<a class="sourceLine" id="cb319-16" data-line-number="16"><span class="kw">library</span>(EnsDb.Hsapiens.v86)</a>
<a class="sourceLine" id="cb319-17" data-line-number="17">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(EnsDb.Hsapiens.v86, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb319-18" data-line-number="18">    <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb319-19" data-line-number="19"></a>
<a class="sourceLine" id="cb319-20" data-line-number="20">### cell-detection <span class="al">###</span></a>
<a class="sourceLine" id="cb319-21" data-line-number="21"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb319-22" data-line-number="22">e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb319-23" data-line-number="23">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="kw">which</span>(e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>)]</a>
<a class="sourceLine" id="cb319-24" data-line-number="24"></a>
<a class="sourceLine" id="cb319-25" data-line-number="25">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb319-26" data-line-number="26">sce.pbmc &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce.pbmc, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(location<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb319-27" data-line-number="27">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.pbmc<span class="op">$</span>pct_counts_Mito, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb319-28" data-line-number="28">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb319-29" data-line-number="29"></a>
<a class="sourceLine" id="cb319-30" data-line-number="30">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb319-31" data-line-number="31"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb319-32" data-line-number="32"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb319-33" data-line-number="33">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb319-34" data-line-number="34">sce.pbmc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.pbmc, <span class="dt">min.mean=</span><span class="fl">0.1</span>, <span class="dt">cluster=</span>clusters)</a>
<a class="sourceLine" id="cb319-35" data-line-number="35">sce.pbmc &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb319-36" data-line-number="36"></a>
<a class="sourceLine" id="cb319-37" data-line-number="37">### feature-selection <span class="al">###</span></a>
<a class="sourceLine" id="cb319-38" data-line-number="38">fit.pbmc &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce.pbmc, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb319-39" data-line-number="39">dec.pbmc &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(<span class="dt">fit=</span>fit.pbmc)</a>
<a class="sourceLine" id="cb319-40" data-line-number="40">o &lt;-<span class="st"> </span><span class="kw">order</span>(dec.pbmc<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb319-41" data-line-number="41">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec.pbmc)[<span class="kw">head</span>(o, <span class="dv">2000</span>)]</a>
<a class="sourceLine" id="cb319-42" data-line-number="42"></a>
<a class="sourceLine" id="cb319-43" data-line-number="43">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb319-44" data-line-number="44"><span class="kw">set.seed</span>(<span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb319-45" data-line-number="45">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce.pbmc, <span class="dt">feature_set=</span>chosen.hvgs, <span class="dt">ncomponents=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb319-46" data-line-number="46">    <span class="dt">BSPARAM=</span>BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>())</a>
<a class="sourceLine" id="cb319-47" data-line-number="47"></a>
<a class="sourceLine" id="cb319-48" data-line-number="48"><span class="kw">set.seed</span>(<span class="dv">100000</span>)</a>
<a class="sourceLine" id="cb319-49" data-line-number="49">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.pbmc, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb319-50" data-line-number="50"></a>
<a class="sourceLine" id="cb319-51" data-line-number="51"><span class="kw">set.seed</span>(<span class="dv">1000000</span>)</a>
<a class="sourceLine" id="cb319-52" data-line-number="52">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce.pbmc, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb319-53" data-line-number="53"></a>
<a class="sourceLine" id="cb319-54" data-line-number="54">### clustering <span class="al">###</span></a>
<a class="sourceLine" id="cb319-55" data-line-number="55">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb319-56" data-line-number="56">clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb319-57" data-line-number="57">sce.pbmc<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb320-1" data-line-number="1">sce.pbmc</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 3922 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
## rowData names(10): ID Symbol ... total_counts log10_total_counts
## colnames(3922): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCACAGGTCCAC-1 TTTGTCATCCCAAGAT-1
## colData names(39): Sample Barcode ...
##   pct_counts_in_top_500_features_Mito cluster
## reducedDimNames(3): PCA TSNE UMAP
## spikeNames(0):</code></pre>
</div>
<div id="using-pairwise-t-tests" class="section level2">
<h2><span class="header-section-number">13.2</span> Using pairwise <span class="math inline">\(t\)</span>-tests</h2>
<div id="standard-application" class="section level3">
<h3><span class="header-section-number">13.2.1</span> Standard application</h3>
<p>The Welch <span class="math inline">\(t\)</span>-test is an obvious choice of statistical method to test for differences in expression between clusters.
It is quickly computed and has good statistical properties for large numbers of cells [soneson2018bias].
We use the <code>findMarkers()</code> function to perform pairwise comparisons between clusters for each gene.
This yields a list of <code>DataFrame</code>s contained ranked candidate markers for each cluster.</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb322-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb322-2" data-line-number="2">markers.pbmc &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster)</a>
<a class="sourceLine" id="cb322-3" data-line-number="3">markers.pbmc</a></code></pre></div>
<pre><code>## DataFrameList of length 12
## names(12): 1 2 3 4 5 6 7 8 9 10 11 12</code></pre>
<p>To demonstrate, we use cluster 9 as our cluster of interest for this section.
The relevant <code>DataFrame</code> contains log<sub>2</sub>-fold changes of expression in cluster 9 over each other cluster, along with several statistics obtained by combining <span class="math inline">\(p\)</span>-values <span class="citation">(Simes <a href="#ref-simes1986improved">1986</a>)</span> across the pairwise comparisons involving 9.</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb324-1" data-line-number="1">chosen &lt;-<span class="st"> &quot;9&quot;</span></a>
<a class="sourceLine" id="cb324-2" data-line-number="2">interesting &lt;-<span class="st"> </span>markers.pbmc[[chosen]]</a>
<a class="sourceLine" id="cb324-3" data-line-number="3"><span class="kw">colnames</span>(interesting)</a></code></pre></div>
<pre><code>##  [1] &quot;Top&quot;      &quot;p.value&quot;  &quot;FDR&quot;      &quot;logFC.1&quot;  &quot;logFC.2&quot;  &quot;logFC.3&quot; 
##  [7] &quot;logFC.4&quot;  &quot;logFC.5&quot;  &quot;logFC.6&quot;  &quot;logFC.7&quot;  &quot;logFC.8&quot;  &quot;logFC.10&quot;
## [13] &quot;logFC.11&quot; &quot;logFC.12&quot;</code></pre>
<p>Of particular interest is the <code>Top</code> field, which contains the highest<a href="#fn17" class="footnote-ref" id="fnref17"><sup>17</sup></a> rank for each gene across all pairwise comparisons involving cluster 9.
The set of genes with <code>Top</code> values of 1 contains the gene with the lowest <span class="math inline">\(p\)</span>-value from each comparison.
Similarly, the set of genes with <code>Top</code> values less than or equal to 10 contains the top 10 genes from each comparison.
Each <code>DataFrame</code> produced by <code>findMarkers()</code> will order genes based on the <code>Top</code> value.</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb326-1" data-line-number="1">interesting[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 3 columns
##                Top               p.value                   FDR
##          &lt;integer&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## LEF1             1 6.20939421496419e-167 2.09219328679006e-162
## GZMA             1 2.85174646001885e-163 4.80433726119381e-159
## ACAP1            1 6.15083691785694e-163  6.9082099703425e-159
## IGHD             1 7.47621860433727e-135 5.03807419309093e-131
## TGFBI            1  5.05258071032724e-82  4.25604136134411e-79
## FCGR3A           1  2.52077860009527e-70   1.4395782059595e-67
## HLA-DQA1         1  3.97224203642212e-60  1.48711914639119e-57
## SEC61B           1  6.04861464944905e-38  7.25274099638915e-36
## PF4              1  2.35117342678427e-28  1.53231020197426e-26
## LYAR             2 4.83449973651891e-131 2.71489390203784e-127</code></pre>
<p>We use the <code>Top</code> value to identify a set of genes that is guaranteed to distinguish cluster 9 from any other cluster.
Here, we examine the top 5 genes from each pairwise comparison (Figure <a href="marker-gene-detection.html#fig:heat-basic-pbmc">13.1</a>).
Some inspection of the most upregulated genes suggest that cluster 9 contains platelets or their precursors, based on the expression of platelet factor 4 (<em>PF4</em>) and pro-platelet basic protein (<em>PPBP</em>).</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb328-1" data-line-number="1">best.set &lt;-<span class="st"> </span>interesting[interesting<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5</span>,]</a>
<a class="sourceLine" id="cb328-2" data-line-number="2">logFCs &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(best.set[,<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)])</a>
<a class="sourceLine" id="cb328-3" data-line-number="3"><span class="kw">colnames</span>(logFCs) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;logFC.&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(logFCs))</a>
<a class="sourceLine" id="cb328-4" data-line-number="4"></a>
<a class="sourceLine" id="cb328-5" data-line-number="5"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb328-6" data-line-number="6"><span class="kw">pheatmap</span>(logFCs, <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="dt">length.out=</span><span class="dv">101</span>))</a></code></pre></div>
<div class="figure"><span id="fig:heat-basic-pbmc"></span>
<img src="P2_W09.marker-detection_files/figure-html/heat-basic-pbmc-1.png" alt="Heatmap of log-fold changes for cluster `r chosen` over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range." width="672" />
<p class="caption">
Figure 13.1: Heatmap of log-fold changes for cluster <code>r chosen</code> over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<div id="using-the-log-fold-change" class="section level3">
<h3><span class="header-section-number">13.2.2</span> Using the log-fold change</h3>
<p>Our previous <code>findMarkers()</code> call considers both up- and downregulated genes to be potential markers.
However, downregulated genes are less appealing as markers as it is more difficult to interpret and experimentally validate an absence of expression.
To focus on up-regulated markers, we can instead perform a one-sided <span class="math inline">\(t\)</span>-test to identify genes that are upregulated in each cluster compared to the others.
This is achieved by setting <code>direction=&quot;up&quot;</code> in the <code>findMarkers()</code> call.</p>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb329-1" data-line-number="1">markers.pbmc.up &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb329-2" data-line-number="2">interesting.up &lt;-<span class="st"> </span>markers.pbmc.up[[chosen]]</a>
<a class="sourceLine" id="cb329-3" data-line-number="3">interesting.up[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 3 columns
##                 Top              p.value                  FDR
##           &lt;integer&gt;            &lt;numeric&gt;            &lt;numeric&gt;
## PF4               1 1.17558671339212e-28 3.96102187210342e-24
## TMSB4X            2 1.64674064685164e-25 2.77426396775095e-21
## SDPR              2 3.55136790784062e-23 3.98865967622607e-19
## GPX1              3 4.58923373210941e-22 3.50460237357863e-18
## PPBP              3 5.20063271439818e-22 3.50460237357863e-18
## NRGN              4 1.28482173823981e-21 7.21513060804201e-18
## TAGLN2            4  4.3101937823381e-21 2.07468099002999e-17
## GNG11             5 8.00725077665062e-20 3.37245384585583e-16
## HIST1H2AC         6 1.02722989599352e-19 3.84572045728949e-16
## TUBB1             7 1.52910079823675e-19 5.15215222957892e-16</code></pre>
<p>The <span class="math inline">\(t\)</span>-test also allows us to specify a non-zero log-fold change as the null hypothesis.
This allows us to consider the magnitude of the log-fold change in our <span class="math inline">\(p\)</span>-value calculations, in a manner that is more rigorous than simply filtering directly on the log-fold changes <span class="citation">(McCarthy and Smyth <a href="#ref-mccarthy2009treat">2009</a>)</span>.
(Specifically, a simple threshold does not consider the variance and can enrich for genes that have both large log-fold changes and large variances.)
We perform this by setting <code>lfc=</code> in our <code>findMarkers()</code> call - when combined with <code>direction=</code>, this tests for genes with log-fold changes that are significantly greater than 1:</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb331-1" data-line-number="1">markers.pbmc.up2 &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster, </a>
<a class="sourceLine" id="cb331-2" data-line-number="2">    <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>, <span class="dt">lfc=</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb331-3" data-line-number="3">interesting.up2 &lt;-<span class="st"> </span>markers.pbmc.up2[[chosen]]</a>
<a class="sourceLine" id="cb331-4" data-line-number="4">interesting.up2[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 3 columns
##                 Top              p.value                  FDR
##           &lt;integer&gt;            &lt;numeric&gt;            &lt;numeric&gt;
## PF4               1 6.02669690930859e-25 2.03063525662244e-20
## TMSB4X            2 7.97170089366702e-19 8.95328299704057e-15
## SDPR              2 3.73547685699784e-19 6.29315786098425e-15
## GPX1              3 6.07463154377111e-18 4.09357270471648e-14
## PPBP              3 1.33107797557639e-18 1.12123353272677e-14
## NRGN              4 4.05970433364242e-17 2.27979463029579e-13
## TAGLN2            4 2.25018630384581e-16 1.08311110459687e-12
## GNG11             5 5.50609782505896e-16 2.31903075146921e-12
## HIST1H2AC         6 6.53594082728356e-16 2.44691100260547e-12
## TUBB1             7 2.07146248785668e-15 6.97958570658431e-12</code></pre>
<p>These two settings yield a more focused set of candidate marker genes that are upregulated in cluster 9 (Figure <a href="#heat-focused-pbmc"><strong>??</strong></a>).</p>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb333-1" data-line-number="1">best.set &lt;-<span class="st"> </span>interesting.up2[interesting.up2<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5</span>,]</a>
<a class="sourceLine" id="cb333-2" data-line-number="2">logFCs &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(best.set[,<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)])</a>
<a class="sourceLine" id="cb333-3" data-line-number="3"><span class="kw">colnames</span>(logFCs) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;logFC.&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(logFCs))</a>
<a class="sourceLine" id="cb333-4" data-line-number="4"></a>
<a class="sourceLine" id="cb333-5" data-line-number="5"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb333-6" data-line-number="6"><span class="kw">pheatmap</span>(logFCs, <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="dt">length.out=</span><span class="dv">101</span>))</a></code></pre></div>
<div class="figure"><span id="fig:heat-focused-pbmc"></span>
<img src="P2_W09.marker-detection_files/figure-html/heat-focused-pbmc-1.png" alt="Heatmap of log-fold changes for cluster 9 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range." width="672" />
<p class="caption">
Figure 13.2: Heatmap of log-fold changes for cluster 9 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
<p>Of course, this increased stringency is not without cost.
If only upregulated genes are requested from <code>findMarkers()</code>, any cluster defined by downregulation of a marker gene will not contain that gene among the top set of features in its <code>DataFrame</code>.
This is occasionally relevant for subtypes or other states that are distinguished by high versus low expression of particular genes<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a>.
Similarly, setting an excessively high log-fold change threshold may discard otherwise useful genes.
For example, a gene upregulated in a small proportion of cells of a cluster will have a small log-fold change but can still be an effective marker if the focus is on specificity rather than sensitivity.</p>
</div>
<div id="finding-cluster-specific-markers" class="section level3">
<h3><span class="header-section-number">13.2.3</span> Finding cluster-specific markers</h3>
<p>By default, <code>findMarkers()</code> will give a high ranking to genes that are differentially expressed in any pairwise comparison.
This is because a gene only needs a very low <span class="math inline">\(p\)</span>-value in a single pairwise comparison to achieve a low <code>Top</code> value.
A more stringent approach would onlly consider genes that are differentially expressed in all pairwise comparisons involving the cluster of interest.
To achieve this, we set <code>pval.type=&quot;all&quot;</code> in <code>findMarkers()</code> to use an intersection-union test <span class="citation">(Berger and Hsu <a href="#ref-berger1996bioequivalence">1996</a>)</span> where the combined <span class="math inline">\(p\)</span>-value for each gene is the maximum of the <span class="math inline">\(p\)</span>-values from all pairwise comparisons.
A gene will only achieve a low combined <span class="math inline">\(p\)</span>-value if it is strongly DE in all comparisons to other clusters.</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb334-1" data-line-number="1"><span class="co"># We can combine this with &#39;direction=&#39;.</span></a>
<a class="sourceLine" id="cb334-2" data-line-number="2">markers.pbmc.up3 &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster, </a>
<a class="sourceLine" id="cb334-3" data-line-number="3">    <span class="dt">pval.type=</span><span class="st">&quot;all&quot;</span>, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb334-4" data-line-number="4">interesting.up3 &lt;-<span class="st"> </span>markers.pbmc.up3[[chosen]]</a>
<a class="sourceLine" id="cb334-5" data-line-number="5">interesting.up3[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 2 columns
##                        p.value                  FDR
##                      &lt;numeric&gt;            &lt;numeric&gt;
## PF4       1.80014139122098e-28 6.06539640357997e-24
## SDPR      3.55136790784062e-23 5.98298951433908e-19
## PPBP      6.48169985592535e-22 7.27981316485164e-18
## NRGN      5.86340715476335e-21 4.93904101681492e-17
## GNG11     8.95542696697954e-20 6.03488312450816e-16
## HIST1H2AC 1.39989539303288e-19   7.360217470827e-16
## TUBB1     1.52910079823675e-19   7.360217470827e-16
## TMSB4X    1.11117788046074e-16 4.68000343803052e-13
## ACRBP     7.90465526790601e-15 2.95932727329806e-11
## TAGLN2    1.13799040801938e-14  3.8343448807805e-11</code></pre>
<p>This strategy will only report genes that are highly specific to the cluster of interest.
When it works, it can be highly effective as it generates a small focused set of candidate markers.
However, any gene that is expressed at the same level in two or more clusters will simply not be detected.
This is likely to discard many interesting genes, especially if the clusters are finely resolved with weak separation.
To give a concrete example, consider a mixed population of CD4<sup>+</sup>-only, CD8<sup>+</sup>-only, double-positive and double-negative T cells.
With <code>pval.type=&quot;all&quot;</code>, neither <em>Cd4</em> or <em>Cd8</em> would be detected as subpopulation-specific markers because each gene is expressed in two subpopulations.
In comparison, <code>pval.type=&quot;any&quot;</code> will detect both of these genes as they will be DE between at least one pair of subpopulations.
This reliability motivates the use of the latter setting as the default in <code>findMarkers()</code>.</p>
</div>
</div>
<div id="alternative-testing-regimes" class="section level2">
<h2><span class="header-section-number">13.3</span> Alternative testing regimes</h2>
<div id="using-the-wilcoxon-rank-sum-test" class="section level3">
<h3><span class="header-section-number">13.3.1</span> Using the Wilcoxon rank sum test</h3>
<p>The Wilcoxon rank sum test (also known as the Wilcoxon-Mann-Whitney test, or WMW test) is another widely used method for pairwise comparisons between groups of observations.
Its strength lies in the fact that it directly assesses separation between the expression distributions of different clusters.
The WMW test statistic is proportional to the area-under-the-curve (AUC), i.e., the concordance probability, which is the probability of a random cell from one cluster having higher expression than a random cell from another cluster.
In a pairwise comparison, AUCs of 1 or 0 indicate that the two clusters have perfectly separated expression distributions.
Thus, the WMW test directly addresses the most desirable property of a candidate marker gene, while the <span class="math inline">\(t\)</span> test only does so indirectly via the difference in the means and the intra-group variance.</p>
<p>We perform WMW tests using the <code>overlapExprs()</code> function.
This returns a list of <code>DataFrame</code>s containing ranked candidate markers for each cluster.
The <code>direction=</code>, <code>lfc=</code> and <code>pval.type=</code> arguments can be specified and have the same interpretation as described for <span class="math inline">\(t\)</span> tests.
We demonstrate below by detecting upregulated genes in each cluster with <code>direction=&quot;up&quot;</code>.</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb336-1" data-line-number="1"><span class="co"># </span><span class="al">TODO</span><span class="co">: make findMarkers() accept a method= option that</span></a>
<a class="sourceLine" id="cb336-2" data-line-number="2"><span class="co"># switches between these things, rather than having a </span></a>
<a class="sourceLine" id="cb336-3" data-line-number="3"><span class="co"># different function altogether. Also need a lfc= option.</span></a>
<a class="sourceLine" id="cb336-4" data-line-number="4">markers.pbmc.wmw &lt;-<span class="st"> </span><span class="kw">overlapExprs</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb336-5" data-line-number="5"><span class="kw">names</span>(markers.pbmc.wmw)</a></code></pre></div>
<pre><code>##  [1] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;4&quot;  &quot;5&quot;  &quot;6&quot;  &quot;7&quot;  &quot;8&quot;  &quot;9&quot;  &quot;10&quot; &quot;11&quot; &quot;12&quot;</code></pre>
<p>To explore the results in more detail, we focus on the <code>DataFrame</code> for cluster 9.
The interpretation of <code>Top</code> is the same as described for <span class="math inline">\(t\)</span> tests, and Simes’ method is again used to combine <span class="math inline">\(p\)</span>-values across pairwise comparisons.</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb338-1" data-line-number="1">interesting.wmw &lt;-<span class="st"> </span>markers.pbmc.wmw[[chosen]]</a>
<a class="sourceLine" id="cb338-2" data-line-number="2">interesting.wmw[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 3 columns
##                 Top               p.value                   FDR
##           &lt;integer&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## PF4               1  1.0202929049051e-187 3.43777491378731e-183
## TMSB4X            1  1.01281466445067e-33  3.25007402895248e-31
## SDPR              2  3.5648749185145e-184 6.00574477522145e-180
## PPBP              3 4.69409849189378e-163 5.27209848619572e-159
## TUBB1             3  8.5842665088848e-162 7.23095689375918e-158
## GNG11             3 3.27662779920279e-159 2.20805394132683e-155
## NRGN              3 1.42552621019204e-155  8.0052800210352e-152
## CLU               5 1.29715673377618e-149 5.46329987348186e-146
## HIST1H2AC         5 2.48168081125816e-106 5.57451688363549e-103
## TAGLN2            5  2.32628433861945e-34  7.60988587431489e-32</code></pre>
<p>The <code>DataFrame</code> contains the AUCs from comparing cluster 9 to every other cluster (Figure <a href="marker-gene-detection.html#fig:heat-wmw-pbmc">13.3</a>).
A value greater than 0.5 indicates that the gene is upregulated in the current cluster compared to the other cluster,
while values less than 0.5 correspond to downregulation.
We would typically expect AUCs of 0.7-0.8 for a strongly upregulated candidate marker.</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb340-1" data-line-number="1">best.set &lt;-<span class="st"> </span>interesting.wmw[interesting.wmw<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5</span>,]</a>
<a class="sourceLine" id="cb340-2" data-line-number="2">AUCs &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(best.set[,<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)])</a>
<a class="sourceLine" id="cb340-3" data-line-number="3"><span class="kw">colnames</span>(AUCs) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;AUC.&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(AUCs))</a>
<a class="sourceLine" id="cb340-4" data-line-number="4"></a>
<a class="sourceLine" id="cb340-5" data-line-number="5"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb340-6" data-line-number="6"><span class="kw">pheatmap</span>(AUCs, <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">length.out=</span><span class="dv">21</span>),</a>
<a class="sourceLine" id="cb340-7" data-line-number="7">    <span class="dt">color=</span>viridis<span class="op">::</span><span class="kw">viridis</span>(<span class="dv">21</span>))</a></code></pre></div>
<div class="figure"><span id="fig:heat-wmw-pbmc"></span>
<img src="P2_W09.marker-detection_files/figure-html/heat-wmw-pbmc-1.png" alt="Heatmap of AUCs for cluster `r chosen` over all other clusters." width="672" />
<p class="caption">
Figure 13.3: Heatmap of AUCs for cluster <code>r chosen</code> over all other clusters.
</p>
</div>
<p>The main disadvantage of the WMW test is that the AUCs are much slower to compute compared to <span class="math inline">\(t\)</span>-statistics.
This may be inconvenient for interactive analyses involving multiple iterations of marker detection.
We can mitigate this to some extent by parallelizing these calculations using the <code>BPPARAM=</code> argument in <code>overlapExprs()</code>.</p>
</div>
<div id="using-a-binomial-test" class="section level3">
<h3><span class="header-section-number">13.3.2</span> Using a binomial test</h3>
<p>The binomial test identifies genes that differ in the proportion of expressing cells between clusters.
(For the purposes of this section, a cell is considered to express a gene simply if it has non-zero expression for that gene.)
This represents a much more stringent definition of marker genes compared to the other methods, as differences in expression between clusters are effectively ignored if both distributions of expression values are not near zero.
The premise is that genes are more likely to contribute to important biological decisions if they were active in one cluster and silent in another, compared to more subtle “tuning” effects from changing the expression of an active gene.
From a practical perspective, a binary measure of presence/absence is easier to validate.</p>
<p>We perform pairwise binomial tests between clusters using the <code>pairwiseBinom()</code> function.
This returns a list of <code>DataFrame</code>s containing marker statistics for each cluster, such as the <code>Top</code> rank and its <span class="math inline">\(p\)</span>-value.
Here, the effect size is reported as the log-fold change in this proportion between each pair of clusters.
Large positive log-fold changes indicate that the gene is more frequently expressed in one cluster compared to the other.
Here, we focus on genes that are upregulated in each cluster compared to the others by setting <code>direction=&quot;up&quot;</code>.</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb341-1" data-line-number="1"><span class="co"># </span><span class="al">TODO</span><span class="co">: push this into another findMarkers() option. </span></a>
<a class="sourceLine" id="cb341-2" data-line-number="2">b.pairs &lt;-<span class="st"> </span><span class="kw">pairwiseBinom</span>(sce.pbmc, sce.pbmc<span class="op">$</span>cluster, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb341-3" data-line-number="3">markers.pbmc.binom &lt;-<span class="st"> </span><span class="kw">combineMarkers</span>(b.pairs<span class="op">$</span>statistics, b.pairs<span class="op">$</span>pairs)</a>
<a class="sourceLine" id="cb341-4" data-line-number="4"><span class="kw">names</span>(markers.pbmc.binom)</a></code></pre></div>
<pre><code>##  [1] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;4&quot;  &quot;5&quot;  &quot;6&quot;  &quot;7&quot;  &quot;8&quot;  &quot;9&quot;  &quot;10&quot; &quot;11&quot; &quot;12&quot;</code></pre>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb343-1" data-line-number="1">interesting.binom &lt;-<span class="st"> </span>markers.pbmc.binom[[chosen]]</a>
<a class="sourceLine" id="cb343-2" data-line-number="2"><span class="kw">colnames</span>(interesting.binom)</a></code></pre></div>
<pre><code>##  [1] &quot;Top&quot;      &quot;p.value&quot;  &quot;FDR&quot;      &quot;logFC.1&quot;  &quot;logFC.2&quot;  &quot;logFC.3&quot; 
##  [7] &quot;logFC.4&quot;  &quot;logFC.5&quot;  &quot;logFC.6&quot;  &quot;logFC.7&quot;  &quot;logFC.8&quot;  &quot;logFC.10&quot;
## [13] &quot;logFC.11&quot; &quot;logFC.12&quot;</code></pre>
<p>Figure (fig:viol-de-binom) confirms that the top genes exhibit strong differences in the proportion of expressing cells in cluster 9 compared to the others.</p>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb345-1" data-line-number="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb345-2" data-line-number="2">top.genes &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">rownames</span>(interesting.binom))</a>
<a class="sourceLine" id="cb345-3" data-line-number="3"><span class="kw">plotExpression</span>(sce.pbmc, <span class="dt">x=</span><span class="st">&quot;cluster&quot;</span>, <span class="dt">features=</span>top.genes)</a></code></pre></div>
<div class="figure"><span id="fig:viol-de-binom"></span>
<img src="P2_W09.marker-detection_files/figure-html/viol-de-binom-1.png" alt="Distribution of log-normalized expression values for the top 10 DE genes involving cluster 9 with the binomial test, stratified by cluster assignment and coloured by the plate of origin for each cell." width="672" />
<p class="caption">
Figure 13.4: Distribution of log-normalized expression values for the top 10 DE genes involving cluster 9 with the binomial test, stratified by cluster assignment and coloured by the plate of origin for each cell.
</p>
</div>
<p>The disadvantage of the binomial test is that its increased stringency can lead to the loss of good candidate markers.
For example, <em>GCG</em> is a known marker for pancreatic alpha cells but is expressed in almost every other cell of the <span class="citation">Lawlor et al. (<a href="#ref-lawlor2017singlecell">2017</a>)</span> pancreas data (Figure <a href="marker-gene-detection.html#fig:viol-gcg-lawlor">13.5</a>) and would not be highly ranked by the binomial test.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb346-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb346-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb346-3" data-line-number="3">sce.lawlor &lt;-<span class="st"> </span><span class="kw">LawlorPancreasData</span>()</a>
<a class="sourceLine" id="cb346-4" data-line-number="4"></a>
<a class="sourceLine" id="cb346-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb346-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb346-7" data-line-number="7">edb &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73881&quot;</span>]]</a>
<a class="sourceLine" id="cb346-8" data-line-number="8">anno &lt;-<span class="st"> </span><span class="kw">select</span>(edb, <span class="dt">keys=</span><span class="kw">rownames</span>(sce.lawlor), <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, </a>
<a class="sourceLine" id="cb346-9" data-line-number="9">    <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;SEQNAME&quot;</span>))</a>
<a class="sourceLine" id="cb346-10" data-line-number="10"><span class="kw">rowData</span>(sce.lawlor) &lt;-<span class="st"> </span>anno[<span class="kw">match</span>(<span class="kw">rownames</span>(sce.lawlor), anno[,<span class="dv">1</span>]),<span class="op">-</span><span class="dv">1</span>]</a>
<a class="sourceLine" id="cb346-11" data-line-number="11"></a>
<a class="sourceLine" id="cb346-12" data-line-number="12">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb346-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb346-14" data-line-number="14">sce.lawlor &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce.lawlor, <span class="dt">compact=</span><span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb346-15" data-line-number="15">    <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(<span class="kw">rowData</span>(sce.lawlor)<span class="op">$</span>SEQNAME<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb346-16" data-line-number="16">NFeatures &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.lawlor<span class="op">$</span>scater_qc<span class="op">$</span>all<span class="op">$</span>total_features_by_counts, </a>
<a class="sourceLine" id="cb346-17" data-line-number="17">    <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmads=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb346-18" data-line-number="18">LibSize &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.lawlor<span class="op">$</span>scater_qc<span class="op">$</span>all<span class="op">$</span>total_counts, </a>
<a class="sourceLine" id="cb346-19" data-line-number="19">    <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">nmads=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb346-20" data-line-number="20">Mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce.lawlor<span class="op">$</span>scater_qc<span class="op">$</span>feature_control_Mito<span class="op">$</span>pct_counts,</a>
<a class="sourceLine" id="cb346-21" data-line-number="21">    <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>, <span class="dt">nmads=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb346-22" data-line-number="22">discard &lt;-<span class="st"> </span>NFeatures <span class="op">|</span><span class="st"> </span>LibSize <span class="op">|</span><span class="st"> </span>Mito</a>
<a class="sourceLine" id="cb346-23" data-line-number="23">sce.lawlor &lt;-<span class="st"> </span>sce.lawlor[,<span class="op">!</span>discard]</a>
<a class="sourceLine" id="cb346-24" data-line-number="24"></a>
<a class="sourceLine" id="cb346-25" data-line-number="25">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb346-26" data-line-number="26"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb346-27" data-line-number="27"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb346-28" data-line-number="28">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.lawlor)</a>
<a class="sourceLine" id="cb346-29" data-line-number="29">sce.lawlor &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.lawlor, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb346-30" data-line-number="30">sce.lawlor &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce.lawlor)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb347-1" data-line-number="1"><span class="kw">plotExpression</span>(sce.lawlor, <span class="dt">x=</span><span class="st">&quot;cell type&quot;</span>, <span class="dt">features=</span><span class="st">&quot;ENSG00000115263&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:viol-gcg-lawlor"></span>
<img src="P2_W09.marker-detection_files/figure-html/viol-gcg-lawlor-1.png" alt="Distribution of log-normalized expression values for _GCG_ across different pancreatic cell types in GSE86469." width="672" />
<p class="caption">
Figure 13.5: Distribution of log-normalized expression values for <em>GCG</em> across different pancreatic cell types in GSE86469.
</p>
</div>
<p>Another property of the binomial test is that it will not respond to scaling normalization.
Systematic differences in library size between clusters will not be considered when computing <span class="math inline">\(p\)</span>-values or effect sizes.
This is not necessarily problematic for marker gene detection -
users can treat this as retaining information about the total RNA content, analogous to spike-in normalization.</p>
</div>
<div id="using-custom-de-methods" class="section level3">
<h3><span class="header-section-number">13.3.3</span> Using custom DE methods</h3>
<p>It is possible to perform marker gene detection based on precomputed DE statistics.
This allows us to take advantage of more sophisticated tests in dedicated DE analysis packages.
To demonstrate, consider the <code>voom()</code> approach from the <em><a href="https://bioconductor.org/packages/3.10/limma">limma</a></em> package <span class="citation">(Law et al. <a href="#ref-law2014voom">2014</a>)</span>.
We first process our <code>SingleCellExperiment</code> to obtain a <code>fit</code> object as shown below.</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb348-1" data-line-number="1"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb348-2" data-line-number="2">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>cluster, <span class="dt">data=</span><span class="kw">colData</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb348-3" data-line-number="3"><span class="kw">colnames</span>(design)</a></code></pre></div>
<pre><code>##  [1] &quot;cluster1&quot;  &quot;cluster2&quot;  &quot;cluster3&quot;  &quot;cluster4&quot;  &quot;cluster5&quot; 
##  [6] &quot;cluster6&quot;  &quot;cluster7&quot;  &quot;cluster8&quot;  &quot;cluster9&quot;  &quot;cluster10&quot;
## [11] &quot;cluster11&quot; &quot;cluster12&quot;</code></pre>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb350-1" data-line-number="1"><span class="co"># Removing very low-abundance genes.</span></a>
<a class="sourceLine" id="cb350-2" data-line-number="2">keep &lt;-<span class="st"> </span><span class="kw">calculateAverage</span>(sce.pbmc) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.1</span> </a>
<a class="sourceLine" id="cb350-3" data-line-number="3"><span class="kw">summary</span>(keep)</a></code></pre></div>
<pre><code>##    Mode   FALSE    TRUE 
## logical   29387    4307</code></pre>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb352-1" data-line-number="1">y &lt;-<span class="st"> </span><span class="kw">convertTo</span>(sce.pbmc, <span class="dt">subset.row=</span>keep)</a>
<a class="sourceLine" id="cb352-2" data-line-number="2">v &lt;-<span class="st"> </span><span class="kw">voom</span>(y, design)</a>
<a class="sourceLine" id="cb352-3" data-line-number="3">fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(v, design)</a></code></pre></div>
<p>We then perform pairwise comparisons between clusters using the TREAT strategy <span class="citation">(McCarthy and Smyth <a href="#ref-mccarthy2009treat">2009</a>)</span> to test for log-fold changes that are significantly greater than 0.5.
For each comparison, we store the corresponding data frame of statistics in <code>all.results</code>, along with the identities of the clusters involved in <code>all.pairs</code>.</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb353-1" data-line-number="1">nclust &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(sce.pbmc<span class="op">$</span>cluster))</a>
<a class="sourceLine" id="cb353-2" data-line-number="2">all.results &lt;-<span class="st"> </span>all.pairs &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb353-3" data-line-number="3">counter &lt;-<span class="st"> </span>1L</a>
<a class="sourceLine" id="cb353-4" data-line-number="4"></a>
<a class="sourceLine" id="cb353-5" data-line-number="5"><span class="co"># Iterating across the first &#39;nclust&#39; coefficients in design,</span></a>
<a class="sourceLine" id="cb353-6" data-line-number="6"><span class="co"># and comparing them to each other in a pairwise manner.</span></a>
<a class="sourceLine" id="cb353-7" data-line-number="7"><span class="cf">for</span> (x <span class="cf">in</span> <span class="kw">seq_len</span>(nclust)) {</a>
<a class="sourceLine" id="cb353-8" data-line-number="8">    <span class="cf">for</span> (y <span class="cf">in</span> <span class="kw">seq_len</span>(x<span class="op">-</span>1L)) {</a>
<a class="sourceLine" id="cb353-9" data-line-number="9">        con &lt;-<span class="st"> </span><span class="kw">integer</span>(<span class="kw">ncol</span>(design))</a>
<a class="sourceLine" id="cb353-10" data-line-number="10">        con[x] &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb353-11" data-line-number="11">        con[y] &lt;-<span class="st"> </span><span class="dv">-1</span></a>
<a class="sourceLine" id="cb353-12" data-line-number="12">        fit2 &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit, con)</a>
<a class="sourceLine" id="cb353-13" data-line-number="13">        fit2 &lt;-<span class="st"> </span><span class="kw">treat</span>(fit2, <span class="dt">robust=</span><span class="ot">TRUE</span>, <span class="dt">lfc=</span><span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb353-14" data-line-number="14"></a>
<a class="sourceLine" id="cb353-15" data-line-number="15">        res &lt;-<span class="st"> </span><span class="kw">topTreat</span>(fit2, <span class="dt">n=</span><span class="ot">Inf</span>, <span class="dt">sort.by=</span><span class="st">&quot;none&quot;</span>)</a>
<a class="sourceLine" id="cb353-16" data-line-number="16">        all.results[[counter]] &lt;-<span class="st"> </span>res</a>
<a class="sourceLine" id="cb353-17" data-line-number="17">        all.pairs[[counter]] &lt;-<span class="st"> </span><span class="kw">colnames</span>(design)[<span class="kw">c</span>(x, y)]</a>
<a class="sourceLine" id="cb353-18" data-line-number="18">        counter &lt;-<span class="st"> </span>counter<span class="op">+</span>1L</a>
<a class="sourceLine" id="cb353-19" data-line-number="19"></a>
<a class="sourceLine" id="cb353-20" data-line-number="20">        <span class="co"># Also filling the reverse comparison.</span></a>
<a class="sourceLine" id="cb353-21" data-line-number="21">        res<span class="op">$</span>logFC &lt;-<span class="st"> </span><span class="op">-</span>res<span class="op">$</span>logFC</a>
<a class="sourceLine" id="cb353-22" data-line-number="22">        all.results[[counter]] &lt;-<span class="st"> </span>res</a>
<a class="sourceLine" id="cb353-23" data-line-number="23">        all.pairs[[counter]] &lt;-<span class="st"> </span><span class="kw">colnames</span>(design)[<span class="kw">c</span>(y, x)]</a>
<a class="sourceLine" id="cb353-24" data-line-number="24">        counter &lt;-<span class="st"> </span>counter<span class="op">+</span>1L</a>
<a class="sourceLine" id="cb353-25" data-line-number="25">    }</a>
<a class="sourceLine" id="cb353-26" data-line-number="26">}</a></code></pre></div>
<p>These custom results are consolidated into a single marker list for each cluster with the <code>combineMarkers()</code> function.
This combines test statistics across all pairwise comparisons involving a single cluster,
yielding a per-cluster <code>DataFrame</code> that can be interpreted in the same manner as discussed previously.</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb354-1" data-line-number="1">all.pairs &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, all.pairs)</a>
<a class="sourceLine" id="cb354-2" data-line-number="2">combined &lt;-<span class="st"> </span><span class="kw">combineMarkers</span>(all.results, all.pairs, <span class="dt">pval.field=</span><span class="st">&quot;P.Value&quot;</span>)</a>
<a class="sourceLine" id="cb354-3" data-line-number="3"></a>
<a class="sourceLine" id="cb354-4" data-line-number="4"><span class="co"># Inspecting results for our cluster of interest again.</span></a>
<a class="sourceLine" id="cb354-5" data-line-number="5">interesting.voom &lt;-<span class="st"> </span>combined[[<span class="kw">paste0</span>(<span class="st">&quot;cluster&quot;</span>, chosen)]] </a>
<a class="sourceLine" id="cb354-6" data-line-number="6"><span class="kw">colnames</span>(interesting.voom)</a></code></pre></div>
<pre><code>##  [1] &quot;Top&quot;             &quot;p.value&quot;         &quot;FDR&quot;            
##  [4] &quot;logFC.cluster1&quot;  &quot;logFC.cluster2&quot;  &quot;logFC.cluster3&quot; 
##  [7] &quot;logFC.cluster4&quot;  &quot;logFC.cluster5&quot;  &quot;logFC.cluster6&quot; 
## [10] &quot;logFC.cluster7&quot;  &quot;logFC.cluster8&quot;  &quot;logFC.cluster10&quot;
## [13] &quot;logFC.cluster11&quot; &quot;logFC.cluster12&quot;</code></pre>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb356-1" data-line-number="1"><span class="kw">head</span>(interesting.voom[,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 3 columns
##                Top   p.value       FDR
##          &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## PDZK1IP1         1         0         0
## RGS18            1         0         0
## SLC40A1          1         0         0
## SDPR             1         0         0
## PF4              1         0         0
## PPBP             1         0         0</code></pre>
<p>By default, we do not use custom DE methods to perform marker detection, for several reasons.
Many of these methods rely on empirical Bayes shrinkage to share information across genes in the presence of limited replication.
However, this is unnecessary when one treats the many cells in a group as “replicates” of each other (see Section ???).
These methods also make stronger assumptions about the data (e.g., equal variances for linear models, the distribution of variances during empirical Bayes) that are more likely to be violated in noisy scRNA-seq contexts.
From a practical perspective, they require more work to set up and take more time to run.
Nonetheless, some custom methods (e.g., <em><a href="https://bioconductor.org/packages/3.10/MAST">MAST</a></em>) may provide a useful point of difference from the simpler tests, in which case they can be converted into a marker detection scheme as described above.</p>
</div>
</div>
<div id="handling-blocking-factors" class="section level2">
<h2><span class="header-section-number">13.4</span> Handling blocking factors</h2>
</div>
<div id="using-the-block-argument" class="section level2">
<h2><span class="header-section-number">13.5</span> Using the <code>block=</code> argument</h2>
<p>Large studies may contain factors of variation that are known and not interesting (e.g., batch effects, sex differences).
If these are not modelled, they can interfere with marker gene detection - most obviously by inflating the variance within each cluster, but also by distorting the log-fold changes if the cluster composition varies across levels of the blocking factor.
To avoid these issues, we set the <code>block=</code> argument in the <code>findMarkers()</code> call, as demonstrated below for the 416B data set.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb358-1" data-line-number="1">### loading <span class="al">###</span></a>
<a class="sourceLine" id="cb358-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb358-3" data-line-number="3">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">LunSpikeInData</span>(<span class="dt">which=</span><span class="st">&quot;416b&quot;</span>) </a>
<a class="sourceLine" id="cb358-4" data-line-number="4"></a>
<a class="sourceLine" id="cb358-5" data-line-number="5">### gene-annotation <span class="al">###</span></a>
<a class="sourceLine" id="cb358-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb358-7" data-line-number="7">ens.mm.v97 &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73905&quot;</span>]]</a>
<a class="sourceLine" id="cb358-8" data-line-number="8"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL &lt;-<span class="st"> </span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb358-9" data-line-number="9"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb358-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SYMBOL&quot;</span>)</a>
<a class="sourceLine" id="cb358-11" data-line-number="11"><span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rownames</span>(sce<span class="fl">.416</span>b),</a>
<a class="sourceLine" id="cb358-12" data-line-number="12">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>)</a>
<a class="sourceLine" id="cb358-13" data-line-number="13"></a>
<a class="sourceLine" id="cb358-14" data-line-number="14"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb358-15" data-line-number="15"><span class="kw">rownames</span>(sce<span class="fl">.416</span>b) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>ENSEMBL, </a>
<a class="sourceLine" id="cb358-16" data-line-number="16">    <span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SYMBOL)</a>
<a class="sourceLine" id="cb358-17" data-line-number="17"></a>
<a class="sourceLine" id="cb358-18" data-line-number="18">### quality-control <span class="al">###</span></a>
<a class="sourceLine" id="cb358-19" data-line-number="19">mito &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rowData</span>(sce<span class="fl">.416</span>b)<span class="op">$</span>SEQNAME<span class="op">==</span><span class="st">&quot;MT&quot;</span>)</a>
<a class="sourceLine" id="cb358-20" data-line-number="20">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce<span class="fl">.416</span>b, <span class="dt">feature_controls=</span><span class="kw">list</span>(<span class="dt">Mt=</span>mito))</a>
<a class="sourceLine" id="cb358-21" data-line-number="21"></a>
<a class="sourceLine" id="cb358-22" data-line-number="22">combined &lt;-<span class="st"> </span><span class="kw">paste0</span>(sce<span class="fl">.416</span>b<span class="op">$</span>block, <span class="st">&quot;:&quot;</span>, sce<span class="fl">.416</span>b<span class="op">$</span>phenotype)</a>
<a class="sourceLine" id="cb358-23" data-line-number="23">libsize.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_counts, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, </a>
<a class="sourceLine" id="cb358-24" data-line-number="24">    <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb358-25" data-line-number="25">feature.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>total_features_by_counts, <span class="dt">nmads=</span><span class="dv">3</span>, </a>
<a class="sourceLine" id="cb358-26" data-line-number="26">    <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb358-27" data-line-number="27">spike.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(sce<span class="fl">.416</span>b<span class="op">$</span>pct_counts_ERCC, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>,</a>
<a class="sourceLine" id="cb358-28" data-line-number="28">    <span class="dt">batch=</span>combined)</a>
<a class="sourceLine" id="cb358-29" data-line-number="29"></a>
<a class="sourceLine" id="cb358-30" data-line-number="30">keep &lt;-<span class="st"> </span><span class="op">!</span>(libsize.drop <span class="op">|</span><span class="st"> </span>feature.drop <span class="op">|</span><span class="st"> </span>spike.drop)</a>
<a class="sourceLine" id="cb358-31" data-line-number="31">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span>sce<span class="fl">.416</span>b[,keep]</a>
<a class="sourceLine" id="cb358-32" data-line-number="32"></a>
<a class="sourceLine" id="cb358-33" data-line-number="33">### normalization <span class="al">###</span></a>
<a class="sourceLine" id="cb358-34" data-line-number="34"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb358-35" data-line-number="35">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb358-36" data-line-number="36">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">computeSpikeFactors</span>(sce<span class="fl">.416</span>b, <span class="dt">general.use=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb358-37" data-line-number="37">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb358-38" data-line-number="38"></a>
<a class="sourceLine" id="cb358-39" data-line-number="39">### variance-modelling <span class="al">###</span></a>
<a class="sourceLine" id="cb358-40" data-line-number="40">fit<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce<span class="fl">.416</span>b, <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb358-41" data-line-number="41">dec<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce<span class="fl">.416</span>b, fit<span class="fl">.416</span>b)</a>
<a class="sourceLine" id="cb358-42" data-line-number="42"></a>
<a class="sourceLine" id="cb358-43" data-line-number="43"><span class="co"># Taking all of the genes with positive biological components.</span></a>
<a class="sourceLine" id="cb358-44" data-line-number="44">chosen.hvgs &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec<span class="fl">.416</span>b)[dec<span class="fl">.416</span>b<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb358-45" data-line-number="45"></a>
<a class="sourceLine" id="cb358-46" data-line-number="46">### batch-correction <span class="al">###</span></a>
<a class="sourceLine" id="cb358-47" data-line-number="47"><span class="co"># Composition of the plates is expected to be the same,</span></a>
<a class="sourceLine" id="cb358-48" data-line-number="48"><span class="co"># hence the use of a simple &#39;removeBatchEffect&#39;.</span></a>
<a class="sourceLine" id="cb358-49" data-line-number="49"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb358-50" data-line-number="50"><span class="kw">assay</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;corrected&quot;</span>) &lt;-<span class="st"> </span><span class="kw">removeBatchEffect</span>(<span class="kw">logcounts</span>(sce<span class="fl">.416</span>b), </a>
<a class="sourceLine" id="cb358-51" data-line-number="51">    <span class="dt">design=</span><span class="kw">model.matrix</span>(<span class="op">~</span>sce<span class="fl">.416</span>b<span class="op">$</span>phenotype), <span class="dt">batch=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb358-52" data-line-number="52"></a>
<a class="sourceLine" id="cb358-53" data-line-number="53">### dimensionality-reduction <span class="al">###</span></a>
<a class="sourceLine" id="cb358-54" data-line-number="54">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sce<span class="fl">.416</span>b, <span class="dt">technical=</span>dec<span class="fl">.416</span>b, </a>
<a class="sourceLine" id="cb358-55" data-line-number="55">    <span class="dt">subset.row=</span>chosen.hvgs, <span class="dt">assay.type=</span><span class="st">&quot;corrected&quot;</span>)</a>
<a class="sourceLine" id="cb358-56" data-line-number="56"></a>
<a class="sourceLine" id="cb358-57" data-line-number="57"><span class="kw">set.seed</span>(<span class="dv">1010</span>)</a>
<a class="sourceLine" id="cb358-58" data-line-number="58">sce<span class="fl">.416</span>b &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce<span class="fl">.416</span>b, <span class="dt">use_dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb358-59" data-line-number="59"></a>
<a class="sourceLine" id="cb358-60" data-line-number="60">### clustering <span class="al">###</span></a>
<a class="sourceLine" id="cb358-61" data-line-number="61">my.dist &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">reducedDim</span>(sce<span class="fl">.416</span>b, <span class="st">&quot;PCA&quot;</span>))</a>
<a class="sourceLine" id="cb358-62" data-line-number="62">my.tree &lt;-<span class="st"> </span><span class="kw">hclust</span>(my.dist, <span class="dt">method=</span><span class="st">&quot;ward.D2&quot;</span>)</a>
<a class="sourceLine" id="cb358-63" data-line-number="63"></a>
<a class="sourceLine" id="cb358-64" data-line-number="64"><span class="kw">library</span>(dynamicTreeCut)</a>
<a class="sourceLine" id="cb358-65" data-line-number="65">my.clusters &lt;-<span class="st"> </span><span class="kw">unname</span>(<span class="kw">cutreeDynamic</span>(my.tree, <span class="dt">distM=</span><span class="kw">as.matrix</span>(my.dist),</a>
<a class="sourceLine" id="cb358-66" data-line-number="66">    <span class="dt">minClusterSize=</span><span class="dv">10</span>, <span class="dt">verbose=</span><span class="dv">0</span>))</a>
<a class="sourceLine" id="cb358-67" data-line-number="67">sce<span class="fl">.416</span>b<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">factor</span>(my.clusters)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb359-1" data-line-number="1">m.out &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce<span class="fl">.416</span>b, sce<span class="fl">.416</span>b<span class="op">$</span>cluster, </a>
<a class="sourceLine" id="cb359-2" data-line-number="2">                     <span class="dt">block=</span>sce<span class="fl">.416</span>b<span class="op">$</span>block, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>) </a></code></pre></div>
<p>For each gene, each pairwise comparion between clusters is performed separately in each level of the blocking factor - in this case, the plate of origin.
The function will then combine <span class="math inline">\(p\)</span>-values from different plates using Stouffer’s Z method to obtain a single <span class="math inline">\(p\)</span>-value per pairwise comparison.
(These <span class="math inline">\(p\)</span>-values are further combined across comparisons to obtain a single <span class="math inline">\(p\)</span>-value per gene, using either Simes’ method or an intersection-union test depending on the value of <code>pval.type=</code>.)
This approach favours genes that exhibit consistent DE in the same direction in each plate.</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb360-1" data-line-number="1">demo &lt;-<span class="st"> </span>m.out[[<span class="st">&quot;1&quot;</span>]] </a>
<a class="sourceLine" id="cb360-2" data-line-number="2">demo[demo<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 24 rows and 3 columns
##                          Top              p.value                  FDR
##                    &lt;integer&gt;            &lt;numeric&gt;            &lt;numeric&gt;
## Cenpk                      1 3.64842075891759e-38 1.70031001048595e-33
## Uhrf1                      1 1.27436092404157e-31 2.96951582520165e-27
## Pkp4                       1 2.10780869575213e-27 1.63720527428053e-23
## ENSMUSG00000082279         1 1.62145081562279e-23 5.03697869568323e-20
## Hbb-bt                     1 1.47303160270372e-21 2.14528640038763e-18
## ...                      ...                  ...                  ...
## Rfc5                       5 1.43735740166193e-26 8.37332554338155e-23
## Elp4                       5 1.85254656015266e-23 5.07859293466794e-20
## ENSMUSG00000082043         5 4.23815536237289e-19 3.46517530715835e-16
## Cxcr5                      5 2.68967679921673e-15 9.35445504109675e-13
## Kcnk1                      5  7.8504116345369e-13 1.35880722663217e-10</code></pre>
<p>The <code>block=</code> argument works with all tests shown above and is robust to difference in the log-fold changes or variance between batches.
However, it assumes that each pair of clusters is present in at least one batch.
In scenarios where cells from two clusters never co-occur in the same batch, the comparison will be impossible and <code>NA</code>s will be reported in the output.</p>
</div>
<div id="using-the-design-argument" class="section level2">
<h2><span class="header-section-number">13.6</span> Using the <code>design=</code> argument</h2>
<p>Another approach is to define a design matrix containing the batch of origin as the sole factor.
<code>findMarkers()</code> will then fit a linear model to the log-expression values, similar to the use of <em><a href="https://bioconductor.org/packages/3.10/limma">limma</a></em> for bulk RNA sequencing data <span class="citation">(Ritchie et al. <a href="#ref-ritchie2015limma">2015</a>)</span>.
This handles situations where multiple batches contain unique clusters, as comparisons can be implicitly performed via shared cell types in each batch.
There is also a slight increase in power when information is shared across clusters for variance estimation.</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb362-1" data-line-number="1"><span class="co"># Setting up the design matrix (we remove intercept for full rank</span></a>
<a class="sourceLine" id="cb362-2" data-line-number="2"><span class="co"># in the final design matrix with the cluster-specific terms).</span></a>
<a class="sourceLine" id="cb362-3" data-line-number="3">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span>sce<span class="fl">.416</span>b<span class="op">$</span>block)</a>
<a class="sourceLine" id="cb362-4" data-line-number="4">design &lt;-<span class="st"> </span>design[,<span class="op">-</span><span class="dv">1</span>,drop=<span class="ot">FALSE</span>]</a>
<a class="sourceLine" id="cb362-5" data-line-number="5"></a>
<a class="sourceLine" id="cb362-6" data-line-number="6">m.alt &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce<span class="fl">.416</span>b, sce<span class="fl">.416</span>b<span class="op">$</span>cluster, </a>
<a class="sourceLine" id="cb362-7" data-line-number="7">    <span class="dt">design=</span>design, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb362-8" data-line-number="8">demo &lt;-<span class="st"> </span>m.alt[[<span class="st">&quot;1&quot;</span>]]</a>
<a class="sourceLine" id="cb362-9" data-line-number="9">demo[demo<span class="op">$</span>Top <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## DataFrame with 18 rows and 3 columns
##                          Top              p.value                  FDR
##                    &lt;integer&gt;            &lt;numeric&gt;            &lt;numeric&gt;
## ENSMUSG00000081664         1 1.91285355952216e-41 8.91466272879706e-37
## Pmf1                       1 1.12621582896182e-36  1.3121540623234e-32
## Ccna2                      1  1.4320290224626e-35 1.11230467604745e-31
## Hbb-bt                     1   1.144389533167e-28 1.90475463584696e-25
## ENSMUSG00000082610         1 9.85538881311309e-27 1.17769369293929e-23
## ...                      ...                  ...                  ...
## Tuba8                      4 1.12209603758146e-17 3.09432921511517e-15
## Top2a                      5 2.69761684182245e-36 2.51439470592585e-32
## ENSMUSG00000084220         5 1.41054844370574e-29 2.98805453047556e-26
## Mfsd2b                     5 1.77368012038614e-28 2.85036511484399e-25
## ENSMUSG00000084340         5 3.14925058189259e-13  4.0882360478697e-11</code></pre>
<p>The use of a linear model makes some strong assumptions, necessitating some caution when interpreting the results.
If the batch effect is not consistent across clusters, the variance will be inflated and the log-fold change estimates will be distorted.
Variances are also assumed to be equal across groups, which is not true in general.
In particular, the presence of clusters in which a gene is silent will shrink the residual variance towards zero, preventing the model from penalizing genes with high variance in other clusters.
Thus, we generally recommend the use of <code>block=</code> where possible.</p>
</div>
<div id="invalidity-of-p-values" class="section level2">
<h2><span class="header-section-number">13.7</span> Invalidity of <span class="math inline">\(p\)</span>-values</h2>
<div id="from-data-snooping" class="section level3">
<h3><span class="header-section-number">13.7.1</span> From data snooping</h3>
<p>All of our DE strategies for detecting marker genes between clusters are statistically flawed to some extent.
The DE analysis is performed on the same data used to obtain the clusters, which represents “data dredging” (also known as fishing or data snooping).
The hypothesis of interest - that are there differences between clusters? - is formulated from the data, so we are more likely to get a positive result when we re-use the data set to test that hypothesis.</p>
<p>The practical effect of data dredging is best illustrated with a simple simulation.
We simulate i.i.d. normal values, perform <span class="math inline">\(k\)</span>-means clustering and test for DE between clusters of cells with <code>findMarkers()</code>.
The resulting distribution of <span class="math inline">\(p\)</span>-values is heavily skewed towards low values (Figure <a href="marker-gene-detection.html#fig:pval-dist">13.6</a>).
Thus, we can detect “significant” differences between clusters even in the absence of any real substructure in the data.
This effect arises from the fact that clustering, by definition, yields groups of cells that are separated in expression space.
Testing for DE genes between clusters will inevitably yield some significant results as that is how the clusters were defined.</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb364-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb364-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb364-3" data-line-number="3">y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">100000</span>), <span class="dt">ncol=</span><span class="dv">200</span>)</a>
<a class="sourceLine" id="cb364-4" data-line-number="4">clusters &lt;-<span class="st"> </span><span class="kw">kmeans</span>(<span class="kw">t</span>(y), <span class="dt">centers=</span><span class="dv">2</span>)<span class="op">$</span>cluster</a>
<a class="sourceLine" id="cb364-5" data-line-number="5">out &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(y, clusters)</a>
<a class="sourceLine" id="cb364-6" data-line-number="6"><span class="kw">hist</span>(out[[<span class="dv">1</span>]]<span class="op">$</span>p.value, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;p-value&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:pval-dist"></span>
<img src="P2_W09.marker-detection_files/figure-html/pval-dist-1.png" alt="Distribution of $p$-values from a DE analysis between two clusters in a simulation with no true subpopulation structure." width="672" />
<p class="caption">
Figure 13.6: Distribution of <span class="math inline">\(p\)</span>-values from a DE analysis between two clusters in a simulation with no true subpopulation structure.
</p>
</div>
<p>For marker gene detection, this effect is largely harmless as the <span class="math inline">\(p\)</span>-values are used only for ranking.
However, it becomes an issue when the <span class="math inline">\(p\)</span>-values are used to define “significant differences” between clusters with respect to an error rate threshold.
Meaningful interpretation of error rates require consideration of the long-run behaviour, i.e., the rate of incorrect rejections if the experiment were repeated many times.
The concept of statistical significance for differences between clusters is not applicable if clusters and their interpretations are not stably reproducible across (hypothetical) replicate experiments.</p>
</div>
<div id="nature-of-replication" class="section level3">
<h3><span class="header-section-number">13.7.2</span> Nature of replication</h3>
<p>The naive application of DE analysis methods will treat counts from the same cluster of cells as replicate observations.
This is not the most relevant level of replication when cells are derived from the same biological sample (i.e., cell culture, animal or patient).
DE analyses that treat cells as replicates fail to properly model the sample-to-sample variability <span class="citation">(A. T. L. Lun and Marioni <a href="#ref-lun2017overcoming">2017</a>)</span>.
The latter is arguably the more important level of replication as different samples will necessarily be generated if the experiment is to be replicated.
Indeed, the use of cells as replicates only masks the fact that the sample size is actually one in an experiment involving a single biological sample.
This reinforces the inappropriateness of using the marker gene <span class="math inline">\(p\)</span>-values to make statements about statistical inference.</p>
</div>
</div>
<div id="session-info-7" class="section level2">
<h2><span class="header-section-number">13.8</span> Session Info</h2>
<pre><code>## &lt;button class=&quot;aaron-collapse&quot;&gt;View session info&lt;/button&gt;
## &lt;div class=&quot;aaron-content&quot;&gt;
## ```
## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.6 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] limma_3.41.15               scater_1.13.9              
##  [3] ggplot2_3.2.0               pheatmap_1.0.12            
##  [5] scran_1.13.9                SingleCellExperiment_1.7.0 
##  [7] SummarizedExperiment_1.15.5 DelayedArray_0.11.4        
##  [9] BiocParallel_1.19.0         matrixStats_0.54.0         
## [11] Biobase_2.45.0              GenomicRanges_1.37.14      
## [13] GenomeInfoDb_1.21.1         IRanges_2.19.10            
## [15] S4Vectors_0.23.17           BiocGenerics_0.31.5        
## [17] BiocStyle_2.13.2            Cairo_1.5-10               
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.5.1            dynamicTreeCut_1.63-1   
##  [3] edgeR_3.27.9             BiocSingular_1.1.5      
##  [5] viridisLite_0.3.0        DelayedMatrixStats_1.7.1
##  [7] assertthat_0.2.1         statmod_1.4.32          
##  [9] BiocManager_1.30.4       highr_0.8               
## [11] dqrng_0.2.1              GenomeInfoDbData_1.2.1  
## [13] vipor_0.4.5              yaml_2.2.0              
## [15] pillar_1.4.2             lattice_0.20-38         
## [17] glue_1.3.1               digest_0.6.20           
## [19] RColorBrewer_1.1-2       XVector_0.25.0          
## [21] colorspace_1.4-1         cowplot_1.0.0           
## [23] htmltools_0.3.6          Matrix_1.2-17           
## [25] pkgconfig_2.0.2          bookdown_0.12           
## [27] zlibbioc_1.31.0          purrr_0.3.2             
## [29] scales_1.0.0             tibble_2.1.3            
## [31] withr_2.1.2              lazyeval_0.2.2          
## [33] magrittr_1.5             crayon_1.3.4            
## [35] evaluate_0.14            beeswarm_0.2.3          
## [37] tools_3.6.0              stringr_1.4.0           
## [39] munsell_0.5.0            locfit_1.5-9.1          
## [41] irlba_2.3.3              compiler_3.6.0          
## [43] rsvd_1.0.2               rlang_0.4.0             
## [45] grid_3.6.0               RCurl_1.95-4.12         
## [47] BiocNeighbors_1.3.3      igraph_1.2.4.1          
## [49] labeling_0.3             bitops_1.0-6            
## [51] rmarkdown_1.14           gtable_0.3.0            
## [53] codetools_0.2-16         R6_2.4.0                
## [55] gridExtra_2.3            knitr_1.23              
## [57] dplyr_0.8.3              stringi_1.4.3           
## [59] ggbeeswarm_0.6.0         Rcpp_1.0.2              
## [61] tidyselect_0.2.5         xfun_0.8                
## ```
## &lt;/div&gt;</code></pre>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-berger1996bioequivalence">
<p>Berger, R. L., and J. C. Hsu. 1996. “Bioequivalence Trials, Intersection-Union Tests and Equivalence Confidence Sets.” <em>Statist. Sci.</em> 11 (4). The Institute of Mathematical Statistics:283–319. <a href="https://doi.org/10.1214/ss/1032280304" class="uri">https://doi.org/10.1214/ss/1032280304</a>.</p>
</div>
<div id="ref-law2014voom">
<p>Law, C. W., Y. Chen, W. Shi, and G. K. Smyth. 2014. “voom: Precision weights unlock linear model analysis tools for RNA-seq read counts.” <em>Genome Biol.</em> 15 (2):R29.</p>
</div>
<div id="ref-lawlor2017singlecell">
<p>Lawlor, N., J. George, M. Bolisetty, R. Kursawe, L. Sun, V. Sivakamasundari, I. Kycia, P. Robson, and M. L. Stitzel. 2017. “Single-cell transcriptomes identify human islet cell signatures and reveal cell-type-specific expression changes in type 2 diabetes.” <em>Genome Res.</em> 27 (2):208–22.</p>
</div>
<div id="ref-lun2017overcoming">
<p>Lun, A. T. L., and J. C. Marioni. 2017. “Overcoming confounding plate effects in differential expression analyses of single-cell RNA-seq data.” <em>Biostatistics</em> 18 (3):451–64.</p>
</div>
<div id="ref-mccarthy2009treat">
<p>McCarthy, D. J., and G. K. Smyth. 2009. “Testing significance relative to a fold-change threshold is a TREAT.” <em>Bioinformatics</em> 25 (6):765–71.</p>
</div>
<div id="ref-ritchie2015limma">
<p>Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K. Smyth. 2015. “limma powers differential expression analyses for RNA-sequencing and microarray studies.” <em>Nucleic Acids Res.</em> 43 (7):e47.</p>
</div>
<div id="ref-simes1986improved">
<p>Simes, R. J. 1986. “An Improved Bonferroni Procedure for Multiple Tests of Significance.” <em>Biometrika</em> 73 (3):751–54.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="17">
<li id="fn17"><p>That is, the lowest rank value. So 1st is 1, and 2nd is 2, so 1 is higher than 2. Geez.<a href="marker-gene-detection.html#fnref17" class="footnote-back">↩</a></p></li>
<li id="fn18"><p>Standard operating procedure is to (i) experience a brief but crushing bout of disappointment due to the poor quality of upregulated candidate markers, (ii) rage-quit, and (iii) remember to check the genes that are changing in the other direction.<a href="marker-gene-detection.html#fnref18" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="clustering-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="integrating-datasets-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W09.marker-detection.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
