# Data Infrastructure {#data-infrastructure}



One of the advantages of using Bioconductor packages is that they utilize common data infrastructures which makes analyses interoperable across various packages. Furthermore, much engineering effort is put into making this infrastructure robust and scalable. Here, we describe the _SingleCellExperiment_ object (or `sce` in shorthand) in detail to describe how it is constructed, utilized in downstream analysis, and how it stores various types of primary data and metadata.

## Prerequisites

The Bioconductor package *SingleCellExperiment* provides the *SingleCellExperiment* class for usage. While the package is implicitly installed and loaded when using any package that depends on the *SingleCellExperiment* class, it can be explicitly installed (and loaded) as follows:


```r
BiocManager::install('SingleCellExperiment')
```

Additionally, we use some functions from the `scater` and `scran` packages, as well as the CRAN package `uwot` (which conveniently can also be installed through `BiocManager`). These functions will be accessed through the `<package>::<function>` convention as needed.


```r
BiocManager::install(c('scater', 'scran', 'uwot', 'Rtsne'))
```

For this session, all we will need loaded is the *SingleCellExperiment* package:


```r
library(SingleCellExperiment)
library(scater)
library(scran)
library(uwot)
library(Rtsne)
```


## The *SingleCellExperiment* Class

![Overview of the  _SingleCellExperiment_ class object](images/SingleCellExperiment.png)

### Primary Data: The `assays` Slot

The *SingleCellExperiment* (`sce`) object is the basis of single-cell analytical applications based in Bioconductor. The `sce` object is an [S4 object](https://adv-r.hadley.nz/s4.html), which in essence provides a more formalized approach towards construction and accession of data compared to other methods available in R. The utility of S4 comes from validity checks that ensure safe data manipulation, and most important for our discussion, from its extensibility through _slots_.

If we imagine the `sce` object to be a ship, the _slots_ of `sce` can be thought of as individual cargo boxes - each exists as a separate entity _within_ the `sce` object. Furthermore, each slot contains data that arrives in its own format. To extend the metaphor, we can imagine that different variations of cargo boxes are required for fruits versus bricks. In the case of `sce`, certain slots expect numeric matrices, whereas others may expect data frames.

To construct a rudimentary `sce` object, all we need is a single slot:

* `assays` slot: contains primary data such as counts in a list, where each entry of the list is in a matrix format, where rows correspond to features (genes) and columns correspond to samples (cells) (_Figure 1A, blue box_)

Let's start simple by generating three cells worth of count data across ten genes.


```r
counts_matrix <- data.frame(cell_1 = rpois(10, 10), 
                    cell_2 = rpois(10, 10), 
                    cell_3 = rpois(10, 30))
rownames(counts_matrix) <- paste0("gene_", 1:10)
counts_matrix <- as.matrix(counts_matrix) # must be a matrix object!
```

From this, we can now construct our first *SingleCellExperiment* object, using the defined _constructor_, `SingleCellExperiment()`. Note that we provide our data as a *named* _list_, and each entry of the list is a _matrix_. Here, we name the `counts_matrix` entry as simply `counts` within the list.


```r
sce <- SingleCellExperiment(assays = list(counts = counts_matrix))
```

To inspect the object, we can simply type `sce` into the console to see some pertinent information, which will display an overview of the various slots available to us (which may or may not have any data).


```r
sce
```

```
## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

To access the count data we just supplied, we can do any one of the following:

* `assay(sce, "counts")` - this is the most general method, where we can supply the name of the assay as the second argument.
* `counts(sce)` - this is the same as the above, but *only* works for assays with the special name `"counts"`.


```r
counts(sce)
```

```
##         cell_1 cell_2 cell_3
## gene_1      12     11     30
## gene_2       9     13     25
## gene_3       9     17     33
## gene_4       9     11     25
## gene_5       8     13     21
## gene_6      12      7     43
## gene_7      10     11     24
## gene_8      13     15     33
## gene_9      10      6     25
## gene_10     11     14     34
```

```r
## assay(sce, "counts") ## same as above in this special case
```

### Extending the `assays` Slot

What makes the `assay` slot especially powerful is that it can hold _multiple_ representations of the primary data. This is especially useful for storing the raw as well as a _normalized_ version of the data. We can do just that as shown below, using the `scran` and `scater` packages to compute a normalized and log-transformed representation of the initial primary data.

Note that here, we overwrite our previous `sce` upon reassigning the results to `sce` - this is because these functions _return_ a *SingleCellExperiment* object. Some functions - especially those outside of single-cell oriented Bioconductor packages - do not, in which case you will need to append your results to the `sce` object (see below).


```r
sce <- computeSumFactors(sce)
sce <- logNormCounts(sce)
```

Viewing the object again, we see that these functions added some new entries:


```r
sce
```

```
## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(2): counts logcounts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

Specifically, we see that the `assays` slot has grown to be comprised of two entries: `counts` (our initial data) and `logcounts` (the normalized data). Similar to `counts`, the `logcounts` name is a special name which lets us access it simply by typing `logcounts(sce)`, although the longhand version works just as well.


```r
logcounts(sce)
```

```
##         cell_1 cell_2 cell_3
## gene_1    4.39   4.09   4.21
## gene_2    4.00   4.31   3.97
## gene_3    4.00   4.68   4.34
## gene_4    4.00   4.09   3.97
## gene_5    3.84   4.31   3.73
## gene_6    4.39   3.48   4.71
## gene_7    4.14   4.09   3.91
## gene_8    4.50   4.51   4.34
## gene_9    4.14   3.28   3.97
## gene_10   4.27   4.41   4.38
```

```r
## assay(sce, "logcounts") ## same as above
```

Notice that the data before had a severe discrepancy in counts between cells 1/2 versus 3, and that normalization has ameliorated this difference.

To look at all the available assays within `sce`, we can type:


```r
assays(sce)
```

```
## List of length 2
## names(2): counts logcounts
```

While the functions above demonstrate automatic addition of assays to our `sce` object, there may be cases where we want to perform our own calculations and save the result into the `assays` slot. In particular, this is important for using functions that do _not_ return a *SingleCellExperiment* object. 

Let's append a new version of the data that has been offset by `+100`.


```r
counts_100 <- assay(sce, "counts") + 100
assay(sce, "counts_100") <- counts_100 # assign a new entry to assays slot
```

Then we can use the accessor `assays()` (notice this is plural!) to see all our entries into the `assay` slot that we have made so far. Note that to see all the assays, we use the plural `assays()` accessor, and to retrieve a single assay entry (as a matrix) we use the singular `assay()` accessor, providing the name of the assay we wish to retrieve as above.
 

```r
assays(sce)
```

```
## List of length 3
## names(3): counts logcounts counts_100
```

These entries are also seen on the default view of `sce`:


```r
sce
```

```
## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(3): counts logcounts counts_100
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

This sort of extension of the `assays` slot is represented graphically in _Figure 1B (dark blue box)_, showing the addition of the `logcounts` matrix into the `assays` slot.

In a similar manner, many of the slots of `sce` are extendable through assignment as shown above, thus allowing for myriad custom functionality as needed for interoperability with functions outside of single-cell oriented Bioconductor packages.


### Column (Meta)Data: `colData` Slot

To further annotate our `sce` object, one of the first and most useful pieces of information is adding metadata that describes the columns of our primary data, e.g., the samples or cells of our experiment. This data is entered into the `colData` slot:

* `colData` slot: metadata that describes that samples (cells) provided as a `data.frame` (or `DataFrame` if appending), where rows correspond to cells, and columns correspond to the sample (cells) metadata features (e.g. id, batch, author, etc.) (_Figure 1A, orange box_).

So, let's come up with some metadata for the cells, starting with a batch variable, where cells 1 and 2 are in batch 1, and cell 3 is from batch 2.


```r
cell_metadata <- data.frame(batch = c(1, 1, 2))
rownames(cell_metadata) <- paste0("cell_", 1:3)
```

Now, we can take two approaches - either append the `cell_metadata` to our existing `sce`, or start from scratch via the `SingleCellExperiment()` constructor and provide it from the get go. We'll start from scratch for now, but will also show how to append the data:


```r
## From scratch:
sce <- SingleCellExperiment(assays = list(counts = counts_matrix),
                           colData = cell_metadata)

## Appending to existing object (requires DataFrame() coercion)
## colData(sce) <- DataFrame(cell_metadata)
```

Similar to `assays`, we can see our colData is now populated from the default view of `sce`:


```r
sce
```

```
## class: SingleCellExperiment 
## dim: 10 3 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(3): cell_1 cell_2 cell_3
## colData names(1): batch
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

And furthermore access our column (meta)data with the accessor, `colData()`:


```r
colData(sce)
```

```
## DataFrame with 3 rows and 1 column
##            batch
##        <numeric>
## cell_1         1
## cell_2         1
## cell_3         2
```

Finally, some packages automatically add to the `colData` slot, for example, the `scater` package features a function, `addPerCellQC()`, which appends a lot of quality control data. Here we show the first five columns of `colData(sce)` with the quality control metrics appended to it.


```r
sce <- addPerCellQC(sce)
colData(sce)[, 1:5]
```

```
## DataFrame with 3 rows and 5 columns
##            batch       sum  detected percent_top_50 percent_top_100
##        <numeric> <integer> <integer>      <numeric>       <numeric>
## cell_1         1       103        10            100             100
## cell_2         1       118        10            100             100
## cell_3         2       293        10            100             100
```

#### Using `colData` for Subsetting

A common operation with `colData` is its use in subsetting. One simple way to access `colData` is through the use of the `$` operator, which is a shortcut for accessing a variable within the `colData` slot:


```r
sce$batch
```

```
## [1] 1 1 2
```

```r
## colData(sce)$batch # same as above
```

If we only wanted cells within batch 1, we could subset our sce object as follows (remember, we subset on the _columns_ in this case because we are filtering by cells/samples here).


```r
sce[, sce$batch == 1]
```

```
## class: SingleCellExperiment 
## dim: 10 2 
## metadata(0):
## assays(1): counts
## rownames(10): gene_1 gene_2 ... gene_9 gene_10
## rowData names(0):
## colnames(2): cell_1 cell_2
## colData names(8): batch sum ... percent_top_500 total
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

### Feature Metadata: `rowData`/`rowRanges`

Lastly, the rows also have their own metadata slot to store information that pertains to the features of the `sce` object:

* `rowData` slot: contains data in a `data.frame` (`DataFrame`) format that describes aspects of the data corresponding to the rows of the primary data (_Figure 1A, green box_). 

Furthermore, there is a special slot which pertains to features with genomic coordinates:

* `rowRanges` slot: contains data in a `GRangesList` (where each entry is a `GenomicRanges` format) that describes the chromosome, start, and end coordinates of the features (genes, genomic regions).

Both of these can be accessed via their respective accessors, `rowRanges()` and `rowData()`. In our case, `rowRanges(sce)` produces an empty list:


```r
rowRanges(sce) # empty
```

```
## GRangesList object of length 10:
## $gene_1
## GRanges object with 0 ranges and 0 metadata columns:
##    seqnames    ranges strand
##       <Rle> <IRanges>  <Rle>
##   -------
##   seqinfo: no sequences
## 
## $gene_2
## GRanges object with 0 ranges and 0 metadata columns:
##    seqnames    ranges strand
##       <Rle> <IRanges>  <Rle>
##   -------
##   seqinfo: no sequences
## 
## $gene_3
## GRanges object with 0 ranges and 0 metadata columns:
##    seqnames    ranges strand
##       <Rle> <IRanges>  <Rle>
##   -------
##   seqinfo: no sequences
## 
## ...
## <7 more elements>
```

Currently the `rowData` slot is also empty.  However, analogous to our call to `addPerCellQC()` in the prior section, the function `addPerFeatureQC()` will insert values in the `rowData` slot of our `sce` object, as we can see below:


```r
sce <- addPerFeatureQC(sce)
rowData(sce)
```

```
## DataFrame with 10 rows and 2 columns
##              mean  detected
##         <numeric> <numeric>
## gene_1    17.6667       100
## gene_2    15.6667       100
## gene_3    19.6667       100
## gene_4    15.0000       100
## gene_5    14.0000       100
## gene_6    20.6667       100
## gene_7    15.0000       100
## gene_8    20.3333       100
## gene_9    13.6667       100
## gene_10   19.6667       100
```

In a similar fashion to the `colData` slot, such feature metadata could be provided at the onset when creating the *SingleCellExperiment* object, which we leave up to the reader as an exercise.


#### Subsetting by Rows

To subset an `sce` object down at the feature/gene level, we can do a row subsetting operation similar to other R objects, by supplying either numeric indices or a vector of names:


```r
sce[c("gene_1", "gene_4"), ]
```

```
## class: SingleCellExperiment 
## dim: 2 3 
## metadata(0):
## assays(1): counts
## rownames(2): gene_1 gene_4
## rowData names(2): mean detected
## colnames(3): cell_1 cell_2 cell_3
## colData names(8): batch sum ... percent_top_500 total
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

```r
## sce[c(1, 4), ] # same as above in this case
```

### Size Factors Slot: `sizeFactors`

Briefly, we already encountered this via the `scran::computeSumFactors(sce)` call, which adds a `sizeFactors` slot:

* `sizeFactors` slot: contains information in a numeric vector regarding the sample/cell normalization factors used to produce a normalized data representation (_Figure 1B, brown box_)


```r
sce <- scran::computeSumFactors(sce)
sce <- scater::logNormCounts(sce)
sizeFactors(sce)
```

```
## [1] 0.601 0.689 1.710
```

## A Brief Recap: From `se` to `sce`

So far, we have covered the `assays` (primary data), `colData` (sample metadata), `rowData`/`rowRanges` (feature metadata), and `sizeFactors` slots of *SingleCellExperiment*. 

What is important to note is that the *SingleCellExperiment* class _derives_ from the `SummarizedExperiment` (`se`) class, its predecessor, and in particular inherits the `assays`, `colData` and `rowData`/`rowRanges` slots. As such, much of the `SummarizedExperiment` functionality is retained in *SingleCellExperiment*. This allows existing methods that work with `SummarizedExperiment` to work similarly on *SingleCellExperiment* objects.

So what's new about the *SingleCellExperiment* class then? For our discussion, the most important change is the addition of a new slot called `reducedDims`.

## The `reducedDims` Slot

The `reducedDims` slot is a new addition which is specially designed to store reduced dimensionality representations of primary data, obtained by methods such as PCA, tSNE, UMAP, and others. 

* `reducedDims` slot: contains a list of numeric `matrix` entries which describe dimensionality reduced representations of the primary data, such that rows represent the columns of the primary data (aka the samples/cells), and columns represent the dimensions

Most importantly, just like the `assays` slot, the `reducedDims` slot can hold a list of many entries. So, it can hold a PCA, TSNE, and UMAP representation of a given dataset all within the `reducedDims` slot.

In our example, we can calculate a PCA representation of our data as follows using the `scater` package function `runPCA()`. We see that the `sce` now shows a new `reducedDim` and that the accessor `reducedDim()` produces the results of running PCA on the normalized data from `logcounts(sce)`.


```r
sce <- scater::runPCA(sce)
reducedDim(sce, "PCA")
```

```
##           PC1     PC2
## cell_1 -0.495 -0.3154
## cell_2  1.047  0.0111
## cell_3 -0.552  0.3043
## attr(,"percentVar")
## [1] 89.5 10.5
```

From this, we can also calculate a tSNE representation using the `scater` package function `runTSNE()`, and see that it can be seen both in the default view of `sce` and via accession:


```r
sce <- scater::runTSNE(sce, perplexity = 0.1)
```

```
## Perplexity should be lower than K!
```

```r
reducedDim(sce, "TSNE")
```

```
##         [,1]  [,2]
## cell_1   783  5637
## cell_2 -5278 -2133
## cell_3  4495 -3504
```

We can view the names of all our entries in the `reducedDims` slot via the accessor, `reducedDims()` (notice that this is plural, and thus not the same as `reducedDim()`:


```r
reducedDims(sce)
```

```
## List of length 2
## names(2): PCA TSNE
```

Now, say we have a different dimensionality reduction algorithm we'd like to try that is not yet implemented with direct compatibility for *SingleCellExperiment*, e.g. with an associated wrapper in packages such as `scater`. To accommodate this case (or alternately, when we want to directly run the dimensionality reduction method on the data ourselves rather than through a wrapper), we can directly add to the `reducedDims` slot. This is similar to how we previously extended the `assays` slot with our own custom entry of `counts_100`.

Below, we show how to run `umap()` as implemented in the `uwot` package directly, rather than through the associated wrapper function implemented in `scater` via `runUMAP()`, saving the intermediate results, and then adding them into the `sce` object we had previously.


```r
u <- uwot::umap(t(logcounts(sce)), n_neighbors = 2)
reducedDim(sce, "UMAP_uwot") <- u

reducedDim(sce, "UMAP_uwot")
```

```
##          [,1]    [,2]
## cell_1 -0.168 -0.3907
## cell_2  0.647 -0.0126
## cell_3 -0.479  0.4034
## attr(,"scaled:center")
## [1] 11.23 -8.34
```

And now we can also see its entry when we look at the `reducedDims()` accessor output:


```r
reducedDims(sce)
```

```
## List of length 3
## names(3): PCA TSNE UMAP_uwot
```

## The `metadata` Slot

Some analyses produce results that do not fit into the aforementioned slots. Thankfully, there is a slot just for this type of messy data, and in fact, can accommodate any type of data, so long as it is in a named list:

* `metadata` slot: a named list of entries, where each entry in the list can be anything you want it to be

For example, say we have some favorite genes, such as highly variable genes, we want to save inside of `sce` for use in our analysis at a later point. We can do this simply by appending to the metadata slot as follows:


```r
my_genes <- c("gene_1", "gene_5")
metadata(sce) <- list(favorite_genes = my_genes)
metadata(sce)
```

```
## $favorite_genes
## [1] "gene_1" "gene_5"
```

Similarly, we can append more information via the `$` operator:


```r
your_genes <- c("gene_4", "gene_8")
metadata(sce)$your_genes <- your_genes
metadata(sce)
```

```
## $favorite_genes
## [1] "gene_1" "gene_5"
## 
## $your_genes
## [1] "gene_4" "gene_8"
```

## About Spike-Ins

You might have noticed that the `sce` default view produces an entry with `spikeNames`. The *SingleCellExperiment* object contains some special considerations for experiments with spike-in (ERCC) controls. We leave this to the interested reader to learn more about in the [*SingleCellExperiment* introductory vignette](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html).

## Recording Analyses in _SingleCellExperiment_

In subsequent sections, we will show an example workflow that uses the *SingleCellExperiment* object as its base, and similar to our walkthrough of the *SingleCellExperiment* class above, continually appends new entries to save the results of the analysis. The *SingleCellExperiment* thus can serve as a record of analysis in this manner. This makes it especially useful for collaboration, as the object can be transferred, analyzed, and even visualized via graphical user interfaces such as *iSEE*.

## Conclusion 

The *SingleCellExperiment* class to the Bioc-verse single-cell related packages. It is this connection to *SingleCellExperiment* that makes many of these packages easily interoperable and modular over the span of an scRNA-seq analysis. Furthermore, it allows anybody to implement their own analytical tools for scRNA-seq that build on the *SingleCellExperiment*, enabling decentralized scientific developments.




<!-- ## Multimodal Data: _MultiAssayExperiment_ -->

<!-- Recent advances in technology and protocols allow the simultaneous collection of DNA and RNA from the same cells, enabling single-cell multi-modal analysis. These data present new challenges in the complexity of statistical analyses, which are addressed in Bioconductor through the *MultiAssayExperiment* container. The *MultiAssayExperiment* class integrates all major Bioconductor experimental data containers, and any containers derived from those, including *SingleCellExperiment*. It provides harmonized data management for heterogeneous assays, including subsetting by genomic identifiers, genomic coordinates, or sample/cell attributes such as cell type. The user interface mimics that of *SingleCellExperiment*, with comparable actions working across all assays. -->

<!-- Multi-modal profiling is an emergent area of single-cell biology with many exciting technologies coming online, such as gene expression profiling in tandem with protein via CITE-seq/REAP-seq and adaptive repertoire sequencing. While we won't cover multimodal data analysis further in this online book as of this writing, we anticipate infrastructure and statistical methodology advances in this area in the near future. -->

<!-- \cite{Macaulay2016gtseq, Dey2015-eh, Stoeckius2017-se} -->



<!-- ![Graph network of package dependencies linking to the *SingleCellExperiment* package (class). Packages are filtered by biocView "singleCell".](images/SCE-dependency-graph.png) -->

<!-- ```{r, include=FALSE, eval=FALSE} -->
<!-- ## Script used to generate above image -->
<!-- library(tidyverse) -->
<!-- library(BiocPkgTools) -->
<!-- library(igraph) -->
<!-- library(tidygraph) -->
<!-- library(ggraph) -->

<!-- tt <- read_tsv('table-package-types.txt', col_names = FALSE) %>% -->
<!--     dplyr::rename(Category = X1, Package = X2) %>% -->
<!--     unique() %>% -->
<!--     filter(!duplicated(Package)) -->

<!-- bpl = biocPkgList() -->
<!-- bpl_sc = bpl[grepl('SingleCell', bpl$biocViews),] -->

<!-- dep_df = buildPkgDependencyDataFrame() %>%  -->
<!--     filter(Package %in% bpl_sc$Package & dependency %in% bpl_sc$Package) %>% -->
<!--     filter(dependency %in% bpl$Package) -->

<!-- tbl_graph <- as_tbl_graph(dep_df[, 1:3], directed = TRUE) %>% -->
<!--     left_join(tt, by = c("name" = "Package")) %>% -->
<!--     mutate(Category = as.factor(ifelse(is.na(Category), 'Other', Category)), -->
<!--            degree = degree(.)) -->

<!-- ggraph(tbl_graph) + #, layout = 'kk', kkconst = sqrt(vcount(tbl_graph))) + -->
<!--     geom_edge_fan2(edge_width = 0.05, alpha = 0.5) + -->
<!--     geom_node_point(aes(colour = Category, size = degree^2)) + -->
<!-- ##    geom_node_point(aes(color = Category), size = 10) + -->
<!--     geom_node_text(aes(label = name), size = 4, repel=TRUE) + #1.6 is good -->
<!--     theme_graph(base_family = 'Helvetica') + -->
<!-- #    theme(legend.position = 'none') + -->
<!--     scale_colour_brewer(type = 'qual', palette = 2) -->
<!-- ``` -->
