---
output:
  html_document
bibliography: ../ref.bib
---

# Integrating Datasets

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>

<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>

As scRNA-seq approaches continue to gain in popularity, projects that combine independently generated datasets will become increasingly common. While generalized linear modeling frameworks can be used to integrate disparate datasets, the performance of these frameworks in the scRNA-seq context may be sub-optimal, often due to an underlying assumption that the composition of cell populations is either known or identical across batches of cells. This has inspired a new class of methods that leverage the unique properties of scRNA-seq to improve the amelioration of batch effects, which we demonstrate here.

Here we will work with two PBMC datasets from the *TENxPBMCData* package - the `pbmc3k` and `pbmc4k` datasets - that come from different donors, and integrate them to produce an integrated representation of the data.


## Preprocessing

Before proceeding to the integration section, we will first load the data, and subsequently run it through a preprocessing pipeline that ensures we have a common set of genes across both experiments, filters out poor quality cells, and normalizes the expression matrix. The code for this is shown below for completeness. 
Note that here we make use of the *purrr* package both for code succinctness and to apply the same function to each PBMC dataset contained within the `pbmc_l` list. (see base `lapply()` and `purrrr::map()` for more details).

Furthermore, to speed up computation, we will subsample each of these datasets down to 1000 cells each.


```r
## Load the data
library(TENxPBMCData)
pbmc3k <- TENxPBMCData('pbmc3k')
pbmc4k <- TENxPBMCData('pbmc4k')
pbmc_l <- list(pbmc3k = pbmc3k, pbmc4k = pbmc4k)

## Subsample data to 1000 cells each
library(purrr)
set.seed(1234)
pbmc_l <- map(pbmc_l, function(x) { x[, sample(ncol(x), 1000)] })

## Retain common genes (rows); reorder rows by common genes
common_genes <- reduce(map(pbmc_l, rownames), intersect)
pbmc_l <- map(pbmc_l, function(x) { x[common_genes, ] })

## Save rowData from first object
## remove from each individual object else `cbind` will throw an error
rd <- rowData(pbmc_l[[1]])
pbmc_l <- map(pbmc_l, function(x) {
    rowData(x) <- NULL
    return(x)
})

## Combine the objects via `cbind`
sce <- reduce(pbmc_l, cbind)

## Replace rowData within combined object
## Uniquify row ids to use mostly human readable rownames
rowData(sce) <- rd
rownames(sce) <- scater::uniquifyFeatureNames(rd$ENSEMBL_ID, rd$Symbol_TENx)
```

<!-- ## HOTFIX NEEDED - RECAST ASSAY INTO MATRIX  -->



```r
## Calculate and apply QC metrics; filter out cells with low lib or low genes detected
library(scater)
sce <- calculateQCMetrics(sce)
low_lib <- isOutlier(sce$log10_total_counts, type = "lower", nmad = 3)
low_genes <- isOutlier(sce$log10_total_features_by_counts, type = "lower", nmad = 3)
sce <- sce[, !(low_lib | low_genes)]

## normalize data
sce <- normalize(sce)

## feature selection
library(scran)
fit <- trendVar(sce, use.spikes = FALSE)
dec <- decomposeVar(sce, fit)
hvg <- rownames(dec)[dec$bio > 0] # save gene names
```

For faster calculations, here we will subset the `hvg` object to the top 2000 genes:


```r
hvg <- hvg[1:2000]
```

From the combined dataset encapsulated within `sce`, we note that the origin of each cell is annotated within `colData` under the column `Sample`. This will be used as the batch variable throughout.


```r
table(sce$Sample) # alternately: colData(sce)$Sample
```

```
## 
## pbmc3k pbmc4k 
##    990   1000
```


## Fast MNN

The *batchelor* package provides the `fastMNN()` function, which calculates a dimension reduced representation based on the original principal components derived directly from the normalized expression data. These MNN coordinates can then be used for further dimensionality reduction via UMAP as shown below.


```r
library(batchelor)

## create new object from sce for mnn simple workflow
sce_mnn_simple <- sce

## run MNN with no advanced preprocessing
## BSPARAM/BNPARAM/BPPARAM explained in "adaptations to large scale data"
mnn_simple_out <- batchelor::fastMNN(sce_mnn_simple,
                                     subset.row = hvg,
                                     batch = sce_mnn_simple$Sample)
                                     ## BSPARAM = BiocSingular::IrlbaParam(),
                                     ## BNPARAM = BiocNeighbors::AnnoyParam())

## add MNN_simple coordinates to the new sce object
reducedDim(sce_mnn_simple, 'MNN') <- reducedDim(mnn_simple_out, 'corrected')

## run UMAP on the "simple MNN" version
sce_mnn_simple <- runUMAP(sce_mnn_simple, use_dimred = 'MNN')
```


```r
library(patchwork)
p1 <- plotReducedDim(sce_mnn_simple, 'MNN', colour_by = 'Sample')
p2 <- plotReducedDim(sce_mnn_simple, 'UMAP', colour_by = 'Sample')
wrap_plots(p1, p2, nrow = 1)
```

<div class="figure">
<img src="P2_W08.data-integration_files/figure-html/unnamed-chunk-7-1.png" alt="Simple MNN method. First two MNN coordinates (left) and UMAP derived from the full MNN coordinate set (right)." width="672" />
<p class="caption">(\#fig:unnamed-chunk-7)Simple MNN method. First two MNN coordinates (left) and UMAP derived from the full MNN coordinate set (right).</p>
</div>

## Harmony

Harmony provides a unified framework for data visualization, analysis, and interpretation of scRNA-seq data. Included is a method for integrating different batches of scRNA-seq. This produces a new dimension reduction representation, saved as `HARMONY` within the *SingleCellExperiment* object, as shown below. Furthermore, this Harmony embedding can be used to produce a UMAP representation as well:


```r
library(harmony) # devtools::install_github('immunogenomics/harmony')
library(BiocSingular)

## create a new sce for harmony
sce_harmony <- sce

## Add PCA to sce, then run Harmony as its dependent on PCA
sce_harmony <- runPCA(sce_harmony,
                      feature_set = hvg,
                      BSPARAM = BiocSingular::IrlbaParam(),
                      BPPARAM = BiocParallel::MulticoreParam())
sce_harmony <- RunHarmony(sce_harmony, group.by.vars = "Sample")

## Calculate UMAP on Harmony embeddings
sce_harmony <- runUMAP(sce_harmony, use_dimred = 'HARMONY')
```



```r
p1 <- plotReducedDim(sce_harmony, 'HARMONY', colour_by = 'Sample')
p2 <- plotReducedDim(sce_harmony, 'UMAP', colour_by = 'Sample')
wrap_plots(p1, p2, nrow = 1)
```

<div class="figure">
<img src="P2_W08.data-integration_files/figure-html/unnamed-chunk-9-1.png" alt="Plot of the first two components of the Harmony embedding (left). UMAP calculated on the Harmony embeddings. (right)" width="672" />
<p class="caption">(\#fig:unnamed-chunk-9)Plot of the first two components of the Harmony embedding (left). UMAP calculated on the Harmony embeddings. (right)</p>
</div>


## Advanced MNN Workflow

Here we work through a more advanced workflow using MNN as above, but this time walking through the various steps that are otherwise implicit within the `fastMNN()` function. While not always necessary, it may be informative for those interested in the finer details of a batch correction workflow. Note that this follows the 

For more details, we refer to the [`compareSingleCell` vignette](https://github.com/MarioniLab/compareSingleCell/) vignette by Aaron Lun.

<!-- ### Feature Selection -->

<!-- We briefly highlight feature selection within the preprocessing, as the choice of genes used in the integration can have an outsize effect on end result.  -->

<!-- For brevity, here we use the `scran` package to identify the genes with biological coefficients of variation greater than zero, but leave the choice of method up to the reader. Note that, compared to the basic analysis workflow, we use the `multiBlockVar()` function and block on the origin (`pbmc3k` vs `pbmc4k`). This function models the variance of expression in each block separately. -->

<!-- Additionally, we turn off weighting such that each sample contributes equally to the results, regardless of the number of cells per sample. This prevents one condition from biasing the combined statistics.  -->

<!-- For more details, we refer to the [`compareSingleCell` merge](https://github.com/MarioniLab/compareSingleCell/blob/master/vignettes/embryo_merge.Rmd) vignette. -->



```r
## create a new sce for advanced mnn
sce_mnn_advanced <- sce
```

<!-- ```{r} -->
<!-- library(scran) -->
<!-- dec <- multiBlockVar(sce_mnn_advanced, block = sce_mnn_advanced$Sample, -->
<!--                      make.tech.trend = TRUE, -->
<!--                      weighted = FALSE) -->

<!-- hvg <- rownames(dec)[dec$bio > 0] -->
<!-- ``` -->

<!-- In this example, we get back 6001 genes. -->


<!-- ```{r} -->
<!-- hvg_subset <- hvg[1:1000] -->
<!-- ``` -->

### Dimensionality Reduction

Principal components analyses can be used to both reduce computational work and remove high-dimensional noise, as discussed in the [simpleSingleCell vignette](https://bioconductor.org/packages/3.10/simpleSingleCell/vignettes/reads.html#denoising-expression-values-using-pca). 

To accomplish this, the `multiBatchPCA()` function from *scran* ensures that each sample contributes equally to the new coordinate space. For further discussion on this point, see the [simpleSingleCell vignette](https://bioconductor.org/packages/3.10/simpleSingleCell/vignettes/batch.html#hierarchical-merging) section on hierarchical merging.

We then calculate the number of PCs to retain by comparing the variance explained per principal component versus the technical noise and total variance in the data. This will leave us with a subset of PCA components.

Note that we will use the features selected from the preproprocessing step earlier in this chapter.



```r
library(batchelor)
library(BiocSingular) # provide approximate pca via IrlbaParam()

## Calculate PCA by batch
set.seed(1234) # for Random/IrlbaParam backends

## run batchelor version of multiBatchPCA (use :: notation to override namespace)
pcs <- batchelor::multiBatchPCA(sce_mnn_advanced,
                                batch = sce_mnn_advanced$Sample,
                                subset.row = hvg,
                                get.variance = TRUE,
                                BSPARAM = BiocSingular::IrlbaParam(),
                                BPPARAM = BiocParallel::MulticoreParam())

## Retain only the top PCs based on variance explained > technical noise
retain <- denoisePCANumber(
    metadata(pcs)$var.explained, # variance explained per PC.
    sum(dec[hvg, ]$tech),       # technical noise in subset of genes.
    metadata(pcs)$var.total      # total variance in the data
)

retain
```

```
## [1] 3
```

From the number of PCs to retain, we can filter the PCs as follows:


```r
## Subset PCA number of components up to the number in retain
top_pcs <- map(as.list(pcs), ~ .[, 1:retain])
```

Then, using the retained components as input, we can input them directly into the `fastMNN()` method as follows, and furthermore produce a UMAP representation based on these MNN coordinates:


```r
## Calculate MNN coordinates, add into reducedDims
mnn_advanced_out <- batchelor::fastMNN(top_pcs$pbmc3k,
                                       top_pcs$pbmc4k,
                                       pc.input = TRUE)
reducedDim(sce_mnn_advanced, 'MNN') <- mnn_advanced_out$corrected

## Run UMAP on MNN coordinates
sce_mnn_advanced <- runUMAP(sce_mnn_advanced, use_dimred = 'MNN')
```

The MNN coordinates and the UMAP coordinates derived from them are visualized below:


```r
p1 <- plotReducedDim(sce_mnn_advanced, 'MNN', colour_by = 'Sample')
p2 <- plotReducedDim(sce_mnn_advanced, 'UMAP', colour_by = 'Sample')
wrap_plots(p1, p2, nrow = 1)
```

<div class="figure">
<img src="P2_W08.data-integration_files/figure-html/unnamed-chunk-14-1.png" alt="Plot of first two MNN coordinates (left). UMAP calculated on the MNN coordinates (right). Colour denotes origin of sample." width="672" />
<p class="caption">(\#fig:unnamed-chunk-14)Plot of first two MNN coordinates (left). UMAP calculated on the MNN coordinates (right). Colour denotes origin of sample.</p>
</div>

Comparing to the simpler MNN workflow shown above (which only used the `fastMNN()` function on the barely processed *SingleCellExperiment* class object `sce`), we see that the two results are fairly comparable, differing only by the number of principal components and the selected number of features (here we chose 1000 for computational efficacy; we leave it to the interested reader to explore the results with a larger feature space).


## Naive Method Without Correction

For completeness, we show here the results of performing the correction solely on the gene expression matrix.


```r
## create a new sce for naive method
sce_naive <- sce
```


```r
## run PCA on only the normalized gene expression data
sce_naive <- runPCA(sce_naive,
                    feature_set = hvg,
                    BSPARAM = BiocSingular::IrlbaParam(),
                    BPPARAM = BiocParallel::MulticoreParam())

## Run UMAP on the norm gene exp. derived PCA coordinates - use only top PCs
sce_naive <- runUMAP(sce_naive, use_dimred = 'PCA')
```


```r
p1 <- plotPCA(sce_naive, colour_by = 'Sample')
p2 <- plotReducedDim(sce_naive, 'UMAP', colour_by = 'Sample')
wrap_plots(p1, p2, nrow = 1)
```

<div class="figure">
<img src="P2_W08.data-integration_files/figure-html/unnamed-chunk-17-1.png" alt="First two components of PCA calculated directly from the gene expression matrix (left). UMAP calculated on the PCA derived from the normalized gene expression matrix (right)." width="672" />
<p class="caption">(\#fig:unnamed-chunk-17)First two components of PCA calculated directly from the gene expression matrix (left). UMAP calculated on the PCA derived from the normalized gene expression matrix (right).</p>
</div>

While the PCA looks well integrated, the UMAP representation - which encapsulates a fuller PCA space - distinctly shows the residual batch effect present.


## Limma Batch Correction

The `limma` package, a popular framework for the statistical analysis of RNA-seq, has a function `removeBatchEffect()` which will be used here to correct the normalized expression matrix `logcounts` across the two batches. The result will be assigned into the `assays` slot of the `sce` object as `limma_corrected`, and then used for PCA, saving the result in the `reducedDim` slot as `"PCA_limma"`.


```r
## create a new sce for limma method
sce_limma <- sce
```


```r
library(limma)
limma_corrected <- removeBatchEffect(logcounts(sce_limma), batch = sce_limma$Sample)
assay(sce_limma, "logcounts_limma") <- limma_corrected ## add new assay

## calc a PCA on limma values; save separate to prevent overwriting
sce_limma <- runPCA(sce_limma,
                    feature_set = hvg,
                    exprs_values = "logcounts_limma",
                    BSPARAM = BiocSingular::IrlbaParam(),
                    BPPARAM = BiocParallel::MulticoreParam())

## run UMAP; save separate to prevent overwriting
sce_limma <- runUMAP(sce_limma, use_dimred = 'PCA')
```

The resulting PCA and UMAP from the limma batch correction method are shown below:


```r
p1 <- plotReducedDim(sce_limma, 'PCA', colour_by = 'Sample')
p2 <- plotReducedDim(sce_limma, 'UMAP', colour_by = 'Sample')
wrap_plots(p1, p2, nrow = 1)
```

<div class="figure">
<img src="P2_W08.data-integration_files/figure-html/unnamed-chunk-20-1.png" alt="PCA calculated on the limma batch-corrected gene expression matrix (left). UMAP calculated via the PCA derived from limma batch-corrected gene expression matrix (right)." width="672" />
<p class="caption">(\#fig:unnamed-chunk-20)PCA calculated on the limma batch-corrected gene expression matrix (left). UMAP calculated via the PCA derived from limma batch-corrected gene expression matrix (right).</p>
</div>


## Session Info


```
## <button class="aaron-collapse">View session info</button>
## <div class="aaron-content">
## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.6 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] limma_3.41.6                BiocSingular_1.1.5         
##  [3] harmony_0.99.9              Rcpp_1.0.1                 
##  [5] patchwork_0.0.1             batchelor_1.1.4            
##  [7] scran_1.13.9                scater_1.13.9              
##  [9] ggplot2_3.2.0               purrr_0.3.2                
## [11] TENxPBMCData_1.3.0          HDF5Array_1.13.4           
## [13] rhdf5_2.29.0                SingleCellExperiment_1.5.3 
## [15] SummarizedExperiment_1.15.5 DelayedArray_0.11.4        
## [17] BiocParallel_1.19.0         matrixStats_0.54.0         
## [19] Biobase_2.45.0              GenomicRanges_1.37.14      
## [21] GenomeInfoDb_1.21.1         IRanges_2.19.10            
## [23] S4Vectors_0.23.17           BiocGenerics_0.31.5        
## [25] BiocStyle_2.13.2            Cairo_1.5-10               
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  bit64_0.9-7                  
##  [3] httr_1.4.0                    dynamicTreeCut_1.63-1        
##  [5] tools_3.6.0                   backports_1.1.4              
##  [7] R6_2.4.0                      irlba_2.3.3                  
##  [9] vipor_0.4.5                   DBI_1.0.0                    
## [11] lazyeval_0.2.2                colorspace_1.4-1             
## [13] withr_2.1.2                   tidyselect_0.2.5             
## [15] gridExtra_2.3                 bit_1.1-14                   
## [17] curl_4.0                      compiler_3.6.0               
## [19] BiocNeighbors_1.3.2           bookdown_0.12                
## [21] scales_1.0.0                  rappdirs_0.3.1               
## [23] stringr_1.4.0                 digest_0.6.20                
## [25] rmarkdown_1.14                XVector_0.25.0               
## [27] pkgconfig_2.0.2               htmltools_0.3.6              
## [29] dbplyr_1.4.2                  rlang_0.4.0                  
## [31] RSQLite_2.1.1                 shiny_1.3.2                  
## [33] DelayedMatrixStats_1.7.1      dplyr_0.8.3                  
## [35] RCurl_1.95-4.12               magrittr_1.5                 
## [37] GenomeInfoDbData_1.2.1        Matrix_1.2-17                
## [39] ggbeeswarm_0.6.0              munsell_0.5.0                
## [41] Rhdf5lib_1.7.2                viridis_0.5.1                
## [43] edgeR_3.27.6                  stringi_1.4.3                
## [45] yaml_2.2.0                    zlibbioc_1.31.0              
## [47] BiocFileCache_1.9.1           AnnotationHub_2.17.3         
## [49] grid_3.6.0                    blob_1.2.0                   
## [51] dqrng_0.2.1                   promises_1.0.1               
## [53] ExperimentHub_1.11.1          crayon_1.3.4                 
## [55] lattice_0.20-38               cowplot_1.0.0                
## [57] locfit_1.5-9.1                zeallot_0.1.0                
## [59] knitr_1.23                    pillar_1.4.2                 
## [61] igraph_1.2.4.1                codetools_0.2-16             
## [63] glue_1.3.1                    evaluate_0.14                
## [65] BiocManager_1.30.4            vctrs_0.2.0                  
## [67] httpuv_1.5.1                  gtable_0.3.0                 
## [69] assertthat_0.2.1              xfun_0.8                     
## [71] rsvd_1.0.1                    mime_0.7                     
## [73] xtable_1.8-4                  later_0.8.0                  
## [75] viridisLite_0.3.0             tibble_2.1.3                 
## [77] AnnotationDbi_1.47.0          beeswarm_0.2.3               
## [79] memoise_1.1.0                 statmod_1.4.32               
## [81] interactiveDisplayBase_1.23.0
## </div>
```

