---
output:
  html_document
bibliography: ../ref.bib
---

# Adaptations for Large-scale Data

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>

<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>

Large datasets such as the Human Cell Atlas (with over 1.3 million cells) have the potential to benefit from special adaptations that enable analysis in compute-constrained environments (such as personal laptops). Here, we briefly cover topics that aim to ease working with scRNA-seq data to make it faster and more tractable.

Do note however that these adaptations do not universally result in improved computational efficiency. For example, parallelization does incur costs, and disk-backed data representation will generally be slower than purely in-memory representations. In both cases, hardware specifics may dictate any potential gains in efficiency, as can be imagined in the case of a solid state drive (SSD) being faster for disk-backed data representations relative to a hard disk drive (HDD).

Thus, with the right compute environment, data, and task, these adaptations can yield significant computational improvements.


## Approximate Methods

A general approach that works across all scales of data involves changing the task itself. Some methods - such as PCA or nearest neighbor searches - have been extended to include versions that provide approximate results. Generally, these adaptations result in acceptable losses in accuracy for significant computational gains. In some cases, approximate methods may even be desirable for the results themselves, as has been shown by the FIt-SNE approach.

Some example packages that provide approximate versions of popular methods include:

* *BiocSingular* via the `IrlbaParam()` and `RandomParam()` for approximate singular value decomposition (SVD)
* *BiocNeighbors* via the `AnnoyParam()` for approximate nearest neighbor searches

These packages provide users (and developers) a common interface, enabling modular swapping of key algorithms within functions. For example, we can see the immediate benefit of using an approximate method for PCA as provided by the `BSPARAM` argument in the following code, which utilizes the *bench* package for profiling:


```r
library(SingleCellExperiment)
library(scater)

## Simulate a dataset with 1k genes and 1k cells
mat <- matrix(rpois(1e6, 100), nrow = 1000)
tiny_sce <- SingleCellExperiment(assays = list(counts = mat))
tiny_sce <- normalize(tiny_sce)
```


```r
library(bench)
library(BiocSingular)

## simple function to show only cols of interest
.show_bench <- function(b) {
    b[, c('expression', 'min', 'median', 'mem_alloc')]
}

bm <- bench::mark(
           runPCA(tiny_sce, BSPARAM = IrlbaParam()),
           runPCA(tiny_sce, BSPARAM = ExactParam()), 
           check = FALSE)

.show_bench(bm)
```

```
## # A tibble: 2 x 4
##   expression                                    min   median mem_alloc
##   <bch:expr>                               <bch:tm> <bch:tm> <bch:byt>
## 1 runPCA(tiny_sce, BSPARAM = IrlbaParam())    139ms    151ms        0B
## 2 runPCA(tiny_sce, BSPARAM = ExactParam())    165ms    369ms        0B
```

We can see that the approximate, irlba based implementation is about 6 times faster than the exact version in this case.

We can also try this with nearest neighbor searches. Here we provide the `BNPARAM` argument to build a shared nearest neighbors graph using different algorithms under the hood:


```r
library(scran)
library(BiocNeighbors)

## Calculate PCA before testing NN back-ends
tiny_sce_pca <- runPCA(tiny_sce, BSPARAM = IrlbaParam())

bm <- bench::mark(
           buildSNNGraph(tiny_sce_pca, BNPARAM = AnnoyParam()),
           buildSNNGraph(tiny_sce_pca, BNPARAM = KmknnParam()),
           check = FALSE)

.show_bench(bm)
```

```
## # A tibble: 2 x 4
##   expression                                               min   median
##   <bch:expr>                                          <bch:tm> <bch:tm>
## 1 buildSNNGraph(tiny_sce_pca, BNPARAM = AnnoyParam())    1.19s    1.19s
## 2 buildSNNGraph(tiny_sce_pca, BNPARAM = KmknnParam()) 801.83ms 801.83ms
## # â€¦ with 1 more variable: mem_alloc <bch:byt>
```

We can see from the above benchmark that in our tiny dataset, we don't see much, if any, benefit of using an approximate method (via the `AnnoyParam()`, which uses the Annoy library).

However, if we increase our dataset to something larger..


```r
## Simulate a dataset with 1k genes and 10k cells
mat <- matrix(rpois(10e6, 100), nrow = 1000)
big_sce <- SingleCellExperiment(assays = list(counts = mat))
big_sce <- normalize(big_sce)
```


```r
## Calculate PCA before testing NN back-ends
big_sce <- runPCA(big_sce, BSPARAM = IrlbaParam())

## NN search
bm <- bench::mark(
           buildSNNGraph(big_sce, BNPARAM = AnnoyParam()),
           buildSNNGraph(big_sce, BNPARAM = KmknnParam()),
           check = FALSE)

.show_bench(bm)
```

```
## # A tibble: 2 x 4
##   expression                                          min  median mem_alloc
##   <bch:expr>                                     <bch:tm> <bch:t> <bch:byt>
## 1 buildSNNGraph(big_sce, BNPARAM = AnnoyParam())    7.28s   7.28s        0B
## 2 buildSNNGraph(big_sce, BNPARAM = KmknnParam())   12.32s  12.32s        0B
```

We see (more consistently) that we have realized some gains from using the approximate-based nearest neighbors search.


## Parallelization

Many tasks that arise in the analysis of scRNA-seq data are able to be parallelized. In other words, the problem can be broken down into smaller pieces that can be solved independently. Parallel computing can be performed in computing environments with access to multiple cores.

Bioconductor has reduced the complexity of implementing parallelized software via the [*BiocParallel*](https://bioconductor.org/packages/BiocParallel) package. This enables a common interface across Bioconductor software packages for parallel computing.

Across many Bioconductor packages, an argument that will often be present in functions that are parallelizable is the `BPPARAM` argument. Here, it is possible to specify the parallel back-end that can be used for evaluation.

Again, note that parallelization does incur an overhead cost in splitting up the data, sending it off, and combining the results, and thus your mileage may vary depending on the specifics. 

Below is an example usage of the *BiocParallel* library supplying `BPPARAM` argument for constructing the UMAP representation from the PCA results:


```r
library(BiocParallel)

bm <- bench::mark(
           runUMAP(tiny_sce_pca, BPPARAM = SerialParam()),           
           runUMAP(tiny_sce_pca, BPPARAM = MulticoreParam()),
           check = FALSE)

.show_bench(bm)
```

```
## # A tibble: 2 x 4
##   expression                                           min median mem_alloc
##   <bch:expr>                                        <bch:> <bch:> <bch:byt>
## 1 runUMAP(tiny_sce_pca, BPPARAM = SerialParam())     3.86s  3.86s        0B
## 2 runUMAP(tiny_sce_pca, BPPARAM = MulticoreParam())  3.92s  3.92s        0B
```

On this dataset and equipment, there's not much, if any, benefit to parallelization, but your mileage may vary dependent on those aspects.


## On-Disk Data

The matrix of data from a single-cell experiment can be on the order of tens to hundreds of gigabytes, depending on the number of features and cells measured. A standard approach to import and represent these matrices has been to load the entire data set into memory using either the *matrix* object in base R or sparse and dense matrix classes from the *Matrix* R package. This approach however may prove intractable in computing environments with limited memory.

Disk-backed representations such as HDF5 free us from having to load an entire dataset into memory, and thus make it possible to work with large-scale scRNA-seq data.


```r
library(rhdf5)
library(HDF5Array)
```


## Session Info

<button class="aaron-collapse">View session info</button>
<div class="aaron-content">
R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.6 LTS

Matrix products: default
BLAS/LAPACK: /app/easybuild/software/OpenBLAS/0.2.18-GCC-5.4.0-2.26-LAPACK-3.6.1/lib/libopenblas_prescottp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] HDF5Array_1.13.4            rhdf5_2.29.0               
 [3] BiocNeighbors_1.3.2         scran_1.13.9               
 [5] BiocSingular_1.1.5          bench_1.0.2                
 [7] scater_1.13.9               ggplot2_3.2.0              
 [9] SingleCellExperiment_1.5.3  SummarizedExperiment_1.15.5
[11] DelayedArray_0.11.4         BiocParallel_1.19.0        
[13] matrixStats_0.54.0          Biobase_2.45.0             
[15] GenomicRanges_1.37.14       GenomeInfoDb_1.21.1        
[17] IRanges_2.19.10             S4Vectors_0.23.17          
[19] BiocGenerics_0.31.5         BiocStyle_2.13.2           

loaded via a namespace (and not attached):
 [1] viridis_0.5.1            dynamicTreeCut_1.63-1   
 [3] edgeR_3.27.6             viridisLite_0.3.0       
 [5] DelayedMatrixStats_1.7.1 assertthat_0.2.1        
 [7] statmod_1.4.32           BiocManager_1.30.4      
 [9] dqrng_0.2.1              GenomeInfoDbData_1.2.1  
[11] vipor_0.4.5              yaml_2.2.0              
[13] pillar_1.4.2             lattice_0.20-38         
[15] glue_1.3.1               limma_3.41.6            
[17] digest_0.6.20            XVector_0.25.0          
[19] colorspace_1.4-1         htmltools_0.3.6         
[21] Matrix_1.2-17            pkgconfig_2.0.2         
[23] bookdown_0.12            zlibbioc_1.31.0         
[25] purrr_0.3.2              scales_1.0.0            
[27] tibble_2.1.3             withr_2.1.2             
[29] lazyeval_0.2.2           magrittr_1.5            
[31] crayon_1.3.4             evaluate_0.14           
[33] beeswarm_0.2.3           tools_3.6.0             
[35] stringr_1.4.0            Rhdf5lib_1.7.2          
[37] munsell_0.5.0            locfit_1.5-9.1          
[39] irlba_2.3.3              compiler_3.6.0          
[41] rsvd_1.0.1               rlang_0.4.0             
[43] grid_3.6.0               RCurl_1.95-4.12         
[45] igraph_1.2.4.1           bitops_1.0-6            
[47] rmarkdown_1.14           gtable_0.3.0            
[49] R6_2.4.0                 gridExtra_2.3           
[51] knitr_1.23               dplyr_0.8.3             
[53] stringi_1.4.3            ggbeeswarm_0.6.0        
[55] Rcpp_1.0.1               tidyselect_0.2.5        
[57] xfun_0.8                
</div>

<!-- ## Notes on Data Representation -->


<!-- However, the explosion of data from single-cell assays has led to alternative data representations that are compatible with high-performance C++ code used for computationally intensive tasks in existing Bioconductor packages. This enables memory-efficient data manipulation and operations in R. For example, the *beachmat* Bioconductor package is a C++ interface for accessing single-cell data that is interoperable with sparse, dense and file-backed matrices, such as the HDF5 file format, which allows users to only load a subset of the data into memory at a time. In addition, Bioconductor has developed the infrastructure to read and write HDF5 files from R using the *rhdf5* package, to efficiently work with data in HDF5 files using array-like containers in the *HDF5Array*, *DelayedArray*, and *DelayedMatrixStats* packages. Much of this infrastructure was motivated by previous work done in the context of flow cytometry and whole genome bisulfite sequencing. -->


<!-- At a low-level, the main interface between HDF5 and Bioconductor is implemented in the packages `rhdf5`, which provides read/write functionalities, `Rhdf5lib`, which provides C and C++ HDF5 libraries, and `beachmat`, which provides a consistent C++ class interface for a variety of commonly used matrix types, including sparse and HDF5-backed matrices. These packages are useful for developers that want to develop methods able to interact with HDF5 data sets. -->
