# Merged human pancreas datasets {#merged-pancreas}

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>

<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>

## Introduction

For a period in 2016, there was a great deal of interest in using scRNA-seq to profile the human pancreas at cellular resolution
[@muraro2016singlecell;@grun2016denovo;@lawlor2017singlecell;@segerstolpe2016singlecell].
As a consequence, we have a surplus of human pancreas datasets generated by different authors with different technologies,
which provides an ideal use case for demonstrating more complex data integration strategies.
This represents a more challenging application than the PBMC dataset in Chapter \@ref(integrating-datasets) as it involves different sequencing protocols and different patients, most likely with differences in cell type composition.

## The good

We start by considering only two datasets from @muraro2016singlecell and @grun2016denovo.
This is a relatively simple scenario involving very similar protocols (CEL-seq and CEL-seq2) and a similar set of authors.

<button class="aaron-collapse">View history</button>
<div class="aaron-content">
   
```r
#--- setup ---#
library(OSCAUtils)
chapterPreamble(use_cache = TRUE)

#--- loading ---#
library(scRNAseq)
sce.grun <- GrunPancreasData()

#--- gene-annotation ---#
library(org.Hs.eg.db)
gene.ids <- mapIds(org.Hs.eg.db, keys=rowData(sce.grun)$symbol,
    keytype="SYMBOL", column="ENSEMBL")

keep <- !is.na(gene.ids) & !duplicated(gene.ids)
sce.grun <- sce.grun[keep,]
rownames(sce.grun) <- gene.ids[keep]

#--- quality-control ---#
library(scater)
stats <- perCellQCMetrics(sce.grun)
qc <- quickPerCellQC(stats, percent_subsets="altexps_ERCC_percent")
sce.grun <- sce.grun[,!qc$discard]

#--- normalization ---#
library(scran)
set.seed(1000) # for irlba. 
clusters <- quickCluster(sce.grun)
sce.grun <- computeSumFactors(sce.grun, clusters=clusters)
sce.grun <- logNormCounts(sce.grun)

#--- variance-modelling ---#
block <- paste0(sce.grun$sample, "_", sce.grun$donor)
dec.grun <- modelGeneVarWithSpikes(sce.grun, spikes="ERCC", block=block)
top.grun <- getTopHVGs(dec.grun, prop=0.1)
```

</div>


```r
sce.grun
```

```
## class: SingleCellExperiment 
## dim: 17692 1290 
## metadata(0):
## assays(2): counts logcounts
## rownames(17692): ENSG00000268895 ENSG00000121410 ...
##   ENSG00000074755 ENSG00000036549
## rowData names(2): symbol chr
## colnames(1290): D2ex_1 D2ex_2 ... D17TGFB_94 D17TGFB_95
## colData names(2): donor sample
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(1): ERCC
```

<button class="aaron-collapse">View history</button>
<div class="aaron-content">
   
```r
#--- setup ---#
library(OSCAUtils)
chapterPreamble(use_cache = TRUE)

#--- loading ---#
library(scRNAseq)
sce.muraro <- MuraroPancreasData()

#--- gene-annotation ---#
library(AnnotationHub)
edb <- AnnotationHub()[["AH73881"]]
gene.symb <- sub("__chr.*$", "", rownames(sce.muraro))
gene.ids <- mapIds(edb, keys=gene.symb, 
    keytype="SYMBOL", column="GENEID")

# Removing duplicated genes or genes without Ensembl IDs.
keep <- !is.na(gene.ids) & !duplicated(gene.ids)
sce.muraro <- sce.muraro[keep,]
rownames(sce.muraro) <- gene.ids[keep]

#--- quality-control ---#
library(scater)
stats <- perCellQCMetrics(sce.muraro)
qc <- quickPerCellQC(stats, percent_subsets="altexps_ERCC_percent")
sce.muraro <- sce.muraro[,!qc$discard]

#--- normalization ---#
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.muraro)
sce.muraro <- computeSumFactors(sce.muraro, clusters=clusters)
sce.muraro <- logNormCounts(sce.muraro)

#--- variance-modelling ---#
block <- paste0(sce.muraro$plate, "_", sce.muraro$donor)
dec.muraro <- modelGeneVarWithSpikes(sce.muraro, "ERCC", block=block)
top.muraro <- getTopHVGs(dec.muraro, prop=0.1)
```

</div>


```r
sce.muraro
```

```
## class: SingleCellExperiment 
## dim: 16940 2346 
## metadata(0):
## assays(2): counts logcounts
## rownames(16940): ENSG00000268895 ENSG00000121410 ...
##   ENSG00000159840 ENSG00000074755
## rowData names(2): symbol chr
## colnames(2346): D28-1_1 D28-1_2 ... D30-8_93 D30-8_94
## colData names(3): label donor plate
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(1): ERCC
```

We subset both batches to their common universe of genes;
adjust their scaling to equalize sequencing coverage (not really necessary in this case, as the coverage is already similar, but we will do so anyway for consistency);
and select those genes with positive average biological components for further use.


```r
universe <- intersect(rownames(sce.grun), rownames(sce.muraro))
sce.grun2 <- sce.grun[universe,]
dec.grun2 <- dec.grun[universe,]
sce.muraro2 <- sce.muraro[universe,]
dec.muraro2 <- dec.muraro[universe,]

library(batchelor)
normed.pancreas <- multiBatchNorm(sce.grun2, sce.muraro2)
sce.grun2 <- normed.pancreas[[1]]
sce.muraro2 <- normed.pancreas[[2]]

library(scran)
combined.pan <- combineVar(dec.grun2, dec.muraro2)
chosen.genes <- rownames(combined.pan)[combined.pan$bio > 0]
```

We observe that `rescaleBatches()` is unable to align cells from different batches in Figure \@ref(fig:tsne-pancreas-rescaled).
This is attributable to differences in population composition between batches, with additional complications from non-linearities in the batch effect, e.g., when the magnitude or direction of the batch effect differs between cell types.


```r
library(scater)
rescaled.pancreas <- rescaleBatches(sce.grun2, sce.muraro2)

set.seed(100101)
rescaled.pancreas <- runPCA(rescaled.pancreas, subset_row=chosen.genes,
    exprs_values="corrected")

rescaled.pancreas <- runTSNE(rescaled.pancreas, dimred="PCA")
plotTSNE(rescaled.pancreas, colour_by="batch")
```

<div class="figure">
<img src="P3_W10.merged-pancreas_files/figure-html/tsne-pancreas-rescaled-1.png" alt="$t$-SNE plot of the two pancreas datasets after correction with `rescaleBatches()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">(\#fig:tsne-pancreas-rescaled)$t$-SNE plot of the two pancreas datasets after correction with `rescaleBatches()`. Each point represents a cell and is colored according to the batch of origin.</p>
</div>

Here, we use `fastMNN()` to merge together the two human pancreas datasets described earlier.
Clustering on the merged datasets yields fewer batch-specific clusters, which is recapitulated as greater intermingling between batches in Figure \@ref(fig:tsne-pancreas-mnn).
This improvement over Figure \@ref(fig:tsne-pancreas-rescaled) represents the ability of `fastMNN()` to adapt to more complex situations involving differences in population composition between batches.


```r
set.seed(1011011)
mnn.pancreas <- fastMNN(sce.grun2, sce.muraro2, subset.row=chosen.genes)

snn.gr <- buildSNNGraph(mnn.pancreas, use.dimred="corrected")
clusters <- igraph::cluster_walktrap(snn.gr)$membership
tab <- table(Cluster=clusters, Batch=mnn.pancreas$batch)
tab
```

```
##        Batch
## Cluster   1   2
##      1  287 280
##      2   55   0
##      3  214 401
##      4  336 255
##      5  218 853
##      6   61 197
##      7   25 108
##      8   57  82
##      9   22 127
##      10   0  18
##      11   8   4
##      12   7  21
```

```r
mnn.pancreas <- runTSNE(mnn.pancreas, dimred="corrected")
plotTSNE(mnn.pancreas, colour_by="batch")
```

<div class="figure">
<img src="P3_W10.merged-pancreas_files/figure-html/tsne-pancreas-mnn-1.png" alt="$t$-SNE plot of the two pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin." width="672" />
<p class="caption">(\#fig:tsne-pancreas-mnn)$t$-SNE plot of the two pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin.</p>
</div>



## The bad

Flushed with our previous success, we now attempt to merge the other datasets from @lawlor2017singlecell and @segerstolpe2016singlecell.
This is a more challenging task as it involves different technologies, mixtures of UMI and read count data and a more diverse set of authors (presumably with greater differences in the patient population).

<button class="aaron-collapse">View history</button>
<div class="aaron-content">
   
```r
#--- setup ---#
library(OSCAUtils)
chapterPreamble(use_cache = TRUE)

#--- loading ---#
library(scRNAseq)
sce.lawlor <- LawlorPancreasData()

#--- gene-annotation ---#
library(AnnotationHub)
edb <- AnnotationHub()[["AH73881"]]
anno <- select(edb, keys=rownames(sce.lawlor), keytype="GENEID", 
    columns=c("SYMBOL", "SEQNAME"))
rowData(sce.lawlor) <- anno[match(rownames(sce.lawlor), anno[,1]),-1]

#--- quality-control ---#
library(scater)
stats <- perCellQCMetrics(sce.lawlor, 
    subsets=list(Mito=which(rowData(sce.lawlor)$SEQNAME=="MT")))
qc <- quickPerCellQC(stats, percent_subsets="subsets_Mito_percent")
sce.lawlor <- sce.lawlor[,!qc$discard]

#--- normalization ---#
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.lawlor)
sce.lawlor <- computeSumFactors(sce.lawlor, clusters=clusters)
sce.lawlor <- logNormCounts(sce.lawlor)

#--- variance-modelling ---#
dec.lawlor <- modelGeneVar(sce.lawlor, block=sce.lawlor$`islet unos id`)
chosen.genes <- getTopHVGs(dec.lawlor, n=2000)
```

</div>


```r
sce.lawlor
```

```
## class: SingleCellExperiment 
## dim: 26616 615 
## metadata(0):
## assays(2): counts logcounts
## rownames(26616): ENSG00000229483 ENSG00000232849 ...
##   ENSG00000251576 ENSG00000082898
## rowData names(2): SYMBOL SEQNAME
## colnames(615): 10th_C11_S96 10th_C13_S61 ... 9th-C96_S81
##   9th-C9_S13
## colData names(8): title age ... race Sex
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(0):
```

<button class="aaron-collapse">View history</button>
<div class="aaron-content">
   
```r
#--- setup ---#
library(OSCAUtils)
chapterPreamble(use_cache = TRUE)

#--- loading ---#
library(scRNAseq)
sce.seger <- SegerstolpePancreasData()

#--- gene-annotation ---#
library(AnnotationHub)
edb <- AnnotationHub()[["AH73881"]]
symbols <- rowData(sce.seger)$symbol
ens.id <- mapIds(edb, keys=symbols, keytype="SYMBOL", column="GENEID")
ens.id <- ifelse(is.na(ens.id), symbols, ens.id)

# Removing duplicated rows.
keep <- !duplicated(ens.id)
sce.seger <- sce.seger[keep,]
rownames(sce.seger) <- ens.id[keep]

#--- sample-annotation ---#
emtab.meta <- colData(sce.seger)[,c("cell type", 
    "individual", "single cell well quality")]
colnames(emtab.meta) <- c("CellType", "Donor", "Quality")
colData(sce.seger) <- emtab.meta

sce.seger$CellType <- gsub(" cell", "", sce.seger$CellType)
sce.seger$CellType <- paste0(
    toupper(substr(sce.seger$CellType, 1, 1)),
    substring(sce.seger$CellType, 2))

#--- quality-control ---#
low.qual <- sce.seger$Quality == "low quality cell"

library(scater)
stats <- perCellQCMetrics(sce.seger)
qc <- quickPerCellQC(stats, percent_subsets="altexps_ERCC_percent")
sce.seger <- sce.seger[,!(qc$discard | low.qual)]

#--- normalization ---#
library(scran)
clusters <- quickCluster(sce.seger)
sce.seger <- computeSumFactors(sce.seger, clusters=clusters)
sce.seger <- logNormCounts(sce.seger) 

#--- variance-modelling ---#
for.hvg <- sce.seger[,librarySizeFactors(altExp(sce.seger)) > 0
    & sce.seger$Donor!="AZ"]
dec.seger <- modelGeneVarWithSpikes(for.hvg, "ERCC", block=for.hvg$Donor)
chosen.hvgs <- getTopHVGs(dec.seger, n=2000)
```

</div>


```r
sce.seger
```

```
## class: SingleCellExperiment 
## dim: 25454 2169 
## metadata(0):
## assays(2): counts logcounts
## rownames(25454): ENSG00000118473 ENSG00000142920 ...
##   ENSG00000278306 eGFP
## rowData names(2): symbol refseq
## colnames(2169): HP1502401_H13 HP1502401_J14 ... HP1526901T2D_N8
##   HP1526901T2D_A8
## colData names(3): CellType Donor Quality
## reducedDimNames(0):
## spikeNames(0):
## altExpNames(1): ERCC
```

We perform the usual routine to obtain re-normalized values and a set of HVGs. 


```r
universe <- Reduce(intersect, list(rownames(sce.grun), rownames(sce.muraro),
    rownames(sce.lawlor), rownames(sce.seger)))

# Could use a loop here... but I like typing.
sce.grun2 <- sce.grun[universe,]
dec.grun2 <- dec.grun[universe,]
sce.muraro2 <- sce.muraro[universe,]
dec.muraro2 <- dec.muraro[universe,]
sce.lawlor2 <- sce.lawlor[universe,]
dec.lawlor2 <- dec.lawlor[universe,]
sce.seger2 <- sce.seger[universe,]
dec.seger2 <- dec.seger[universe,]

normed.pancreas <- multiBatchNorm(sce.grun2, sce.muraro2, sce.lawlor2, sce.seger2)
sce.grun2 <- normed.pancreas[[1]]
sce.muraro2 <- normed.pancreas[[2]]
sce.lawlor2 <- normed.pancreas[[3]]
sce.seger2 <- normed.pancreas[[4]]

combined.pan <- combineVar(dec.grun2, dec.muraro2, dec.lawlor2, dec.seger2)
chosen.genes <- rownames(combined.pan)[combined.pan$bio > 0]
```

We observe that the merge is generally successful, with many clusters containing contributions from each batch (Figure \@ref(fig:tsne-pancreas-mnn-many)).
There are few clusters that are specific to the Segerstolpe dataset, and if we were naive, we might consider them to represent interesting subpopulations that are not present in the other datasets.


```r
set.seed(1011110)
mnn.pancreas <- fastMNN(Grun=sce.grun2, Muraro=sce.muraro2, 
    Lawlor=sce.lawlor2, Seger=sce.seger2, subset.row=chosen.genes)

# Bumping up 'k' to get broader clusters,
# 'smooth over' any remaining batch effect.
snn.gr <- buildSNNGraph(mnn.pancreas, use.dimred="corrected", k=20)
clusters <- igraph::cluster_walktrap(snn.gr)$membership
tab <- table(Cluster=clusters, Batch=mnn.pancreas$batch)
tab
```

```
##        Batch
## Cluster Grun Lawlor Muraro Seger
##      1   189    256    475   216
##      2   305     11    239   134
##      3     0      0      0   315
##      4   340     26    256   391
##      5   240    243    852   362
##      6    33     11     44    54
##      7    62     20    198   112
##      8    12      2     25     7
##      9     0      0      0    52
##      10    0      0      0   220
##      11   23     19    128   195
##      12   54      0      0     0
##      13    0      0      0    38
##      14   25     19    108    56
##      15    7      8     21    17
```

```r
mnn.pancreas <- runTSNE(mnn.pancreas, dimred="corrected")
gridExtra::grid.arrange(
    plotTSNE(mnn.pancreas, colour_by="batch", text_by=I(clusters)),
    plotTSNE(mnn.pancreas, colour_by=I(factor(clusters)), text_by=I(clusters)),
    ncol=2
)
```

<div class="figure">
<img src="P3_W10.merged-pancreas_files/figure-html/tsne-pancreas-mnn-many-1.png" alt="$t$-SNE plots of the four pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin (left) or the assigned cluster (right). The cluster label is shown at the median location across all cells in the cluster." width="672" />
<p class="caption">(\#fig:tsne-pancreas-mnn-many)$t$-SNE plots of the four pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin (left) or the assigned cluster (right). The cluster label is shown at the median location across all cells in the cluster.</p>
</div>



Fortunately, we are battle-hardened and cynical, so we are sure to check for other sources of variation. 
The most obvious candidate is the donor of origin for each cell (Figure \@ref(fig:tsne-pancreas-mnn-donor)), which correlates strongly to these Segerstolpe-only clusters.
This is not surprising given the large differences between humans in the wild, but donor-level variation is not interesting for the purposes of cell type characterization.
(That said, preservation of the within-dataset donor effects is the technically correct course of action here, as a batch correction method should try to avoid removing heterogeneity within each of its defined batches.)


```r
donors <- c(sce.grun2$donor, sce.muraro2$donor,
    sce.lawlor2$`islet unos id`, sce.seger2$Donor)
seger.donors <- donors
seger.donors[mnn.pancreas$batch!="Seger"] <- NA
plotTSNE(mnn.pancreas, colour_by=I(seger.donors))
```

<div class="figure">
<img src="P3_W10.merged-pancreas_files/figure-html/tsne-pancreas-mnn-donor-1.png" alt="$t$-SNE plots of the four pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the donor of origin for the Segerstolpe dataset." width="672" />
<p class="caption">(\#fig:tsne-pancreas-mnn-donor)$t$-SNE plots of the four pancreas datasets after correction with `fastMNN()`. Each point represents a cell and is colored according to the donor of origin for the Segerstolpe dataset.</p>
</div>

## The ugly

Given these results, the most prudent course of action is to remove the donor effects within each dataset in addition to the batch effects across datasets.
This involves a bit more work to properly specify the two levels of unwanted heterogeneity.
To make it a bit easier, we use the `noCorrect()` utility to combine all batches into a single `SingleCellExperiment` object.


```r
combined <- noCorrect(Grun=sce.grun2, Muraro=sce.muraro2, 
    Lawlor=sce.lawlor2, Seger=sce.seger2)
assayNames(combined) <- "logcounts"
combined$donor <- donors
```

For each donor, we determine the number of donors from the same dataset.
This is used to weight the PCA step to ensure that each dataset contributes equally to the definition of the PC space, regardless of the number of donors in that dataset.
(Similarly, the default behavior of `fastMNN()` ensures that each batch contributes equally to the PCA, regardless of the number of cells.)


```r
donors.by.batches <- lapply(split(combined$donor, combined$batch), unique)
donors.by.batches
```

```
## $Grun
## [1] "D2"  "D3"  "D7"  "D10" "D17"
## 
## $Lawlor
## [1] "ACIW009"  "ACJV399"  "ACCG268"  "ACCR015A" "ACEK420A" "ACEL337" 
## [7] "ACHY057"  "ACIB065" 
## 
## $Muraro
## [1] "D28" "D29" "D31" "D30"
## 
## $Seger
##  [1] "HP1502401"    "HP1504101T2D" "AZ"           "HP1508501T2D"
##  [5] "HP1506401"    "HP1507101"    "HP1509101"    "HP1504901"   
##  [9] "HP1525301T2D" "HP1526901T2D"
```

```r
ndonors <- rep(lengths(donors.by.batches), lengths(donors.by.batches))
names(ndonors) <- unlist(donors.by.batches)
ndonors
```

```
##           D2           D3           D7          D10          D17 
##            5            5            5            5            5 
##      ACIW009      ACJV399      ACCG268     ACCR015A     ACEK420A 
##            8            8            8            8            8 
##      ACEL337      ACHY057      ACIB065          D28          D29 
##            8            8            8            4            4 
##          D31          D30    HP1502401 HP1504101T2D           AZ 
##            4            4           10           10           10 
## HP1508501T2D    HP1506401    HP1507101    HP1509101    HP1504901 
##           10           10           10           10           10 
## HP1525301T2D HP1526901T2D 
##           10           10
```

We then call `fastMNN()` on the combined object, using the `batch=` argument to specify which cells belong to which donors.
This will progressively merge cells from each donor until all cells are mapped onto a common coordinate space.


```r
set.seed(1010100)
multiout <- fastMNN(combined, batch=donors, subset.row=chosen.genes,
    weights=1/ndonors)

# Renaming things for easier communication later.
multiout$dataset <- combined$batch
multiout$donor <- multiout$batch
multiout$batch <- NULL
```

With this approach, we see that the Segerstolpe-only clusters have disappeared (Figure \@ref(fig:tsne-pancreas-mnn-donor-all)).
Visually, there also seems to be much greater mixing between cells from different Segerstolpe donors.
This suggests that we have removed most of the donor effect, which simplifies the interpretation of our clusters.


```r
library(scater)
g <- buildSNNGraph(multiout, use.dimred=1, k=20)
clusters <- igraph::cluster_walktrap(g)$membership
tab <- table(clusters, multiout$dataset)
tab
```

```
##         
## clusters Grun Lawlor Muraro Seger
##       1   336     22    285   188
##       2   206    260    480   314
##       3   344     25    256   391
##       4   219    241    849   889
##       5    11      2     24     8
##       6    65     20    197   115
##       7    26     18    108    56
##       8    52      0      0     0
##       9     7      9     21    17
##       10   24     18    126   191
```

```r
multiout <- runTSNE(multiout, dimred="corrected")
gridExtra::grid.arrange(
    plotTSNE(multiout, colour_by="dataset", text_by=I(clusters)),
    plotTSNE(multiout, colour_by=I(seger.donors)),
    ncol=2
)
```

<div class="figure">
<img src="P3_W10.merged-pancreas_files/figure-html/tsne-pancreas-mnn-donor-all-1.png" alt="$t$-SNE plots of the four pancreas datasets after donor-level correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin (left) or the donor of origin for the Segerstolpe-derived cells (right). The cluster label is shown at the median location across all cells in the cluster." width="960" />
<p class="caption">(\#fig:tsne-pancreas-mnn-donor-all)$t$-SNE plots of the four pancreas datasets after donor-level correction with `fastMNN()`. Each point represents a cell and is colored according to the batch of origin (left) or the donor of origin for the Segerstolpe-derived cells (right). The cluster label is shown at the median location across all cells in the cluster.</p>
</div>



Our clusters compare well to the published annotations, indicating that we did not inadvertently discard important factors of variation during correction.
(Though in this case, the cell types are so well defined, it would be quite a feat to fail to separate them!)


```r
proposed <- c(rep(NA, ncol(sce.grun)), 
    sce.muraro$label,
    sce.lawlor$`cell type`,
    sce.seger$CellType)

proposed <- tolower(proposed)
proposed[proposed=="gamma/pp"] <- "gamma"
proposed[proposed=="pp"] <- "gamma"
proposed[proposed=="duct"] <- "ductal"
proposed[proposed=="psc"] <- "stellate"
table(proposed, clusters)
```

```
##                         clusters
## proposed                    1    2    3    4    5    6    7    8    9   10
##   acinar                  424    0    0    0    1    0    0    0    1    0
##   alpha                     1   24    6 1874    1    1    0    0    1    0
##   beta                      3  945    5    3    1    0    2    0    1    7
##   co-expression             0   38    0    1    0    0    0    0    0    0
##   delta                     0    3    2    1    0  317    0    0    1    0
##   ductal                   10    4  614    1   13    0    5    0    0    1
##   endothelial               0    0    0    0    0    0    1    0   35    0
##   epsilon                   0    1    0    6    0    0    0    0    0    3
##   gamma                     2    0    0    1    0    0    0    0    0  302
##   mast                      0    0    0    0    2    0    0    0    0    0
##   mesenchymal               0    0    1    0    0    0   79    0    0    0
##   mhc class ii              0    0    0    0    5    0    0    0    0    0
##   nana                      2    2    9   16    0    1    0    0    1    1
##   none/other                0    1    3    5    1    0    0    0    4    0
##   stellate                  0    0    0    0    0    0   72    0    1    0
##   unclassified              0    0    0    0    0    0    2    0    0    0
##   unclassified endocrine    0    6    1    4    0    3    0    0    0    1
##   unclear                   0    0    4    0    0    0    0    0    0    0
```



## Session Info {-}

<button class="aaron-collapse">View session info</button>
<div class="aaron-content">
```
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.6 LTS

Matrix products: default
BLAS:   /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRblas.so
LAPACK: /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] fossil_0.3.7                shapefiles_0.7             
 [3] foreign_0.8-72              maps_3.3.0                 
 [5] sp_1.3-1                    scater_1.13.27             
 [7] ggplot2_3.2.1               scran_1.13.32              
 [9] batchelor_1.1.18            SingleCellExperiment_1.7.11
[11] SummarizedExperiment_1.15.9 DelayedArray_0.11.8        
[13] BiocParallel_1.19.5         matrixStats_0.55.0         
[15] Biobase_2.45.1              GenomicRanges_1.37.17      
[17] GenomeInfoDb_1.21.2         IRanges_2.19.17            
[19] S4Vectors_0.23.25           BiocGenerics_0.31.6        
[21] Cairo_1.5-10                BiocStyle_2.13.2           
[23] OSCAUtils_0.0.1            

loaded via a namespace (and not attached):
 [1] viridis_0.5.1            edgeR_3.27.14           
 [3] BiocSingular_1.1.7       viridisLite_0.3.0       
 [5] DelayedMatrixStats_1.7.2 assertthat_0.2.1        
 [7] statmod_1.4.32           highr_0.8               
 [9] BiocManager_1.30.8       dqrng_0.2.1             
[11] GenomeInfoDbData_1.2.1   vipor_0.4.5             
[13] yaml_2.2.0               pillar_1.4.2            
[15] lattice_0.20-38          glue_1.3.1              
[17] limma_3.41.18            digest_0.6.22           
[19] XVector_0.25.0           colorspace_1.4-1        
[21] cowplot_1.0.0            htmltools_0.4.0         
[23] Matrix_1.2-17            pkgconfig_2.0.3         
[25] bookdown_0.14            zlibbioc_1.31.0         
[27] purrr_0.3.3              scales_1.0.0            
[29] Rtsne_0.15               tibble_2.1.3            
[31] withr_2.1.2              lazyeval_0.2.2          
[33] magrittr_1.5             crayon_1.3.4            
[35] evaluate_0.14            beeswarm_0.2.3          
[37] tools_3.6.1              stringr_1.4.0           
[39] munsell_0.5.0            locfit_1.5-9.1          
[41] irlba_2.3.3              compiler_3.6.1          
[43] rsvd_1.0.2               rlang_0.4.0             
[45] grid_3.6.1               RCurl_1.95-4.12         
[47] BiocNeighbors_1.3.5      igraph_1.2.4.1          
[49] labeling_0.3             bitops_1.0-6            
[51] rmarkdown_1.16           gtable_0.3.0            
[53] codetools_0.2-16         R6_2.4.0                
[55] gridExtra_2.3            knitr_1.25              
[57] dplyr_0.8.3              stringi_1.4.3           
[59] ggbeeswarm_0.6.0         Rcpp_1.0.2              
[61] tidyselect_0.2.5         xfun_0.10               
```
</div>
