# Human pancreas dataset (Segerstolpe)

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>

<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>

## Introduction

This performs an analysis of the @segerstolpe2016singlcell dataset,
consisting of human pancreas cells from various donors.

## Analysis code

### Data loading


```r
library(scRNAseq)
sce.seger <- SegerstolpePancreasData()
```

### Gene annotation


```r
library(AnnotationHub)
edb <- AnnotationHub()[["AH73881"]]
symbols <- rowData(sce.seger)$symbol
ens.id <- mapIds(edb, keys=symbols, keytype="SYMBOL", column="GENEID")
ens.id <- ifelse(is.na(ens.id), symbols, ens.id)

# Removing duplicated rows.
keep <- !duplicated(ens.id)
sce.seger <- sce.seger[keep,]
rownames(sce.seger) <- ens.id[keep]
```

### Sample annotation

We simplify the names of some of the relevant column metadata fields for ease of access.
Some editing of the cell type labels is necessary for consistency with other data sets.


```r
emtab.meta <- colData(sce.seger)[,c("cell type", 
    "individual", "single cell well quality")]
colnames(emtab.meta) <- c("CellType", "Donor", "Quality")
colData(sce.seger) <- emtab.meta

sce.seger$CellType <- gsub(" cell", "", sce.seger$CellType)
sce.seger$CellType <- paste0(
    toupper(substr(sce.seger$CellType, 1, 1)),
    substring(sce.seger$CellType, 2))
```

### Quality control

We remove low quality cells that were marked by the authors.
We then perform additional quality control, as some of the remaining 
still have very low counts and numbers of detected features.


```r
unfiltered <- sce.seger
```


```r
low.qual <- sce.seger$Quality == "low quality cell"

library(scater)
stats <- perCellQCMetrics(sce.seger)
qc <- quickCellQC(stats, nmads=3, percent_subsets="altexps_ERCC_percent")
sce.seger <- sce.seger[,!(qc$discard | low.qual)]
```

### Normalization

We don't normalize the spike-ins as there are some cells with no spike-in counts.


```r
library(scran)
clusters <- quickCluster(sce.seger)
sce.seger <- computeSumFactors(sce.seger, clusters=clusters)
sce.seger <- logNormCounts(sce.seger, use_altexps=FALSE)
```

### Variance modelling

We do not use cells with no spike-ins for variance modelling.
Donor AZ also has very low spike-in counts and is subsequently ignored.


```r
for.hvg <- sce.seger[,librarySizeFactors(altExp(sce.seger)) > 0
    & sce.seger$Donor!="AZ"]
dec.seger <- modelGeneVarWithSpikes(for.hvg, "ERCC", block=for.hvg$Donor)
chosen.hvgs <- head(order(dec.seger$bio, decreasing=TRUE), 2000)
```

### Dimensionality reduction


```r
library(BiocSingular)
set.seed(101011001)
sce.seger <- runPCA(sce.seger, subset_row=chosen.hvgs, 
    ncomponents=25, BSPARAM=IrlbaParam())
sce.seger <- runTSNE(sce.seger, dimred="PCA")
```

### Clustering


```r
snn.gr <- buildSNNGraph(sce.seger, use.dimred="PCA")
sce.seger$cluster <- factor(igraph::cluster_walktrap(snn.gr)$membership)
```

## Results

### Quality control statistics


```r
colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard

gridExtra::grid.arrange(
    plotColData(unfiltered, x="Donor", y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, x="Donor", y="detected", colour_by="discard") +
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, x="Donor", y="altexps_ERCC_percent",
        colour_by="discard") + ggtitle("ERCC percent"),
    ncol=2
)
```

<img src="P3_W09.segerstolpe-pancreas_files/figure-html/unnamed-chunk-2-1.png" width="672" />


```r
colSums(as.matrix(qc))
```

```
##              low_lib_size            low_n_features 
##                       197                       587 
## high_altexps_ERCC_percent                   discard 
##                       902                      1028
```

### Normalization


```r
summary(sizeFactors(sce.seger))
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   0.005   0.392   0.709   1.000   1.318  11.239
```


```r
plot(librarySizeFactors(sce.seger), sizeFactors(sce.seger), pch=16,
    xlab="Library size factors", ylab="Deconvolution factors", log="xy")
```

<img src="P3_W09.segerstolpe-pancreas_files/figure-html/unnamed-chunk-5-1.png" width="672" />

### Variance modelling


```r
par(mfrow=c(3,3))
blocked.stats <- dec.seger$per.block
for (i in colnames(blocked.stats)) {
    current <- blocked.stats[[i]]
    plot(current$mean, current$total, main=i, pch=16, cex=0.5,
        xlab="Mean of log-expression", ylab="Variance of log-expression")
    curfit <- metadata(current)
    points(curfit$mean, curfit$var, col="red", pch=16)
    curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2)
}
```

<img src="P3_W09.segerstolpe-pancreas_files/figure-html/unnamed-chunk-6-1.png" width="672" />

### Dimensionality reduction


```r
ncol(reducedDim(sce.seger, "PCA"))
```

```
## [1] 25
```

### Clustering


```r
table(sce.seger$cluster)
```

```
## 
##   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18 
## 124 125 104 361 330 137 300 181  68  49  94  39  14  79  42  26  34  17 
##  19 
##  45
```


```r
plotTSNE(sce.seger, colour_by="cluster")
```

<img src="P3_W09.segerstolpe-pancreas_files/figure-html/unnamed-chunk-9-1.png" width="672" />

```r
plotTSNE(sce.seger, colour_by="Donor")
```

<img src="P3_W09.segerstolpe-pancreas_files/figure-html/unnamed-chunk-9-2.png" width="672" />


## Session Info {-}

<button class="aaron-collapse">View session info</button>
<div class="aaron-content">
```
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.6 LTS

Matrix products: default
BLAS:   /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRblas.so
LAPACK: /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] BiocSingular_1.1.7          scran_1.13.25              
 [3] scater_1.13.24              ggplot2_3.2.1              
 [5] ensembldb_2.9.6             AnnotationFilter_1.9.0     
 [7] GenomicFeatures_1.37.4      AnnotationDbi_1.47.1       
 [9] AnnotationHub_2.17.10       BiocFileCache_1.9.1        
[11] dbplyr_1.4.2                scRNAseq_1.99.6            
[13] SingleCellExperiment_1.7.11 SummarizedExperiment_1.15.9
[15] DelayedArray_0.11.6         BiocParallel_1.19.3        
[17] matrixStats_0.55.0          Biobase_2.45.1             
[19] GenomicRanges_1.37.16       GenomeInfoDb_1.21.2        
[21] IRanges_2.19.16             S4Vectors_0.23.24          
[23] BiocGenerics_0.31.6         Cairo_1.5-10               
[25] BiocStyle_2.13.2            OSCAUtils_0.0.1            

loaded via a namespace (and not attached):
 [1] Rtsne_0.15                    ggbeeswarm_0.6.0             
 [3] colorspace_1.4-1              XVector_0.25.0               
 [5] BiocNeighbors_1.3.5           bit64_0.9-7                  
 [7] interactiveDisplayBase_1.23.0 codetools_0.2-16             
 [9] knitr_1.25                    zeallot_0.1.0                
[11] Rsamtools_2.1.5               shiny_1.3.2                  
[13] BiocManager_1.30.4            compiler_3.6.1               
[15] httr_1.4.1                    dqrng_0.2.1                  
[17] backports_1.1.5               assertthat_0.2.1             
[19] Matrix_1.2-17                 lazyeval_0.2.2               
[21] limma_3.41.17                 later_0.8.0                  
[23] htmltools_0.3.6               prettyunits_1.0.2            
[25] tools_3.6.1                   rsvd_1.0.2                   
[27] igraph_1.2.4.1                gtable_0.3.0                 
[29] glue_1.3.1                    GenomeInfoDbData_1.2.1       
[31] dplyr_0.8.3                   rappdirs_0.3.1               
[33] Rcpp_1.0.2                    vctrs_0.2.0                  
[35] Biostrings_2.53.2             ExperimentHub_1.11.6         
[37] rtracklayer_1.45.6            DelayedMatrixStats_1.7.2     
[39] xfun_0.10                     stringr_1.4.0                
[41] mime_0.7                      irlba_2.3.3                  
[43] statmod_1.4.32                XML_3.98-1.20                
[45] edgeR_3.27.13                 zlibbioc_1.31.0              
[47] scales_1.0.0                  hms_0.5.1                    
[49] promises_1.0.1                ProtGenerics_1.17.4          
[51] yaml_2.2.0                    curl_4.2                     
[53] memoise_1.1.0                 gridExtra_2.3                
[55] biomaRt_2.41.9                stringi_1.4.3                
[57] RSQLite_2.1.2                 rlang_0.4.0                  
[59] pkgconfig_2.0.3               bitops_1.0-6                 
[61] evaluate_0.14                 lattice_0.20-38              
[63] purrr_0.3.2                   labeling_0.3                 
[65] GenomicAlignments_1.21.7      cowplot_1.0.0                
[67] bit_1.1-14                    tidyselect_0.2.5             
[69] magrittr_1.5                  bookdown_0.14                
[71] R6_2.4.0                      DBI_1.0.0                    
[73] pillar_1.4.2                  withr_2.1.2                  
[75] RCurl_1.95-4.12               tibble_2.1.3                 
[77] crayon_1.3.4                  rmarkdown_1.16               
[79] viridis_0.5.1                 progress_1.2.2               
[81] locfit_1.5-9.1                grid_3.6.1                   
[83] blob_1.2.0                    digest_0.6.21                
[85] xtable_1.8-4                  httpuv_1.5.2                 
[87] openssl_1.4.1                 munsell_0.5.0                
[89] beeswarm_0.2.3                viridisLite_0.3.0            
[91] vipor_0.4.5                   askpass_1.1                  
```
</div>
