# Human pancreas dataset (Grun)

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
          content.style.display = "none";
        } else {
          content.style.display = "block";
        }
    }
})
</script>

<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>

## Introduction

This performs an analysis of the @grun2016denovo CEL-seq2 dataset,
consisting of human pancreas cells from various donors.

## Analysis code

### Data loading


```r
library(scRNAseq)
sce.grun <- GrunPancreasData()
```

### Gene annotation

We convert to Ensembl identifiers, and we remove duplicated genes or genes without Ensembl IDs.


```r
library(org.Hs.eg.db)
gene.ids <- mapIds(org.Hs.eg.db, keys=rowData(sce.grun)$symbol,
    keytype="SYMBOL", column="ENSEMBL")

keep <- !is.na(gene.ids) & !duplicated(gene.ids)
sce.grun <- sce.grun[keep,]
rownames(sce.grun) <- gene.ids[keep]
```

### Quality control

This dataset lacks mitochondrial genes so we will do without.


```r
unfiltered <- sce.grun
```


```r
library(scater)
stats <- perCellQCMetrics(sce.grun)
qc <- quickPerCellQC(stats, percent_subsets="altexps_ERCC_percent")
sce.grun <- sce.grun[,!qc$discard]
```

### Normalization 


```r
library(scran)
set.seed(1000) # for irlba. 
clusters <- quickCluster(sce.grun)
sce.grun <- computeSumFactors(sce.grun, clusters=clusters)
sce.grun <- logNormCounts(sce.grun)
```

### Variance modelling

We block on a combined plate and donor factor.


```r
block <- paste0(sce.grun$sample, "_", sce.grun$donor)
dec.grun <- modelGeneVarWithSpikes(sce.grun, spikes="ERCC", block=block)
top.grun <- getTopHVGs(dec.grun, prop=0.1)
```

### Data integration


```r
library(batchelor)
set.seed(1001010)
merged.grun <- fastMNN(sce.grun, subset.row=top.grun, batch=sce.grun$donor)
```

### Dimensionality reduction


```r
set.seed(100111)
merged.grun <- runTSNE(merged.grun, dimred="corrected")
```

### Clustering


```r
snn.gr <- buildSNNGraph(merged.grun, use.dimred="corrected")
merged.grun$cluster <- factor(igraph::cluster_walktrap(snn.gr)$membership)
```

## Results

### Quality control statistics


```r
colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard

gridExtra::grid.arrange(
    plotColData(unfiltered, x="donor", y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, x="donor", y="detected", colour_by="discard") +
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, x="donor", y="altexps_ERCC_percent",
        colour_by="discard") + ggtitle("ERCC percent"),
    ncol=2
)
```

<img src="P3_W06.grun-pancreas_files/figure-html/unnamed-chunk-2-1.png" width="672" />


```r
colSums(as.matrix(qc))
```

```
##              low_lib_size            low_n_features high_altexps_ERCC_percent 
##                       101                       149                        NA 
##                   discard 
##                       438
```

### Normalization


```r
summary(sizeFactors(sce.grun))
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   0.006   0.427   0.795   1.000   1.294   9.122
```


```r
plot(librarySizeFactors(sce.grun), sizeFactors(sce.grun), pch=16,
    xlab="Library size factors", ylab="Deconvolution factors", log="xy")
```

<img src="P3_W06.grun-pancreas_files/figure-html/unnamed-chunk-5-1.png" width="672" />

### Variance modelling


```r
par(mfrow=c(6,3))
blocked.stats <- dec.grun$per.block
for (i in colnames(blocked.stats)) {
    current <- blocked.stats[[i]]
    plot(current$mean, current$total, main=i, pch=16, cex=0.5,
        xlab="Mean of log-expression", ylab="Variance of log-expression")
    curfit <- metadata(current)
    points(curfit$mean, curfit$var, col="red", pch=16)
    curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2)
}
```

<img src="P3_W06.grun-pancreas_files/figure-html/unnamed-chunk-6-1.png" width="672" />

### Data integration


```r
metadata(merged.grun)$merge.info$lost.var
```

```
##           D10      D17       D2      D3      D7
## [1,] 0.063990 0.046806 0.000000 0.00000 0.00000
## [2,] 0.004310 0.012395 0.037910 0.00000 0.00000
## [3,] 0.006735 0.008700 0.006976 0.06123 0.00000
## [4,] 0.005845 0.009381 0.007911 0.01107 0.05828
```

### Clustering


```r
table(Cluster=merged.grun$cluster, Donor=merged.grun$batch)
```

```
##        Donor
## Cluster D10 D17  D2  D3  D7
##      1   34  73  32 122  28
##      2   17  39   9  17  16
##      3   10  51   0   3  11
##      4   11 112   0  10  55
##      5   12  68  27  13  77
##      6   14  36   3  12  67
##      7   32   5   4  13   3
##      8    6  14   0   4  10
##      9   16  34  15  25  45
##      10   4  14   0   3   1
##      11   6  17   0   7  33
```


```r
plotTSNE(merged.grun, colour_by="cluster")
```

<img src="P3_W06.grun-pancreas_files/figure-html/unnamed-chunk-9-1.png" width="672" />

```r
plotTSNE(merged.grun, colour_by="batch")
```

<img src="P3_W06.grun-pancreas_files/figure-html/unnamed-chunk-9-2.png" width="672" />

## Session Info {-}

<button class="aaron-collapse">View session info</button>
<div class="aaron-content">
```
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS:   /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRblas.so
LAPACK: /home/ramezqui/Rbuild/danbuild/R-3.6.1/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=C               LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] batchelor_1.2.2             scran_1.14.3                scater_1.14.3              
 [4] ggplot2_3.2.1               org.Hs.eg.db_3.10.0         AnnotationDbi_1.48.0       
 [7] scRNAseq_2.0.2              SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.0
[10] DelayedArray_0.12.0         BiocParallel_1.20.0         matrixStats_0.55.0         
[13] Biobase_2.46.0              GenomicRanges_1.38.0        GenomeInfoDb_1.22.0        
[16] IRanges_2.20.0              S4Vectors_0.24.0            BiocGenerics_0.32.0        
[19] Cairo_1.5-10                BiocStyle_2.14.0            OSCAUtils_0.0.1            

loaded via a namespace (and not attached):
 [1] bitops_1.0-6                  bit64_0.9-7                   httr_1.4.1                   
 [4] tools_3.6.1                   backports_1.1.5               R6_2.4.1                     
 [7] irlba_2.3.3                   vipor_0.4.5                   DBI_1.0.0                    
[10] lazyeval_0.2.2                colorspace_1.4-1              withr_2.1.2                  
[13] tidyselect_0.2.5              gridExtra_2.3                 bit_1.1-14                   
[16] curl_4.2                      compiler_3.6.1                BiocNeighbors_1.4.0          
[19] labeling_0.3                  bookdown_0.15                 scales_1.0.0                 
[22] rappdirs_0.3.1                stringr_1.4.0                 digest_0.6.22                
[25] rmarkdown_1.17                XVector_0.26.0                pkgconfig_2.0.3              
[28] htmltools_0.4.0               limma_3.42.0                  dbplyr_1.4.2                 
[31] fastmap_1.0.1                 rlang_0.4.1                   RSQLite_2.1.2                
[34] shiny_1.4.0                   DelayedMatrixStats_1.8.0      dplyr_0.8.3                  
[37] RCurl_1.95-4.12               magrittr_1.5                  BiocSingular_1.2.0           
[40] GenomeInfoDbData_1.2.2        Matrix_1.2-17                 Rcpp_1.0.3                   
[43] ggbeeswarm_0.6.0              munsell_0.5.0                 viridis_0.5.1                
[46] edgeR_3.28.0                  stringi_1.4.3                 yaml_2.2.0                   
[49] zlibbioc_1.32.0               Rtsne_0.15                    BiocFileCache_1.10.2         
[52] AnnotationHub_2.18.0          grid_3.6.1                    blob_1.2.0                   
[55] dqrng_0.2.1                   promises_1.1.0                ExperimentHub_1.12.0         
[58] crayon_1.3.4                  lattice_0.20-38               cowplot_1.0.0                
[61] locfit_1.5-9.1                zeallot_0.1.0                 knitr_1.26                   
[64] pillar_1.4.2                  igraph_1.2.4.1                codetools_0.2-16             
[67] glue_1.3.1                    BiocVersion_3.10.1            evaluate_0.14                
[70] BiocManager_1.30.9            vctrs_0.2.0                   httpuv_1.5.2                 
[73] gtable_0.3.0                  purrr_0.3.3                   assertthat_0.2.1             
[76] xfun_0.11                     rsvd_1.0.2                    mime_0.7                     
[79] xtable_1.8-4                  later_1.0.0                   viridisLite_0.3.0            
[82] tibble_2.1.3                  beeswarm_0.2.3                memoise_1.1.0                
[85] statmod_1.4.32                interactiveDisplayBase_1.24.0
```
</div>
