<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Quick Start | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Quick Start | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Quick Start | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  



<meta name="date" content="2019-06-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="analytical-workflow-overview.html">
<link rel="next" href="a-basic-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-singlecellexperiment-class"><i class="fa fa-check"></i><b>4.2</b> The <em>SingleCellExperiment</em> Class</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
<li class="chapter" data-level="4.8" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-centrality-of-singlecellexperiment"><i class="fa fa-check"></i><b>4.8</b> The Centrality of SingleCellExperiment</a></li>
<li class="chapter" data-level="4.9" data-path="data-infrastructure.html"><a href="data-infrastructure.html#multimodal-data-multiassayexperiment"><i class="fa fa-check"></i><b>4.9</b> Multimodal Data: <em>MultiAssayExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html"><i class="fa fa-check"></i><b>5</b> Analytical Workflow Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#experimental-design"><i class="fa fa-check"></i><b>5.1</b> Experimental Design</a></li>
<li class="chapter" data-level="5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#preprocessing"><i class="fa fa-check"></i><b>5.2</b> Preprocessing</a></li>
<li class="chapter" data-level="5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#import-to-r"><i class="fa fa-check"></i><b>5.3</b> Import to R</a></li>
<li class="chapter" data-level="5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#data-processing"><i class="fa fa-check"></i><b>5.4</b> Data Processing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#quality-control-metrics"><i class="fa fa-check"></i><b>5.4.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="5.4.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#normalizing-data"><i class="fa fa-check"></i><b>5.4.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="5.4.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#feature-selection"><i class="fa fa-check"></i><b>5.4.3</b> Feature Selection</a></li>
<li class="chapter" data-level="5.4.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#imputation"><i class="fa fa-check"></i><b>5.4.4</b> Imputation</a></li>
<li class="chapter" data-level="5.4.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.4.5</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="5.4.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#integrating-datasets"><i class="fa fa-check"></i><b>5.4.6</b> Integrating Datasets</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#downstream-statistical-analyses"><i class="fa fa-check"></i><b>5.5</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="5.5.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#clustering"><i class="fa fa-check"></i><b>5.5.1</b> Clustering</a></li>
<li class="chapter" data-level="5.5.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#differential-expression"><i class="fa fa-check"></i><b>5.5.2</b> Differential Expression</a></li>
<li class="chapter" data-level="5.5.3" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#trajectory-analysis"><i class="fa fa-check"></i><b>5.5.3</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="5.5.4" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#annotation"><i class="fa fa-check"></i><b>5.5.4</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#accessible-reproducible-analysis"><i class="fa fa-check"></i><b>5.6</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="5.6.1" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#interactive-data-visualization"><i class="fa fa-check"></i><b>5.6.1</b> Interactive Data Visualization</a></li>
<li class="chapter" data-level="5.6.2" data-path="analytical-workflow-overview.html"><a href="analytical-workflow-overview.html#report-generation"><i class="fa fa-check"></i><b>5.6.2</b> Report Generation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>6</b> Quick Start</a><ul>
<li class="chapter" data-level="6.1" data-path="quick-start.html"><a href="quick-start.html#code"><i class="fa fa-check"></i><b>6.1</b> Code</a></li>
<li class="chapter" data-level="6.2" data-path="quick-start.html"><a href="quick-start.html#visualizations"><i class="fa fa-check"></i><b>6.2</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html"><i class="fa fa-check"></i><b>7</b> A Basic Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#preprocessing-import-to-r"><i class="fa fa-check"></i><b>7.1</b> Preprocessing &amp; Import to R</a></li>
<li class="chapter" data-level="7.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#constructing-the-singlecellexperiment"><i class="fa fa-check"></i><b>7.2</b> Constructing the SingleCellExperiment</a><ul>
<li class="chapter" data-level="7.2.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-scratch"><i class="fa fa-check"></i><b>7.2.1</b> From Scratch</a></li>
<li class="chapter" data-level="7.2.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#from-publicly-available-data"><i class="fa fa-check"></i><b>7.2.2</b> From Publicly Available Data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#data-processing-1"><i class="fa fa-check"></i><b>7.3</b> Data Processing</a><ul>
<li class="chapter" data-level="7.3.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#quality-control-metrics-1"><i class="fa fa-check"></i><b>7.3.1</b> Quality Control Metrics</a></li>
<li class="chapter" data-level="7.3.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#normalizing-data-1"><i class="fa fa-check"></i><b>7.3.2</b> Normalizing Data</a></li>
<li class="chapter" data-level="7.3.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#feature-selection-1"><i class="fa fa-check"></i><b>7.3.3</b> Feature Selection</a></li>
<li class="chapter" data-level="7.3.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>7.3.4</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#downstream-statistical-analyses-1"><i class="fa fa-check"></i><b>7.4</b> Downstream Statistical Analyses</a><ul>
<li class="chapter" data-level="7.4.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#clustering-1"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#differential-expression-1"><i class="fa fa-check"></i><b>7.4.2</b> Differential Expression</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#annotation-1"><i class="fa fa-check"></i><b>7.4.3</b> Annotation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#accessible-reproducible-analysis-1"><i class="fa fa-check"></i><b>7.5</b> Accessible &amp; Reproducible Analysis</a><ul>
<li class="chapter" data-level="7.5.1" data-path="a-basic-analysis.html"><a href="a-basic-analysis.html#interactive-data-visualization-1"><i class="fa fa-check"></i><b>7.5.1</b> Interactive Data Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>8</b> Quality Control</a><ul>
<li class="chapter" data-level="8.1" data-path="quality-control.html"><a href="quality-control.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="quality-control.html"><a href="quality-control.html#introducing-the-dataset"><i class="fa fa-check"></i><b>8.2</b> Introducing the dataset</a></li>
<li class="chapter" data-level="8.3" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>8.3</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="8.4" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>8.4</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="8.4.1" data-path="quality-control.html"><a href="quality-control.html#with-fixed-thresholds"><i class="fa fa-check"></i><b>8.4.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="8.4.2" data-path="quality-control.html"><a href="quality-control.html#with-adaptive-thresholds"><i class="fa fa-check"></i><b>8.4.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="8.4.3" data-path="quality-control.html"><a href="quality-control.html#considering-experimental-factors"><i class="fa fa-check"></i><b>8.4.3</b> Considering experimental factors</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="quality-control.html"><a href="quality-control.html#checking-diagnostic-plots"><i class="fa fa-check"></i><b>8.5</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="8.6" data-path="quality-control.html"><a href="quality-control.html#droplet-specific-procedures"><i class="fa fa-check"></i><b>8.6</b> Droplet-specific procedures</a><ul>
<li class="chapter" data-level="8.6.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>8.6.1</b> Background</a></li>
<li class="chapter" data-level="8.6.2" data-path="quality-control.html"><a href="quality-control.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>8.6.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="8.6.3" data-path="quality-control.html"><a href="quality-control.html#examining-cell-calling-diagnostics"><i class="fa fa-check"></i><b>8.6.3</b> Examining cell-calling diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="quality-control.html"><a href="quality-control.html#assumptions-of-outlier-detection"><i class="fa fa-check"></i><b>8.7</b> Assumptions of outlier detection</a><ul>
<li class="chapter" data-level="8.7.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>8.7.1</b> Overview</a></li>
<li class="chapter" data-level="8.7.2" data-path="quality-control.html"><a href="quality-control.html#checking-for-discarded-cell-types"><i class="fa fa-check"></i><b>8.7.2</b> Checking for discarded cell types</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>9</b> Normalization</a><ul>
<li class="chapter" data-level="9.1" data-path="normalization.html"><a href="normalization.html#motivation-1"><i class="fa fa-check"></i><b>9.1</b> Motivation</a></li>
<li class="chapter" data-level="9.2" data-path="normalization.html"><a href="normalization.html#setting-up-the-data"><i class="fa fa-check"></i><b>9.2</b> Setting up the data</a></li>
<li class="chapter" data-level="9.3" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>9.3</b> Library size normalization</a></li>
<li class="chapter" data-level="9.4" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>9.4</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="9.5" data-path="normalization.html"><a href="normalization.html#normalization-by-spike-ins"><i class="fa fa-check"></i><b>9.5</b> Normalization by spike-ins</a><ul>
<li class="chapter" data-level="9.5.1" data-path="normalization.html"><a href="normalization.html#differences-in-total-rna-content"><i class="fa fa-check"></i><b>9.5.1</b> Differences in total RNA content</a></li>
<li class="chapter" data-level="9.5.2" data-path="normalization.html"><a href="normalization.html#differences-in-cell-damage"><i class="fa fa-check"></i><b>9.5.2</b> Differences in cell damage</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="normalization.html"><a href="normalization.html#transformation"><i class="fa fa-check"></i><b>9.6</b> Transformation</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering-2.html"><a href="clustering-2.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="clustering-2.html"><a href="clustering-2.html#loading-the-data"><i class="fa fa-check"></i><b>10.1</b> Loading the Data</a></li>
<li class="chapter" data-level="10.2" data-path="clustering-2.html"><a href="clustering-2.html#manual-clustering"><i class="fa fa-check"></i><b>10.2</b> Manual Clustering</a><ul>
<li class="chapter" data-level="10.2.1" data-path="clustering-2.html"><a href="clustering-2.html#varying-k"><i class="fa fa-check"></i><b>10.2.1</b> Varying <code>k</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="clustering-2.html"><a href="clustering-2.html#varying-clustering-algorithms"><i class="fa fa-check"></i><b>10.2.2</b> Varying Clustering Algorithms</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="clustering-2.html"><a href="clustering-2.html#automated-clustering"><i class="fa fa-check"></i><b>10.3</b> Automated Clustering</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html"><i class="fa fa-check"></i><b>11</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="11.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#preprocessing-1"><i class="fa fa-check"></i><b>11.1</b> Preprocessing</a></li>
<li class="chapter" data-level="11.2" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#fast-mnn"><i class="fa fa-check"></i><b>11.2</b> Fast MNN</a></li>
<li class="chapter" data-level="11.3" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#harmony"><i class="fa fa-check"></i><b>11.3</b> Harmony</a></li>
<li class="chapter" data-level="11.4" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#advanced-mnn-workflow"><i class="fa fa-check"></i><b>11.4</b> Advanced MNN Workflow</a><ul>
<li class="chapter" data-level="11.4.1" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>11.4.1</b> Dimensionality Reduction</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#naive-method-without-correction"><i class="fa fa-check"></i><b>11.5</b> Naive Method Without Correction</a></li>
<li class="chapter" data-level="11.6" data-path="integrating-datasets-1.html"><a href="integrating-datasets-1.html#limma-batch-correction"><i class="fa fa-check"></i><b>11.6</b> Limma Batch Correction</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="differential-expression-2.html"><a href="differential-expression-2.html"><i class="fa fa-check"></i><b>12</b> Differential Expression</a></li>
<li class="chapter" data-level="13" data-path="trajectory-analysis-1.html"><a href="trajectory-analysis-1.html"><i class="fa fa-check"></i><b>13</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="14" data-path="annotation-2.html"><a href="annotation-2.html"><i class="fa fa-check"></i><b>14</b> Annotation</a></li>
<li class="chapter" data-level="15" data-path="sharing-and-interactive-interfaces.html"><a href="sharing-and-interactive-interfaces.html"><i class="fa fa-check"></i><b>15</b> Sharing and Interactive Interfaces</a></li>
<li class="chapter" data-level="16" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html"><i class="fa fa-check"></i><b>16</b> Adaptations for Large-scale Data</a><ul>
<li class="chapter" data-level="16.1" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#approximate-methods"><i class="fa fa-check"></i><b>16.1</b> Approximate Methods</a></li>
<li class="chapter" data-level="16.2" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#parallelization"><i class="fa fa-check"></i><b>16.2</b> Parallelization</a></li>
<li class="chapter" data-level="16.3" data-path="adaptations-for-large-scale-data.html"><a href="adaptations-for-large-scale-data.html#on-disk-data"><i class="fa fa-check"></i><b>16.3</b> On-Disk Data</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="data-import-and-export.html"><a href="data-import-and-export.html"><i class="fa fa-check"></i><b>17</b> Data Import and Export</a></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="18" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>18</b> About the Data</a><ul>
<li class="chapter" data-level="18.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>18.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="18.2" data-path="about-the-data.html"><a href="about-the-data.html#cellbench_data"><i class="fa fa-check"></i><b>18.2</b> Cellbench_data</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="about-the-contributors.html"><a href="about-the-contributors.html"><i class="fa fa-check"></i><b>19</b> About the Contributors</a><ul>
<li class="chapter" data-level="19.0.1" data-path="about-the-contributors.html"><a href="about-the-contributors.html#stephanie-hicks-phd"><i class="fa fa-check"></i><b>19.0.1</b> <em>Stephanie Hicks, PhD</em></a></li>
<li class="chapter" data-level="19.0.2" data-path="about-the-contributors.html"><a href="about-the-contributors.html#raphael-gottardo-phd"><i class="fa fa-check"></i><b>19.0.2</b> <em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="19.0.3" data-path="about-the-contributors.html"><a href="about-the-contributors.html#robert-amezquita-phd"><i class="fa fa-check"></i><b>19.0.3</b> <em>Robert Amezquita, PhD</em></a></li>
<li class="chapter" data-level="19.0.4" data-path="about-the-contributors.html"><a href="about-the-contributors.html#aaron-lun-phd"><i class="fa fa-check"></i><b>19.0.4</b> <em>Aaron Lun, PhD</em></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="quick-start" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Quick Start</h1>
<p>To make it as easy as possible to get started fast, here we simply provide a script that walks through a typical, basic scRNA-seq analysis in code, with prose as comments (<code>#</code>), and all visualization held until the end of the script. The next chapter - “A Basic Analysis” - will provide more commentary on the various steps throughout, as well as relevant intermediate plotting results.</p>
<p>Here, we use an example dataset from the <a href="https://preview.data.humancellatlas.org"><em>Human Cell Atlas</em> immune cell profiling project on bone marrow</a>. This dataset is loaded via the <em>HCAData</em> package, which provides a ready to use <em>SingleCellExperiment</em> object.</p>
<p>Note that the <em>HCAData</em> bone marrow dataset is comprised of 8 donors, so we have added an integration step to ameliorate batch effects caused by different donors. However, for use cases where integration is not necessary (e.g. no expected batch effects), we note in the code what to skip and relevant arguments to replace.</p>
<p>Lastly, note that some arguments are added for the sake of reducing computational runtime and can be modified or removed. These include parallelization via <code>BPPARAM</code>, and different algorithms for SVD and nearest-neighbor via <code>BSPARAM</code> and <code>BNPARAM</code>. See the “Adaptations for Large-scale Data” chapter for more information on these arguments.</p>
<div id="code" class="section level2">
<h2><span class="header-section-number">6.1</span> Code</h2>
<pre class="sourceCode r"><code class="sourceCode r">## Setup ---------------------------------------------------
## not run - uncomment these lines to install necessary pkgs

## install.packages(&#39;BiocManager&#39;)
## BiocManager::install(version = &#39;devel&#39;) # devel=3.10
## BiocManager::install(c(
##   &#39;HCAData&#39;,                       # dataset
##   &#39;scater&#39;, &#39;scran&#39;, &#39;batchelor&#39;,  # processing
##   &#39;igraph&#39;,                        # clustering
##   &#39;MAST&#39;, &#39;slingshot&#39;,             # DE + Trajectory
##   &#39;iSEE&#39;                           # interactive viz
## ))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Import data into R --------------------------------------

## For reading in data directly from CellRanger output
## use the lines below and replace with proper paths to data
## append any cell metadata as needed to colData()
## library(DropletUtils)
## sce &lt;- read10xCounts(&#39;/path/to/cellranger/outs/&#39;)

## For this quick-start: Human Cell Atlas (HCA) data
<span class="kw">library</span>(HCAData)
sce &lt;-<span class="st"> </span><span class="kw">HCAData</span>(<span class="st">&#39;ica_bone_marrow&#39;</span>)

## subsample for better brevity of compilation
<span class="kw">set.seed</span>(<span class="dv">1234</span>)
sce &lt;-<span class="st"> </span>sce[, <span class="kw">sample</span>(<span class="kw">ncol</span>(sce), <span class="dv">10000</span>)]

## Split out donor based on barcode
Donor &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce<span class="op">$</span>Barcode, strsplit, <span class="st">&#39;_&#39;</span>)
Donor &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(Donor, <span class="cf">function</span>(x) { x[[<span class="dv">1</span>]][<span class="dv">1</span>] }))
sce<span class="op">$</span>Donor &lt;-<span class="st"> </span>Donor

## Convert DelayedArray to regular matrix
<span class="kw">counts</span>(sce) &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">counts</span>(sce))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Quality Control -----------------------------------------
<span class="kw">library</span>(scater)
sce &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(sce,
                          <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

## remove &quot;bad&quot; cells by total counts/features per cell
filt &lt;-<span class="st"> </span>sce<span class="op">$</span>total_counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">500</span> <span class="op">&amp;</span><span class="st"> </span>sce<span class="op">$</span>total_features_by_counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">100</span>
sce &lt;-<span class="st"> </span>sce[, filt]

## to ease computation, remove low frequency genes from `sce`
num_reads &lt;-<span class="st"> </span><span class="dv">1</span>                  <span class="co"># minimum 1 read</span>
num_cells &lt;-<span class="st"> </span><span class="fl">0.025</span> <span class="op">*</span><span class="st"> </span><span class="kw">ncol</span>(sce)   <span class="co"># in at least 2.5% of all cells</span>
keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">counts</span>(sce) <span class="op">&gt;=</span><span class="st"> </span>num_reads) <span class="op">&gt;=</span><span class="st"> </span>num_cells
sce &lt;-<span class="st"> </span>sce[keep, ]

## for readability, use Symbols in lieu of IDs as rownames
uniq_feats &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(<span class="dt">ID =</span> <span class="kw">rowData</span>(sce)<span class="op">$</span>ID,
                                   <span class="dt">names =</span> <span class="kw">rowData</span>(sce)<span class="op">$</span>Symbol)
<span class="kw">rownames</span>(sce) &lt;-<span class="st"> </span>uniq_feats</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Normalization -------------------------------------------
sce &lt;-<span class="st"> </span><span class="kw">normalize</span>(sce)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Feature Selection ---------------------------------------
<span class="kw">library</span>(scran)

fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce, <span class="dt">use.spikes =</span> <span class="ot">FALSE</span>)
dec &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce, fit)
hvg &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec)[dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>] <span class="co"># save gene names</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Integration ---------------------------------------------
## only perform this section if there is a batch effect
<span class="kw">library</span>(batchelor)

<span class="kw">set.seed</span>(<span class="dv">1234</span>)
mnn &lt;-<span class="st"> </span><span class="kw">fastMNN</span>(sce, <span class="dt">batch =</span> sce<span class="op">$</span>Donor,
               <span class="dt">subset.row =</span> hvg,
               <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(<span class="dt">deferred =</span> <span class="ot">TRUE</span>),
               <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
               <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

<span class="kw">reducedDim</span>(sce, <span class="st">&#39;MNN&#39;</span>) &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(mnn, <span class="st">&#39;corrected&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Dimensionality Reduction --------------------------------
## note on `use_dimred` arg: specifies which precomputed
## dimension reduction to use in `sce`; if there is none,
## it will first calculate and save PCA to `sce` then UMAP
<span class="kw">set.seed</span>(<span class="dv">1234</span>)
sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce,
               <span class="dt">use_dimred =</span> <span class="st">&#39;MNN&#39;</span>, <span class="co"># omit if `fastMNN()` not run</span>
               <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
               <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>(),
               ## unnecessary options, only used to make a pretty graph
               <span class="dt">min_dist =</span> <span class="fl">0.5</span>, <span class="dt">repulsion_strength =</span> <span class="fl">0.25</span>,
               <span class="dt">spread =</span> <span class="fl">0.7</span>,
               <span class="dt">n_neighbors =</span> <span class="dv">15</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Clustering ----------------------------------------------
<span class="kw">library</span>(igraph)

## replace `use.dimred` with &#39;PCA&#39; if no integration was performed
## this will be automatically added via `runUMAP` above
<span class="kw">set.seed</span>(<span class="dv">1234</span>)
g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce, <span class="dt">use.dimred =</span> <span class="st">&#39;MNN&#39;</span>,
                   <span class="dt">k =</span> <span class="dv">30</span>,  <span class="co"># higher = bigger clusters</span>
                   <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
                   <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())
clusters &lt;-<span class="st"> </span><span class="kw">as.factor</span>(igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership)
sce<span class="op">$</span>clusters &lt;-<span class="st"> </span>clusters</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Differential Expression ---------------------------------
## pval.type = &#39;all&#39; : only get globally unique markers
markers &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce, <span class="dt">clusters =</span> sce<span class="op">$</span>clusters,
                       <span class="dt">block =</span> sce<span class="op">$</span>Donor, <span class="co"># use to get within-donor DE</span>
                       <span class="dt">direction =</span> <span class="st">&#39;up&#39;</span>, <span class="dt">lfc =</span> <span class="fl">1.5</span>,
                       <span class="dt">pval.type =</span> <span class="st">&quot;all&quot;</span>, <span class="co"># get cluster-unique markers</span>
                       <span class="dt">subset.row =</span> hvg,
                       <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())

## pval.type = &#39;any&#39; : get all potential markers of any direction/comparison
markers_any &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce, <span class="dt">clusters =</span> sce<span class="op">$</span>clusters,
                           <span class="dt">block =</span> sce<span class="op">$</span>Donor, <span class="co"># use to get within-donor DE</span>
                           <span class="dt">direction =</span> <span class="st">&#39;any&#39;</span>, <span class="dt">lfc =</span> <span class="dv">0</span>,
                           <span class="dt">pval.type =</span> <span class="st">&quot;any&quot;</span>, <span class="co"># get all potential markers</span>
                           <span class="dt">subset.row =</span> hvg,
                           <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Annotation ----------------------------------------------
## Get mappings of ENTREZID to Symbol
<span class="kw">library</span>(org.Hs.eg.db)

keys_entrez &lt;-<span class="st"> </span><span class="kw">keys</span>(org.Hs.eg.db, <span class="st">&#39;ENTREZID&#39;</span>)

mapping_es &lt;-<span class="st"> </span>AnnotationDbi<span class="op">::</span><span class="kw">select</span>(org.Hs.eg.db,
                                    <span class="dt">keys =</span> keys_entrez,
                                    <span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">&#39;ENTREZID&#39;</span>, <span class="st">&#39;SYMBOL&#39;</span>),
                                    <span class="dt">keytype =</span> <span class="st">&#39;ENTREZID&#39;</span>)
mapping_es<span class="op">$</span>ENTREZID &lt;-<span class="st"> </span><span class="kw">as.integer</span>(mapping_es<span class="op">$</span>ENTREZID)

## Get pathways of interest - convert to list with symbol
## devtools::install_github(&#39;stephenturner/msigdf&#39;)
<span class="kw">library</span>(msigdf)
<span class="kw">library</span>(dplyr)

mdb &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">inner_join</span>(msigdf.human, mapping_es,
                         <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&#39;entrez&#39;</span> =<span class="st"> &#39;ENTREZID&#39;</span>)) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">filter</span>(collection <span class="op">==</span><span class="st"> &#39;c7&#39;</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>collection, <span class="op">-</span>entrez) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">group_nest</span>(geneset)

pathways &lt;-<span class="st"> </span>purrr<span class="op">::</span><span class="kw">map</span>(mdb<span class="op">$</span>data, <span class="cf">function</span>(x) {
    <span class="kw">as.character</span>(x<span class="op">$</span>SYMBOL)
})
<span class="kw">names</span>(pathways) &lt;-<span class="st"> </span>mdb<span class="op">$</span>geneset

## Get stats based on markers search - compare clusters 3 vs 2
stats &lt;-<span class="st"> </span>markers_any[[<span class="dv">3</span>]]<span class="op">$</span>logFC<span class="fl">.2</span>
<span class="kw">names</span>(stats) &lt;-<span class="st"> </span><span class="kw">rownames</span>(markers_any[[<span class="dv">3</span>]])

## Run fast gene set enrichment analysis (see plot at bottom)
<span class="kw">library</span>(fgsea)

fgseaRes &lt;-<span class="st"> </span><span class="kw">fgsea</span>(<span class="dt">pathways =</span> pathways, 
                  <span class="dt">stats =</span> stats,
                  <span class="dt">minSize =</span> <span class="dv">15</span>,
                  <span class="dt">maxSize =</span> <span class="dv">500</span>,
                  <span class="dt">nperm =</span> <span class="dv">10000</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Trajectory Analysis -------------------------------------
<span class="kw">library</span>(slingshot)

slc &lt;-<span class="st"> </span><span class="kw">slingshot</span>(sce[, sce<span class="op">$</span>clusters <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">9</span>)],
                 <span class="dt">clusterLabels =</span> <span class="st">&#39;clusters&#39;</span>,
                 <span class="dt">reducedDim =</span> <span class="st">&#39;UMAP&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Interactive Exploration ---------------------------------
## library(iSEE)
## iSEE(sce) ## not run; opens a web browser GUI</code></pre>
</div>
<div id="visualizations" class="section level2">
<h2><span class="header-section-number">6.2</span> Visualizations</h2>
<pre class="sourceCode r"><code class="sourceCode r">## Visualizations ------------------------------------------
## todo: quality control: histogram of total counts
## todo: normalization: ???
## todo: feature selection: variance v mean expression?
##
## PCA/UMAP: before and after batch correction
## UMAP (no colours)
## UMAP (with clusters colours)
## UMAP (with cell type labels)
## heatmap (top marker genes)
## trajectory plot (all or on subset of cells/within a clust)
## DE (volcano plot)
## Gene set enrichment plot</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## Trajectory analysis -------------------------------------
## Slingshot trajectory plot
<span class="kw">library</span>(RColorBrewer)
colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">brewer.pal</span>(<span class="dv">11</span>, <span class="st">&#39;Spectral&#39;</span>)[<span class="op">-</span><span class="dv">6</span>])(<span class="dv">100</span>)
plotcol &lt;-<span class="st"> </span>colors[<span class="kw">cut</span>(slc<span class="op">$</span>slingPseudotime_<span class="dv">1</span>, <span class="dt">breaks =</span> <span class="dv">100</span>)]

<span class="kw">plot</span>(<span class="kw">reducedDims</span>(slc)<span class="op">$</span>UMAP, <span class="dt">col =</span> plotcol, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">asp =</span> <span class="dv">1</span>)
<span class="kw">lines</span>(<span class="kw">SlingshotDataSet</span>(slc), <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&#39;black&#39;</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## UMAP based plots ----------------------------------------
## UMAP (no colours)
<span class="kw">plotUMAP</span>(sce)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Pre vs post batch correction
tmp &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce,
              <span class="dt">BSPARAM =</span> BiocSingular<span class="op">::</span><span class="kw">IrlbaParam</span>(),
              <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>())
tmp &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(tmp,
               <span class="dt">BNPARAM =</span> BiocNeighbors<span class="op">::</span><span class="kw">AnnoyParam</span>(),
               <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>(),
               ## unnecessary options, only used to make a pretty graph
               <span class="dt">min_dist =</span> <span class="fl">0.5</span>, <span class="dt">repulsion_strength =</span> <span class="fl">0.25</span>,
               <span class="dt">spread =</span> <span class="fl">0.7</span>,
               <span class="dt">n_neighbors =</span> <span class="dv">15</span>)
p0 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(tmp, <span class="dt">colour_by =</span> <span class="st">&#39;Donor&#39;</span>)
p1 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;Donor&#39;</span>)
patchwork<span class="op">::</span><span class="kw">wrap_plots</span>(p0, p1, <span class="dt">nrow =</span> <span class="dv">2</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-15-2.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Gene expression on UMAP plots
p2 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;CD3E&#39;</span>)
p3 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;CD79A&#39;</span>)
p4 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;LYZ&#39;</span>)
p5 &lt;-<span class="st"> </span><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;NKG7&#39;</span>)

patchwork<span class="op">::</span><span class="kw">wrap_plots</span>(p2, p3, p4, p5, <span class="dt">nrow =</span> <span class="dv">2</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-15-3.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Clusters on UMAP
<span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by =</span> <span class="st">&#39;clusters&#39;</span>, <span class="dt">text_by =</span> <span class="st">&#39;clusters&#39;</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-15-4.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Gene expression plots -----------------------------------
## Heatmap: Top global markers per cluster
top_markers &lt;-<span class="st"> </span><span class="kw">lapply</span>(markers, <span class="cf">function</span>(x) {
    <span class="kw">rownames</span>(x)[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]
})

top_markers &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(<span class="kw">unlist</span>(top_markers)))
top_markers &lt;-<span class="st"> </span>top_markers[<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;MT-|^RP&#39;</span>, top_markers)]

<span class="kw">set.seed</span>(<span class="dv">1234</span>)
<span class="kw">plotHeatmap</span>(sce[, <span class="kw">sample</span>(<span class="kw">ncol</span>(sce), <span class="dv">5000</span>)],
            <span class="dt">features =</span> top_markers,
            <span class="dt">color =</span> viridis<span class="op">::</span><span class="kw">viridis</span>(<span class="dv">101</span>, <span class="dt">option =</span> <span class="st">&#39;A&#39;</span>),
##            symmetric = TRUE, zlim = c(-5, 5),
            <span class="dt">colour_columns_by =</span> <span class="st">&#39;clusters&#39;</span>,
            <span class="dt">clustering_method =</span> <span class="st">&#39;ward.D2&#39;</span>,
            <span class="dt">show_colnames =</span> <span class="ot">FALSE</span>,
            <span class="dt">fontsize_row =</span> <span class="dv">6</span>
            )</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Volcano plot
marker_tbl_<span class="dv">3</span> &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(markers_any[[<span class="dv">3</span>]])

marker_tbl_<span class="dv">3</span> <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> logFC<span class="fl">.2</span>, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(FDR))) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="kw">log</span>(<span class="fl">1.5</span>), <span class="kw">log</span>(<span class="fl">1.5</span>)), <span class="dt">linetype =</span> <span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_classic</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">coord_cartesian</span>(<span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">expand =</span> <span class="ot">FALSE</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-16-2.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Gene set enrichment -------------------------------------
## Plot multiple pathways enrichment plots
topPathwaysUp &lt;-<span class="st"> </span>fgseaRes[ES <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>][<span class="kw">head</span>(<span class="kw">order</span>(pval), <span class="dt">n=</span><span class="dv">10</span>), pathway]
topPathwaysDown &lt;-<span class="st"> </span>fgseaRes[ES <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span>][<span class="kw">head</span>(<span class="kw">order</span>(pval), <span class="dt">n=</span><span class="dv">10</span>), pathway]
topPathways &lt;-<span class="st"> </span><span class="kw">c</span>(topPathwaysUp, <span class="kw">rev</span>(topPathwaysDown))
<span class="kw">plotGseaTable</span>(pathways[topPathways], stats,
              fgseaRes, 
              <span class="dt">gseaParam =</span> <span class="fl">0.5</span>)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">## Traditional GSEA plot
<span class="kw">plotEnrichment</span>(pathways[[<span class="st">&quot;GSE29618_MONOCYTE_VS_PDC_UP&quot;</span>]],
               stats)</code></pre>
<p><img src="P2_W0.quick-start_files/figure-html/unnamed-chunk-17-2.png" width="672" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="analytical-workflow-overview.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="a-basic-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis/P2_W0.quick-start.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
