<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 7 Workflow: Clustering | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 7 Workflow: Clustering | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Workflow: Clustering | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online companion to ‘Orchestrating Single-Cell Analysis with Bioconductor’ manuscript by the Bioconductor team." />
  

<meta name="author" content="Robert A. Amezquita">
<meta name="author" content="Bioconductor">


<meta name="date" content="2019-03-05">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="workflow-integrating-datasets.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Orchestrating Single-Cell Analysis with Bioconductor</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#prerequisites"><i class="fa fa-check"></i><b>4.1</b> Prerequisites</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-essentials-of-sce"><i class="fa fa-check"></i><b>4.2</b> The Essentials of <code>sce</code></a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#primary-data-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Primary Data: The <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#extending-the-assays-slot"><i class="fa fa-check"></i><b>4.2.2</b> Extending the <code>assays</code> Slot</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-metadata-coldata-slot"><i class="fa fa-check"></i><b>4.2.3</b> Column (Meta)Data: <code>colData</code> Slot</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#feature-metadata-rowdatarowranges"><i class="fa fa-check"></i><b>4.2.4</b> Feature Metadata: <code>rowData</code>/<code>rowRanges</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors-slot-sizefactors"><i class="fa fa-check"></i><b>4.2.5</b> Size Factors Slot: <code>sizeFactors</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#a-brief-recap-from-se-to-sce"><i class="fa fa-check"></i><b>4.3</b> A Brief Recap: From <code>se</code> to <code>sce</code></a></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#the-reduceddims-slot"><i class="fa fa-check"></i><b>4.4</b> The <code>reducedDims</code> Slot</a></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#one-more-thing-metadata-slot"><i class="fa fa-check"></i><b>4.5</b> One More Thing: <code>metadata</code> Slot</a></li>
<li class="chapter" data-level="4.6" data-path="data-infrastructure.html"><a href="data-infrastructure.html#about-spike-ins"><i class="fa fa-check"></i><b>4.6</b> About Spike-Ins</a></li>
<li class="chapter" data-level="4.7" data-path="data-infrastructure.html"><a href="data-infrastructure.html#working-with-singlecellexperiment"><i class="fa fa-check"></i><b>4.7</b> Working with <em>SingleCellExperiment</em></a></li>
</ul></li>
<li class="part"><span><b>II Workflows</b></span></li>
<li class="chapter" data-level="5" data-path="about-the-data.html"><a href="about-the-data.html"><i class="fa fa-check"></i><b>5</b> About the Data</a><ul>
<li class="chapter" data-level="5.1" data-path="about-the-data.html"><a href="about-the-data.html#x-genomics-pbmc-data"><i class="fa fa-check"></i><b>5.1</b> 10X Genomics PBMC Data</a></li>
<li class="chapter" data-level="5.2" data-path="about-the-data.html"><a href="about-the-data.html#human-cell-atlas"><i class="fa fa-check"></i><b>5.2</b> Human Cell Atlas</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Workflow: Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#package-requirements"><i class="fa fa-check"></i><b>6.1</b> Package Requirements</a></li>
<li class="chapter" data-level="6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#loading-the-data"><i class="fa fa-check"></i><b>6.2</b> Loading the Data</a></li>
<li class="chapter" data-level="6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#preprocessing"><i class="fa fa-check"></i><b>6.3</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.3.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#working-with-common-genes"><i class="fa fa-check"></i><b>6.3.1</b> Working with Common Genes</a></li>
<li class="chapter" data-level="6.3.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>6.3.2</b> Cell and Gene Quality Control</a></li>
<li class="chapter" data-level="6.3.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#normalization"><i class="fa fa-check"></i><b>6.3.3</b> Normalization</a></li>
<li class="chapter" data-level="6.3.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#multibatch-normalization"><i class="fa fa-check"></i><b>6.3.4</b> Multibatch Normalization</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a></li>
<li class="chapter" data-level="6.5" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#combining-the-datasets"><i class="fa fa-check"></i><b>6.5</b> Combining the Datasets</a></li>
<li class="chapter" data-level="6.6" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#integrating-datasets"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.6.1" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#naive-approach"><i class="fa fa-check"></i><b>6.6.1</b> Naive Approach</a></li>
<li class="chapter" data-level="6.6.2" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#limma-batch-correction"><i class="fa fa-check"></i><b>6.6.2</b> Limma Batch Correction</a></li>
<li class="chapter" data-level="6.6.3" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#mnn-approach"><i class="fa fa-check"></i><b>6.6.3</b> MNN Approach</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="workflow-integrating-datasets.html"><a href="workflow-integrating-datasets.html#session-info"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="workflow-clustering.html"><a href="workflow-clustering.html"><i class="fa fa-check"></i><b>7</b> Workflow: Clustering</a><ul>
<li class="chapter" data-level="7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#package-requirements-1"><i class="fa fa-check"></i><b>7.1</b> Package Requirements</a></li>
<li class="chapter" data-level="7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#loading-the-data-1"><i class="fa fa-check"></i><b>7.2</b> Loading the Data</a></li>
<li class="chapter" data-level="7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#understanding-the-data"><i class="fa fa-check"></i><b>7.3</b> Understanding the Data</a></li>
<li class="chapter" data-level="7.4" data-path="workflow-clustering.html"><a href="workflow-clustering.html#preprocessing-1"><i class="fa fa-check"></i><b>7.4</b> Preprocessing</a></li>
<li class="chapter" data-level="7.5" data-path="workflow-clustering.html"><a href="workflow-clustering.html#feature-selection-1"><i class="fa fa-check"></i><b>7.5</b> Feature Selection</a></li>
<li class="chapter" data-level="7.6" data-path="workflow-clustering.html"><a href="workflow-clustering.html#dimension-reduction"><i class="fa fa-check"></i><b>7.6</b> Dimension Reduction</a></li>
<li class="chapter" data-level="7.7" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clustering"><i class="fa fa-check"></i><b>7.7</b> Clustering</a><ul>
<li class="chapter" data-level="7.7.1" data-path="workflow-clustering.html"><a href="workflow-clustering.html#sc3"><i class="fa fa-check"></i><b>7.7.1</b> SC3</a></li>
<li class="chapter" data-level="7.7.2" data-path="workflow-clustering.html"><a href="workflow-clustering.html#clusterexperiment"><i class="fa fa-check"></i><b>7.7.2</b> clusterExperiment</a></li>
<li class="chapter" data-level="7.7.3" data-path="workflow-clustering.html"><a href="workflow-clustering.html#biocneighbors"><i class="fa fa-check"></i><b>7.7.3</b> BiocNeighbors</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="workflow-clustering" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Workflow: Clustering</h1>
<p>The purpose of this case study is to demonstrate various approaches to clustering scRNA-seq datasets using R/Bioconductor packages. In this workflow, we go from preprocessing the data to clustering the data. Furthermore, we highlight methods which are especially suitable for large datasets.</p>
<p>Here we will be working with a dataset from the <a href="https://github.com/LuyiTian/CellBench_data">CellBench</a> scRNA-seq benchmarking dataset collection. Specifically, we will be working with the <code>sc_10x_5cl</code> dataset, which contains 5 sorted cell lines that were sequencing on the 10X Genomics platform. We will use this to showcase our different clustering strategies.</p>
<div id="package-requirements-1" class="section level2">
<h2><span class="header-section-number">7.1</span> Package Requirements</h2>
<p>These packages will be required for working through the vignette, and can be installed by running the code below:</p>
<pre class="sourceCode r"><code class="sourceCode r">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&#39;scater&#39;</span>, <span class="st">&#39;scran&#39;</span>,
                      <span class="st">&#39;SC3&#39;</span>, <span class="st">&#39;clusterExperiment&#39;</span>, <span class="st">&#39;BiocNeighbors&#39;</span>,
                      <span class="st">&#39;BiocParallel&#39;</span>))

<span class="co">## suggested for combining plots</span>
devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&#39;thomasp85/patchwork&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)
<span class="kw">library</span>(scran)
<span class="kw">library</span>(SC3)
<span class="kw">library</span>(clusterExperiment)
<span class="kw">library</span>(BiocNeighbors)
<span class="kw">library</span>(BiocParallel)
<span class="kw">library</span>(patchwork)</code></pre>
</div>
<div id="loading-the-data-1" class="section level2">
<h2><span class="header-section-number">7.2</span> Loading the Data</h2>
<p>To follow along with the workflow, we use the <a href="https://github.com/LuyiTian/CellBench_data">CellBench_data</a> Github repository <code>data</code> folder’s <code>sincell_with_class_5cl.RData</code> workspace, which contains an object called <code>sc_10x_5cl_qc</code>. While this data will eventually be submitted to ExperimentHub, a central repository for datasets, it is currently as yet unavailable there.</p>
<p>For the time being, the data can be found in this book’s Github repo in the <code>_rfiles/_data</code> folder. The data arrives as a <code>SingleCellExperiment</code> class object.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;_rfiles/_data/cellbench_sce_sc_10x_5cl_qc.rds&quot;</span>)</code></pre>
</div>
<div id="understanding-the-data" class="section level2">
<h2><span class="header-section-number">7.3</span> Understanding the Data</h2>
<p>Within the cell metadata stored in the <code>colData</code> slot of our <code>sce</code> object, we see that information regarding the cell line of origin for each cell is stored under the column <code>cell_line</code>, and furthermore can count the number of instances of each cell line present in the dataset as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(sce<span class="op">$</span>cell_line)</code></pre>
<pre><code>## 
##   A549  H1975  H2228   H838 HCC827 
##   1244    515    751    840    568</code></pre>
<p>We will be using this information to illustrate our dimensionality reduction by highlighting each of the cell lines, and furthermore using this information to verify our clustering performance. While ground truth is not often known in practice, such benchmarking datasets are essential for validation of methods. While simulations are another source of establishing ground truth, this dataset is particularly appealing as it is derived under realistic experimental conditions.</p>
</div>
<div id="preprocessing-1" class="section level2">
<h2><span class="header-section-number">7.4</span> Preprocessing</h2>
<p>This dataset has already undergone cell quality control, so in this case we will skip this step (for this, we refer to the <a href="workflow-integrating-datasets.html#workflow-integrating-datasets">Workflow on Integrating Datasets</a>. Thus, we will begin with the feature selection step and go on from there.</p>
</div>
<div id="feature-selection-1" class="section level2">
<h2><span class="header-section-number">7.5</span> Feature Selection</h2>
<p>In order to improve performance in terms of both speed and quality of results, we will start with defining a set of highly variable genes based on the biological coefficient of variability (see the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/scran/inst/doc/scran.html#5_variance_modelling">scran vignette</a> for details).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Calculating highly variable genes -------------------</span>
fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce, <span class="dt">parametric=</span><span class="ot">TRUE</span>, <span class="dt">use.spikes =</span> <span class="ot">FALSE</span>)
dec &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(sce, fit)
hvg_genes &lt;-<span class="st"> </span><span class="kw">rownames</span>(dec[dec<span class="op">$</span>FDR <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.00001</span>, ]) <span class="co"># 1874 genes</span>

<span class="co">## remove uninteresting/unknown genes (ribosomal, mitochondrial)</span>
bad_patterns &lt;-<span class="st"> &#39;RP[0-9]|^MT|[0-9][0-9][0-9][0-9]|^RPL|^RPS&#39;</span>
hvg_genes &lt;-<span class="st"> </span>hvg_genes[<span class="op">!</span><span class="kw">grepl</span>(bad_patterns, hvg_genes)] <span class="co"># 1713 genes</span>

<span class="co">## save the decomposed variance table and hvg_genes into metadata for safekeeping</span>
<span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes &lt;-<span class="st"> </span>hvg_genes
<span class="kw">metadata</span>(sce)<span class="op">$</span>dec_var &lt;-<span class="st"> </span>dec</code></pre>
</div>
<div id="dimension-reduction" class="section level2">
<h2><span class="header-section-number">7.6</span> Dimension Reduction</h2>
<p>Here we will use PCA to calculate the first 20 principal components with our highly variable geneset, using the <code>irlba</code> method for a faster approximate version of the calculation. These PCs will then form the basis of our subsequent tSNE calculation.</p>
<pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">method =</span> <span class="st">&quot;irlba&quot;</span>,
             <span class="dt">ncomponents =</span> <span class="dv">20</span>,
             <span class="dt">feature_set =</span> <span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes)
sce &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce,
              <span class="dt">perplexity =</span> <span class="dv">30</span>,
              <span class="dt">feature_set =</span> <span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes)
sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce,
              <span class="dt">n_neighbors =</span> <span class="dv">15</span>,
              <span class="dt">feature_set =</span> <span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes)</code></pre>
<p>Here, we see that PCA nicely separates most of the cell lines, with the exception of HCC827 and H1975. On the other hand, tSNE nicely manages to separate the 5 cell lines, but interestingly produces two separate clusters for cell line H1975.</p>
<p>We leave it as an exercise to the interested reader to modify the essential tuning parameters, <code>perplexity</code>, and <code>n_neighbors</code> for TSNE and UMAP respectively to see how the dimensionality reduction embeddings change. Here we have used the default parameters, but simply made them explicit. Information on these and other parameters can be seen by accessing the help for each function.</p>
<p>Note that clustering performance is <em>not</em> necessarily dependent on the dimensionality reduction results (to answer this, it will depend on the specific clustering technique at hand).</p>
<p>However, clustering performance indeed <em>may</em> be reflected in the dimensionality reduction results. For example, given the PCA result, we might expect some confusion between HCC827 and H1975, and based on the UMAP results, we might possibly even see two clusters within the H1975 cell line. Note however that these dimensionality reduction embeddings are also a function of the tuning parameters noted above.</p>
<pre class="sourceCode r"><code class="sourceCode r">gPCA &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="st">&quot;PCA&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;cell_line&quot;</span>)
gTSNE &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;cell_line&quot;</span>)
gUMAP &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="st">&quot;UMAP&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;cell_line&quot;</span>)
patchwork<span class="op">::</span><span class="kw">wrap_plots</span>(gPCA, gTSNE, gUMAP, <span class="dt">heights =</span> <span class="dv">1</span>, <span class="dt">widths =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">3</span>)</code></pre>
<p><img src="0Y.clustering-datasets_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="clustering" class="section level2">
<h2><span class="header-section-number">7.7</span> Clustering</h2>
<p>Here we highlight different frameworks for clustering. The first two, using the <code>sc3</code> and <code>clusterExperiment</code> Bioconductor packages, are fuller implementations that can test across multiple parametrizations and furthermore inspect the quality of the clustering results. The <code>BiocNeighbors</code> package is briefly highlighted to show a minimal alternative to clustering that emphasizes speed.</p>
<div id="sc3" class="section level3">
<h3><span class="header-section-number">7.7.1</span> SC3</h3>
<p>The <code>sc3</code> package provides a simple framework that allows users to test for many k’s, e.g. numbers of clusters, and subsequently compare the results from these differing k’s in both quantitative and qualitative ways to pick an optimal k result. Below, we first set an essential <code>rowData</code> feature required by the package to run. Subsequently, we run the <code>sc3()</code> function in this example using a subset of the data (only the highly variable genes, <code>hvg_genes</code>, testing k’s 3 through 6.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## SC3 requires this column to be appended</span>
<span class="kw">rowData</span>(sce)<span class="op">$</span>feature_symbol &lt;-<span class="st"> </span><span class="kw">rownames</span>(sce) 

<span class="co">## SC3 will return an SCE object with appended &quot;sc3_&quot; columns</span>
sce &lt;-<span class="st"> </span><span class="kw">sc3</span>(sce[<span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes, ],
          <span class="dt">ks =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">6</span>,
          <span class="dt">k_estimator =</span> <span class="ot">TRUE</span>)</code></pre>
<p>After using <code>sc3()</code>, the function returns the original <code>SingleCellExperiment</code> object, but with new columns in <code>colData(sce)</code> corresponding to the different <code>ks</code> supplied to the function, as well as a full representation of the analysis that is stored in <code>metadata(sce)$sc3</code>, which includes an estimate of the optimal <code>k</code> (as dictated by the <code>k_estimator = TRUE</code> argument above).</p>
<p>Below, we show the clustering results of the <code>ks</code> we supplied, 3 through 6, shown on the TSNE representation of the data. We use the <code>scater</code> package function <code>plotReducedDim()</code> to produce the individual plots, and for the sake of this vignette, wrap them together into a single plot using the <code>patchwork</code> package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Basic code:</span>
g3 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_3_clusters&quot;</span>)
g4 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_4_clusters&quot;</span>)
g5 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_5_clusters&quot;</span>)
g6 &lt;-<span class="st"> </span><span class="kw">plotReducedDim</span>(sce, <span class="dt">use_dimred =</span> <span class="st">&quot;TSNE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_6_clusters&quot;</span>)
patchwork<span class="op">::</span><span class="kw">wrap_plots</span>(g3, g4, g5, g6, <span class="dt">widths =</span> <span class="dv">1</span>, <span class="dt">heights =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">1</span>)</code></pre>
<p><img src="0Y.clustering-datasets_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>The <code>sc3</code> package contains many more utilities for exploring the stability of clustering and can even produce differential expression analysis results using the <code>biology = TRUE</code> argument within the <code>sc3()</code> function. We leave it to the interested reader to <a href="https://bioconductor.org/packages/release/bioc/html/SC3.html">learn more about <code>sc3</code></a> via their vignette.</p>
</div>
<div id="clusterexperiment" class="section level3">
<h3><span class="header-section-number">7.7.2</span> clusterExperiment</h3>
<p>The <code>clusterExperiment</code> package uses the function <code>RSEC()</code> to calculate clusters across various parametrizations, and contains multiple parameters to do so. Here, we specify the parametrizations to iterate over using the <code>alphas</code> and <code>k0s</code> arguments below. We refer the reader to the help page for <code>?RSEC</code> to learn more about the different parameters. In the end, this exhaustive exercise is used to ultimately determine a consensus clustering that combines the information learned from the various parametrizations.</p>
<pre class="sourceCode r"><code class="sourceCode r">rsec &lt;-<span class="st"> </span><span class="kw">RSEC</span>(sce[<span class="kw">metadata</span>(sce)<span class="op">$</span>hvg_genes, ],
            <span class="dt">reduceMethod =</span> <span class="st">&quot;PCA&quot;</span>,
            <span class="dt">nReducedDims =</span> <span class="dv">20</span>,
            <span class="dt">alphas =</span> <span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.3</span>), <span class="dt">k0s =</span> <span class="dv">4</span><span class="op">:</span><span class="dv">6</span>,
            <span class="dt">consensusMinSize =</span> <span class="dv">50</span>)

<span class="kw">metadata</span>(sce)<span class="op">$</span>rsec &lt;-<span class="st"> </span>rsec
<span class="kw">colData</span>(sce)<span class="op">$</span>rsec_consensus_clusters &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">clusterMatrix</span>(rsec)<span class="op">$</span>mergeClusters)</code></pre>
<p>Once again, we refer the interested reader to the <a href="https://bioconductor.org/packages/release/bioc/html/clusterExperiment.html">vignette for <code>clusterExperiment</code></a> to learn more about extended functionality for visualizing the data and about the specifics of the clustering workflow.</p>
</div>
<div id="biocneighbors" class="section level3">
<h3><span class="header-section-number">7.7.3</span> BiocNeighbors</h3>
<p>The <code>BiocNeighbors</code> package provides a lower-level method for clustering of data. Under the hood, it provides various types of clustering algorithms, as specified through the <code>BNPARAM</code> class object. This <code>BNPARAM</code> object can be passed into the <code>BiocNeighbors</code> clustering function, <code>findKNN()</code>, as well as other functions that support <code>BNPARAM</code>.</p>
<p>The <code>BiocNeighbors</code> packaged is especially designed for developers to enable a unified interface for clustering algorithm specification, allowing for the separation of clustering functionality from the main function.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="workflow-integrating-datasets.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["OSCA.pdf", "OSCA.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
