<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 5 Geneset Analysis and Gene Expression Networks | Orchestrating Single-Cell Analysis</title>
  <meta name="description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 5 Geneset Analysis and Gene Expression Networks | Orchestrating Single-Cell Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Geneset Analysis and Gene Expression Networks | Orchestrating Single-Cell Analysis" />
  
  <meta name="twitter:description" content="This is a companion of workflows to the ‘Orchestrating Single-Cell Analysis’ paper by the Bioconductor team." />
  

<meta name="author" content="Bioconductor">


<meta name="date" content="2019-01-07">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="trajectory-analysis-workflow.html">
<link rel="next" href="integrating-datasets.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Prerequisites</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html"><i class="fa fa-check"></i><b>3</b> Clustering and Differential Expression Workflow</a><ul>
<li class="chapter" data-level="3.1" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#introduction"><i class="fa fa-check"></i><b>3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#learning-objectives"><i class="fa fa-check"></i><b>3.2</b> Learning objectives</a></li>
<li class="chapter" data-level="3.3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#package-requirements"><i class="fa fa-check"></i><b>3.3</b> Package requirements</a></li>
<li class="chapter" data-level="3.4" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#loading-the-data"><i class="fa fa-check"></i><b>3.4</b> Loading the data</a></li>
<li class="chapter" data-level="3.5" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#preprocessing"><i class="fa fa-check"></i><b>3.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="3.5.1" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#cell-and-gene-quality-control"><i class="fa fa-check"></i><b>3.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="3.5.2" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#normalization"><i class="fa fa-check"></i><b>3.5.2</b> Normalization</a></li>
<li class="chapter" data-level="3.5.3" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#feature-selection"><i class="fa fa-check"></i><b>3.5.3</b> Feature selection</a></li>
<li class="chapter" data-level="3.5.4" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#dimensionality-reduction"><i class="fa fa-check"></i><b>3.5.4</b> Dimensionality reduction</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#clustering"><i class="fa fa-check"></i><b>3.6</b> Clustering</a></li>
<li class="chapter" data-level="3.7" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#differential-expression"><i class="fa fa-check"></i><b>3.7</b> Differential Expression</a></li>
<li class="chapter" data-level="3.8" data-path="clustering-and-differential-expression-workflow.html"><a href="clustering-and-differential-expression-workflow.html#session-info"><i class="fa fa-check"></i><b>3.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html"><i class="fa fa-check"></i><b>4</b> Trajectory Analysis Workflow</a><ul>
<li class="chapter" data-level="4.1" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#learning-objectives-1"><i class="fa fa-check"></i><b>4.2</b> Learning objectives</a></li>
<li class="chapter" data-level="4.3" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#package-requirements-1"><i class="fa fa-check"></i><b>4.3</b> Package requirements</a></li>
<li class="chapter" data-level="4.4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#loading-the-data-1"><i class="fa fa-check"></i><b>4.4</b> Loading the data</a></li>
<li class="chapter" data-level="4.5" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#preprocessing-1"><i class="fa fa-check"></i><b>4.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="4.5.1" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#cell-and-gene-quality-control-1"><i class="fa fa-check"></i><b>4.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="4.5.2" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#normalization-1"><i class="fa fa-check"></i><b>4.5.2</b> Normalization</a></li>
<li class="chapter" data-level="4.5.3" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>4.5.3</b> Feature selection</a></li>
<li class="chapter" data-level="4.5.4" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>4.5.4</b> Dimensionality reduction</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#trajectory-analysis"><i class="fa fa-check"></i><b>4.6</b> Trajectory Analysis</a></li>
<li class="chapter" data-level="4.7" data-path="trajectory-analysis-workflow.html"><a href="trajectory-analysis-workflow.html#session-info-1"><i class="fa fa-check"></i><b>4.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html"><i class="fa fa-check"></i><b>5</b> Geneset Analysis and Gene Expression Networks</a><ul>
<li class="chapter" data-level="5.1" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#learning-objectives-2"><i class="fa fa-check"></i><b>5.2</b> Learning objectives</a></li>
<li class="chapter" data-level="5.3" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#package-requirements-2"><i class="fa fa-check"></i><b>5.3</b> Package requirements</a></li>
<li class="chapter" data-level="5.4" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#loading-the-data-2"><i class="fa fa-check"></i><b>5.4</b> Loading the data</a></li>
<li class="chapter" data-level="5.5" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#preprocessing-2"><i class="fa fa-check"></i><b>5.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="5.5.1" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#cell-and-gene-quality-control-2"><i class="fa fa-check"></i><b>5.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="5.5.2" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#normalization-2"><i class="fa fa-check"></i><b>5.5.2</b> Normalization</a></li>
<li class="chapter" data-level="5.5.3" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#feature-selection-2"><i class="fa fa-check"></i><b>5.5.3</b> Feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#geneset-analysis"><i class="fa fa-check"></i><b>5.6</b> Geneset Analysis</a></li>
<li class="chapter" data-level="5.7" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#gene-expression-networks"><i class="fa fa-check"></i><b>5.7</b> Gene Expression Networks</a></li>
<li class="chapter" data-level="5.8" data-path="geneset-analysis-and-gene-expression-networks.html"><a href="geneset-analysis-and-gene-expression-networks.html#session-info-2"><i class="fa fa-check"></i><b>5.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="integrating-datasets.html"><a href="integrating-datasets.html"><i class="fa fa-check"></i><b>6</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="6.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#introduction-3"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#learning-objectives-3"><i class="fa fa-check"></i><b>6.2</b> Learning objectives</a></li>
<li class="chapter" data-level="6.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#package-requirements-3"><i class="fa fa-check"></i><b>6.3</b> Package requirements</a></li>
<li class="chapter" data-level="6.4" data-path="integrating-datasets.html"><a href="integrating-datasets.html#loading-the-data-3"><i class="fa fa-check"></i><b>6.4</b> Loading the data</a></li>
<li class="chapter" data-level="6.5" data-path="integrating-datasets.html"><a href="integrating-datasets.html#preprocessing-3"><i class="fa fa-check"></i><b>6.5</b> Preprocessing</a><ul>
<li class="chapter" data-level="6.5.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#cell-and-gene-quality-control-3"><i class="fa fa-check"></i><b>6.5.1</b> Cell and gene quality control</a></li>
<li class="chapter" data-level="6.5.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#normalization-3"><i class="fa fa-check"></i><b>6.5.2</b> Normalization</a></li>
<li class="chapter" data-level="6.5.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#feature-selection-3"><i class="fa fa-check"></i><b>6.5.3</b> Feature selection</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="integrating-datasets.html"><a href="integrating-datasets.html#integrating-datasets-1"><i class="fa fa-check"></i><b>6.6</b> Integrating Datasets</a></li>
<li class="chapter" data-level="6.7" data-path="integrating-datasets.html"><a href="integrating-datasets.html#session-info-3"><i class="fa fa-check"></i><b>6.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="geneset-analysis-and-gene-expression-networks" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Geneset Analysis and Gene Expression Networks</h1>
<p><em>Author</em>: Author1, affil, Author2, affil</p>
<div id="introduction-2" class="section level2">
<h2><span class="header-section-number">5.1</span> Introduction</h2>
<p>Introduce the main purpose of this vignette in paragraph form, why these steps are important, what the end products allow for.</p>
</div>
<div id="learning-objectives-2" class="section level2">
<h2><span class="header-section-number">5.2</span> Learning objectives</h2>
<ul>
<li>In bullet form, illustrate what the main outputs are of this vignette</li>
<li>A couple of bullet points are fine</li>
</ul>
</div>
<div id="package-requirements-2" class="section level2">
<h2><span class="header-section-number">5.3</span> Package requirements</h2>
<p>Packages that are required to go through the vignette and how to install them, should largely be through BiocManager interface.</p>
<pre class="sourceCode r"><code class="sourceCode r">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&#39;SC3&#39;</span>, <span class="st">&#39;clusterExperiment&#39;</span>, <span class="st">&#39;BiocParallel&#39;</span>, <span class="st">&#39;BEARscc&#39;</span>, <span class="st">&#39;zinbwave&#39;</span>, <span class="st">&#39;edgeR&#39;</span>, <span class="st">&#39;DESeq2&#39;</span>, <span class="st">&#39;MAST&#39;</span>, <span class="st">&#39;scDD&#39;</span>))</code></pre>
</div>
<div id="loading-the-data-2" class="section level2">
<h2><span class="header-section-number">5.4</span> Loading the data</h2>
<p>Here the data should be loaded if in package form and briefly explored to illustrate the contents of the data.</p>
<p>If data comes from “raw” form, then a simple pipeline for creating the processed data should be illustrated as succinctly as possible.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TENxPBMCData)</code></pre>
<pre><code>## Loading required package: SingleCellExperiment</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, Map, mapply, match, mget, order, paste,
##     pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,
##     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,
##     tapply, union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## Loading required package: BiocParallel</code></pre>
<pre><code>## 
## Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     aperm, apply</code></pre>
<pre><code>## Loading required package: HDF5Array</code></pre>
<pre><code>## Loading required package: rhdf5</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)</code></pre>
<pre><code>## snapshotDate(): 2019-01-04</code></pre>
<pre><code>## see ?TENxPBMCData and browseVignettes(&#39;TENxPBMCData&#39;) for documentation</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/ramezqui//.ExperimentHub/1613&#39;</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 4340 
## metadata(0):
## assays(1): counts
## rownames(33694): ENSG00000243485 ENSG00000237613 ...
##   ENSG00000277475 ENSG00000268674
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames: NULL
## colData names(11): Sample Barcode ... Individual Date_published
## reducedDimNames(0):
## spikeNames(0):</code></pre>
</div>
<div id="preprocessing-2" class="section level2">
<h2><span class="header-section-number">5.5</span> Preprocessing</h2>
<p>All steps required to preprocess the data into the clean expression matrix, split it up into subsections as necessary. Subsection splits should be informed by the paper (sub)sections. This is all steps prior to the main strategy of interest.</p>
<div id="cell-and-gene-quality-control-2" class="section level3">
<h3><span class="header-section-number">5.5.1</span> Cell and gene quality control</h3>
<p>Removal of “bad cells”, low abundance genes.</p>
</div>
<div id="normalization-2" class="section level3">
<h3><span class="header-section-number">5.5.2</span> Normalization</h3>
<p>Application of normalization approach, even if its just a log normalization on the count data. Cell cycle normalization may be applicable here.</p>
</div>
<div id="feature-selection-2" class="section level3">
<h3><span class="header-section-number">5.5.3</span> Feature selection</h3>
<p>Identifying the subset of the clean matrix to work with.</p>
</div>
</div>
<div id="geneset-analysis" class="section level2">
<h2><span class="header-section-number">5.6</span> Geneset Analysis</h2>
<p>Strategy 1 of interest. Use BiocParallel. May need to save intermediate data within package to facilitate fast construction of Rmd.</p>
</div>
<div id="gene-expression-networks" class="section level2">
<h2><span class="header-section-number">5.7</span> Gene Expression Networks</h2>
<p>Strategy 2 of interest. Differential expression approaches. Use one of the outputs from above to perform differential expression analyses comparison.</p>
</div>
<div id="session-info-2" class="section level2">
<h2><span class="header-section-number">5.8</span> Session Info</h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="trajectory-analysis-workflow.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="integrating-datasets.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["orchestratingSingleCellAnalysis.pdf", "orchestratingSingleCellAnalysis.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
